{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# DIRAC Comparisons between LC M001 Liver Proteomics and Related Transcriptomics in GOBP Modules"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***by Kengo Watanabe***  \n",
    "\n",
    "The differential rank conservation (DIRAC; Eddy, J.A. et al. PLoS Comput. Biol. 2010) analyses were performed on (1) the preprocessed Longevity Consortium (LC) M001 proteomics dataset (adjusted with sex and age; analytes detected in all samples; sample-based robust Z-score followed by analyte-based robust Z-score) using the retrieved a priori module set (Gene Ontology (Biological Process) (GOBP) derived by EMBL-EBI QuickGO API; ≥4 analytes and ≥50% coverage) and (2) the preprocessed LC M001-related transcriptomics dataset (Tyshkovskiy, A. et al. Cell Metab. 2019; adjusted with sex and age; analytes detected in all samples; sample-based robust Z-score followed by analyte-based robust Z-score) using the retrieved a priori module set (GOBP derived by R org.Mm.eg.db package; ≥4 analytes and ≥50% coverage).  \n",
    "–> This Jupyter Notebook (with Python 3 kernel) compared the DIRAC results between the LC M001 proteomics and M001-related transcriptomics data.  \n",
    "\n",
    "Input files:  \n",
    "- Module metadata (proteomics): 230214_LC-M001-proteomics-DIRAC-ver7-2_DIRAC-GOBP_module-metadata.tsv  \n",
    "- Module metadata (transcriptomics): 230217_LC-M001-related-TrOmics-DIRAC-ver3_DIRAC-GOBP_onWenc_module-metadata.tsv  \n",
    "- Sample–mouse metadata (proteomics): 230213_LC-M001-proteomics-DIRAC-ver7-2_Preprocessing_sample-metadata.tsv  \n",
    "–> Preprocessed analyte data (proteomics): 230213_LC-M001-proteomics-DIRAC-ver7-2_Preprocessing_normalized-data.tsv  \n",
    "- Sample–mouse metadata (transcriptomics): 230215_LC-M001-related-TrOmics-DIRAC-ver3_Preprocessing_onWenc_sample-metadata.tsv  \n",
    "–> Preprocessed analyte data (transcriptomics): 230215_LC-M001-related-TrOmics-DIRAC-ver3_Preprocessing_onWenc_normalized-data.tsv  \n",
    "- DIRAC RMS data (proteomics): 230214_LC-M001-proteomics-DIRAC-ver7-2_DIRAC-GOBP_RankMatchingScore-BS-combined.tsv  \n",
    "- DIRAC RCI data (proteomics): 230214_LC-M001-proteomics-DIRAC-ver7-2_DIRAC-GOBP_RankConservationIndex-BS-combined.tsv  \n",
    "- DIRAC RMS data (transcriptomics): 230217_LC-M001-related-TrOmics-DIRAC-ver3_DIRAC-GOBP_onWenc_RankMatchingScore-BS-combined.tsv  \n",
    "- DIRAC RCI data (transcriptomics): 230217_LC-M001-related-TrOmics-DIRAC-ver3_DIRAC-GOBP_onWenc_RankConservationIndex-BS-combined.tsv  \n",
    "\n",
    "Output figures and tables:  \n",
    "- Figure 4b–d, 5d, 5g  \n",
    "- Supplementary Data 6  \n",
    "\n",
    "Original notebook (memo for my future tracing):  \n",
    "- dalek:\\[JupyterLab HOME\\]/230224_LC-M001-PrOmics-vs-TrOmics-DIRAC-ver3/230224_LC-M001-PrOmics-vs-TrOmics-DIRAC-ver3_GOBP.ipynb  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# packages in environment at /opt/conda/envs/arivale-py3:\n",
      "#\n",
      "# Name                    Version                   Build  Channel\n",
      "_libgcc_mutex             0.1                 conda_forge    conda-forge\n",
      "_openmp_mutex             4.5                       1_gnu    conda-forge\n",
      "analytics                 0.1                      pypi_0    pypi\n",
      "argon2-cffi               21.1.0           py39h3811e60_0    conda-forge\n",
      "arivale-data-interface    0.1.0                    pypi_0    pypi\n",
      "async_generator           1.10                       py_0    conda-forge\n",
      "atk-1.0                   2.36.0               h3371d22_4    conda-forge\n",
      "attrs                     21.2.0             pyhd8ed1ab_0    conda-forge\n",
      "backcall                  0.2.0              pyh9f0ad1d_0    conda-forge\n",
      "backports                 1.0                        py_2    conda-forge\n",
      "backports.functools_lru_cache 1.6.4              pyhd8ed1ab_0    conda-forge\n",
      "biopython                 1.79             py39h3811e60_0    conda-forge\n",
      "bleach                    4.1.0              pyhd8ed1ab_0    conda-forge\n",
      "bokeh                     2.4.1            py39hf3d152e_1    conda-forge\n",
      "boost-cpp                 1.74.0               h312852a_4    conda-forge\n",
      "boto                      2.49.0                     py_0    conda-forge\n",
      "boto3                     1.19.2                   pypi_0    pypi\n",
      "botocore                  1.22.2                   pypi_0    pypi\n",
      "brotlipy                  0.7.0           py39h3811e60_1001    conda-forge\n",
      "bzip2                     1.0.8                h7f98852_4    conda-forge\n",
      "c-ares                    1.17.2               h7f98852_0    conda-forge\n",
      "ca-certificates           2022.12.7            ha878542_0    conda-forge\n",
      "cachecontrol              0.12.6                     py_0    conda-forge\n",
      "cairo                     1.16.0            h6cf1ce9_1008    conda-forge\n",
      "certifi                   2022.12.7          pyhd8ed1ab_0    conda-forge\n",
      "cffi                      1.14.6           py39h4bc2ebd_1    conda-forge\n",
      "chardet                   4.0.0            py39hf3d152e_1    conda-forge\n",
      "charset-normalizer        2.0.0              pyhd8ed1ab_0    conda-forge\n",
      "click                     8.0.3            py39hf3d152e_0    conda-forge\n",
      "colorama                  0.4.4              pyh9f0ad1d_0    conda-forge\n",
      "cryptography              35.0.0           py39h95dcef6_1    conda-forge\n",
      "cycler                    0.10.0                     py_2    conda-forge\n",
      "cython                    0.29.24          py39he80948d_0    conda-forge\n",
      "debugpy                   1.4.1            py39he80948d_0    conda-forge\n",
      "decorator                 4.4.2                      py_0    conda-forge\n",
      "defusedxml                0.7.1              pyhd8ed1ab_0    conda-forge\n",
      "english                   2020.7.0           pyhd8ed1ab_0    conda-forge\n",
      "entrypoints               0.3             py39hde42818_1002    conda-forge\n",
      "et_xmlfile                1.0.1                   py_1001    conda-forge\n",
      "expat                     2.4.1                h9c3ff4c_0    conda-forge\n",
      "font-ttf-dejavu-sans-mono 2.37                 hab24e00_0    conda-forge\n",
      "font-ttf-inconsolata      3.000                h77eed37_0    conda-forge\n",
      "font-ttf-source-code-pro  2.038                h77eed37_0    conda-forge\n",
      "font-ttf-ubuntu           0.83                 hab24e00_0    conda-forge\n",
      "fontconfig                2.13.1            hba837de_1005    conda-forge\n",
      "fonts-conda-ecosystem     1                             0    conda-forge\n",
      "fonts-conda-forge         1                             0    conda-forge\n",
      "freetype                  2.10.4               h0708190_1    conda-forge\n",
      "fribidi                   1.0.10               h36c2ea0_0    conda-forge\n",
      "future                    0.18.2           py39hf3d152e_3    conda-forge\n",
      "gdk-pixbuf                2.42.6               h04a7f16_0    conda-forge\n",
      "gettext                   0.19.8.1          h73d1719_1008    conda-forge\n",
      "giflib                    5.2.1                h36c2ea0_2    conda-forge\n",
      "goatools                  1.2.3              pyh7cba7a3_2    bioconda\n",
      "graphite2                 1.3.13            h58526e2_1001    conda-forge\n",
      "graphviz                  2.49.1               h85b4f2f_0    conda-forge\n",
      "gtk2                      2.24.33              h539f30e_1    conda-forge\n",
      "gts                       0.7.6                h64030ff_2    conda-forge\n",
      "harfbuzz                  3.0.0                h83ec7ef_1    conda-forge\n",
      "hdmedians                 0.14.2           py39hce5d2b2_0    conda-forge\n",
      "icu                       68.1                 h58526e2_0    conda-forge\n",
      "idna                      3.1                pyhd3deb0d_0    conda-forge\n",
      "importlib-metadata        4.8.1            py39hf3d152e_0    conda-forge\n",
      "iniconfig                 1.1.1              pyh9f0ad1d_0    conda-forge\n",
      "interchange               2021.0.4           pyhd8ed1ab_0    conda-forge\n",
      "ipykernel                 6.4.2            py39hef51801_0    conda-forge\n",
      "ipython                   7.28.0           py39hef51801_0    conda-forge\n",
      "ipython_genutils          0.2.0                      py_1    conda-forge\n",
      "ipywidgets                7.6.5              pyhd8ed1ab_0    conda-forge\n",
      "jbig                      2.1               h7f98852_2003    conda-forge\n",
      "jedi                      0.18.0           py39hf3d152e_2    conda-forge\n",
      "jinja2                    3.0.2              pyhd8ed1ab_0    conda-forge\n",
      "jmespath                  0.10.0                   pypi_0    pypi\n",
      "joblib                    1.1.0              pyhd8ed1ab_0    conda-forge\n",
      "jpeg                      9d                   h36c2ea0_0    conda-forge\n",
      "jsonschema                4.1.2              pyhd8ed1ab_0    conda-forge\n",
      "jupyter_client            7.0.6              pyhd8ed1ab_0    conda-forge\n",
      "jupyter_core              4.8.1            py39hf3d152e_0    conda-forge\n",
      "jupyterlab_pygments       0.1.2              pyh9f0ad1d_0    conda-forge\n",
      "jupyterlab_widgets        1.0.2              pyhd8ed1ab_0    conda-forge\n",
      "kiwisolver                1.3.2            py39h1a9c180_0    conda-forge\n",
      "krb5                      1.19.2               hcc1bbae_2    conda-forge\n",
      "lcms2                     2.12                 hddcbb42_0    conda-forge\n",
      "ld_impl_linux-64          2.36.1               hea4e1c9_2    conda-forge\n",
      "lerc                      2.2.1                h9c3ff4c_0    conda-forge\n",
      "libblas                   3.9.0           12_linux64_openblas    conda-forge\n",
      "libcblas                  3.9.0           12_linux64_openblas    conda-forge\n",
      "libcurl                   7.79.1               h2574ce0_1    conda-forge\n",
      "libdeflate                1.7                  h7f98852_5    conda-forge\n",
      "libedit                   3.1.20191231         he28a2e2_2    conda-forge\n",
      "libev                     4.33                 h516909a_1    conda-forge\n",
      "libffi                    3.4.2                h9c3ff4c_4    conda-forge\n",
      "libgcc-ng                 11.2.0              h1d223b6_11    conda-forge\n",
      "libgd                     2.3.3                h6ad9fb6_0    conda-forge\n",
      "libgfortran-ng            11.2.0              h69a702a_11    conda-forge\n",
      "libgfortran5              11.2.0              h5c6108e_11    conda-forge\n",
      "libglib                   2.70.0               h174f98d_1    conda-forge\n",
      "libgomp                   11.2.0              h1d223b6_11    conda-forge\n",
      "libiconv                  1.16                 h516909a_0    conda-forge\n",
      "liblapack                 3.9.0           12_linux64_openblas    conda-forge\n",
      "libnghttp2                1.43.0               h812cca2_1    conda-forge\n",
      "libopenblas               0.3.18          pthreads_h8fe5266_0    conda-forge\n",
      "libpng                    1.6.37               h21135ba_2    conda-forge\n",
      "libpq                     13.3                 hd57d9b9_1    conda-forge\n",
      "librsvg                   2.52.2               hc3c00ef_0    conda-forge\n",
      "libsodium                 1.0.18               h36c2ea0_1    conda-forge\n",
      "libssh2                   1.10.0               ha56f1ee_2    conda-forge\n",
      "libstdcxx-ng              11.2.0              he4da1e4_11    conda-forge\n",
      "libtiff                   4.3.0                hf544144_1    conda-forge\n",
      "libtool                   2.4.6             h9c3ff4c_1008    conda-forge\n",
      "libuuid                   2.32.1            h7f98852_1000    conda-forge\n",
      "libwebp                   1.2.1                h3452ae3_0    conda-forge\n",
      "libwebp-base              1.2.1                h7f98852_0    conda-forge\n",
      "libxcb                    1.13              h7f98852_1003    conda-forge\n",
      "libxml2                   2.9.12               h72842e0_0    conda-forge\n",
      "libxslt                   1.1.33               h15afd5d_2    conda-forge\n",
      "libzlib                   1.2.11            h36c2ea0_1013    conda-forge\n",
      "lockfile                  0.12.2                     py_1    conda-forge\n",
      "lxml                      4.6.3            py39h107f48f_0    conda-forge\n",
      "lz4-c                     1.9.3                h9c3ff4c_1    conda-forge\n",
      "markupsafe                2.0.1            py39h3811e60_0    conda-forge\n",
      "matplotlib-base           3.4.3            py39h2fa2bec_1    conda-forge\n",
      "matplotlib-inline         0.1.3              pyhd8ed1ab_0    conda-forge\n",
      "matplotlib-venn           0.11.7             pyhd8ed1ab_0    conda-forge\n",
      "missingpy                 0.2.0                    pypi_0    pypi\n",
      "mistune                   0.8.4           py39h3811e60_1004    conda-forge\n",
      "more-itertools            8.10.0             pyhd8ed1ab_0    conda-forge\n",
      "mscorefonts               0.0.1                         3    conda-forge\n",
      "msgpack-python            1.0.2            py39h1a9c180_1    conda-forge\n",
      "natsort                   7.1.1              pyhd8ed1ab_0    conda-forge\n",
      "nbclient                  0.5.4              pyhd8ed1ab_0    conda-forge\n",
      "nbconvert                 6.2.0            py39hf3d152e_0    conda-forge\n",
      "nbformat                  5.1.3              pyhd8ed1ab_0    conda-forge\n",
      "ncurses                   6.2                  h58526e2_4    conda-forge\n",
      "neo4j-python-driver       4.3.7              pyhd8ed1ab_0    conda-forge\n",
      "neobolt                   1.7.17           py39h3811e60_2    conda-forge\n",
      "neotime                   1.7.4                      py_0    conda-forge\n",
      "nest-asyncio              1.5.1              pyhd8ed1ab_0    conda-forge\n",
      "networkx                  2.6.3              pyhd8ed1ab_0    conda-forge\n",
      "notebook                  6.4.5              pyha770c72_0    conda-forge\n",
      "numpy                     1.21.3           py39hdbf815f_0    conda-forge\n",
      "olefile                   0.46               pyh9f0ad1d_1    conda-forge\n",
      "openjpeg                  2.4.0                hb52868f_1    conda-forge\n",
      "openpyxl                  3.0.9              pyhd8ed1ab_0    conda-forge\n",
      "openssl                   1.1.1o               h166bdaf_0    conda-forge\n",
      "packaging                 21.0               pyhd8ed1ab_0    conda-forge\n",
      "pandas                    1.3.4            py39hde0f152_0    conda-forge\n",
      "pandoc                    2.15                 h7f98852_0    conda-forge\n",
      "pandocfilters             1.5.0              pyhd8ed1ab_0    conda-forge\n",
      "pango                     1.48.10              h54213e6_2    conda-forge\n",
      "parso                     0.8.2              pyhd8ed1ab_0    conda-forge\n",
      "patsy                     0.5.2              pyhd8ed1ab_0    conda-forge\n",
      "pcre                      8.45                 h9c3ff4c_0    conda-forge\n",
      "pdf2image                 1.16.2                   pypi_0    pypi\n",
      "pexpect                   4.8.0              pyh9f0ad1d_2    conda-forge\n",
      "pickleshare               0.7.5           py39hde42818_1002    conda-forge\n",
      "pillow                    8.3.2            py39ha612740_0    conda-forge\n",
      "pip                       21.3.1             pyhd8ed1ab_0    conda-forge\n",
      "pixman                    0.40.0               h36c2ea0_0    conda-forge\n",
      "plotly                    5.3.1              pyhd8ed1ab_0    conda-forge\n",
      "pluggy                    1.0.0            py39hf3d152e_1    conda-forge\n",
      "poppler                   21.09.0              h0e1ea10_2    conda-forge\n",
      "poppler-data              0.4.12               hd8ed1ab_0    conda-forge\n",
      "prometheus_client         0.11.0             pyhd8ed1ab_0    conda-forge\n",
      "prompt-toolkit            3.0.21             pyha770c72_0    conda-forge\n",
      "prompt_toolkit            3.0.21               hd8ed1ab_0    conda-forge\n",
      "psycopg2                  2.9.1            py39h3811e60_0    conda-forge\n",
      "pthread-stubs             0.4               h36c2ea0_1001    conda-forge\n",
      "ptyprocess                0.7.0              pyhd3deb0d_0    conda-forge\n",
      "py                        1.10.0             pyhd3deb0d_0    conda-forge\n",
      "py2neo                    2021.2.3           pyhd8ed1ab_0    conda-forge\n",
      "pycparser                 2.20               pyh9f0ad1d_2    conda-forge\n",
      "pydot                     1.4.2            py39hf3d152e_3    conda-forge\n",
      "pygments                  2.10.0             pyhd8ed1ab_0    conda-forge\n",
      "pygraphviz                1.7              py39h78163bd_0    conda-forge\n",
      "pyopenssl                 21.0.0             pyhd8ed1ab_0    conda-forge\n",
      "pyparsing                 2.4.7              pyh9f0ad1d_0    conda-forge\n",
      "pyrsistent                0.17.3           py39h3811e60_2    conda-forge\n",
      "pysam                     0.17.0           py39h051187c_0    bioconda\n",
      "pysocks                   1.7.1            py39hf3d152e_3    conda-forge\n",
      "pytabix                   0.1              py39h98c8e45_1    bioconda\n",
      "pytest                    6.2.5            py39hf3d152e_0    conda-forge\n",
      "python                    3.9.7           hb7a2778_3_cpython    conda-forge\n",
      "python-dateutil           2.8.2              pyhd8ed1ab_0    conda-forge\n",
      "python-graphviz           0.17               pyhaef67bd_0    conda-forge\n",
      "python-wget               3.2                        py_0    conda-forge\n",
      "python_abi                3.9                      2_cp39    conda-forge\n",
      "pytz                      2021.3             pyhd8ed1ab_0    conda-forge\n",
      "pyvcf                     0.6.8           py39hde42818_1002    conda-forge\n",
      "pyyaml                    6.0              py39h3811e60_0    conda-forge\n",
      "pyzmq                     22.3.0           py39h37b5a0c_0    conda-forge\n",
      "readline                  8.1                  h46c0cb4_0    conda-forge\n",
      "requests                  2.26.0             pyhd8ed1ab_0    conda-forge\n",
      "s3transfer                0.5.0                    pypi_0    pypi\n",
      "scikit-bio                0.5.6            py39h16ac069_4    conda-forge\n",
      "scikit-learn              1.0.1            py39h7c5d8c9_0    conda-forge\n",
      "scipy                     1.7.1            py39hee8e79c_0    conda-forge\n",
      "seaborn                   0.11.2               hd8ed1ab_0    conda-forge\n",
      "seaborn-base              0.11.2             pyhd8ed1ab_0    conda-forge\n",
      "send2trash                1.8.0              pyhd8ed1ab_0    conda-forge\n",
      "setuptools                58.2.0           py39hf3d152e_0    conda-forge\n",
      "six                       1.16.0             pyh6c4a22f_0    conda-forge\n",
      "sqlite                    3.36.0               h9cd32fc_2    conda-forge\n",
      "statsmodels               0.13.0           py39hce5d2b2_0    conda-forge\n",
      "tenacity                  8.0.1              pyhd8ed1ab_0    conda-forge\n",
      "terminado                 0.12.1           py39hf3d152e_0    conda-forge\n",
      "testpath                  0.5.0              pyhd8ed1ab_0    conda-forge\n",
      "threadpoolctl             3.0.0              pyh8a188c0_0    conda-forge\n",
      "tk                        8.6.11               h27826a3_1    conda-forge\n",
      "toml                      0.10.2             pyhd8ed1ab_0    conda-forge\n",
      "tornado                   6.1              py39h3811e60_1    conda-forge\n",
      "traitlets                 5.1.0              pyhd8ed1ab_0    conda-forge\n",
      "typing_extensions         3.10.0.2           pyha770c72_0    conda-forge\n",
      "tzdata                    2021e                he74cb21_0    conda-forge\n",
      "urllib3                   1.26.7             pyhd8ed1ab_0    conda-forge\n",
      "venn                      0.1.3                    pypi_0    pypi\n",
      "wcwidth                   0.2.5              pyh9f0ad1d_2    conda-forge\n",
      "webencodings              0.5.1                      py_1    conda-forge\n",
      "wheel                     0.37.0             pyhd8ed1ab_1    conda-forge\n",
      "widgetsnbextension        3.5.1            py39hf3d152e_4    conda-forge\n",
      "wordcloud                 1.8.1                    pypi_0    pypi\n",
      "xlrd                      2.0.1              pyhd8ed1ab_3    conda-forge\n",
      "xlsxwriter                3.0.3              pyhd8ed1ab_0    conda-forge\n",
      "xorg-kbproto              1.0.7             h7f98852_1002    conda-forge\n",
      "xorg-libice               1.0.10               h7f98852_0    conda-forge\n",
      "xorg-libsm                1.2.3             hd9c2040_1000    conda-forge\n",
      "xorg-libx11               1.7.2                h7f98852_0    conda-forge\n",
      "xorg-libxau               1.0.9                h7f98852_0    conda-forge\n",
      "xorg-libxdmcp             1.1.3                h7f98852_0    conda-forge\n",
      "xorg-libxext              1.3.4                h7f98852_1    conda-forge\n",
      "xorg-libxrender           0.9.10            h7f98852_1003    conda-forge\n",
      "xorg-renderproto          0.11.1            h7f98852_1002    conda-forge\n",
      "xorg-xextproto            7.3.0             h7f98852_1002    conda-forge\n",
      "xorg-xproto               7.0.31            h7f98852_1007    conda-forge\n",
      "xz                        5.2.5                h516909a_1    conda-forge\n",
      "yaml                      0.2.5                h516909a_0    conda-forge\n",
      "zeromq                    4.3.4                h9c3ff4c_1    conda-forge\n",
      "zipp                      3.6.0              pyhd8ed1ab_0    conda-forge\n",
      "zlib                      1.2.11            h36c2ea0_1013    conda-forge\n",
      "zstd                      1.5.0                ha95c52a_0    conda-forge\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scipy.stats as stats\n",
    "import matplotlib.pyplot as plt\n",
    "%matplotlib inline\n",
    "import seaborn as sns\n",
    "#For Arial font\n",
    "#!conda install -c conda-forge -y mscorefonts\n",
    "##-> The below was also needed in matplotlib 3.4.2\n",
    "#import shutil\n",
    "#import matplotlib\n",
    "#shutil.rmtree(matplotlib.get_cachedir())\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')\n",
    "from IPython.display import display\n",
    "import time\n",
    "#For exporting .pdf file with editable text\n",
    "import matplotlib\n",
    "matplotlib.rcParams['pdf.fonttype']=42\n",
    "matplotlib.rcParams['ps.fonttype']=42\n",
    "\n",
    "import statsmodels.formula.api as smf\n",
    "from statsmodels.stats.anova import anova_lm\n",
    "from statsmodels.stats import weightstats\n",
    "from statsmodels.stats import multitest as multi\n",
    "from decimal import Decimal, ROUND_HALF_UP\n",
    "import re\n",
    "import matplotlib.patches as mpatches\n",
    "from matplotlib.offsetbox import AnchoredText\n",
    "#!pip install venn\n",
    "from venn import venn\n",
    "from textwrap import wrap\n",
    "\n",
    "!conda list"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. Prepare metadata and DIRAC results"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> The transcriptomics-related files were copied from wenceslaus server in advance.  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1-1. Module–analyte metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Import the cleaned module metadata for proteomics\n",
    "fileDir = '../230206_LC-M001-proteomics-DIRAC-ver7/ExportData/'\n",
    "ipynbName = '230214_LC-M001-proteomics-DIRAC-ver7-2_DIRAC-GOBP_'\n",
    "fileName = 'module-metadata.tsv'\n",
    "tempDF1 = pd.read_csv(fileDir+ipynbName+fileName, sep='\\t').set_index('ModuleID')\n",
    "print('Proteomics modules:', len(tempDF1))\n",
    "\n",
    "#Import the cleaned module metadata for tran\n",
    "fileDir = './ImportData/'\n",
    "ipynbName = '230217_LC-M001-related-TrOmics-DIRAC-ver3_DIRAC-GOBP_onWenc_'\n",
    "fileName = 'module-metadata.tsv'\n",
    "tempDF2 = pd.read_csv(fileDir+ipynbName+fileName, sep='\\t').set_index('ModuleID')\n",
    "print('Transcriptomics modules:', len(tempDF2))\n",
    "\n",
    "#Retrieve common modules\n",
    "tempL = ['ModuleName', 'ModuleType']\n",
    "tempDF = tempDF1.loc[tempDF1.index.isin(tempDF2.index.tolist()), tempL]\n",
    "##Clean\n",
    "tempDF1 = tempDF1.drop(columns=tempL)\n",
    "tempDF1.columns = 'PrOmics_'+tempDF1.columns\n",
    "tempDF = pd.merge(tempDF, tempDF1, left_index=True, right_index=True, how='left')\n",
    "tempDF2 = tempDF2.drop(columns=tempL)\n",
    "tempDF2.columns = 'TrOmics_'+tempDF2.columns\n",
    "tempDF = pd.merge(tempDF, tempDF2, left_index=True, right_index=True, how='left')\n",
    "print('Common modules:', len(tempDF))\n",
    "display(tempDF)\n",
    "\n",
    "#Save\n",
    "fileDir = './ExportData/'\n",
    "ipynbName = '230224_LC-M001-PrOmics-vs-TrOmics-DIRAC-ver3_GOBP_'\n",
    "fileName = 'module-metadata.tsv'\n",
    "tempDF.to_csv(fileDir+ipynbName+fileName, index=True, sep='\\t')\n",
    "\n",
    "moduleDF = tempDF"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1-2. Sample–mouse metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tempL1 = ['Ctrl', 'Aca', 'Rapa']#Target groups to be comparaed\n",
    "tempL2 = ['Dataset', 'Intervention', 'Sex', 'Age', 'Phenotype']\n",
    "\n",
    "#Prepare sample-mouse metadata for proteomics\n",
    "fileDir = '../230206_LC-M001-proteomics-DIRAC-ver7/ExportData/'\n",
    "ipynbName = '230213_LC-M001-proteomics-DIRAC-ver7-2_Preprocessing_'\n",
    "fileName = 'sample-metadata.tsv'\n",
    "tempDF1 = pd.read_csv(fileDir+ipynbName+fileName, sep='\\t').set_index('SampleID')\n",
    "##Clean\n",
    "tempDF1['Phenotype'] = tempDF1['Intervention']#Sex-and-age-pooled for DIRAC template (rank consensus)\n",
    "tempDF1['Dataset'] = 'PrOmics'\n",
    "tempDF1['Age'] = '12M'#Forced conversion to month-level\n",
    "tempDF1 = tempDF1[tempL2]\n",
    "##Select the assessed samples\n",
    "fileName = 'normalized-data.tsv'\n",
    "tempDF = pd.read_csv(fileDir+ipynbName+fileName, sep='\\t').set_index('UniProtID')\n",
    "tempDF1 = tempDF1.loc[tempDF1.index.isin(tempDF.columns)]\n",
    "##Take only the target groups\n",
    "tempDF1 = tempDF1.loc[tempDF1['Intervention'].isin(tempL1)]\n",
    "\n",
    "#Prepare sample-mouse metadata for transcriptomics\n",
    "fileDir = './ImportData/'\n",
    "ipynbName = '230215_LC-M001-related-TrOmics-DIRAC-ver3_Preprocessing_onWenc_'\n",
    "fileName = 'sample-metadata.tsv'\n",
    "tempDF2 = pd.read_csv(fileDir+ipynbName+fileName, sep='\\t').set_index('SampleID')\n",
    "##Clean\n",
    "tempDF2['Phenotype'] = tempDF2['Intervention']#Sex-and-age-pooled for DIRAC template (rank consensus)\n",
    "tempDF2['Dataset'] = 'TrOmics'\n",
    "tempDF2['Intervention'] = tempDF2['Intervention'].str.replace('Ctrl1', 'Ctrl')#For consistency\n",
    "tempDF2 = tempDF2[tempL2]\n",
    "##Select the assessed samples\n",
    "fileName = 'normalized-data.tsv'\n",
    "tempDF = pd.read_csv(fileDir+ipynbName+fileName, sep='\\t').set_index('EnsemblID')\n",
    "tempDF2 = tempDF2.loc[tempDF2.index.isin(tempDF.columns)]\n",
    "##Take only the target groups\n",
    "tempDF2 = tempDF2.loc[tempDF2['Intervention'].isin(tempL1)]\n",
    "\n",
    "#Merge\n",
    "tempDF = pd.concat([tempDF1, tempDF2], axis=0)\n",
    "tempDF['Group'] = tempDF['Dataset'].str.slice(start=0, stop=1)+':'+tempDF['Intervention']\n",
    "\n",
    "display(tempDF)\n",
    "display(tempDF['Group'].value_counts())\n",
    "\n",
    "#Save\n",
    "fileDir = './ExportData/'\n",
    "ipynbName = '230224_LC-M001-PrOmics-vs-TrOmics-DIRAC-ver3_GOBP_'\n",
    "fileName = 'sample-metadata.tsv'\n",
    "tempDF.to_csv(fileDir+ipynbName+fileName, index=True, sep='\\t')\n",
    "\n",
    "sampleDF = tempDF"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1-3. DIRAC metrics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Prepare the DIRAC results for proteomics\n",
    "fileDir = '../230206_LC-M001-proteomics-DIRAC-ver7/ExportData/'\n",
    "ipynbName = '230214_LC-M001-proteomics-DIRAC-ver7-2_DIRAC-GOBP_'\n",
    "tempD1 = {'RMS':'RankMatchingScore-BS-combined.tsv',\n",
    "          'RCI':'RankConservationIndex-BS-combined.tsv'}\n",
    "dataset = 'PrOmics'\n",
    "tempD2 = {}\n",
    "for metric in tempD1.keys():\n",
    "    #Import the combined DIRAC results\n",
    "    fileName = tempD1[metric]\n",
    "    tempDF = pd.read_csv(fileDir+ipynbName+fileName, sep='\\t')\n",
    "    print(metric)\n",
    "    print('Original DF:', tempDF.shape)\n",
    "    print(' - Unique modules:', len(tempDF['ModuleID'].unique()))\n",
    "    print(' - Unique templates:', len(tempDF['Template'].unique()))\n",
    "    \n",
    "    #Take only the target modules and samples/groups\n",
    "    tempDF = tempDF.loc[tempDF['ModuleID'].isin(moduleDF.index.tolist())]\n",
    "    tempDF1 = sampleDF.loc[sampleDF['Dataset']==dataset]\n",
    "    tempDF = tempDF.loc[tempDF['Template'].isin(tempDF1['Phenotype'].unique())]\n",
    "    tempDF = tempDF.set_index(['ModuleID', 'Template'])\n",
    "    if metric=='RMS':\n",
    "        tempDF = tempDF.loc[:, tempDF.columns.isin(tempDF1.index.tolist())]\n",
    "    elif metric=='RCI':\n",
    "        tempDF = tempDF.loc[:, tempDF.columns.isin(tempDF1['Phenotype'].unique())]\n",
    "    tempDF = tempDF.reset_index()\n",
    "    \n",
    "    #Rename labels\n",
    "    tempS = tempDF1.reset_index().set_index('Phenotype')['Group'].drop_duplicates(keep='first')\n",
    "    tempDF['Template'] = tempDF['Template'].map(tempS)\n",
    "    if metric=='RCI':\n",
    "        tempDF = tempDF.set_index(['ModuleID', 'Template'])\n",
    "        tempDF.columns = tempDF.columns.map(tempS)\n",
    "        tempDF = tempDF.reset_index()\n",
    "    \n",
    "    print('Cleaned DF:', tempDF.shape)\n",
    "    print(' - Unique modules:', len(tempDF['ModuleID'].unique()))\n",
    "    print(' - Unique templates:', len(tempDF['Template'].unique()))\n",
    "    display(tempDF)\n",
    "    tempD2[metric] = tempDF\n",
    "    print('')\n",
    "\n",
    "#Save\n",
    "fileDir = './ExportData/'\n",
    "ipynbName = '230224_LC-M001-PrOmics-vs-TrOmics-DIRAC-ver3_GOBP_'\n",
    "for metric in tempD2.keys():\n",
    "    tempDF = tempD2[metric]\n",
    "    fileName = dataset+'-'+metric+'.tsv'\n",
    "    tempDF.to_csv(fileDir+ipynbName+fileName, index=True, sep='\\t')\n",
    "\n",
    "rmsDF_p = tempD2['RMS']\n",
    "rciDF_p = tempD2['RCI']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Prepare the DIRAC results for transcriptomics\n",
    "fileDir = './ImportData/'\n",
    "ipynbName = '230217_LC-M001-related-TrOmics-DIRAC-ver3_DIRAC-GOBP_onWenc_'\n",
    "tempD1 = {'RMS':'RankMatchingScore-BS-combined.tsv',\n",
    "          'RCI':'RankConservationIndex-BS-combined.tsv'}\n",
    "dataset = 'TrOmics'\n",
    "tempD2 = {}\n",
    "for metric in tempD1.keys():\n",
    "    #Import the combined DIRAC results\n",
    "    fileName = tempD1[metric]\n",
    "    tempDF = pd.read_csv(fileDir+ipynbName+fileName, sep='\\t')\n",
    "    print(metric)\n",
    "    print('Original DF:', tempDF.shape)\n",
    "    print(' - Unique modules:', len(tempDF['ModuleID'].unique()))\n",
    "    print(' - Unique templates:', len(tempDF['Template'].unique()))\n",
    "    \n",
    "    #Take only the target modules and samples/groups\n",
    "    tempDF = tempDF.loc[tempDF['ModuleID'].isin(moduleDF.index.tolist())]\n",
    "    tempDF1 = sampleDF.loc[sampleDF['Dataset']==dataset]\n",
    "    tempDF = tempDF.loc[tempDF['Template'].isin(tempDF1['Phenotype'].unique())]\n",
    "    tempDF = tempDF.set_index(['ModuleID', 'Template'])\n",
    "    if metric=='RMS':\n",
    "        tempDF = tempDF.loc[:, tempDF.columns.isin(tempDF1.index.tolist())]\n",
    "    elif metric=='RCI':\n",
    "        tempDF = tempDF.loc[:, tempDF.columns.isin(tempDF1['Phenotype'].unique())]\n",
    "    tempDF = tempDF.reset_index()\n",
    "    \n",
    "    #Rename labels\n",
    "    tempS = tempDF1.reset_index().set_index('Phenotype')['Group'].drop_duplicates(keep='first')\n",
    "    tempDF['Template'] = tempDF['Template'].map(tempS)\n",
    "    if metric=='RCI':\n",
    "        tempDF = tempDF.set_index(['ModuleID', 'Template'])\n",
    "        tempDF.columns = tempDF.columns.map(tempS)\n",
    "        tempDF = tempDF.reset_index()\n",
    "    \n",
    "    print('Cleaned DF:', tempDF.shape)\n",
    "    print(' - Unique modules:', len(tempDF['ModuleID'].unique()))\n",
    "    print(' - Unique templates:', len(tempDF['Template'].unique()))\n",
    "    display(tempDF)\n",
    "    tempD2[metric] = tempDF\n",
    "    print('')\n",
    "\n",
    "#Save\n",
    "fileDir = './ExportData/'\n",
    "ipynbName = '230224_LC-M001-PrOmics-vs-TrOmics-DIRAC-ver3_GOBP_'\n",
    "for metric in tempD2.keys():\n",
    "    tempDF = tempD2[metric]\n",
    "    fileName = dataset+'-'+metric+'.tsv'\n",
    "    tempDF.to_csv(fileDir+ipynbName+fileName, index=True, sep='\\t')\n",
    "\n",
    "rmsDF_t = tempD2['RMS']\n",
    "rciDF_t = tempD2['RCI']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. Rank conservation index: general pattern"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2-1. Extract RCI (the mean of RMSs under the own phenotype consensus)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Extract RCI whose template phenotype corresponds to the own phenotype\n",
    "rciDF_kk = pd.DataFrame(index=moduleDF.index)\n",
    "for tempDF in [rciDF_p, rciDF_t]:\n",
    "    phenotypeL = tempDF.drop(columns=['ModuleID', 'Template']).columns.tolist()\n",
    "    tempDF = tempDF.set_index('ModuleID')\n",
    "    for k in phenotypeL:\n",
    "        tempS = tempDF[k].loc[tempDF['Template']==k]\n",
    "        rciDF_kk = pd.merge(rciDF_kk, tempS, left_index=True, right_index=True, how='left')\n",
    "##Sort\n",
    "tempL1 = ['Ctrl', 'Aca', 'Rapa']\n",
    "tempL2 = ['PrOmics', 'TrOmics']\n",
    "tempL = [dataset[0]+':'+intervention for dataset in tempL2 for intervention in tempL1]\n",
    "rciDF_kk = rciDF_kk[tempL]\n",
    "display(rciDF_kk)\n",
    "display(rciDF_kk.describe())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2-2. Mann–Whitney U-test"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> Note that the scipy API (scipy.stats.mannwhitneyu) is used, because only the one-sided test seems implemented in the current statsmodels API (statsmodels.stats.nonparametric.rank_compare_2indep). Actually, the output objects are same b/w the two APIs, which is contrast to the case of t-test (degrees of freedom is not reported in the scipy API).  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tempDF = rciDF_kk\n",
    "tempL1 = ['Ctrl', 'Aca', 'Rapa']\n",
    "tempL2 = ['PrOmics', 'TrOmics']\n",
    "tempL = [dataset[0]+':'+intervention for dataset in tempL2 for intervention in tempL1]\n",
    "control_suffix = ':Ctrl'\n",
    "\n",
    "#Statistical tests\n",
    "tempDF1 = pd.DataFrame(columns=['Ustat', 'Pval'])\n",
    "for contrast in tempL:\n",
    "    dataset = re.sub(':.*$', '', contrast)\n",
    "    control = dataset+control_suffix\n",
    "    if control!=contrast:\n",
    "        tempS1 = tempDF[control]\n",
    "        tempS2 = tempDF[contrast]\n",
    "        #Two-sided Mann–Whitney U-test\n",
    "        ustat, pval = stats.mannwhitneyu(tempS2, tempS1,#U-statistic corresponds to the contrast\n",
    "                                         use_continuity=True, alternative='two-sided', method='auto')\n",
    "        tempDF1.loc[contrast+'-vs-'+control] = [ustat, pval]\n",
    "##P-value adjustment by using Benjamini–Hochberg method\n",
    "tempDF1['AdjPval'] = multi.multipletests(tempDF1['Pval'], alpha=0.05, method='fdr_bh',\n",
    "                                         is_sorted=False, returnsorted=False)[1]\n",
    "tempDF1.index.rename('ComparisonLabel', inplace=True)\n",
    "display(tempDF1)\n",
    "\n",
    "#Calculate general statistics\n",
    "tempDF2 = pd.DataFrame(columns=['N', 'RCImedian', 'RCImad'])\n",
    "for group in tempL:\n",
    "    tempS = tempDF[group]\n",
    "    size = len(tempS)\n",
    "    median = tempS.median()\n",
    "    mad = stats.median_absolute_deviation(tempS)#Cf. pd.Series.mad() is not median absolute deviation but mean absolute deviation\n",
    "    tempDF2.loc[group] = [size, median, mad]\n",
    "tempDF2.index.rename('GroupLabel', inplace=True)\n",
    "display(tempDF2)\n",
    "\n",
    "#Clean\n",
    "##Reformat while renaming column names\n",
    "tempD1 = {'Comparison':tempDF1, 'Group':tempDF2}\n",
    "tempD2 = {}\n",
    "for target in tempD1.keys():\n",
    "    tempDF3 = tempD1[target]\n",
    "    tempL1 = tempDF3.index.tolist()#For sorting later\n",
    "    tempL2 = tempDF3.columns.tolist()#For sorting later\n",
    "    tempDF3 = tempDF3.reset_index().melt(var_name='Variable', value_name='Value', id_vars=target+'Label')\n",
    "    tempDF3['Variable'] = tempDF3[target+'Label']+'_'+tempDF3['Variable']\n",
    "    tempDF3['ModuleID'] = 'All'#Dummy\n",
    "    tempDF3 = tempDF3.pivot(index='ModuleID', columns='Variable', values='Value')\n",
    "    tempDF3.columns.name = None#Erase 'Variable'\n",
    "    tempL3 = [label+'_'+variable for label in tempL1 for variable in tempL2]\n",
    "    tempDF3 = tempDF3[tempL3]#Sort\n",
    "    tempD2[target] = tempDF3\n",
    "##Merge\n",
    "tempDF3 = pd.merge(tempD2['Group'], tempD2['Comparison'], left_index=True, right_index=True, how='inner')\n",
    "##Convert data type\n",
    "for col_n in tempDF3.columns.tolist():\n",
    "    if re.search('_N$', col_n):\n",
    "        tempDF3[col_n] = tempDF3[col_n].astype(int)\n",
    "#Add dummy module name\n",
    "tempL1 = tempDF3.columns.tolist()#For sorting later\n",
    "tempDF3['ModuleName'] = 'General pattern'\n",
    "tempL2 = [col_n for sublist in [['ModuleName'], tempL1] for col_n in sublist]\n",
    "tempDF3 = tempDF3[tempL2]\n",
    "display(tempDF3)\n",
    "\n",
    "#Save\n",
    "fileDir = './ExportData/'\n",
    "ipynbName = '230224_LC-M001-PrOmics-vs-TrOmics-DIRAC-ver3_GOBP_'\n",
    "fileName = 'inter-group-comparison_RCIdistribution.tsv'\n",
    "tempDF3.to_csv(fileDir+ipynbName+fileName, sep='\\t', index=True)\n",
    "\n",
    "statDF = tempDF3"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2-3. Visualization: boxplot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#Prepare DF and color\n",
    "tempD0 = {'P:Ctrl':'P-Control', 'P:Aca':'P-Acarbose', 'P:Rapa':'P-Rapamycin',\n",
    "          'T:Ctrl':'T-Control', 'T:Aca':'T-Acarbose', 'T:Rapa':'T-Rapamycin'}\n",
    "tempDF = rciDF_kk.rename(columns=tempD0)\n",
    "tempDF = tempDF.reset_index().melt(var_name='Group', value_name='RCI', id_vars='ModuleID')\n",
    "tempD = {'P-Control':'tab:blue', 'P-Acarbose':'tab:red', 'P-Rapamycin':'tab:purple',\n",
    "         'T-Control':plt.get_cmap('tab20')(1), 'T-Acarbose':plt.get_cmap('tab20')(7),\n",
    "         'T-Rapamycin':plt.get_cmap('tab20')(9)}\n",
    "\n",
    "#Prepare significance labels\n",
    "##Retrieve statistical significance\n",
    "module = 'All'\n",
    "tempS = statDF.loc[module, statDF.columns.str.contains('AdjPval')]\n",
    "tempS.index = tempS.index.str.replace('_AdjPval', '')\n",
    "tempS.name = 'AdjPval'\n",
    "##Clean\n",
    "tempDF1 = tempS.index.to_series().str.split(pat='-vs-', expand=True)\n",
    "tempDF1 = tempDF1.rename(columns={0:'Contrast', 1:'Baseline'})\n",
    "tempDF1 = pd.merge(tempDF1, tempS, left_index=True, right_index=True, how='left')\n",
    "tempDF1['Contrast'] = tempDF1['Contrast'].map(tempD0)\n",
    "tempDF1['Baseline'] = tempDF1['Baseline'].map(tempD0)\n",
    "##Convert p-value to label\n",
    "tempL = []\n",
    "for row_i in range(len(tempDF1)):\n",
    "    pval = tempDF1['AdjPval'].iloc[row_i]\n",
    "    if pval<0.001:\n",
    "        tempL.append('***')\n",
    "    elif pval<0.01:\n",
    "        tempL.append('**')\n",
    "    elif pval<0.05:\n",
    "        tempL.append('*')\n",
    "    else:\n",
    "        pval_text = Decimal(str(pval)).quantize(Decimal('0.01'), rounding=ROUND_HALF_UP)\n",
    "        tempL.append(r'$P$ = '+str(pval_text))\n",
    "tempDF1['SignifLabel'] = tempL\n",
    "##Add the y-position level in figure\n",
    "tempDF1['YposLevel'] = [0, 1, 0, 1]\n",
    "display(tempDF1)\n",
    "\n",
    "#Visualization\n",
    "ymax = 1.0\n",
    "ymin = 0.5\n",
    "yinter = 0.1\n",
    "ymargin_t = 0.06\n",
    "ymargin_b = 0.01\n",
    "aline_ymin = 0.95\n",
    "aline_ymargin = 0.05\n",
    "sns.set(style='ticks', font='Arial', context='talk')\n",
    "plt.figure(figsize=(3, 4))\n",
    "p = sns.boxplot(data=tempDF, y='RCI', x='Group', order=list(tempD.keys()), palette=tempD, dodge=False,\n",
    "                showfliers=True, flierprops={'marker':'o', 'markerfacecolor':'gray', 'alpha':0.4},\n",
    "                showcaps=True, notch=True)\n",
    "p.set(ylim=(ymin-ymargin_b, ymax+ymargin_t), yticks=np.arange(ymin, ymax + yinter/10, yinter))\n",
    "##Add border line\n",
    "p.axvline(x=2.5, **{'linestyle':'dotted', 'color':'black', 'zorder':0})\n",
    "##Add significance labels\n",
    "lines = p.axes.get_lines()#Line2D: [[Q1, Q1-1.5IQR], [Q3, Q3+1.5IQR], [Q1, Q1], [Q3, Q3], [Med, Med], [flier]]\n",
    "lines_unit = 5 + int(True)#showfliers=True\n",
    "for row_i in range(len(tempDF1)):\n",
    "    #Baseline\n",
    "    group_0 = tempDF1['Baseline'].iloc[row_i]\n",
    "    index_0 = list(tempD.keys()).index(group_0)\n",
    "    whisker_0 = lines[index_0*lines_unit + 1]\n",
    "    xcoord_0 = whisker_0._x[1]#Q3+1.5IQR\n",
    "    #ycoord_0 = whisker_0._y[1]#Q3+1.5IQR\n",
    "    #Contrast\n",
    "    group_1 = tempDF1['Contrast'].iloc[row_i]\n",
    "    index_1 = list(tempD.keys()).index(group_1)\n",
    "    whisker_1 = lines[index_1*lines_unit + 1]\n",
    "    xcoord_1 = whisker_1._x[1]#Q3+1.5IQR\n",
    "    #ycoord_1 = whisker_1._y[1]#Q3+1.5IQR\n",
    "    #Standard point of marker\n",
    "    xcoord = (xcoord_0+xcoord_1)/2\n",
    "    #ycoord = max(ycoord_0, ycoord_1)\n",
    "    ycoord = aline_ymin + aline_ymargin*tempDF1['YposLevel'].iloc[row_i]\n",
    "    label = tempDF1['SignifLabel'].iloc[row_i]\n",
    "    #Add annotation lines\n",
    "    aline_offset = yinter/10\n",
    "    aline_length = yinter/10 + aline_offset\n",
    "    plt.plot([xcoord_0, xcoord_0, xcoord_1, xcoord_1],\n",
    "             [ycoord+aline_offset, ycoord+aline_length, ycoord+aline_length, ycoord+aline_offset],\n",
    "             lw=1.5, c='k')\n",
    "    #Add annotation text\n",
    "    if label in ['***', '**', '*']:\n",
    "        text_offset = yinter/25\n",
    "        p.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                   horizontalalignment='center', verticalalignment='bottom',\n",
    "                   fontsize='medium', color='k')\n",
    "    else:\n",
    "        text_offset = yinter/5\n",
    "        p.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                   horizontalalignment='center', verticalalignment='bottom',\n",
    "                   fontsize='x-small', color='k')\n",
    "sns.despine()\n",
    "plt.xlabel('')\n",
    "plt.ylabel('Module RCI')\n",
    "plt.xticks(rotation=70, horizontalalignment='right', verticalalignment='center', rotation_mode='anchor')\n",
    "##Save\n",
    "fileDir = './ExportFigures/'\n",
    "ipynbName = '230224_LC-M001-PrOmics-vs-TrOmics-DIRAC-ver3_GOBP_'\n",
    "fileName = 'RCI-boxplot.pdf'\n",
    "plt.gcf().savefig(fileDir+ipynbName+fileName, dpi=300, bbox_inches='tight', pad_inches=0.04, transparent=True)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3. Rank conservation index: inter-group module comparison"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> Test specific hypothesis: control RCI == intervention RCI (i.e., inter-group module comparison).  \n",
    "> 1. Testing the main effect of intervention on rank mathing scores (RMSs) for each module using ANOVA model  \n",
    "> 2. Then, performing post-hoc comparisons of RMSs between control vs. each intervention using Welch's t-tests  \n",
    ">  \n",
    "> Basically, statistical strategy is same with the one used in each dataset analysis. Because RMS/RCI was not normalized (i.e., the expected mean and variance could be different between datasets due to different number of mapped analytes), dataset and its interaction term are NOT included in ANOVA model; instead, ANOVA model is generated per dataset. The p-value adjustment is performed in a conservative manner: the P-values in ANOVA tests are adjusted across all models (= modules x datasets), and those in post-hoc tests are adjusted across all comparisons (interventions x datasets) only within the module (not across modules).  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3-1. Extract RMS under the own phenotype consensus"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Extract RMS whose template phenotype corresponds to the own phenotype\n",
    "rmsDF_kk = pd.DataFrame(index=moduleDF.index)\n",
    "for tempDF1, tempDF2 in zip([rmsDF_p, rmsDF_t], [rciDF_p, rciDF_t]):\n",
    "    phenotypeL = tempDF2.drop(columns=['ModuleID', 'Template']).columns.tolist()\n",
    "    tempDF1 = tempDF1.set_index('ModuleID')\n",
    "    for k in phenotypeL:\n",
    "        tempL = sampleDF.loc[sampleDF['Group']==k].index.tolist()\n",
    "        tempDF = tempDF1[tempL].loc[tempDF1['Template']==k]\n",
    "        rmsDF_kk = pd.merge(rmsDF_kk, tempDF, left_index=True, right_index=True, how='left')\n",
    "\n",
    "display(rmsDF_kk)\n",
    "display(rmsDF_kk.describe())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3-2. ANOVA test (RMS ~ Intervention), followed by Welch's t-tests (Intervention)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 3-2-1. Perform all statistical tests"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tempL1 = ['PrOmics', 'TrOmics']#Target dataset\n",
    "tempL2 = ['Ctrl', 'Aca', 'Rapa']#Target sample groups to be assessed\n",
    "control = 'Ctrl'#For post-hoc comparisons\n",
    "tempDF1 = rmsDF_kk\n",
    "tempDF2 = sampleDF.loc[sampleDF['Intervention'].isin(tempL2)]\n",
    "tempI = moduleDF.index\n",
    "formula = 'RMS ~ C(Intervention)'\n",
    "tempL3 = ['C(Intervention)']#For variables of interest in ANOVA\n",
    "\n",
    "#Statistical tests per dataset\n",
    "tempD1 = {}\n",
    "tempD2 = {}\n",
    "for dataset in tempL1:\n",
    "    #Statistical tests per module\n",
    "    t_start = time.time()\n",
    "    tempL4 = []#For ANOVA table\n",
    "    tempL5 = []#For post-hoc test table\n",
    "    for module in tempI.tolist():\n",
    "        #Select the target module RMSs\n",
    "        tempS = tempDF1.loc[module]\n",
    "        tempS.name = 'RMS'\n",
    "        #Add metadata while selecting the target samples\n",
    "        tempDF = tempDF2.loc[tempDF2['Dataset']==dataset]\n",
    "        tempDF = pd.merge(tempS, tempDF, left_index=True, right_index=True, how='inner')\n",
    "        \n",
    "        #ANOVA\n",
    "        model = smf.ols(formula, data=tempDF).fit()\n",
    "        anovaDF = anova_lm(model, typ=2)#ANOVA type doesn't matter in this case\n",
    "        ##Take the results per variable\n",
    "        tempDF3 = pd.DataFrame(columns=['DoF', 'Fstat', 'Pval'])\n",
    "        for variable in tempL3:\n",
    "            dof1 = int(anovaDF.at[variable, 'df'])#Between-groups\n",
    "            dof2 = int(anovaDF.at['Residual', 'df'])#Within-groups\n",
    "            dof = (dof1, dof2)\n",
    "            fstat = anovaDF.at[variable, 'F']\n",
    "            pval = anovaDF.at[variable, 'PR(>F)']\n",
    "            tempDF3.loc[variable] = [dof, fstat, pval]\n",
    "        tempDF3['AdjPval'] = 1.0#Add dummy column for now\n",
    "        ##Convert to wide-format\n",
    "        tempS = pd.Series(len(tempDF), index=['N'], name=module)\n",
    "        tempL = [tempS]\n",
    "        for variable in tempDF3.index.tolist():\n",
    "            tempS = tempDF3.loc[variable]\n",
    "            tempS.index = variable+'_'+tempS.index\n",
    "            tempS.name = module\n",
    "            tempL.append(tempS)\n",
    "        tempS = pd.concat(tempL, axis=0)\n",
    "        tempL4.append(tempS)\n",
    "        \n",
    "        #Post-hoc tests per control vs. contrast\n",
    "        tempDF4 = pd.DataFrame(columns=['DoF', 'tStat', 'Pval'])\n",
    "        for contrast in tempL2:\n",
    "            if control!=contrast:\n",
    "                tempS1 = tempDF['RMS'].loc[tempDF['Intervention']==control]\n",
    "                tempS2 = tempDF['RMS'].loc[tempDF['Intervention']==contrast]\n",
    "                #Two-sided Welch's t-test\n",
    "                tstat, pval, dof = weightstats.ttest_ind(tempS2, tempS1,#t-statistic reflects direction from the baseline\n",
    "                                                         alternative='two-sided', usevar='unequal')\n",
    "                tempDF4.loc[contrast+'-vs-'+control] = [dof, tstat, pval]\n",
    "        tempDF4['AdjPval'] = 1.0#Add dummy column for now\n",
    "        ##Convert to wide-format\n",
    "        tempL = []\n",
    "        for comparison in tempDF4.index.tolist():\n",
    "            tempS = tempDF4.loc[comparison]\n",
    "            tempS.index = comparison+'_'+tempS.index\n",
    "            tempS.name = module\n",
    "            tempL.append(tempS)\n",
    "        tempS = pd.concat(tempL, axis=0)\n",
    "        tempL5.append(tempS)\n",
    "    t_elapsed = time.time() - t_start\n",
    "    print(dataset)\n",
    "    print('Elapsed time for', len(tempI), 'ANOVA and',\n",
    "          (len(tempL2)-1)*len(tempI), 'post-hoc tests (',\n",
    "          len(tempL2)-1, 'comparisons x', len(tempI), 'modules):',\n",
    "          round(t_elapsed//60), 'min', round(t_elapsed%60, 1), 'sec')\n",
    "    \n",
    "    #Generate ANOVA table\n",
    "    tempDF3 = pd.concat(tempL4, axis=1).T\n",
    "    tempDF3.index.name = tempI.name\n",
    "    tempD1[dataset] = tempDF3\n",
    "    \n",
    "    #Generate post-hoc test table\n",
    "    tempDF4 = pd.concat(tempL5, axis=1).T\n",
    "    tempDF4.index.name = tempI.name\n",
    "    tempD2[dataset] = tempDF4\n",
    "\n",
    "#Clean all ANOVA tables\n",
    "tempDF3 = pd.DataFrame()\n",
    "for dataset in tempD1.keys():\n",
    "    tempDF = tempD1[dataset]\n",
    "    tempDF['Dataset'] = dataset\n",
    "    tempDF = tempDF.reset_index().set_index(['Dataset', 'ModuleID'])\n",
    "    tempDF3 = pd.concat([tempDF3, tempDF], axis=0)\n",
    "##P-value adjustment across all tests (modules x datasets) by using Benjamini–Hochberg method\n",
    "for variable in tempL3:\n",
    "    #Overwrite the dummy values\n",
    "    tempDF3[variable+'_AdjPval'] = multi.multipletests(tempDF3[variable+'_Pval'], alpha=0.05, method='fdr_bh',\n",
    "                                                       is_sorted=False, returnsorted=False)[1]\n",
    "##Convert back dtypes (due to the forced change during wide-format)\n",
    "for col_n in tempDF3.columns.tolist():\n",
    "    if 'N'==col_n:\n",
    "        tempDF3[col_n] = tempDF3[col_n].astype(int)\n",
    "    elif 'DoF' in col_n:\n",
    "        tempDF3[col_n] = tempDF3[col_n].astype(str)\n",
    "    else:\n",
    "        tempDF3[col_n] = tempDF3[col_n].astype(float)\n",
    "##Rename columns (because only one variable in this case)\n",
    "tempDF3.columns = 'ANOVA_'+tempDF3.columns.str.replace('^.*_', '', regex=True)\n",
    "display(tempDF3)\n",
    "\n",
    "#Clean all post-hoc test tables\n",
    "tempDF4 = pd.DataFrame()\n",
    "for dataset in tempD2.keys():\n",
    "    tempDF = tempD2[dataset]\n",
    "    tempDF['Dataset'] = dataset\n",
    "    tempDF = tempDF.reset_index().set_index(['Dataset', 'ModuleID'])\n",
    "    tempDF4 = pd.concat([tempDF4, tempDF], axis=0)\n",
    "##P-value adjustment across all tests per module (comparisons x datasets) by using Benjamini–Hochberg method\n",
    "for module in tempI.tolist():\n",
    "    tempL = tempDF4.loc[:, tempDF4.columns.str.contains('_Pval$', regex=True)].columns.tolist()\n",
    "    tempDF = tempDF4.reset_index().melt(var_name='Comparison', value_name='Pval', value_vars=tempL,\n",
    "                                        id_vars=['Dataset', 'ModuleID'])\n",
    "    tempDF = tempDF.loc[tempDF['ModuleID']==module]\n",
    "    tempDF['AdjPval'] = multi.multipletests(tempDF['Pval'], alpha=0.05, method='fdr_bh',\n",
    "                                            is_sorted=False, returnsorted=False)[1]\n",
    "    tempDF = tempDF.pivot(index=['Dataset', 'ModuleID'], columns='Comparison', values='AdjPval')\n",
    "    tempDF.columns = tempDF.columns.str.replace('_Pval', '_AdjPval')\n",
    "    #Replace the dummy values with the adjusted p-values\n",
    "    tempL = tempDF4.loc[:, tempDF4.columns.str.contains('_AdjPval$', regex=True)].columns.tolist()\n",
    "    for col_n in tempL:\n",
    "        for dataset in tempL1:\n",
    "            tempDF4.loc[(dataset, module), col_n] = tempDF.loc[(dataset, module), col_n]\n",
    "display(tempDF4)\n",
    "\n",
    "statDF1 = tempDF3\n",
    "statDF2 = tempDF4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tempL1 = ['PrOmics', 'TrOmics']#Target dataset\n",
    "tempL2 = ['Ctrl', 'Aca', 'Rapa']#Target sample groups to be summarized\n",
    "tempDF1 = rmsDF_kk\n",
    "tempDF2 = sampleDF.loc[sampleDF['Intervention'].isin(tempL2)]\n",
    "\n",
    "#Calculate general statistics per dataset\n",
    "tempD = {}\n",
    "for dataset in tempL1:\n",
    "    #Calculate general statistics per intervention group\n",
    "    tempL3 = []\n",
    "    for intervention in tempL2:\n",
    "        #Select the target samples\n",
    "        tempL = tempDF2.loc[(tempDF2['Dataset']==dataset)&\n",
    "                            (tempDF2['Intervention']==intervention)].index.tolist()\n",
    "        tempDF = tempDF1[tempL]\n",
    "        #Calculate general statistics\n",
    "        tempS1 = len(tempL) - tempDF.isnull().sum(axis=1)\n",
    "        tempS1.name = intervention+'_N'\n",
    "        tempS2 = tempDF.mean(axis=1)\n",
    "        tempS2.name = intervention+'_RMSmean'\n",
    "        tempS3 = tempDF.sem(axis=1, ddof=1)\n",
    "        tempS3.name = intervention+'_RMSsem'\n",
    "        #Merge\n",
    "        tempDF = pd.concat([tempS1, tempS2, tempS3], axis=1)\n",
    "        tempL3.append(tempDF)\n",
    "    tempDF = pd.concat(tempL3, axis=1)\n",
    "    tempD[dataset] = tempDF\n",
    "##Clean all general statistics tables\n",
    "tempDF3 = pd.DataFrame()\n",
    "for dataset in tempD.keys():\n",
    "    tempDF = tempD[dataset]\n",
    "    tempDF['Dataset'] = dataset\n",
    "    tempDF = tempDF.reset_index().set_index(['Dataset', 'ModuleID'])\n",
    "    tempDF3 = pd.concat([tempDF3, tempDF], axis=0)\n",
    "display(tempDF3)\n",
    "\n",
    "#Merge all the tables\n",
    "print('General statistics table:', tempDF3.shape)\n",
    "print('ANOVA table:', statDF1.shape)\n",
    "print('Post-hoc test table:', statDF2.shape)\n",
    "tempDF = pd.merge(moduleDF['ModuleName'], tempDF3.reset_index(), on='ModuleID', how='right')\n",
    "tempDF = pd.concat([tempDF.set_index(['Dataset', 'ModuleID']), statDF1, statDF2], axis=1)\n",
    "\n",
    "#Sort\n",
    "tempDF = tempDF.sort_values(by='ANOVA_Pval', ascending=True)\n",
    "display(tempDF)\n",
    "\n",
    "#Save per rank consensus\n",
    "fileDir = './ExportData/'\n",
    "ipynbName = '230224_LC-M001-PrOmics-vs-TrOmics-DIRAC-ver3_GOBP_'\n",
    "for dataset in tempL1:\n",
    "    fileName = 'inter-group-comparison_'+dataset+'-RCI.tsv'\n",
    "    tempDF3 = tempDF.loc[dataset]\n",
    "    tempDF3.to_csv(fileDir+ipynbName+fileName, sep='\\t', index=True)\n",
    "    print('Saved .tsv table for '+dataset)\n",
    "    display(tempDF3)\n",
    "\n",
    "statDF = tempDF"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 3-2-2. Changed modules (ANOVA)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Prepare variables in the model\n",
    "#-> In this case, only the one variable (intervention) was included.\n",
    "variableL = ['ANOVA']\n",
    "\n",
    "#Changed modules\n",
    "for variable in variableL:\n",
    "    tempDF = statDF.loc[statDF[variable+'_AdjPval']<0.05]\n",
    "    tempDF = tempDF.sort_values(by=variable+'_AdjPval', ascending=True)\n",
    "    tempL1 = tempDF.loc[:, tempDF.columns.str.contains('_RMSmean')].columns.tolist()\n",
    "    tempL2 = tempDF.loc[:, tempDF.columns.str.contains('^'+variable+'_')].columns.tolist()\n",
    "    tempDF = tempDF[[col_n for subL in [['ModuleName'], tempL1, tempL2] for col_n in subL]]\n",
    "    print(variable+' (adjusted P < 0.05):', len(tempDF))\n",
    "    tempL = tempDF.index.to_frame()['ModuleID'].unique().tolist()\n",
    "    print(' -> Unique module:', len(tempL))\n",
    "    tempL = tempDF.index.to_frame()['Dataset'].unique().tolist()\n",
    "    for dataset in tempL:\n",
    "        tempDF1 = tempDF.loc[dataset]\n",
    "        print(' -> '+dataset+':', len(tempDF1))\n",
    "    display(tempDF)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 4-2-3. Changed modules by each intervention (Welch's t-test)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Spread statDF by template to flatten multi-index\n",
    "tempDF = moduleDF['ModuleName']#pd.Series() for now\n",
    "for dataset in ['PrOmics', 'TrOmics']:\n",
    "    tempDF1 = statDF.loc[dataset].drop(columns=['ModuleName'])\n",
    "    tempDF1.columns = dataset+'_'+tempDF1.columns\n",
    "    tempDF = pd.merge(tempDF, tempDF1, left_index=True, right_index=True, how='left')\n",
    "\n",
    "#Extract only the changed modules\n",
    "variable = 'ANOVA'\n",
    "tempS1 = tempDF['PrOmics_'+variable+'_AdjPval']<0.05\n",
    "tempS2 = tempDF['TrOmics_'+variable+'_AdjPval']<0.05\n",
    "tempDF = tempDF.loc[tempS1|tempS2]\n",
    "tempDF = tempDF.sort_values(by=['PrOmics_'+variable+'_AdjPval', 'TrOmics_'+variable+'_AdjPval'], ascending=True)\n",
    "print(variable+' (adjusted P < 0.05):', len(tempDF))\n",
    "print(' -> Proteomics:', sum(tempS1))\n",
    "print(' -> Transcriptomics:', sum(tempS2))\n",
    "\n",
    "#Take adjusted P-value\n",
    "tempDF1 = tempDF.loc[:, tempDF.columns.str.contains('-vs-.*_AdjPval$')]\n",
    "tempDF1.columns = tempDF1.columns.str.replace('_AdjPval$', '')\n",
    "tempDF1 = pd.merge(tempDF[['ModuleName', 'PrOmics_'+variable+'_AdjPval', 'TrOmics_'+variable+'_AdjPval']],\n",
    "                   tempDF1, left_index=True, right_index=True, how='left')\n",
    "print('Adjusted P-value:')\n",
    "display(tempDF1)\n",
    "display(tempDF1.describe())\n",
    "\n",
    "#Take effect size\n",
    "tempDF2 = tempDF.loc[:, tempDF.columns.str.contains('-vs-.*_tStat$')]\n",
    "tempDF2.columns = tempDF2.columns.str.replace('_tStat$', '')\n",
    "tempDF2 = pd.merge(tempDF[['ModuleName', 'PrOmics_'+variable+'_AdjPval', 'TrOmics_'+variable+'_AdjPval']],\n",
    "                   tempDF2, left_index=True, right_index=True, how='left')\n",
    "print('Changed direction (t-statistic):')\n",
    "display(tempDF2)\n",
    "display(tempDF2.describe())\n",
    "\n",
    "pvalDF = tempDF1\n",
    "diffDF = tempDF2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> Check the changed modules (based on the nominal P-value for the main effect and the adjusted P-values for the post-hoc tests) as reference.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Spread statDF by template to flatten multi-index\n",
    "tempDF = moduleDF['ModuleName']#pd.Series() for now\n",
    "for dataset in ['PrOmics', 'TrOmics']:\n",
    "    tempDF1 = statDF.loc[dataset].drop(columns=['ModuleName'])\n",
    "    tempDF1.columns = dataset+'_'+tempDF1.columns\n",
    "    tempDF = pd.merge(tempDF, tempDF1, left_index=True, right_index=True, how='left')\n",
    "\n",
    "#Extract only the changed modules\n",
    "variable = 'ANOVA'\n",
    "tempS1 = tempDF['PrOmics_'+variable+'_Pval']<0.05\n",
    "tempS2 = tempDF['TrOmics_'+variable+'_Pval']<0.05\n",
    "tempDF = tempDF.loc[tempS1|tempS2]\n",
    "tempDF = tempDF.sort_values(by=['PrOmics_'+variable+'_Pval',\n",
    "                                'TrOmics_'+variable+'_Pval'], ascending=True)\n",
    "print(variable+' (nominal P < 0.05):', len(tempDF))\n",
    "print(' -> Proteomics:', sum(tempS1))\n",
    "print(' -> Transcriptomics:', sum(tempS2))\n",
    "\n",
    "#Take adjusted P-value\n",
    "tempDF1 = tempDF.loc[:, tempDF.columns.str.contains('-vs-.*_AdjPval$')]\n",
    "tempDF1.columns = tempDF1.columns.str.replace('_AdjPval$', '')\n",
    "tempDF1 = pd.merge(tempDF[['ModuleName', 'PrOmics_'+variable+'_AdjPval', 'TrOmics_'+variable+'_AdjPval']],\n",
    "                   tempDF1, left_index=True, right_index=True, how='left')\n",
    "print('Adjusted P-value:')\n",
    "display(tempDF1)\n",
    "display(tempDF1.describe())\n",
    "\n",
    "#Take effect size\n",
    "tempDF2 = tempDF.loc[:, tempDF.columns.str.contains('-vs-.*_tStat$')]\n",
    "tempDF2.columns = tempDF2.columns.str.replace('_tStat$', '')\n",
    "tempDF2 = pd.merge(tempDF[['ModuleName', 'PrOmics_'+variable+'_AdjPval', 'TrOmics_'+variable+'_AdjPval']],\n",
    "                   tempDF2, left_index=True, right_index=True, how='left')\n",
    "#print('Changed direction (t-statistic):')\n",
    "#display(tempDF2)\n",
    "#display(tempDF2.describe())\n",
    "\n",
    "pvalDF_ref = tempDF1\n",
    "diffDF_ref = tempDF2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "### 3-3. Visualization: clustermap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tempD = {'P:Ctrl':'P-Control', 'P:Aca':'P-Acarbose', 'P:Rapa':'P-Rapamycin',\n",
    "         'T:Ctrl':'T-Control', 'T:Aca':'T-Acarbose', 'T:Rapa':'T-Rapamycin'}\n",
    "tempDF1 = rciDF_kk.rename(columns=tempD)\n",
    "regulation = 'Tightened'\n",
    "tempD0 = {'Ctrl':'Control', 'Aca':'Acarbose', 'Rapa':'Rapamycin'}\n",
    "tempD1 = {'Control':'tab:blue', 'Acarbose':'tab:red', 'Rapamycin':'tab:purple'}\n",
    "tempD2 = {'Control':plt.get_cmap('tab20')(1), 'Acarbose':plt.get_cmap('tab20')(7),\n",
    "          'Rapamycin':plt.get_cmap('tab20')(9)}\n",
    "\n",
    "#Prepare color labels for changed module set\n",
    "tempDF2 = pvalDF_ref.loc[:, pvalDF_ref.columns.str.contains('-vs-')]#Based on nominal ANOVA P-values\n",
    "for col_n in tempDF2.columns.tolist():\n",
    "    tempS1 = pvalDF_ref[col_n]#Adjusted P-values in post-hoc tests\n",
    "    tempS2 = diffDF_ref[col_n]\n",
    "    tempS3 = pvalDF[col_n]#Adjusted P-values in post-hoc tests\n",
    "    tempS4 = diffDF[col_n]\n",
    "    if regulation=='Changed':\n",
    "        tempS2 = tempS2.loc[(tempS1<0.05)]\n",
    "        tempS4 = tempS4.loc[(tempS3<0.05)]\n",
    "    elif regulation=='Tightened':\n",
    "        tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2>0)]\n",
    "        tempS4 = tempS4.loc[(tempS3<0.05)&(tempS4>0)]\n",
    "    elif regulation=='Loosened':\n",
    "        tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2<0)]\n",
    "        tempS4 = tempS4.loc[(tempS3<0.05)&(tempS4<0)]\n",
    "    label = re.sub('-vs-.*', '', col_n)\n",
    "    group = tempD0[re.sub('.*_', '', label)]\n",
    "    dataset = re.sub('_.*', '', label)\n",
    "    tempL = []\n",
    "    for module in tempDF2.index.tolist():\n",
    "        if module in tempS2.index.tolist():\n",
    "            if module in tempS4.index.tolist():#Based on adjusted ANOVA P-values\n",
    "                tempL.append(tempD1[group])\n",
    "            else:#Based on nominal ANOVA P-values\n",
    "                tempL.append(tempD2[group])\n",
    "        else:\n",
    "            tempL.append('white')\n",
    "    tempDF2[dataset[0]+'-'+group] = tempL\n",
    "    print(regulation+' module in '+col_n)\n",
    "    print(' -> in adjusted ANOVA P < 0.05:', len(tempS4))\n",
    "    print(' -> in nominal ANOVA P < 0.05:', len(tempS2))\n",
    "tempDF2 = tempDF2[['P-Acarbose', 'P-Rapamycin', 'T-Acarbose', 'T-Rapamycin']]#Sort\n",
    "\n",
    "#Prepare color labels for samples\n",
    "tempA = np.repeat(['tab:pink', 'tab:cyan'], 3)\n",
    "tempDF3 = pd.DataFrame({'Data':tempA, 'Group':np.tile(list(tempD1.values()), 2)},\n",
    "                       index=list(tempD.values()))\n",
    "\n",
    "#Clustermap\n",
    "sns.set(style='ticks', font='Arial', context='talk')\n",
    "cm = sns.clustermap(tempDF1.T, method='ward', metric='euclidean', cmap='afmhot',\n",
    "                    row_cluster=True, col_cluster=True, row_linkage=None, col_linkage=None,\n",
    "                    row_colors=tempDF3, col_colors=tempDF2, xticklabels=False, yticklabels=True,\n",
    "                    dendrogram_ratio=(0.025, 0.2), colors_ratio=(0.025, 0.075),\n",
    "                    cbar_pos=(0.05, -0.05, 0.3, 0.075), cbar_kws={'orientation': 'horizontal'},\n",
    "                    figsize=(12, 4), **{'vmin':0.5, 'vmax':1})\n",
    "cm.cax.set_title('Module RCI', size='medium',\n",
    "                 verticalalignment='bottom', horizontalalignment='center')\n",
    "cm.cax.tick_params(labelsize='small')\n",
    "bottom, top = cm.ax_heatmap.get_ylim()\n",
    "#cm.ax_heatmap.set_ylim(bottom + 0.5, top - 0.5)##To avoid half cut of first and last rows\n",
    "hm = cm.ax_heatmap.get_position()\n",
    "rd = cm.ax_row_dendrogram.get_position()\n",
    "cd = cm.ax_col_dendrogram.get_position()\n",
    "cm.ax_heatmap.set_position([hm.x0, hm.y0, hm.width, hm.height])\n",
    "cm.ax_row_dendrogram.set_position([rd.x0, rd.y0, rd.width, rd.height])\n",
    "cm.ax_col_dendrogram.set_position([cd.x0, cd.y0, cd.width, cd.height])\n",
    "cm.ax_heatmap.set_xlabel('GOBP module')\n",
    "cm.ax_heatmap.set_ylabel('')\n",
    "##row/column color bar legend (axis is same with cm.cax!)\n",
    "tempL = []\n",
    "for group in tempD1.keys():\n",
    "    if group!='Control':\n",
    "        tempL.append(mpatches.Patch(color=tempD1[group],\n",
    "                                    label='by '+group+' (adjusted '+r'$P$'+' < 0.05)'))\n",
    "for group in tempD2.keys():\n",
    "    if group!='Control':\n",
    "        tempL.append(mpatches.Patch(color=tempD2[group], label='(nominal '+r'$P$'+' < 0.05)'))\n",
    "legend1 = plt.legend(handles=tempL, fontsize='small', labelspacing=0.2, ncol=2,\n",
    "                     title='Tightened module (vs. Control)', title_fontsize='medium',\n",
    "                     bbox_to_anchor=(1, 0.5), loc='center left', borderaxespad=3.5, frameon=False)\n",
    "plt.gca().add_artist(legend1)\n",
    "##Save\n",
    "fileDir = './ExportFigures/'\n",
    "ipynbName = '230224_LC-M001-PrOmics-vs-TrOmics-DIRAC-ver3_GOBP_'\n",
    "fileName = 'RCI-clustermap.pdf'\n",
    "#plt.gcf().savefig(fileDir+ipynbName+fileName, dpi=300, bbox_inches='tight', pad_inches=0.04, transparent=True)\n",
    "plt.show()\n",
    "\n",
    "#Save label order\n",
    "tempDF = moduleDF.loc[rciDF_kk.index[cm.dendrogram_col.reordered_ind]]\n",
    "tempDF1 = rciDF_kk.copy()\n",
    "tempDF = pd.merge(tempDF['ModuleName'], tempDF1, left_index=True, right_index=True, how='left')\n",
    "tempDF1 = pvalDF.loc[:, pvalDF.columns.str.contains('-vs-')]#Adjusted P-value\n",
    "tempDF = pd.merge(tempDF, tempDF1, left_index=True, right_index=True, how='left')\n",
    "tempDF = tempDF.reset_index()\n",
    "tempDF.index.name = 'Xcoord'\n",
    "display(tempDF)\n",
    "fileDir = './ExportData/'\n",
    "ipynbName = '230224_LC-M001-PrOmics-vs-TrOmics-DIRAC-ver3_GOBP_'\n",
    "fileName = 'RCI-clustermap-xticks.tsv'\n",
    "#tempDF.to_csv(fileDir+ipynbName+fileName, index=True, sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> –> Simplify.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tempD = {'P:Ctrl':'P-Control', 'P:Aca':'P-Acarbose', 'P:Rapa':'P-Rapamycin',\n",
    "         'T:Ctrl':'T-Control', 'T:Aca':'T-Acarbose', 'T:Rapa':'T-Rapamycin'}\n",
    "tempDF1 = rciDF_kk.rename(columns=tempD)\n",
    "regulation = 'Tightened'\n",
    "tempD0 = {'Ctrl':'Control', 'Aca':'Acarbose', 'Rapa':'Rapamycin'}\n",
    "tempD1 = {'Control':'tab:blue', 'Acarbose':'tab:red', 'Rapamycin':'tab:purple'}\n",
    "\n",
    "#Prepare color labels for changed module set\n",
    "tempDF2 = pvalDF.loc[:, pvalDF.columns.str.contains('-vs-')]#Based on adjusted ANOVA P-values\n",
    "for col_n in tempDF2.columns.tolist():\n",
    "    tempS1 = pvalDF[col_n]#Adjusted P-values in post-hoc tests\n",
    "    tempS2 = diffDF[col_n]\n",
    "    if regulation=='Changed':\n",
    "        tempS2 = tempS2.loc[(tempS1<0.05)]\n",
    "    elif regulation=='Tightened':\n",
    "        tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2>0)]\n",
    "    elif regulation=='Loosened':\n",
    "        tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2<0)]\n",
    "    label = re.sub('-vs-.*', '', col_n)\n",
    "    group = tempD0[re.sub('.*_', '', label)]\n",
    "    dataset = re.sub('_.*', '', label)\n",
    "    tempL = []\n",
    "    for module in tempDF2.index.tolist():\n",
    "        if module in tempS2.index.tolist():\n",
    "            tempL.append(tempD1[group])\n",
    "        else:\n",
    "            tempL.append('white')\n",
    "    tempDF2[dataset[0]+'-'+group] = tempL\n",
    "    print(regulation+' module in '+col_n)\n",
    "    print(' -> in adjusted ANOVA P < 0.05:', len(tempS2))\n",
    "tempDF2 = tempDF2[['P-Acarbose', 'P-Rapamycin', 'T-Acarbose', 'T-Rapamycin']]#Sort\n",
    "\n",
    "#Prepare color labels for samples\n",
    "tempA = np.repeat(['tab:pink', 'tab:cyan'], 3)\n",
    "tempDF3 = pd.DataFrame({'Data':tempA, 'Group':np.tile(list(tempD1.values()), 2)},\n",
    "                       index=list(tempD.values()))\n",
    "\n",
    "#Clustermap\n",
    "sns.set(style='ticks', font='Arial', context='talk')\n",
    "cm = sns.clustermap(tempDF1.T, method='ward', metric='euclidean', cmap='afmhot',\n",
    "                    row_cluster=True, col_cluster=True, row_linkage=None, col_linkage=None,\n",
    "                    row_colors=tempDF3, col_colors=tempDF2, xticklabels=False, yticklabels=True,\n",
    "                    dendrogram_ratio=(0.025, 0.2), colors_ratio=(0.025, 0.075),\n",
    "                    cbar_pos=(0.175, -0.15, 0.3, 0.075), cbar_kws={'orientation': 'horizontal'},\n",
    "                    figsize=(12, 4), **{'vmin':0.5, 'vmax':1})\n",
    "cm.cax.set_title('Module RCI', size='medium',\n",
    "                 verticalalignment='bottom', horizontalalignment='center')\n",
    "cm.cax.tick_params(labelsize='small')\n",
    "bottom, top = cm.ax_heatmap.get_ylim()\n",
    "#cm.ax_heatmap.set_ylim(bottom + 0.5, top - 0.5)##To avoid half cut of first and last rows\n",
    "hm = cm.ax_heatmap.get_position()\n",
    "rd = cm.ax_row_dendrogram.get_position()\n",
    "cd = cm.ax_col_dendrogram.get_position()\n",
    "cm.ax_heatmap.set_position([hm.x0, hm.y0, hm.width, hm.height])\n",
    "cm.ax_row_dendrogram.set_position([rd.x0, rd.y0, rd.width, rd.height])\n",
    "cm.ax_col_dendrogram.set_position([cd.x0, cd.y0, cd.width, cd.height])\n",
    "cm.ax_heatmap.set_xlabel('GOBP module')\n",
    "cm.ax_heatmap.set_ylabel('')\n",
    "##row/column color bar legend (axis is same with cm.cax!)\n",
    "tempL = []\n",
    "for group in tempD1.keys():\n",
    "    if group!='Control':\n",
    "        tempL.append(mpatches.Patch(color=tempD1[group], label='by '+group))\n",
    "legend1 = plt.legend(handles=tempL, fontsize='small', labelspacing=0.2,\n",
    "                     title='Tightened module (vs. Control)', title_fontsize='medium',\n",
    "                     bbox_to_anchor=(1, 0.5), loc='center left', borderaxespad=3.5, frameon=False)\n",
    "plt.gca().add_artist(legend1)\n",
    "##Save\n",
    "fileDir = './ExportFigures/'\n",
    "ipynbName = '230224_LC-M001-PrOmics-vs-TrOmics-DIRAC-ver3_GOBP_'\n",
    "fileName = 'RCI-clustermap.pdf'\n",
    "plt.gcf().savefig(fileDir+ipynbName+fileName, dpi=300, bbox_inches='tight', pad_inches=0.04, transparent=True)\n",
    "plt.show()\n",
    "\n",
    "#Save label order\n",
    "tempDF = moduleDF.loc[rciDF_kk.index[cm.dendrogram_col.reordered_ind]]\n",
    "tempDF1 = rciDF_kk.copy()\n",
    "tempDF = pd.merge(tempDF['ModuleName'], tempDF1, left_index=True, right_index=True, how='left')\n",
    "tempDF1 = pvalDF.loc[:, pvalDF.columns.str.contains('-vs-')]#Adjusted P-value\n",
    "tempDF = pd.merge(tempDF, tempDF1, left_index=True, right_index=True, how='left')\n",
    "tempDF = tempDF.reset_index()\n",
    "tempDF.index.name = 'Xcoord'\n",
    "display(tempDF)\n",
    "fileDir = './ExportData/'\n",
    "ipynbName = '230224_LC-M001-PrOmics-vs-TrOmics-DIRAC-ver3_GOBP_'\n",
    "fileName = 'RCI-clustermap-xticks.tsv'\n",
    "tempDF.to_csv(fileDir+ipynbName+fileName, index=True, sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3-4. Visualization: venn diagram"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Prepare label and color\n",
    "tempD0 = {'Ctrl':'Control', 'Aca':'Acarbose', 'Rapa':'Rapamycin'}\n",
    "tempD1 = {'T-Acarbose':plt.get_cmap('tab20')(7), 'P-Acarbose':'tab:red',\n",
    "          'P-Rapamycin':'tab:purple', 'T-Rapamycin':plt.get_cmap('tab20')(9)}\n",
    "\n",
    "#Visualization per direction\n",
    "for regulation in ['Changed', 'Tightened', 'Loosened']:\n",
    "    #Prepare module sets\n",
    "    tempD2 = {}\n",
    "    tempDF = pvalDF.loc[:, pvalDF.columns.str.contains('-vs-')]#Based on adjusted ANOVA P-values\n",
    "    for col_n in tempDF.columns.tolist():\n",
    "        tempS1 = pvalDF[col_n]\n",
    "        tempS2 = diffDF[col_n]\n",
    "        if regulation=='Changed':\n",
    "            tempS2 = tempS2.loc[(tempS1<0.05)]\n",
    "        elif regulation=='Tightened':\n",
    "            tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2>0)]\n",
    "        elif regulation=='Loosened':\n",
    "            tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2<0)]\n",
    "        label = re.sub('-vs-.*', '', col_n)\n",
    "        group = tempD0[re.sub('.*_', '', label)]\n",
    "        dataset = re.sub('_.*', '', label)\n",
    "        tempD2[dataset[0]+'-'+group] = set(tempS2.index.tolist())\n",
    "    ##Sort to make consistent order in manual legend generation\n",
    "    tempD = {}\n",
    "    for label in tempD1.keys():\n",
    "        tempD[label] = tempD2[label]\n",
    "    \n",
    "    #Not significant in all contrasts\n",
    "    tempDF = pvalDF.copy()#Based on adjusted ANOVA P-values\n",
    "    for moduleS in tempD.values():\n",
    "        tempDF = tempDF.loc[~tempDF.index.isin(moduleS)]\n",
    "    count = len(tempDF)\n",
    "    print(regulation+' modules (vs. Control):')\n",
    "    print(' -> Not significant in all contrasts:', count)\n",
    "    \n",
    "    #Skip the followings if no significant module\n",
    "    if count==len(pvalDF):\n",
    "        continue\n",
    "    \n",
    "    #Venn diagram\n",
    "    sns.set(style='ticks', font='Arial', context='talk')\n",
    "    fig, ax = plt.subplots(figsize=(4, 4))\n",
    "    venn(tempD, fmt='{size:,}', cmap=list(tempD1.values()), legend_loc=None, ax=ax)\n",
    "    plt.setp(ax, ylim=(0.1, 0.875))#Otherwise, weird space...\n",
    "    ##Add legend annotation\n",
    "    x_coord = [0.1, 0.1, 0.9, 0.9]\n",
    "    y_coord = [0.25, 0.7, 0.7, 0.25]\n",
    "    h_align = ['right', 'right', 'left', 'left']\n",
    "    v_align = ['top', 'bottom', 'bottom', 'top']\n",
    "    for i in range(len(tempD1)):\n",
    "        key = list(tempD1.keys())[i]\n",
    "        total = f'{len(tempD[key]):,}'\n",
    "        ax.text(x_coord[i], y_coord[i], key+'\\n('+total+' modules)',\n",
    "                fontsize='small', multialignment='center',\n",
    "                horizontalalignment=h_align[i], verticalalignment=v_align[i],\n",
    "                bbox={'boxstyle':'round', 'facecolor':tempD1[key], 'pad':0.2, 'alpha':0.5})\n",
    "    title = regulation+' modules (vs. Control)'\n",
    "    ax.set_title(title, fontsize='medium')\n",
    "    ##Save\n",
    "    if regulation!='Changed':\n",
    "        fileDir = './ExportFigures/'\n",
    "        ipynbName = '230224_LC-M001-PrOmics-vs-TrOmics-DIRAC-ver3_GOBP_'\n",
    "        fileName = 'RCI-venn-'+regulation.lower()+'.pdf'\n",
    "        plt.gcf().savefig(fileDir+ipynbName+fileName, dpi=300, bbox_inches='tight', pad_inches=0.04, transparent=True)\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Export module list in each subset in the venn diagram\n",
    "for regulation in ['Tightened', 'Loosened']:\n",
    "    #Prepare module sets\n",
    "    tempD = {}\n",
    "    tempDF = pvalDF.loc[:, pvalDF.columns.str.contains('-vs-')]#Based on adjusted ANOVA P-values\n",
    "    for col_n in tempDF.columns.tolist():\n",
    "        tempS1 = pvalDF[col_n]\n",
    "        tempS2 = diffDF[col_n]\n",
    "        if regulation=='Changed':\n",
    "            tempS2 = tempS2.loc[(tempS1<0.05)]\n",
    "        elif regulation=='Tightened':\n",
    "            tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2>0)]\n",
    "        elif regulation=='Loosened':\n",
    "            tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2<0)]\n",
    "        tempD[col_n] = set(tempS2.index.tolist())\n",
    "    \n",
    "    #Not significant in all contrasts\n",
    "    tempDF = pvalDF.copy()#Based on adjusted ANOVA P-values\n",
    "    for moduleS in tempD.values():\n",
    "        tempDF = tempDF.loc[~tempDF.index.isin(moduleS)]\n",
    "    count = len(tempDF)\n",
    "    print(regulation+' modules (vs. Control):')\n",
    "    print(' -> Not significant in all contrasts:', count)\n",
    "    \n",
    "    #Skip the followings if no significant module\n",
    "    if count==len(pvalDF):\n",
    "        continue\n",
    "    \n",
    "    #Prepare a new .xlsx file (dummy README)\n",
    "    tempL1 = [len(tempD[key]) for key in tempD.keys()]\n",
    "    tempDF = pd.DataFrame({'Group':tempD.keys(), 'nModules':tempL1})\n",
    "    tempDF = tempDF.reset_index().rename(columns={'index':'VennOrder'})\n",
    "    tempDF['VennOrder'] = tempDF['VennOrder'] + 1\n",
    "    fileDir = './ExportData/'\n",
    "    ipynbName = '230224_LC-M001-PrOmics-vs-TrOmics-DIRAC-ver3_GOBP_'\n",
    "    fileName = 'RCI-venn-'+regulation.lower()+'.xlsx'\n",
    "    tempDF.to_excel(fileDir+ipynbName+fileName, sheet_name='README', header=True, index=False)\n",
    "    display(tempDF)#Check\n",
    "    \n",
    "    #Prepare saving data (statDF spread by dataset)\n",
    "    tempDF = moduleDF['ModuleName']#pd.Series() for now\n",
    "    for dataset in ['PrOmics', 'TrOmics']:\n",
    "        tempDF1 = statDF.loc[dataset].drop(columns=['ModuleName'])\n",
    "        tempDF1.columns = dataset+'_'+tempDF1.columns\n",
    "        tempDF = pd.merge(tempDF, tempDF1, left_index=True, right_index=True, how='left')\n",
    "    \n",
    "    t_start = time.time()\n",
    "    #Extract overall set\n",
    "    for key_i in range(len(tempD)):\n",
    "        key = list(tempD.keys())[key_i]\n",
    "        tempS = tempD[key]\n",
    "        tempDF1 = tempDF.loc[tempDF.index.isin(tempS)]\n",
    "        #Save summary table by appended to the above .xlsx file\n",
    "        ##Prepare sheet name\n",
    "        tempL1 = ['NA' for i in range(len(tempD))]\n",
    "        tempL1[key_i] = '1'\n",
    "        setName = '('+','.join(tempL1)+')'\n",
    "        with pd.ExcelWriter(fileDir+ipynbName+fileName, mode='a', engine='openpyxl') as writer:\n",
    "            tempDF1.to_excel(writer, sheet_name=setName, header=True, index=True)\n",
    "        print(' - '+setName+':', len(tempDF1))\n",
    "    \n",
    "    #Extract subset\n",
    "    tempL1 = ['1', '0']\n",
    "    tempL2 = [[k1, k2, k3, k4] for k1 in tempL1 for k2 in tempL1 for k3 in tempL1 for k4 in tempL1]\n",
    "    #tempL2.remove(['0', '0', '0', '0'])\n",
    "    for tempL1 in tempL2:\n",
    "        #Positive module set\n",
    "        tempL3 = [list(tempD.values())[key_i] for key_i, binary in enumerate(tempL1) if binary=='1']\n",
    "        tempS1 = set(pvalDF.index.tolist())#Initialize\n",
    "        for tempS in tempL3:\n",
    "            tempS1 = tempS1 & tempS\n",
    "        #Negative module set\n",
    "        tempL3 = [list(tempD.values())[key_i] for key_i, binary in enumerate(tempL1) if binary=='0']\n",
    "        tempS2 = set()#Initialize\n",
    "        for tempS in tempL3:\n",
    "            tempS2 = tempS2 | tempS\n",
    "        #Extract subset\n",
    "        tempS = tempS1 - tempS2\n",
    "        tempDF1 = tempDF.loc[tempDF.index.isin(tempS)]\n",
    "        #Save summary table by appended to the above .xlsx file\n",
    "        ##Prepare sheet name\n",
    "        setName = '('+','.join(tempL1)+')'\n",
    "        with pd.ExcelWriter(fileDir+ipynbName+fileName, mode='a', engine='openpyxl') as writer:\n",
    "            tempDF1.to_excel(writer, sheet_name=setName, header=True, index=True)\n",
    "        print(' - '+setName+':', len(tempDF1))\n",
    "    \n",
    "    t_elapsed = time.time() - t_start\n",
    "    print(' - Elapsed time:', round(t_elapsed//60), 'min', round(t_elapsed%60, 1), 'sec')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3-5. Visualization: pointplot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Spread statDF by template to flatten multi-index\n",
    "tempDF = moduleDF['ModuleName']#pd.Series() for now\n",
    "for dataset in ['PrOmics', 'TrOmics']:\n",
    "    tempDF1 = statDF.loc[dataset].drop(columns=['ModuleName'])\n",
    "    tempDF1.columns = dataset+'_'+tempDF1.columns\n",
    "    tempDF = pd.merge(tempDF, tempDF1, left_index=True, right_index=True, how='left')\n",
    "display(tempDF)\n",
    "\n",
    "statDF_flatten = tempDF"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 3-5-1. Modules tightened by all interventions and datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "posL = ['PrOmics_Aca-vs-Ctrl', 'PrOmics_Rapa-vs-Ctrl', 'TrOmics_Aca-vs-Ctrl', 'TrOmics_Rapa-vs-Ctrl']\n",
    "negL = ['']\n",
    "regulation = 'Tightened'\n",
    "nPlots = 10\n",
    "sort_varL = ['PrOmics_ANOVA_AdjPval', 'TrOmics_ANOVA_AdjPval']\n",
    "\n",
    "#Prepare the target module set\n",
    "tempS = pd.Series(np.repeat(True, len(pvalDF)), index=pvalDF.index)#Initialize\n",
    "tempDF = pvalDF.loc[:, pvalDF.columns.str.contains('-vs-')]#Based on adjusted ANOVA P-values\n",
    "for col_n in tempDF.columns.tolist():\n",
    "    tempS1 = pvalDF[col_n]\n",
    "    tempS2 = diffDF[col_n]\n",
    "    if col_n in posL:\n",
    "        if regulation=='Tightened':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2>0)\n",
    "        elif regulation=='Loosened':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2<0)\n",
    "    elif col_n in negL:\n",
    "        tempS3 = (tempS1>=0.05)\n",
    "        #Significance for inverse regulation\n",
    "        if regulation=='Tightened':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2<0)\n",
    "        elif regulation=='Loosened':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2>0)\n",
    "        tempS1 = tempS3 | tempS1\n",
    "    else:\n",
    "        tempS1 = (tempS1>=0.0)\n",
    "    #Update True\n",
    "    tempS = tempS & tempS1\n",
    "tempL = tempS.loc[tempS.tolist()].index.tolist()\n",
    "print(len(tempL), regulation.lower()+' modules with significance in', posL, 'but not in', negL)\n",
    "\n",
    "#Select representatives and sort the target modules\n",
    "tempDF = statDF_flatten.loc[:, statDF_flatten.columns.str.contains('Pval$')]\n",
    "tempDF = pd.merge(moduleDF['ModuleName'], tempDF, left_index=True, right_index=True, how='left')\n",
    "tempDF = tempDF.loc[tempL]\n",
    "tempDF = tempDF.sort_values(by=sort_varL, ascending=True)\n",
    "topX = np.min([30, len(tempL)])\n",
    "print('Top', topX, 'modules (sort by', sort_varL, '):')\n",
    "display(tempDF.iloc[:topX])\n",
    "plotL = tempDF.index.tolist()[:np.min([nPlots, len(tempL)])]\n",
    "\n",
    "#Prepare RMS DF for plot\n",
    "tempDF1 = rmsDF_kk.reset_index().melt(var_name='SampleID', value_name='RMS', id_vars='ModuleID')\n",
    "tempDF = sampleDF.reset_index()[['SampleID', 'Dataset', 'Intervention']]\n",
    "tempDF1 = pd.merge(tempDF1, tempDF, on='SampleID', how='left')\n",
    "\n",
    "#Prepare label and color\n",
    "tempD0 = {'Ctrl':'Control', 'Aca':'Acarbose', 'Rapa':'Rapamycin',\n",
    "          'PrOmics':'Proteins', 'TrOmics':'Transcripts'}\n",
    "tempDF1['Dataset'] = tempDF1['Dataset'].map(tempD0)\n",
    "tempDF1['Group'] = tempDF1['Intervention'].map(tempD0)\n",
    "tempD1 = {'Control':'tab:blue', 'Acarbose':'tab:red', 'Rapamycin':'tab:purple'}\n",
    "tempD2 = {'Proteins':plt.get_cmap('tab20')(13), 'Transcripts':plt.get_cmap('tab20')(19)}\n",
    "\n",
    "#Prepare P-value DF for plot\n",
    "tempDF2 = statDF_flatten.loc[:, statDF_flatten.columns.str.contains('-vs-.*_AdjPval$')]\n",
    "tempDF2.columns = tempDF2.columns.str.replace('_AdjPval$', '')\n",
    "\n",
    "#Visualize each module\n",
    "for rank_i in range(len(plotL)):\n",
    "    print(' - Rank '+str(rank_i+1)+' (sort by', sort_varL, '):')\n",
    "    module = plotL[rank_i]\n",
    "    #Check module summary\n",
    "    tempDF = pd.DataFrame(moduleDF.loc[module]).T\n",
    "    display(tempDF)\n",
    "    \n",
    "    #Select RMS\n",
    "    tempDF3 = tempDF1.loc[tempDF1['ModuleID']==module]\n",
    "    \n",
    "    #Check RMS summary\n",
    "    tempDF = tempDF3.groupby(['Dataset', 'Group'])['RMS'].agg(['count', 'mean', 'std'])\n",
    "    tempL1 = []\n",
    "    tempL2 = []\n",
    "    for row_n in tempDF.index.tolist():\n",
    "        count, mean, std = tempDF.loc[row_n]\n",
    "        tempL1.append(mean - 1.96*std/np.sqrt(count))\n",
    "        tempL2.append(mean + 1.96*std/np.sqrt(count))\n",
    "    tempDF['0.025'] = tempL1\n",
    "    tempDF['0.975'] = tempL2\n",
    "    ##Multiindex sort\n",
    "    tempDF = tempDF.reset_index()\n",
    "    tempDF['Dataset'] = pd.Categorical(tempDF['Dataset'], categories=list(tempD2.keys()))\n",
    "    tempDF['Group'] = pd.Categorical(tempDF['Group'], categories=list(tempD1.keys()))\n",
    "    tempDF = tempDF.sort_values(by=['Dataset', 'Group']).set_index(['Dataset', 'Group'])\n",
    "    display(tempDF)\n",
    "    \n",
    "    #Prepare significance labels\n",
    "    tempS = tempDF2.loc[module]\n",
    "    tempS.name = 'AdjPval'\n",
    "    ##Clean\n",
    "    tempDF = tempS.index.to_series().str.split(pat='_', expand=True)\n",
    "    tempDF = tempDF.rename(columns={0:'Dataset', 1:'Comparison'})\n",
    "    tempS1 = tempDF['Dataset']\n",
    "    tempDF = tempDF['Comparison'].str.split(pat='-vs-', expand=True)\n",
    "    tempDF = tempDF.rename(columns={0:'Contrast', 1:'Baseline'})\n",
    "    tempDF = pd.merge(tempS1, tempDF, left_index=True, right_index=True, how='left')\n",
    "    tempDF4 = pd.merge(tempDF, tempS, left_index=True, right_index=True, how='left')\n",
    "    tempDF4['Dataset'] = tempDF4['Dataset'].map(tempD0)\n",
    "    tempDF4['Contrast'] = tempDF4['Contrast'].map(tempD0)\n",
    "    tempDF4['Baseline'] = tempDF4['Baseline'].map(tempD0)\n",
    "    ##Convert p-value to label\n",
    "    tempL = []\n",
    "    for row_i in range(len(tempDF4)):\n",
    "        pval = tempDF4['AdjPval'].iloc[row_i]\n",
    "        if pval<0.001:\n",
    "            tempL.append('***')\n",
    "        elif pval<0.01:\n",
    "            tempL.append('**')\n",
    "        elif pval<0.05:\n",
    "            tempL.append('*')\n",
    "        else:\n",
    "            pval_text = Decimal(str(pval)).quantize(Decimal('0.01'), rounding=ROUND_HALF_UP)\n",
    "            tempL.append(r'$P$ = '+str(pval_text))\n",
    "    tempDF4['SignifLabel'] = tempL\n",
    "    display(tempDF4)\n",
    "    \n",
    "    #Visualization\n",
    "    ymax = 1.0\n",
    "    ymin = 0.0\n",
    "    yinter = 0.2\n",
    "    ymargin_t = 0.275\n",
    "    ymargin_b = 0.05\n",
    "    aline_ymin = 1.0\n",
    "    aline_ymargin = 0.125\n",
    "    sns.set(style='ticks', font='Arial', context='talk')\n",
    "    fig, axes = plt.subplots(nrows=1, ncols=2, figsize=(4.75, 5), sharex=True, sharey=True,\n",
    "                             gridspec_kw={'width_ratios':[2, 2]})\n",
    "    for ax_i, ax in enumerate(axes.flat):\n",
    "        dataset = list(tempD2.keys())[ax_i]\n",
    "        tempDF = tempDF3.loc[tempDF3['Dataset']==dataset]\n",
    "        sns.pointplot(data=tempDF, x='Group', y='RMS', order=list(tempD1.keys()), palette=tempD1,\n",
    "                      markers='o', dodge=False, join=False, capsize=0.6, estimator=np.mean, ci=95, ax=ax)\n",
    "        sns.stripplot(data=tempDF, x='Group', y='RMS',\n",
    "                      order=list(tempD1.keys()), palette=tempD1, dodge=False, jitter=0.15,\n",
    "                      size=5, edgecolor='black', linewidth=1, **{'marker':'o', 'alpha':0.5}, ax=ax)\n",
    "        #Set axis\n",
    "        sns.despine()\n",
    "        ax.set(ylim=(ymin-ymargin_b, ymax+ymargin_t), yticks=np.arange(ymin, ymax + yinter/10, yinter))\n",
    "        plt.setp(ax.get_xticklabels(), rotation=70, horizontalalignment='right',\n",
    "                 verticalalignment='center', rotation_mode='anchor')\n",
    "        if ax_i==0:\n",
    "            plt.setp(ax, xlabel='', ylabel='Sample RMS\\n(Mean = Module RCI)')\n",
    "        else:\n",
    "            plt.setp(ax.get_yticklabels(), visible=False)\n",
    "            plt.setp(ax, xlabel='', ylabel='')\n",
    "        #Add significance labels\n",
    "        tempDF = tempDF4.loc[tempDF4['Dataset']==dataset]\n",
    "        for row_i in range(len(tempDF)):\n",
    "            #Baseline\n",
    "            group_0 = tempDF['Baseline'].iloc[row_i]\n",
    "            index_0 = list(tempD1.keys()).index(group_0)\n",
    "            xcoord_0 = index_0\n",
    "            #Contrast\n",
    "            group_1 = tempDF['Contrast'].iloc[row_i]\n",
    "            index_1 = list(tempD1.keys()).index(group_1)\n",
    "            xcoord_1 = index_1\n",
    "            #Standard point of marker\n",
    "            xcoord = (xcoord_0+xcoord_1)/2\n",
    "            ycoord = aline_ymin + aline_ymargin*row_i\n",
    "            label = tempDF['SignifLabel'].iloc[row_i]\n",
    "            #Add annotation lines\n",
    "            aline_offset = yinter/5\n",
    "            aline_length = yinter/5 + aline_offset/2\n",
    "            ax.plot([xcoord_0, xcoord_0, xcoord_1, xcoord_1],\n",
    "                    [ycoord+aline_offset, ycoord+aline_length, ycoord+aline_length, ycoord+aline_offset],\n",
    "                    lw=1.5, c='k')\n",
    "            #Add annotation text\n",
    "            if label in ['***', '**', '*']:\n",
    "                text_offset = yinter/21\n",
    "                ax.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                            horizontalalignment='center', verticalalignment='bottom',\n",
    "                            fontsize='medium', color='k')\n",
    "            else:\n",
    "                text_offset = yinter/3.5\n",
    "                ax.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                            horizontalalignment='center', verticalalignment='bottom',\n",
    "                            fontsize='x-small', color='k')\n",
    "        #Add annotation\n",
    "        ax.set_title(dataset, {'fontsize':'small'})\n",
    "        xoff = 0.025\n",
    "        yoff = 0.01\n",
    "        rect = plt.Rectangle((xoff, 1), 1-xoff, 0.1,#Manual adjustment\n",
    "                             transform=ax.transAxes, facecolor=tempD2[dataset],\n",
    "                             clip_on=False, linewidth=0, zorder=0.5)\n",
    "        ax.add_patch(rect)\n",
    "        if ax_i==1:\n",
    "            at = AnchoredText('Consensus: own group', loc='center',\n",
    "                              bbox_to_anchor=(1.15, 0.505), bbox_transform=ax.transAxes,#Minor manual adjustment\n",
    "                              frameon=False, prop={'size':'small', 'rotation':90})\n",
    "            ax.add_artist(at)\n",
    "            rect = plt.Rectangle((1+xoff, yoff), 0.225, 1-2*yoff,#Manual adjustment\n",
    "                                 transform=ax.transAxes, facecolor=plt.get_cmap('tab20')(15),\n",
    "                                 clip_on=False, linewidth=0, zorder=0.5)\n",
    "            ax.add_patch(rect)\n",
    "    fig.tight_layout()\n",
    "    #Set title\n",
    "    modulename = moduleDF.loc[module, 'ModuleName']\n",
    "    initial = modulename[0].capitalize()\n",
    "    title = re.sub('^.', initial, modulename)+' ('+module+')'\n",
    "    title = '\\n'.join(wrap(title, 40))#Because the below wrap=True didn't work\n",
    "    fig.suptitle(title, size='small',\n",
    "                 verticalalignment='bottom', horizontalalignment='center', wrap=True, y=0.93)\n",
    "    ##Save\n",
    "    fileDir = './ExportFigures/'\n",
    "    ipynbName = '230224_LC-M001-PrOmics-vs-TrOmics-DIRAC-ver3_GOBP_'\n",
    "    fileName = 'RCI-pointplot-'+module.replace('GO:', 'GO')+'.pdf'\n",
    "    #plt.gcf().savefig(fileDir+ipynbName+fileName, dpi=300, bbox_inches='tight', pad_inches=0.04, transparent=True)\n",
    "    plt.show()\n",
    "    print('')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 3-5-2. Modules tightened by all interventions specifically in proteomics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "posL = ['PrOmics_Aca-vs-Ctrl', 'PrOmics_Rapa-vs-Ctrl']\n",
    "negL = ['TrOmics_Aca-vs-Ctrl', 'TrOmics_Rapa-vs-Ctrl']\n",
    "regulation = 'Tightened'\n",
    "nPlots = 10\n",
    "sort_varL = ['PrOmics_ANOVA_AdjPval', 'TrOmics_ANOVA_AdjPval']\n",
    "\n",
    "#Prepare the target module set\n",
    "tempS = pd.Series(np.repeat(True, len(pvalDF)), index=pvalDF.index)#Initialize\n",
    "tempDF = pvalDF.loc[:, pvalDF.columns.str.contains('-vs-')]#Based on adjusted ANOVA P-values\n",
    "for col_n in tempDF.columns.tolist():\n",
    "    tempS1 = pvalDF[col_n]\n",
    "    tempS2 = diffDF[col_n]\n",
    "    if col_n in posL:\n",
    "        if regulation=='Tightened':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2>0)\n",
    "        elif regulation=='Loosened':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2<0)\n",
    "    elif col_n in negL:\n",
    "        tempS3 = (tempS1>=0.05)\n",
    "        #Significance for inverse regulation\n",
    "        if regulation=='Tightened':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2<0)\n",
    "        elif regulation=='Loosened':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2>0)\n",
    "        tempS1 = tempS3 | tempS1\n",
    "    else:\n",
    "        tempS1 = (tempS1>=0.0)\n",
    "    #Update True\n",
    "    tempS = tempS & tempS1\n",
    "tempL = tempS.loc[tempS.tolist()].index.tolist()\n",
    "print(len(tempL), regulation.lower()+' modules with significance in', posL, 'but not in', negL)\n",
    "\n",
    "#Select representatives and sort the target modules\n",
    "tempDF = statDF_flatten.loc[:, statDF_flatten.columns.str.contains('Pval$')]\n",
    "tempDF = pd.merge(moduleDF['ModuleName'], tempDF, left_index=True, right_index=True, how='left')\n",
    "tempDF = tempDF.loc[tempL]\n",
    "tempDF = tempDF.sort_values(by=sort_varL, ascending=True)\n",
    "topX = np.min([30, len(tempL)])\n",
    "print('Top', topX, 'modules (sort by', sort_varL, '):')\n",
    "display(tempDF.iloc[:topX])\n",
    "plotL = tempDF.index.tolist()[:np.min([nPlots, len(tempL)])]\n",
    "\n",
    "#Prepare RMS DF for plot\n",
    "tempDF1 = rmsDF_kk.reset_index().melt(var_name='SampleID', value_name='RMS', id_vars='ModuleID')\n",
    "tempDF = sampleDF.reset_index()[['SampleID', 'Dataset', 'Intervention']]\n",
    "tempDF1 = pd.merge(tempDF1, tempDF, on='SampleID', how='left')\n",
    "\n",
    "#Prepare label and color\n",
    "tempD0 = {'Ctrl':'Control', 'Aca':'Acarbose', 'Rapa':'Rapamycin',\n",
    "          'PrOmics':'Proteins', 'TrOmics':'Transcripts'}\n",
    "tempDF1['Dataset'] = tempDF1['Dataset'].map(tempD0)\n",
    "tempDF1['Group'] = tempDF1['Intervention'].map(tempD0)\n",
    "tempD1 = {'Control':'tab:blue', 'Acarbose':'tab:red', 'Rapamycin':'tab:purple'}\n",
    "tempD2 = {'Proteins':plt.get_cmap('tab20')(13), 'Transcripts':plt.get_cmap('tab20')(19)}\n",
    "\n",
    "#Prepare P-value DF for plot\n",
    "tempDF2 = statDF_flatten.loc[:, statDF_flatten.columns.str.contains('-vs-.*_AdjPval$')]\n",
    "tempDF2.columns = tempDF2.columns.str.replace('_AdjPval$', '')\n",
    "\n",
    "#Visualize each module\n",
    "for rank_i in range(len(plotL)):\n",
    "    print(' - Rank '+str(rank_i+1)+' (sort by', sort_varL, '):')\n",
    "    module = plotL[rank_i]\n",
    "    #Check module summary\n",
    "    tempDF = pd.DataFrame(moduleDF.loc[module]).T\n",
    "    display(tempDF)\n",
    "    \n",
    "    #Select RMS\n",
    "    tempDF3 = tempDF1.loc[tempDF1['ModuleID']==module]\n",
    "    \n",
    "    #Check RMS summary\n",
    "    tempDF = tempDF3.groupby(['Dataset', 'Group'])['RMS'].agg(['count', 'mean', 'std'])\n",
    "    tempL1 = []\n",
    "    tempL2 = []\n",
    "    for row_n in tempDF.index.tolist():\n",
    "        count, mean, std = tempDF.loc[row_n]\n",
    "        tempL1.append(mean - 1.96*std/np.sqrt(count))\n",
    "        tempL2.append(mean + 1.96*std/np.sqrt(count))\n",
    "    tempDF['0.025'] = tempL1\n",
    "    tempDF['0.975'] = tempL2\n",
    "    ##Multiindex sort\n",
    "    tempDF = tempDF.reset_index()\n",
    "    tempDF['Dataset'] = pd.Categorical(tempDF['Dataset'], categories=list(tempD2.keys()))\n",
    "    tempDF['Group'] = pd.Categorical(tempDF['Group'], categories=list(tempD1.keys()))\n",
    "    tempDF = tempDF.sort_values(by=['Dataset', 'Group']).set_index(['Dataset', 'Group'])\n",
    "    display(tempDF)\n",
    "    \n",
    "    #Prepare significance labels\n",
    "    tempS = tempDF2.loc[module]\n",
    "    tempS.name = 'AdjPval'\n",
    "    ##Clean\n",
    "    tempDF = tempS.index.to_series().str.split(pat='_', expand=True)\n",
    "    tempDF = tempDF.rename(columns={0:'Dataset', 1:'Comparison'})\n",
    "    tempS1 = tempDF['Dataset']\n",
    "    tempDF = tempDF['Comparison'].str.split(pat='-vs-', expand=True)\n",
    "    tempDF = tempDF.rename(columns={0:'Contrast', 1:'Baseline'})\n",
    "    tempDF = pd.merge(tempS1, tempDF, left_index=True, right_index=True, how='left')\n",
    "    tempDF4 = pd.merge(tempDF, tempS, left_index=True, right_index=True, how='left')\n",
    "    tempDF4['Dataset'] = tempDF4['Dataset'].map(tempD0)\n",
    "    tempDF4['Contrast'] = tempDF4['Contrast'].map(tempD0)\n",
    "    tempDF4['Baseline'] = tempDF4['Baseline'].map(tempD0)\n",
    "    ##Convert p-value to label\n",
    "    tempL = []\n",
    "    for row_i in range(len(tempDF4)):\n",
    "        pval = tempDF4['AdjPval'].iloc[row_i]\n",
    "        if pval<0.001:\n",
    "            tempL.append('***')\n",
    "        elif pval<0.01:\n",
    "            tempL.append('**')\n",
    "        elif pval<0.05:\n",
    "            tempL.append('*')\n",
    "        else:\n",
    "            pval_text = Decimal(str(pval)).quantize(Decimal('0.01'), rounding=ROUND_HALF_UP)\n",
    "            tempL.append(r'$P$ = '+str(pval_text))\n",
    "    tempDF4['SignifLabel'] = tempL\n",
    "    display(tempDF4)\n",
    "    \n",
    "    #Visualization\n",
    "    ymax = 1.0\n",
    "    ymin = 0.0\n",
    "    yinter = 0.2\n",
    "    ymargin_t = 0.275\n",
    "    ymargin_b = 0.05\n",
    "    aline_ymin = 1.0\n",
    "    aline_ymargin = 0.125\n",
    "    sns.set(style='ticks', font='Arial', context='talk')\n",
    "    fig, axes = plt.subplots(nrows=1, ncols=2, figsize=(4.75, 5), sharex=True, sharey=True,\n",
    "                             gridspec_kw={'width_ratios':[2, 2]})\n",
    "    for ax_i, ax in enumerate(axes.flat):\n",
    "        dataset = list(tempD2.keys())[ax_i]\n",
    "        tempDF = tempDF3.loc[tempDF3['Dataset']==dataset]\n",
    "        sns.pointplot(data=tempDF, x='Group', y='RMS', order=list(tempD1.keys()), palette=tempD1,\n",
    "                      markers='o', dodge=False, join=False, capsize=0.6, estimator=np.mean, ci=95, ax=ax)\n",
    "        sns.stripplot(data=tempDF, x='Group', y='RMS',\n",
    "                      order=list(tempD1.keys()), palette=tempD1, dodge=False, jitter=0.15,\n",
    "                      size=5, edgecolor='black', linewidth=1, **{'marker':'o', 'alpha':0.5}, ax=ax)\n",
    "        #Set axis\n",
    "        sns.despine()\n",
    "        ax.set(ylim=(ymin-ymargin_b, ymax+ymargin_t), yticks=np.arange(ymin, ymax + yinter/10, yinter))\n",
    "        plt.setp(ax.get_xticklabels(), rotation=70, horizontalalignment='right',\n",
    "                 verticalalignment='center', rotation_mode='anchor')\n",
    "        if ax_i==0:\n",
    "            plt.setp(ax, xlabel='', ylabel='Sample RMS\\n(Mean = Module RCI)')\n",
    "        else:\n",
    "            plt.setp(ax.get_yticklabels(), visible=False)\n",
    "            plt.setp(ax, xlabel='', ylabel='')\n",
    "        #Add significance labels\n",
    "        tempDF = tempDF4.loc[tempDF4['Dataset']==dataset]\n",
    "        for row_i in range(len(tempDF)):\n",
    "            #Baseline\n",
    "            group_0 = tempDF['Baseline'].iloc[row_i]\n",
    "            index_0 = list(tempD1.keys()).index(group_0)\n",
    "            xcoord_0 = index_0\n",
    "            #Contrast\n",
    "            group_1 = tempDF['Contrast'].iloc[row_i]\n",
    "            index_1 = list(tempD1.keys()).index(group_1)\n",
    "            xcoord_1 = index_1\n",
    "            #Standard point of marker\n",
    "            xcoord = (xcoord_0+xcoord_1)/2\n",
    "            ycoord = aline_ymin + aline_ymargin*row_i\n",
    "            label = tempDF['SignifLabel'].iloc[row_i]\n",
    "            #Add annotation lines\n",
    "            aline_offset = yinter/5\n",
    "            aline_length = yinter/5 + aline_offset/2\n",
    "            ax.plot([xcoord_0, xcoord_0, xcoord_1, xcoord_1],\n",
    "                    [ycoord+aline_offset, ycoord+aline_length, ycoord+aline_length, ycoord+aline_offset],\n",
    "                    lw=1.5, c='k')\n",
    "            #Add annotation text\n",
    "            if label in ['***', '**', '*']:\n",
    "                text_offset = yinter/21\n",
    "                ax.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                            horizontalalignment='center', verticalalignment='bottom',\n",
    "                            fontsize='medium', color='k')\n",
    "            else:\n",
    "                text_offset = yinter/3.5\n",
    "                ax.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                            horizontalalignment='center', verticalalignment='bottom',\n",
    "                            fontsize='x-small', color='k')\n",
    "        #Add annotation\n",
    "        ax.set_title(dataset, {'fontsize':'small'})\n",
    "        xoff = 0.025\n",
    "        yoff = 0.01\n",
    "        rect = plt.Rectangle((xoff, 1), 1-xoff, 0.1,#Manual adjustment\n",
    "                             transform=ax.transAxes, facecolor=tempD2[dataset],\n",
    "                             clip_on=False, linewidth=0, zorder=0.5)\n",
    "        ax.add_patch(rect)\n",
    "        if ax_i==1:\n",
    "            at = AnchoredText('Consensus: own group', loc='center',\n",
    "                              bbox_to_anchor=(1.15, 0.505), bbox_transform=ax.transAxes,#Minor manual adjustment\n",
    "                              frameon=False, prop={'size':'small', 'rotation':90})\n",
    "            ax.add_artist(at)\n",
    "            rect = plt.Rectangle((1+xoff, yoff), 0.225, 1-2*yoff,#Manual adjustment\n",
    "                                 transform=ax.transAxes, facecolor=plt.get_cmap('tab20')(15),\n",
    "                                 clip_on=False, linewidth=0, zorder=0.5)\n",
    "            ax.add_patch(rect)\n",
    "    fig.tight_layout()\n",
    "    #Set title\n",
    "    modulename = moduleDF.loc[module, 'ModuleName']\n",
    "    initial = modulename[0].capitalize()\n",
    "    title = re.sub('^.', initial, modulename)+' ('+module+')'\n",
    "    title = '\\n'.join(wrap(title, 40))#Because the below wrap=True didn't work\n",
    "    fig.suptitle(title, size='small',\n",
    "                 verticalalignment='bottom', horizontalalignment='center', wrap=True, y=0.93)\n",
    "    ##Save\n",
    "    fileDir = './ExportFigures/'\n",
    "    ipynbName = '230224_LC-M001-PrOmics-vs-TrOmics-DIRAC-ver3_GOBP_'\n",
    "    fileName = 'RCI-pointplot-'+module.replace('GO:', 'GO')+'.pdf'\n",
    "    #plt.gcf().savefig(fileDir+ipynbName+fileName, dpi=300, bbox_inches='tight', pad_inches=0.04, transparent=True)\n",
    "    plt.show()\n",
    "    print('')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 3-5-3. Modules tightened by all interventions specifically in transcriptomics"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> Because proteins are the players for cellular functions, this module is not important even if positive, though.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "posL = ['TrOmics_Aca-vs-Ctrl', 'TrOmics_Rapa-vs-Ctrl']\n",
    "negL = ['PrOmics_Aca-vs-Ctrl', 'PrOmics_Rapa-vs-Ctrl']\n",
    "regulation = 'Tightened'\n",
    "nPlots = 10\n",
    "sort_varL = ['PrOmics_ANOVA_AdjPval', 'TrOmics_ANOVA_AdjPval']\n",
    "\n",
    "#Prepare the target module set\n",
    "tempS = pd.Series(np.repeat(True, len(pvalDF)), index=pvalDF.index)#Initialize\n",
    "tempDF = pvalDF.loc[:, pvalDF.columns.str.contains('-vs-')]#Based on adjusted ANOVA P-values\n",
    "for col_n in tempDF.columns.tolist():\n",
    "    tempS1 = pvalDF[col_n]\n",
    "    tempS2 = diffDF[col_n]\n",
    "    if col_n in posL:\n",
    "        if regulation=='Tightened':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2>0)\n",
    "        elif regulation=='Loosened':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2<0)\n",
    "    elif col_n in negL:\n",
    "        tempS3 = (tempS1>=0.05)\n",
    "        #Significance for inverse regulation\n",
    "        if regulation=='Tightened':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2<0)\n",
    "        elif regulation=='Loosened':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2>0)\n",
    "        tempS1 = tempS3 | tempS1\n",
    "    else:\n",
    "        tempS1 = (tempS1>=0.0)\n",
    "    #Update True\n",
    "    tempS = tempS & tempS1\n",
    "tempL = tempS.loc[tempS.tolist()].index.tolist()\n",
    "print(len(tempL), regulation.lower()+' modules with significance in', posL, 'but not in', negL)\n",
    "\n",
    "#Select representatives and sort the target modules\n",
    "tempDF = statDF_flatten.loc[:, statDF_flatten.columns.str.contains('Pval$')]\n",
    "tempDF = pd.merge(moduleDF['ModuleName'], tempDF, left_index=True, right_index=True, how='left')\n",
    "tempDF = tempDF.loc[tempL]\n",
    "tempDF = tempDF.sort_values(by=sort_varL, ascending=True)\n",
    "topX = np.min([30, len(tempL)])\n",
    "print('Top', topX, 'modules (sort by', sort_varL, '):')\n",
    "display(tempDF.iloc[:topX])\n",
    "plotL = tempDF.index.tolist()[:np.min([nPlots, len(tempL)])]\n",
    "\n",
    "#Prepare RMS DF for plot\n",
    "tempDF1 = rmsDF_kk.reset_index().melt(var_name='SampleID', value_name='RMS', id_vars='ModuleID')\n",
    "tempDF = sampleDF.reset_index()[['SampleID', 'Dataset', 'Intervention']]\n",
    "tempDF1 = pd.merge(tempDF1, tempDF, on='SampleID', how='left')\n",
    "\n",
    "#Prepare label and color\n",
    "tempD0 = {'Ctrl':'Control', 'Aca':'Acarbose', 'Rapa':'Rapamycin',\n",
    "          'PrOmics':'Proteins', 'TrOmics':'Transcripts'}\n",
    "tempDF1['Dataset'] = tempDF1['Dataset'].map(tempD0)\n",
    "tempDF1['Group'] = tempDF1['Intervention'].map(tempD0)\n",
    "tempD1 = {'Control':'tab:blue', 'Acarbose':'tab:red', 'Rapamycin':'tab:purple'}\n",
    "tempD2 = {'Proteins':plt.get_cmap('tab20')(13), 'Transcripts':plt.get_cmap('tab20')(19)}\n",
    "\n",
    "#Prepare P-value DF for plot\n",
    "tempDF2 = statDF_flatten.loc[:, statDF_flatten.columns.str.contains('-vs-.*_AdjPval$')]\n",
    "tempDF2.columns = tempDF2.columns.str.replace('_AdjPval$', '')\n",
    "\n",
    "#Visualize each module\n",
    "for rank_i in range(len(plotL)):\n",
    "    print(' - Rank '+str(rank_i+1)+' (sort by', sort_varL, '):')\n",
    "    module = plotL[rank_i]\n",
    "    #Check module summary\n",
    "    tempDF = pd.DataFrame(moduleDF.loc[module]).T\n",
    "    display(tempDF)\n",
    "    \n",
    "    #Select RMS\n",
    "    tempDF3 = tempDF1.loc[tempDF1['ModuleID']==module]\n",
    "    \n",
    "    #Check RMS summary\n",
    "    tempDF = tempDF3.groupby(['Dataset', 'Group'])['RMS'].agg(['count', 'mean', 'std'])\n",
    "    tempL1 = []\n",
    "    tempL2 = []\n",
    "    for row_n in tempDF.index.tolist():\n",
    "        count, mean, std = tempDF.loc[row_n]\n",
    "        tempL1.append(mean - 1.96*std/np.sqrt(count))\n",
    "        tempL2.append(mean + 1.96*std/np.sqrt(count))\n",
    "    tempDF['0.025'] = tempL1\n",
    "    tempDF['0.975'] = tempL2\n",
    "    ##Multiindex sort\n",
    "    tempDF = tempDF.reset_index()\n",
    "    tempDF['Dataset'] = pd.Categorical(tempDF['Dataset'], categories=list(tempD2.keys()))\n",
    "    tempDF['Group'] = pd.Categorical(tempDF['Group'], categories=list(tempD1.keys()))\n",
    "    tempDF = tempDF.sort_values(by=['Dataset', 'Group']).set_index(['Dataset', 'Group'])\n",
    "    display(tempDF)\n",
    "    \n",
    "    #Prepare significance labels\n",
    "    tempS = tempDF2.loc[module]\n",
    "    tempS.name = 'AdjPval'\n",
    "    ##Clean\n",
    "    tempDF = tempS.index.to_series().str.split(pat='_', expand=True)\n",
    "    tempDF = tempDF.rename(columns={0:'Dataset', 1:'Comparison'})\n",
    "    tempS1 = tempDF['Dataset']\n",
    "    tempDF = tempDF['Comparison'].str.split(pat='-vs-', expand=True)\n",
    "    tempDF = tempDF.rename(columns={0:'Contrast', 1:'Baseline'})\n",
    "    tempDF = pd.merge(tempS1, tempDF, left_index=True, right_index=True, how='left')\n",
    "    tempDF4 = pd.merge(tempDF, tempS, left_index=True, right_index=True, how='left')\n",
    "    tempDF4['Dataset'] = tempDF4['Dataset'].map(tempD0)\n",
    "    tempDF4['Contrast'] = tempDF4['Contrast'].map(tempD0)\n",
    "    tempDF4['Baseline'] = tempDF4['Baseline'].map(tempD0)\n",
    "    ##Convert p-value to label\n",
    "    tempL = []\n",
    "    for row_i in range(len(tempDF4)):\n",
    "        pval = tempDF4['AdjPval'].iloc[row_i]\n",
    "        if pval<0.001:\n",
    "            tempL.append('***')\n",
    "        elif pval<0.01:\n",
    "            tempL.append('**')\n",
    "        elif pval<0.05:\n",
    "            tempL.append('*')\n",
    "        else:\n",
    "            pval_text = Decimal(str(pval)).quantize(Decimal('0.01'), rounding=ROUND_HALF_UP)\n",
    "            tempL.append(r'$P$ = '+str(pval_text))\n",
    "    tempDF4['SignifLabel'] = tempL\n",
    "    display(tempDF4)\n",
    "    \n",
    "    #Visualization\n",
    "    ymax = 1.0\n",
    "    ymin = 0.0\n",
    "    yinter = 0.2\n",
    "    ymargin_t = 0.275\n",
    "    ymargin_b = 0.05\n",
    "    aline_ymin = 1.0\n",
    "    aline_ymargin = 0.125\n",
    "    sns.set(style='ticks', font='Arial', context='talk')\n",
    "    fig, axes = plt.subplots(nrows=1, ncols=2, figsize=(4.75, 5), sharex=True, sharey=True,\n",
    "                             gridspec_kw={'width_ratios':[2, 2]})\n",
    "    for ax_i, ax in enumerate(axes.flat):\n",
    "        dataset = list(tempD2.keys())[ax_i]\n",
    "        tempDF = tempDF3.loc[tempDF3['Dataset']==dataset]\n",
    "        sns.pointplot(data=tempDF, x='Group', y='RMS', order=list(tempD1.keys()), palette=tempD1,\n",
    "                      markers='o', dodge=False, join=False, capsize=0.6, estimator=np.mean, ci=95, ax=ax)\n",
    "        sns.stripplot(data=tempDF, x='Group', y='RMS',\n",
    "                      order=list(tempD1.keys()), palette=tempD1, dodge=False, jitter=0.15,\n",
    "                      size=5, edgecolor='black', linewidth=1, **{'marker':'o', 'alpha':0.5}, ax=ax)\n",
    "        #Set axis\n",
    "        sns.despine()\n",
    "        ax.set(ylim=(ymin-ymargin_b, ymax+ymargin_t), yticks=np.arange(ymin, ymax + yinter/10, yinter))\n",
    "        plt.setp(ax.get_xticklabels(), rotation=70, horizontalalignment='right',\n",
    "                 verticalalignment='center', rotation_mode='anchor')\n",
    "        if ax_i==0:\n",
    "            plt.setp(ax, xlabel='', ylabel='Sample RMS\\n(Mean = Module RCI)')\n",
    "        else:\n",
    "            plt.setp(ax.get_yticklabels(), visible=False)\n",
    "            plt.setp(ax, xlabel='', ylabel='')\n",
    "        #Add significance labels\n",
    "        tempDF = tempDF4.loc[tempDF4['Dataset']==dataset]\n",
    "        for row_i in range(len(tempDF)):\n",
    "            #Baseline\n",
    "            group_0 = tempDF['Baseline'].iloc[row_i]\n",
    "            index_0 = list(tempD1.keys()).index(group_0)\n",
    "            xcoord_0 = index_0\n",
    "            #Contrast\n",
    "            group_1 = tempDF['Contrast'].iloc[row_i]\n",
    "            index_1 = list(tempD1.keys()).index(group_1)\n",
    "            xcoord_1 = index_1\n",
    "            #Standard point of marker\n",
    "            xcoord = (xcoord_0+xcoord_1)/2\n",
    "            ycoord = aline_ymin + aline_ymargin*row_i\n",
    "            label = tempDF['SignifLabel'].iloc[row_i]\n",
    "            #Add annotation lines\n",
    "            aline_offset = yinter/5\n",
    "            aline_length = yinter/5 + aline_offset/2\n",
    "            ax.plot([xcoord_0, xcoord_0, xcoord_1, xcoord_1],\n",
    "                    [ycoord+aline_offset, ycoord+aline_length, ycoord+aline_length, ycoord+aline_offset],\n",
    "                    lw=1.5, c='k')\n",
    "            #Add annotation text\n",
    "            if label in ['***', '**', '*']:\n",
    "                text_offset = yinter/21\n",
    "                ax.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                            horizontalalignment='center', verticalalignment='bottom',\n",
    "                            fontsize='medium', color='k')\n",
    "            else:\n",
    "                text_offset = yinter/3.5\n",
    "                ax.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                            horizontalalignment='center', verticalalignment='bottom',\n",
    "                            fontsize='x-small', color='k')\n",
    "        #Add annotation\n",
    "        ax.set_title(dataset, {'fontsize':'small'})\n",
    "        xoff = 0.025\n",
    "        yoff = 0.01\n",
    "        rect = plt.Rectangle((xoff, 1), 1-xoff, 0.1,#Manual adjustment\n",
    "                             transform=ax.transAxes, facecolor=tempD2[dataset],\n",
    "                             clip_on=False, linewidth=0, zorder=0.5)\n",
    "        ax.add_patch(rect)\n",
    "        if ax_i==1:\n",
    "            at = AnchoredText('Consensus: own group', loc='center',\n",
    "                              bbox_to_anchor=(1.15, 0.505), bbox_transform=ax.transAxes,#Minor manual adjustment\n",
    "                              frameon=False, prop={'size':'small', 'rotation':90})\n",
    "            ax.add_artist(at)\n",
    "            rect = plt.Rectangle((1+xoff, yoff), 0.225, 1-2*yoff,#Manual adjustment\n",
    "                                 transform=ax.transAxes, facecolor=plt.get_cmap('tab20')(15),\n",
    "                                 clip_on=False, linewidth=0, zorder=0.5)\n",
    "            ax.add_patch(rect)\n",
    "    fig.tight_layout()\n",
    "    #Set title\n",
    "    modulename = moduleDF.loc[module, 'ModuleName']\n",
    "    initial = modulename[0].capitalize()\n",
    "    title = re.sub('^.', initial, modulename)+' ('+module+')'\n",
    "    title = '\\n'.join(wrap(title, 40))#Because the below wrap=True didn't work\n",
    "    fig.suptitle(title, size='small',\n",
    "                 verticalalignment='bottom', horizontalalignment='center', wrap=True, y=0.93)\n",
    "    ##Save\n",
    "    fileDir = './ExportFigures/'\n",
    "    ipynbName = '230224_LC-M001-PrOmics-vs-TrOmics-DIRAC-ver3_GOBP_'\n",
    "    fileName = 'RCI-pointplot-'+module.replace('GO:', 'GO')+'.pdf'\n",
    "    #plt.gcf().savefig(fileDir+ipynbName+fileName, dpi=300, bbox_inches='tight', pad_inches=0.04, transparent=True)\n",
    "    plt.show()\n",
    "    print('')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 3-5-4. Modules tightened by Aca in both datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "posL = ['PrOmics_Aca-vs-Ctrl', 'TrOmics_Aca-vs-Ctrl']\n",
    "negL = ['']\n",
    "regulation = 'Tightened'\n",
    "nPlots = 3\n",
    "sort_varL = [col_prefix+'_AdjPval' for col_prefix in posL]\n",
    "\n",
    "#Prepare the target module set\n",
    "tempS = pd.Series(np.repeat(True, len(pvalDF)), index=pvalDF.index)#Initialize\n",
    "tempDF = pvalDF.loc[:, pvalDF.columns.str.contains('-vs-')]#Based on adjusted ANOVA P-values\n",
    "for col_n in tempDF.columns.tolist():\n",
    "    tempS1 = pvalDF[col_n]\n",
    "    tempS2 = diffDF[col_n]\n",
    "    if col_n in posL:\n",
    "        if regulation=='Tightened':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2>0)\n",
    "        elif regulation=='Loosened':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2<0)\n",
    "    elif col_n in negL:\n",
    "        tempS3 = (tempS1>=0.05)\n",
    "        #Significance for inverse regulation\n",
    "        if regulation=='Tightened':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2<0)\n",
    "        elif regulation=='Loosened':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2>0)\n",
    "        tempS1 = tempS3 | tempS1\n",
    "    else:\n",
    "        tempS1 = (tempS1>=0.0)\n",
    "    #Update True\n",
    "    tempS = tempS & tempS1\n",
    "tempL = tempS.loc[tempS.tolist()].index.tolist()\n",
    "print(len(tempL), regulation.lower()+' modules with significance in', posL, 'but not in', negL)\n",
    "\n",
    "#Select representatives and sort the target modules\n",
    "tempDF = statDF_flatten.loc[:, statDF_flatten.columns.str.contains('Pval$')]\n",
    "tempDF = pd.merge(moduleDF['ModuleName'], tempDF, left_index=True, right_index=True, how='left')\n",
    "tempDF = tempDF.loc[tempL]\n",
    "tempDF = tempDF.sort_values(by=sort_varL, ascending=True)\n",
    "topX = np.min([30, len(tempL)])\n",
    "print('Top', topX, 'modules (sort by', sort_varL, '):')\n",
    "display(tempDF.iloc[:topX])\n",
    "plotL = tempDF.index.tolist()[:np.min([nPlots, len(tempL)])]\n",
    "\n",
    "#Prepare RMS DF for plot\n",
    "tempDF1 = rmsDF_kk.reset_index().melt(var_name='SampleID', value_name='RMS', id_vars='ModuleID')\n",
    "tempDF = sampleDF.reset_index()[['SampleID', 'Dataset', 'Intervention']]\n",
    "tempDF1 = pd.merge(tempDF1, tempDF, on='SampleID', how='left')\n",
    "\n",
    "#Prepare label and color\n",
    "tempD0 = {'Ctrl':'Control', 'Aca':'Acarbose', 'Rapa':'Rapamycin',\n",
    "          'PrOmics':'Proteins', 'TrOmics':'Transcripts'}\n",
    "tempDF1['Dataset'] = tempDF1['Dataset'].map(tempD0)\n",
    "tempDF1['Group'] = tempDF1['Intervention'].map(tempD0)\n",
    "tempD1 = {'Control':'tab:blue', 'Acarbose':'tab:red', 'Rapamycin':'tab:purple'}\n",
    "tempD2 = {'Proteins':plt.get_cmap('tab20')(13), 'Transcripts':plt.get_cmap('tab20')(19)}\n",
    "\n",
    "#Prepare P-value DF for plot\n",
    "tempDF2 = statDF_flatten.loc[:, statDF_flatten.columns.str.contains('-vs-.*_AdjPval$')]\n",
    "tempDF2.columns = tempDF2.columns.str.replace('_AdjPval$', '')\n",
    "\n",
    "#Visualize each module\n",
    "for rank_i in range(len(plotL)):\n",
    "    print(' - Rank '+str(rank_i+1)+' (sort by', sort_varL, '):')\n",
    "    module = plotL[rank_i]\n",
    "    #Check module summary\n",
    "    tempDF = pd.DataFrame(moduleDF.loc[module]).T\n",
    "    display(tempDF)\n",
    "    \n",
    "    #Select RMS\n",
    "    tempDF3 = tempDF1.loc[tempDF1['ModuleID']==module]\n",
    "    \n",
    "    #Check RMS summary\n",
    "    tempDF = tempDF3.groupby(['Dataset', 'Group'])['RMS'].agg(['count', 'mean', 'std'])\n",
    "    tempL1 = []\n",
    "    tempL2 = []\n",
    "    for row_n in tempDF.index.tolist():\n",
    "        count, mean, std = tempDF.loc[row_n]\n",
    "        tempL1.append(mean - 1.96*std/np.sqrt(count))\n",
    "        tempL2.append(mean + 1.96*std/np.sqrt(count))\n",
    "    tempDF['0.025'] = tempL1\n",
    "    tempDF['0.975'] = tempL2\n",
    "    ##Multiindex sort\n",
    "    tempDF = tempDF.reset_index()\n",
    "    tempDF['Dataset'] = pd.Categorical(tempDF['Dataset'], categories=list(tempD2.keys()))\n",
    "    tempDF['Group'] = pd.Categorical(tempDF['Group'], categories=list(tempD1.keys()))\n",
    "    tempDF = tempDF.sort_values(by=['Dataset', 'Group']).set_index(['Dataset', 'Group'])\n",
    "    display(tempDF)\n",
    "    \n",
    "    #Prepare significance labels\n",
    "    tempS = tempDF2.loc[module]\n",
    "    tempS.name = 'AdjPval'\n",
    "    ##Clean\n",
    "    tempDF = tempS.index.to_series().str.split(pat='_', expand=True)\n",
    "    tempDF = tempDF.rename(columns={0:'Dataset', 1:'Comparison'})\n",
    "    tempS1 = tempDF['Dataset']\n",
    "    tempDF = tempDF['Comparison'].str.split(pat='-vs-', expand=True)\n",
    "    tempDF = tempDF.rename(columns={0:'Contrast', 1:'Baseline'})\n",
    "    tempDF = pd.merge(tempS1, tempDF, left_index=True, right_index=True, how='left')\n",
    "    tempDF4 = pd.merge(tempDF, tempS, left_index=True, right_index=True, how='left')\n",
    "    tempDF4['Dataset'] = tempDF4['Dataset'].map(tempD0)\n",
    "    tempDF4['Contrast'] = tempDF4['Contrast'].map(tempD0)\n",
    "    tempDF4['Baseline'] = tempDF4['Baseline'].map(tempD0)\n",
    "    ##Convert p-value to label\n",
    "    tempL = []\n",
    "    for row_i in range(len(tempDF4)):\n",
    "        pval = tempDF4['AdjPval'].iloc[row_i]\n",
    "        if pval<0.001:\n",
    "            tempL.append('***')\n",
    "        elif pval<0.01:\n",
    "            tempL.append('**')\n",
    "        elif pval<0.05:\n",
    "            tempL.append('*')\n",
    "        else:\n",
    "            pval_text = Decimal(str(pval)).quantize(Decimal('0.01'), rounding=ROUND_HALF_UP)\n",
    "            tempL.append(r'$P$ = '+str(pval_text))\n",
    "    tempDF4['SignifLabel'] = tempL\n",
    "    display(tempDF4)\n",
    "    \n",
    "    #Visualization\n",
    "    ymax = 1.0\n",
    "    ymin = 0.0\n",
    "    yinter = 0.2\n",
    "    ymargin_t = 0.275\n",
    "    ymargin_b = 0.05\n",
    "    aline_ymin = 1.0\n",
    "    aline_ymargin = 0.125\n",
    "    sns.set(style='ticks', font='Arial', context='talk')\n",
    "    fig, axes = plt.subplots(nrows=1, ncols=2, figsize=(4.75, 5), sharex=True, sharey=True,\n",
    "                             gridspec_kw={'width_ratios':[2, 2]})\n",
    "    for ax_i, ax in enumerate(axes.flat):\n",
    "        dataset = list(tempD2.keys())[ax_i]\n",
    "        tempDF = tempDF3.loc[tempDF3['Dataset']==dataset]\n",
    "        sns.pointplot(data=tempDF, x='Group', y='RMS', order=list(tempD1.keys()), palette=tempD1,\n",
    "                      markers='o', dodge=False, join=False, capsize=0.6, estimator=np.mean, ci=95, ax=ax)\n",
    "        sns.stripplot(data=tempDF, x='Group', y='RMS',\n",
    "                      order=list(tempD1.keys()), palette=tempD1, dodge=False, jitter=0.15,\n",
    "                      size=5, edgecolor='black', linewidth=1, **{'marker':'o', 'alpha':0.5}, ax=ax)\n",
    "        #Set axis\n",
    "        sns.despine()\n",
    "        ax.set(ylim=(ymin-ymargin_b, ymax+ymargin_t), yticks=np.arange(ymin, ymax + yinter/10, yinter))\n",
    "        plt.setp(ax.get_xticklabels(), rotation=70, horizontalalignment='right',\n",
    "                 verticalalignment='center', rotation_mode='anchor')\n",
    "        if ax_i==0:\n",
    "            plt.setp(ax, xlabel='', ylabel='Sample RMS\\n(Mean = Module RCI)')\n",
    "        else:\n",
    "            plt.setp(ax.get_yticklabels(), visible=False)\n",
    "            plt.setp(ax, xlabel='', ylabel='')\n",
    "        #Add significance labels\n",
    "        tempDF = tempDF4.loc[tempDF4['Dataset']==dataset]\n",
    "        for row_i in range(len(tempDF)):\n",
    "            #Baseline\n",
    "            group_0 = tempDF['Baseline'].iloc[row_i]\n",
    "            index_0 = list(tempD1.keys()).index(group_0)\n",
    "            xcoord_0 = index_0\n",
    "            #Contrast\n",
    "            group_1 = tempDF['Contrast'].iloc[row_i]\n",
    "            index_1 = list(tempD1.keys()).index(group_1)\n",
    "            xcoord_1 = index_1\n",
    "            #Standard point of marker\n",
    "            xcoord = (xcoord_0+xcoord_1)/2\n",
    "            ycoord = aline_ymin + aline_ymargin*row_i\n",
    "            label = tempDF['SignifLabel'].iloc[row_i]\n",
    "            #Add annotation lines\n",
    "            aline_offset = yinter/5\n",
    "            aline_length = yinter/5 + aline_offset/2\n",
    "            ax.plot([xcoord_0, xcoord_0, xcoord_1, xcoord_1],\n",
    "                    [ycoord+aline_offset, ycoord+aline_length, ycoord+aline_length, ycoord+aline_offset],\n",
    "                    lw=1.5, c='k')\n",
    "            #Add annotation text\n",
    "            if label in ['***', '**', '*']:\n",
    "                text_offset = yinter/21\n",
    "                ax.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                            horizontalalignment='center', verticalalignment='bottom',\n",
    "                            fontsize='medium', color='k')\n",
    "            else:\n",
    "                text_offset = yinter/3.5\n",
    "                ax.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                            horizontalalignment='center', verticalalignment='bottom',\n",
    "                            fontsize='x-small', color='k')\n",
    "        #Add annotation\n",
    "        ax.set_title(dataset, {'fontsize':'small'})\n",
    "        xoff = 0.025\n",
    "        yoff = 0.01\n",
    "        rect = plt.Rectangle((xoff, 1), 1-xoff, 0.1,#Manual adjustment\n",
    "                             transform=ax.transAxes, facecolor=tempD2[dataset],\n",
    "                             clip_on=False, linewidth=0, zorder=0.5)\n",
    "        ax.add_patch(rect)\n",
    "        if ax_i==1:\n",
    "            at = AnchoredText('Consensus: own group', loc='center',\n",
    "                              bbox_to_anchor=(1.15, 0.505), bbox_transform=ax.transAxes,#Minor manual adjustment\n",
    "                              frameon=False, prop={'size':'small', 'rotation':90})\n",
    "            ax.add_artist(at)\n",
    "            rect = plt.Rectangle((1+xoff, yoff), 0.225, 1-2*yoff,#Manual adjustment\n",
    "                                 transform=ax.transAxes, facecolor=plt.get_cmap('tab20')(15),\n",
    "                                 clip_on=False, linewidth=0, zorder=0.5)\n",
    "            ax.add_patch(rect)\n",
    "    fig.tight_layout()\n",
    "    #Set title\n",
    "    modulename = moduleDF.loc[module, 'ModuleName']\n",
    "    initial = modulename[0].capitalize()\n",
    "    title = re.sub('^.', initial, modulename)+' ('+module+')'\n",
    "    title = '\\n'.join(wrap(title, 40))#Because the below wrap=True didn't work\n",
    "    fig.suptitle(title, size='small',\n",
    "                 verticalalignment='bottom', horizontalalignment='center', wrap=True, y=0.93)\n",
    "    ##Save\n",
    "    fileDir = './ExportFigures/'\n",
    "    ipynbName = '230224_LC-M001-PrOmics-vs-TrOmics-DIRAC-ver3_GOBP_'\n",
    "    fileName = 'RCI-pointplot-'+module.replace('GO:', 'GO')+'.pdf'\n",
    "    #plt.gcf().savefig(fileDir+ipynbName+fileName, dpi=300, bbox_inches='tight', pad_inches=0.04, transparent=True)\n",
    "    plt.show()\n",
    "    print('')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 3-5-5. Modules tightened by Aca specifically in proteomics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "posL = ['PrOmics_Aca-vs-Ctrl']\n",
    "negL = ['TrOmics_Aca-vs-Ctrl']\n",
    "regulation = 'Tightened'\n",
    "nPlots = 3\n",
    "sort_varL = [col_prefix+'_AdjPval' for col_prefix in posL]\n",
    "\n",
    "#Prepare the target module set\n",
    "tempS = pd.Series(np.repeat(True, len(pvalDF)), index=pvalDF.index)#Initialize\n",
    "tempDF = pvalDF.loc[:, pvalDF.columns.str.contains('-vs-')]#Based on adjusted ANOVA P-values\n",
    "for col_n in tempDF.columns.tolist():\n",
    "    tempS1 = pvalDF[col_n]\n",
    "    tempS2 = diffDF[col_n]\n",
    "    if col_n in posL:\n",
    "        if regulation=='Tightened':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2>0)\n",
    "        elif regulation=='Loosened':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2<0)\n",
    "    elif col_n in negL:\n",
    "        tempS3 = (tempS1>=0.05)\n",
    "        #Significance for inverse regulation\n",
    "        if regulation=='Tightened':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2<0)\n",
    "        elif regulation=='Loosened':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2>0)\n",
    "        tempS1 = tempS3 | tempS1\n",
    "    else:\n",
    "        tempS1 = (tempS1>=0.0)\n",
    "    #Update True\n",
    "    tempS = tempS & tempS1\n",
    "tempL = tempS.loc[tempS.tolist()].index.tolist()\n",
    "print(len(tempL), regulation.lower()+' modules with significance in', posL, 'but not in', negL)\n",
    "\n",
    "#Select representatives and sort the target modules\n",
    "tempDF = statDF_flatten.loc[:, statDF_flatten.columns.str.contains('Pval$')]\n",
    "tempDF = pd.merge(moduleDF['ModuleName'], tempDF, left_index=True, right_index=True, how='left')\n",
    "tempDF = tempDF.loc[tempL]\n",
    "tempDF = tempDF.sort_values(by=sort_varL, ascending=True)\n",
    "topX = np.min([30, len(tempL)])\n",
    "print('Top', topX, 'modules (sort by', sort_varL, '):')\n",
    "display(tempDF.iloc[:topX])\n",
    "plotL = tempDF.index.tolist()[:np.min([nPlots, len(tempL)])]\n",
    "\n",
    "#Prepare RMS DF for plot\n",
    "tempDF1 = rmsDF_kk.reset_index().melt(var_name='SampleID', value_name='RMS', id_vars='ModuleID')\n",
    "tempDF = sampleDF.reset_index()[['SampleID', 'Dataset', 'Intervention']]\n",
    "tempDF1 = pd.merge(tempDF1, tempDF, on='SampleID', how='left')\n",
    "\n",
    "#Prepare label and color\n",
    "tempD0 = {'Ctrl':'Control', 'Aca':'Acarbose', 'Rapa':'Rapamycin',\n",
    "          'PrOmics':'Proteins', 'TrOmics':'Transcripts'}\n",
    "tempDF1['Dataset'] = tempDF1['Dataset'].map(tempD0)\n",
    "tempDF1['Group'] = tempDF1['Intervention'].map(tempD0)\n",
    "tempD1 = {'Control':'tab:blue', 'Acarbose':'tab:red', 'Rapamycin':'tab:purple'}\n",
    "tempD2 = {'Proteins':plt.get_cmap('tab20')(13), 'Transcripts':plt.get_cmap('tab20')(19)}\n",
    "\n",
    "#Prepare P-value DF for plot\n",
    "tempDF2 = statDF_flatten.loc[:, statDF_flatten.columns.str.contains('-vs-.*_AdjPval$')]\n",
    "tempDF2.columns = tempDF2.columns.str.replace('_AdjPval$', '')\n",
    "\n",
    "#Visualize each module\n",
    "for rank_i in range(len(plotL)):\n",
    "    print(' - Rank '+str(rank_i+1)+' (sort by', sort_varL, '):')\n",
    "    module = plotL[rank_i]\n",
    "    #Check module summary\n",
    "    tempDF = pd.DataFrame(moduleDF.loc[module]).T\n",
    "    display(tempDF)\n",
    "    \n",
    "    #Select RMS\n",
    "    tempDF3 = tempDF1.loc[tempDF1['ModuleID']==module]\n",
    "    \n",
    "    #Check RMS summary\n",
    "    tempDF = tempDF3.groupby(['Dataset', 'Group'])['RMS'].agg(['count', 'mean', 'std'])\n",
    "    tempL1 = []\n",
    "    tempL2 = []\n",
    "    for row_n in tempDF.index.tolist():\n",
    "        count, mean, std = tempDF.loc[row_n]\n",
    "        tempL1.append(mean - 1.96*std/np.sqrt(count))\n",
    "        tempL2.append(mean + 1.96*std/np.sqrt(count))\n",
    "    tempDF['0.025'] = tempL1\n",
    "    tempDF['0.975'] = tempL2\n",
    "    ##Multiindex sort\n",
    "    tempDF = tempDF.reset_index()\n",
    "    tempDF['Dataset'] = pd.Categorical(tempDF['Dataset'], categories=list(tempD2.keys()))\n",
    "    tempDF['Group'] = pd.Categorical(tempDF['Group'], categories=list(tempD1.keys()))\n",
    "    tempDF = tempDF.sort_values(by=['Dataset', 'Group']).set_index(['Dataset', 'Group'])\n",
    "    display(tempDF)\n",
    "    \n",
    "    #Prepare significance labels\n",
    "    tempS = tempDF2.loc[module]\n",
    "    tempS.name = 'AdjPval'\n",
    "    ##Clean\n",
    "    tempDF = tempS.index.to_series().str.split(pat='_', expand=True)\n",
    "    tempDF = tempDF.rename(columns={0:'Dataset', 1:'Comparison'})\n",
    "    tempS1 = tempDF['Dataset']\n",
    "    tempDF = tempDF['Comparison'].str.split(pat='-vs-', expand=True)\n",
    "    tempDF = tempDF.rename(columns={0:'Contrast', 1:'Baseline'})\n",
    "    tempDF = pd.merge(tempS1, tempDF, left_index=True, right_index=True, how='left')\n",
    "    tempDF4 = pd.merge(tempDF, tempS, left_index=True, right_index=True, how='left')\n",
    "    tempDF4['Dataset'] = tempDF4['Dataset'].map(tempD0)\n",
    "    tempDF4['Contrast'] = tempDF4['Contrast'].map(tempD0)\n",
    "    tempDF4['Baseline'] = tempDF4['Baseline'].map(tempD0)\n",
    "    ##Convert p-value to label\n",
    "    tempL = []\n",
    "    for row_i in range(len(tempDF4)):\n",
    "        pval = tempDF4['AdjPval'].iloc[row_i]\n",
    "        if pval<0.001:\n",
    "            tempL.append('***')\n",
    "        elif pval<0.01:\n",
    "            tempL.append('**')\n",
    "        elif pval<0.05:\n",
    "            tempL.append('*')\n",
    "        else:\n",
    "            pval_text = Decimal(str(pval)).quantize(Decimal('0.01'), rounding=ROUND_HALF_UP)\n",
    "            tempL.append(r'$P$ = '+str(pval_text))\n",
    "    tempDF4['SignifLabel'] = tempL\n",
    "    display(tempDF4)\n",
    "    \n",
    "    #Visualization\n",
    "    ymax = 1.0\n",
    "    ymin = 0.0\n",
    "    yinter = 0.2\n",
    "    ymargin_t = 0.275\n",
    "    ymargin_b = 0.05\n",
    "    aline_ymin = 1.0\n",
    "    aline_ymargin = 0.125\n",
    "    sns.set(style='ticks', font='Arial', context='talk')\n",
    "    fig, axes = plt.subplots(nrows=1, ncols=2, figsize=(4.75, 5), sharex=True, sharey=True,\n",
    "                             gridspec_kw={'width_ratios':[2, 2]})\n",
    "    for ax_i, ax in enumerate(axes.flat):\n",
    "        dataset = list(tempD2.keys())[ax_i]\n",
    "        tempDF = tempDF3.loc[tempDF3['Dataset']==dataset]\n",
    "        sns.pointplot(data=tempDF, x='Group', y='RMS', order=list(tempD1.keys()), palette=tempD1,\n",
    "                      markers='o', dodge=False, join=False, capsize=0.6, estimator=np.mean, ci=95, ax=ax)\n",
    "        sns.stripplot(data=tempDF, x='Group', y='RMS',\n",
    "                      order=list(tempD1.keys()), palette=tempD1, dodge=False, jitter=0.15,\n",
    "                      size=5, edgecolor='black', linewidth=1, **{'marker':'o', 'alpha':0.5}, ax=ax)\n",
    "        #Set axis\n",
    "        sns.despine()\n",
    "        ax.set(ylim=(ymin-ymargin_b, ymax+ymargin_t), yticks=np.arange(ymin, ymax + yinter/10, yinter))\n",
    "        plt.setp(ax.get_xticklabels(), rotation=70, horizontalalignment='right',\n",
    "                 verticalalignment='center', rotation_mode='anchor')\n",
    "        if ax_i==0:\n",
    "            plt.setp(ax, xlabel='', ylabel='Sample RMS\\n(Mean = Module RCI)')\n",
    "        else:\n",
    "            plt.setp(ax.get_yticklabels(), visible=False)\n",
    "            plt.setp(ax, xlabel='', ylabel='')\n",
    "        #Add significance labels\n",
    "        tempDF = tempDF4.loc[tempDF4['Dataset']==dataset]\n",
    "        for row_i in range(len(tempDF)):\n",
    "            #Baseline\n",
    "            group_0 = tempDF['Baseline'].iloc[row_i]\n",
    "            index_0 = list(tempD1.keys()).index(group_0)\n",
    "            xcoord_0 = index_0\n",
    "            #Contrast\n",
    "            group_1 = tempDF['Contrast'].iloc[row_i]\n",
    "            index_1 = list(tempD1.keys()).index(group_1)\n",
    "            xcoord_1 = index_1\n",
    "            #Standard point of marker\n",
    "            xcoord = (xcoord_0+xcoord_1)/2\n",
    "            ycoord = aline_ymin + aline_ymargin*row_i\n",
    "            label = tempDF['SignifLabel'].iloc[row_i]\n",
    "            #Add annotation lines\n",
    "            aline_offset = yinter/5\n",
    "            aline_length = yinter/5 + aline_offset/2\n",
    "            ax.plot([xcoord_0, xcoord_0, xcoord_1, xcoord_1],\n",
    "                    [ycoord+aline_offset, ycoord+aline_length, ycoord+aline_length, ycoord+aline_offset],\n",
    "                    lw=1.5, c='k')\n",
    "            #Add annotation text\n",
    "            if label in ['***', '**', '*']:\n",
    "                text_offset = yinter/21\n",
    "                ax.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                            horizontalalignment='center', verticalalignment='bottom',\n",
    "                            fontsize='medium', color='k')\n",
    "            else:\n",
    "                text_offset = yinter/3.5\n",
    "                ax.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                            horizontalalignment='center', verticalalignment='bottom',\n",
    "                            fontsize='x-small', color='k')\n",
    "        #Add annotation\n",
    "        ax.set_title(dataset, {'fontsize':'small'})\n",
    "        xoff = 0.025\n",
    "        yoff = 0.01\n",
    "        rect = plt.Rectangle((xoff, 1), 1-xoff, 0.1,#Manual adjustment\n",
    "                             transform=ax.transAxes, facecolor=tempD2[dataset],\n",
    "                             clip_on=False, linewidth=0, zorder=0.5)\n",
    "        ax.add_patch(rect)\n",
    "        if ax_i==1:\n",
    "            at = AnchoredText('Consensus: own group', loc='center',\n",
    "                              bbox_to_anchor=(1.15, 0.505), bbox_transform=ax.transAxes,#Minor manual adjustment\n",
    "                              frameon=False, prop={'size':'small', 'rotation':90})\n",
    "            ax.add_artist(at)\n",
    "            rect = plt.Rectangle((1+xoff, yoff), 0.225, 1-2*yoff,#Manual adjustment\n",
    "                                 transform=ax.transAxes, facecolor=plt.get_cmap('tab20')(15),\n",
    "                                 clip_on=False, linewidth=0, zorder=0.5)\n",
    "            ax.add_patch(rect)\n",
    "    fig.tight_layout()\n",
    "    #Set title\n",
    "    modulename = moduleDF.loc[module, 'ModuleName']\n",
    "    initial = modulename[0].capitalize()\n",
    "    title = re.sub('^.', initial, modulename)+' ('+module+')'\n",
    "    title = '\\n'.join(wrap(title, 40))#Because the below wrap=True didn't work\n",
    "    fig.suptitle(title, size='small',\n",
    "                 verticalalignment='bottom', horizontalalignment='center', wrap=True, y=0.93)\n",
    "    ##Save\n",
    "    fileDir = './ExportFigures/'\n",
    "    ipynbName = '230224_LC-M001-PrOmics-vs-TrOmics-DIRAC-ver3_GOBP_'\n",
    "    fileName = 'RCI-pointplot-'+module.replace('GO:', 'GO')+'.pdf'\n",
    "    #plt.gcf().savefig(fileDir+ipynbName+fileName, dpi=300, bbox_inches='tight', pad_inches=0.04, transparent=True)\n",
    "    plt.show()\n",
    "    print('')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 3-5-6. Modules tightened by Rapa in both datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "posL = ['PrOmics_Rapa-vs-Ctrl', 'TrOmics_Rapa-vs-Ctrl']\n",
    "negL = ['']\n",
    "regulation = 'Tightened'\n",
    "nPlots = 3\n",
    "sort_varL = [col_prefix+'_AdjPval' for col_prefix in posL]\n",
    "\n",
    "#Prepare the target module set\n",
    "tempS = pd.Series(np.repeat(True, len(pvalDF)), index=pvalDF.index)#Initialize\n",
    "tempDF = pvalDF.loc[:, pvalDF.columns.str.contains('-vs-')]#Based on adjusted ANOVA P-values\n",
    "for col_n in tempDF.columns.tolist():\n",
    "    tempS1 = pvalDF[col_n]\n",
    "    tempS2 = diffDF[col_n]\n",
    "    if col_n in posL:\n",
    "        if regulation=='Tightened':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2>0)\n",
    "        elif regulation=='Loosened':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2<0)\n",
    "    elif col_n in negL:\n",
    "        tempS3 = (tempS1>=0.05)\n",
    "        #Significance for inverse regulation\n",
    "        if regulation=='Tightened':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2<0)\n",
    "        elif regulation=='Loosened':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2>0)\n",
    "        tempS1 = tempS3 | tempS1\n",
    "    else:\n",
    "        tempS1 = (tempS1>=0.0)\n",
    "    #Update True\n",
    "    tempS = tempS & tempS1\n",
    "tempL = tempS.loc[tempS.tolist()].index.tolist()\n",
    "print(len(tempL), regulation.lower()+' modules with significance in', posL, 'but not in', negL)\n",
    "\n",
    "#Select representatives and sort the target modules\n",
    "tempDF = statDF_flatten.loc[:, statDF_flatten.columns.str.contains('Pval$')]\n",
    "tempDF = pd.merge(moduleDF['ModuleName'], tempDF, left_index=True, right_index=True, how='left')\n",
    "tempDF = tempDF.loc[tempL]\n",
    "tempDF = tempDF.sort_values(by=sort_varL, ascending=True)\n",
    "topX = np.min([30, len(tempL)])\n",
    "print('Top', topX, 'modules (sort by', sort_varL, '):')\n",
    "display(tempDF.iloc[:topX])\n",
    "plotL = tempDF.index.tolist()[:np.min([nPlots, len(tempL)])]\n",
    "\n",
    "#Prepare RMS DF for plot\n",
    "tempDF1 = rmsDF_kk.reset_index().melt(var_name='SampleID', value_name='RMS', id_vars='ModuleID')\n",
    "tempDF = sampleDF.reset_index()[['SampleID', 'Dataset', 'Intervention']]\n",
    "tempDF1 = pd.merge(tempDF1, tempDF, on='SampleID', how='left')\n",
    "\n",
    "#Prepare label and color\n",
    "tempD0 = {'Ctrl':'Control', 'Aca':'Acarbose', 'Rapa':'Rapamycin',\n",
    "          'PrOmics':'Proteins', 'TrOmics':'Transcripts'}\n",
    "tempDF1['Dataset'] = tempDF1['Dataset'].map(tempD0)\n",
    "tempDF1['Group'] = tempDF1['Intervention'].map(tempD0)\n",
    "tempD1 = {'Control':'tab:blue', 'Acarbose':'tab:red', 'Rapamycin':'tab:purple'}\n",
    "tempD2 = {'Proteins':plt.get_cmap('tab20')(13), 'Transcripts':plt.get_cmap('tab20')(19)}\n",
    "\n",
    "#Prepare P-value DF for plot\n",
    "tempDF2 = statDF_flatten.loc[:, statDF_flatten.columns.str.contains('-vs-.*_AdjPval$')]\n",
    "tempDF2.columns = tempDF2.columns.str.replace('_AdjPval$', '')\n",
    "\n",
    "#Visualize each module\n",
    "for rank_i in range(len(plotL)):\n",
    "    print(' - Rank '+str(rank_i+1)+' (sort by', sort_varL, '):')\n",
    "    module = plotL[rank_i]\n",
    "    #Check module summary\n",
    "    tempDF = pd.DataFrame(moduleDF.loc[module]).T\n",
    "    display(tempDF)\n",
    "    \n",
    "    #Select RMS\n",
    "    tempDF3 = tempDF1.loc[tempDF1['ModuleID']==module]\n",
    "    \n",
    "    #Check RMS summary\n",
    "    tempDF = tempDF3.groupby(['Dataset', 'Group'])['RMS'].agg(['count', 'mean', 'std'])\n",
    "    tempL1 = []\n",
    "    tempL2 = []\n",
    "    for row_n in tempDF.index.tolist():\n",
    "        count, mean, std = tempDF.loc[row_n]\n",
    "        tempL1.append(mean - 1.96*std/np.sqrt(count))\n",
    "        tempL2.append(mean + 1.96*std/np.sqrt(count))\n",
    "    tempDF['0.025'] = tempL1\n",
    "    tempDF['0.975'] = tempL2\n",
    "    ##Multiindex sort\n",
    "    tempDF = tempDF.reset_index()\n",
    "    tempDF['Dataset'] = pd.Categorical(tempDF['Dataset'], categories=list(tempD2.keys()))\n",
    "    tempDF['Group'] = pd.Categorical(tempDF['Group'], categories=list(tempD1.keys()))\n",
    "    tempDF = tempDF.sort_values(by=['Dataset', 'Group']).set_index(['Dataset', 'Group'])\n",
    "    display(tempDF)\n",
    "    \n",
    "    #Prepare significance labels\n",
    "    tempS = tempDF2.loc[module]\n",
    "    tempS.name = 'AdjPval'\n",
    "    ##Clean\n",
    "    tempDF = tempS.index.to_series().str.split(pat='_', expand=True)\n",
    "    tempDF = tempDF.rename(columns={0:'Dataset', 1:'Comparison'})\n",
    "    tempS1 = tempDF['Dataset']\n",
    "    tempDF = tempDF['Comparison'].str.split(pat='-vs-', expand=True)\n",
    "    tempDF = tempDF.rename(columns={0:'Contrast', 1:'Baseline'})\n",
    "    tempDF = pd.merge(tempS1, tempDF, left_index=True, right_index=True, how='left')\n",
    "    tempDF4 = pd.merge(tempDF, tempS, left_index=True, right_index=True, how='left')\n",
    "    tempDF4['Dataset'] = tempDF4['Dataset'].map(tempD0)\n",
    "    tempDF4['Contrast'] = tempDF4['Contrast'].map(tempD0)\n",
    "    tempDF4['Baseline'] = tempDF4['Baseline'].map(tempD0)\n",
    "    ##Convert p-value to label\n",
    "    tempL = []\n",
    "    for row_i in range(len(tempDF4)):\n",
    "        pval = tempDF4['AdjPval'].iloc[row_i]\n",
    "        if pval<0.001:\n",
    "            tempL.append('***')\n",
    "        elif pval<0.01:\n",
    "            tempL.append('**')\n",
    "        elif pval<0.05:\n",
    "            tempL.append('*')\n",
    "        else:\n",
    "            pval_text = Decimal(str(pval)).quantize(Decimal('0.01'), rounding=ROUND_HALF_UP)\n",
    "            tempL.append(r'$P$ = '+str(pval_text))\n",
    "    tempDF4['SignifLabel'] = tempL\n",
    "    display(tempDF4)\n",
    "    \n",
    "    #Visualization\n",
    "    ymax = 1.0\n",
    "    ymin = 0.0\n",
    "    yinter = 0.2\n",
    "    ymargin_t = 0.275\n",
    "    ymargin_b = 0.05\n",
    "    aline_ymin = 1.0\n",
    "    aline_ymargin = 0.125\n",
    "    sns.set(style='ticks', font='Arial', context='talk')\n",
    "    fig, axes = plt.subplots(nrows=1, ncols=2, figsize=(4.75, 5), sharex=True, sharey=True,\n",
    "                             gridspec_kw={'width_ratios':[2, 2]})\n",
    "    for ax_i, ax in enumerate(axes.flat):\n",
    "        dataset = list(tempD2.keys())[ax_i]\n",
    "        tempDF = tempDF3.loc[tempDF3['Dataset']==dataset]\n",
    "        sns.pointplot(data=tempDF, x='Group', y='RMS', order=list(tempD1.keys()), palette=tempD1,\n",
    "                      markers='o', dodge=False, join=False, capsize=0.6, estimator=np.mean, ci=95, ax=ax)\n",
    "        sns.stripplot(data=tempDF, x='Group', y='RMS',\n",
    "                      order=list(tempD1.keys()), palette=tempD1, dodge=False, jitter=0.15,\n",
    "                      size=5, edgecolor='black', linewidth=1, **{'marker':'o', 'alpha':0.5}, ax=ax)\n",
    "        #Set axis\n",
    "        sns.despine()\n",
    "        ax.set(ylim=(ymin-ymargin_b, ymax+ymargin_t), yticks=np.arange(ymin, ymax + yinter/10, yinter))\n",
    "        plt.setp(ax.get_xticklabels(), rotation=70, horizontalalignment='right',\n",
    "                 verticalalignment='center', rotation_mode='anchor')\n",
    "        if ax_i==0:\n",
    "            plt.setp(ax, xlabel='', ylabel='Sample RMS\\n(Mean = Module RCI)')\n",
    "        else:\n",
    "            plt.setp(ax.get_yticklabels(), visible=False)\n",
    "            plt.setp(ax, xlabel='', ylabel='')\n",
    "        #Add significance labels\n",
    "        tempDF = tempDF4.loc[tempDF4['Dataset']==dataset]\n",
    "        for row_i in range(len(tempDF)):\n",
    "            #Baseline\n",
    "            group_0 = tempDF['Baseline'].iloc[row_i]\n",
    "            index_0 = list(tempD1.keys()).index(group_0)\n",
    "            xcoord_0 = index_0\n",
    "            #Contrast\n",
    "            group_1 = tempDF['Contrast'].iloc[row_i]\n",
    "            index_1 = list(tempD1.keys()).index(group_1)\n",
    "            xcoord_1 = index_1\n",
    "            #Standard point of marker\n",
    "            xcoord = (xcoord_0+xcoord_1)/2\n",
    "            ycoord = aline_ymin + aline_ymargin*row_i\n",
    "            label = tempDF['SignifLabel'].iloc[row_i]\n",
    "            #Add annotation lines\n",
    "            aline_offset = yinter/5\n",
    "            aline_length = yinter/5 + aline_offset/2\n",
    "            ax.plot([xcoord_0, xcoord_0, xcoord_1, xcoord_1],\n",
    "                    [ycoord+aline_offset, ycoord+aline_length, ycoord+aline_length, ycoord+aline_offset],\n",
    "                    lw=1.5, c='k')\n",
    "            #Add annotation text\n",
    "            if label in ['***', '**', '*']:\n",
    "                text_offset = yinter/21\n",
    "                ax.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                            horizontalalignment='center', verticalalignment='bottom',\n",
    "                            fontsize='medium', color='k')\n",
    "            else:\n",
    "                text_offset = yinter/3.5\n",
    "                ax.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                            horizontalalignment='center', verticalalignment='bottom',\n",
    "                            fontsize='x-small', color='k')\n",
    "        #Add annotation\n",
    "        ax.set_title(dataset, {'fontsize':'small'})\n",
    "        xoff = 0.025\n",
    "        yoff = 0.01\n",
    "        rect = plt.Rectangle((xoff, 1), 1-xoff, 0.1,#Manual adjustment\n",
    "                             transform=ax.transAxes, facecolor=tempD2[dataset],\n",
    "                             clip_on=False, linewidth=0, zorder=0.5)\n",
    "        ax.add_patch(rect)\n",
    "        if ax_i==1:\n",
    "            at = AnchoredText('Consensus: own group', loc='center',\n",
    "                              bbox_to_anchor=(1.15, 0.505), bbox_transform=ax.transAxes,#Minor manual adjustment\n",
    "                              frameon=False, prop={'size':'small', 'rotation':90})\n",
    "            ax.add_artist(at)\n",
    "            rect = plt.Rectangle((1+xoff, yoff), 0.225, 1-2*yoff,#Manual adjustment\n",
    "                                 transform=ax.transAxes, facecolor=plt.get_cmap('tab20')(15),\n",
    "                                 clip_on=False, linewidth=0, zorder=0.5)\n",
    "            ax.add_patch(rect)\n",
    "    fig.tight_layout()\n",
    "    #Set title\n",
    "    modulename = moduleDF.loc[module, 'ModuleName']\n",
    "    initial = modulename[0].capitalize()\n",
    "    title = re.sub('^.', initial, modulename)+' ('+module+')'\n",
    "    title = '\\n'.join(wrap(title, 40))#Because the below wrap=True didn't work\n",
    "    fig.suptitle(title, size='small',\n",
    "                 verticalalignment='bottom', horizontalalignment='center', wrap=True, y=0.93)\n",
    "    ##Save\n",
    "    fileDir = './ExportFigures/'\n",
    "    ipynbName = '230224_LC-M001-PrOmics-vs-TrOmics-DIRAC-ver3_GOBP_'\n",
    "    fileName = 'RCI-pointplot-'+module.replace('GO:', 'GO')+'.pdf'\n",
    "    #plt.gcf().savefig(fileDir+ipynbName+fileName, dpi=300, bbox_inches='tight', pad_inches=0.04, transparent=True)\n",
    "    plt.show()\n",
    "    print('')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 3-5-7. Modules tightened by Rapa specifically in proteomics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "posL = ['PrOmics_Rapa-vs-Ctrl']\n",
    "negL = ['TrOmics_Rapa-vs-Ctrl']\n",
    "regulation = 'Tightened'\n",
    "nPlots = 3\n",
    "sort_varL = [col_prefix+'_AdjPval' for col_prefix in posL]\n",
    "\n",
    "#Prepare the target module set\n",
    "tempS = pd.Series(np.repeat(True, len(pvalDF)), index=pvalDF.index)#Initialize\n",
    "tempDF = pvalDF.loc[:, pvalDF.columns.str.contains('-vs-')]#Based on adjusted ANOVA P-values\n",
    "for col_n in tempDF.columns.tolist():\n",
    "    tempS1 = pvalDF[col_n]\n",
    "    tempS2 = diffDF[col_n]\n",
    "    if col_n in posL:\n",
    "        if regulation=='Tightened':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2>0)\n",
    "        elif regulation=='Loosened':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2<0)\n",
    "    elif col_n in negL:\n",
    "        tempS3 = (tempS1>=0.05)\n",
    "        #Significance for inverse regulation\n",
    "        if regulation=='Tightened':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2<0)\n",
    "        elif regulation=='Loosened':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2>0)\n",
    "        tempS1 = tempS3 | tempS1\n",
    "    else:\n",
    "        tempS1 = (tempS1>=0.0)\n",
    "    #Update True\n",
    "    tempS = tempS & tempS1\n",
    "tempL = tempS.loc[tempS.tolist()].index.tolist()\n",
    "print(len(tempL), regulation.lower()+' modules with significance in', posL, 'but not in', negL)\n",
    "\n",
    "#Select representatives and sort the target modules\n",
    "tempDF = statDF_flatten.loc[:, statDF_flatten.columns.str.contains('Pval$')]\n",
    "tempDF = pd.merge(moduleDF['ModuleName'], tempDF, left_index=True, right_index=True, how='left')\n",
    "tempDF = tempDF.loc[tempL]\n",
    "tempDF = tempDF.sort_values(by=sort_varL, ascending=True)\n",
    "topX = np.min([30, len(tempL)])\n",
    "print('Top', topX, 'modules (sort by', sort_varL, '):')\n",
    "display(tempDF.iloc[:topX])\n",
    "plotL = tempDF.index.tolist()[:np.min([nPlots, len(tempL)])]\n",
    "\n",
    "#Prepare RMS DF for plot\n",
    "tempDF1 = rmsDF_kk.reset_index().melt(var_name='SampleID', value_name='RMS', id_vars='ModuleID')\n",
    "tempDF = sampleDF.reset_index()[['SampleID', 'Dataset', 'Intervention']]\n",
    "tempDF1 = pd.merge(tempDF1, tempDF, on='SampleID', how='left')\n",
    "\n",
    "#Prepare label and color\n",
    "tempD0 = {'Ctrl':'Control', 'Aca':'Acarbose', 'Rapa':'Rapamycin',\n",
    "          'PrOmics':'Proteins', 'TrOmics':'Transcripts'}\n",
    "tempDF1['Dataset'] = tempDF1['Dataset'].map(tempD0)\n",
    "tempDF1['Group'] = tempDF1['Intervention'].map(tempD0)\n",
    "tempD1 = {'Control':'tab:blue', 'Acarbose':'tab:red', 'Rapamycin':'tab:purple'}\n",
    "tempD2 = {'Proteins':plt.get_cmap('tab20')(13), 'Transcripts':plt.get_cmap('tab20')(19)}\n",
    "\n",
    "#Prepare P-value DF for plot\n",
    "tempDF2 = statDF_flatten.loc[:, statDF_flatten.columns.str.contains('-vs-.*_AdjPval$')]\n",
    "tempDF2.columns = tempDF2.columns.str.replace('_AdjPval$', '')\n",
    "\n",
    "#Visualize each module\n",
    "for rank_i in range(len(plotL)):\n",
    "    print(' - Rank '+str(rank_i+1)+' (sort by', sort_varL, '):')\n",
    "    module = plotL[rank_i]\n",
    "    #Check module summary\n",
    "    tempDF = pd.DataFrame(moduleDF.loc[module]).T\n",
    "    display(tempDF)\n",
    "    \n",
    "    #Select RMS\n",
    "    tempDF3 = tempDF1.loc[tempDF1['ModuleID']==module]\n",
    "    \n",
    "    #Check RMS summary\n",
    "    tempDF = tempDF3.groupby(['Dataset', 'Group'])['RMS'].agg(['count', 'mean', 'std'])\n",
    "    tempL1 = []\n",
    "    tempL2 = []\n",
    "    for row_n in tempDF.index.tolist():\n",
    "        count, mean, std = tempDF.loc[row_n]\n",
    "        tempL1.append(mean - 1.96*std/np.sqrt(count))\n",
    "        tempL2.append(mean + 1.96*std/np.sqrt(count))\n",
    "    tempDF['0.025'] = tempL1\n",
    "    tempDF['0.975'] = tempL2\n",
    "    ##Multiindex sort\n",
    "    tempDF = tempDF.reset_index()\n",
    "    tempDF['Dataset'] = pd.Categorical(tempDF['Dataset'], categories=list(tempD2.keys()))\n",
    "    tempDF['Group'] = pd.Categorical(tempDF['Group'], categories=list(tempD1.keys()))\n",
    "    tempDF = tempDF.sort_values(by=['Dataset', 'Group']).set_index(['Dataset', 'Group'])\n",
    "    display(tempDF)\n",
    "    \n",
    "    #Prepare significance labels\n",
    "    tempS = tempDF2.loc[module]\n",
    "    tempS.name = 'AdjPval'\n",
    "    ##Clean\n",
    "    tempDF = tempS.index.to_series().str.split(pat='_', expand=True)\n",
    "    tempDF = tempDF.rename(columns={0:'Dataset', 1:'Comparison'})\n",
    "    tempS1 = tempDF['Dataset']\n",
    "    tempDF = tempDF['Comparison'].str.split(pat='-vs-', expand=True)\n",
    "    tempDF = tempDF.rename(columns={0:'Contrast', 1:'Baseline'})\n",
    "    tempDF = pd.merge(tempS1, tempDF, left_index=True, right_index=True, how='left')\n",
    "    tempDF4 = pd.merge(tempDF, tempS, left_index=True, right_index=True, how='left')\n",
    "    tempDF4['Dataset'] = tempDF4['Dataset'].map(tempD0)\n",
    "    tempDF4['Contrast'] = tempDF4['Contrast'].map(tempD0)\n",
    "    tempDF4['Baseline'] = tempDF4['Baseline'].map(tempD0)\n",
    "    ##Convert p-value to label\n",
    "    tempL = []\n",
    "    for row_i in range(len(tempDF4)):\n",
    "        pval = tempDF4['AdjPval'].iloc[row_i]\n",
    "        if pval<0.001:\n",
    "            tempL.append('***')\n",
    "        elif pval<0.01:\n",
    "            tempL.append('**')\n",
    "        elif pval<0.05:\n",
    "            tempL.append('*')\n",
    "        else:\n",
    "            pval_text = Decimal(str(pval)).quantize(Decimal('0.01'), rounding=ROUND_HALF_UP)\n",
    "            tempL.append(r'$P$ = '+str(pval_text))\n",
    "    tempDF4['SignifLabel'] = tempL\n",
    "    display(tempDF4)\n",
    "    \n",
    "    #Visualization\n",
    "    ymax = 1.0\n",
    "    ymin = 0.0\n",
    "    yinter = 0.2\n",
    "    ymargin_t = 0.275\n",
    "    ymargin_b = 0.05\n",
    "    aline_ymin = 1.0\n",
    "    aline_ymargin = 0.125\n",
    "    sns.set(style='ticks', font='Arial', context='talk')\n",
    "    fig, axes = plt.subplots(nrows=1, ncols=2, figsize=(4.75, 5), sharex=True, sharey=True,\n",
    "                             gridspec_kw={'width_ratios':[2, 2]})\n",
    "    for ax_i, ax in enumerate(axes.flat):\n",
    "        dataset = list(tempD2.keys())[ax_i]\n",
    "        tempDF = tempDF3.loc[tempDF3['Dataset']==dataset]\n",
    "        sns.pointplot(data=tempDF, x='Group', y='RMS', order=list(tempD1.keys()), palette=tempD1,\n",
    "                      markers='o', dodge=False, join=False, capsize=0.6, estimator=np.mean, ci=95, ax=ax)\n",
    "        sns.stripplot(data=tempDF, x='Group', y='RMS',\n",
    "                      order=list(tempD1.keys()), palette=tempD1, dodge=False, jitter=0.15,\n",
    "                      size=5, edgecolor='black', linewidth=1, **{'marker':'o', 'alpha':0.5}, ax=ax)\n",
    "        #Set axis\n",
    "        sns.despine()\n",
    "        ax.set(ylim=(ymin-ymargin_b, ymax+ymargin_t), yticks=np.arange(ymin, ymax + yinter/10, yinter))\n",
    "        plt.setp(ax.get_xticklabels(), rotation=70, horizontalalignment='right',\n",
    "                 verticalalignment='center', rotation_mode='anchor')\n",
    "        if ax_i==0:\n",
    "            plt.setp(ax, xlabel='', ylabel='Sample RMS\\n(Mean = Module RCI)')\n",
    "        else:\n",
    "            plt.setp(ax.get_yticklabels(), visible=False)\n",
    "            plt.setp(ax, xlabel='', ylabel='')\n",
    "        #Add significance labels\n",
    "        tempDF = tempDF4.loc[tempDF4['Dataset']==dataset]\n",
    "        for row_i in range(len(tempDF)):\n",
    "            #Baseline\n",
    "            group_0 = tempDF['Baseline'].iloc[row_i]\n",
    "            index_0 = list(tempD1.keys()).index(group_0)\n",
    "            xcoord_0 = index_0\n",
    "            #Contrast\n",
    "            group_1 = tempDF['Contrast'].iloc[row_i]\n",
    "            index_1 = list(tempD1.keys()).index(group_1)\n",
    "            xcoord_1 = index_1\n",
    "            #Standard point of marker\n",
    "            xcoord = (xcoord_0+xcoord_1)/2\n",
    "            ycoord = aline_ymin + aline_ymargin*row_i\n",
    "            label = tempDF['SignifLabel'].iloc[row_i]\n",
    "            #Add annotation lines\n",
    "            aline_offset = yinter/5\n",
    "            aline_length = yinter/5 + aline_offset/2\n",
    "            ax.plot([xcoord_0, xcoord_0, xcoord_1, xcoord_1],\n",
    "                    [ycoord+aline_offset, ycoord+aline_length, ycoord+aline_length, ycoord+aline_offset],\n",
    "                    lw=1.5, c='k')\n",
    "            #Add annotation text\n",
    "            if label in ['***', '**', '*']:\n",
    "                text_offset = yinter/21\n",
    "                ax.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                            horizontalalignment='center', verticalalignment='bottom',\n",
    "                            fontsize='medium', color='k')\n",
    "            else:\n",
    "                text_offset = yinter/3.5\n",
    "                ax.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                            horizontalalignment='center', verticalalignment='bottom',\n",
    "                            fontsize='x-small', color='k')\n",
    "        #Add annotation\n",
    "        ax.set_title(dataset, {'fontsize':'small'})\n",
    "        xoff = 0.025\n",
    "        yoff = 0.01\n",
    "        rect = plt.Rectangle((xoff, 1), 1-xoff, 0.1,#Manual adjustment\n",
    "                             transform=ax.transAxes, facecolor=tempD2[dataset],\n",
    "                             clip_on=False, linewidth=0, zorder=0.5)\n",
    "        ax.add_patch(rect)\n",
    "        if ax_i==1:\n",
    "            at = AnchoredText('Consensus: own group', loc='center',\n",
    "                              bbox_to_anchor=(1.15, 0.505), bbox_transform=ax.transAxes,#Minor manual adjustment\n",
    "                              frameon=False, prop={'size':'small', 'rotation':90})\n",
    "            ax.add_artist(at)\n",
    "            rect = plt.Rectangle((1+xoff, yoff), 0.225, 1-2*yoff,#Manual adjustment\n",
    "                                 transform=ax.transAxes, facecolor=plt.get_cmap('tab20')(15),\n",
    "                                 clip_on=False, linewidth=0, zorder=0.5)\n",
    "            ax.add_patch(rect)\n",
    "    fig.tight_layout()\n",
    "    #Set title\n",
    "    modulename = moduleDF.loc[module, 'ModuleName']\n",
    "    initial = modulename[0].capitalize()\n",
    "    title = re.sub('^.', initial, modulename)+' ('+module+')'\n",
    "    title = '\\n'.join(wrap(title, 40))#Because the below wrap=True didn't work\n",
    "    fig.suptitle(title, size='small',\n",
    "                 verticalalignment='bottom', horizontalalignment='center', wrap=True, y=0.93)\n",
    "    ##Save\n",
    "    fileDir = './ExportFigures/'\n",
    "    ipynbName = '230224_LC-M001-PrOmics-vs-TrOmics-DIRAC-ver3_GOBP_'\n",
    "    fileName = 'RCI-pointplot-'+module.replace('GO:', 'GO')+'.pdf'\n",
    "    #plt.gcf().savefig(fileDir+ipynbName+fileName, dpi=300, bbox_inches='tight', pad_inches=0.04, transparent=True)\n",
    "    plt.show()\n",
    "    print('')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 3-5-8. Manually selected modules"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Manual selection based on all the DIRAC results\n",
    "##Previous list\n",
    "tempL = ['GO:0006635', 'GO:0031998', 'GO:0016558', 'GO:0006625',#Common in M001\n",
    "         'GO:0006637', 'GO:0006734', 'GO:0001561',#Related to fatty acid oxidation (mentioned in earlier version)\n",
    "         'GO:0023035', 'GO:0039529',#Immune response (mentioned in earlier version)\n",
    "         'GO:0010638', 'GO:0002181', 'GO:0045899',#Mentioned in earlier version\n",
    "         'GO:0006635', 'GO:1990126', 'GO:0098761',#Common in proteomics and transcriptomics\n",
    "         'GO:0061732', 'GO:0006086', 'GO:0019441',#Specific to proteomics (Aca)\n",
    "         'GO:0031998', 'GO:0044794', 'GO:0006625', 'GO:0016558', 'GO:0019441', 'GO:0033572',#Specific to proteomics (Rapa)\n",
    "         'GO:0006635', 'GO:0002181', 'GO:0000028', 'GO:0016558', 'GO:0006734', 'GO:0098761',#Common in M001 + M004\n",
    "         'GO:0015986', 'GO:0042776', 'GO:0070934', 'GO:0006703', 'GO:0034354', 'GO:0045899']#Similar to 4EGI-1\n",
    "print('Previous list:', len(tempL))\n",
    "##New list (based on adjusted dataset and parametric test)\n",
    "###Common among interventions in M001 proteomics (version 7-2)\n",
    "tempS1 = {'GO:0017144', 'GO:0031998', 'GO:0006102', 'GO:0006749', 'GO:0006625',\n",
    "          'GO:0016558', 'GO:0006635', 'GO:0034063', 'GO:0006103', 'GO:0042866',\n",
    "          'GO:0002181', 'GO:0098761', 'GO:0006637', 'GO:0140374', 'GO:1902416'}\n",
    "print('Common among interventions in M001 proteomics:', len(tempS1))\n",
    "###Common among interventions in M001 proteomics and transcriptomics (version 3-1)\n",
    "tempS2 = {'GO:0016558', 'GO:0006749', 'GO:0006635', 'GO:0098761', 'GO:0150093',\n",
    "          'GO:0043691', 'GO:0006796', 'GO:0006734', 'GO:0015909'}\n",
    "print('Common among interventions in M001 proteomics and transcriptomics:', len(tempS2))\n",
    "###Common among interventions but specifically in M001 proteomics (version 3-1)\n",
    "tempS3 = {'GO:0006625', 'GO:0031998', 'GO:0042866', 'GO:0033572', 'GO:0000028'}\n",
    "print('Common among interventions but specifically in M001 proteomics:', len(tempS3))\n",
    "###Similar to M004 4EGI-1 in any intervention of M001 proteomics (version 3-1)\n",
    "tempS4 = {'GO:0010637', 'GO:0006102', 'GO:0006739', 'GO:0046826', 'GO:0023035'}\n",
    "print('Similar to M004 4EGI-1 in any intervention of M001 proteomics:', len(tempS4))\n",
    "###Dissimilar to M004 4EGI-1 in any intervention of M001 proteomics (version 3-1)\n",
    "tempS5 = {'GO:0044794', 'GO:0034372', 'GO:0006544', 'GO:0008210', 'GO:0000038',\n",
    "          'GO:0072378', 'GO:0070189'}\n",
    "print('Dissimilar to M004 4EGI-1 in any intervention of M001 proteomics:', len(tempS5))\n",
    "##Merge\n",
    "tempL = [item for sublist in [tempL, tempS1, tempS2, tempS3, tempS4, tempS5] for item in sublist]\n",
    "\n",
    "#Resolve duplicates and judge whether it was included in this analysis\n",
    "tempL = list(set(tempL))\n",
    "tempL.sort()\n",
    "for module in tempL:\n",
    "    if module not in moduleDF.index.tolist():\n",
    "        print(module+' was NOT included in this analysis.')\n",
    "        tempL.remove(module)\n",
    "\n",
    "print('Assessed modules:', len(tempL))\n",
    "display(moduleDF.loc[moduleDF.index.isin(tempL)])\n",
    "\n",
    "targetL = tempL"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tempL = targetL\n",
    "topX = len(targetL)\n",
    "nPlots = len(targetL)\n",
    "sort_varL = ['PrOmics_ANOVA_AdjPval', 'TrOmics_ANOVA_AdjPval']\n",
    "\n",
    "#Sort the target modules\n",
    "tempDF = statDF_flatten.loc[:, statDF_flatten.columns.str.contains('Pval$')]\n",
    "tempDF = pd.merge(moduleDF['ModuleName'], tempDF, left_index=True, right_index=True, how='left')\n",
    "tempDF = tempDF.loc[tempL]\n",
    "tempDF = tempDF.sort_values(by=sort_varL, ascending=True)\n",
    "#topX = np.min([30, len(tempL)])\n",
    "print('Top', topX, 'modules (sort by', sort_varL, '):')\n",
    "display(tempDF.iloc[:topX])\n",
    "plotL = tempDF.index.tolist()[:np.min([nPlots, len(tempL)])]\n",
    "\n",
    "#Prepare RMS DF for plot\n",
    "tempDF1 = rmsDF_kk.reset_index().melt(var_name='SampleID', value_name='RMS', id_vars='ModuleID')\n",
    "tempDF = sampleDF.reset_index()[['SampleID', 'Dataset', 'Intervention']]\n",
    "tempDF1 = pd.merge(tempDF1, tempDF, on='SampleID', how='left')\n",
    "\n",
    "#Prepare label and color\n",
    "tempD0 = {'Ctrl':'Control', 'Aca':'Acarbose', 'Rapa':'Rapamycin',\n",
    "          'PrOmics':'Proteins', 'TrOmics':'Transcripts'}\n",
    "tempDF1['Dataset'] = tempDF1['Dataset'].map(tempD0)\n",
    "tempDF1['Group'] = tempDF1['Intervention'].map(tempD0)\n",
    "tempD1 = {'Control':'tab:blue', 'Acarbose':'tab:red', 'Rapamycin':'tab:purple'}\n",
    "tempD2 = {'Proteins':plt.get_cmap('tab20')(13), 'Transcripts':plt.get_cmap('tab20')(19)}\n",
    "\n",
    "#Prepare P-value DF for plot\n",
    "tempDF2 = statDF_flatten.loc[:, statDF_flatten.columns.str.contains('-vs-.*_AdjPval$')]\n",
    "tempDF2.columns = tempDF2.columns.str.replace('_AdjPval$', '')\n",
    "\n",
    "#Visualize each module\n",
    "for rank_i in range(len(plotL)):\n",
    "    print(' - Rank '+str(rank_i+1)+' (sort by', sort_varL, '):')\n",
    "    module = plotL[rank_i]\n",
    "    #Check module summary\n",
    "    tempDF = pd.DataFrame(moduleDF.loc[module]).T\n",
    "    display(tempDF)\n",
    "    \n",
    "    #Select RMS\n",
    "    tempDF3 = tempDF1.loc[tempDF1['ModuleID']==module]\n",
    "    \n",
    "    #Check RMS summary\n",
    "    tempDF = tempDF3.groupby(['Dataset', 'Group'])['RMS'].agg(['count', 'mean', 'std'])\n",
    "    tempL1 = []\n",
    "    tempL2 = []\n",
    "    for row_n in tempDF.index.tolist():\n",
    "        count, mean, std = tempDF.loc[row_n]\n",
    "        tempL1.append(mean - 1.96*std/np.sqrt(count))\n",
    "        tempL2.append(mean + 1.96*std/np.sqrt(count))\n",
    "    tempDF['0.025'] = tempL1\n",
    "    tempDF['0.975'] = tempL2\n",
    "    ##Multiindex sort\n",
    "    tempDF = tempDF.reset_index()\n",
    "    tempDF['Dataset'] = pd.Categorical(tempDF['Dataset'], categories=list(tempD2.keys()))\n",
    "    tempDF['Group'] = pd.Categorical(tempDF['Group'], categories=list(tempD1.keys()))\n",
    "    tempDF = tempDF.sort_values(by=['Dataset', 'Group']).set_index(['Dataset', 'Group'])\n",
    "    display(tempDF)\n",
    "    \n",
    "    #Prepare significance labels\n",
    "    tempS = tempDF2.loc[module]\n",
    "    tempS.name = 'AdjPval'\n",
    "    ##Clean\n",
    "    tempDF = tempS.index.to_series().str.split(pat='_', expand=True)\n",
    "    tempDF = tempDF.rename(columns={0:'Dataset', 1:'Comparison'})\n",
    "    tempS1 = tempDF['Dataset']\n",
    "    tempDF = tempDF['Comparison'].str.split(pat='-vs-', expand=True)\n",
    "    tempDF = tempDF.rename(columns={0:'Contrast', 1:'Baseline'})\n",
    "    tempDF = pd.merge(tempS1, tempDF, left_index=True, right_index=True, how='left')\n",
    "    tempDF4 = pd.merge(tempDF, tempS, left_index=True, right_index=True, how='left')\n",
    "    tempDF4['Dataset'] = tempDF4['Dataset'].map(tempD0)\n",
    "    tempDF4['Contrast'] = tempDF4['Contrast'].map(tempD0)\n",
    "    tempDF4['Baseline'] = tempDF4['Baseline'].map(tempD0)\n",
    "    ##Convert p-value to label\n",
    "    tempL = []\n",
    "    for row_i in range(len(tempDF4)):\n",
    "        pval = tempDF4['AdjPval'].iloc[row_i]\n",
    "        if pval<0.001:\n",
    "            tempL.append('***')\n",
    "        elif pval<0.01:\n",
    "            tempL.append('**')\n",
    "        elif pval<0.05:\n",
    "            tempL.append('*')\n",
    "        else:\n",
    "            pval_text = Decimal(str(pval)).quantize(Decimal('0.01'), rounding=ROUND_HALF_UP)\n",
    "            tempL.append(r'$P$ = '+str(pval_text))\n",
    "    tempDF4['SignifLabel'] = tempL\n",
    "    display(tempDF4)\n",
    "    \n",
    "    #Visualization\n",
    "    ymax = 1.0\n",
    "    ymin = 0.0\n",
    "    yinter = 0.2\n",
    "    ymargin_t = 0.275\n",
    "    ymargin_b = 0.05\n",
    "    aline_ymin = 1.0\n",
    "    aline_ymargin = 0.125\n",
    "    sns.set(style='ticks', font='Arial', context='talk')\n",
    "    fig, axes = plt.subplots(nrows=1, ncols=2, figsize=(4.75, 5), sharex=True, sharey=True,\n",
    "                             gridspec_kw={'width_ratios':[2, 2]})\n",
    "    for ax_i, ax in enumerate(axes.flat):\n",
    "        dataset = list(tempD2.keys())[ax_i]\n",
    "        tempDF = tempDF3.loc[tempDF3['Dataset']==dataset]\n",
    "        sns.pointplot(data=tempDF, x='Group', y='RMS', order=list(tempD1.keys()), palette=tempD1,\n",
    "                      markers='o', dodge=False, join=False, capsize=0.6, estimator=np.mean, ci=95, ax=ax)\n",
    "        sns.stripplot(data=tempDF, x='Group', y='RMS',\n",
    "                      order=list(tempD1.keys()), palette=tempD1, dodge=False, jitter=0.15,\n",
    "                      size=5, edgecolor='black', linewidth=1, **{'marker':'o', 'alpha':0.5}, ax=ax)\n",
    "        #Set axis\n",
    "        sns.despine()\n",
    "        ax.set(ylim=(ymin-ymargin_b, ymax+ymargin_t), yticks=np.arange(ymin, ymax + yinter/10, yinter))\n",
    "        plt.setp(ax.get_xticklabels(), rotation=70, horizontalalignment='right',\n",
    "                 verticalalignment='center', rotation_mode='anchor')\n",
    "        if ax_i==0:\n",
    "            plt.setp(ax, xlabel='', ylabel='Sample RMS\\n(Mean = Module RCI)')\n",
    "        else:\n",
    "            plt.setp(ax.get_yticklabels(), visible=False)\n",
    "            plt.setp(ax, xlabel='', ylabel='')\n",
    "        #Add significance labels\n",
    "        tempDF = tempDF4.loc[tempDF4['Dataset']==dataset]\n",
    "        for row_i in range(len(tempDF)):\n",
    "            #Baseline\n",
    "            group_0 = tempDF['Baseline'].iloc[row_i]\n",
    "            index_0 = list(tempD1.keys()).index(group_0)\n",
    "            xcoord_0 = index_0\n",
    "            #Contrast\n",
    "            group_1 = tempDF['Contrast'].iloc[row_i]\n",
    "            index_1 = list(tempD1.keys()).index(group_1)\n",
    "            xcoord_1 = index_1\n",
    "            #Standard point of marker\n",
    "            xcoord = (xcoord_0+xcoord_1)/2\n",
    "            ycoord = aline_ymin + aline_ymargin*row_i\n",
    "            label = tempDF['SignifLabel'].iloc[row_i]\n",
    "            #Add annotation lines\n",
    "            aline_offset = yinter/5\n",
    "            aline_length = yinter/5 + aline_offset/2\n",
    "            ax.plot([xcoord_0, xcoord_0, xcoord_1, xcoord_1],\n",
    "                    [ycoord+aline_offset, ycoord+aline_length, ycoord+aline_length, ycoord+aline_offset],\n",
    "                    lw=1.5, c='k')\n",
    "            #Add annotation text\n",
    "            if label in ['***', '**', '*']:\n",
    "                text_offset = yinter/21\n",
    "                ax.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                            horizontalalignment='center', verticalalignment='bottom',\n",
    "                            fontsize='medium', color='k')\n",
    "            else:\n",
    "                text_offset = yinter/3.5\n",
    "                ax.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                            horizontalalignment='center', verticalalignment='bottom',\n",
    "                            fontsize='x-small', color='k')\n",
    "        #Add annotation\n",
    "        ax.set_title(dataset, {'fontsize':'small'})\n",
    "        xoff = 0.025\n",
    "        yoff = 0.01\n",
    "        rect = plt.Rectangle((xoff, 1), 1-xoff, 0.1,#Manual adjustment\n",
    "                             transform=ax.transAxes, facecolor=tempD2[dataset],\n",
    "                             clip_on=False, linewidth=0, zorder=0.5)\n",
    "        ax.add_patch(rect)\n",
    "        if ax_i==1:\n",
    "            at = AnchoredText('Consensus: own group', loc='center',\n",
    "                              bbox_to_anchor=(1.15, 0.505), bbox_transform=ax.transAxes,#Minor manual adjustment\n",
    "                              frameon=False, prop={'size':'small', 'rotation':90})\n",
    "            ax.add_artist(at)\n",
    "            rect = plt.Rectangle((1+xoff, yoff), 0.225, 1-2*yoff,#Manual adjustment\n",
    "                                 transform=ax.transAxes, facecolor=plt.get_cmap('tab20')(15),\n",
    "                                 clip_on=False, linewidth=0, zorder=0.5)\n",
    "            ax.add_patch(rect)\n",
    "    fig.tight_layout()\n",
    "    #Set title\n",
    "    modulename = moduleDF.loc[module, 'ModuleName']\n",
    "    initial = modulename[0].capitalize()\n",
    "    title = re.sub('^.', initial, modulename)+' ('+module+')'\n",
    "    title = '\\n'.join(wrap(title, 40))#Because the below wrap=True didn't work\n",
    "    fig.suptitle(title, size='small',\n",
    "                 verticalalignment='bottom', horizontalalignment='center', wrap=True, y=0.93)\n",
    "    ##Save\n",
    "    fileDir = './ExportFigures/'\n",
    "    ipynbName = '230224_LC-M001-PrOmics-vs-TrOmics-DIRAC-ver3_GOBP_'\n",
    "    fileName = 'RCI-pointplot-'+module.replace('GO:', 'GO')+'.pdf'\n",
    "    plt.gcf().savefig(fileDir+ipynbName+fileName, dpi=300, bbox_inches='tight', pad_inches=0.04, transparent=True)\n",
    "    plt.show()\n",
    "    print('')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 5. Rank matching score under a fixed consensus: inter-group module comparison"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> The comparison of RMSs under a fixed rank consensus is skipped, since the pattern similarity across datasets is out of current interest.  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# — End of notebook —"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "arivale-py3 - Python",
   "language": "python",
   "name": "conda-env-arivale-py3-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  },
  "widgets": {
   "application/vnd.jupyter.widget-state+json": {
    "state": {},
    "version_major": 2,
    "version_minor": 0
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
