{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# DIRAC Analysis of LC M001 Liver Proteomics — DIRAC with GOBP Modules"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***by Kengo Watanabe***  \n",
    "\n",
    "This Jupyter Notebook (with Python 3 kernel) performed the differential rank conservation (DIRAC; Eddy, J.A. et al. PLoS Comput. Biol. 2010) analysis on the preprocessed Longevity Consortium (LC) M001 proteomics dataset (adjusted with sex and age; analytes detected in all samples; sample-based robust Z-score followed by analyte-based robust Z-score) using the retrieved a priori module set (Gene Ontology (Biological Process) derived by EMBL-EBI QuickGO API; ≥4 analytes and ≥50% coverage).  \n",
    "\n",
    "Input files:  \n",
    "- Preprocessed analyte data: 230213_LC-M001-proteomics-DIRAC-ver7-2_Preprocessing_normalized-data.tsv  \n",
    "- Module–analyte metadata: 220427_LCproteomics-M001-DIRAC-ver6-2_Preprocessing_module-metadata_QuickGO-GOBP-min-n4-cov50.tsv  \n",
    "- Analyte metadata: 220427_LCproteomics-M001-DIRAC-ver6-2_Preprocessing_analyte-metadata_UniProt.tsv  \n",
    "- Sample–mouse metadata: 230213_LC-M001-proteomics-DIRAC-ver7-2_Preprocessing_sample-metadata.tsv  \n",
    "\n",
    "Output figures and tables:  \n",
    "- Figure 2  \n",
    "- Supplementary Figure 1  \n",
    "- Supplementary Data 1  \n",
    "\n",
    "Original notebook (memo for my future tracing):  \n",
    "- dalek:\\[JupyterLab HOME\\]/230206_LC-M001-proteomics-DIRAC-ver7/230214_LC-M001-proteomics-DIRAC-ver7-2_DIRAC-GOBP.ipynb  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# packages in environment at /opt/conda/envs/arivale-py3:\n",
      "#\n",
      "# Name                    Version                   Build  Channel\n",
      "_libgcc_mutex             0.1                 conda_forge    conda-forge\n",
      "_openmp_mutex             4.5                       1_gnu    conda-forge\n",
      "analytics                 0.1                      pypi_0    pypi\n",
      "argon2-cffi               21.1.0           py39h3811e60_0    conda-forge\n",
      "arivale-data-interface    0.1.0                    pypi_0    pypi\n",
      "async_generator           1.10                       py_0    conda-forge\n",
      "atk-1.0                   2.36.0               h3371d22_4    conda-forge\n",
      "attrs                     21.2.0             pyhd8ed1ab_0    conda-forge\n",
      "backcall                  0.2.0              pyh9f0ad1d_0    conda-forge\n",
      "backports                 1.0                        py_2    conda-forge\n",
      "backports.functools_lru_cache 1.6.4              pyhd8ed1ab_0    conda-forge\n",
      "biopython                 1.79             py39h3811e60_0    conda-forge\n",
      "bleach                    4.1.0              pyhd8ed1ab_0    conda-forge\n",
      "bokeh                     2.4.1            py39hf3d152e_1    conda-forge\n",
      "boost-cpp                 1.74.0               h312852a_4    conda-forge\n",
      "boto                      2.49.0                     py_0    conda-forge\n",
      "boto3                     1.19.2                   pypi_0    pypi\n",
      "botocore                  1.22.2                   pypi_0    pypi\n",
      "brotlipy                  0.7.0           py39h3811e60_1001    conda-forge\n",
      "bzip2                     1.0.8                h7f98852_4    conda-forge\n",
      "c-ares                    1.17.2               h7f98852_0    conda-forge\n",
      "ca-certificates           2022.12.7            ha878542_0    conda-forge\n",
      "cachecontrol              0.12.6                     py_0    conda-forge\n",
      "cairo                     1.16.0            h6cf1ce9_1008    conda-forge\n",
      "certifi                   2022.12.7          pyhd8ed1ab_0    conda-forge\n",
      "cffi                      1.14.6           py39h4bc2ebd_1    conda-forge\n",
      "chardet                   4.0.0            py39hf3d152e_1    conda-forge\n",
      "charset-normalizer        2.0.0              pyhd8ed1ab_0    conda-forge\n",
      "click                     8.0.3            py39hf3d152e_0    conda-forge\n",
      "colorama                  0.4.4              pyh9f0ad1d_0    conda-forge\n",
      "cryptography              35.0.0           py39h95dcef6_1    conda-forge\n",
      "cycler                    0.10.0                     py_2    conda-forge\n",
      "cython                    0.29.24          py39he80948d_0    conda-forge\n",
      "debugpy                   1.4.1            py39he80948d_0    conda-forge\n",
      "decorator                 4.4.2                      py_0    conda-forge\n",
      "defusedxml                0.7.1              pyhd8ed1ab_0    conda-forge\n",
      "english                   2020.7.0           pyhd8ed1ab_0    conda-forge\n",
      "entrypoints               0.3             py39hde42818_1002    conda-forge\n",
      "et_xmlfile                1.0.1                   py_1001    conda-forge\n",
      "expat                     2.4.1                h9c3ff4c_0    conda-forge\n",
      "font-ttf-dejavu-sans-mono 2.37                 hab24e00_0    conda-forge\n",
      "font-ttf-inconsolata      3.000                h77eed37_0    conda-forge\n",
      "font-ttf-source-code-pro  2.038                h77eed37_0    conda-forge\n",
      "font-ttf-ubuntu           0.83                 hab24e00_0    conda-forge\n",
      "fontconfig                2.13.1            hba837de_1005    conda-forge\n",
      "fonts-conda-ecosystem     1                             0    conda-forge\n",
      "fonts-conda-forge         1                             0    conda-forge\n",
      "freetype                  2.10.4               h0708190_1    conda-forge\n",
      "fribidi                   1.0.10               h36c2ea0_0    conda-forge\n",
      "future                    0.18.2           py39hf3d152e_3    conda-forge\n",
      "gdk-pixbuf                2.42.6               h04a7f16_0    conda-forge\n",
      "gettext                   0.19.8.1          h73d1719_1008    conda-forge\n",
      "giflib                    5.2.1                h36c2ea0_2    conda-forge\n",
      "goatools                  1.2.3              pyh7cba7a3_2    bioconda\n",
      "graphite2                 1.3.13            h58526e2_1001    conda-forge\n",
      "graphviz                  2.49.1               h85b4f2f_0    conda-forge\n",
      "gtk2                      2.24.33              h539f30e_1    conda-forge\n",
      "gts                       0.7.6                h64030ff_2    conda-forge\n",
      "harfbuzz                  3.0.0                h83ec7ef_1    conda-forge\n",
      "hdmedians                 0.14.2           py39hce5d2b2_0    conda-forge\n",
      "icu                       68.1                 h58526e2_0    conda-forge\n",
      "idna                      3.1                pyhd3deb0d_0    conda-forge\n",
      "importlib-metadata        4.8.1            py39hf3d152e_0    conda-forge\n",
      "iniconfig                 1.1.1              pyh9f0ad1d_0    conda-forge\n",
      "interchange               2021.0.4           pyhd8ed1ab_0    conda-forge\n",
      "ipykernel                 6.4.2            py39hef51801_0    conda-forge\n",
      "ipython                   7.28.0           py39hef51801_0    conda-forge\n",
      "ipython_genutils          0.2.0                      py_1    conda-forge\n",
      "ipywidgets                7.6.5              pyhd8ed1ab_0    conda-forge\n",
      "jbig                      2.1               h7f98852_2003    conda-forge\n",
      "jedi                      0.18.0           py39hf3d152e_2    conda-forge\n",
      "jinja2                    3.0.2              pyhd8ed1ab_0    conda-forge\n",
      "jmespath                  0.10.0                   pypi_0    pypi\n",
      "joblib                    1.1.0              pyhd8ed1ab_0    conda-forge\n",
      "jpeg                      9d                   h36c2ea0_0    conda-forge\n",
      "jsonschema                4.1.2              pyhd8ed1ab_0    conda-forge\n",
      "jupyter_client            7.0.6              pyhd8ed1ab_0    conda-forge\n",
      "jupyter_core              4.8.1            py39hf3d152e_0    conda-forge\n",
      "jupyterlab_pygments       0.1.2              pyh9f0ad1d_0    conda-forge\n",
      "jupyterlab_widgets        1.0.2              pyhd8ed1ab_0    conda-forge\n",
      "kiwisolver                1.3.2            py39h1a9c180_0    conda-forge\n",
      "krb5                      1.19.2               hcc1bbae_2    conda-forge\n",
      "lcms2                     2.12                 hddcbb42_0    conda-forge\n",
      "ld_impl_linux-64          2.36.1               hea4e1c9_2    conda-forge\n",
      "lerc                      2.2.1                h9c3ff4c_0    conda-forge\n",
      "libblas                   3.9.0           12_linux64_openblas    conda-forge\n",
      "libcblas                  3.9.0           12_linux64_openblas    conda-forge\n",
      "libcurl                   7.79.1               h2574ce0_1    conda-forge\n",
      "libdeflate                1.7                  h7f98852_5    conda-forge\n",
      "libedit                   3.1.20191231         he28a2e2_2    conda-forge\n",
      "libev                     4.33                 h516909a_1    conda-forge\n",
      "libffi                    3.4.2                h9c3ff4c_4    conda-forge\n",
      "libgcc-ng                 11.2.0              h1d223b6_11    conda-forge\n",
      "libgd                     2.3.3                h6ad9fb6_0    conda-forge\n",
      "libgfortran-ng            11.2.0              h69a702a_11    conda-forge\n",
      "libgfortran5              11.2.0              h5c6108e_11    conda-forge\n",
      "libglib                   2.70.0               h174f98d_1    conda-forge\n",
      "libgomp                   11.2.0              h1d223b6_11    conda-forge\n",
      "libiconv                  1.16                 h516909a_0    conda-forge\n",
      "liblapack                 3.9.0           12_linux64_openblas    conda-forge\n",
      "libnghttp2                1.43.0               h812cca2_1    conda-forge\n",
      "libopenblas               0.3.18          pthreads_h8fe5266_0    conda-forge\n",
      "libpng                    1.6.37               h21135ba_2    conda-forge\n",
      "libpq                     13.3                 hd57d9b9_1    conda-forge\n",
      "librsvg                   2.52.2               hc3c00ef_0    conda-forge\n",
      "libsodium                 1.0.18               h36c2ea0_1    conda-forge\n",
      "libssh2                   1.10.0               ha56f1ee_2    conda-forge\n",
      "libstdcxx-ng              11.2.0              he4da1e4_11    conda-forge\n",
      "libtiff                   4.3.0                hf544144_1    conda-forge\n",
      "libtool                   2.4.6             h9c3ff4c_1008    conda-forge\n",
      "libuuid                   2.32.1            h7f98852_1000    conda-forge\n",
      "libwebp                   1.2.1                h3452ae3_0    conda-forge\n",
      "libwebp-base              1.2.1                h7f98852_0    conda-forge\n",
      "libxcb                    1.13              h7f98852_1003    conda-forge\n",
      "libxml2                   2.9.12               h72842e0_0    conda-forge\n",
      "libxslt                   1.1.33               h15afd5d_2    conda-forge\n",
      "libzlib                   1.2.11            h36c2ea0_1013    conda-forge\n",
      "lockfile                  0.12.2                     py_1    conda-forge\n",
      "lxml                      4.6.3            py39h107f48f_0    conda-forge\n",
      "lz4-c                     1.9.3                h9c3ff4c_1    conda-forge\n",
      "markupsafe                2.0.1            py39h3811e60_0    conda-forge\n",
      "matplotlib-base           3.4.3            py39h2fa2bec_1    conda-forge\n",
      "matplotlib-inline         0.1.3              pyhd8ed1ab_0    conda-forge\n",
      "matplotlib-venn           0.11.7             pyhd8ed1ab_0    conda-forge\n",
      "missingpy                 0.2.0                    pypi_0    pypi\n",
      "mistune                   0.8.4           py39h3811e60_1004    conda-forge\n",
      "more-itertools            8.10.0             pyhd8ed1ab_0    conda-forge\n",
      "mscorefonts               0.0.1                         3    conda-forge\n",
      "msgpack-python            1.0.2            py39h1a9c180_1    conda-forge\n",
      "natsort                   7.1.1              pyhd8ed1ab_0    conda-forge\n",
      "nbclient                  0.5.4              pyhd8ed1ab_0    conda-forge\n",
      "nbconvert                 6.2.0            py39hf3d152e_0    conda-forge\n",
      "nbformat                  5.1.3              pyhd8ed1ab_0    conda-forge\n",
      "ncurses                   6.2                  h58526e2_4    conda-forge\n",
      "neo4j-python-driver       4.3.7              pyhd8ed1ab_0    conda-forge\n",
      "neobolt                   1.7.17           py39h3811e60_2    conda-forge\n",
      "neotime                   1.7.4                      py_0    conda-forge\n",
      "nest-asyncio              1.5.1              pyhd8ed1ab_0    conda-forge\n",
      "networkx                  2.6.3              pyhd8ed1ab_0    conda-forge\n",
      "notebook                  6.4.5              pyha770c72_0    conda-forge\n",
      "numpy                     1.21.3           py39hdbf815f_0    conda-forge\n",
      "olefile                   0.46               pyh9f0ad1d_1    conda-forge\n",
      "openjpeg                  2.4.0                hb52868f_1    conda-forge\n",
      "openpyxl                  3.0.9              pyhd8ed1ab_0    conda-forge\n",
      "openssl                   1.1.1o               h166bdaf_0    conda-forge\n",
      "packaging                 21.0               pyhd8ed1ab_0    conda-forge\n",
      "pandas                    1.3.4            py39hde0f152_0    conda-forge\n",
      "pandoc                    2.15                 h7f98852_0    conda-forge\n",
      "pandocfilters             1.5.0              pyhd8ed1ab_0    conda-forge\n",
      "pango                     1.48.10              h54213e6_2    conda-forge\n",
      "parso                     0.8.2              pyhd8ed1ab_0    conda-forge\n",
      "patsy                     0.5.2              pyhd8ed1ab_0    conda-forge\n",
      "pcre                      8.45                 h9c3ff4c_0    conda-forge\n",
      "pdf2image                 1.16.2                   pypi_0    pypi\n",
      "pexpect                   4.8.0              pyh9f0ad1d_2    conda-forge\n",
      "pickleshare               0.7.5           py39hde42818_1002    conda-forge\n",
      "pillow                    8.3.2            py39ha612740_0    conda-forge\n",
      "pip                       21.3.1             pyhd8ed1ab_0    conda-forge\n",
      "pixman                    0.40.0               h36c2ea0_0    conda-forge\n",
      "plotly                    5.3.1              pyhd8ed1ab_0    conda-forge\n",
      "pluggy                    1.0.0            py39hf3d152e_1    conda-forge\n",
      "poppler                   21.09.0              h0e1ea10_2    conda-forge\n",
      "poppler-data              0.4.12               hd8ed1ab_0    conda-forge\n",
      "prometheus_client         0.11.0             pyhd8ed1ab_0    conda-forge\n",
      "prompt-toolkit            3.0.21             pyha770c72_0    conda-forge\n",
      "prompt_toolkit            3.0.21               hd8ed1ab_0    conda-forge\n",
      "psycopg2                  2.9.1            py39h3811e60_0    conda-forge\n",
      "pthread-stubs             0.4               h36c2ea0_1001    conda-forge\n",
      "ptyprocess                0.7.0              pyhd3deb0d_0    conda-forge\n",
      "py                        1.10.0             pyhd3deb0d_0    conda-forge\n",
      "py2neo                    2021.2.3           pyhd8ed1ab_0    conda-forge\n",
      "pycparser                 2.20               pyh9f0ad1d_2    conda-forge\n",
      "pydot                     1.4.2            py39hf3d152e_3    conda-forge\n",
      "pygments                  2.10.0             pyhd8ed1ab_0    conda-forge\n",
      "pygraphviz                1.7              py39h78163bd_0    conda-forge\n",
      "pyopenssl                 21.0.0             pyhd8ed1ab_0    conda-forge\n",
      "pyparsing                 2.4.7              pyh9f0ad1d_0    conda-forge\n",
      "pyrsistent                0.17.3           py39h3811e60_2    conda-forge\n",
      "pysam                     0.17.0           py39h051187c_0    bioconda\n",
      "pysocks                   1.7.1            py39hf3d152e_3    conda-forge\n",
      "pytabix                   0.1              py39h98c8e45_1    bioconda\n",
      "pytest                    6.2.5            py39hf3d152e_0    conda-forge\n",
      "python                    3.9.7           hb7a2778_3_cpython    conda-forge\n",
      "python-dateutil           2.8.2              pyhd8ed1ab_0    conda-forge\n",
      "python-graphviz           0.17               pyhaef67bd_0    conda-forge\n",
      "python-wget               3.2                        py_0    conda-forge\n",
      "python_abi                3.9                      2_cp39    conda-forge\n",
      "pytz                      2021.3             pyhd8ed1ab_0    conda-forge\n",
      "pyvcf                     0.6.8           py39hde42818_1002    conda-forge\n",
      "pyyaml                    6.0              py39h3811e60_0    conda-forge\n",
      "pyzmq                     22.3.0           py39h37b5a0c_0    conda-forge\n",
      "readline                  8.1                  h46c0cb4_0    conda-forge\n",
      "requests                  2.26.0             pyhd8ed1ab_0    conda-forge\n",
      "s3transfer                0.5.0                    pypi_0    pypi\n",
      "scikit-bio                0.5.6            py39h16ac069_4    conda-forge\n",
      "scikit-learn              1.0.1            py39h7c5d8c9_0    conda-forge\n",
      "scipy                     1.7.1            py39hee8e79c_0    conda-forge\n",
      "seaborn                   0.11.2               hd8ed1ab_0    conda-forge\n",
      "seaborn-base              0.11.2             pyhd8ed1ab_0    conda-forge\n",
      "send2trash                1.8.0              pyhd8ed1ab_0    conda-forge\n",
      "setuptools                58.2.0           py39hf3d152e_0    conda-forge\n",
      "six                       1.16.0             pyh6c4a22f_0    conda-forge\n",
      "sqlite                    3.36.0               h9cd32fc_2    conda-forge\n",
      "statsmodels               0.13.0           py39hce5d2b2_0    conda-forge\n",
      "tenacity                  8.0.1              pyhd8ed1ab_0    conda-forge\n",
      "terminado                 0.12.1           py39hf3d152e_0    conda-forge\n",
      "testpath                  0.5.0              pyhd8ed1ab_0    conda-forge\n",
      "threadpoolctl             3.0.0              pyh8a188c0_0    conda-forge\n",
      "tk                        8.6.11               h27826a3_1    conda-forge\n",
      "toml                      0.10.2             pyhd8ed1ab_0    conda-forge\n",
      "tornado                   6.1              py39h3811e60_1    conda-forge\n",
      "traitlets                 5.1.0              pyhd8ed1ab_0    conda-forge\n",
      "typing_extensions         3.10.0.2           pyha770c72_0    conda-forge\n",
      "tzdata                    2021e                he74cb21_0    conda-forge\n",
      "urllib3                   1.26.7             pyhd8ed1ab_0    conda-forge\n",
      "venn                      0.1.3                    pypi_0    pypi\n",
      "wcwidth                   0.2.5              pyh9f0ad1d_2    conda-forge\n",
      "webencodings              0.5.1                      py_1    conda-forge\n",
      "wheel                     0.37.0             pyhd8ed1ab_1    conda-forge\n",
      "widgetsnbextension        3.5.1            py39hf3d152e_4    conda-forge\n",
      "wordcloud                 1.8.1                    pypi_0    pypi\n",
      "xlrd                      2.0.1              pyhd8ed1ab_3    conda-forge\n",
      "xlsxwriter                3.0.3              pyhd8ed1ab_0    conda-forge\n",
      "xorg-kbproto              1.0.7             h7f98852_1002    conda-forge\n",
      "xorg-libice               1.0.10               h7f98852_0    conda-forge\n",
      "xorg-libsm                1.2.3             hd9c2040_1000    conda-forge\n",
      "xorg-libx11               1.7.2                h7f98852_0    conda-forge\n",
      "xorg-libxau               1.0.9                h7f98852_0    conda-forge\n",
      "xorg-libxdmcp             1.1.3                h7f98852_0    conda-forge\n",
      "xorg-libxext              1.3.4                h7f98852_1    conda-forge\n",
      "xorg-libxrender           0.9.10            h7f98852_1003    conda-forge\n",
      "xorg-renderproto          0.11.1            h7f98852_1002    conda-forge\n",
      "xorg-xextproto            7.3.0             h7f98852_1002    conda-forge\n",
      "xorg-xproto               7.0.31            h7f98852_1007    conda-forge\n",
      "xz                        5.2.5                h516909a_1    conda-forge\n",
      "yaml                      0.2.5                h516909a_0    conda-forge\n",
      "zeromq                    4.3.4                h9c3ff4c_1    conda-forge\n",
      "zipp                      3.6.0              pyhd8ed1ab_0    conda-forge\n",
      "zlib                      1.2.11            h36c2ea0_1013    conda-forge\n",
      "zstd                      1.5.0                ha95c52a_0    conda-forge\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scipy.stats as stats\n",
    "import matplotlib.pyplot as plt\n",
    "%matplotlib inline\n",
    "import seaborn as sns\n",
    "#For Arial font\n",
    "#!conda install -c conda-forge -y mscorefonts\n",
    "##-> The below was also needed in matplotlib 3.4.2\n",
    "#import shutil\n",
    "#import matplotlib\n",
    "#shutil.rmtree(matplotlib.get_cachedir())\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')\n",
    "from IPython.display import display\n",
    "import time\n",
    "#For exporting .pdf file with editable text\n",
    "import matplotlib\n",
    "matplotlib.rcParams['pdf.fonttype']=42\n",
    "matplotlib.rcParams['ps.fonttype']=42\n",
    "\n",
    "from itertools import combinations\n",
    "import math\n",
    "from multiprocessing import Pool\n",
    "import statsmodels.formula.api as smf\n",
    "from statsmodels.stats.anova import anova_lm\n",
    "from statsmodels.stats import weightstats\n",
    "from statsmodels.stats import multitest as multi\n",
    "from decimal import Decimal, ROUND_HALF_UP\n",
    "import re\n",
    "import matplotlib.patches as mpatches\n",
    "#!pip install venn\n",
    "from venn import venn\n",
    "from textwrap import wrap\n",
    "\n",
    "!conda list"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 0. DIRAC code"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> The original code for DIRAC was written in MATLAB. Hence, it is re-written in Python 3 here.  \n",
    "> <– I don't care about computational cost here; rather, the code adheres to the story in the original paper.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def network_ranking(DF, networkS):\n",
    "    # This function calculates the pairwise ordering of network genes (i.e., network ranking).\n",
    "    ## Ref. Eddy, J. A. et al. PLoS Comput. Biol. 2010 (Figure 1 at a glance)\n",
    "    # Requirements:\n",
    "    ## import numpy as np: confirmed with versions 1.17.5 and 1.21.1\n",
    "    ## import pandas as pd: confirmed with versions 0.25.3 and 1.3.1\n",
    "    ## from itertools import combinations: confirmed with Python 3.7.6 and 3.9.6\n",
    "    # Input:\n",
    "    ## DF: pd.DataFrame containing expression values (X_gn) with gene (g; 1-G) indices and sample (n; 1-N) columns\n",
    "    ## networkS: pd.Series containing genes (g; 1-G) with network (m; 1-M) indices (i.e., 1-on-1 long-format)\n",
    "    # Output:\n",
    "    ## pd.DataFrame containing binary values of the networking ranking comparison for each sample\n",
    "    ## -> row: comparison_id (indicated by network (m) and ordering (g_i < g_j) columns)\n",
    "    ## -> column: sample (n) (with exception of the NetworkID and Ordering columns)\n",
    "    # Note:\n",
    "    ## If items in network m and gene g contain ' : ' or ' < ', this code would produce error or unexpected output.\n",
    "    ## If an item in sample n is 'NetworkID' or 'Ordering', this code would produce error or unexpected output.\n",
    "    \n",
    "    #Calculate binary values of the network ranking comparison for each sample\n",
    "    rankDF = pd.DataFrame()\n",
    "    sampleL = DF.columns.tolist()\n",
    "    networkL = networkS.index.unique()\n",
    "    for n in sampleL:\n",
    "        rankDF_n = pd.DataFrame()\n",
    "        for m in networkL:\n",
    "            #Pairs of genes (g_i, g_j) in the network m\n",
    "            networkS_m = networkS.loc[m]\n",
    "            pairL_m = list(combinations(range(0, len(networkS_m)), 2))\n",
    "            pairL_m_i, pairL_m_j = [[pair[x] for pair in pairL_m] for x in (0,1)]\n",
    "            rankDF_m = pd.DataFrame({'g_i':networkS_m.iloc[pairL_m_i],\n",
    "                                     'g_j':networkS_m.iloc[pairL_m_j]})#Hold network (m; 1-M) index\n",
    "            #Compare the expression values (X_gn) between pairwise genes (g_i vs. g_j)\n",
    "            tempL = []\n",
    "            for pair_i in range(0, len(rankDF_m)):\n",
    "                g_i = rankDF_m.iloc[pair_i, 0]\n",
    "                g_j = rankDF_m.iloc[pair_i, 1]\n",
    "                X_i = DF.loc[g_i, n]\n",
    "                X_j = DF.loc[g_j, n]\n",
    "                #If X_i < X_j is true, add 1; otherwise (X_i >= X_j), add 0\n",
    "                if X_i < X_j:\n",
    "                    tempL.append(1)\n",
    "                else:\n",
    "                    tempL.append(0)\n",
    "            rankDF_m['X_i<X_j'] = tempL\n",
    "            #Update the network ranking dataframe of sample n\n",
    "            rankDF_m.index.set_names('NetworkID', inplace=True)#Set/reset index name\n",
    "            rankDF_m = rankDF_m.reset_index()\n",
    "            rankDF_n = pd.concat([rankDF_n, rankDF_m], axis=0)\n",
    "        #Updata the network ranking dataframe of all samples\n",
    "        rankDF_n['Sample'] = n\n",
    "        rankDF = pd.concat([rankDF, rankDF_n], axis=0)\n",
    "    ##Prepare dummy index and clean dataframe\n",
    "    rankDF['ComparisonID'] = rankDF['NetworkID'] + ' : ' + rankDF['g_i'] + ' < ' + rankDF['g_j']\n",
    "    rankDF = rankDF.pivot(index='ComparisonID', columns='Sample', values='X_i<X_j')#Sorted by index during this\n",
    "    rankDF = rankDF.reset_index()#Index becomes row number here\n",
    "    tempDF = rankDF['ComparisonID'].str.split(pat=' : ', expand=True)\n",
    "    tempDF = tempDF.rename(columns={0:'NetworkID', 1:'Ordering'})\n",
    "    rankDF = pd.concat([tempDF, rankDF], axis=1)#Dropping columns name 'Sample' during this\n",
    "    rankDF = rankDF.drop(columns='ComparisonID')\n",
    "    return rankDF\n",
    "\n",
    "def rank_template(rankDF, phenotypeS):\n",
    "    # This function generates the rank template (T) presenting the expected network ranking in a phenotype.\n",
    "    ## Ref. Eddy, J. A. et al. PLoS Comput. Biol. 2010 (Figure 1 at a galance)\n",
    "    # Requirements:\n",
    "    ## import numpy as np: confirmed with versions 1.17.5 and 1.21.1\n",
    "    ## import pandas as pd: confirmed with versions 0.25.3 and 1.3.1\n",
    "    # Input:\n",
    "    ## rankDF: pd.DataFrame obtained from the above network_ranking() function\n",
    "    ## phenotypeS: pd.Series containing phenotypes (k; 1-K) with sample (n; 1-N) indices (i.e., 1-on-1 long-format)\n",
    "    # Output:\n",
    "    ## pd.DataFrame containing the expected binary values of network ranking comparison for each phenotype (T_mk)\n",
    "    ## -> row: comparison_id (indicated by network (m) and ordering (g_i < g_j) columns)\n",
    "    ## -> column: phenotype (k) (with exception of the NetworkID and Ordering columns)\n",
    "    # Note:\n",
    "    ## True rate = 0.5 is assigned to 0 in this code.\n",
    "    ## If an item in phenotype k is 'NetworkID' or 'Ordering', this code would produce error or unexpected output.\n",
    "    \n",
    "    #Calculate the expected binary values of network ranking comparison for each phenotype (T_mk)\n",
    "    templateDF = rankDF[['NetworkID', 'Ordering']]\n",
    "    phenotypeL = phenotypeS.unique().tolist()\n",
    "    for k in phenotypeL:\n",
    "        sampleL_k = phenotypeS.loc[phenotypeS==k].index.tolist()\n",
    "        tempDF = rankDF[sampleL_k]\n",
    "        tempS = tempDF.mean(axis=1)#True (=1) rate\n",
    "        tempS = (tempS>0.5).astype('int64')#If true rate > 0.5, 1; otherwise (<= 0.5), 0\n",
    "        templateDF[k] = tempS\n",
    "    return templateDF\n",
    "\n",
    "def rank_matching_score(rankDF, templateDF):\n",
    "    # This function calculates the rank matching score (R) of a sample.\n",
    "    ## Ref. Eddy, J. A. et al. PLoS Comput. Biol. 2010 (Figure 1 at a galance)\n",
    "    # Requirements:\n",
    "    ## import numpy as np: confirmed with versions 1.17.5 and 1.21.1\n",
    "    ## import pandas as pd: confirmed with versions 0.25.3 and 1.3.1\n",
    "    # Input:\n",
    "    ## rankDF: pd.DataFrame obtained from the above network_ranking() function\n",
    "    ## templateDF: pd.DataFrame obtained from the above rank_template() function\n",
    "    # Output:\n",
    "    ## pd.DataFrame containing the rates of gene pairs matching to a rank template in a sample (R_mkn)\n",
    "    ## -> row: rank template (T_mk) (indicated by network (m) and template phenotype (k) columns)\n",
    "    ## -> column: sample (n) (with exception of the NetworkID and Template columns)\n",
    "    # Note:\n",
    "    ## True rate = 0.5 was assigned to 0 in the above rank_template() function.\n",
    "    ## -> 'Match (1)' and 'Mismatch (0)' are evenly assigned to samples in the phenotype.\n",
    "    ##    (i.e., 'Match' for (X_i < X_j) = 0 and 'Mismatch' for (X_i < X_j) = 1 in the tie case)\n",
    "    ## If items in network m and phenotype k contain ' : ', this code would produce error or unexpected output.\n",
    "    \n",
    "    #Calculate the rates of gene pairs matching to a rank template in a sample (R_mkn)\n",
    "    scoreDF = pd.DataFrame()\n",
    "    sampleL = rankDF.drop(columns=['NetworkID', 'Ordering']).columns.tolist()\n",
    "    phenotypeL = templateDF.drop(columns=['NetworkID', 'Ordering']).columns.tolist()\n",
    "    for n in sampleL:\n",
    "        scoreDF_n = pd.DataFrame()\n",
    "        for k_template in phenotypeL:\n",
    "            tempDF = rankDF[['NetworkID', 'Ordering']]\n",
    "            tempS = (rankDF[n]==templateDF[k_template]).astype('int64')#If matching, 1; otherwise, 0\n",
    "            tempDF['Match'] = tempS\n",
    "            #Calculate the rank matching score\n",
    "            scoreDF_k = tempDF.groupby(by='NetworkID', as_index=False, sort=False).mean()\n",
    "            #Update the rank matching score dataframe of sample n\n",
    "            scoreDF_k['Template'] = k_template\n",
    "            scoreDF_k = scoreDF_k[['NetworkID', 'Template', 'Match']]\n",
    "            scoreDF_n = pd.concat([scoreDF_n, scoreDF_k], axis=0)\n",
    "        #Update the rank matching score dataframe of all samples\n",
    "        scoreDF_n['Sample'] = n\n",
    "        scoreDF = pd.concat([scoreDF, scoreDF_n], axis=0)\n",
    "    ##Prepare dummy index and clean dataframe\n",
    "    scoreDF['RMSmkID'] = scoreDF['NetworkID'].str.cat(scoreDF['Template'], sep=' : ')\n",
    "    scoreDF = scoreDF.pivot(index='RMSmkID', columns='Sample', values='Match')#Sorted by index during this\n",
    "    scoreDF = scoreDF.reset_index()#Index becomes row number here\n",
    "    tempDF = scoreDF['RMSmkID'].str.split(pat=' : ', expand=True)\n",
    "    tempDF = tempDF.rename(columns={0:'NetworkID', 1:'Template'})\n",
    "    scoreDF = pd.concat([tempDF, scoreDF], axis=1)#Dropping columns name 'Sample' during this\n",
    "    scoreDF = scoreDF.drop(columns='RMSmkID')\n",
    "    return scoreDF\n",
    "\n",
    "def rank_conservation_index(scoreDF, phenotypeS):\n",
    "    # This function calculates the rank conservation index (muR) of a phenotype.\n",
    "    ## Ref. Eddy, J. A. et al. PLoS Comput. Biol. 2010 (Figure 1 at a galance)\n",
    "    # Requirements:\n",
    "    ## import numpy as np: confirmed with versions 1.17.5 and 1.21.1\n",
    "    ## import pandas as pd: confirmed with versions 0.25.3 and 1.3.1\n",
    "    # Input:\n",
    "    ## scoreDF: pd.DataFrame obtained from the above rank_matching_score() function\n",
    "    ## phenotypeS: pd.Series containing phenotypes (k; 1-K) with sample (n; 1-N) indices (i.e., 1-on-1 long-format)\n",
    "    # Output:\n",
    "    ## pd.DataFrame containing the mean values of rank matching scores (RMSs) in a phenotype (muR_mkk)\n",
    "    ## -> row: rank template (T_mk) (indicated by network (m) and template phenotype (k) columns)\n",
    "    ## -> column: phenotype (k) (with exception of the NetworkID and Template columns)\n",
    "    # Note:\n",
    "    ## Rank conservation index (RCI) is used with a broader stance here, but following a strict interpretation,\n",
    "    ## the term RCI is used for the mean of RMSs only when template phenotype is same with the sample phenotype.\n",
    "    ## If an item in phenotype k is 'NetworkID' or 'Template', this code would produce error or unexpected output.\n",
    "    \n",
    "    #Calculate the mean values of RMSs in a phenotype (muR_mkk)\n",
    "    conservationDF = scoreDF[['NetworkID', 'Template']]\n",
    "    phenotypeL = phenotypeS.unique().tolist()\n",
    "    for k in phenotypeL:\n",
    "        sampleL_k = phenotypeS.loc[phenotypeS==k].index.tolist()\n",
    "        tempS = scoreDF[sampleL_k].mean(axis=1)\n",
    "        conservationDF[k] = tempS\n",
    "    return conservationDF"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. Prepare dataset and metadata"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1-1. Analyte data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Import analyte data\n",
    "fileDir = './ExportData/'\n",
    "ipynbName = '230213_LC-M001-proteomics-DIRAC-ver7-2_Preprocessing_'\n",
    "fileName = 'normalized-data.tsv'\n",
    "tempDF = pd.read_csv(fileDir+ipynbName+fileName, sep='\\t').set_index('UniProtID')\n",
    "display(tempDF)\n",
    "\n",
    "analyteDF = tempDF"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1-2. Module–analyte metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Import module-analyte metadata\n",
    "fileDir = '../210126_LCproteomics-M001-DIRAC-ver6/ExportData/'\n",
    "ipynbName = '220427_LCproteomics-M001-DIRAC-ver6-2_Preprocessing_'\n",
    "fileName = 'module-metadata_QuickGO-GOBP-min-n4-cov50.tsv'\n",
    "tempDF = pd.read_csv(fileDir+ipynbName+fileName, sep='\\t').set_index('ModuleID')\n",
    "print(' - Unique analytes with module:', len(tempDF['UniProtID'].unique()))\n",
    "print(' - Unique modules with analytes:', len(tempDF.index.unique()))\n",
    "\n",
    "#Prepare moduleS\n",
    "moduleS = tempDF['UniProtID']\n",
    "display(moduleS)\n",
    "\n",
    "#Retrieve module metadata\n",
    "tempDF = tempDF[['ModuleName', 'nAnalytes', 'nBackgrounds', 'Coverage']]\n",
    "moduleDF = tempDF.reset_index().drop_duplicates(keep='first').set_index('ModuleID')\n",
    "display(moduleDF)\n",
    "display(moduleDF.describe(include='all'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> –> Add the mapped analytes to module metadata.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Import analyte metadata\n",
    "fileDir = '../210126_LCproteomics-M001-DIRAC-ver6/ExportData/'\n",
    "ipynbName = '220427_LCproteomics-M001-DIRAC-ver6-2_Preprocessing_'\n",
    "fileName = 'analyte-metadata_UniProt.tsv'\n",
    "tempDF = pd.read_csv(fileDir+ipynbName+fileName, sep='\\t')\n",
    "tempDF = pd.merge(moduleS.reset_index(), tempDF, on='UniProtID', how='left').set_index('ModuleID')\n",
    "display(tempDF)\n",
    "print(' - Unique analytes:', len(tempDF['UniProtID'].unique()))\n",
    "print(' - Unique labels:', len(tempDF['GeneLabel'].unique()))\n",
    "\n",
    "#Concatenate labels\n",
    "t_start = time.time()\n",
    "moduleDF['MappedAnalyteIDs'] = ''#Initialize\n",
    "moduleDF['MappedAnalyteGeneLabels'] = ''#Initialize\n",
    "for module in moduleDF.index.tolist():\n",
    "    tempS = tempDF['UniProtID'].loc[tempDF.index.isin([module])]#Retrieve as pd.Series\n",
    "    label = tempS.str.cat(sep=';')\n",
    "    moduleDF.loc[module, 'MappedAnalyteIDs'] = label\n",
    "    tempS = tempDF['GeneLabel'].loc[tempDF.index.isin([module])]#Retrieve as pd.Series\n",
    "    label = tempS.str.cat(sep=';')\n",
    "    moduleDF.loc[module, 'MappedAnalyteGeneLabels'] = label\n",
    "t_elapsed = time.time() - t_start\n",
    "print('Elapsed time:', round(t_elapsed//60), 'min', round(t_elapsed%60, 1), 'sec')\n",
    "\n",
    "#Clean\n",
    "moduleDF['ModuleType'] = 'Gene Ontology (Biological Process)'\n",
    "moduleDF['Source'] = 'EMBL-EBI QuickGO API'\n",
    "moduleDF = moduleDF[['ModuleName', 'ModuleType', 'MappedAnalyteIDs', 'MappedAnalyteGeneLabels',\n",
    "                     'nAnalytes', 'nBackgrounds', 'Coverage', 'Source']]\n",
    "moduleDF = moduleDF.sort_index(ascending=True)\n",
    "\n",
    "#Save\n",
    "fileDir = './ExportData/'\n",
    "ipynbName = '230214_LC-M001-proteomics-DIRAC-ver7-2_DIRAC-GOBP_'\n",
    "fileName = 'module-metadata.tsv'\n",
    "moduleDF.to_csv(fileDir+ipynbName+fileName, index=True, sep='\\t')\n",
    "\n",
    "display(moduleDF)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1-3. Sample–mouse metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Import sample-mouse metadata\n",
    "fileDir = './ExportData/'\n",
    "ipynbName = '230213_LC-M001-proteomics-DIRAC-ver7-2_Preprocessing_'\n",
    "fileName = 'sample-metadata.tsv'\n",
    "tempDF = pd.read_csv(fileDir+ipynbName+fileName, sep='\\t')\n",
    "tempDF = tempDF.loc[tempDF['SampleID'].isin(analyteDF.columns)]\n",
    "tempDF = tempDF.set_index('SampleID')\n",
    "\n",
    "#Prepare phenotypeS\n",
    "tempDF['Phenotype'] = tempDF['Intervention']#Sex-pooled\n",
    "display(tempDF)\n",
    "display(tempDF['Phenotype'].value_counts())\n",
    "\n",
    "sampleDF = tempDF"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. Perform DIRAC with sex-pooled rank consensus"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> When applied to overall dataset, the DIRAC code often stops runnning due to memory limit (e.g., transcriptomics).  \n",
    "> –> Divide moduleS into subsets while considering the number of comparisons, and compute DIRAC in parallel.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Divide moduleS\n",
    "cutoff = 250#The maximum number of analytes per module\n",
    "tempDF = moduleDF.sort_values(by='nAnalytes', ascending=True)\n",
    "nwsubL = []\n",
    "tempL = []#Initialize\n",
    "count = 0#Initialize\n",
    "for module in tempDF.index.tolist():\n",
    "    nanalytes = tempDF.loc[module, 'nAnalytes']\n",
    "    if nanalytes>cutoff:\n",
    "        nwsubL.append([module])\n",
    "    else:\n",
    "        tempL.append(module)\n",
    "        count += nanalytes\n",
    "        if count>cutoff:\n",
    "            nwsubL.append(tempL)\n",
    "            tempL = []#Initialize\n",
    "            count = 0#Initialize\n",
    "if len(tempL)>0:#The last one but still count<=cutoff\n",
    "    nwsubL.append(tempL)\n",
    "nSub = len(nwsubL)\n",
    "print('nSublists: ', nSub)\n",
    "print('nModules per sublist:', [len(sublist) for sublist in nwsubL])\n",
    "\n",
    "#Check nAnalytes distribution\n",
    "tempS = pd.Series(name='Subset')\n",
    "for list_i in range(nSub):\n",
    "    tempL = moduleS.loc[nwsubL[list_i]].index.unique().tolist()\n",
    "    tempS1 = pd.Series(np.repeat('Subset '+str(list_i+1).zfill(3), len(tempL)), index=tempL, name='Subset')\n",
    "    tempS = pd.concat([tempS, tempS1], axis=0)\n",
    "tempS.index.set_names('ModuleID', inplace=True)\n",
    "tempDF = pd.merge(moduleDF, tempS, left_index=True, right_index=True, how='left')\n",
    "tempDF = tempDF.sort_values(by='nAnalytes', ascending=False)\n",
    "display(tempDF)\n",
    "tempDF = tempDF.groupby('Subset').agg({'nAnalytes':[len, sum, max, np.median]})\n",
    "tempDF = tempDF.sort_values(by=('nAnalytes', 'sum'), ascending=False)\n",
    "print('Subset summary:')\n",
    "display(tempDF.describe())\n",
    "print(' -> Check subset which will need high computational cost:')\n",
    "display(tempDF.loc[tempDF[('nAnalytes', 'max')]>cutoff])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "nprocessors = 4\n",
    "fileDir = './ExportData/'\n",
    "ipynbName = '230214_LC-M001-proteomics-DIRAC-ver7-2_DIRAC-GOBP_'\n",
    "\n",
    "#Wrap as a single function\n",
    "def parallel_dirac(list_i):\n",
    "    tempS = moduleS.loc[nwsubL[list_i]]\n",
    "    \n",
    "    #Calculate the pairwise ordering of network genes (i.e., network ranking)\n",
    "    t_start1 = time.time()\n",
    "    rankingDF = network_ranking(analyteDF, tempS)\n",
    "    t_elapsed1 = time.time() - t_start1\n",
    "    \n",
    "    #Generate the rank template (T) presenting the expected network ranking in a phenotype\n",
    "    t_start2 = time.time()\n",
    "    templateDF = rank_template(rankingDF, sampleDF['Phenotype'])\n",
    "    t_elapsed2 = time.time() - t_start2\n",
    "    \n",
    "    #Calculate the rank matching score (R) of a sample\n",
    "    t_start3 = time.time()\n",
    "    rmsDF = rank_matching_score(rankingDF, templateDF)\n",
    "    t_elapsed3 = time.time() - t_start3\n",
    "    \n",
    "    #Calculate the rank conservation index (muR) of a phenotype\n",
    "    t_start4 = time.time()\n",
    "    rciDF = rank_conservation_index(rmsDF, sampleDF['Phenotype'])\n",
    "    t_elapsed4 = time.time() - t_start4\n",
    "    \n",
    "    #Save\n",
    "    fileName = 'NetworkRanking-'+str(list_i+1).zfill(3)+'.tsv'\n",
    "    rankingDF.to_csv(fileDir+ipynbName+fileName, index=False, sep='\\t')\n",
    "    fileName = 'RankTemplate-BS-'+str(list_i+1).zfill(3)+'.tsv'\n",
    "    templateDF.to_csv(fileDir+ipynbName+fileName, index=False, sep='\\t')\n",
    "    fileName = 'RankMatchingScore-BS-'+str(list_i+1).zfill(3)+'.tsv'\n",
    "    rmsDF.to_csv(fileDir+ipynbName+fileName, index=False, sep='\\t')\n",
    "    fileName = 'RankConservationIndex-BS-'+str(list_i+1).zfill(3)+'.tsv'\n",
    "    rciDF.to_csv(fileDir+ipynbName+fileName, index=False, sep='\\t')\n",
    "    \n",
    "    #Check results\n",
    "    print('Subset '+str(list_i+1).zfill(3)+':',\n",
    "          len(tempS.index.unique()), 'modules,', len(tempS.unique()), 'analytes')\n",
    "    print(' - Network ranking dataframe:')\n",
    "    print('   - DF shape:', rankingDF.shape)\n",
    "    print('   - Elapsed time:', round(t_elapsed1//60), 'min', round(t_elapsed1%60, 1), 'sec')\n",
    "    print(' - Rank template dataframe:')\n",
    "    print('   - DF shape:', templateDF.shape)\n",
    "    print('   - Elapsed time:', round(t_elapsed2//60), 'min', round(t_elapsed2%60, 1), 'sec')\n",
    "    print(' - Rank matching score dataframe:')\n",
    "    print('   - DF shape:', rmsDF.shape)\n",
    "    print('   - Elapsed time:', round(t_elapsed3//60), 'min', round(t_elapsed3%60, 1), 'sec')\n",
    "    print(' - Rank conservation index dataframe:')\n",
    "    print('   - DF shape:', rciDF.shape)\n",
    "    print('   - Elapsed time:', round(t_elapsed4//60), 'min', round(t_elapsed4%60, 1), 'sec')\n",
    "    print('')\n",
    "\n",
    "#Parallel computing\n",
    "if __name__=='__main__':\n",
    "    t_start = time.time()\n",
    "    p = Pool(nprocessors)\n",
    "    p.map(parallel_dirac, range(nSub))\n",
    "    t_finish = time.time()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Record as reference\n",
    "print(nSub, 'sublists with', nprocessors, 'processors')\n",
    "print(' - Start:', time.ctime(t_start))\n",
    "print(' - Finish:', time.ctime(t_finish))\n",
    "t_elapsed = (t_finish - t_start)\n",
    "print(' - Elapsed time:', round(t_elapsed//60), 'min', round(t_elapsed%60, 1), 'sec')\n",
    "t_elapsed = (t_finish - t_start) * nprocessors\n",
    "print(' - Total (approximate) elapsed time:', round(t_elapsed//60), 'min', round(t_elapsed%60, 1), 'sec')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Combine each result\n",
    "rmsDF = pd.DataFrame()\n",
    "rciDF = pd.DataFrame()\n",
    "for list_i in range(nSub):\n",
    "    fileName = 'RankMatchingScore-BS-'+str(list_i+1).zfill(3)+'.tsv'\n",
    "    tempDF = pd.read_csv(fileDir+ipynbName+fileName, sep='\\t')\n",
    "    rmsDF = pd.concat([rmsDF, tempDF], axis=0, ignore_index=True)\n",
    "    \n",
    "    fileName = 'RankConservationIndex-BS-'+str(list_i+1).zfill(3)+'.tsv'\n",
    "    tempDF = pd.read_csv(fileDir+ipynbName+fileName, sep='\\t')\n",
    "    rciDF = pd.concat([rciDF, tempDF], axis=0, ignore_index=True)\n",
    "\n",
    "#Clean\n",
    "tempD1 = {'RMS':rmsDF, 'RCI':rciDF}\n",
    "tempD2 = {}\n",
    "for metric in tempD1.keys():\n",
    "    tempDF = tempD1[metric]\n",
    "    tempDF = tempDF.rename(columns={'NetworkID':'ModuleID'})\n",
    "    tempD2[metric] = tempDF\n",
    "    \n",
    "    print(metric+' dataframe:', tempDF.shape)\n",
    "    print(' - Unique modules:', len(tempDF['ModuleID'].unique()))\n",
    "    print(' - Unique templates:', len(tempDF['Template'].unique()))\n",
    "    display(tempDF)\n",
    "    print('')\n",
    "##Update\n",
    "rmsDF = tempD2['RMS']\n",
    "rciDF = tempD2['RCI']\n",
    "\n",
    "#Save\n",
    "fileName = 'RankMatchingScore-BS-combined.tsv'\n",
    "rmsDF.to_csv(fileDir+ipynbName+fileName, index=False, sep='\\t')\n",
    "fileName = 'RankConservationIndex-BS-combined.tsv'\n",
    "rciDF.to_csv(fileDir+ipynbName+fileName, index=False, sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3. Rank conservation index: general pattern"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3-1. Extract RCI (the mean of RMSs under the own phenotype consensus)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Extract RCI whose template phenotype corresponds to the own phenotype\n",
    "phenotypeL = rciDF.drop(columns=['ModuleID', 'Template']).columns.tolist()\n",
    "rciDF_kk = pd.DataFrame(index=pd.Index(rciDF['ModuleID'].unique(), name='ModuleID'))\n",
    "tempDF = rciDF.set_index('ModuleID')\n",
    "for k in phenotypeL:\n",
    "    tempS = tempDF[k].loc[tempDF['Template']==k]\n",
    "    rciDF_kk = pd.merge(rciDF_kk, tempS, left_index=True, right_index=True, how='left')\n",
    "##Sort\n",
    "tempL = ['Ctrl', 'Aca', '17aE2', 'Rapa']\n",
    "rciDF_kk = rciDF_kk[tempL]\n",
    "display(rciDF_kk)\n",
    "display(rciDF_kk.describe())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3-2. Mann–Whitney U-test"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> Note that the scipy API (scipy.stats.mannwhitneyu) is used, because only the one-sided test seems implemented in the current statsmodels API (statsmodels.stats.nonparametric.rank_compare_2indep). Actually, the output objects are same b/w the two APIs, which is contrast to the case of t-test (degrees of freedom is not reported in the scipy API).  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tempDF = rciDF_kk\n",
    "tempL = ['Ctrl', 'Aca', '17aE2', 'Rapa']\n",
    "control = 'Ctrl'\n",
    "\n",
    "#Statistical tests\n",
    "tempDF1 = pd.DataFrame(columns=['Ustat', 'Pval'])\n",
    "for contrast in tempL:\n",
    "    if control!=contrast:\n",
    "        tempS1 = tempDF[control]\n",
    "        tempS2 = tempDF[contrast]\n",
    "        #Two-sided Mann–Whitney U-test\n",
    "        ustat, pval = stats.mannwhitneyu(tempS2, tempS1,#U-statistic corresponds to the contrast\n",
    "                                         use_continuity=True, alternative='two-sided', method='auto')\n",
    "        tempDF1.loc[contrast+'-vs-'+control] = [ustat, pval]\n",
    "##P-value adjustment by using Benjamini–Hochberg method\n",
    "tempDF1['AdjPval'] = multi.multipletests(tempDF1['Pval'], alpha=0.05, method='fdr_bh',\n",
    "                                         is_sorted=False, returnsorted=False)[1]\n",
    "tempDF1.index.rename('ComparisonLabel', inplace=True)\n",
    "display(tempDF1)\n",
    "\n",
    "#Calculate general statistics\n",
    "tempDF2 = pd.DataFrame(columns=['N', 'RCImedian', 'RCImad'])\n",
    "for group in tempL:\n",
    "    tempS = tempDF[group]\n",
    "    size = len(tempS)\n",
    "    median = tempS.median()\n",
    "    mad = stats.median_absolute_deviation(tempS)#Cf. pd.Series.mad() is not median absolute deviation but mean absolute deviation\n",
    "    tempDF2.loc[group] = [size, median, mad]\n",
    "tempDF2.index.rename('GroupLabel', inplace=True)\n",
    "display(tempDF2)\n",
    "\n",
    "#Clean\n",
    "##Reformat while renaming column names\n",
    "tempD1 = {'Comparison':tempDF1, 'Group':tempDF2}\n",
    "tempD2 = {}\n",
    "for target in tempD1.keys():\n",
    "    tempDF3 = tempD1[target]\n",
    "    tempL1 = tempDF3.index.tolist()#For sorting later\n",
    "    tempL2 = tempDF3.columns.tolist()#For sorting later\n",
    "    tempDF3 = tempDF3.reset_index().melt(var_name='Variable', value_name='Value', id_vars=target+'Label')\n",
    "    tempDF3['Variable'] = tempDF3[target+'Label']+'_'+tempDF3['Variable']\n",
    "    tempDF3['ModuleID'] = 'All'#Dummy\n",
    "    tempDF3 = tempDF3.pivot(index='ModuleID', columns='Variable', values='Value')\n",
    "    tempDF3.columns.name = None#Erase 'Variable'\n",
    "    tempL3 = [label+'_'+variable for label in tempL1 for variable in tempL2]\n",
    "    tempDF3 = tempDF3[tempL3]#Sort\n",
    "    tempD2[target] = tempDF3\n",
    "##Merge\n",
    "tempDF3 = pd.merge(tempD2['Group'], tempD2['Comparison'], left_index=True, right_index=True, how='inner')\n",
    "##Convert data type\n",
    "for col_n in tempDF3.columns.tolist():\n",
    "    if re.search('_N$', col_n):\n",
    "        tempDF3[col_n] = tempDF3[col_n].astype(int)\n",
    "#Add dummy module name\n",
    "tempL1 = tempDF3.columns.tolist()#For sorting later\n",
    "tempDF3['ModuleName'] = 'General pattern'\n",
    "tempL2 = [col_n for sublist in [['ModuleName'], tempL1] for col_n in sublist]\n",
    "tempDF3 = tempDF3[tempL2]\n",
    "display(tempDF3)\n",
    "\n",
    "#Save\n",
    "fileDir = './ExportData/'\n",
    "ipynbName = '230214_LC-M001-proteomics-DIRAC-ver7-2_DIRAC-GOBP_'\n",
    "fileName = 'inter-group-comparison_RCIdistribution.tsv'\n",
    "tempDF3.to_csv(fileDir+ipynbName+fileName, sep='\\t', index=True)\n",
    "\n",
    "statDF = tempDF3"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3-3. Visualization: boxplot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#Prepare DF and color\n",
    "tempD0 = {'Ctrl':'Control', 'Aca':'Acarbose',\n",
    "          '17aE2':'17'+r'$\\alpha$'+'-Estradiol', 'Rapa':'Rapamycin'}\n",
    "tempDF = rciDF_kk.rename(columns=tempD0)\n",
    "tempDF = tempDF.reset_index().melt(var_name='Group', value_name='RCI', id_vars='ModuleID')\n",
    "tempD = {'Control':'tab:blue', 'Acarbose':'tab:red',\n",
    "         '17'+r'$\\alpha$'+'-Estradiol':'tab:green', 'Rapamycin':'tab:purple'}\n",
    "\n",
    "#Prepare significance labels\n",
    "##Retrieve statistical significance\n",
    "module = 'All'\n",
    "tempS = statDF.loc[module, statDF.columns.str.contains('AdjPval')]\n",
    "tempS.index = tempS.index.str.replace('_AdjPval', '')\n",
    "tempS.name = 'AdjPval'\n",
    "##Clean\n",
    "tempDF1 = tempS.index.to_series().str.split(pat='-vs-', expand=True)\n",
    "tempDF1 = tempDF1.rename(columns={0:'Contrast', 1:'Baseline'})\n",
    "tempDF1 = pd.merge(tempDF1, tempS, left_index=True, right_index=True, how='left')\n",
    "tempDF1['Contrast'] = tempDF1['Contrast'].map(tempD0)\n",
    "tempDF1['Baseline'] = tempDF1['Baseline'].map(tempD0)\n",
    "##Convert p-value to label\n",
    "tempL = []\n",
    "for row_i in range(len(tempDF1)):\n",
    "    pval = tempDF1['AdjPval'].iloc[row_i]\n",
    "    if pval<0.001:\n",
    "        tempL.append('***')\n",
    "    elif pval<0.01:\n",
    "        tempL.append('**')\n",
    "    elif pval<0.05:\n",
    "        tempL.append('*')\n",
    "    else:\n",
    "        pval_text = Decimal(str(pval)).quantize(Decimal('0.01'), rounding=ROUND_HALF_UP)\n",
    "        tempL.append(r'$P$ = '+str(pval_text))\n",
    "tempDF1['SignifLabel'] = tempL\n",
    "display(tempDF1)\n",
    "\n",
    "#Visualization\n",
    "ymax = 1.0\n",
    "ymin = 0.5\n",
    "yinter = 0.1\n",
    "ymargin_t = 0.06\n",
    "ymargin_b = 0.01\n",
    "aline_ymin = 0.9\n",
    "aline_ymargin = 0.05\n",
    "sns.set(style='ticks', font='Arial', context='talk')\n",
    "plt.figure(figsize=(2, 4))\n",
    "p = sns.boxplot(data=tempDF, y='RCI', x='Group', order=list(tempD.keys()), palette=tempD, dodge=False,\n",
    "                showfliers=True, flierprops={'marker':'o', 'markerfacecolor':'gray', 'alpha':0.4},\n",
    "                showcaps=True, notch=True)\n",
    "p.set(ylim=(ymin-ymargin_b, ymax+ymargin_t), yticks=np.arange(ymin, ymax + yinter/10, yinter))\n",
    "##Add significance labels\n",
    "lines = p.axes.get_lines()#Line2D: [[Q1, Q1-1.5IQR], [Q3, Q3+1.5IQR], [Q1, Q1], [Q3, Q3], [Med, Med], [flier]]\n",
    "lines_unit = 5 + int(True)#showfliers=True\n",
    "for row_i in range(len(tempDF1)):\n",
    "    #Baseline\n",
    "    group_0 = tempDF1['Baseline'].iloc[row_i]\n",
    "    index_0 = list(tempD.keys()).index(group_0)\n",
    "    whisker_0 = lines[index_0*lines_unit + 1]\n",
    "    xcoord_0 = whisker_0._x[1]#Q3+1.5IQR\n",
    "    #ycoord_0 = whisker_0._y[1]#Q3+1.5IQR\n",
    "    #Contrast\n",
    "    group_1 = tempDF1['Contrast'].iloc[row_i]\n",
    "    index_1 = list(tempD.keys()).index(group_1)\n",
    "    whisker_1 = lines[index_1*lines_unit + 1]\n",
    "    xcoord_1 = whisker_1._x[1]#Q3+1.5IQR\n",
    "    #ycoord_1 = whisker_1._y[1]#Q3+1.5IQR\n",
    "    #Standard point of marker\n",
    "    xcoord = (xcoord_0+xcoord_1)/2\n",
    "    #ycoord = max(ycoord_0, ycoord_1)\n",
    "    ycoord = aline_ymin + aline_ymargin*row_i\n",
    "    label = tempDF1['SignifLabel'].iloc[row_i]\n",
    "    #Add annotation lines\n",
    "    aline_offset = yinter/10\n",
    "    aline_length = yinter/10 + aline_offset\n",
    "    plt.plot([xcoord_0, xcoord_0, xcoord_1, xcoord_1],\n",
    "             [ycoord+aline_offset, ycoord+aline_length, ycoord+aline_length, ycoord+aline_offset],\n",
    "             lw=1.5, c='k')\n",
    "    #Add annotation text\n",
    "    if label in ['***', '**', '*']:\n",
    "        text_offset = yinter/25\n",
    "        p.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                   horizontalalignment='center', verticalalignment='bottom',\n",
    "                   fontsize='medium', color='k')\n",
    "    else:\n",
    "        text_offset = yinter/5\n",
    "        p.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                   horizontalalignment='center', verticalalignment='bottom',\n",
    "                   fontsize='x-small', color='k')\n",
    "sns.despine()\n",
    "plt.xlabel('')\n",
    "plt.ylabel('Module RCI')\n",
    "plt.xticks(rotation=70, horizontalalignment='right', verticalalignment='center', rotation_mode='anchor')\n",
    "##Save\n",
    "fileDir = './ExportFigures/'\n",
    "ipynbName = '230214_LC-M001-proteomics-DIRAC-ver7-2_DIRAC-GOBP_'\n",
    "fileName = 'RCI-boxplot.pdf'\n",
    "plt.gcf().savefig(fileDir+ipynbName+fileName, dpi=300, bbox_inches='tight', pad_inches=0.04, transparent=True)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4. Rank conservation index: inter-group module comparison"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> Test specific hypothesis: control RCI == intervention RCI (i.e., inter-group module comparison).  \n",
    "> 1. Testing the main effect of intervention on rank mathing scores (RMSs) for each module using ANOVA model  \n",
    "> 2. Then, performing post-hoc comparisons of RMSs between control vs. each intervention using Welch's t-tests  \n",
    ">  \n",
    "> Since RCI is the mean of RMSs under the own phenotype consensus, not RCI but RMS is practically used to obtain variance around group consensus. As well as using sex-pooled rank consensus, sex is NOT included in the ANOVA model. Note that the dataset was already adjusted with sex and age in this version 7. Although tricky, the P-value adjustment in (1) is performed across modules under a conservative assumption that modules are completely independent. Because the post-hoc comparisons (2) are to address the effect of each intervention within a specific module, the P-values are adjusted across interventions only within the module (not across modules).  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 4-1. Extract RMS under the own phenotype consensus"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Extract RMS whose template phenotype corresponds to the own phenotype\n",
    "phenotypeL = rciDF.drop(columns=['ModuleID', 'Template']).columns.tolist()\n",
    "rmsDF_kk = pd.DataFrame(index=pd.Index(rmsDF['ModuleID'].unique(), name='ModuleID'))\n",
    "tempDF = rmsDF.set_index('ModuleID')\n",
    "for k in phenotypeL:\n",
    "    tempL = sampleDF.loc[sampleDF['Phenotype']==k].index.tolist()\n",
    "    tempDF1 = tempDF[tempL].loc[tempDF['Template']==k]\n",
    "    rmsDF_kk = pd.merge(rmsDF_kk, tempDF1, left_index=True, right_index=True, how='left')\n",
    "\n",
    "display(rmsDF_kk)\n",
    "display(rmsDF_kk.describe())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 4-2. ANOVA test (RMS ~ Intervention), followed by Welch's t-tests (Intervention)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 4-2-1. Perform all statistical tests"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tempL1 = ['Ctrl', 'Aca', '17aE2', 'Rapa']#Target sample groups to be assessed\n",
    "control = 'Ctrl'#For post-hoc comparisons\n",
    "tempDF1 = rmsDF_kk\n",
    "tempDF2 = sampleDF.loc[sampleDF['Phenotype'].isin(tempL1)]\n",
    "tempI = moduleDF.index\n",
    "formula = 'RMS ~ C(Intervention)'\n",
    "tempL2 = ['C(Intervention)']#For variables of interest in ANOVA\n",
    "\n",
    "#Statistical tests per module\n",
    "t_start = time.time()\n",
    "tempL3 = []#For ANOVA table\n",
    "tempL4 = []#For post-hoc test table\n",
    "for module in tempI.tolist():\n",
    "    #Select the target module\n",
    "    tempS = tempDF1.loc[module]\n",
    "    tempS.name = 'RMS'\n",
    "    #Add metadata while selecting the target sample samples\n",
    "    tempDF = pd.merge(tempS, tempDF2, left_index=True, right_index=True, how='inner')\n",
    "    \n",
    "    #ANOVA\n",
    "    model = smf.ols(formula, data=tempDF).fit()\n",
    "    anovaDF = anova_lm(model, typ=2)#ANOVA type doesn't matter in this case\n",
    "    ##Take the results per variable\n",
    "    tempDF3 = pd.DataFrame(columns=['DoF', 'Fstat', 'Pval'])\n",
    "    for variable in tempL2:\n",
    "        dof1 = int(anovaDF.at[variable, 'df'])#Between-groups\n",
    "        dof2 = int(anovaDF.at['Residual', 'df'])#Within-groups\n",
    "        dof = (dof1, dof2)\n",
    "        fstat = anovaDF.at[variable, 'F']\n",
    "        pval = anovaDF.at[variable, 'PR(>F)']\n",
    "        tempDF3.loc[variable] = [dof, fstat, pval]\n",
    "    tempDF3['AdjPval'] = 1.0#Add dummy column for now\n",
    "    ##Convert to wide-format\n",
    "    tempS = pd.Series(len(tempDF), index=['N'], name=module)\n",
    "    tempL = [tempS]\n",
    "    for variable in tempDF3.index.tolist():\n",
    "        tempS = tempDF3.loc[variable]\n",
    "        tempS.index = variable+'_'+tempS.index\n",
    "        tempS.name = module\n",
    "        tempL.append(tempS)\n",
    "    tempS = pd.concat(tempL, axis=0)\n",
    "    tempL3.append(tempS)\n",
    "    \n",
    "    #Post-hoc tests per control vs. contrast\n",
    "    tempDF4 = pd.DataFrame(columns=['DoF', 'tStat', 'Pval'])\n",
    "    for contrast in tempL1:\n",
    "        if control!=contrast:\n",
    "            tempS1 = tempDF['RMS'].loc[tempDF['Phenotype']==control]\n",
    "            tempS2 = tempDF['RMS'].loc[tempDF['Phenotype']==contrast]\n",
    "            #Two-sided Welch's t-test\n",
    "            tstat, pval, dof = weightstats.ttest_ind(tempS2, tempS1,#t-statistic reflects direction from the baseline\n",
    "                                                     alternative='two-sided', usevar='unequal')\n",
    "            tempDF4.loc[contrast+'-vs-'+control] = [dof, tstat, pval]\n",
    "    ##P-value adjustment across all comparisons per module by using Benjamini–Hochberg method\n",
    "    tempDF4['AdjPval'] = multi.multipletests(tempDF4['Pval'], alpha=0.05, method='fdr_bh',\n",
    "                                             is_sorted=False, returnsorted=False)[1]\n",
    "    ##Convert to wide-format\n",
    "    tempL = []\n",
    "    for comparison in tempDF4.index.tolist():\n",
    "        tempS = tempDF4.loc[comparison]\n",
    "        tempS.index = comparison+'_'+tempS.index\n",
    "        tempS.name = module\n",
    "        tempL.append(tempS)\n",
    "    tempS = pd.concat(tempL, axis=0)\n",
    "    tempL4.append(tempS)\n",
    "t_elapsed = time.time() - t_start\n",
    "print('Elapsed time for', len(tempI), 'ANOVA and',\n",
    "      (len(tempL1)-1)*len(tempI), 'post-hoc tests (',\n",
    "      len(tempL1)-1, 'comparisons x', len(tempI), 'modules):',\n",
    "      round(t_elapsed//60), 'min', round(t_elapsed%60, 1), 'sec')\n",
    "\n",
    "#Generate ANOVA table\n",
    "tempDF3 = pd.concat(tempL3, axis=1).T\n",
    "tempDF3.index.name = tempI.name\n",
    "##P-value adjustment across all tests by using Benjamini–Hochberg method\n",
    "for variable in tempL2:\n",
    "    #Overwrite the dummy values\n",
    "    tempDF3[variable+'_AdjPval'] = multi.multipletests(tempDF3[variable+'_Pval'], alpha=0.05, method='fdr_bh',\n",
    "                                                       is_sorted=False, returnsorted=False)[1]\n",
    "##Convert back dtypes (due to the forced change during wide-format)\n",
    "for col_n in tempDF3.columns.tolist():\n",
    "    if 'N'==col_n:\n",
    "        tempDF3[col_n] = tempDF3[col_n].astype(int)\n",
    "    elif 'DoF' in col_n:\n",
    "        tempDF3[col_n] = tempDF3[col_n].astype(str)\n",
    "    else:\n",
    "        tempDF3[col_n] = tempDF3[col_n].astype(float)\n",
    "##Rename columns (because only one variable in this case)\n",
    "tempDF3.columns = 'ANOVA_'+tempDF3.columns.str.replace('^.*_', '', regex=True)\n",
    "display(tempDF3)\n",
    "\n",
    "#Generate post-hoc test table\n",
    "tempDF4 = pd.concat(tempL4, axis=1).T\n",
    "tempDF4.index.name = tempI.name\n",
    "display(tempDF4)\n",
    "\n",
    "statDF1 = tempDF3\n",
    "statDF2 = tempDF4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tempL1 = ['Ctrl', 'Aca', '17aE2', 'Rapa']#Target sample groups to be summarized\n",
    "tempDF1 = rmsDF_kk\n",
    "tempDF2 = sampleDF.loc[sampleDF['Phenotype'].isin(tempL1)]\n",
    "\n",
    "#Calculate general statistics per intervention group\n",
    "tempL2 = []\n",
    "for phenotype in tempL1:\n",
    "    #Select the target samples\n",
    "    tempL = tempDF2.loc[tempDF2['Phenotype']==phenotype].index.tolist()\n",
    "    tempDF = tempDF1[tempL]\n",
    "    #Calculate general statistics\n",
    "    tempS1 = len(tempL) - tempDF.isnull().sum(axis=1)\n",
    "    tempS1.name = phenotype+'_N'\n",
    "    tempS2 = tempDF.mean(axis=1)\n",
    "    tempS2.name = phenotype+'_RMSmean'\n",
    "    tempS3 = tempDF.sem(axis=1, ddof=1)\n",
    "    tempS3.name = phenotype+'_RMSsem'\n",
    "    #Merge\n",
    "    tempDF = pd.concat([tempS1, tempS2, tempS3], axis=1)\n",
    "    tempL2.append(tempDF)\n",
    "tempDF = pd.concat(tempL2, axis=1)\n",
    "display(tempDF)\n",
    "\n",
    "#Merge all the tables\n",
    "print('General statistics table:', tempDF.shape)\n",
    "print('ANOVA table:', statDF1.shape)\n",
    "print('Post-hoc test table:', statDF2.shape)\n",
    "tempDF = pd.concat([moduleDF['ModuleName'], tempDF, statDF1, statDF2], axis=1)\n",
    "\n",
    "#Sort\n",
    "tempDF = tempDF.sort_values(by='ANOVA_Pval', ascending=True)\n",
    "display(tempDF)\n",
    "\n",
    "#Save\n",
    "fileDir = './ExportData/'\n",
    "ipynbName = '230214_LC-M001-proteomics-DIRAC-ver7-2_DIRAC-GOBP_'\n",
    "fileName = 'inter-group-comparison_RCI.tsv'\n",
    "tempDF.to_csv(fileDir+ipynbName+fileName, sep='\\t', index=True)\n",
    "\n",
    "statDF = tempDF"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 4-2-2. Changed modules (ANOVA)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Prepare variables in the model\n",
    "#-> In this case, only the one variable (intervention) was included.\n",
    "variableL = ['ANOVA']\n",
    "\n",
    "#Changed modules\n",
    "for variable in variableL:\n",
    "    tempDF = statDF.loc[statDF[variable+'_AdjPval']<0.05]\n",
    "    tempDF = tempDF.sort_values(by=variable+'_AdjPval', ascending=True)\n",
    "    tempL1 = tempDF.loc[:, tempDF.columns.str.contains('_RMSmean')].columns.tolist()\n",
    "    tempL2 = tempDF.loc[:, tempDF.columns.str.contains('^'+variable+'_')].columns.tolist()\n",
    "    tempDF = tempDF[[col_n for subL in [['ModuleName'], tempL1, tempL2] for col_n in subL]]\n",
    "    print(variable+' (adjusted P < 0.05):', len(tempDF))\n",
    "    display(tempDF)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 4-2-3. Changed modules by each intervention (Welch's t-test)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Extract only the changed modules\n",
    "variable = 'ANOVA'\n",
    "tempDF = statDF.loc[statDF[variable+'_AdjPval']<0.05]\n",
    "tempDF = tempDF.sort_values(by=variable+'_AdjPval', ascending=True)\n",
    "print(variable+' (adjusted P < 0.05):', len(tempDF))\n",
    "\n",
    "#Take adjusted P-value\n",
    "tempDF1 = tempDF.loc[:, tempDF.columns.str.contains('-vs-.*_AdjPval$')]\n",
    "tempDF1.columns = tempDF1.columns.str.replace('_AdjPval$', '')\n",
    "tempDF1 = pd.merge(tempDF[['ModuleName', variable+'_AdjPval']], tempDF1,\n",
    "                   left_index=True, right_index=True, how='right')\n",
    "print('Adjusted P-value:')\n",
    "display(tempDF1)\n",
    "display(tempDF1.describe())\n",
    "\n",
    "#Take t-statistic (for direction)\n",
    "tempDF2 = tempDF.loc[:, tempDF.columns.str.contains('-vs-.*_tStat$')]\n",
    "tempDF2.columns = tempDF2.columns.str.replace('_tStat$', '')\n",
    "tempDF2 = pd.merge(tempDF[['ModuleName', variable+'_AdjPval']], tempDF2,\n",
    "                   left_index=True, right_index=True, how='right')\n",
    "print('Changed direction (t-statistic):')\n",
    "display(tempDF2)\n",
    "display(tempDF2.describe())\n",
    "\n",
    "pvalDF = tempDF1\n",
    "diffDF = tempDF2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> Check the changed modules (nominal P-value < 0.05) as reference.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Extract only the changed modules\n",
    "variable = 'ANOVA'\n",
    "tempDF = statDF.loc[statDF[variable+'_Pval']<0.05]\n",
    "tempDF = tempDF.sort_values(by=variable+'_Pval', ascending=True)\n",
    "print(variable+' (nominal P < 0.05):', len(tempDF))\n",
    "\n",
    "#Take adjusted P-value\n",
    "tempDF1 = tempDF.loc[:, tempDF.columns.str.contains('-vs-.*_AdjPval$')]\n",
    "tempDF1.columns = tempDF1.columns.str.replace('_AdjPval$', '')\n",
    "tempDF1 = pd.merge(tempDF[['ModuleName', variable+'_AdjPval']], tempDF1,\n",
    "                   left_index=True, right_index=True, how='right')\n",
    "print('Adjusted P-value:')\n",
    "display(tempDF1)\n",
    "display(tempDF1.describe())\n",
    "\n",
    "#Take t-statistic (for direction)\n",
    "tempDF2 = tempDF.loc[:, tempDF.columns.str.contains('-vs-.*_tStat$')]\n",
    "tempDF2.columns = tempDF2.columns.str.replace('_tStat$', '')\n",
    "tempDF2 = pd.merge(tempDF[['ModuleName', variable+'_AdjPval']], tempDF2,\n",
    "                   left_index=True, right_index=True, how='right')\n",
    "#print('Changed direction (t-statistic):')\n",
    "#display(tempDF2)\n",
    "#display(tempDF2.describe())\n",
    "\n",
    "pvalDF_ref = tempDF1\n",
    "diffDF_ref = tempDF2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "### 4-3. Visualization: clustermap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tempD0 = {'Ctrl':'Control', 'Aca':'Acarbose',\n",
    "          '17aE2':'17'+r'$\\alpha$'+'-Estradiol', 'Rapa':'Rapamycin'}\n",
    "tempDF1 = rciDF_kk.rename(columns=tempD0)\n",
    "regulation = 'Tightened'\n",
    "tempD1 = {'Acarbose':'tab:red',\n",
    "          '17'+r'$\\alpha$'+'-Estradiol':'tab:green', 'Rapamycin':'tab:purple'}\n",
    "tempD2 = {'Acarbose':plt.get_cmap('tab20')(7),\n",
    "          '17'+r'$\\alpha$'+'-Estradiol':plt.get_cmap('tab20')(5),\n",
    "          'Rapamycin':plt.get_cmap('tab20')(9)}\n",
    "\n",
    "#Prepare color labels for tightened module set\n",
    "tempDF2 = pvalDF_ref.loc[:, pvalDF_ref.columns.str.contains('-vs-')]#Nominal P-value\n",
    "for col_n in tempDF2.columns.tolist():\n",
    "    tempS1 = pvalDF_ref[col_n]\n",
    "    tempS2 = diffDF_ref[col_n]\n",
    "    if regulation=='Changed':\n",
    "        tempS2 = tempS2.loc[(tempS1<0.05)]\n",
    "    elif regulation=='Tightened':\n",
    "        tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2>0)]\n",
    "    elif regulation=='Loosened':\n",
    "        tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2<0)]\n",
    "    label = tempD0[re.sub('-vs-.*', '', col_n)]\n",
    "    tempL = []\n",
    "    count = 0#just for checking\n",
    "    for module in tempDF2.index.tolist():\n",
    "        if module in tempS2.index.tolist():\n",
    "            if module in pvalDF.index.tolist():#Adjusted P-value\n",
    "                tempL.append(tempD1[label])\n",
    "                count += 1#Just for checking\n",
    "            else:#Nominal P-value\n",
    "                tempL.append(tempD2[label])\n",
    "        else:\n",
    "            tempL.append('white')\n",
    "    tempDF2[label] = tempL\n",
    "    print(regulation+' module in '+col_n)\n",
    "    print(' -> adjusted P < 0.05:', count)\n",
    "    print(' -> nominal P < 0.05:', len(tempS2))\n",
    "tempDF2 = tempDF2[list(tempD1.keys())]\n",
    "\n",
    "#Clustermap\n",
    "sns.set(style='ticks', font='Arial', context='talk')\n",
    "cm = sns.clustermap(tempDF1.T, method='ward', metric='euclidean', cmap='afmhot',\n",
    "                    row_cluster=True, col_cluster=True, row_linkage=None, col_linkage=None,\n",
    "                    row_colors=None, col_colors=tempDF2, xticklabels=False, yticklabels=True,\n",
    "                    dendrogram_ratio=(0.025, 0.2), colors_ratio=(0.01, 0.1),\n",
    "                    cbar_pos=(0.05, -0.15, 0.3, 0.075), cbar_kws={'orientation': 'horizontal'},\n",
    "                    figsize=(12, 3), **{'vmin':0.5, 'vmax':1})\n",
    "cm.cax.set_title('Module RCI', size='medium',\n",
    "                 verticalalignment='bottom', horizontalalignment='center')\n",
    "cm.cax.tick_params(labelsize='small')\n",
    "bottom, top = cm.ax_heatmap.get_ylim()\n",
    "#cm.ax_heatmap.set_ylim(bottom + 0.5, top - 0.5)##To avoid half cut of first and last rows\n",
    "hm = cm.ax_heatmap.get_position()\n",
    "rd = cm.ax_row_dendrogram.get_position()\n",
    "cd = cm.ax_col_dendrogram.get_position()\n",
    "cm.ax_heatmap.set_position([hm.x0, hm.y0, hm.width, hm.height])\n",
    "cm.ax_row_dendrogram.set_position([rd.x0, rd.y0, rd.width, rd.height])\n",
    "cm.ax_col_dendrogram.set_position([cd.x0, cd.y0, cd.width, cd.height])\n",
    "cm.ax_heatmap.set_xlabel('GOBP module')\n",
    "cm.ax_heatmap.set_ylabel('')\n",
    "##row/column color bar legend (axis is same with cm.cax!)\n",
    "tempL = []\n",
    "for group in tempD1.keys():\n",
    "    tempL.append(mpatches.Patch(color=tempD1[group],\n",
    "                                label='by '+group+' (adjusted '+r'$P$'+' < 0.05)'))\n",
    "for group in tempD2.keys():\n",
    "    tempL.append(mpatches.Patch(color=tempD2[group], label='(nominal '+r'$P$'+' < 0.05)'))\n",
    "legend1 = plt.legend(handles=tempL, fontsize='small', labelspacing=0.2, ncol=2,\n",
    "                     title='Tightened module (vs. Control)', title_fontsize='medium',\n",
    "                     bbox_to_anchor=(1, 0.5), loc='center left', borderaxespad=3.5, frameon=False)\n",
    "plt.gca().add_artist(legend1)\n",
    "##Save\n",
    "fileDir = './ExportFigures/'\n",
    "ipynbName = '230214_LC-M001-proteomics-DIRAC-ver7-2_DIRAC-GOBP_'\n",
    "fileName = 'RCI-clustermap.pdf'\n",
    "#plt.gcf().savefig(fileDir+ipynbName+fileName, dpi=300, bbox_inches='tight', pad_inches=0.04, transparent=True)\n",
    "plt.show()\n",
    "\n",
    "#Save label order\n",
    "tempDF = moduleDF.loc[rciDF_kk.index[cm.dendrogram_col.reordered_ind]]\n",
    "tempDF1 = rciDF_kk.copy()\n",
    "tempDF = pd.merge(tempDF['ModuleName'], tempDF1, left_index=True, right_index=True, how='left')\n",
    "tempDF1 = pvalDF.loc[:, pvalDF.columns.str.contains('-vs-')]#Adjusted P-value\n",
    "tempDF = pd.merge(tempDF, tempDF1, left_index=True, right_index=True, how='left')\n",
    "tempDF = tempDF.reset_index()\n",
    "tempDF.index.name = 'Xcoord'\n",
    "display(tempDF)\n",
    "fileDir = './ExportData/'\n",
    "ipynbName = '230214_LC-M001-proteomics-DIRAC-ver7-2_DIRAC-GOBP_'\n",
    "fileName = 'RCI-clustermap-xticks.tsv'\n",
    "#tempDF.to_csv(fileDir+ipynbName+fileName, index=True, sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> –> Simplify.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tempD0 = {'Ctrl':'Control', 'Aca':'Acarbose',\n",
    "          '17aE2':'17'+r'$\\alpha$'+'-Estradiol', 'Rapa':'Rapamycin'}\n",
    "tempDF1 = rciDF_kk.rename(columns=tempD0)\n",
    "regulation = 'Tightened'\n",
    "tempD1 = {'Acarbose':'tab:red',\n",
    "          '17'+r'$\\alpha$'+'-Estradiol':'tab:green', 'Rapamycin':'tab:purple'}\n",
    "\n",
    "#Prepare color labels for tightened module set\n",
    "tempDF2 = pvalDF.loc[:, pvalDF.columns.str.contains('-vs-')]#Adjusted P-value\n",
    "for col_n in tempDF2.columns.tolist():\n",
    "    tempS1 = pvalDF[col_n]\n",
    "    tempS2 = diffDF[col_n]\n",
    "    if regulation=='Changed':\n",
    "        tempS2 = tempS2.loc[(tempS1<0.05)]\n",
    "    elif regulation=='Tightened':\n",
    "        tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2>0)]\n",
    "    elif regulation=='Loosened':\n",
    "        tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2<0)]\n",
    "    label = tempD0[re.sub('-vs-.*', '', col_n)]\n",
    "    tempL = []\n",
    "    count = 0#just for checking\n",
    "    for module in tempDF2.index.tolist():\n",
    "        if module in tempS2.index.tolist():\n",
    "            tempL.append(tempD1[label])\n",
    "            count += 1#Just for checking\n",
    "        else:\n",
    "            tempL.append('white')\n",
    "    tempDF2[label] = tempL\n",
    "    print(regulation+' module in '+col_n)\n",
    "    print(' -> adjusted P < 0.05:', count)\n",
    "tempDF2 = tempDF2[list(tempD1.keys())]\n",
    "\n",
    "#Clustermap\n",
    "sns.set(style='ticks', font='Arial', context='talk')\n",
    "cm = sns.clustermap(tempDF1.T, method='ward', metric='euclidean', cmap='afmhot',\n",
    "                    row_cluster=True, col_cluster=True, row_linkage=None, col_linkage=None,\n",
    "                    row_colors=None, col_colors=tempDF2, xticklabels=False, yticklabels=True,\n",
    "                    dendrogram_ratio=(0.025, 0.2), colors_ratio=(0.01, 0.1),\n",
    "                    cbar_pos=(0.175, -0.15, 0.3, 0.075), cbar_kws={'orientation': 'horizontal'},\n",
    "                    figsize=(12, 3), **{'vmin':0.5, 'vmax':1})\n",
    "cm.cax.set_title('Module RCI', size='medium',\n",
    "                 verticalalignment='bottom', horizontalalignment='center')\n",
    "cm.cax.tick_params(labelsize='small')\n",
    "bottom, top = cm.ax_heatmap.get_ylim()\n",
    "#cm.ax_heatmap.set_ylim(bottom + 0.5, top - 0.5)##To avoid half cut of first and last rows\n",
    "hm = cm.ax_heatmap.get_position()\n",
    "rd = cm.ax_row_dendrogram.get_position()\n",
    "cd = cm.ax_col_dendrogram.get_position()\n",
    "cm.ax_heatmap.set_position([hm.x0, hm.y0, hm.width, hm.height])\n",
    "cm.ax_row_dendrogram.set_position([rd.x0, rd.y0, rd.width, rd.height])\n",
    "cm.ax_col_dendrogram.set_position([cd.x0, cd.y0, cd.width, cd.height])\n",
    "cm.ax_heatmap.set_xlabel('GOBP module')\n",
    "cm.ax_heatmap.set_ylabel('')\n",
    "##row/column color bar legend (axis is same with cm.cax!)\n",
    "tempL = []\n",
    "for group in tempD1.keys():\n",
    "    tempL.append(mpatches.Patch(color=tempD1[group], label='by '+group))\n",
    "legend1 = plt.legend(handles=tempL, fontsize='small', labelspacing=0.2,\n",
    "                     title='Tightened module (vs. Control)', title_fontsize='medium',\n",
    "                     bbox_to_anchor=(1, 0.5), loc='center left', borderaxespad=3.5, frameon=False)\n",
    "plt.gca().add_artist(legend1)\n",
    "##Save\n",
    "fileDir = './ExportFigures/'\n",
    "ipynbName = '230214_LC-M001-proteomics-DIRAC-ver7-2_DIRAC-GOBP_'\n",
    "fileName = 'RCI-clustermap.pdf'\n",
    "plt.gcf().savefig(fileDir+ipynbName+fileName, dpi=300, bbox_inches='tight', pad_inches=0.04, transparent=True)\n",
    "plt.show()\n",
    "\n",
    "#Save label order\n",
    "tempDF = moduleDF.loc[rciDF_kk.index[cm.dendrogram_col.reordered_ind]]\n",
    "tempDF1 = rciDF_kk.copy()\n",
    "tempDF = pd.merge(tempDF['ModuleName'], tempDF1, left_index=True, right_index=True, how='left')\n",
    "tempDF1 = pvalDF.loc[:, pvalDF.columns.str.contains('-vs-')]#Adjusted P-value\n",
    "tempDF = pd.merge(tempDF, tempDF1, left_index=True, right_index=True, how='left')\n",
    "tempDF = tempDF.reset_index()\n",
    "tempDF.index.name = 'Xcoord'\n",
    "display(tempDF)\n",
    "fileDir = './ExportData/'\n",
    "ipynbName = '230214_LC-M001-proteomics-DIRAC-ver7-2_DIRAC-GOBP_'\n",
    "fileName = 'RCI-clustermap-xticks.tsv'\n",
    "tempDF.to_csv(fileDir+ipynbName+fileName, index=True, sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 4-4. Visualization: venn diagram"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Prepare label and color\n",
    "tempD0 = {'Ctrl':'Control', 'Aca':'Acarbose',\n",
    "          '17aE2':'17'+r'$\\alpha$'+'-Estradiol', 'Rapa':'Rapamycin'}\n",
    "tempD1 = {'Acarbose':'tab:red',\n",
    "          '17'+r'$\\alpha$'+'-Estradiol':'tab:green', 'Rapamycin':'tab:purple'}\n",
    "\n",
    "#Visualization per direction\n",
    "for regulation in ['Changed', 'Tightened', 'Loosened']:\n",
    "    #Prepare module sets\n",
    "    tempD2 = {}\n",
    "    tempDF = pvalDF.loc[:, pvalDF.columns.str.contains('-vs-')]#Adjusted P-value\n",
    "    for col_n in tempDF.columns.tolist():\n",
    "        tempS1 = pvalDF[col_n]\n",
    "        tempS2 = diffDF[col_n]\n",
    "        if regulation=='Changed':\n",
    "            tempS2 = tempS2.loc[(tempS1<0.05)]\n",
    "        elif regulation=='Tightened':\n",
    "            tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2>0)]\n",
    "        elif regulation=='Loosened':\n",
    "            tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2<0)]\n",
    "        label = tempD0[re.sub('-vs-.*$', '', col_n)]\n",
    "        tempD2[label] = set(tempS2.index.tolist())\n",
    "    ##Sort to make consistent order in manual legend generation\n",
    "    tempD = {}\n",
    "    for label in tempD1.keys():\n",
    "        tempD[label] = tempD2[label]\n",
    "    \n",
    "    #Not significant in all contrasts\n",
    "    tempDF = pvalDF.copy()\n",
    "    for moduleS in tempD.values():\n",
    "        tempDF = tempDF.loc[~tempDF.index.isin(moduleS)]\n",
    "    count = len(tempDF)\n",
    "    print(regulation+' modules (vs. Control):')\n",
    "    print(' -> Not significant in all contrasts:', count)\n",
    "    \n",
    "    #Skip the followings if no significant module\n",
    "    if count==len(pvalDF):\n",
    "        continue\n",
    "    \n",
    "    #Venn diagram\n",
    "    sns.set(style='ticks', font='Arial', context='talk')\n",
    "    fig, ax = plt.subplots(figsize=(4, 4))\n",
    "    venn(tempD, fmt='{size:,}', cmap=list(tempD1.values()), legend_loc=None, ax=ax)\n",
    "    plt.setp(ax, ylim=(0.05, 0.975))#Otherwise, weird space...\n",
    "    ##Add legend annotation\n",
    "    x_coord = [0.1, 0.9, 0.8]\n",
    "    y_coord = [0.8, 0.8, 0.25]\n",
    "    h_align = ['right', 'left', 'left']\n",
    "    v_align = ['bottom', 'bottom', 'top']\n",
    "    for i in range(len(tempD1)):\n",
    "        key = list(tempD1.keys())[i]\n",
    "        total = f'{len(tempD[key]):,}'\n",
    "        ax.text(x_coord[i], y_coord[i], key+'\\n('+total+' modules)',\n",
    "                fontsize='small', multialignment='center',\n",
    "                horizontalalignment=h_align[i], verticalalignment=v_align[i],\n",
    "                bbox={'boxstyle':'round', 'facecolor':tempD1[key], 'pad':0.2, 'alpha':0.5})\n",
    "    title = regulation+' modules (vs. Control)'\n",
    "    ax.set_title(title, fontsize='medium')\n",
    "    ##Save\n",
    "    if regulation!='Changed':\n",
    "        fileDir = './ExportFigures/'\n",
    "        ipynbName = '230214_LC-M001-proteomics-DIRAC-ver7-2_DIRAC-GOBP_'\n",
    "        fileName = 'RCI-venn-'+regulation.lower()+'.pdf'\n",
    "        plt.gcf().savefig(fileDir+ipynbName+fileName, dpi=300, bbox_inches='tight', pad_inches=0.04, transparent=True)\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Export module list in each subset in the venn diagram\n",
    "for regulation in ['Tightened', 'Loosened']:\n",
    "    #Prepare module sets\n",
    "    tempD = {}\n",
    "    tempDF = pvalDF.loc[:, pvalDF.columns.str.contains('-vs-')]#Adjusted P-value\n",
    "    for col_n in tempDF.columns.tolist():\n",
    "        tempS1 = pvalDF[col_n]\n",
    "        tempS2 = diffDF[col_n]\n",
    "        if regulation=='Changed':\n",
    "            tempS2 = tempS2.loc[(tempS1<0.05)]\n",
    "        elif regulation=='Tightened':\n",
    "            tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2>0)]\n",
    "        elif regulation=='Loosened':\n",
    "            tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2<0)]\n",
    "        tempD[col_n] = set(tempS2.index.tolist())\n",
    "    \n",
    "    #Not significant in all contrasts\n",
    "    tempDF = pvalDF.copy()\n",
    "    for moduleS in tempD.values():\n",
    "        tempDF = tempDF.loc[~tempDF.index.isin(moduleS)]\n",
    "    count = len(tempDF)\n",
    "    print(regulation+' modules (vs. Control):')\n",
    "    print(' -> Not significant in all contrasts:', count)\n",
    "    \n",
    "    #Skip the followings if no significant module\n",
    "    if count==len(pvalDF):\n",
    "        continue\n",
    "    \n",
    "    #Prepare a new .xlsx file (dummy README)\n",
    "    tempL1 = [len(tempD[key]) for key in tempD.keys()]\n",
    "    tempDF = pd.DataFrame({'Group':tempD.keys(), 'nModules':tempL1})\n",
    "    tempDF = tempDF.reset_index().rename(columns={'index':'VennOrder'})\n",
    "    tempDF['VennOrder'] = tempDF['VennOrder'] + 1\n",
    "    fileDir = './ExportData/'\n",
    "    ipynbName = '230214_LC-M001-proteomics-DIRAC-ver7-2_DIRAC-GOBP_'\n",
    "    fileName = 'RCI-venn-'+regulation.lower()+'.xlsx'\n",
    "    tempDF.to_excel(fileDir+ipynbName+fileName, sheet_name='README', header=True, index=False)\n",
    "    display(tempDF)#Check\n",
    "    \n",
    "    t_start = time.time()\n",
    "    #Extract overall set\n",
    "    for key_i in range(len(tempD)):\n",
    "        key = list(tempD.keys())[key_i]\n",
    "        tempS = tempD[key]\n",
    "        tempDF = statDF.loc[statDF.index.isin(tempS)]\n",
    "        #Save summary table by appended to the above .xlsx file\n",
    "        ##Prepare sheet name\n",
    "        tempL1 = ['NA' for i in range(len(tempD))]\n",
    "        tempL1[key_i] = '1'\n",
    "        setName = '('+','.join(tempL1)+')'\n",
    "        with pd.ExcelWriter(fileDir+ipynbName+fileName, mode='a', engine='openpyxl') as writer:\n",
    "            tempDF.to_excel(writer, sheet_name=setName, header=True, index=True)\n",
    "        print(' - '+setName+':', len(tempDF))\n",
    "    \n",
    "    #Extract subset\n",
    "    tempL1 = ['1', '0']\n",
    "    tempL2 = [[k1, k2, k3] for k1 in tempL1 for k2 in tempL1 for k3 in tempL1]\n",
    "    #tempL2.remove(['0', '0', '0'])\n",
    "    for tempL1 in tempL2:\n",
    "        #Positive module set\n",
    "        tempL3 = [list(tempD.values())[key_i] for key_i, binary in enumerate(tempL1) if binary=='1']\n",
    "        tempS1 = set(pvalDF.index.tolist())#Initialize\n",
    "        for tempS in tempL3:\n",
    "            tempS1 = tempS1 & tempS\n",
    "        #Negative module set\n",
    "        tempL3 = [list(tempD.values())[key_i] for key_i, binary in enumerate(tempL1) if binary=='0']\n",
    "        tempS2 = set()#Initialize\n",
    "        for tempS in tempL3:\n",
    "            tempS2 = tempS2 | tempS\n",
    "        #Extract subset\n",
    "        tempS = tempS1 - tempS2\n",
    "        tempDF = statDF.loc[statDF.index.isin(tempS)]\n",
    "        #Save summary table by appended to the above .xlsx file\n",
    "        ##Prepare sheet name\n",
    "        setName = '('+','.join(tempL1)+')'\n",
    "        with pd.ExcelWriter(fileDir+ipynbName+fileName, mode='a', engine='openpyxl') as writer:\n",
    "            tempDF.to_excel(writer, sheet_name=setName, header=True, index=True)\n",
    "        print(' - '+setName+':', len(tempDF))\n",
    "    \n",
    "    t_elapsed = time.time() - t_start\n",
    "    print(' - Elapsed time:', round(t_elapsed//60), 'min', round(t_elapsed%60, 1), 'sec')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 4-5. Visualization: pointplot"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> More modules are checked later (in a separate notebook), and so representative modules are checked for now.  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 4-5-1. Modules tightened by all interventions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Prepare the target module set\n",
    "posL = ['Aca-vs-Ctrl', '17aE2-vs-Ctrl', 'Rapa-vs-Ctrl']\n",
    "negL = ['']\n",
    "regulation = 'Tightened'\n",
    "tempS = pd.Series(np.repeat(True, len(pvalDF)), index=pvalDF.index)#Initialize\n",
    "tempDF = pvalDF.loc[:, pvalDF.columns.str.contains('-vs-')]#Adjusted P-value\n",
    "for col_n in tempDF.columns.tolist():\n",
    "    tempS1 = pvalDF[col_n]\n",
    "    tempS2 = diffDF[col_n]\n",
    "    if col_n in posL:\n",
    "        if regulation=='Tightened':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2>0)\n",
    "        elif regulation=='Loosened':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2<0)\n",
    "    elif col_n in negL:\n",
    "        tempS3 = (tempS1>=0.05)\n",
    "        #Significance for inverse regulation\n",
    "        if regulation=='Tightened':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2<0)\n",
    "        elif regulation=='Loosened':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2>0)\n",
    "        tempS1 = tempS3 | tempS1\n",
    "    else:\n",
    "        tempS1 = (tempS1>=0.0)\n",
    "    #Update True\n",
    "    tempS = tempS & tempS1\n",
    "tempL = tempS.loc[tempS.tolist()].index.tolist()\n",
    "print(len(tempL), regulation.lower()+' modules with significance in', posL, 'but not in', negL)\n",
    "\n",
    "#Select representatives\n",
    "topX = np.min([30, len(tempL)])\n",
    "topX_plot = np.min([10, len(tempL)])\n",
    "tempDF = statDF.loc[:, statDF.columns.str.contains('Pval$')]\n",
    "tempDF = pd.merge(statDF['ModuleName'], tempDF, left_index=True, right_index=True, how='left')\n",
    "tempDF = tempDF.loc[tempL].sort_values(by=variable+'_AdjPval', ascending=True)\n",
    "print('Top', topX, 'modules (sort by the main effect of '+variable+'):')\n",
    "display(tempDF.iloc[:topX])\n",
    "plotL = tempDF.index.tolist()[:topX_plot]\n",
    "\n",
    "#Prepare DF for plot\n",
    "tempDF = rmsDF_kk.reset_index().melt(var_name='SampleID', value_name='RMS', id_vars='ModuleID')\n",
    "tempDF1 = sampleDF.reset_index()[['SampleID', 'Phenotype']]\n",
    "tempDF = pd.merge(tempDF, tempDF1, on='SampleID', how='left')\n",
    "\n",
    "#Prepare label and color\n",
    "tempD0 = {'Ctrl':'Control', 'Aca':'Acarbose',\n",
    "          '17aE2':'17'+r'$\\alpha$'+'-Estradiol', 'Rapa':'Rapamycin'}\n",
    "tempDF['Group'] = tempDF['Phenotype'].map(tempD0)\n",
    "tempD1 = {'Control':'tab:blue', 'Acarbose':'tab:red',\n",
    "          '17'+r'$\\alpha$'+'-Estradiol':'tab:green', 'Rapamycin':'tab:purple'}\n",
    "\n",
    "#Visualize each representative\n",
    "for rank_i in range(len(plotL)):\n",
    "    print(' - Rank '+str(rank_i+1)+' (sort by the main effect of '+variable+'):')\n",
    "    module = plotL[rank_i]\n",
    "    #Check module summary\n",
    "    tempDF1 = pd.DataFrame(moduleDF.loc[module]).T\n",
    "    display(tempDF1)\n",
    "    \n",
    "    #Select RMS\n",
    "    tempDF1 = tempDF.loc[tempDF['ModuleID']==module]\n",
    "    \n",
    "    #Check RMS summary\n",
    "    tempDF2 = tempDF1.groupby(['Group'])['RMS'].agg(['count', 'mean', 'std'])\n",
    "    tempL1 = []\n",
    "    tempL2 = []\n",
    "    for row_n in tempDF2.index.tolist():\n",
    "        count, mean, std = tempDF2.loc[row_n]\n",
    "        tempL1.append(mean - 1.96*std/np.sqrt(count))\n",
    "        tempL2.append(mean + 1.96*std/np.sqrt(count))\n",
    "    tempDF2['0.025'] = tempL1\n",
    "    tempDF2['0.975'] = tempL2\n",
    "    tempDF2 = tempDF2.loc[list(tempD1.keys())]#Sort\n",
    "    display(tempDF2)\n",
    "    \n",
    "    #Prepare significance labels\n",
    "    ##Retrieve statistical significance\n",
    "    tempS = pvalDF.loc[module, pvalDF.columns.str.contains('-vs-')]#Adjusted P-value\n",
    "    tempS.name = 'AdjPval'\n",
    "    ##Clean\n",
    "    tempDF2 = tempS.index.to_series().str.split(pat='-vs-', expand=True)\n",
    "    tempDF2 = tempDF2.rename(columns={0:'Contrast', 1:'Baseline'})\n",
    "    tempDF2 = pd.merge(tempDF2, tempS, left_index=True, right_index=True, how='left')\n",
    "    tempDF2['Contrast'] = tempDF2['Contrast'].map(tempD0)\n",
    "    tempDF2['Baseline'] = tempDF2['Baseline'].map(tempD0)\n",
    "    ##Convert p-value to label\n",
    "    tempL = []\n",
    "    for row_i in range(len(tempDF2)):\n",
    "        pval = tempDF2['AdjPval'].iloc[row_i]\n",
    "        if pval<0.001:\n",
    "            tempL.append('***')\n",
    "        elif pval<0.01:\n",
    "            tempL.append('**')\n",
    "        elif pval<0.05:\n",
    "            tempL.append('*')\n",
    "        else:\n",
    "            pval_text = Decimal(str(pval)).quantize(Decimal('0.01'), rounding=ROUND_HALF_UP)\n",
    "            tempL.append(r'$P$ = '+str(pval_text))\n",
    "    tempDF2['SignifLabel'] = tempL\n",
    "    display(tempDF2)\n",
    "    \n",
    "    #Visualization\n",
    "    ymax = 1.0\n",
    "    ymin = 0.0\n",
    "    yinter = 0.2\n",
    "    ymargin_t = 0.4\n",
    "    ymargin_b = 0.05\n",
    "    aline_ymin = 1.0\n",
    "    aline_ymargin = 0.125\n",
    "    sns.set(style='ticks', font='Arial', context='talk')\n",
    "    plt.figure(figsize=(2, 4))\n",
    "    sns.pointplot(data=tempDF1, x='Group', y='RMS', order=list(tempD1.keys()), palette=tempD1,\n",
    "                  markers='o', dodge=False, join=False, capsize=0.6, estimator=np.mean, ci=95)\n",
    "    p = sns.stripplot(data=tempDF1, x='Group', y='RMS',\n",
    "                      order=list(tempD1.keys()), palette=tempD1, dodge=False, jitter=0.15,\n",
    "                      size=5, edgecolor='black', linewidth=1, **{'marker':'o', 'alpha':0.5})\n",
    "    ##Set axis\n",
    "    sns.despine()\n",
    "    p.set(ylim=(ymin-ymargin_b, ymax+ymargin_t), yticks=np.arange(ymin, ymax + yinter/10, yinter))\n",
    "    plt.setp(p.get_xticklabels(), rotation=70, horizontalalignment='right',\n",
    "             verticalalignment='center', rotation_mode='anchor')\n",
    "    ##Add significance labels\n",
    "    for row_i in range(len(tempDF2)):\n",
    "        #Baseline\n",
    "        group_0 = tempDF2['Baseline'].iloc[row_i]\n",
    "        index_0 = list(tempD1.keys()).index(group_0)\n",
    "        xcoord_0 = index_0\n",
    "        #Contrast\n",
    "        group_1 = tempDF2['Contrast'].iloc[row_i]\n",
    "        index_1 = list(tempD1.keys()).index(group_1)\n",
    "        xcoord_1 = index_1\n",
    "        #Standard point of marker\n",
    "        xcoord = (xcoord_0+xcoord_1)/2\n",
    "        ycoord = aline_ymin + aline_ymargin*row_i\n",
    "        label = tempDF2['SignifLabel'].iloc[row_i]\n",
    "        #Add annotation lines\n",
    "        aline_offset = yinter/5\n",
    "        aline_length = yinter/5 + aline_offset/2\n",
    "        plt.plot([xcoord_0, xcoord_0, xcoord_1, xcoord_1],\n",
    "                 [ycoord+aline_offset, ycoord+aline_length, ycoord+aline_length, ycoord+aline_offset],\n",
    "                 lw=1.5, c='k')\n",
    "        #Add annotation text\n",
    "        if label in ['***', '**', '*']:\n",
    "            text_offset = yinter/21\n",
    "            p.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                       horizontalalignment='center', verticalalignment='bottom',\n",
    "                       fontsize='medium', color='k')\n",
    "        else:\n",
    "            text_offset = yinter/3.5\n",
    "            p.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                       horizontalalignment='center', verticalalignment='bottom',\n",
    "                       fontsize='x-small', color='k')\n",
    "    ##Add annotation\n",
    "    p.set_title('Consensus:\\nown group', {'fontsize':'small'})\n",
    "    xoff = 0.015\n",
    "    yoff = 0.01\n",
    "    rect = plt.Rectangle((xoff, 1), 1-xoff, 0.155,#Manual adjustment\n",
    "                         transform=p.axes.transAxes, facecolor=plt.get_cmap('tab20')(15),\n",
    "                         clip_on=False, linewidth=0, zorder=0.5)\n",
    "    p.add_patch(rect)\n",
    "    ##Set axis label and title\n",
    "    plt.setp(p, xlabel='', ylabel='Sample RMS\\n(Mean = Module RCI)')\n",
    "    plt.suptitle(module, size='small',\n",
    "                 verticalalignment='bottom', horizontalalignment='center', wrap=True, y=1.0)\n",
    "    ##Save\n",
    "    #fileDir = './ExportFigures/'\n",
    "    #ipynbName = '230214_LC-M001-proteomics-DIRAC-ver7-2_DIRAC-GOBP_'\n",
    "    #fileName = 'RCI-pointplot-'+module.replace('GO:', 'GO')+'.pdf'\n",
    "    #plt.gcf().savefig(fileDir+ipynbName+fileName, dpi=300, bbox_inches='tight', pad_inches=0.04, transparent=True)\n",
    "    plt.show()\n",
    "    print('')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 4-5-2. Modules tightened specifically by Aca"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Prepare the target module set\n",
    "posL = ['Aca-vs-Ctrl']\n",
    "negL = ['17aE2-vs-Ctrl', 'Rapa-vs-Ctrl']\n",
    "regulation = 'Tightened'\n",
    "tempS = pd.Series(np.repeat(True, len(pvalDF)), index=pvalDF.index)#Initialize\n",
    "tempDF = pvalDF.loc[:, pvalDF.columns.str.contains('-vs-')]#Adjusted P-value\n",
    "for col_n in tempDF.columns.tolist():\n",
    "    tempS1 = pvalDF[col_n]\n",
    "    tempS2 = diffDF[col_n]\n",
    "    if col_n in posL:\n",
    "        if regulation=='Tightened':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2>0)\n",
    "        elif regulation=='Loosened':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2<0)\n",
    "    elif col_n in negL:\n",
    "        tempS3 = (tempS1>=0.05)\n",
    "        #Significance for inverse regulation\n",
    "        if regulation=='Tightened':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2<0)\n",
    "        elif regulation=='Loosened':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2>0)\n",
    "        tempS1 = tempS3 | tempS1\n",
    "    else:\n",
    "        tempS1 = (tempS1>=0.0)\n",
    "    #Update True\n",
    "    tempS = tempS & tempS1\n",
    "tempL = tempS.loc[tempS.tolist()].index.tolist()\n",
    "print(len(tempL), regulation.lower()+' modules with significance in', posL, 'but not in', negL)\n",
    "\n",
    "#Select representatives\n",
    "topX = np.min([30, len(tempL)])\n",
    "topX_plot = np.min([10, len(tempL)])\n",
    "tempDF = statDF.loc[:, statDF.columns.str.contains('Pval$')]\n",
    "tempDF = pd.merge(statDF['ModuleName'], tempDF, left_index=True, right_index=True, how='left')\n",
    "tempDF = tempDF.loc[tempL].sort_values(by=variable+'_AdjPval', ascending=True)\n",
    "print('Top', topX, 'modules (sort by the main effect of '+variable+'):')\n",
    "display(tempDF.iloc[:topX])\n",
    "plotL = tempDF.index.tolist()[:topX_plot]\n",
    "\n",
    "#Prepare DF for plot\n",
    "tempDF = rmsDF_kk.reset_index().melt(var_name='SampleID', value_name='RMS', id_vars='ModuleID')\n",
    "tempDF1 = sampleDF.reset_index()[['SampleID', 'Phenotype']]\n",
    "tempDF = pd.merge(tempDF, tempDF1, on='SampleID', how='left')\n",
    "\n",
    "#Prepare label and color\n",
    "tempD0 = {'Ctrl':'Control', 'Aca':'Acarbose',\n",
    "          '17aE2':'17'+r'$\\alpha$'+'-Estradiol', 'Rapa':'Rapamycin'}\n",
    "tempDF['Group'] = tempDF['Phenotype'].map(tempD0)\n",
    "tempD1 = {'Control':'tab:blue', 'Acarbose':'tab:red',\n",
    "          '17'+r'$\\alpha$'+'-Estradiol':'tab:green', 'Rapamycin':'tab:purple'}\n",
    "\n",
    "#Visualize each representative\n",
    "for rank_i in range(len(plotL)):\n",
    "    print(' - Rank '+str(rank_i+1)+' (sort by the main effect of '+variable+'):')\n",
    "    module = plotL[rank_i]\n",
    "    #Check module summary\n",
    "    tempDF1 = pd.DataFrame(moduleDF.loc[module]).T\n",
    "    display(tempDF1)\n",
    "    \n",
    "    #Select RMS\n",
    "    tempDF1 = tempDF.loc[tempDF['ModuleID']==module]\n",
    "    \n",
    "    #Check RMS summary\n",
    "    tempDF2 = tempDF1.groupby(['Group'])['RMS'].agg(['count', 'mean', 'std'])\n",
    "    tempL1 = []\n",
    "    tempL2 = []\n",
    "    for row_n in tempDF2.index.tolist():\n",
    "        count, mean, std = tempDF2.loc[row_n]\n",
    "        tempL1.append(mean - 1.96*std/np.sqrt(count))\n",
    "        tempL2.append(mean + 1.96*std/np.sqrt(count))\n",
    "    tempDF2['0.025'] = tempL1\n",
    "    tempDF2['0.975'] = tempL2\n",
    "    tempDF2 = tempDF2.loc[list(tempD1.keys())]#Sort\n",
    "    display(tempDF2)\n",
    "    \n",
    "    #Prepare significance labels\n",
    "    ##Retrieve statistical significance\n",
    "    tempS = pvalDF.loc[module, pvalDF.columns.str.contains('-vs-')]#Adjusted P-value\n",
    "    tempS.name = 'AdjPval'\n",
    "    ##Clean\n",
    "    tempDF2 = tempS.index.to_series().str.split(pat='-vs-', expand=True)\n",
    "    tempDF2 = tempDF2.rename(columns={0:'Contrast', 1:'Baseline'})\n",
    "    tempDF2 = pd.merge(tempDF2, tempS, left_index=True, right_index=True, how='left')\n",
    "    tempDF2['Contrast'] = tempDF2['Contrast'].map(tempD0)\n",
    "    tempDF2['Baseline'] = tempDF2['Baseline'].map(tempD0)\n",
    "    ##Convert p-value to label\n",
    "    tempL = []\n",
    "    for row_i in range(len(tempDF2)):\n",
    "        pval = tempDF2['AdjPval'].iloc[row_i]\n",
    "        if pval<0.001:\n",
    "            tempL.append('***')\n",
    "        elif pval<0.01:\n",
    "            tempL.append('**')\n",
    "        elif pval<0.05:\n",
    "            tempL.append('*')\n",
    "        else:\n",
    "            pval_text = Decimal(str(pval)).quantize(Decimal('0.01'), rounding=ROUND_HALF_UP)\n",
    "            tempL.append(r'$P$ = '+str(pval_text))\n",
    "    tempDF2['SignifLabel'] = tempL\n",
    "    display(tempDF2)\n",
    "    \n",
    "    #Visualization\n",
    "    ymax = 1.0\n",
    "    ymin = 0.0\n",
    "    yinter = 0.2\n",
    "    ymargin_t = 0.4\n",
    "    ymargin_b = 0.05\n",
    "    aline_ymin = 1.0\n",
    "    aline_ymargin = 0.125\n",
    "    sns.set(style='ticks', font='Arial', context='talk')\n",
    "    plt.figure(figsize=(2, 4))\n",
    "    sns.pointplot(data=tempDF1, x='Group', y='RMS', order=list(tempD1.keys()), palette=tempD1,\n",
    "                  markers='o', dodge=False, join=False, capsize=0.6, estimator=np.mean, ci=95)\n",
    "    p = sns.stripplot(data=tempDF1, x='Group', y='RMS',\n",
    "                      order=list(tempD1.keys()), palette=tempD1, dodge=False, jitter=0.15,\n",
    "                      size=5, edgecolor='black', linewidth=1, **{'marker':'o', 'alpha':0.5})\n",
    "    ##Set axis\n",
    "    sns.despine()\n",
    "    p.set(ylim=(ymin-ymargin_b, ymax+ymargin_t), yticks=np.arange(ymin, ymax + yinter/10, yinter))\n",
    "    plt.setp(p.get_xticklabels(), rotation=70, horizontalalignment='right',\n",
    "             verticalalignment='center', rotation_mode='anchor')\n",
    "    ##Add significance labels\n",
    "    for row_i in range(len(tempDF2)):\n",
    "        #Baseline\n",
    "        group_0 = tempDF2['Baseline'].iloc[row_i]\n",
    "        index_0 = list(tempD1.keys()).index(group_0)\n",
    "        xcoord_0 = index_0\n",
    "        #Contrast\n",
    "        group_1 = tempDF2['Contrast'].iloc[row_i]\n",
    "        index_1 = list(tempD1.keys()).index(group_1)\n",
    "        xcoord_1 = index_1\n",
    "        #Standard point of marker\n",
    "        xcoord = (xcoord_0+xcoord_1)/2\n",
    "        ycoord = aline_ymin + aline_ymargin*row_i\n",
    "        label = tempDF2['SignifLabel'].iloc[row_i]\n",
    "        #Add annotation lines\n",
    "        aline_offset = yinter/5\n",
    "        aline_length = yinter/5 + aline_offset/2\n",
    "        plt.plot([xcoord_0, xcoord_0, xcoord_1, xcoord_1],\n",
    "                 [ycoord+aline_offset, ycoord+aline_length, ycoord+aline_length, ycoord+aline_offset],\n",
    "                 lw=1.5, c='k')\n",
    "        #Add annotation text\n",
    "        if label in ['***', '**', '*']:\n",
    "            text_offset = yinter/21\n",
    "            p.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                       horizontalalignment='center', verticalalignment='bottom',\n",
    "                       fontsize='medium', color='k')\n",
    "        else:\n",
    "            text_offset = yinter/3.5\n",
    "            p.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                       horizontalalignment='center', verticalalignment='bottom',\n",
    "                       fontsize='x-small', color='k')\n",
    "    ##Add annotation\n",
    "    p.set_title('Consensus:\\nown group', {'fontsize':'small'})\n",
    "    xoff = 0.015\n",
    "    yoff = 0.01\n",
    "    rect = plt.Rectangle((xoff, 1), 1-xoff, 0.155,#Manual adjustment\n",
    "                         transform=p.axes.transAxes, facecolor=plt.get_cmap('tab20')(15),\n",
    "                         clip_on=False, linewidth=0, zorder=0.5)\n",
    "    p.add_patch(rect)\n",
    "    ##Set axis label and title\n",
    "    plt.setp(p, xlabel='', ylabel='Sample RMS\\n(Mean = Module RCI)')\n",
    "    plt.suptitle(module, size='small',\n",
    "                 verticalalignment='bottom', horizontalalignment='center', wrap=True, y=1.0)\n",
    "    ##Save\n",
    "    #fileDir = './ExportFigures/'\n",
    "    #ipynbName = '230214_LC-M001-proteomics-DIRAC-ver7-2_DIRAC-GOBP_'\n",
    "    #fileName = 'RCI-pointplot-'+module.replace('GO:', 'GO')+'.pdf'\n",
    "    #plt.gcf().savefig(fileDir+ipynbName+fileName, dpi=300, bbox_inches='tight', pad_inches=0.04, transparent=True)\n",
    "    plt.show()\n",
    "    print('')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 4-5-3. Modules tightened specifically by 17aE2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Prepare the target module set\n",
    "posL = ['17aE2-vs-Ctrl']\n",
    "negL = ['Aca-vs-Ctrl', 'Rapa-vs-Ctrl']\n",
    "regulation = 'Tightened'\n",
    "tempS = pd.Series(np.repeat(True, len(pvalDF)), index=pvalDF.index)#Initialize\n",
    "tempDF = pvalDF.loc[:, pvalDF.columns.str.contains('-vs-')]#Adjusted P-value\n",
    "for col_n in tempDF.columns.tolist():\n",
    "    tempS1 = pvalDF[col_n]\n",
    "    tempS2 = diffDF[col_n]\n",
    "    if col_n in posL:\n",
    "        if regulation=='Tightened':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2>0)\n",
    "        elif regulation=='Loosened':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2<0)\n",
    "    elif col_n in negL:\n",
    "        tempS3 = (tempS1>=0.05)\n",
    "        #Significance for inverse regulation\n",
    "        if regulation=='Tightened':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2<0)\n",
    "        elif regulation=='Loosened':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2>0)\n",
    "        tempS1 = tempS3 | tempS1\n",
    "    else:\n",
    "        tempS1 = (tempS1>=0.0)\n",
    "    #Update True\n",
    "    tempS = tempS & tempS1\n",
    "tempL = tempS.loc[tempS.tolist()].index.tolist()\n",
    "print(len(tempL), regulation.lower()+' modules with significance in', posL, 'but not in', negL)\n",
    "\n",
    "#Select representatives\n",
    "topX = np.min([30, len(tempL)])\n",
    "topX_plot = np.min([10, len(tempL)])\n",
    "tempDF = statDF.loc[:, statDF.columns.str.contains('Pval$')]\n",
    "tempDF = pd.merge(statDF['ModuleName'], tempDF, left_index=True, right_index=True, how='left')\n",
    "tempDF = tempDF.loc[tempL].sort_values(by=variable+'_AdjPval', ascending=True)\n",
    "print('Top', topX, 'modules (sort by the main effect of '+variable+'):')\n",
    "display(tempDF.iloc[:topX])\n",
    "plotL = tempDF.index.tolist()[:topX_plot]\n",
    "\n",
    "#Prepare DF for plot\n",
    "tempDF = rmsDF_kk.reset_index().melt(var_name='SampleID', value_name='RMS', id_vars='ModuleID')\n",
    "tempDF1 = sampleDF.reset_index()[['SampleID', 'Phenotype']]\n",
    "tempDF = pd.merge(tempDF, tempDF1, on='SampleID', how='left')\n",
    "\n",
    "#Prepare label and color\n",
    "tempD0 = {'Ctrl':'Control', 'Aca':'Acarbose',\n",
    "          '17aE2':'17'+r'$\\alpha$'+'-Estradiol', 'Rapa':'Rapamycin'}\n",
    "tempDF['Group'] = tempDF['Phenotype'].map(tempD0)\n",
    "tempD1 = {'Control':'tab:blue', 'Acarbose':'tab:red',\n",
    "          '17'+r'$\\alpha$'+'-Estradiol':'tab:green', 'Rapamycin':'tab:purple'}\n",
    "\n",
    "#Visualize each representative\n",
    "for rank_i in range(len(plotL)):\n",
    "    print(' - Rank '+str(rank_i+1)+' (sort by the main effect of '+variable+'):')\n",
    "    module = plotL[rank_i]\n",
    "    #Check module summary\n",
    "    tempDF1 = pd.DataFrame(moduleDF.loc[module]).T\n",
    "    display(tempDF1)\n",
    "    \n",
    "    #Select RMS\n",
    "    tempDF1 = tempDF.loc[tempDF['ModuleID']==module]\n",
    "    \n",
    "    #Check RMS summary\n",
    "    tempDF2 = tempDF1.groupby(['Group'])['RMS'].agg(['count', 'mean', 'std'])\n",
    "    tempL1 = []\n",
    "    tempL2 = []\n",
    "    for row_n in tempDF2.index.tolist():\n",
    "        count, mean, std = tempDF2.loc[row_n]\n",
    "        tempL1.append(mean - 1.96*std/np.sqrt(count))\n",
    "        tempL2.append(mean + 1.96*std/np.sqrt(count))\n",
    "    tempDF2['0.025'] = tempL1\n",
    "    tempDF2['0.975'] = tempL2\n",
    "    tempDF2 = tempDF2.loc[list(tempD1.keys())]#Sort\n",
    "    display(tempDF2)\n",
    "    \n",
    "    #Prepare significance labels\n",
    "    ##Retrieve statistical significance\n",
    "    tempS = pvalDF.loc[module, pvalDF.columns.str.contains('-vs-')]#Adjusted P-value\n",
    "    tempS.name = 'AdjPval'\n",
    "    ##Clean\n",
    "    tempDF2 = tempS.index.to_series().str.split(pat='-vs-', expand=True)\n",
    "    tempDF2 = tempDF2.rename(columns={0:'Contrast', 1:'Baseline'})\n",
    "    tempDF2 = pd.merge(tempDF2, tempS, left_index=True, right_index=True, how='left')\n",
    "    tempDF2['Contrast'] = tempDF2['Contrast'].map(tempD0)\n",
    "    tempDF2['Baseline'] = tempDF2['Baseline'].map(tempD0)\n",
    "    ##Convert p-value to label\n",
    "    tempL = []\n",
    "    for row_i in range(len(tempDF2)):\n",
    "        pval = tempDF2['AdjPval'].iloc[row_i]\n",
    "        if pval<0.001:\n",
    "            tempL.append('***')\n",
    "        elif pval<0.01:\n",
    "            tempL.append('**')\n",
    "        elif pval<0.05:\n",
    "            tempL.append('*')\n",
    "        else:\n",
    "            pval_text = Decimal(str(pval)).quantize(Decimal('0.01'), rounding=ROUND_HALF_UP)\n",
    "            tempL.append(r'$P$ = '+str(pval_text))\n",
    "    tempDF2['SignifLabel'] = tempL\n",
    "    display(tempDF2)\n",
    "    \n",
    "    #Visualization\n",
    "    ymax = 1.0\n",
    "    ymin = 0.0\n",
    "    yinter = 0.2\n",
    "    ymargin_t = 0.4\n",
    "    ymargin_b = 0.05\n",
    "    aline_ymin = 1.0\n",
    "    aline_ymargin = 0.125\n",
    "    sns.set(style='ticks', font='Arial', context='talk')\n",
    "    plt.figure(figsize=(2, 4))\n",
    "    sns.pointplot(data=tempDF1, x='Group', y='RMS', order=list(tempD1.keys()), palette=tempD1,\n",
    "                  markers='o', dodge=False, join=False, capsize=0.6, estimator=np.mean, ci=95)\n",
    "    p = sns.stripplot(data=tempDF1, x='Group', y='RMS',\n",
    "                      order=list(tempD1.keys()), palette=tempD1, dodge=False, jitter=0.15,\n",
    "                      size=5, edgecolor='black', linewidth=1, **{'marker':'o', 'alpha':0.5})\n",
    "    ##Set axis\n",
    "    sns.despine()\n",
    "    p.set(ylim=(ymin-ymargin_b, ymax+ymargin_t), yticks=np.arange(ymin, ymax + yinter/10, yinter))\n",
    "    plt.setp(p.get_xticklabels(), rotation=70, horizontalalignment='right',\n",
    "             verticalalignment='center', rotation_mode='anchor')\n",
    "    ##Add significance labels\n",
    "    for row_i in range(len(tempDF2)):\n",
    "        #Baseline\n",
    "        group_0 = tempDF2['Baseline'].iloc[row_i]\n",
    "        index_0 = list(tempD1.keys()).index(group_0)\n",
    "        xcoord_0 = index_0\n",
    "        #Contrast\n",
    "        group_1 = tempDF2['Contrast'].iloc[row_i]\n",
    "        index_1 = list(tempD1.keys()).index(group_1)\n",
    "        xcoord_1 = index_1\n",
    "        #Standard point of marker\n",
    "        xcoord = (xcoord_0+xcoord_1)/2\n",
    "        ycoord = aline_ymin + aline_ymargin*row_i\n",
    "        label = tempDF2['SignifLabel'].iloc[row_i]\n",
    "        #Add annotation lines\n",
    "        aline_offset = yinter/5\n",
    "        aline_length = yinter/5 + aline_offset/2\n",
    "        plt.plot([xcoord_0, xcoord_0, xcoord_1, xcoord_1],\n",
    "                 [ycoord+aline_offset, ycoord+aline_length, ycoord+aline_length, ycoord+aline_offset],\n",
    "                 lw=1.5, c='k')\n",
    "        #Add annotation text\n",
    "        if label in ['***', '**', '*']:\n",
    "            text_offset = yinter/21\n",
    "            p.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                       horizontalalignment='center', verticalalignment='bottom',\n",
    "                       fontsize='medium', color='k')\n",
    "        else:\n",
    "            text_offset = yinter/3.5\n",
    "            p.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                       horizontalalignment='center', verticalalignment='bottom',\n",
    "                       fontsize='x-small', color='k')\n",
    "    ##Add annotation\n",
    "    p.set_title('Consensus:\\nown group', {'fontsize':'small'})\n",
    "    xoff = 0.015\n",
    "    yoff = 0.01\n",
    "    rect = plt.Rectangle((xoff, 1), 1-xoff, 0.155,#Manual adjustment\n",
    "                         transform=p.axes.transAxes, facecolor=plt.get_cmap('tab20')(15),\n",
    "                         clip_on=False, linewidth=0, zorder=0.5)\n",
    "    p.add_patch(rect)\n",
    "    ##Set axis label and title\n",
    "    plt.setp(p, xlabel='', ylabel='Sample RMS\\n(Mean = Module RCI)')\n",
    "    plt.suptitle(module, size='small',\n",
    "                 verticalalignment='bottom', horizontalalignment='center', wrap=True, y=1.0)\n",
    "    ##Save\n",
    "    #fileDir = './ExportFigures/'\n",
    "    #ipynbName = '230214_LC-M001-proteomics-DIRAC-ver7-2_DIRAC-GOBP_'\n",
    "    #fileName = 'RCI-pointplot-'+module.replace('GO:', 'GO')+'.pdf'\n",
    "    #plt.gcf().savefig(fileDir+ipynbName+fileName, dpi=300, bbox_inches='tight', pad_inches=0.04, transparent=True)\n",
    "    plt.show()\n",
    "    print('')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 4-5-4. Modules tightened specifically by Rapa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Prepare the target module set\n",
    "posL = ['Rapa-vs-Ctrl']\n",
    "negL = ['Aca-vs-Ctrl', '17aE2-vs-Ctrl']\n",
    "regulation = 'Tightened'\n",
    "tempS = pd.Series(np.repeat(True, len(pvalDF)), index=pvalDF.index)#Initialize\n",
    "tempDF = pvalDF.loc[:, pvalDF.columns.str.contains('-vs-')]#Adjusted P-value\n",
    "for col_n in tempDF.columns.tolist():\n",
    "    tempS1 = pvalDF[col_n]\n",
    "    tempS2 = diffDF[col_n]\n",
    "    if col_n in posL:\n",
    "        if regulation=='Tightened':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2>0)\n",
    "        elif regulation=='Loosened':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2<0)\n",
    "    elif col_n in negL:\n",
    "        tempS3 = (tempS1>=0.05)\n",
    "        #Significance for inverse regulation\n",
    "        if regulation=='Tightened':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2<0)\n",
    "        elif regulation=='Loosened':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2>0)\n",
    "        tempS1 = tempS3 | tempS1\n",
    "    else:\n",
    "        tempS1 = (tempS1>=0.0)\n",
    "    #Update True\n",
    "    tempS = tempS & tempS1\n",
    "tempL = tempS.loc[tempS.tolist()].index.tolist()\n",
    "print(len(tempL), regulation.lower()+' modules with significance in', posL, 'but not in', negL)\n",
    "\n",
    "#Select representatives\n",
    "topX = np.min([30, len(tempL)])\n",
    "topX_plot = np.min([10, len(tempL)])\n",
    "tempDF = statDF.loc[:, statDF.columns.str.contains('Pval$')]\n",
    "tempDF = pd.merge(statDF['ModuleName'], tempDF, left_index=True, right_index=True, how='left')\n",
    "tempDF = tempDF.loc[tempL].sort_values(by=variable+'_AdjPval', ascending=True)\n",
    "print('Top', topX, 'modules (sort by the main effect of '+variable+'):')\n",
    "display(tempDF.iloc[:topX])\n",
    "plotL = tempDF.index.tolist()[:topX_plot]\n",
    "\n",
    "#Prepare DF for plot\n",
    "tempDF = rmsDF_kk.reset_index().melt(var_name='SampleID', value_name='RMS', id_vars='ModuleID')\n",
    "tempDF1 = sampleDF.reset_index()[['SampleID', 'Phenotype']]\n",
    "tempDF = pd.merge(tempDF, tempDF1, on='SampleID', how='left')\n",
    "\n",
    "#Prepare label and color\n",
    "tempD0 = {'Ctrl':'Control', 'Aca':'Acarbose',\n",
    "          '17aE2':'17'+r'$\\alpha$'+'-Estradiol', 'Rapa':'Rapamycin'}\n",
    "tempDF['Group'] = tempDF['Phenotype'].map(tempD0)\n",
    "tempD1 = {'Control':'tab:blue', 'Acarbose':'tab:red',\n",
    "          '17'+r'$\\alpha$'+'-Estradiol':'tab:green', 'Rapamycin':'tab:purple'}\n",
    "\n",
    "#Visualize each representative\n",
    "for rank_i in range(len(plotL)):\n",
    "    print(' - Rank '+str(rank_i+1)+' (sort by the main effect of '+variable+'):')\n",
    "    module = plotL[rank_i]\n",
    "    #Check module summary\n",
    "    tempDF1 = pd.DataFrame(moduleDF.loc[module]).T\n",
    "    display(tempDF1)\n",
    "    \n",
    "    #Select RMS\n",
    "    tempDF1 = tempDF.loc[tempDF['ModuleID']==module]\n",
    "    \n",
    "    #Check RMS summary\n",
    "    tempDF2 = tempDF1.groupby(['Group'])['RMS'].agg(['count', 'mean', 'std'])\n",
    "    tempL1 = []\n",
    "    tempL2 = []\n",
    "    for row_n in tempDF2.index.tolist():\n",
    "        count, mean, std = tempDF2.loc[row_n]\n",
    "        tempL1.append(mean - 1.96*std/np.sqrt(count))\n",
    "        tempL2.append(mean + 1.96*std/np.sqrt(count))\n",
    "    tempDF2['0.025'] = tempL1\n",
    "    tempDF2['0.975'] = tempL2\n",
    "    tempDF2 = tempDF2.loc[list(tempD1.keys())]#Sort\n",
    "    display(tempDF2)\n",
    "    \n",
    "    #Prepare significance labels\n",
    "    ##Retrieve statistical significance\n",
    "    tempS = pvalDF.loc[module, pvalDF.columns.str.contains('-vs-')]#Adjusted P-value\n",
    "    tempS.name = 'AdjPval'\n",
    "    ##Clean\n",
    "    tempDF2 = tempS.index.to_series().str.split(pat='-vs-', expand=True)\n",
    "    tempDF2 = tempDF2.rename(columns={0:'Contrast', 1:'Baseline'})\n",
    "    tempDF2 = pd.merge(tempDF2, tempS, left_index=True, right_index=True, how='left')\n",
    "    tempDF2['Contrast'] = tempDF2['Contrast'].map(tempD0)\n",
    "    tempDF2['Baseline'] = tempDF2['Baseline'].map(tempD0)\n",
    "    ##Convert p-value to label\n",
    "    tempL = []\n",
    "    for row_i in range(len(tempDF2)):\n",
    "        pval = tempDF2['AdjPval'].iloc[row_i]\n",
    "        if pval<0.001:\n",
    "            tempL.append('***')\n",
    "        elif pval<0.01:\n",
    "            tempL.append('**')\n",
    "        elif pval<0.05:\n",
    "            tempL.append('*')\n",
    "        else:\n",
    "            pval_text = Decimal(str(pval)).quantize(Decimal('0.01'), rounding=ROUND_HALF_UP)\n",
    "            tempL.append(r'$P$ = '+str(pval_text))\n",
    "    tempDF2['SignifLabel'] = tempL\n",
    "    display(tempDF2)\n",
    "    \n",
    "    #Visualization\n",
    "    ymax = 1.0\n",
    "    ymin = 0.0\n",
    "    yinter = 0.2\n",
    "    ymargin_t = 0.4\n",
    "    ymargin_b = 0.05\n",
    "    aline_ymin = 1.0\n",
    "    aline_ymargin = 0.125\n",
    "    sns.set(style='ticks', font='Arial', context='talk')\n",
    "    plt.figure(figsize=(2, 4))\n",
    "    sns.pointplot(data=tempDF1, x='Group', y='RMS', order=list(tempD1.keys()), palette=tempD1,\n",
    "                  markers='o', dodge=False, join=False, capsize=0.6, estimator=np.mean, ci=95)\n",
    "    p = sns.stripplot(data=tempDF1, x='Group', y='RMS',\n",
    "                      order=list(tempD1.keys()), palette=tempD1, dodge=False, jitter=0.15,\n",
    "                      size=5, edgecolor='black', linewidth=1, **{'marker':'o', 'alpha':0.5})\n",
    "    ##Set axis\n",
    "    sns.despine()\n",
    "    p.set(ylim=(ymin-ymargin_b, ymax+ymargin_t), yticks=np.arange(ymin, ymax + yinter/10, yinter))\n",
    "    plt.setp(p.get_xticklabels(), rotation=70, horizontalalignment='right',\n",
    "             verticalalignment='center', rotation_mode='anchor')\n",
    "    ##Add significance labels\n",
    "    for row_i in range(len(tempDF2)):\n",
    "        #Baseline\n",
    "        group_0 = tempDF2['Baseline'].iloc[row_i]\n",
    "        index_0 = list(tempD1.keys()).index(group_0)\n",
    "        xcoord_0 = index_0\n",
    "        #Contrast\n",
    "        group_1 = tempDF2['Contrast'].iloc[row_i]\n",
    "        index_1 = list(tempD1.keys()).index(group_1)\n",
    "        xcoord_1 = index_1\n",
    "        #Standard point of marker\n",
    "        xcoord = (xcoord_0+xcoord_1)/2\n",
    "        ycoord = aline_ymin + aline_ymargin*row_i\n",
    "        label = tempDF2['SignifLabel'].iloc[row_i]\n",
    "        #Add annotation lines\n",
    "        aline_offset = yinter/5\n",
    "        aline_length = yinter/5 + aline_offset/2\n",
    "        plt.plot([xcoord_0, xcoord_0, xcoord_1, xcoord_1],\n",
    "                 [ycoord+aline_offset, ycoord+aline_length, ycoord+aline_length, ycoord+aline_offset],\n",
    "                 lw=1.5, c='k')\n",
    "        #Add annotation text\n",
    "        if label in ['***', '**', '*']:\n",
    "            text_offset = yinter/21\n",
    "            p.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                       horizontalalignment='center', verticalalignment='bottom',\n",
    "                       fontsize='medium', color='k')\n",
    "        else:\n",
    "            text_offset = yinter/3.5\n",
    "            p.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                       horizontalalignment='center', verticalalignment='bottom',\n",
    "                       fontsize='x-small', color='k')\n",
    "    ##Add annotation\n",
    "    p.set_title('Consensus:\\nown group', {'fontsize':'small'})\n",
    "    xoff = 0.015\n",
    "    yoff = 0.01\n",
    "    rect = plt.Rectangle((xoff, 1), 1-xoff, 0.155,#Manual adjustment\n",
    "                         transform=p.axes.transAxes, facecolor=plt.get_cmap('tab20')(15),\n",
    "                         clip_on=False, linewidth=0, zorder=0.5)\n",
    "    p.add_patch(rect)\n",
    "    ##Set axis label and title\n",
    "    plt.setp(p, xlabel='', ylabel='Sample RMS\\n(Mean = Module RCI)')\n",
    "    plt.suptitle(module, size='small',\n",
    "                 verticalalignment='bottom', horizontalalignment='center', wrap=True, y=1.0)\n",
    "    ##Save\n",
    "    #fileDir = './ExportFigures/'\n",
    "    #ipynbName = '230214_LC-M001-proteomics-DIRAC-ver7-2_DIRAC-GOBP_'\n",
    "    #fileName = 'RCI-pointplot-'+module.replace('GO:', 'GO')+'.pdf'\n",
    "    #plt.gcf().savefig(fileDir+ipynbName+fileName, dpi=300, bbox_inches='tight', pad_inches=0.04, transparent=True)\n",
    "    plt.show()\n",
    "    print('')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 4-5-5. Top changed modules by Aca"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Prepare the target module set\n",
    "target = 'Aca-vs-Ctrl'\n",
    "tempL = pvalDF.loc[pvalDF[target]<0.05].index.tolist()\n",
    "print(len(tempL), 'changed modules with significance in '+target)\n",
    "\n",
    "#Select representatives\n",
    "topX = np.min([30, len(tempL)])\n",
    "topX_plot = np.min([3, len(tempL)])\n",
    "tempDF = statDF.loc[:, statDF.columns.str.contains('Pval$')]\n",
    "tempDF = pd.merge(statDF['ModuleName'], tempDF, left_index=True, right_index=True, how='left')\n",
    "tempDF = tempDF.loc[tempL].sort_values(by=target+'_AdjPval', ascending=True)\n",
    "print('Top', topX, 'modules (sort by significance in '+target+'):')\n",
    "display(tempDF.iloc[:topX])\n",
    "plotL = tempDF.index.tolist()[:topX_plot]\n",
    "\n",
    "#Prepare DF for plot\n",
    "tempDF = rmsDF_kk.reset_index().melt(var_name='SampleID', value_name='RMS', id_vars='ModuleID')\n",
    "tempDF1 = sampleDF.reset_index()[['SampleID', 'Phenotype']]\n",
    "tempDF = pd.merge(tempDF, tempDF1, on='SampleID', how='left')\n",
    "\n",
    "#Prepare label and color\n",
    "tempD0 = {'Ctrl':'Control', 'Aca':'Acarbose',\n",
    "          '17aE2':'17'+r'$\\alpha$'+'-Estradiol', 'Rapa':'Rapamycin'}\n",
    "tempDF['Group'] = tempDF['Phenotype'].map(tempD0)\n",
    "tempD1 = {'Control':'tab:blue', 'Acarbose':'tab:red',\n",
    "          '17'+r'$\\alpha$'+'-Estradiol':'tab:green', 'Rapamycin':'tab:purple'}\n",
    "\n",
    "#Visualize each representative\n",
    "for rank_i in range(len(plotL)):\n",
    "    print(' - Rank '+str(rank_i+1)+' (sort by the main effect of '+variable+'):')\n",
    "    module = plotL[rank_i]\n",
    "    #Check module summary\n",
    "    tempDF1 = pd.DataFrame(moduleDF.loc[module]).T\n",
    "    display(tempDF1)\n",
    "    \n",
    "    #Select RMS\n",
    "    tempDF1 = tempDF.loc[tempDF['ModuleID']==module]\n",
    "    \n",
    "    #Check RMS summary\n",
    "    tempDF2 = tempDF1.groupby(['Group'])['RMS'].agg(['count', 'mean', 'std'])\n",
    "    tempL1 = []\n",
    "    tempL2 = []\n",
    "    for row_n in tempDF2.index.tolist():\n",
    "        count, mean, std = tempDF2.loc[row_n]\n",
    "        tempL1.append(mean - 1.96*std/np.sqrt(count))\n",
    "        tempL2.append(mean + 1.96*std/np.sqrt(count))\n",
    "    tempDF2['0.025'] = tempL1\n",
    "    tempDF2['0.975'] = tempL2\n",
    "    tempDF2 = tempDF2.loc[list(tempD1.keys())]#Sort\n",
    "    display(tempDF2)\n",
    "    \n",
    "    #Prepare significance labels\n",
    "    ##Retrieve statistical significance\n",
    "    tempS = pvalDF.loc[module, pvalDF.columns.str.contains('-vs-')]#Adjusted P-value\n",
    "    tempS.name = 'AdjPval'\n",
    "    ##Clean\n",
    "    tempDF2 = tempS.index.to_series().str.split(pat='-vs-', expand=True)\n",
    "    tempDF2 = tempDF2.rename(columns={0:'Contrast', 1:'Baseline'})\n",
    "    tempDF2 = pd.merge(tempDF2, tempS, left_index=True, right_index=True, how='left')\n",
    "    tempDF2['Contrast'] = tempDF2['Contrast'].map(tempD0)\n",
    "    tempDF2['Baseline'] = tempDF2['Baseline'].map(tempD0)\n",
    "    ##Convert p-value to label\n",
    "    tempL = []\n",
    "    for row_i in range(len(tempDF2)):\n",
    "        pval = tempDF2['AdjPval'].iloc[row_i]\n",
    "        if pval<0.001:\n",
    "            tempL.append('***')\n",
    "        elif pval<0.01:\n",
    "            tempL.append('**')\n",
    "        elif pval<0.05:\n",
    "            tempL.append('*')\n",
    "        else:\n",
    "            pval_text = Decimal(str(pval)).quantize(Decimal('0.01'), rounding=ROUND_HALF_UP)\n",
    "            tempL.append(r'$P$ = '+str(pval_text))\n",
    "    tempDF2['SignifLabel'] = tempL\n",
    "    display(tempDF2)\n",
    "    \n",
    "    #Visualization\n",
    "    ymax = 1.0\n",
    "    ymin = 0.0\n",
    "    yinter = 0.2\n",
    "    ymargin_t = 0.4\n",
    "    ymargin_b = 0.05\n",
    "    aline_ymin = 1.0\n",
    "    aline_ymargin = 0.125\n",
    "    sns.set(style='ticks', font='Arial', context='talk')\n",
    "    plt.figure(figsize=(2, 4))\n",
    "    sns.pointplot(data=tempDF1, x='Group', y='RMS', order=list(tempD1.keys()), palette=tempD1,\n",
    "                  markers='o', dodge=False, join=False, capsize=0.6, estimator=np.mean, ci=95)\n",
    "    p = sns.stripplot(data=tempDF1, x='Group', y='RMS',\n",
    "                      order=list(tempD1.keys()), palette=tempD1, dodge=False, jitter=0.15,\n",
    "                      size=5, edgecolor='black', linewidth=1, **{'marker':'o', 'alpha':0.5})\n",
    "    ##Set axis\n",
    "    sns.despine()\n",
    "    p.set(ylim=(ymin-ymargin_b, ymax+ymargin_t), yticks=np.arange(ymin, ymax + yinter/10, yinter))\n",
    "    plt.setp(p.get_xticklabels(), rotation=70, horizontalalignment='right',\n",
    "             verticalalignment='center', rotation_mode='anchor')\n",
    "    ##Add significance labels\n",
    "    for row_i in range(len(tempDF2)):\n",
    "        #Baseline\n",
    "        group_0 = tempDF2['Baseline'].iloc[row_i]\n",
    "        index_0 = list(tempD1.keys()).index(group_0)\n",
    "        xcoord_0 = index_0\n",
    "        #Contrast\n",
    "        group_1 = tempDF2['Contrast'].iloc[row_i]\n",
    "        index_1 = list(tempD1.keys()).index(group_1)\n",
    "        xcoord_1 = index_1\n",
    "        #Standard point of marker\n",
    "        xcoord = (xcoord_0+xcoord_1)/2\n",
    "        ycoord = aline_ymin + aline_ymargin*row_i\n",
    "        label = tempDF2['SignifLabel'].iloc[row_i]\n",
    "        #Add annotation lines\n",
    "        aline_offset = yinter/5\n",
    "        aline_length = yinter/5 + aline_offset/2\n",
    "        plt.plot([xcoord_0, xcoord_0, xcoord_1, xcoord_1],\n",
    "                 [ycoord+aline_offset, ycoord+aline_length, ycoord+aline_length, ycoord+aline_offset],\n",
    "                 lw=1.5, c='k')\n",
    "        #Add annotation text\n",
    "        if label in ['***', '**', '*']:\n",
    "            text_offset = yinter/21\n",
    "            p.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                       horizontalalignment='center', verticalalignment='bottom',\n",
    "                       fontsize='medium', color='k')\n",
    "        else:\n",
    "            text_offset = yinter/3.5\n",
    "            p.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                       horizontalalignment='center', verticalalignment='bottom',\n",
    "                       fontsize='x-small', color='k')\n",
    "    ##Add annotation\n",
    "    p.set_title('Consensus:\\nown group', {'fontsize':'small'})\n",
    "    xoff = 0.015\n",
    "    yoff = 0.01\n",
    "    rect = plt.Rectangle((xoff, 1), 1-xoff, 0.155,#Manual adjustment\n",
    "                         transform=p.axes.transAxes, facecolor=plt.get_cmap('tab20')(15),\n",
    "                         clip_on=False, linewidth=0, zorder=0.5)\n",
    "    p.add_patch(rect)\n",
    "    ##Set axis label and title\n",
    "    plt.setp(p, xlabel='', ylabel='Sample RMS\\n(Mean = Module RCI)')\n",
    "    plt.suptitle(module, size='small',\n",
    "                 verticalalignment='bottom', horizontalalignment='center', wrap=True, y=1.0)\n",
    "    ##Save\n",
    "    #fileDir = './ExportFigures/'\n",
    "    #ipynbName = '230214_LC-M001-proteomics-DIRAC-ver7-2_DIRAC-GOBP_'\n",
    "    #fileName = 'RCI-pointplot-'+module.replace('GO:', 'GO')+'.pdf'\n",
    "    #plt.gcf().savefig(fileDir+ipynbName+fileName, dpi=300, bbox_inches='tight', pad_inches=0.04, transparent=True)\n",
    "    plt.show()\n",
    "    print('')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 4-5-6. Top changed modules by 17aE2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Prepare the target module set\n",
    "target = '17aE2-vs-Ctrl'\n",
    "tempL = pvalDF.loc[pvalDF[target]<0.05].index.tolist()\n",
    "print(len(tempL), 'changed modules with significance in '+target)\n",
    "\n",
    "#Select representatives\n",
    "topX = np.min([30, len(tempL)])\n",
    "topX_plot = np.min([3, len(tempL)])\n",
    "tempDF = statDF.loc[:, statDF.columns.str.contains('Pval$')]\n",
    "tempDF = pd.merge(statDF['ModuleName'], tempDF, left_index=True, right_index=True, how='left')\n",
    "tempDF = tempDF.loc[tempL].sort_values(by=target+'_AdjPval', ascending=True)\n",
    "print('Top', topX, 'modules (sort by significance in '+target+'):')\n",
    "display(tempDF.iloc[:topX])\n",
    "plotL = tempDF.index.tolist()[:topX_plot]\n",
    "\n",
    "#Prepare DF for plot\n",
    "tempDF = rmsDF_kk.reset_index().melt(var_name='SampleID', value_name='RMS', id_vars='ModuleID')\n",
    "tempDF1 = sampleDF.reset_index()[['SampleID', 'Phenotype']]\n",
    "tempDF = pd.merge(tempDF, tempDF1, on='SampleID', how='left')\n",
    "\n",
    "#Prepare label and color\n",
    "tempD0 = {'Ctrl':'Control', 'Aca':'Acarbose',\n",
    "          '17aE2':'17'+r'$\\alpha$'+'-Estradiol', 'Rapa':'Rapamycin'}\n",
    "tempDF['Group'] = tempDF['Phenotype'].map(tempD0)\n",
    "tempD1 = {'Control':'tab:blue', 'Acarbose':'tab:red',\n",
    "          '17'+r'$\\alpha$'+'-Estradiol':'tab:green', 'Rapamycin':'tab:purple'}\n",
    "\n",
    "#Visualize each representative\n",
    "for rank_i in range(len(plotL)):\n",
    "    print(' - Rank '+str(rank_i+1)+' (sort by the main effect of '+variable+'):')\n",
    "    module = plotL[rank_i]\n",
    "    #Check module summary\n",
    "    tempDF1 = pd.DataFrame(moduleDF.loc[module]).T\n",
    "    display(tempDF1)\n",
    "    \n",
    "    #Select RMS\n",
    "    tempDF1 = tempDF.loc[tempDF['ModuleID']==module]\n",
    "    \n",
    "    #Check RMS summary\n",
    "    tempDF2 = tempDF1.groupby(['Group'])['RMS'].agg(['count', 'mean', 'std'])\n",
    "    tempL1 = []\n",
    "    tempL2 = []\n",
    "    for row_n in tempDF2.index.tolist():\n",
    "        count, mean, std = tempDF2.loc[row_n]\n",
    "        tempL1.append(mean - 1.96*std/np.sqrt(count))\n",
    "        tempL2.append(mean + 1.96*std/np.sqrt(count))\n",
    "    tempDF2['0.025'] = tempL1\n",
    "    tempDF2['0.975'] = tempL2\n",
    "    tempDF2 = tempDF2.loc[list(tempD1.keys())]#Sort\n",
    "    display(tempDF2)\n",
    "    \n",
    "    #Prepare significance labels\n",
    "    ##Retrieve statistical significance\n",
    "    tempS = pvalDF.loc[module, pvalDF.columns.str.contains('-vs-')]#Adjusted P-value\n",
    "    tempS.name = 'AdjPval'\n",
    "    ##Clean\n",
    "    tempDF2 = tempS.index.to_series().str.split(pat='-vs-', expand=True)\n",
    "    tempDF2 = tempDF2.rename(columns={0:'Contrast', 1:'Baseline'})\n",
    "    tempDF2 = pd.merge(tempDF2, tempS, left_index=True, right_index=True, how='left')\n",
    "    tempDF2['Contrast'] = tempDF2['Contrast'].map(tempD0)\n",
    "    tempDF2['Baseline'] = tempDF2['Baseline'].map(tempD0)\n",
    "    ##Convert p-value to label\n",
    "    tempL = []\n",
    "    for row_i in range(len(tempDF2)):\n",
    "        pval = tempDF2['AdjPval'].iloc[row_i]\n",
    "        if pval<0.001:\n",
    "            tempL.append('***')\n",
    "        elif pval<0.01:\n",
    "            tempL.append('**')\n",
    "        elif pval<0.05:\n",
    "            tempL.append('*')\n",
    "        else:\n",
    "            pval_text = Decimal(str(pval)).quantize(Decimal('0.01'), rounding=ROUND_HALF_UP)\n",
    "            tempL.append(r'$P$ = '+str(pval_text))\n",
    "    tempDF2['SignifLabel'] = tempL\n",
    "    display(tempDF2)\n",
    "    \n",
    "    #Visualization\n",
    "    ymax = 1.0\n",
    "    ymin = 0.0\n",
    "    yinter = 0.2\n",
    "    ymargin_t = 0.4\n",
    "    ymargin_b = 0.05\n",
    "    aline_ymin = 1.0\n",
    "    aline_ymargin = 0.125\n",
    "    sns.set(style='ticks', font='Arial', context='talk')\n",
    "    plt.figure(figsize=(2, 4))\n",
    "    sns.pointplot(data=tempDF1, x='Group', y='RMS', order=list(tempD1.keys()), palette=tempD1,\n",
    "                  markers='o', dodge=False, join=False, capsize=0.6, estimator=np.mean, ci=95)\n",
    "    p = sns.stripplot(data=tempDF1, x='Group', y='RMS',\n",
    "                      order=list(tempD1.keys()), palette=tempD1, dodge=False, jitter=0.15,\n",
    "                      size=5, edgecolor='black', linewidth=1, **{'marker':'o', 'alpha':0.5})\n",
    "    ##Set axis\n",
    "    sns.despine()\n",
    "    p.set(ylim=(ymin-ymargin_b, ymax+ymargin_t), yticks=np.arange(ymin, ymax + yinter/10, yinter))\n",
    "    plt.setp(p.get_xticklabels(), rotation=70, horizontalalignment='right',\n",
    "             verticalalignment='center', rotation_mode='anchor')\n",
    "    ##Add significance labels\n",
    "    for row_i in range(len(tempDF2)):\n",
    "        #Baseline\n",
    "        group_0 = tempDF2['Baseline'].iloc[row_i]\n",
    "        index_0 = list(tempD1.keys()).index(group_0)\n",
    "        xcoord_0 = index_0\n",
    "        #Contrast\n",
    "        group_1 = tempDF2['Contrast'].iloc[row_i]\n",
    "        index_1 = list(tempD1.keys()).index(group_1)\n",
    "        xcoord_1 = index_1\n",
    "        #Standard point of marker\n",
    "        xcoord = (xcoord_0+xcoord_1)/2\n",
    "        ycoord = aline_ymin + aline_ymargin*row_i\n",
    "        label = tempDF2['SignifLabel'].iloc[row_i]\n",
    "        #Add annotation lines\n",
    "        aline_offset = yinter/5\n",
    "        aline_length = yinter/5 + aline_offset/2\n",
    "        plt.plot([xcoord_0, xcoord_0, xcoord_1, xcoord_1],\n",
    "                 [ycoord+aline_offset, ycoord+aline_length, ycoord+aline_length, ycoord+aline_offset],\n",
    "                 lw=1.5, c='k')\n",
    "        #Add annotation text\n",
    "        if label in ['***', '**', '*']:\n",
    "            text_offset = yinter/21\n",
    "            p.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                       horizontalalignment='center', verticalalignment='bottom',\n",
    "                       fontsize='medium', color='k')\n",
    "        else:\n",
    "            text_offset = yinter/3.5\n",
    "            p.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                       horizontalalignment='center', verticalalignment='bottom',\n",
    "                       fontsize='x-small', color='k')\n",
    "    ##Add annotation\n",
    "    p.set_title('Consensus:\\nown group', {'fontsize':'small'})\n",
    "    xoff = 0.015\n",
    "    yoff = 0.01\n",
    "    rect = plt.Rectangle((xoff, 1), 1-xoff, 0.155,#Manual adjustment\n",
    "                         transform=p.axes.transAxes, facecolor=plt.get_cmap('tab20')(15),\n",
    "                         clip_on=False, linewidth=0, zorder=0.5)\n",
    "    p.add_patch(rect)\n",
    "    ##Set axis label and title\n",
    "    plt.setp(p, xlabel='', ylabel='Sample RMS\\n(Mean = Module RCI)')\n",
    "    plt.suptitle(module, size='small',\n",
    "                 verticalalignment='bottom', horizontalalignment='center', wrap=True, y=1.0)\n",
    "    ##Save\n",
    "    #fileDir = './ExportFigures/'\n",
    "    #ipynbName = '230214_LC-M001-proteomics-DIRAC-ver7-2_DIRAC-GOBP_'\n",
    "    #fileName = 'RCI-pointplot-'+module.replace('GO:', 'GO')+'.pdf'\n",
    "    #plt.gcf().savefig(fileDir+ipynbName+fileName, dpi=300, bbox_inches='tight', pad_inches=0.04, transparent=True)\n",
    "    plt.show()\n",
    "    print('')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 4-5-7. Top changed modules by Rapa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Prepare the target module set\n",
    "target = 'Rapa-vs-Ctrl'\n",
    "tempL = pvalDF.loc[pvalDF[target]<0.05].index.tolist()\n",
    "print(len(tempL), 'changed modules with significance in '+target)\n",
    "\n",
    "#Select representatives\n",
    "topX = np.min([30, len(tempL)])\n",
    "topX_plot = np.min([3, len(tempL)])\n",
    "tempDF = statDF.loc[:, statDF.columns.str.contains('Pval$')]\n",
    "tempDF = pd.merge(statDF['ModuleName'], tempDF, left_index=True, right_index=True, how='left')\n",
    "tempDF = tempDF.loc[tempL].sort_values(by=target+'_AdjPval', ascending=True)\n",
    "print('Top', topX, 'modules (sort by significance in '+target+'):')\n",
    "display(tempDF.iloc[:topX])\n",
    "plotL = tempDF.index.tolist()[:topX_plot]\n",
    "\n",
    "#Prepare DF for plot\n",
    "tempDF = rmsDF_kk.reset_index().melt(var_name='SampleID', value_name='RMS', id_vars='ModuleID')\n",
    "tempDF1 = sampleDF.reset_index()[['SampleID', 'Phenotype']]\n",
    "tempDF = pd.merge(tempDF, tempDF1, on='SampleID', how='left')\n",
    "\n",
    "#Prepare label and color\n",
    "tempD0 = {'Ctrl':'Control', 'Aca':'Acarbose',\n",
    "          '17aE2':'17'+r'$\\alpha$'+'-Estradiol', 'Rapa':'Rapamycin'}\n",
    "tempDF['Group'] = tempDF['Phenotype'].map(tempD0)\n",
    "tempD1 = {'Control':'tab:blue', 'Acarbose':'tab:red',\n",
    "          '17'+r'$\\alpha$'+'-Estradiol':'tab:green', 'Rapamycin':'tab:purple'}\n",
    "\n",
    "#Visualize each representative\n",
    "for rank_i in range(len(plotL)):\n",
    "    print(' - Rank '+str(rank_i+1)+' (sort by the main effect of '+variable+'):')\n",
    "    module = plotL[rank_i]\n",
    "    #Check module summary\n",
    "    tempDF1 = pd.DataFrame(moduleDF.loc[module]).T\n",
    "    display(tempDF1)\n",
    "    \n",
    "    #Select RMS\n",
    "    tempDF1 = tempDF.loc[tempDF['ModuleID']==module]\n",
    "    \n",
    "    #Check RMS summary\n",
    "    tempDF2 = tempDF1.groupby(['Group'])['RMS'].agg(['count', 'mean', 'std'])\n",
    "    tempL1 = []\n",
    "    tempL2 = []\n",
    "    for row_n in tempDF2.index.tolist():\n",
    "        count, mean, std = tempDF2.loc[row_n]\n",
    "        tempL1.append(mean - 1.96*std/np.sqrt(count))\n",
    "        tempL2.append(mean + 1.96*std/np.sqrt(count))\n",
    "    tempDF2['0.025'] = tempL1\n",
    "    tempDF2['0.975'] = tempL2\n",
    "    tempDF2 = tempDF2.loc[list(tempD1.keys())]#Sort\n",
    "    display(tempDF2)\n",
    "    \n",
    "    #Prepare significance labels\n",
    "    ##Retrieve statistical significance\n",
    "    tempS = pvalDF.loc[module, pvalDF.columns.str.contains('-vs-')]#Adjusted P-value\n",
    "    tempS.name = 'AdjPval'\n",
    "    ##Clean\n",
    "    tempDF2 = tempS.index.to_series().str.split(pat='-vs-', expand=True)\n",
    "    tempDF2 = tempDF2.rename(columns={0:'Contrast', 1:'Baseline'})\n",
    "    tempDF2 = pd.merge(tempDF2, tempS, left_index=True, right_index=True, how='left')\n",
    "    tempDF2['Contrast'] = tempDF2['Contrast'].map(tempD0)\n",
    "    tempDF2['Baseline'] = tempDF2['Baseline'].map(tempD0)\n",
    "    ##Convert p-value to label\n",
    "    tempL = []\n",
    "    for row_i in range(len(tempDF2)):\n",
    "        pval = tempDF2['AdjPval'].iloc[row_i]\n",
    "        if pval<0.001:\n",
    "            tempL.append('***')\n",
    "        elif pval<0.01:\n",
    "            tempL.append('**')\n",
    "        elif pval<0.05:\n",
    "            tempL.append('*')\n",
    "        else:\n",
    "            pval_text = Decimal(str(pval)).quantize(Decimal('0.01'), rounding=ROUND_HALF_UP)\n",
    "            tempL.append(r'$P$ = '+str(pval_text))\n",
    "    tempDF2['SignifLabel'] = tempL\n",
    "    display(tempDF2)\n",
    "    \n",
    "    #Visualization\n",
    "    ymax = 1.0\n",
    "    ymin = 0.0\n",
    "    yinter = 0.2\n",
    "    ymargin_t = 0.4\n",
    "    ymargin_b = 0.05\n",
    "    aline_ymin = 1.0\n",
    "    aline_ymargin = 0.125\n",
    "    sns.set(style='ticks', font='Arial', context='talk')\n",
    "    plt.figure(figsize=(2, 4))\n",
    "    sns.pointplot(data=tempDF1, x='Group', y='RMS', order=list(tempD1.keys()), palette=tempD1,\n",
    "                  markers='o', dodge=False, join=False, capsize=0.6, estimator=np.mean, ci=95)\n",
    "    p = sns.stripplot(data=tempDF1, x='Group', y='RMS',\n",
    "                      order=list(tempD1.keys()), palette=tempD1, dodge=False, jitter=0.15,\n",
    "                      size=5, edgecolor='black', linewidth=1, **{'marker':'o', 'alpha':0.5})\n",
    "    ##Set axis\n",
    "    sns.despine()\n",
    "    p.set(ylim=(ymin-ymargin_b, ymax+ymargin_t), yticks=np.arange(ymin, ymax + yinter/10, yinter))\n",
    "    plt.setp(p.get_xticklabels(), rotation=70, horizontalalignment='right',\n",
    "             verticalalignment='center', rotation_mode='anchor')\n",
    "    ##Add significance labels\n",
    "    for row_i in range(len(tempDF2)):\n",
    "        #Baseline\n",
    "        group_0 = tempDF2['Baseline'].iloc[row_i]\n",
    "        index_0 = list(tempD1.keys()).index(group_0)\n",
    "        xcoord_0 = index_0\n",
    "        #Contrast\n",
    "        group_1 = tempDF2['Contrast'].iloc[row_i]\n",
    "        index_1 = list(tempD1.keys()).index(group_1)\n",
    "        xcoord_1 = index_1\n",
    "        #Standard point of marker\n",
    "        xcoord = (xcoord_0+xcoord_1)/2\n",
    "        ycoord = aline_ymin + aline_ymargin*row_i\n",
    "        label = tempDF2['SignifLabel'].iloc[row_i]\n",
    "        #Add annotation lines\n",
    "        aline_offset = yinter/5\n",
    "        aline_length = yinter/5 + aline_offset/2\n",
    "        plt.plot([xcoord_0, xcoord_0, xcoord_1, xcoord_1],\n",
    "                 [ycoord+aline_offset, ycoord+aline_length, ycoord+aline_length, ycoord+aline_offset],\n",
    "                 lw=1.5, c='k')\n",
    "        #Add annotation text\n",
    "        if label in ['***', '**', '*']:\n",
    "            text_offset = yinter/21\n",
    "            p.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                       horizontalalignment='center', verticalalignment='bottom',\n",
    "                       fontsize='medium', color='k')\n",
    "        else:\n",
    "            text_offset = yinter/3.5\n",
    "            p.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                       horizontalalignment='center', verticalalignment='bottom',\n",
    "                       fontsize='x-small', color='k')\n",
    "    ##Add annotation\n",
    "    p.set_title('Consensus:\\nown group', {'fontsize':'small'})\n",
    "    xoff = 0.015\n",
    "    yoff = 0.01\n",
    "    rect = plt.Rectangle((xoff, 1), 1-xoff, 0.155,#Manual adjustment\n",
    "                         transform=p.axes.transAxes, facecolor=plt.get_cmap('tab20')(15),\n",
    "                         clip_on=False, linewidth=0, zorder=0.5)\n",
    "    p.add_patch(rect)\n",
    "    ##Set axis label and title\n",
    "    plt.setp(p, xlabel='', ylabel='Sample RMS\\n(Mean = Module RCI)')\n",
    "    plt.suptitle(module, size='small',\n",
    "                 verticalalignment='bottom', horizontalalignment='center', wrap=True, y=1.0)\n",
    "    ##Save\n",
    "    #fileDir = './ExportFigures/'\n",
    "    #ipynbName = '230214_LC-M001-proteomics-DIRAC-ver7-2_DIRAC-GOBP_'\n",
    "    #fileName = 'RCI-pointplot-'+module.replace('GO:', 'GO')+'.pdf'\n",
    "    #plt.gcf().savefig(fileDir+ipynbName+fileName, dpi=300, bbox_inches='tight', pad_inches=0.04, transparent=True)\n",
    "    plt.show()\n",
    "    print('')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 4-5-8. Top RCI modules in Ctrl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "target = 'Ctrl'\n",
    "\n",
    "#Select representatives\n",
    "topX = 30\n",
    "topX_plot = 3\n",
    "tempDF = statDF.loc[:, statDF.columns.str.contains('Pval$')]\n",
    "tempDF = pd.merge(statDF['ModuleName'], tempDF, left_index=True, right_index=True, how='left')\n",
    "tempDF = pd.merge(tempDF, rciDF_kk, left_index=True, right_index=True, how='left')\n",
    "tempDF = tempDF.sort_values(by=target, ascending=False)\n",
    "print('Top', topX, 'modules (sort by RCI of '+target+'):')\n",
    "display(tempDF.iloc[:topX])\n",
    "plotL = tempDF.index.tolist()[:topX_plot]\n",
    "\n",
    "#Prepare DF for plot\n",
    "tempDF = rmsDF_kk.reset_index().melt(var_name='SampleID', value_name='RMS', id_vars='ModuleID')\n",
    "tempDF1 = sampleDF.reset_index()[['SampleID', 'Phenotype']]\n",
    "tempDF = pd.merge(tempDF, tempDF1, on='SampleID', how='left')\n",
    "\n",
    "#Prepare label and color\n",
    "tempD0 = {'Ctrl':'Control', 'Aca':'Acarbose',\n",
    "          '17aE2':'17'+r'$\\alpha$'+'-Estradiol', 'Rapa':'Rapamycin'}\n",
    "tempDF['Group'] = tempDF['Phenotype'].map(tempD0)\n",
    "tempD1 = {'Control':'tab:blue', 'Acarbose':'tab:red',\n",
    "          '17'+r'$\\alpha$'+'-Estradiol':'tab:green', 'Rapamycin':'tab:purple'}\n",
    "\n",
    "#Visualize each representative\n",
    "for rank_i in range(len(plotL)):\n",
    "    print(' - Rank '+str(rank_i+1)+' (sort by RCI of '+target+'):')\n",
    "    module = plotL[rank_i]\n",
    "    #Check module summary\n",
    "    tempDF1 = pd.DataFrame(moduleDF.loc[module]).T\n",
    "    display(tempDF1)\n",
    "    \n",
    "    #Select RMS\n",
    "    tempDF1 = tempDF.loc[tempDF['ModuleID']==module]\n",
    "    \n",
    "    #Check RMS summary\n",
    "    tempDF2 = tempDF1.groupby(['Group'])['RMS'].agg(['count', 'mean', 'std'])\n",
    "    tempL1 = []\n",
    "    tempL2 = []\n",
    "    for row_n in tempDF2.index.tolist():\n",
    "        count, mean, std = tempDF2.loc[row_n]\n",
    "        tempL1.append(mean - 1.96*std/np.sqrt(count))\n",
    "        tempL2.append(mean + 1.96*std/np.sqrt(count))\n",
    "    tempDF2['0.025'] = tempL1\n",
    "    tempDF2['0.975'] = tempL2\n",
    "    tempDF2 = tempDF2.loc[list(tempD1.keys())]#Sort\n",
    "    display(tempDF2)\n",
    "    \n",
    "    #Prepare significance labels\n",
    "    ##Re-prepare p-values because these modules can be filtered out in pvalDF\n",
    "    tempDF2 = statDF.loc[:, statDF.columns.str.contains('-vs-.*_AdjPval$')]\n",
    "    tempDF2.columns = tempDF2.columns.str.replace('_AdjPval$', '')\n",
    "    ##Retrieve statistical significance\n",
    "    tempS = tempDF2.loc[module]\n",
    "    tempS.name = 'AdjPval'\n",
    "    ##Clean\n",
    "    tempDF2 = tempS.index.to_series().str.split(pat='-vs-', expand=True)\n",
    "    tempDF2 = tempDF2.rename(columns={0:'Contrast', 1:'Baseline'})\n",
    "    tempDF2 = pd.merge(tempDF2, tempS, left_index=True, right_index=True, how='left')\n",
    "    tempDF2['Contrast'] = tempDF2['Contrast'].map(tempD0)\n",
    "    tempDF2['Baseline'] = tempDF2['Baseline'].map(tempD0)\n",
    "    ##Convert p-value to label\n",
    "    tempL = []\n",
    "    for row_i in range(len(tempDF2)):\n",
    "        pval = tempDF2['AdjPval'].iloc[row_i]\n",
    "        if pval<0.001:\n",
    "            tempL.append('***')\n",
    "        elif pval<0.01:\n",
    "            tempL.append('**')\n",
    "        elif pval<0.05:\n",
    "            tempL.append('*')\n",
    "        else:\n",
    "            pval_text = Decimal(str(pval)).quantize(Decimal('0.01'), rounding=ROUND_HALF_UP)\n",
    "            tempL.append(r'$P$ = '+str(pval_text))\n",
    "    tempDF2['SignifLabel'] = tempL\n",
    "    display(tempDF2)\n",
    "    \n",
    "    #Visualization\n",
    "    ymax = 1.0\n",
    "    ymin = 0.0\n",
    "    yinter = 0.2\n",
    "    ymargin_t = 0.4\n",
    "    ymargin_b = 0.05\n",
    "    aline_ymin = 1.0\n",
    "    aline_ymargin = 0.125\n",
    "    sns.set(style='ticks', font='Arial', context='talk')\n",
    "    plt.figure(figsize=(2, 4))\n",
    "    sns.pointplot(data=tempDF1, x='Group', y='RMS', order=list(tempD1.keys()), palette=tempD1,\n",
    "                  markers='o', dodge=False, join=False, capsize=0.6, estimator=np.mean, ci=95)\n",
    "    p = sns.stripplot(data=tempDF1, x='Group', y='RMS',\n",
    "                      order=list(tempD1.keys()), palette=tempD1, dodge=False, jitter=0.15,\n",
    "                      size=5, edgecolor='black', linewidth=1, **{'marker':'o', 'alpha':0.5})\n",
    "    ##Set axis\n",
    "    sns.despine()\n",
    "    p.set(ylim=(ymin-ymargin_b, ymax+ymargin_t), yticks=np.arange(ymin, ymax + yinter/10, yinter))\n",
    "    plt.setp(p.get_xticklabels(), rotation=70, horizontalalignment='right',\n",
    "             verticalalignment='center', rotation_mode='anchor')\n",
    "    ##Add significance labels\n",
    "    for row_i in range(len(tempDF2)):\n",
    "        #Baseline\n",
    "        group_0 = tempDF2['Baseline'].iloc[row_i]\n",
    "        index_0 = list(tempD1.keys()).index(group_0)\n",
    "        xcoord_0 = index_0\n",
    "        #Contrast\n",
    "        group_1 = tempDF2['Contrast'].iloc[row_i]\n",
    "        index_1 = list(tempD1.keys()).index(group_1)\n",
    "        xcoord_1 = index_1\n",
    "        #Standard point of marker\n",
    "        xcoord = (xcoord_0+xcoord_1)/2\n",
    "        ycoord = aline_ymin + aline_ymargin*row_i\n",
    "        label = tempDF2['SignifLabel'].iloc[row_i]\n",
    "        #Add annotation lines\n",
    "        aline_offset = yinter/5\n",
    "        aline_length = yinter/5 + aline_offset/2\n",
    "        plt.plot([xcoord_0, xcoord_0, xcoord_1, xcoord_1],\n",
    "                 [ycoord+aline_offset, ycoord+aline_length, ycoord+aline_length, ycoord+aline_offset],\n",
    "                 lw=1.5, c='k')\n",
    "        #Add annotation text\n",
    "        if label in ['***', '**', '*']:\n",
    "            text_offset = yinter/21\n",
    "            p.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                       horizontalalignment='center', verticalalignment='bottom',\n",
    "                       fontsize='medium', color='k')\n",
    "        else:\n",
    "            text_offset = yinter/3.5\n",
    "            p.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                       horizontalalignment='center', verticalalignment='bottom',\n",
    "                       fontsize='x-small', color='k')\n",
    "    ##Add annotation\n",
    "    p.set_title('Consensus:\\nown group', {'fontsize':'small'})\n",
    "    xoff = 0.015\n",
    "    yoff = 0.01\n",
    "    rect = plt.Rectangle((xoff, 1), 1-xoff, 0.155,#Manual adjustment\n",
    "                         transform=p.axes.transAxes, facecolor=plt.get_cmap('tab20')(15),\n",
    "                         clip_on=False, linewidth=0, zorder=0.5)\n",
    "    p.add_patch(rect)\n",
    "    ##Set axis label and title\n",
    "    plt.setp(p, xlabel='', ylabel='Sample RMS\\n(Mean = Module RCI)')\n",
    "    plt.suptitle(module, size='small',\n",
    "                 verticalalignment='bottom', horizontalalignment='center', wrap=True, y=1.0)\n",
    "    ##Save\n",
    "    #fileDir = './ExportFigures/'\n",
    "    #ipynbName = '230214_LC-M001-proteomics-DIRAC-ver7-2_DIRAC-GOBP_'\n",
    "    #fileName = 'RCI-pointplot-'+module.replace('GO:', 'GO')+'.pdf'\n",
    "    #plt.gcf().savefig(fileDir+ipynbName+fileName, dpi=300, bbox_inches='tight', pad_inches=0.04, transparent=True)\n",
    "    plt.show()\n",
    "    print('')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 5. Rank matching score under a fixed consensus: inter-group module comparison"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> Test specific hypotheses: control RMS mean == intervention RMS mean per a fixed rank consensus for each module (i.e., inter-group module comparison).  \n",
    ">\n",
    "> 3. performing Welch's t-tests per rank consensus for each module.  \n",
    ">  \n",
    "> Because the main interests are whether the changed patterns are similar or not in a changed module (i.e., this is a downstream analysis of RCI), ANOVA and the p-value adjustment across modules are unnecessary. Although tricky (i.e., the hypotheses across rank consensus are not independent), the p-values in (3) are conservatively adjusted across all comparisons per module (interventions x rank consensus group). In addition, the rank consensus group is eliminated from the tests because its mean of RMS is RCI (i.e., expected mean and variance is clearly different) and because the number of hypotheses can be reduced.  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 5-1. Welch's t-tests"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 5-1-1. Perform all statistical tests"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tempL1 = ['Aca', '17aE2', 'Rapa']#Target rank consensus groups\n",
    "tempL2 = ['Ctrl', 'Aca', '17aE2', 'Rapa']#Target sample groups to be assessed\n",
    "control = 'Ctrl'\n",
    "tempDF1 = rmsDF.loc[rmsDF['Template'].isin(tempL1)].set_index(['Template', 'ModuleID'])\n",
    "tempDF2 = sampleDF.loc[sampleDF['Phenotype'].isin(tempL2)]\n",
    "tempI = moduleDF.index\n",
    "\n",
    "#Statistical tests per rank consensus\n",
    "tempD = {}\n",
    "for template in tempL1:\n",
    "    #Statistical tests per module\n",
    "    t_start = time.time()\n",
    "    tempL3 = []#For summary table\n",
    "    for module in tempI.tolist():\n",
    "        #Select the target RMSs\n",
    "        tempS = tempDF1.loc[(template, module)]\n",
    "        tempS.name = 'RMS'\n",
    "        #Add metadata while selecting the target sample samples\n",
    "        tempDF = pd.merge(tempS, tempDF2, left_index=True, right_index=True, how='inner')\n",
    "        \n",
    "        #Tests per control vs. contrast, except for the template group\n",
    "        tempDF4 = pd.DataFrame(columns=['DoF', 'tStat', 'Pval'])\n",
    "        for contrast in tempL2:\n",
    "            if control!=contrast:\n",
    "                if contrast==template:\n",
    "                    tstat, pval, dof = [np.nan, np.nan, np.nan]\n",
    "                else:\n",
    "                    tempS1 = tempDF['RMS'].loc[tempDF['Phenotype']==control]\n",
    "                    tempS2 = tempDF['RMS'].loc[tempDF['Phenotype']==contrast]\n",
    "                    #Two-sided Welch's t-test\n",
    "                    tstat, pval, dof = weightstats.ttest_ind(tempS2, tempS1,#t-statistic reflects direction from the baseline\n",
    "                                                             alternative='two-sided', usevar='unequal')\n",
    "                tempDF4.loc[contrast+'-vs-'+control] = [dof, tstat, pval]\n",
    "        tempDF4['AdjPval'] = 1.0#Add dummy column for now\n",
    "        ##Convert to wide-format\n",
    "        tempL = []\n",
    "        for comparison in tempDF4.index.tolist():\n",
    "            tempS = tempDF4.loc[comparison]\n",
    "            tempS.index = comparison+'_'+tempS.index\n",
    "            tempS.name = module\n",
    "            tempL.append(tempS)\n",
    "        tempS = pd.concat(tempL, axis=0)\n",
    "        tempL3.append(tempS)\n",
    "    t_elapsed = time.time() - t_start\n",
    "    print(template)\n",
    "    print('Elapsed time for', (len(tempL2)-2)*len(tempI), 'tests (',\n",
    "          len(tempL2)-2, 'comparisons x', len(tempI), 'modules):',\n",
    "          round(t_elapsed//60), 'min', round(t_elapsed%60, 1), 'sec')\n",
    "    \n",
    "    #Generate summary table\n",
    "    tempDF4 = pd.concat(tempL3, axis=1).T\n",
    "    tempDF4.index.name = tempI.name\n",
    "    tempD[template] = tempDF4\n",
    "\n",
    "#Clean all summary tables\n",
    "tempDF4 = pd.DataFrame()\n",
    "for template in tempD.keys():\n",
    "    tempDF = tempD[template]\n",
    "    tempDF['Template'] = template\n",
    "    tempDF = tempDF.reset_index().set_index(['Template', 'ModuleID'])\n",
    "    tempDF4 = pd.concat([tempDF4, tempDF], axis=0)\n",
    "##P-value adjustment across all tests per module (comparisons x templates) by using Benjamini–Hochberg method\n",
    "for module in tempI.tolist():\n",
    "    tempL = tempDF4.loc[:, tempDF4.columns.str.contains('_Pval$', regex=True)].columns.tolist()\n",
    "    tempDF = tempDF4.reset_index().melt(var_name='Comparison', value_name='Pval', value_vars=tempL,\n",
    "                                        id_vars=['Template', 'ModuleID'])\n",
    "    tempDF = tempDF.loc[tempDF['ModuleID']==module]\n",
    "    tempDF = tempDF.dropna(axis=0)#NaN due to the eliminated test for the consensus group\n",
    "    tempDF['AdjPval'] = multi.multipletests(tempDF['Pval'], alpha=0.05, method='fdr_bh',\n",
    "                                            is_sorted=False, returnsorted=False)[1]\n",
    "    tempDF = tempDF.pivot(index=['Template', 'ModuleID'], columns='Comparison', values='AdjPval')\n",
    "    tempDF.columns = tempDF.columns.str.replace('_Pval', '_AdjPval')\n",
    "    #Replace the dummy values with the adjusted p-values\n",
    "    tempL = tempDF4.loc[:, tempDF4.columns.str.contains('_AdjPval$', regex=True)].columns.tolist()\n",
    "    for col_n in tempL:\n",
    "        for template in tempL1:\n",
    "            tempDF4.loc[(template, module), col_n] = tempDF.loc[(template, module), col_n]\n",
    "display(tempDF4)\n",
    "\n",
    "statDF2 = tempDF4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tempL1 = ['Aca', '17aE2', 'Rapa']#Target rank consensus groups\n",
    "tempL2 = ['Ctrl', 'Aca', '17aE2', 'Rapa']#Target sample groups to be summarized\n",
    "tempDF1 = rmsDF.loc[rmsDF['Template'].isin(tempL1)].set_index(['Template', 'ModuleID'])\n",
    "tempDF2 = sampleDF.loc[sampleDF['Phenotype'].isin(tempL2)]\n",
    "tempDF3 = statDF#Inter-group RCI comparisons, for sorting purpose\n",
    "\n",
    "#Calculate general statistics per rank consensus\n",
    "tempD = {}\n",
    "for template in tempL1:\n",
    "    #Calculate general statistics per intervention group\n",
    "    tempL3 = []\n",
    "    for phenotype in tempL2:\n",
    "        #Select the target RMSs\n",
    "        tempDF = tempDF1.loc[template]\n",
    "        #Select the target samples\n",
    "        tempL = tempDF2.loc[tempDF2['Phenotype']==phenotype].index.tolist()\n",
    "        tempDF = tempDF[tempL]\n",
    "        #Calculate general statistics\n",
    "        tempS1 = len(tempL) - tempDF.isnull().sum(axis=1)\n",
    "        tempS1.name = phenotype+'_N'\n",
    "        tempS2 = tempDF.mean(axis=1)\n",
    "        tempS2.name = phenotype+'_RMSmean'\n",
    "        tempS3 = tempDF.sem(axis=1, ddof=1)\n",
    "        tempS3.name = phenotype+'_RMSsem'\n",
    "        #Merge\n",
    "        tempDF = pd.concat([tempS1, tempS2, tempS3], axis=1)\n",
    "        tempL3.append(tempDF)\n",
    "    tempDF = pd.concat(tempL3, axis=1)\n",
    "    tempD[template] = tempDF\n",
    "##Clean all general statistics tables\n",
    "tempDF4 = pd.DataFrame()\n",
    "for template in tempD.keys():\n",
    "    tempDF = tempD[template]\n",
    "    tempDF['Template'] = template\n",
    "    tempDF = tempDF.reset_index().set_index(['Template', 'ModuleID'])\n",
    "    tempDF4 = pd.concat([tempDF4, tempDF], axis=0)\n",
    "display(tempDF4)\n",
    "\n",
    "#Merge all the tables\n",
    "print('General statistics table:', tempDF4.shape)\n",
    "print('Test table:', statDF2.shape)\n",
    "tempDF = pd.merge(moduleDF['ModuleName'], tempDF4.reset_index(), on='ModuleID', how='right')\n",
    "tempDF = pd.concat([tempDF.set_index(['Template', 'ModuleID']), statDF2], axis=1)\n",
    "\n",
    "#Sort based on P-value in ANOVA of the inter-group RCI comparisons\n",
    "tempL = [(template, module) for template in tempL1 for module in tempDF3.index.tolist()]\n",
    "tempDF = tempDF.loc[tempL]\n",
    "display(tempDF)\n",
    "\n",
    "#Save per rank consensus\n",
    "fileDir = './ExportData/'\n",
    "ipynbName = '230214_LC-M001-proteomics-DIRAC-ver7-2_DIRAC-GOBP_'\n",
    "for template in tempL1:\n",
    "    fileName = 'inter-group-comparison_'+template+'-fixed-RMSmean.tsv'\n",
    "    tempDF4 = tempDF.loc[template]\n",
    "    tempDF4 = tempDF4.dropna(axis=1)#NaN due to the eliminated test for the consensus group\n",
    "    tempDF4.to_csv(fileDir+ipynbName+fileName, sep='\\t', index=True)\n",
    "    print('Saved .tsv table for '+template)\n",
    "    display(tempDF4)\n",
    "\n",
    "statDF_fixed = tempDF"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 5-1-2. Changed modules"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> Focus only the changed modules based on RCI.  \n",
    "> –> Utilize and update the above pvalDF and diffDF (4-2-3 section).  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Spread statDF by template to flatten multi-index\n",
    "tempDF = moduleDF['ModuleName']#pd.Series() for now\n",
    "for template in ['Aca', '17aE2', 'Rapa']:\n",
    "    tempDF1 = statDF_fixed.loc[template].drop(columns=['ModuleName'])\n",
    "    tempDF1 = tempDF1.dropna(axis=1)#NaN due to the eliminated test for the consensus group\n",
    "    tempDF1.columns = template+'-fixed_'+tempDF1.columns\n",
    "    tempDF = pd.merge(tempDF, tempDF1, left_index=True, right_index=True, how='left')\n",
    "\n",
    "#Extract only the changed modules\n",
    "tempDF = tempDF.loc[pvalDF.index.tolist()]\n",
    "print('Changed modules (adjusted P < 0.05):', len(tempDF))\n",
    "\n",
    "#Take adjusted P-value and add to the previous DF\n",
    "tempDF1 = tempDF.loc[:, tempDF.columns.str.contains('-vs-.*_AdjPval$')]\n",
    "tempDF1.columns = tempDF1.columns.str.replace('_AdjPval$', '')\n",
    "tempDF1 = pd.merge(pvalDF, tempDF1, left_index=True, right_index=True, how='left')\n",
    "print('Adjusted P-value:')\n",
    "display(tempDF1)\n",
    "display(tempDF1.describe())\n",
    "\n",
    "#Take t-statistic (for direction) and add to the previous DF\n",
    "tempDF2 = tempDF.loc[:, tempDF.columns.str.contains('-vs-.*_tStat$')]\n",
    "tempDF2.columns = tempDF2.columns.str.replace('_tStat$', '')\n",
    "tempDF2 = pd.merge(diffDF, tempDF2, left_index=True, right_index=True, how='left')\n",
    "print('Changed direction (t-statistic):')\n",
    "display(tempDF2)\n",
    "display(tempDF2.describe())\n",
    "\n",
    "pvalDF = tempDF1#Update\n",
    "diffDF = tempDF2#Update"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 5-2. Visualization: venn diagram"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 5-2-1. RMS under Aca consensus"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "template = 'Aca'\n",
    "\n",
    "#Prepare label and color\n",
    "tempD0 = {'Ctrl':'Control', 'Aca':'Acarbose',\n",
    "          '17aE2':'17'+r'$\\alpha$'+'-Estradiol', 'Rapa':'Rapamycin'}\n",
    "tempD1 = {'Dissimilarly changed\\nby 17'+r'$\\alpha$'+'-Estradiol':plt.get_cmap('tab20')(5),\n",
    "          'Similarly changed\\nby 17'+r'$\\alpha$'+'-Estradiol':'tab:green',\n",
    "          'Similarly changed\\nby Rapamycin':'tab:purple',\n",
    "          'Dissimilarly changed\\nby Rapamycin':plt.get_cmap('tab20')(9)}\n",
    "target = template+'-vs-Ctrl'\n",
    "\n",
    "#Visualization per direction\n",
    "for regulation in ['Tightened', 'Loosened']:\n",
    "    #Prepare the changed module set\n",
    "    tempS1 = pvalDF[target]\n",
    "    tempS2 = diffDF[target]\n",
    "    if regulation=='Changed':\n",
    "        tempS2 = tempS2.loc[(tempS1<0.05)]\n",
    "    elif regulation=='Tightened':\n",
    "        tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2>0)]\n",
    "    elif regulation=='Loosened':\n",
    "        tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2<0)]\n",
    "    tempL = tempS2.index.tolist()\n",
    "    print(regulation+' modules (vs. Control):', len(tempL))\n",
    "    \n",
    "    #Skip the followings if no changed module\n",
    "    if len(tempL)==0:\n",
    "        continue\n",
    "    \n",
    "    #Prepare similar/dissimilar module set\n",
    "    tempD2 = {}\n",
    "    tempL1 = pvalDF.loc[:, pvalDF.columns.str.contains(template+'-fixed')].columns.tolist()\n",
    "    for col_n in tempL1:\n",
    "        tempS1 = pvalDF.loc[tempL, col_n]\n",
    "        tempS2 = diffDF.loc[tempL, col_n]\n",
    "        for similarity in ['Similar', 'Dissimilar']:\n",
    "            if similarity=='Similar':\n",
    "                tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2>0)]\n",
    "            elif similarity=='Dissimilar':\n",
    "                tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2<0)]\n",
    "            comparison = re.sub('^.*-fixed_', '', col_n)\n",
    "            label = similarity+'ly changed\\nby '+tempD0[re.sub('-vs-.*$', '', comparison)]\n",
    "            tempD2[label] = set(tempS2.index.tolist())\n",
    "    ##Sort to make consistent order in manual legend generation\n",
    "    tempD = {}\n",
    "    for label in tempD1.keys():\n",
    "        tempD[label] = tempD2[label]\n",
    "    \n",
    "    #Not significant in all comparisons\n",
    "    tempDF = pvalDF.loc[tempL]\n",
    "    for moduleS in tempD.values():\n",
    "        tempDF = tempDF.loc[~tempDF.index.isin(moduleS)]\n",
    "    count = len(tempDF)\n",
    "    print(' -> Not similarly/dissimilarly changed in the other interventions:', count)\n",
    "    \n",
    "    #Skip the followings if no similarly/dissimilarly changed module\n",
    "    if count==len(tempL):\n",
    "        continue\n",
    "    \n",
    "    #Venn diagram\n",
    "    sns.set(style='ticks', font='Arial', context='talk')\n",
    "    fig, ax = plt.subplots(figsize=(3, 3))\n",
    "    venn(tempD, fmt='{size:,}', cmap=list(tempD1.values()), legend_loc=None, ax=ax)\n",
    "    plt.setp(ax, ylim=(0.1, 1.05))#Otherwise, weird space...\n",
    "    #Replace text for the n/a regions\n",
    "    for text_i, text in enumerate(ax.texts):\n",
    "        if text_i in [2, 6, 10, 11, 12, 13, 14]:\n",
    "            text.set_text('n/a')\n",
    "    #Add legend annotation\n",
    "    x_coord = [0.1, 0.1, 0.9, 0.9]\n",
    "    y_coord = [0.25, 0.7, 0.7, 0.25]\n",
    "    h_align = ['right', 'right', 'left', 'left']\n",
    "    v_align = ['top', 'bottom', 'bottom', 'top']\n",
    "    for i in range(len(tempD1)):\n",
    "        key = list(tempD1.keys())[i]\n",
    "        total = f'{len(tempD[key]):,}'\n",
    "        ax.text(x_coord[i], y_coord[i], key+'\\n('+total+' modules)',\n",
    "                fontsize='small', multialignment='center',\n",
    "                horizontalalignment=h_align[i], verticalalignment=v_align[i],\n",
    "                bbox={'boxstyle':'round', 'facecolor':tempD1[key], 'pad':0.2, 'alpha':0.5})\n",
    "    title = 'Similarity to '+tempD0[template]+' consensus'\n",
    "    ax.set_title(title, fontsize='medium')\n",
    "    ##Save\n",
    "    if regulation!='Changed':\n",
    "        fileDir = './ExportFigures/'\n",
    "        ipynbName = '230214_LC-M001-proteomics-DIRAC-ver7-2_DIRAC-GOBP_'\n",
    "        fileName = template+'-fixed-RMSmean-venn-'+regulation.lower()+'.pdf'\n",
    "        plt.gcf().savefig(fileDir+ipynbName+fileName, dpi=300, bbox_inches='tight', pad_inches=0.04, transparent=True)\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "target = template+'-vs-Ctrl'\n",
    "\n",
    "#Export module list in each subset in the venn diagram\n",
    "for regulation in ['Tightened', 'Loosened']:\n",
    "    #Prepare the changed module set\n",
    "    contrast = template.replace('-fixed', '-vs-Cont')\n",
    "    tempS1 = pvalDF[target]\n",
    "    tempS2 = diffDF[target]\n",
    "    if regulation=='Changed':\n",
    "        tempS2 = tempS2.loc[(tempS1<0.05)]\n",
    "    elif regulation=='Tightened':\n",
    "        tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2>0)]\n",
    "    elif regulation=='Loosened':\n",
    "        tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2<0)]\n",
    "    tempL = tempS2.index.tolist()\n",
    "    print(regulation+' modules (vs. Control):', len(tempL))\n",
    "    \n",
    "    #Skip the followings if no changed module\n",
    "    if len(tempL)==0:\n",
    "        continue\n",
    "    \n",
    "    #Prepare similar/dissimilar module set\n",
    "    tempD = {}\n",
    "    tempL1 = pvalDF.loc[:, pvalDF.columns.str.contains(template+'-fixed')].columns.tolist()\n",
    "    for col_n in tempL1:\n",
    "        tempS1 = pvalDF.loc[tempL, col_n]\n",
    "        tempS2 = diffDF.loc[tempL, col_n]\n",
    "        for similarity in ['Similar', 'Dissimilar']:\n",
    "            if similarity=='Similar':\n",
    "                tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2>0)]\n",
    "            elif similarity=='Dissimilar':\n",
    "                tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2<0)]\n",
    "            label = col_n+'_'+similarity+'lyChanged'\n",
    "            tempD[label] = set(tempS2.index.tolist())\n",
    "    \n",
    "    #Not significant in all contrasts\n",
    "    tempDF = pvalDF.loc[tempL]\n",
    "    for moduleS in tempD.values():\n",
    "        tempDF = tempDF.loc[~tempDF.index.isin(moduleS)]\n",
    "    count = len(tempDF)\n",
    "    print(' -> Not similarly/dissimilarly changed in the other interventions:', count)\n",
    "    \n",
    "    #Skip the followings if no similarly/dissimilarly changed module\n",
    "    if count==len(tempL):\n",
    "        continue\n",
    "    \n",
    "    #Prepare a new .xlsx file (dummy README)\n",
    "    tempL1 = [len(tempD[key]) for key in tempD.keys()]\n",
    "    tempDF = pd.DataFrame({'Group':tempD.keys(), 'nModules':tempL1})\n",
    "    tempDF = tempDF.reset_index().rename(columns={'index':'VennOrder'})\n",
    "    tempDF['VennOrder'] = tempDF['VennOrder'] + 1\n",
    "    fileDir = './ExportData/'\n",
    "    ipynbName = '230214_LC-M001-proteomics-DIRAC-ver7-2_DIRAC-GOBP_'\n",
    "    fileName = template+'-fixed-RMSmean-venn-'+regulation.lower()+'.xlsx'\n",
    "    tempDF.to_excel(fileDir+ipynbName+fileName, sheet_name='README', header=True, index=False)\n",
    "    display(tempDF)#Check\n",
    "    \n",
    "    #Prepare summary table for the template\n",
    "    tempDF1 = statDF_fixed.loc[template]\n",
    "    tempDF1 = tempDF1.dropna(axis=1)#NaN due to the eliminated test for the consensus group\n",
    "    \n",
    "    t_start = time.time()\n",
    "    #Extract overall set\n",
    "    for key_i in range(len(tempD)):\n",
    "        key = list(tempD.keys())[key_i]\n",
    "        tempS = tempD[key]\n",
    "        tempDF = tempDF1.loc[tempDF1.index.isin(tempS)]\n",
    "        #Save summary table by appended to the above .xlsx file\n",
    "        ##Prepare sheet name\n",
    "        tempL1 = ['NA' for i in range(len(tempD))]\n",
    "        tempL1[key_i] = '1'\n",
    "        setName = '('+','.join(tempL1)+')'\n",
    "        with pd.ExcelWriter(fileDir+ipynbName+fileName, mode='a', engine='openpyxl') as writer:\n",
    "            tempDF.to_excel(writer, sheet_name=setName, header=True, index=True)\n",
    "        print(' - '+setName+':', len(tempDF))\n",
    "    \n",
    "    #Extract subset\n",
    "    tempL1 = ['1', '0']\n",
    "    tempL2 = [[k1, k2, k3, k4] for k1 in tempL1 for k2 in tempL1 for k3 in tempL1 for k4 in tempL1]\n",
    "    #tempL2.remove(['0', '0', '0', '0'])\n",
    "    for tempL1 in tempL2:\n",
    "        #Positive module set\n",
    "        tempL3 = [list(tempD.values())[key_i] for key_i, binary in enumerate(tempL1) if binary=='1']\n",
    "        tempS1 = set(pvalDF.index.tolist())#Initialize\n",
    "        for tempS in tempL3:\n",
    "            tempS1 = tempS1 & tempS\n",
    "        #Negative module set\n",
    "        tempL3 = [list(tempD.values())[key_i] for key_i, binary in enumerate(tempL1) if binary=='0']\n",
    "        tempS2 = set()#Initialize\n",
    "        for tempS in tempL3:\n",
    "            tempS2 = tempS2 | tempS\n",
    "        #Extract subset\n",
    "        tempS = tempS1 - tempS2\n",
    "        tempDF = tempDF1.loc[tempDF1.index.isin(tempS)]\n",
    "        #Save summary table by appended to the above .xlsx file\n",
    "        ##Prepare sheet name\n",
    "        setName = '('+','.join(tempL1)+')'\n",
    "        with pd.ExcelWriter(fileDir+ipynbName+fileName, mode='a', engine='openpyxl') as writer:\n",
    "            tempDF.to_excel(writer, sheet_name=setName, header=True, index=True)\n",
    "        print(' - '+setName+':', len(tempDF))\n",
    "    \n",
    "    t_elapsed = time.time() - t_start\n",
    "    print(' - Elapsed time:', round(t_elapsed//60), 'min', round(t_elapsed%60, 1), 'sec')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 5-2-2. RMS under 17aE2 consensus"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "template = '17aE2'\n",
    "\n",
    "#Prepare label and color\n",
    "tempD0 = {'Ctrl':'Control', 'Aca':'Acarbose',\n",
    "          '17aE2':'17'+r'$\\alpha$'+'-Estradiol', 'Rapa':'Rapamycin'}\n",
    "tempD1 = {'Dissimilarly changed\\nby Rapamycin':plt.get_cmap('tab20')(9),\n",
    "          'Similarly changed\\nby Rapamycin':'tab:purple',\n",
    "          'Similarly changed\\nby Acarbose':'tab:red',\n",
    "          'Dissimilarly changed\\nby Acarbose':plt.get_cmap('tab20')(7)}\n",
    "target = template+'-vs-Ctrl'\n",
    "\n",
    "#Visualization per direction\n",
    "for regulation in ['Tightened', 'Loosened']:\n",
    "    #Prepare the changed module set\n",
    "    tempS1 = pvalDF[target]\n",
    "    tempS2 = diffDF[target]\n",
    "    if regulation=='Changed':\n",
    "        tempS2 = tempS2.loc[(tempS1<0.05)]\n",
    "    elif regulation=='Tightened':\n",
    "        tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2>0)]\n",
    "    elif regulation=='Loosened':\n",
    "        tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2<0)]\n",
    "    tempL = tempS2.index.tolist()\n",
    "    print(regulation+' modules (vs. Control):', len(tempL))\n",
    "    \n",
    "    #Skip the followings if no changed module\n",
    "    if len(tempL)==0:\n",
    "        continue\n",
    "    \n",
    "    #Prepare similar/dissimilar module set\n",
    "    tempD2 = {}\n",
    "    tempL1 = pvalDF.loc[:, pvalDF.columns.str.contains(template+'-fixed')].columns.tolist()\n",
    "    for col_n in tempL1:\n",
    "        tempS1 = pvalDF.loc[tempL, col_n]\n",
    "        tempS2 = diffDF.loc[tempL, col_n]\n",
    "        for similarity in ['Similar', 'Dissimilar']:\n",
    "            if similarity=='Similar':\n",
    "                tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2>0)]\n",
    "            elif similarity=='Dissimilar':\n",
    "                tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2<0)]\n",
    "            comparison = re.sub('^.*-fixed_', '', col_n)\n",
    "            label = similarity+'ly changed\\nby '+tempD0[re.sub('-vs-.*$', '', comparison)]\n",
    "            tempD2[label] = set(tempS2.index.tolist())\n",
    "    ##Sort to make consistent order in manual legend generation\n",
    "    tempD = {}\n",
    "    for label in tempD1.keys():\n",
    "        tempD[label] = tempD2[label]\n",
    "    \n",
    "    #Not significant in all comparisons\n",
    "    tempDF = pvalDF.loc[tempL]\n",
    "    for moduleS in tempD.values():\n",
    "        tempDF = tempDF.loc[~tempDF.index.isin(moduleS)]\n",
    "    count = len(tempDF)\n",
    "    print(' -> Not similarly/dissimilarly changed in the other interventions:', count)\n",
    "    \n",
    "    #Skip the followings if no similarly/dissimilarly changed module\n",
    "    if count==len(tempL):\n",
    "        continue\n",
    "    \n",
    "    #Venn diagram\n",
    "    sns.set(style='ticks', font='Arial', context='talk')\n",
    "    fig, ax = plt.subplots(figsize=(3, 3))\n",
    "    venn(tempD, fmt='{size:,}', cmap=list(tempD1.values()), legend_loc=None, ax=ax)\n",
    "    plt.setp(ax, ylim=(0.1, 1.05))#Otherwise, weird space...\n",
    "    #Replace text for the n/a regions\n",
    "    for text_i, text in enumerate(ax.texts):\n",
    "        if text_i in [2, 6, 10, 11, 12, 13, 14]:\n",
    "            text.set_text('n/a')\n",
    "    #Add legend annotation\n",
    "    x_coord = [0.1, 0.1, 0.9, 0.9]\n",
    "    y_coord = [0.25, 0.7, 0.7, 0.25]\n",
    "    h_align = ['right', 'right', 'left', 'left']\n",
    "    v_align = ['top', 'bottom', 'bottom', 'top']\n",
    "    for i in range(len(tempD1)):\n",
    "        key = list(tempD1.keys())[i]\n",
    "        total = f'{len(tempD[key]):,}'\n",
    "        ax.text(x_coord[i], y_coord[i], key+'\\n('+total+' modules)',\n",
    "                fontsize='small', multialignment='center',\n",
    "                horizontalalignment=h_align[i], verticalalignment=v_align[i],\n",
    "                bbox={'boxstyle':'round', 'facecolor':tempD1[key], 'pad':0.2, 'alpha':0.5})\n",
    "    title = 'Similarity to '+tempD0[template]+' consensus'\n",
    "    ax.set_title(title, fontsize='medium')\n",
    "    ##Save\n",
    "    if regulation!='Changed':\n",
    "        fileDir = './ExportFigures/'\n",
    "        ipynbName = '230214_LC-M001-proteomics-DIRAC-ver7-2_DIRAC-GOBP_'\n",
    "        fileName = template+'-fixed-RMSmean-venn-'+regulation.lower()+'.pdf'\n",
    "        plt.gcf().savefig(fileDir+ipynbName+fileName, dpi=300, bbox_inches='tight', pad_inches=0.04, transparent=True)\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "target = template+'-vs-Ctrl'\n",
    "\n",
    "#Export module list in each subset in the venn diagram\n",
    "for regulation in ['Tightened', 'Loosened']:\n",
    "    #Prepare the changed module set\n",
    "    contrast = template.replace('-fixed', '-vs-Cont')\n",
    "    tempS1 = pvalDF[target]\n",
    "    tempS2 = diffDF[target]\n",
    "    if regulation=='Changed':\n",
    "        tempS2 = tempS2.loc[(tempS1<0.05)]\n",
    "    elif regulation=='Tightened':\n",
    "        tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2>0)]\n",
    "    elif regulation=='Loosened':\n",
    "        tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2<0)]\n",
    "    tempL = tempS2.index.tolist()\n",
    "    print(regulation+' modules (vs. Control):', len(tempL))\n",
    "    \n",
    "    #Skip the followings if no changed module\n",
    "    if len(tempL)==0:\n",
    "        continue\n",
    "    \n",
    "    #Prepare similar/dissimilar module set\n",
    "    tempD = {}\n",
    "    tempL1 = pvalDF.loc[:, pvalDF.columns.str.contains(template+'-fixed')].columns.tolist()\n",
    "    for col_n in tempL1:\n",
    "        tempS1 = pvalDF.loc[tempL, col_n]\n",
    "        tempS2 = diffDF.loc[tempL, col_n]\n",
    "        for similarity in ['Similar', 'Dissimilar']:\n",
    "            if similarity=='Similar':\n",
    "                tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2>0)]\n",
    "            elif similarity=='Dissimilar':\n",
    "                tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2<0)]\n",
    "            label = col_n+'_'+similarity+'lyChanged'\n",
    "            tempD[label] = set(tempS2.index.tolist())\n",
    "    \n",
    "    #Not significant in all contrasts\n",
    "    tempDF = pvalDF.loc[tempL]\n",
    "    for moduleS in tempD.values():\n",
    "        tempDF = tempDF.loc[~tempDF.index.isin(moduleS)]\n",
    "    count = len(tempDF)\n",
    "    print(' -> Not similarly/dissimilarly changed in the other interventions:', count)\n",
    "    \n",
    "    #Skip the followings if no similarly/dissimilarly changed module\n",
    "    if count==len(tempL):\n",
    "        continue\n",
    "    \n",
    "    #Prepare a new .xlsx file (dummy README)\n",
    "    tempL1 = [len(tempD[key]) for key in tempD.keys()]\n",
    "    tempDF = pd.DataFrame({'Group':tempD.keys(), 'nModules':tempL1})\n",
    "    tempDF = tempDF.reset_index().rename(columns={'index':'VennOrder'})\n",
    "    tempDF['VennOrder'] = tempDF['VennOrder'] + 1\n",
    "    fileDir = './ExportData/'\n",
    "    ipynbName = '230214_LC-M001-proteomics-DIRAC-ver7-2_DIRAC-GOBP_'\n",
    "    fileName = template+'-fixed-RMSmean-venn-'+regulation.lower()+'.xlsx'\n",
    "    tempDF.to_excel(fileDir+ipynbName+fileName, sheet_name='README', header=True, index=False)\n",
    "    display(tempDF)#Check\n",
    "    \n",
    "    #Prepare summary table for the template\n",
    "    tempDF1 = statDF_fixed.loc[template]\n",
    "    tempDF1 = tempDF1.dropna(axis=1)#NaN due to the eliminated test for the consensus group\n",
    "    \n",
    "    t_start = time.time()\n",
    "    #Extract overall set\n",
    "    for key_i in range(len(tempD)):\n",
    "        key = list(tempD.keys())[key_i]\n",
    "        tempS = tempD[key]\n",
    "        tempDF = tempDF1.loc[tempDF1.index.isin(tempS)]\n",
    "        #Save summary table by appended to the above .xlsx file\n",
    "        ##Prepare sheet name\n",
    "        tempL1 = ['NA' for i in range(len(tempD))]\n",
    "        tempL1[key_i] = '1'\n",
    "        setName = '('+','.join(tempL1)+')'\n",
    "        with pd.ExcelWriter(fileDir+ipynbName+fileName, mode='a', engine='openpyxl') as writer:\n",
    "            tempDF.to_excel(writer, sheet_name=setName, header=True, index=True)\n",
    "        print(' - '+setName+':', len(tempDF))\n",
    "    \n",
    "    #Extract subset\n",
    "    tempL1 = ['1', '0']\n",
    "    tempL2 = [[k1, k2, k3, k4] for k1 in tempL1 for k2 in tempL1 for k3 in tempL1 for k4 in tempL1]\n",
    "    #tempL2.remove(['0', '0', '0', '0'])\n",
    "    for tempL1 in tempL2:\n",
    "        #Positive module set\n",
    "        tempL3 = [list(tempD.values())[key_i] for key_i, binary in enumerate(tempL1) if binary=='1']\n",
    "        tempS1 = set(pvalDF.index.tolist())#Initialize\n",
    "        for tempS in tempL3:\n",
    "            tempS1 = tempS1 & tempS\n",
    "        #Negative module set\n",
    "        tempL3 = [list(tempD.values())[key_i] for key_i, binary in enumerate(tempL1) if binary=='0']\n",
    "        tempS2 = set()#Initialize\n",
    "        for tempS in tempL3:\n",
    "            tempS2 = tempS2 | tempS\n",
    "        #Extract subset\n",
    "        tempS = tempS1 - tempS2\n",
    "        tempDF = tempDF1.loc[tempDF1.index.isin(tempS)]\n",
    "        #Save summary table by appended to the above .xlsx file\n",
    "        ##Prepare sheet name\n",
    "        setName = '('+','.join(tempL1)+')'\n",
    "        with pd.ExcelWriter(fileDir+ipynbName+fileName, mode='a', engine='openpyxl') as writer:\n",
    "            tempDF.to_excel(writer, sheet_name=setName, header=True, index=True)\n",
    "        print(' - '+setName+':', len(tempDF))\n",
    "    \n",
    "    t_elapsed = time.time() - t_start\n",
    "    print(' - Elapsed time:', round(t_elapsed//60), 'min', round(t_elapsed%60, 1), 'sec')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 5-2-3. RMS under Rapa consensus"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "template = 'Rapa'\n",
    "\n",
    "#Prepare label and color\n",
    "tempD0 = {'Ctrl':'Control', 'Aca':'Acarbose',\n",
    "          '17aE2':'17'+r'$\\alpha$'+'-Estradiol', 'Rapa':'Rapamycin'}\n",
    "tempD1 = {'Dissimilarly changed\\nby Acarbose':plt.get_cmap('tab20')(7),\n",
    "          'Similarly changed\\nby Acarbose':'tab:red',\n",
    "          'Similarly changed\\nby 17'+r'$\\alpha$'+'-Estradiol':'tab:green',\n",
    "          'Dissimilarly changed\\nby 17'+r'$\\alpha$'+'-Estradiol':plt.get_cmap('tab20')(5)}\n",
    "target = template+'-vs-Ctrl'\n",
    "\n",
    "#Visualization per direction\n",
    "for regulation in ['Tightened', 'Loosened']:\n",
    "    #Prepare the changed module set\n",
    "    tempS1 = pvalDF[target]\n",
    "    tempS2 = diffDF[target]\n",
    "    if regulation=='Changed':\n",
    "        tempS2 = tempS2.loc[(tempS1<0.05)]\n",
    "    elif regulation=='Tightened':\n",
    "        tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2>0)]\n",
    "    elif regulation=='Loosened':\n",
    "        tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2<0)]\n",
    "    tempL = tempS2.index.tolist()\n",
    "    print(regulation+' modules (vs. Control):', len(tempL))\n",
    "    \n",
    "    #Skip the followings if no changed module\n",
    "    if len(tempL)==0:\n",
    "        continue\n",
    "    \n",
    "    #Prepare similar/dissimilar module set\n",
    "    tempD2 = {}\n",
    "    tempL1 = pvalDF.loc[:, pvalDF.columns.str.contains(template+'-fixed')].columns.tolist()\n",
    "    for col_n in tempL1:\n",
    "        tempS1 = pvalDF.loc[tempL, col_n]\n",
    "        tempS2 = diffDF.loc[tempL, col_n]\n",
    "        for similarity in ['Similar', 'Dissimilar']:\n",
    "            if similarity=='Similar':\n",
    "                tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2>0)]\n",
    "            elif similarity=='Dissimilar':\n",
    "                tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2<0)]\n",
    "            comparison = re.sub('^.*-fixed_', '', col_n)\n",
    "            label = similarity+'ly changed\\nby '+tempD0[re.sub('-vs-.*$', '', comparison)]\n",
    "            tempD2[label] = set(tempS2.index.tolist())\n",
    "    ##Sort to make consistent order in manual legend generation\n",
    "    tempD = {}\n",
    "    for label in tempD1.keys():\n",
    "        tempD[label] = tempD2[label]\n",
    "    \n",
    "    #Not significant in all comparisons\n",
    "    tempDF = pvalDF.loc[tempL]\n",
    "    for moduleS in tempD.values():\n",
    "        tempDF = tempDF.loc[~tempDF.index.isin(moduleS)]\n",
    "    count = len(tempDF)\n",
    "    print(' -> Not similarly/dissimilarly changed in the other interventions:', count)\n",
    "    \n",
    "    #Skip the followings if no similarly/dissimilarly changed module\n",
    "    if count==len(tempL):\n",
    "        continue\n",
    "    \n",
    "    #Venn diagram\n",
    "    sns.set(style='ticks', font='Arial', context='talk')\n",
    "    fig, ax = plt.subplots(figsize=(3, 3))\n",
    "    venn(tempD, fmt='{size:,}', cmap=list(tempD1.values()), legend_loc=None, ax=ax)\n",
    "    plt.setp(ax, ylim=(0.1, 1.05))#Otherwise, weird space...\n",
    "    #Replace text for the n/a regions\n",
    "    for text_i, text in enumerate(ax.texts):\n",
    "        if text_i in [2, 6, 10, 11, 12, 13, 14]:\n",
    "            text.set_text('n/a')\n",
    "    #Add legend annotation\n",
    "    x_coord = [0.1, 0.1, 0.9, 0.9]\n",
    "    y_coord = [0.25, 0.7, 0.7, 0.25]\n",
    "    h_align = ['right', 'right', 'left', 'left']\n",
    "    v_align = ['top', 'bottom', 'bottom', 'top']\n",
    "    for i in range(len(tempD1)):\n",
    "        key = list(tempD1.keys())[i]\n",
    "        total = f'{len(tempD[key]):,}'\n",
    "        ax.text(x_coord[i], y_coord[i], key+'\\n('+total+' modules)',\n",
    "                fontsize='small', multialignment='center',\n",
    "                horizontalalignment=h_align[i], verticalalignment=v_align[i],\n",
    "                bbox={'boxstyle':'round', 'facecolor':tempD1[key], 'pad':0.2, 'alpha':0.5})\n",
    "    title = 'Similarity to '+tempD0[template]+' consensus'\n",
    "    ax.set_title(title, fontsize='medium')\n",
    "    ##Save\n",
    "    if regulation!='Changed':\n",
    "        fileDir = './ExportFigures/'\n",
    "        ipynbName = '230214_LC-M001-proteomics-DIRAC-ver7-2_DIRAC-GOBP_'\n",
    "        fileName = template+'-fixed-RMSmean-venn-'+regulation.lower()+'.pdf'\n",
    "        plt.gcf().savefig(fileDir+ipynbName+fileName, dpi=300, bbox_inches='tight', pad_inches=0.04, transparent=True)\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "target = template+'-vs-Ctrl'\n",
    "\n",
    "#Export module list in each subset in the venn diagram\n",
    "for regulation in ['Tightened', 'Loosened']:\n",
    "    #Prepare the changed module set\n",
    "    contrast = template.replace('-fixed', '-vs-Cont')\n",
    "    tempS1 = pvalDF[target]\n",
    "    tempS2 = diffDF[target]\n",
    "    if regulation=='Changed':\n",
    "        tempS2 = tempS2.loc[(tempS1<0.05)]\n",
    "    elif regulation=='Tightened':\n",
    "        tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2>0)]\n",
    "    elif regulation=='Loosened':\n",
    "        tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2<0)]\n",
    "    tempL = tempS2.index.tolist()\n",
    "    print(regulation+' modules (vs. Control):', len(tempL))\n",
    "    \n",
    "    #Skip the followings if no changed module\n",
    "    if len(tempL)==0:\n",
    "        continue\n",
    "    \n",
    "    #Prepare similar/dissimilar module set\n",
    "    tempD = {}\n",
    "    tempL1 = pvalDF.loc[:, pvalDF.columns.str.contains(template+'-fixed')].columns.tolist()\n",
    "    for col_n in tempL1:\n",
    "        tempS1 = pvalDF.loc[tempL, col_n]\n",
    "        tempS2 = diffDF.loc[tempL, col_n]\n",
    "        for similarity in ['Similar', 'Dissimilar']:\n",
    "            if similarity=='Similar':\n",
    "                tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2>0)]\n",
    "            elif similarity=='Dissimilar':\n",
    "                tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2<0)]\n",
    "            label = col_n+'_'+similarity+'lyChanged'\n",
    "            tempD[label] = set(tempS2.index.tolist())\n",
    "    \n",
    "    #Not significant in all contrasts\n",
    "    tempDF = pvalDF.loc[tempL]\n",
    "    for moduleS in tempD.values():\n",
    "        tempDF = tempDF.loc[~tempDF.index.isin(moduleS)]\n",
    "    count = len(tempDF)\n",
    "    print(' -> Not similarly/dissimilarly changed in the other interventions:', count)\n",
    "    \n",
    "    #Skip the followings if no similarly/dissimilarly changed module\n",
    "    if count==len(tempL):\n",
    "        continue\n",
    "    \n",
    "    #Prepare a new .xlsx file (dummy README)\n",
    "    tempL1 = [len(tempD[key]) for key in tempD.keys()]\n",
    "    tempDF = pd.DataFrame({'Group':tempD.keys(), 'nModules':tempL1})\n",
    "    tempDF = tempDF.reset_index().rename(columns={'index':'VennOrder'})\n",
    "    tempDF['VennOrder'] = tempDF['VennOrder'] + 1\n",
    "    fileDir = './ExportData/'\n",
    "    ipynbName = '230214_LC-M001-proteomics-DIRAC-ver7-2_DIRAC-GOBP_'\n",
    "    fileName = template+'-fixed-RMSmean-venn-'+regulation.lower()+'.xlsx'\n",
    "    tempDF.to_excel(fileDir+ipynbName+fileName, sheet_name='README', header=True, index=False)\n",
    "    display(tempDF)#Check\n",
    "    \n",
    "    #Prepare summary table for the template\n",
    "    tempDF1 = statDF_fixed.loc[template]\n",
    "    tempDF1 = tempDF1.dropna(axis=1)#NaN due to the eliminated test for the consensus group\n",
    "    \n",
    "    t_start = time.time()\n",
    "    #Extract overall set\n",
    "    for key_i in range(len(tempD)):\n",
    "        key = list(tempD.keys())[key_i]\n",
    "        tempS = tempD[key]\n",
    "        tempDF = tempDF1.loc[tempDF1.index.isin(tempS)]\n",
    "        #Save summary table by appended to the above .xlsx file\n",
    "        ##Prepare sheet name\n",
    "        tempL1 = ['NA' for i in range(len(tempD))]\n",
    "        tempL1[key_i] = '1'\n",
    "        setName = '('+','.join(tempL1)+')'\n",
    "        with pd.ExcelWriter(fileDir+ipynbName+fileName, mode='a', engine='openpyxl') as writer:\n",
    "            tempDF.to_excel(writer, sheet_name=setName, header=True, index=True)\n",
    "        print(' - '+setName+':', len(tempDF))\n",
    "    \n",
    "    #Extract subset\n",
    "    tempL1 = ['1', '0']\n",
    "    tempL2 = [[k1, k2, k3, k4] for k1 in tempL1 for k2 in tempL1 for k3 in tempL1 for k4 in tempL1]\n",
    "    #tempL2.remove(['0', '0', '0', '0'])\n",
    "    for tempL1 in tempL2:\n",
    "        #Positive module set\n",
    "        tempL3 = [list(tempD.values())[key_i] for key_i, binary in enumerate(tempL1) if binary=='1']\n",
    "        tempS1 = set(pvalDF.index.tolist())#Initialize\n",
    "        for tempS in tempL3:\n",
    "            tempS1 = tempS1 & tempS\n",
    "        #Negative module set\n",
    "        tempL3 = [list(tempD.values())[key_i] for key_i, binary in enumerate(tempL1) if binary=='0']\n",
    "        tempS2 = set()#Initialize\n",
    "        for tempS in tempL3:\n",
    "            tempS2 = tempS2 | tempS\n",
    "        #Extract subset\n",
    "        tempS = tempS1 - tempS2\n",
    "        tempDF = tempDF1.loc[tempDF1.index.isin(tempS)]\n",
    "        #Save summary table by appended to the above .xlsx file\n",
    "        ##Prepare sheet name\n",
    "        setName = '('+','.join(tempL1)+')'\n",
    "        with pd.ExcelWriter(fileDir+ipynbName+fileName, mode='a', engine='openpyxl') as writer:\n",
    "            tempDF.to_excel(writer, sheet_name=setName, header=True, index=True)\n",
    "        print(' - '+setName+':', len(tempDF))\n",
    "    \n",
    "    t_elapsed = time.time() - t_start\n",
    "    print(' - Elapsed time:', round(t_elapsed//60), 'min', round(t_elapsed%60, 1), 'sec')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 5-3. Visualization: pointplot"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> More modules are checked later (in a separate notebook), and so representative modules are checked for now.  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 5-3-1. Modules tightened by all interventions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Prepare the target module set\n",
    "posL = ['Aca-vs-Ctrl', '17aE2-vs-Ctrl', 'Rapa-vs-Ctrl']\n",
    "negL = ['']\n",
    "regulation = 'Tightened'\n",
    "tempS = pd.Series(np.repeat(True, len(pvalDF)), index=pvalDF.index)#Initialize\n",
    "tempDF = pvalDF.loc[:, pvalDF.columns.str.contains('-vs-')]#Adjusted P-value\n",
    "tempDF = tempDF.loc[:, ~tempDF.columns.str.contains('-fixed')]\n",
    "for col_n in tempDF.columns.tolist():\n",
    "    tempS1 = pvalDF[col_n]\n",
    "    tempS2 = diffDF[col_n]\n",
    "    if col_n in posL:\n",
    "        if regulation=='Tightened':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2>0)\n",
    "        elif regulation=='Loosened':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2<0)\n",
    "    elif col_n in negL:\n",
    "        tempS3 = (tempS1>=0.05)\n",
    "        #Significance for inverse regulation\n",
    "        if regulation=='Tightened':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2<0)\n",
    "        elif regulation=='Loosened':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2>0)\n",
    "        tempS1 = tempS3 | tempS1\n",
    "    else:\n",
    "        tempS1 = (tempS1>=0.0)\n",
    "    #Update True\n",
    "    tempS = tempS & tempS1\n",
    "tempL = tempS.loc[tempS.tolist()].index.tolist()\n",
    "print(len(tempL), regulation.lower()+' modules with significance in', posL, 'but not in', negL)\n",
    "\n",
    "#Select representatives\n",
    "topX = np.min([30, len(tempL)])\n",
    "topX_plot = np.min([10, len(tempL)])\n",
    "tempDF = statDF.loc[:, statDF.columns.str.contains('Pval$')]\n",
    "tempDF = pd.merge(statDF['ModuleName'], tempDF, left_index=True, right_index=True, how='left')\n",
    "tempDF = tempDF.loc[tempL].sort_values(by=variable+'_AdjPval', ascending=True)\n",
    "print('Top', topX, 'modules (sort by the main effect of '+variable+'):')\n",
    "display(tempDF.iloc[:topX])\n",
    "plotL = tempDF.index.tolist()[:topX_plot]\n",
    "\n",
    "#Prepare DF for plot\n",
    "tempDF = rmsDF.reset_index().melt(var_name='SampleID', value_name='RMS', id_vars=['ModuleID', 'Template'])\n",
    "tempDF1 = sampleDF.reset_index()[['SampleID', 'Phenotype']]\n",
    "tempDF = pd.merge(tempDF, tempDF1, on='SampleID', how='left')\n",
    "\n",
    "#Prepare label and color\n",
    "tempD0 = {'Ctrl':'Control', 'Aca':'Acarbose',\n",
    "          '17aE2':'17'+r'$\\alpha$'+'-Estradiol', 'Rapa':'Rapamycin'}\n",
    "tempDF['Template'] = tempDF['Template'].map(tempD0)\n",
    "tempDF['Group'] = tempDF['Phenotype'].map(tempD0)\n",
    "tempD1 = {'Control':'tab:blue', 'Acarbose':'tab:red',\n",
    "          '17'+r'$\\alpha$'+'-Estradiol':'tab:green', 'Rapamycin':'tab:purple'}\n",
    "tempD2 = {}\n",
    "for label in tempD1.keys():\n",
    "    if tempD1[label]=='tab:blue':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(1)\n",
    "    elif tempD1[label]=='tab:orange':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(3)\n",
    "    elif tempD1[label]=='tab:green':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(5)\n",
    "    elif tempD1[label]=='tab:red':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(7)\n",
    "    elif tempD1[label]=='tab:purple':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(9)\n",
    "    elif tempD1[label]=='tab:olive':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(17)\n",
    "    else:\n",
    "        tempD2[label] = 'Error?'\n",
    "\n",
    "#Visualize each representative\n",
    "for rank_i in range(len(plotL)):\n",
    "    print(' - Rank '+str(rank_i+1)+' (sort by the main effect of '+variable+'):')\n",
    "    module = plotL[rank_i]\n",
    "    #Check module summary\n",
    "    tempDF1 = pd.DataFrame(moduleDF.loc[module]).T\n",
    "    display(tempDF1)\n",
    "    \n",
    "    #Select RMS\n",
    "    tempDF1 = tempDF.loc[tempDF['ModuleID']==module]\n",
    "    \n",
    "    #Check RMS summary\n",
    "    tempDF2 = tempDF1.groupby(['Template', 'Group'])['RMS'].agg(['count', 'mean', 'std'])\n",
    "    tempL1 = []\n",
    "    tempL2 = []\n",
    "    for row_n in tempDF2.index.tolist():\n",
    "        count, mean, std = tempDF2.loc[row_n]\n",
    "        tempL1.append(mean - 1.96*std/np.sqrt(count))\n",
    "        tempL2.append(mean + 1.96*std/np.sqrt(count))\n",
    "    tempDF2['0.025'] = tempL1\n",
    "    tempDF2['0.975'] = tempL2\n",
    "    ##Multiindex sort\n",
    "    tempDF2 = tempDF2.reset_index()\n",
    "    tempDF2['Template'] = pd.Categorical(tempDF2['Template'], categories=list(tempD2.keys()))\n",
    "    tempDF2['Group'] = pd.Categorical(tempDF2['Group'], categories=list(tempD1.keys()))\n",
    "    tempDF2 = tempDF2.sort_values(by=['Template', 'Group']).set_index(['Template', 'Group'])\n",
    "    display(tempDF2)\n",
    "    \n",
    "    #Prepare significance labels\n",
    "    ##Retrieve statistical significance\n",
    "    tempS = pvalDF.loc[module, pvalDF.columns.str.contains('-vs-')]#Adjusted P-value\n",
    "    tempS = tempS.loc[tempS.index.str.contains('-fixed')]\n",
    "    tempS.name = 'AdjPval'\n",
    "    ##Clean\n",
    "    tempDF2 = tempS.index.to_series().str.split(pat='-fixed_', expand=True)\n",
    "    tempDF2 = tempDF2.rename(columns={0:'Template', 1:'Comparison'})\n",
    "    tempS1 = tempDF2['Template']\n",
    "    tempDF2 = tempDF2['Comparison'].str.split(pat='-vs-', expand=True)\n",
    "    tempDF2 = tempDF2.rename(columns={0:'Contrast', 1:'Baseline'})\n",
    "    tempDF2 = pd.merge(tempS1, tempDF2, left_index=True, right_index=True, how='left')\n",
    "    tempDF2 = pd.merge(tempDF2, tempS, left_index=True, right_index=True, how='left')\n",
    "    tempDF2['Template'] = tempDF2['Template'].map(tempD0)\n",
    "    tempDF2['Contrast'] = tempDF2['Contrast'].map(tempD0)\n",
    "    tempDF2['Baseline'] = tempDF2['Baseline'].map(tempD0)\n",
    "    ##Convert p-value to label\n",
    "    tempL = []\n",
    "    for row_i in range(len(tempDF2)):\n",
    "        pval = tempDF2['AdjPval'].iloc[row_i]\n",
    "        if pval<0.001:\n",
    "            tempL.append('***')\n",
    "        elif pval<0.01:\n",
    "            tempL.append('**')\n",
    "        elif pval<0.05:\n",
    "            tempL.append('*')\n",
    "        else:\n",
    "            pval_text = Decimal(str(pval)).quantize(Decimal('0.01'), rounding=ROUND_HALF_UP)\n",
    "            tempL.append(r'$P$ = '+str(pval_text))\n",
    "    tempDF2['SignifLabel'] = tempL\n",
    "    display(tempDF2)\n",
    "    \n",
    "    #Visualization\n",
    "    ymax = 1.0\n",
    "    ymin = 0.0\n",
    "    yinter = 0.2\n",
    "    ymargin_t = 0.275\n",
    "    ymargin_b = 0.05\n",
    "    aline_ymin = 1.0\n",
    "    aline_ymargin = 0.125\n",
    "    sns.set(style='ticks', font='Arial', context='talk')\n",
    "    fig, axes = plt.subplots(nrows=1, ncols=4, figsize=(8.8, 5.6), sharex=True, sharey=True,\n",
    "                             gridspec_kw={'width_ratios':[2.2, 2.2, 2.2, 2.2]})\n",
    "    for ax_i, ax in enumerate(axes.flat):\n",
    "        template = list(tempD2.keys())[ax_i]\n",
    "        tempDF3 = tempDF1.loc[tempDF1['Template']==template]\n",
    "        sns.pointplot(data=tempDF3, x='Group', y='RMS', order=list(tempD1.keys()), palette=tempD1,\n",
    "                      markers='o', dodge=False, join=False, capsize=0.6, estimator=np.mean, ci=95, ax=ax)\n",
    "        sns.stripplot(data=tempDF3, x='Group', y='RMS',\n",
    "                      order=list(tempD1.keys()), palette=tempD1, dodge=False, jitter=0.15,\n",
    "                      size=5, edgecolor='black', linewidth=1, **{'marker':'o', 'alpha':0.5}, ax=ax)\n",
    "        #Add RCI line\n",
    "        rci = tempDF3['RMS'].loc[tempDF3['Group']==template].mean()\n",
    "        ax.axhline(y=rci, **{'linestyle':'--', 'color':tempD1[template], 'zorder':0})\n",
    "        #Set axis\n",
    "        sns.despine()\n",
    "        ax.set(ylim=(ymin-ymargin_b, ymax+ymargin_t), yticks=np.arange(ymin, ymax + yinter/10, yinter))\n",
    "        plt.setp(ax.get_xticklabels(), rotation=70, horizontalalignment='right',\n",
    "                 verticalalignment='center', rotation_mode='anchor')\n",
    "        if ax_i==0:\n",
    "            plt.setp(ax, xlabel='', ylabel='Sample RMS')\n",
    "        else:\n",
    "            plt.setp(ax.get_yticklabels(), visible=False)\n",
    "            plt.setp(ax, xlabel='', ylabel='')\n",
    "        #Add significance labels\n",
    "        tempDF3 = tempDF2.loc[tempDF2['Template']==template]\n",
    "        for row_i in range(len(tempDF3)):\n",
    "            #Baseline\n",
    "            group_0 = tempDF3['Baseline'].iloc[row_i]\n",
    "            index_0 = list(tempD1.keys()).index(group_0)\n",
    "            xcoord_0 = index_0\n",
    "            #Contrast\n",
    "            group_1 = tempDF3['Contrast'].iloc[row_i]\n",
    "            index_1 = list(tempD1.keys()).index(group_1)\n",
    "            xcoord_1 = index_1\n",
    "            #Standard point of marker\n",
    "            xcoord = (xcoord_0+xcoord_1)/2\n",
    "            ycoord = aline_ymin + aline_ymargin*row_i\n",
    "            label = tempDF3['SignifLabel'].iloc[row_i]\n",
    "            #Add annotation lines\n",
    "            aline_offset = yinter/5\n",
    "            aline_length = yinter/5 + aline_offset/2\n",
    "            ax.plot([xcoord_0, xcoord_0, xcoord_1, xcoord_1],\n",
    "                    [ycoord+aline_offset, ycoord+aline_length, ycoord+aline_length, ycoord+aline_offset],\n",
    "                    lw=1.5, c='k')\n",
    "            #Add annotation text\n",
    "            if label in ['***', '**', '*']:\n",
    "                text_offset = yinter/21\n",
    "                ax.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                            horizontalalignment='center', verticalalignment='bottom',\n",
    "                            fontsize='medium', color='k')\n",
    "            else:\n",
    "                text_offset = yinter/3.5\n",
    "                ax.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                            horizontalalignment='center', verticalalignment='bottom',\n",
    "                            fontsize='x-small', color='k')\n",
    "        #Add annotation\n",
    "        ax.set_title('Consensus:\\n'+template, {'fontsize':'small'})\n",
    "        xoff = 0.015\n",
    "        yoff = 0.01\n",
    "        rect = plt.Rectangle((xoff, 1), 1-xoff, 0.17,#Manual adjustment\n",
    "                             transform=ax.transAxes, facecolor=tempD2[template],\n",
    "                             clip_on=False, linewidth=0, zorder=0.5)\n",
    "        ax.add_patch(rect)\n",
    "    fig.tight_layout()\n",
    "    #Set title\n",
    "    modulename = moduleDF.loc[module, 'ModuleName']\n",
    "    initial = modulename[0].capitalize()\n",
    "    title = re.sub('^.', initial, modulename)+' ('+module+')'\n",
    "    title = '\\n'.join(wrap(title, 100))#Because the below wrap=True didn't work\n",
    "    fig.suptitle(title, size='small',\n",
    "                 verticalalignment='baseline', horizontalalignment='center', wrap=True, y=0.925)\n",
    "    ##Save\n",
    "    #fileDir = './ExportFigures/'\n",
    "    #ipynbName = '230214_LC-M001-proteomics-DIRAC-ver7-2_DIRAC-GOBP_'\n",
    "    #fileName = 'fixed-RMSmean-pointplot-'+module.replace('GO:', 'GO')+'.pdf'\n",
    "    #plt.gcf().savefig(fileDir+ipynbName+fileName, dpi=300, bbox_inches='tight', pad_inches=0.04, transparent=True)\n",
    "    plt.show()\n",
    "    print('')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 5-3-2. Top changed modules by Aca"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Prepare the target module set\n",
    "target = 'Aca-vs-Ctrl'\n",
    "tempL = pvalDF.loc[pvalDF[target]<0.05].index.tolist()\n",
    "print(len(tempL), 'changed modules with significance in '+target)\n",
    "\n",
    "#Select representatives\n",
    "topX = np.min([30, len(tempL)])\n",
    "topX_plot = np.min([3, len(tempL)])\n",
    "tempDF = statDF.loc[:, statDF.columns.str.contains('Pval$')]\n",
    "tempDF = pd.merge(statDF['ModuleName'], tempDF, left_index=True, right_index=True, how='left')\n",
    "tempDF = tempDF.loc[tempL].sort_values(by=target+'_AdjPval', ascending=True)\n",
    "print('Top', topX, 'modules (sort by significance in '+target+'):')\n",
    "display(tempDF.iloc[:topX])\n",
    "plotL = tempDF.index.tolist()[:topX_plot]\n",
    "\n",
    "#Prepare DF for plot\n",
    "tempDF = rmsDF.reset_index().melt(var_name='SampleID', value_name='RMS', id_vars=['ModuleID', 'Template'])\n",
    "tempDF1 = sampleDF.reset_index()[['SampleID', 'Phenotype']]\n",
    "tempDF = pd.merge(tempDF, tempDF1, on='SampleID', how='left')\n",
    "\n",
    "#Prepare label and color\n",
    "tempD0 = {'Ctrl':'Control', 'Aca':'Acarbose',\n",
    "          '17aE2':'17'+r'$\\alpha$'+'-Estradiol', 'Rapa':'Rapamycin'}\n",
    "tempDF['Template'] = tempDF['Template'].map(tempD0)\n",
    "tempDF['Group'] = tempDF['Phenotype'].map(tempD0)\n",
    "tempD1 = {'Control':'tab:blue', 'Acarbose':'tab:red',\n",
    "          '17'+r'$\\alpha$'+'-Estradiol':'tab:green', 'Rapamycin':'tab:purple'}\n",
    "tempD2 = {}\n",
    "for label in tempD1.keys():\n",
    "    if tempD1[label]=='tab:blue':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(1)\n",
    "    elif tempD1[label]=='tab:orange':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(3)\n",
    "    elif tempD1[label]=='tab:green':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(5)\n",
    "    elif tempD1[label]=='tab:red':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(7)\n",
    "    elif tempD1[label]=='tab:purple':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(9)\n",
    "    elif tempD1[label]=='tab:olive':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(17)\n",
    "    else:\n",
    "        tempD2[label] = 'Error?'\n",
    "\n",
    "#Visualize each representative\n",
    "for rank_i in range(len(plotL)):\n",
    "    print(' - Rank '+str(rank_i+1)+' (sort by significance in '+target+'):')\n",
    "    module = plotL[rank_i]\n",
    "    #Check module summary\n",
    "    tempDF1 = pd.DataFrame(moduleDF.loc[module]).T\n",
    "    display(tempDF1)\n",
    "    \n",
    "    #Select RMS\n",
    "    tempDF1 = tempDF.loc[tempDF['ModuleID']==module]\n",
    "    \n",
    "    #Check RMS summary\n",
    "    tempDF2 = tempDF1.groupby(['Template', 'Group'])['RMS'].agg(['count', 'mean', 'std'])\n",
    "    tempL1 = []\n",
    "    tempL2 = []\n",
    "    for row_n in tempDF2.index.tolist():\n",
    "        count, mean, std = tempDF2.loc[row_n]\n",
    "        tempL1.append(mean - 1.96*std/np.sqrt(count))\n",
    "        tempL2.append(mean + 1.96*std/np.sqrt(count))\n",
    "    tempDF2['0.025'] = tempL1\n",
    "    tempDF2['0.975'] = tempL2\n",
    "    ##Multiindex sort\n",
    "    tempDF2 = tempDF2.reset_index()\n",
    "    tempDF2['Template'] = pd.Categorical(tempDF2['Template'], categories=list(tempD2.keys()))\n",
    "    tempDF2['Group'] = pd.Categorical(tempDF2['Group'], categories=list(tempD1.keys()))\n",
    "    tempDF2 = tempDF2.sort_values(by=['Template', 'Group']).set_index(['Template', 'Group'])\n",
    "    display(tempDF2)\n",
    "    \n",
    "    #Prepare significance labels\n",
    "    ##Retrieve statistical significance\n",
    "    tempS = pvalDF.loc[module, pvalDF.columns.str.contains('-vs-')]#Adjusted P-value\n",
    "    tempS = tempS.loc[tempS.index.str.contains('-fixed')]\n",
    "    tempS.name = 'AdjPval'\n",
    "    ##Clean\n",
    "    tempDF2 = tempS.index.to_series().str.split(pat='-fixed_', expand=True)\n",
    "    tempDF2 = tempDF2.rename(columns={0:'Template', 1:'Comparison'})\n",
    "    tempS1 = tempDF2['Template']\n",
    "    tempDF2 = tempDF2['Comparison'].str.split(pat='-vs-', expand=True)\n",
    "    tempDF2 = tempDF2.rename(columns={0:'Contrast', 1:'Baseline'})\n",
    "    tempDF2 = pd.merge(tempS1, tempDF2, left_index=True, right_index=True, how='left')\n",
    "    tempDF2 = pd.merge(tempDF2, tempS, left_index=True, right_index=True, how='left')\n",
    "    tempDF2['Template'] = tempDF2['Template'].map(tempD0)\n",
    "    tempDF2['Contrast'] = tempDF2['Contrast'].map(tempD0)\n",
    "    tempDF2['Baseline'] = tempDF2['Baseline'].map(tempD0)\n",
    "    ##Convert p-value to label\n",
    "    tempL = []\n",
    "    for row_i in range(len(tempDF2)):\n",
    "        pval = tempDF2['AdjPval'].iloc[row_i]\n",
    "        if pval<0.001:\n",
    "            tempL.append('***')\n",
    "        elif pval<0.01:\n",
    "            tempL.append('**')\n",
    "        elif pval<0.05:\n",
    "            tempL.append('*')\n",
    "        else:\n",
    "            pval_text = Decimal(str(pval)).quantize(Decimal('0.01'), rounding=ROUND_HALF_UP)\n",
    "            tempL.append(r'$P$ = '+str(pval_text))\n",
    "    tempDF2['SignifLabel'] = tempL\n",
    "    display(tempDF2)\n",
    "    \n",
    "    #Visualization\n",
    "    ymax = 1.0\n",
    "    ymin = 0.0\n",
    "    yinter = 0.2\n",
    "    ymargin_t = 0.275\n",
    "    ymargin_b = 0.05\n",
    "    aline_ymin = 1.0\n",
    "    aline_ymargin = 0.125\n",
    "    sns.set(style='ticks', font='Arial', context='talk')\n",
    "    fig, axes = plt.subplots(nrows=1, ncols=4, figsize=(8.8, 5.6), sharex=True, sharey=True,\n",
    "                             gridspec_kw={'width_ratios':[2.2, 2.2, 2.2, 2.2]})\n",
    "    for ax_i, ax in enumerate(axes.flat):\n",
    "        template = list(tempD2.keys())[ax_i]\n",
    "        tempDF3 = tempDF1.loc[tempDF1['Template']==template]\n",
    "        sns.pointplot(data=tempDF3, x='Group', y='RMS', order=list(tempD1.keys()), palette=tempD1,\n",
    "                      markers='o', dodge=False, join=False, capsize=0.6, estimator=np.mean, ci=95, ax=ax)\n",
    "        sns.stripplot(data=tempDF3, x='Group', y='RMS',\n",
    "                      order=list(tempD1.keys()), palette=tempD1, dodge=False, jitter=0.15,\n",
    "                      size=5, edgecolor='black', linewidth=1, **{'marker':'o', 'alpha':0.5}, ax=ax)\n",
    "        #Add RCI line\n",
    "        rci = tempDF3['RMS'].loc[tempDF3['Group']==template].mean()\n",
    "        ax.axhline(y=rci, **{'linestyle':'--', 'color':tempD1[template], 'zorder':0})\n",
    "        #Set axis\n",
    "        sns.despine()\n",
    "        ax.set(ylim=(ymin-ymargin_b, ymax+ymargin_t), yticks=np.arange(ymin, ymax + yinter/10, yinter))\n",
    "        plt.setp(ax.get_xticklabels(), rotation=70, horizontalalignment='right',\n",
    "                 verticalalignment='center', rotation_mode='anchor')\n",
    "        if ax_i==0:\n",
    "            plt.setp(ax, xlabel='', ylabel='Sample RMS')\n",
    "        else:\n",
    "            plt.setp(ax.get_yticklabels(), visible=False)\n",
    "            plt.setp(ax, xlabel='', ylabel='')\n",
    "        #Add significance labels\n",
    "        tempDF3 = tempDF2.loc[tempDF2['Template']==template]\n",
    "        for row_i in range(len(tempDF3)):\n",
    "            #Baseline\n",
    "            group_0 = tempDF3['Baseline'].iloc[row_i]\n",
    "            index_0 = list(tempD1.keys()).index(group_0)\n",
    "            xcoord_0 = index_0\n",
    "            #Contrast\n",
    "            group_1 = tempDF3['Contrast'].iloc[row_i]\n",
    "            index_1 = list(tempD1.keys()).index(group_1)\n",
    "            xcoord_1 = index_1\n",
    "            #Standard point of marker\n",
    "            xcoord = (xcoord_0+xcoord_1)/2\n",
    "            ycoord = aline_ymin + aline_ymargin*row_i\n",
    "            label = tempDF3['SignifLabel'].iloc[row_i]\n",
    "            #Add annotation lines\n",
    "            aline_offset = yinter/5\n",
    "            aline_length = yinter/5 + aline_offset/2\n",
    "            ax.plot([xcoord_0, xcoord_0, xcoord_1, xcoord_1],\n",
    "                    [ycoord+aline_offset, ycoord+aline_length, ycoord+aline_length, ycoord+aline_offset],\n",
    "                    lw=1.5, c='k')\n",
    "            #Add annotation text\n",
    "            if label in ['***', '**', '*']:\n",
    "                text_offset = yinter/21\n",
    "                ax.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                            horizontalalignment='center', verticalalignment='bottom',\n",
    "                            fontsize='medium', color='k')\n",
    "            else:\n",
    "                text_offset = yinter/3.5\n",
    "                ax.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                            horizontalalignment='center', verticalalignment='bottom',\n",
    "                            fontsize='x-small', color='k')\n",
    "        #Add annotation\n",
    "        ax.set_title('Consensus:\\n'+template, {'fontsize':'small'})\n",
    "        xoff = 0.015\n",
    "        yoff = 0.01\n",
    "        rect = plt.Rectangle((xoff, 1), 1-xoff, 0.17,#Manual adjustment\n",
    "                             transform=ax.transAxes, facecolor=tempD2[template],\n",
    "                             clip_on=False, linewidth=0, zorder=0.5)\n",
    "        ax.add_patch(rect)\n",
    "    fig.tight_layout()\n",
    "    #Set title\n",
    "    modulename = moduleDF.loc[module, 'ModuleName']\n",
    "    initial = modulename[0].capitalize()\n",
    "    title = re.sub('^.', initial, modulename)+' ('+module+')'\n",
    "    title = '\\n'.join(wrap(title, 100))#Because the below wrap=True didn't work\n",
    "    fig.suptitle(title, size='small',\n",
    "                 verticalalignment='baseline', horizontalalignment='center', wrap=True, y=0.925)\n",
    "    ##Save\n",
    "    #fileDir = './ExportFigures/'\n",
    "    #ipynbName = '230214_LC-M001-proteomics-DIRAC-ver7-2_DIRAC-GOBP_'\n",
    "    #fileName = 'fixed-RMSmean-pointplot-'+module.replace('GO:', 'GO')+'.pdf'\n",
    "    #plt.gcf().savefig(fileDir+ipynbName+fileName, dpi=300, bbox_inches='tight', pad_inches=0.04, transparent=True)\n",
    "    plt.show()\n",
    "    print('')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 5-3-3. Top changed modules by 17aE2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Prepare the target module set\n",
    "target = '17aE2-vs-Ctrl'\n",
    "tempL = pvalDF.loc[pvalDF[target]<0.05].index.tolist()\n",
    "print(len(tempL), 'changed modules with significance in '+target)\n",
    "\n",
    "#Select representatives\n",
    "topX = np.min([30, len(tempL)])\n",
    "topX_plot = np.min([3, len(tempL)])\n",
    "tempDF = statDF.loc[:, statDF.columns.str.contains('Pval$')]\n",
    "tempDF = pd.merge(statDF['ModuleName'], tempDF, left_index=True, right_index=True, how='left')\n",
    "tempDF = tempDF.loc[tempL].sort_values(by=target+'_AdjPval', ascending=True)\n",
    "print('Top', topX, 'modules (sort by significance in '+target+'):')\n",
    "display(tempDF.iloc[:topX])\n",
    "plotL = tempDF.index.tolist()[:topX_plot]\n",
    "\n",
    "#Prepare DF for plot\n",
    "tempDF = rmsDF.reset_index().melt(var_name='SampleID', value_name='RMS', id_vars=['ModuleID', 'Template'])\n",
    "tempDF1 = sampleDF.reset_index()[['SampleID', 'Phenotype']]\n",
    "tempDF = pd.merge(tempDF, tempDF1, on='SampleID', how='left')\n",
    "\n",
    "#Prepare label and color\n",
    "tempD0 = {'Ctrl':'Control', 'Aca':'Acarbose',\n",
    "          '17aE2':'17'+r'$\\alpha$'+'-Estradiol', 'Rapa':'Rapamycin'}\n",
    "tempDF['Template'] = tempDF['Template'].map(tempD0)\n",
    "tempDF['Group'] = tempDF['Phenotype'].map(tempD0)\n",
    "tempD1 = {'Control':'tab:blue', 'Acarbose':'tab:red',\n",
    "          '17'+r'$\\alpha$'+'-Estradiol':'tab:green', 'Rapamycin':'tab:purple'}\n",
    "tempD2 = {}\n",
    "for label in tempD1.keys():\n",
    "    if tempD1[label]=='tab:blue':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(1)\n",
    "    elif tempD1[label]=='tab:orange':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(3)\n",
    "    elif tempD1[label]=='tab:green':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(5)\n",
    "    elif tempD1[label]=='tab:red':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(7)\n",
    "    elif tempD1[label]=='tab:purple':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(9)\n",
    "    elif tempD1[label]=='tab:olive':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(17)\n",
    "    else:\n",
    "        tempD2[label] = 'Error?'\n",
    "\n",
    "#Visualize each representative\n",
    "for rank_i in range(len(plotL)):\n",
    "    print(' - Rank '+str(rank_i+1)+' (sort by significance in '+target+'):')\n",
    "    module = plotL[rank_i]\n",
    "    #Check module summary\n",
    "    tempDF1 = pd.DataFrame(moduleDF.loc[module]).T\n",
    "    display(tempDF1)\n",
    "    \n",
    "    #Select RMS\n",
    "    tempDF1 = tempDF.loc[tempDF['ModuleID']==module]\n",
    "    \n",
    "    #Check RMS summary\n",
    "    tempDF2 = tempDF1.groupby(['Template', 'Group'])['RMS'].agg(['count', 'mean', 'std'])\n",
    "    tempL1 = []\n",
    "    tempL2 = []\n",
    "    for row_n in tempDF2.index.tolist():\n",
    "        count, mean, std = tempDF2.loc[row_n]\n",
    "        tempL1.append(mean - 1.96*std/np.sqrt(count))\n",
    "        tempL2.append(mean + 1.96*std/np.sqrt(count))\n",
    "    tempDF2['0.025'] = tempL1\n",
    "    tempDF2['0.975'] = tempL2\n",
    "    ##Multiindex sort\n",
    "    tempDF2 = tempDF2.reset_index()\n",
    "    tempDF2['Template'] = pd.Categorical(tempDF2['Template'], categories=list(tempD2.keys()))\n",
    "    tempDF2['Group'] = pd.Categorical(tempDF2['Group'], categories=list(tempD1.keys()))\n",
    "    tempDF2 = tempDF2.sort_values(by=['Template', 'Group']).set_index(['Template', 'Group'])\n",
    "    display(tempDF2)\n",
    "    \n",
    "    #Prepare significance labels\n",
    "    ##Retrieve statistical significance\n",
    "    tempS = pvalDF.loc[module, pvalDF.columns.str.contains('-vs-')]#Adjusted P-value\n",
    "    tempS = tempS.loc[tempS.index.str.contains('-fixed')]\n",
    "    tempS.name = 'AdjPval'\n",
    "    ##Clean\n",
    "    tempDF2 = tempS.index.to_series().str.split(pat='-fixed_', expand=True)\n",
    "    tempDF2 = tempDF2.rename(columns={0:'Template', 1:'Comparison'})\n",
    "    tempS1 = tempDF2['Template']\n",
    "    tempDF2 = tempDF2['Comparison'].str.split(pat='-vs-', expand=True)\n",
    "    tempDF2 = tempDF2.rename(columns={0:'Contrast', 1:'Baseline'})\n",
    "    tempDF2 = pd.merge(tempS1, tempDF2, left_index=True, right_index=True, how='left')\n",
    "    tempDF2 = pd.merge(tempDF2, tempS, left_index=True, right_index=True, how='left')\n",
    "    tempDF2['Template'] = tempDF2['Template'].map(tempD0)\n",
    "    tempDF2['Contrast'] = tempDF2['Contrast'].map(tempD0)\n",
    "    tempDF2['Baseline'] = tempDF2['Baseline'].map(tempD0)\n",
    "    ##Convert p-value to label\n",
    "    tempL = []\n",
    "    for row_i in range(len(tempDF2)):\n",
    "        pval = tempDF2['AdjPval'].iloc[row_i]\n",
    "        if pval<0.001:\n",
    "            tempL.append('***')\n",
    "        elif pval<0.01:\n",
    "            tempL.append('**')\n",
    "        elif pval<0.05:\n",
    "            tempL.append('*')\n",
    "        else:\n",
    "            pval_text = Decimal(str(pval)).quantize(Decimal('0.01'), rounding=ROUND_HALF_UP)\n",
    "            tempL.append(r'$P$ = '+str(pval_text))\n",
    "    tempDF2['SignifLabel'] = tempL\n",
    "    display(tempDF2)\n",
    "    \n",
    "    #Visualization\n",
    "    ymax = 1.0\n",
    "    ymin = 0.0\n",
    "    yinter = 0.2\n",
    "    ymargin_t = 0.275\n",
    "    ymargin_b = 0.05\n",
    "    aline_ymin = 1.0\n",
    "    aline_ymargin = 0.125\n",
    "    sns.set(style='ticks', font='Arial', context='talk')\n",
    "    fig, axes = plt.subplots(nrows=1, ncols=4, figsize=(8.8, 5.6), sharex=True, sharey=True,\n",
    "                             gridspec_kw={'width_ratios':[2.2, 2.2, 2.2, 2.2]})\n",
    "    for ax_i, ax in enumerate(axes.flat):\n",
    "        template = list(tempD2.keys())[ax_i]\n",
    "        tempDF3 = tempDF1.loc[tempDF1['Template']==template]\n",
    "        sns.pointplot(data=tempDF3, x='Group', y='RMS', order=list(tempD1.keys()), palette=tempD1,\n",
    "                      markers='o', dodge=False, join=False, capsize=0.6, estimator=np.mean, ci=95, ax=ax)\n",
    "        sns.stripplot(data=tempDF3, x='Group', y='RMS',\n",
    "                      order=list(tempD1.keys()), palette=tempD1, dodge=False, jitter=0.15,\n",
    "                      size=5, edgecolor='black', linewidth=1, **{'marker':'o', 'alpha':0.5}, ax=ax)\n",
    "        #Add RCI line\n",
    "        rci = tempDF3['RMS'].loc[tempDF3['Group']==template].mean()\n",
    "        ax.axhline(y=rci, **{'linestyle':'--', 'color':tempD1[template], 'zorder':0})\n",
    "        #Set axis\n",
    "        sns.despine()\n",
    "        ax.set(ylim=(ymin-ymargin_b, ymax+ymargin_t), yticks=np.arange(ymin, ymax + yinter/10, yinter))\n",
    "        plt.setp(ax.get_xticklabels(), rotation=70, horizontalalignment='right',\n",
    "                 verticalalignment='center', rotation_mode='anchor')\n",
    "        if ax_i==0:\n",
    "            plt.setp(ax, xlabel='', ylabel='Sample RMS')\n",
    "        else:\n",
    "            plt.setp(ax.get_yticklabels(), visible=False)\n",
    "            plt.setp(ax, xlabel='', ylabel='')\n",
    "        #Add significance labels\n",
    "        tempDF3 = tempDF2.loc[tempDF2['Template']==template]\n",
    "        for row_i in range(len(tempDF3)):\n",
    "            #Baseline\n",
    "            group_0 = tempDF3['Baseline'].iloc[row_i]\n",
    "            index_0 = list(tempD1.keys()).index(group_0)\n",
    "            xcoord_0 = index_0\n",
    "            #Contrast\n",
    "            group_1 = tempDF3['Contrast'].iloc[row_i]\n",
    "            index_1 = list(tempD1.keys()).index(group_1)\n",
    "            xcoord_1 = index_1\n",
    "            #Standard point of marker\n",
    "            xcoord = (xcoord_0+xcoord_1)/2\n",
    "            ycoord = aline_ymin + aline_ymargin*row_i\n",
    "            label = tempDF3['SignifLabel'].iloc[row_i]\n",
    "            #Add annotation lines\n",
    "            aline_offset = yinter/5\n",
    "            aline_length = yinter/5 + aline_offset/2\n",
    "            ax.plot([xcoord_0, xcoord_0, xcoord_1, xcoord_1],\n",
    "                    [ycoord+aline_offset, ycoord+aline_length, ycoord+aline_length, ycoord+aline_offset],\n",
    "                    lw=1.5, c='k')\n",
    "            #Add annotation text\n",
    "            if label in ['***', '**', '*']:\n",
    "                text_offset = yinter/21\n",
    "                ax.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                            horizontalalignment='center', verticalalignment='bottom',\n",
    "                            fontsize='medium', color='k')\n",
    "            else:\n",
    "                text_offset = yinter/3.5\n",
    "                ax.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                            horizontalalignment='center', verticalalignment='bottom',\n",
    "                            fontsize='x-small', color='k')\n",
    "        #Add annotation\n",
    "        ax.set_title('Consensus:\\n'+template, {'fontsize':'small'})\n",
    "        xoff = 0.015\n",
    "        yoff = 0.01\n",
    "        rect = plt.Rectangle((xoff, 1), 1-xoff, 0.17,#Manual adjustment\n",
    "                             transform=ax.transAxes, facecolor=tempD2[template],\n",
    "                             clip_on=False, linewidth=0, zorder=0.5)\n",
    "        ax.add_patch(rect)\n",
    "    fig.tight_layout()\n",
    "    #Set title\n",
    "    modulename = moduleDF.loc[module, 'ModuleName']\n",
    "    initial = modulename[0].capitalize()\n",
    "    title = re.sub('^.', initial, modulename)+' ('+module+')'\n",
    "    title = '\\n'.join(wrap(title, 100))#Because the below wrap=True didn't work\n",
    "    fig.suptitle(title, size='small',\n",
    "                 verticalalignment='baseline', horizontalalignment='center', wrap=True, y=0.925)\n",
    "    ##Save\n",
    "    #fileDir = './ExportFigures/'\n",
    "    #ipynbName = '230214_LC-M001-proteomics-DIRAC-ver7-2_DIRAC-GOBP_'\n",
    "    #fileName = 'fixed-RMSmean-pointplot-'+module.replace('GO:', 'GO')+'.pdf'\n",
    "    #plt.gcf().savefig(fileDir+ipynbName+fileName, dpi=300, bbox_inches='tight', pad_inches=0.04, transparent=True)\n",
    "    plt.show()\n",
    "    print('')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 5-3-4. Top changed modules by Rapa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Prepare the target module set\n",
    "target = 'Rapa-vs-Ctrl'\n",
    "tempL = pvalDF.loc[pvalDF[target]<0.05].index.tolist()\n",
    "print(len(tempL), 'changed modules with significance in '+target)\n",
    "\n",
    "#Select representatives\n",
    "topX = np.min([30, len(tempL)])\n",
    "topX_plot = np.min([3, len(tempL)])\n",
    "tempDF = statDF.loc[:, statDF.columns.str.contains('Pval$')]\n",
    "tempDF = pd.merge(statDF['ModuleName'], tempDF, left_index=True, right_index=True, how='left')\n",
    "tempDF = tempDF.loc[tempL].sort_values(by=target+'_AdjPval', ascending=True)\n",
    "print('Top', topX, 'modules (sort by significance in '+target+'):')\n",
    "display(tempDF.iloc[:topX])\n",
    "plotL = tempDF.index.tolist()[:topX_plot]\n",
    "\n",
    "#Prepare DF for plot\n",
    "tempDF = rmsDF.reset_index().melt(var_name='SampleID', value_name='RMS', id_vars=['ModuleID', 'Template'])\n",
    "tempDF1 = sampleDF.reset_index()[['SampleID', 'Phenotype']]\n",
    "tempDF = pd.merge(tempDF, tempDF1, on='SampleID', how='left')\n",
    "\n",
    "#Prepare label and color\n",
    "tempD0 = {'Ctrl':'Control', 'Aca':'Acarbose',\n",
    "          '17aE2':'17'+r'$\\alpha$'+'-Estradiol', 'Rapa':'Rapamycin'}\n",
    "tempDF['Template'] = tempDF['Template'].map(tempD0)\n",
    "tempDF['Group'] = tempDF['Phenotype'].map(tempD0)\n",
    "tempD1 = {'Control':'tab:blue', 'Acarbose':'tab:red',\n",
    "          '17'+r'$\\alpha$'+'-Estradiol':'tab:green', 'Rapamycin':'tab:purple'}\n",
    "tempD2 = {}\n",
    "for label in tempD1.keys():\n",
    "    if tempD1[label]=='tab:blue':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(1)\n",
    "    elif tempD1[label]=='tab:orange':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(3)\n",
    "    elif tempD1[label]=='tab:green':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(5)\n",
    "    elif tempD1[label]=='tab:red':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(7)\n",
    "    elif tempD1[label]=='tab:purple':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(9)\n",
    "    elif tempD1[label]=='tab:olive':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(17)\n",
    "    else:\n",
    "        tempD2[label] = 'Error?'\n",
    "\n",
    "#Visualize each representative\n",
    "for rank_i in range(len(plotL)):\n",
    "    print(' - Rank '+str(rank_i+1)+' (sort by significance in '+target+'):')\n",
    "    module = plotL[rank_i]\n",
    "    #Check module summary\n",
    "    tempDF1 = pd.DataFrame(moduleDF.loc[module]).T\n",
    "    display(tempDF1)\n",
    "    \n",
    "    #Select RMS\n",
    "    tempDF1 = tempDF.loc[tempDF['ModuleID']==module]\n",
    "    \n",
    "    #Check RMS summary\n",
    "    tempDF2 = tempDF1.groupby(['Template', 'Group'])['RMS'].agg(['count', 'mean', 'std'])\n",
    "    tempL1 = []\n",
    "    tempL2 = []\n",
    "    for row_n in tempDF2.index.tolist():\n",
    "        count, mean, std = tempDF2.loc[row_n]\n",
    "        tempL1.append(mean - 1.96*std/np.sqrt(count))\n",
    "        tempL2.append(mean + 1.96*std/np.sqrt(count))\n",
    "    tempDF2['0.025'] = tempL1\n",
    "    tempDF2['0.975'] = tempL2\n",
    "    ##Multiindex sort\n",
    "    tempDF2 = tempDF2.reset_index()\n",
    "    tempDF2['Template'] = pd.Categorical(tempDF2['Template'], categories=list(tempD2.keys()))\n",
    "    tempDF2['Group'] = pd.Categorical(tempDF2['Group'], categories=list(tempD1.keys()))\n",
    "    tempDF2 = tempDF2.sort_values(by=['Template', 'Group']).set_index(['Template', 'Group'])\n",
    "    display(tempDF2)\n",
    "    \n",
    "    #Prepare significance labels\n",
    "    ##Retrieve statistical significance\n",
    "    tempS = pvalDF.loc[module, pvalDF.columns.str.contains('-vs-')]#Adjusted P-value\n",
    "    tempS = tempS.loc[tempS.index.str.contains('-fixed')]\n",
    "    tempS.name = 'AdjPval'\n",
    "    ##Clean\n",
    "    tempDF2 = tempS.index.to_series().str.split(pat='-fixed_', expand=True)\n",
    "    tempDF2 = tempDF2.rename(columns={0:'Template', 1:'Comparison'})\n",
    "    tempS1 = tempDF2['Template']\n",
    "    tempDF2 = tempDF2['Comparison'].str.split(pat='-vs-', expand=True)\n",
    "    tempDF2 = tempDF2.rename(columns={0:'Contrast', 1:'Baseline'})\n",
    "    tempDF2 = pd.merge(tempS1, tempDF2, left_index=True, right_index=True, how='left')\n",
    "    tempDF2 = pd.merge(tempDF2, tempS, left_index=True, right_index=True, how='left')\n",
    "    tempDF2['Template'] = tempDF2['Template'].map(tempD0)\n",
    "    tempDF2['Contrast'] = tempDF2['Contrast'].map(tempD0)\n",
    "    tempDF2['Baseline'] = tempDF2['Baseline'].map(tempD0)\n",
    "    ##Convert p-value to label\n",
    "    tempL = []\n",
    "    for row_i in range(len(tempDF2)):\n",
    "        pval = tempDF2['AdjPval'].iloc[row_i]\n",
    "        if pval<0.001:\n",
    "            tempL.append('***')\n",
    "        elif pval<0.01:\n",
    "            tempL.append('**')\n",
    "        elif pval<0.05:\n",
    "            tempL.append('*')\n",
    "        else:\n",
    "            pval_text = Decimal(str(pval)).quantize(Decimal('0.01'), rounding=ROUND_HALF_UP)\n",
    "            tempL.append(r'$P$ = '+str(pval_text))\n",
    "    tempDF2['SignifLabel'] = tempL\n",
    "    display(tempDF2)\n",
    "    \n",
    "    #Visualization\n",
    "    ymax = 1.0\n",
    "    ymin = 0.0\n",
    "    yinter = 0.2\n",
    "    ymargin_t = 0.275\n",
    "    ymargin_b = 0.05\n",
    "    aline_ymin = 1.0\n",
    "    aline_ymargin = 0.125\n",
    "    sns.set(style='ticks', font='Arial', context='talk')\n",
    "    fig, axes = plt.subplots(nrows=1, ncols=4, figsize=(8.8, 5.6), sharex=True, sharey=True,\n",
    "                             gridspec_kw={'width_ratios':[2.2, 2.2, 2.2, 2.2]})\n",
    "    for ax_i, ax in enumerate(axes.flat):\n",
    "        template = list(tempD2.keys())[ax_i]\n",
    "        tempDF3 = tempDF1.loc[tempDF1['Template']==template]\n",
    "        sns.pointplot(data=tempDF3, x='Group', y='RMS', order=list(tempD1.keys()), palette=tempD1,\n",
    "                      markers='o', dodge=False, join=False, capsize=0.6, estimator=np.mean, ci=95, ax=ax)\n",
    "        sns.stripplot(data=tempDF3, x='Group', y='RMS',\n",
    "                      order=list(tempD1.keys()), palette=tempD1, dodge=False, jitter=0.15,\n",
    "                      size=5, edgecolor='black', linewidth=1, **{'marker':'o', 'alpha':0.5}, ax=ax)\n",
    "        #Add RCI line\n",
    "        rci = tempDF3['RMS'].loc[tempDF3['Group']==template].mean()\n",
    "        ax.axhline(y=rci, **{'linestyle':'--', 'color':tempD1[template], 'zorder':0})\n",
    "        #Set axis\n",
    "        sns.despine()\n",
    "        ax.set(ylim=(ymin-ymargin_b, ymax+ymargin_t), yticks=np.arange(ymin, ymax + yinter/10, yinter))\n",
    "        plt.setp(ax.get_xticklabels(), rotation=70, horizontalalignment='right',\n",
    "                 verticalalignment='center', rotation_mode='anchor')\n",
    "        if ax_i==0:\n",
    "            plt.setp(ax, xlabel='', ylabel='Sample RMS')\n",
    "        else:\n",
    "            plt.setp(ax.get_yticklabels(), visible=False)\n",
    "            plt.setp(ax, xlabel='', ylabel='')\n",
    "        #Add significance labels\n",
    "        tempDF3 = tempDF2.loc[tempDF2['Template']==template]\n",
    "        for row_i in range(len(tempDF3)):\n",
    "            #Baseline\n",
    "            group_0 = tempDF3['Baseline'].iloc[row_i]\n",
    "            index_0 = list(tempD1.keys()).index(group_0)\n",
    "            xcoord_0 = index_0\n",
    "            #Contrast\n",
    "            group_1 = tempDF3['Contrast'].iloc[row_i]\n",
    "            index_1 = list(tempD1.keys()).index(group_1)\n",
    "            xcoord_1 = index_1\n",
    "            #Standard point of marker\n",
    "            xcoord = (xcoord_0+xcoord_1)/2\n",
    "            ycoord = aline_ymin + aline_ymargin*row_i\n",
    "            label = tempDF3['SignifLabel'].iloc[row_i]\n",
    "            #Add annotation lines\n",
    "            aline_offset = yinter/5\n",
    "            aline_length = yinter/5 + aline_offset/2\n",
    "            ax.plot([xcoord_0, xcoord_0, xcoord_1, xcoord_1],\n",
    "                    [ycoord+aline_offset, ycoord+aline_length, ycoord+aline_length, ycoord+aline_offset],\n",
    "                    lw=1.5, c='k')\n",
    "            #Add annotation text\n",
    "            if label in ['***', '**', '*']:\n",
    "                text_offset = yinter/21\n",
    "                ax.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                            horizontalalignment='center', verticalalignment='bottom',\n",
    "                            fontsize='medium', color='k')\n",
    "            else:\n",
    "                text_offset = yinter/3.5\n",
    "                ax.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                            horizontalalignment='center', verticalalignment='bottom',\n",
    "                            fontsize='x-small', color='k')\n",
    "        #Add annotation\n",
    "        ax.set_title('Consensus:\\n'+template, {'fontsize':'small'})\n",
    "        xoff = 0.015\n",
    "        yoff = 0.01\n",
    "        rect = plt.Rectangle((xoff, 1), 1-xoff, 0.17,#Manual adjustment\n",
    "                             transform=ax.transAxes, facecolor=tempD2[template],\n",
    "                             clip_on=False, linewidth=0, zorder=0.5)\n",
    "        ax.add_patch(rect)\n",
    "    fig.tight_layout()\n",
    "    #Set title\n",
    "    modulename = moduleDF.loc[module, 'ModuleName']\n",
    "    initial = modulename[0].capitalize()\n",
    "    title = re.sub('^.', initial, modulename)+' ('+module+')'\n",
    "    title = '\\n'.join(wrap(title, 100))#Because the below wrap=True didn't work\n",
    "    fig.suptitle(title, size='small',\n",
    "                 verticalalignment='baseline', horizontalalignment='center', wrap=True, y=0.925)\n",
    "    ##Save\n",
    "    #fileDir = './ExportFigures/'\n",
    "    #ipynbName = '230214_LC-M001-proteomics-DIRAC-ver7-2_DIRAC-GOBP_'\n",
    "    #fileName = 'fixed-RMSmean-pointplot-'+module.replace('GO:', 'GO')+'.pdf'\n",
    "    #plt.gcf().savefig(fileDir+ipynbName+fileName, dpi=300, bbox_inches='tight', pad_inches=0.04, transparent=True)\n",
    "    plt.show()\n",
    "    print('')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 5-3-5. Top RCI modules in Ctrl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "target = 'Ctrl'\n",
    "\n",
    "#Select representatives\n",
    "topX = 30\n",
    "topX_plot = 3\n",
    "tempDF = statDF.loc[:, statDF.columns.str.contains('Pval$')]\n",
    "tempDF = pd.merge(statDF['ModuleName'], tempDF, left_index=True, right_index=True, how='left')\n",
    "tempDF = pd.merge(tempDF, rciDF_kk, left_index=True, right_index=True, how='left')\n",
    "tempDF = tempDF.sort_values(by=target, ascending=False)\n",
    "print('Top', topX, 'modules (sort by RCI of '+target+'):')\n",
    "display(tempDF.iloc[:topX])\n",
    "plotL = tempDF.index.tolist()[:topX_plot]\n",
    "\n",
    "#Prepare DF for plot\n",
    "tempDF = rmsDF.reset_index().melt(var_name='SampleID', value_name='RMS', id_vars=['ModuleID', 'Template'])\n",
    "tempDF1 = sampleDF.reset_index()[['SampleID', 'Phenotype']]\n",
    "tempDF = pd.merge(tempDF, tempDF1, on='SampleID', how='left')\n",
    "\n",
    "#Prepare label and color\n",
    "tempD0 = {'Ctrl':'Control', 'Aca':'Acarbose',\n",
    "          '17aE2':'17'+r'$\\alpha$'+'-Estradiol', 'Rapa':'Rapamycin'}\n",
    "tempDF['Template'] = tempDF['Template'].map(tempD0)\n",
    "tempDF['Group'] = tempDF['Phenotype'].map(tempD0)\n",
    "tempD1 = {'Control':'tab:blue', 'Acarbose':'tab:red',\n",
    "          '17'+r'$\\alpha$'+'-Estradiol':'tab:green', 'Rapamycin':'tab:purple'}\n",
    "tempD2 = {}\n",
    "for label in tempD1.keys():\n",
    "    if tempD1[label]=='tab:blue':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(1)\n",
    "    elif tempD1[label]=='tab:orange':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(3)\n",
    "    elif tempD1[label]=='tab:green':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(5)\n",
    "    elif tempD1[label]=='tab:red':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(7)\n",
    "    elif tempD1[label]=='tab:purple':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(9)\n",
    "    elif tempD1[label]=='tab:olive':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(17)\n",
    "    else:\n",
    "        tempD2[label] = 'Error?'\n",
    "\n",
    "#Re-prepare p-values because these modules can be filtered out in pvalDF\n",
    "tempDF4 = statDF.loc[:, statDF.columns.str.contains('-vs-.*_AdjPval$')]\n",
    "tempDF4.columns = tempDF4.columns.str.replace('_AdjPval$', '')\n",
    "for template in ['Aca', '17aE2', 'Rapa']:\n",
    "    #Reproduce the fixed consensus version of summary table\n",
    "    fileDir = './ExportData/'\n",
    "    ipynbName = '230214_LC-M001-proteomics-DIRAC-ver7-2_DIRAC-GOBP_'\n",
    "    fileName = 'inter-group-comparison_'+template+'-fixed-RMSmean.tsv'\n",
    "    tempDF1 = pd.read_csv(fileDir+ipynbName+fileName, sep='\\t')\n",
    "    tempDF1 = tempDF1.set_index('ModuleID')\n",
    "    #Take adjusted P-value\n",
    "    tempDF1 = tempDF1.loc[:, tempDF1.columns.str.contains('-vs-.*_AdjPval$')]\n",
    "    tempDF1.columns = tempDF1.columns.str.replace('_AdjPval$', '')\n",
    "    tempDF1.columns = template+'-fixed_'+tempDF1.columns\n",
    "    #Update\n",
    "    tempDF4 = pd.merge(tempDF4, tempDF1, left_index=True, right_index=True, how='left')\n",
    "\n",
    "#Visualize each representative\n",
    "for rank_i in range(len(plotL)):\n",
    "    print(' - Rank '+str(rank_i+1)+' (sort by RCI of '+target+'):')\n",
    "    module = plotL[rank_i]\n",
    "    #Check module summary\n",
    "    tempDF1 = pd.DataFrame(moduleDF.loc[module]).T\n",
    "    display(tempDF1)\n",
    "    \n",
    "    #Select RMS\n",
    "    tempDF1 = tempDF.loc[tempDF['ModuleID']==module]\n",
    "    \n",
    "    #Check RMS summary\n",
    "    tempDF2 = tempDF1.groupby(['Template', 'Group'])['RMS'].agg(['count', 'mean', 'std'])\n",
    "    tempL1 = []\n",
    "    tempL2 = []\n",
    "    for row_n in tempDF2.index.tolist():\n",
    "        count, mean, std = tempDF2.loc[row_n]\n",
    "        tempL1.append(mean - 1.96*std/np.sqrt(count))\n",
    "        tempL2.append(mean + 1.96*std/np.sqrt(count))\n",
    "    tempDF2['0.025'] = tempL1\n",
    "    tempDF2['0.975'] = tempL2\n",
    "    ##Multiindex sort\n",
    "    tempDF2 = tempDF2.reset_index()\n",
    "    tempDF2['Template'] = pd.Categorical(tempDF2['Template'], categories=list(tempD2.keys()))\n",
    "    tempDF2['Group'] = pd.Categorical(tempDF2['Group'], categories=list(tempD1.keys()))\n",
    "    tempDF2 = tempDF2.sort_values(by=['Template', 'Group']).set_index(['Template', 'Group'])\n",
    "    display(tempDF2)\n",
    "    \n",
    "    #Prepare significance labels\n",
    "    ##Re-prepare p-values because these modules can be filtered out in pvalDF\n",
    "    ##-> Moved to outside of for-loop\n",
    "    ##Retrieve statistical significance\n",
    "    tempS = tempDF4.loc[module]\n",
    "    tempS = tempS.loc[tempS.index.str.contains('-fixed')]\n",
    "    tempS.name = 'AdjPval'\n",
    "    ##Clean\n",
    "    tempDF2 = tempS.index.to_series().str.split(pat='-fixed_', expand=True)\n",
    "    tempDF2 = tempDF2.rename(columns={0:'Template', 1:'Comparison'})\n",
    "    tempS1 = tempDF2['Template']\n",
    "    tempDF2 = tempDF2['Comparison'].str.split(pat='-vs-', expand=True)\n",
    "    tempDF2 = tempDF2.rename(columns={0:'Contrast', 1:'Baseline'})\n",
    "    tempDF2 = pd.merge(tempS1, tempDF2, left_index=True, right_index=True, how='left')\n",
    "    tempDF2 = pd.merge(tempDF2, tempS, left_index=True, right_index=True, how='left')\n",
    "    tempDF2['Template'] = tempDF2['Template'].map(tempD0)\n",
    "    tempDF2['Contrast'] = tempDF2['Contrast'].map(tempD0)\n",
    "    tempDF2['Baseline'] = tempDF2['Baseline'].map(tempD0)\n",
    "    ##Convert p-value to label\n",
    "    tempL = []\n",
    "    for row_i in range(len(tempDF2)):\n",
    "        pval = tempDF2['AdjPval'].iloc[row_i]\n",
    "        if pval<0.001:\n",
    "            tempL.append('***')\n",
    "        elif pval<0.01:\n",
    "            tempL.append('**')\n",
    "        elif pval<0.05:\n",
    "            tempL.append('*')\n",
    "        else:\n",
    "            pval_text = Decimal(str(pval)).quantize(Decimal('0.01'), rounding=ROUND_HALF_UP)\n",
    "            tempL.append(r'$P$ = '+str(pval_text))\n",
    "    tempDF2['SignifLabel'] = tempL\n",
    "    display(tempDF2)\n",
    "    \n",
    "    #Visualization\n",
    "    ymax = 1.0\n",
    "    ymin = 0.0\n",
    "    yinter = 0.2\n",
    "    ymargin_t = 0.275\n",
    "    ymargin_b = 0.05\n",
    "    aline_ymin = 1.0\n",
    "    aline_ymargin = 0.125\n",
    "    sns.set(style='ticks', font='Arial', context='talk')\n",
    "    fig, axes = plt.subplots(nrows=1, ncols=4, figsize=(8.8, 5.6), sharex=True, sharey=True,\n",
    "                             gridspec_kw={'width_ratios':[2.2, 2.2, 2.2, 2.2]})\n",
    "    for ax_i, ax in enumerate(axes.flat):\n",
    "        template = list(tempD2.keys())[ax_i]\n",
    "        tempDF3 = tempDF1.loc[tempDF1['Template']==template]\n",
    "        sns.pointplot(data=tempDF3, x='Group', y='RMS', order=list(tempD1.keys()), palette=tempD1,\n",
    "                      markers='o', dodge=False, join=False, capsize=0.6, estimator=np.mean, ci=95, ax=ax)\n",
    "        sns.stripplot(data=tempDF3, x='Group', y='RMS',\n",
    "                      order=list(tempD1.keys()), palette=tempD1, dodge=False, jitter=0.15,\n",
    "                      size=5, edgecolor='black', linewidth=1, **{'marker':'o', 'alpha':0.5}, ax=ax)\n",
    "        #Add RCI line\n",
    "        rci = tempDF3['RMS'].loc[tempDF3['Group']==template].mean()\n",
    "        ax.axhline(y=rci, **{'linestyle':'--', 'color':tempD1[template], 'zorder':0})\n",
    "        #Set axis\n",
    "        sns.despine()\n",
    "        ax.set(ylim=(ymin-ymargin_b, ymax+ymargin_t), yticks=np.arange(ymin, ymax + yinter/10, yinter))\n",
    "        plt.setp(ax.get_xticklabels(), rotation=70, horizontalalignment='right',\n",
    "                 verticalalignment='center', rotation_mode='anchor')\n",
    "        if ax_i==0:\n",
    "            plt.setp(ax, xlabel='', ylabel='Sample RMS')\n",
    "        else:\n",
    "            plt.setp(ax.get_yticklabels(), visible=False)\n",
    "            plt.setp(ax, xlabel='', ylabel='')\n",
    "        #Add significance labels\n",
    "        tempDF3 = tempDF2.loc[tempDF2['Template']==template]\n",
    "        for row_i in range(len(tempDF3)):\n",
    "            #Baseline\n",
    "            group_0 = tempDF3['Baseline'].iloc[row_i]\n",
    "            index_0 = list(tempD1.keys()).index(group_0)\n",
    "            xcoord_0 = index_0\n",
    "            #Contrast\n",
    "            group_1 = tempDF3['Contrast'].iloc[row_i]\n",
    "            index_1 = list(tempD1.keys()).index(group_1)\n",
    "            xcoord_1 = index_1\n",
    "            #Standard point of marker\n",
    "            xcoord = (xcoord_0+xcoord_1)/2\n",
    "            ycoord = aline_ymin + aline_ymargin*row_i\n",
    "            label = tempDF3['SignifLabel'].iloc[row_i]\n",
    "            #Add annotation lines\n",
    "            aline_offset = yinter/5\n",
    "            aline_length = yinter/5 + aline_offset/2\n",
    "            ax.plot([xcoord_0, xcoord_0, xcoord_1, xcoord_1],\n",
    "                    [ycoord+aline_offset, ycoord+aline_length, ycoord+aline_length, ycoord+aline_offset],\n",
    "                    lw=1.5, c='k')\n",
    "            #Add annotation text\n",
    "            if label in ['***', '**', '*']:\n",
    "                text_offset = yinter/21\n",
    "                ax.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                            horizontalalignment='center', verticalalignment='bottom',\n",
    "                            fontsize='medium', color='k')\n",
    "            else:\n",
    "                text_offset = yinter/3.5\n",
    "                ax.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                            horizontalalignment='center', verticalalignment='bottom',\n",
    "                            fontsize='x-small', color='k')\n",
    "        #Add annotation\n",
    "        ax.set_title('Consensus:\\n'+template, {'fontsize':'small'})\n",
    "        xoff = 0.015\n",
    "        yoff = 0.01\n",
    "        rect = plt.Rectangle((xoff, 1), 1-xoff, 0.17,#Manual adjustment\n",
    "                             transform=ax.transAxes, facecolor=tempD2[template],\n",
    "                             clip_on=False, linewidth=0, zorder=0.5)\n",
    "        ax.add_patch(rect)\n",
    "    fig.tight_layout()\n",
    "    #Set title\n",
    "    modulename = moduleDF.loc[module, 'ModuleName']\n",
    "    initial = modulename[0].capitalize()\n",
    "    title = re.sub('^.', initial, modulename)+' ('+module+')'\n",
    "    title = '\\n'.join(wrap(title, 100))#Because the below wrap=True didn't work\n",
    "    fig.suptitle(title, size='small',\n",
    "                 verticalalignment='baseline', horizontalalignment='center', wrap=True, y=0.925)\n",
    "    ##Save\n",
    "    #fileDir = './ExportFigures/'\n",
    "    #ipynbName = '230214_LC-M001-proteomics-DIRAC-ver7-2_DIRAC-GOBP_'\n",
    "    #fileName = 'fixed-RMSmean-pointplot-'+module.replace('GO:', 'GO')+'.pdf'\n",
    "    #plt.gcf().savefig(fileDir+ipynbName+fileName, dpi=300, bbox_inches='tight', pad_inches=0.04, transparent=True)\n",
    "    plt.show()\n",
    "    print('')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# — End of notebook —"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "arivale-py3 - Python",
   "language": "python",
   "name": "conda-env-arivale-py3-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  },
  "widgets": {
   "application/vnd.jupyter.widget-state+json": {
    "state": {},
    "version_major": 2,
    "version_minor": 0
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
