{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# DIRAC Analysis of LC M001-related Transcriptomics — DIRAC with GOBP Modules"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***by Kengo Watanabe***  \n",
    "\n",
    "In this notebook, the differential rank conservation (DIRAC; Eddy, J.A. et al. PLoS Comput. Biol. 2010) analysis is performed on the preprocessed Longevity Consortium (LC) M001-related transcriptomics dataset (Tyshkovskiy, A. et al. Cell Metab. 2019; analytes detected in all samples; sample-based robust Z-score followed by analyte-based robust Z-score) using the retrieved a priori module set (Gene Ontology (Biological Process) derived by R org.Mm.eg.db package; ≥4 analytes and ≥50% coverage).  \n",
    "> Because Dunnett's test does not seem available in Python yet, statistical tests are performed in a different notebook with R kernel.  \n",
    "\n",
    "Input:  \n",
    "* Preprocessed analyte data: 201221_LC-M001-related-transcriptomics-DIRAC_cleaned-robustZscored.tsv  \n",
    "* Module–analyte metadata: 220516_LC-M001-related-transcriptomics-DIRAC_Preprocessing_ver2-2_module-metadata_orgMmegdb-GOBP_min-n4-cov50.tsv  \n",
    "* Analyte metadata: 220516_LC-M001-related-transcriptomics-DIRAC_Preprocessing_ver2-2_analyte-metadata_EnsDb.tsv  \n",
    "* Sample–mouse metadata: 201221_LC-M001-related-transcriptomics-DIRAC_metadata.tsv  \n",
    "* Statistical test summary: 220522_LC-M001-related-transcriptomics-DIRAC_StatisticalTest-GOBP_ver2-4_inter-group-comparison.xlsx (Supplementary Data 5)  \n",
    "\n",
    "Output:  \n",
    "* Cleaned module metadata, which is incorporated into Supplementary Data 5 in R sub-notebook  \n",
    "* DIRAC measures, which are used in statistical analysis (R sub-notebook)  \n",
    "* Figure 5a, h  \n",
    "* Supplementary Figure 3a–d, g, j  \n",
    "\n",
    "Original notebook (memo for my future tracing):  \n",
    "* dalek:[JupyterLab HOME]/201221_LC-M001-related-transcriptomics-DIRAC/220522_LC-M001-related-transcriptomics-DIRAC_DIRAC-GOBP_ver2-4.ipynb  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scipy.stats as stats\n",
    "import matplotlib.pyplot as plt\n",
    "%matplotlib inline\n",
    "import seaborn as sns\n",
    "#For Arial font\n",
    "#!conda install -c conda-forge -y mscorefonts\n",
    "##-> The below was also needed in matplotlib 3.4.2\n",
    "#import shutil\n",
    "#import matplotlib\n",
    "#shutil.rmtree(matplotlib.get_cachedir())\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')\n",
    "from IPython.display import display\n",
    "import time\n",
    "\n",
    "#!conda install -c bioconda goatools -y\n",
    "from goatools.obo_parser import GODag\n",
    "from itertools import combinations\n",
    "import math\n",
    "from multiprocessing import Pool\n",
    "from decimal import Decimal, ROUND_HALF_UP\n",
    "import re\n",
    "import matplotlib.patches as mpatches\n",
    "#!pip install venn\n",
    "from venn import venn\n",
    "#!conda install -c conda-forge -y matplotlib-venn\n",
    "from matplotlib_venn import venn3, venn3_circles, venn2, venn2_circles\n",
    "\n",
    "!conda list"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 0. DIRAC code"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> The original code for DIRAC was written in MATLAB. Hence, it is re-written in Python 3 here.  \n",
    "> <– I don't care about computational cost here; rather, the code adheres to the story in the original paper.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def network_ranking(DF, networkS):\n",
    "    # This function calculates the pairwise ordering of network genes (i.e., network ranking).\n",
    "    ## Ref. Eddy, J. A. et al. PLoS Comput. Biol. 2010 (Figure 1 at a glance)\n",
    "    # Requirements:\n",
    "    ## import numpy as np: confirmed with versions 1.17.5 and 1.21.1\n",
    "    ## import pandas as pd: confirmed with versions 0.25.3 and 1.3.1\n",
    "    ## from itertools import combinations: confirmed with Python 3.7.6 and 3.9.6\n",
    "    # Input:\n",
    "    ## DF: pd.DataFrame containing expression values (X_gn) with gene (g; 1-G) indices and sample (n; 1-N) columns\n",
    "    ## networkS: pd.Series containing genes (g; 1-G) with network (m; 1-M) indices (i.e., 1-on-1 long-format)\n",
    "    # Output:\n",
    "    ## pd.DataFrame containing binary values of the networking ranking comparison for each sample\n",
    "    ## -> row: comparison_id (indicated by network (m) and ordering (g_i < g_j) columns)\n",
    "    ## -> column: sample (n) (with exception of the NetworkID and Ordering columns)\n",
    "    # Note:\n",
    "    ## If items in network m and gene g contain ' : ' or ' < ', this code would produce error or unexpected output.\n",
    "    ## If an item in sample n is 'NetworkID' or 'Ordering', this code would produce error or unexpected output.\n",
    "    \n",
    "    #Calculate binary values of the network ranking comparison for each sample\n",
    "    rankDF = pd.DataFrame()\n",
    "    sampleL = DF.columns.tolist()\n",
    "    networkL = networkS.index.unique()\n",
    "    for n in sampleL:\n",
    "        rankDF_n = pd.DataFrame()\n",
    "        for m in networkL:\n",
    "            #Pairs of genes (g_i, g_j) in the network m\n",
    "            networkS_m = networkS.loc[m]\n",
    "            pairL_m = list(combinations(range(0, len(networkS_m)), 2))\n",
    "            pairL_m_i, pairL_m_j = [[pair[x] for pair in pairL_m] for x in (0,1)]\n",
    "            rankDF_m = pd.DataFrame({'g_i':networkS_m.iloc[pairL_m_i],\n",
    "                                     'g_j':networkS_m.iloc[pairL_m_j]})#Hold network (m; 1-M) index\n",
    "            #Compare the expression values (X_gn) between pairwise genes (g_i vs. g_j)\n",
    "            tempL = []\n",
    "            for pair_i in range(0, len(rankDF_m)):\n",
    "                g_i = rankDF_m.iloc[pair_i, 0]\n",
    "                g_j = rankDF_m.iloc[pair_i, 1]\n",
    "                X_i = DF.loc[g_i, n]\n",
    "                X_j = DF.loc[g_j, n]\n",
    "                #If X_i < X_j is true, add 1; otherwise (X_i >= X_j), add 0\n",
    "                if X_i < X_j:\n",
    "                    tempL.append(1)\n",
    "                else:\n",
    "                    tempL.append(0)\n",
    "            rankDF_m['X_i<X_j'] = tempL\n",
    "            #Update the network ranking dataframe of sample n\n",
    "            rankDF_m.index.set_names('NetworkID', inplace=True)#Set/reset index name\n",
    "            rankDF_m = rankDF_m.reset_index()\n",
    "            rankDF_n = pd.concat([rankDF_n, rankDF_m], axis=0)\n",
    "        #Updata the network ranking dataframe of all samples\n",
    "        rankDF_n['Sample'] = n\n",
    "        rankDF = pd.concat([rankDF, rankDF_n], axis=0)\n",
    "    ##Prepare dummy index and clean dataframe\n",
    "    rankDF['ComparisonID'] = rankDF['NetworkID'] + ' : ' + rankDF['g_i'] + ' < ' + rankDF['g_j']\n",
    "    rankDF = rankDF.pivot(index='ComparisonID', columns='Sample', values='X_i<X_j')#Sorted by index during this\n",
    "    rankDF = rankDF.reset_index()#Index becomes row number here\n",
    "    tempDF = rankDF['ComparisonID'].str.split(pat=' : ', expand=True)\n",
    "    tempDF = tempDF.rename(columns={0:'NetworkID', 1:'Ordering'})\n",
    "    rankDF = pd.concat([tempDF, rankDF], axis=1)#Dropping columns name 'Sample' during this\n",
    "    rankDF = rankDF.drop(columns='ComparisonID')\n",
    "    return rankDF\n",
    "\n",
    "def rank_template(rankDF, phenotypeS):\n",
    "    # This function generates the rank template (T) presenting the expected network ranking in a phenotype.\n",
    "    ## Ref. Eddy, J. A. et al. PLoS Comput. Biol. 2010 (Figure 1 at a galance)\n",
    "    # Requirements:\n",
    "    ## import numpy as np: confirmed with versions 1.17.5 and 1.21.1\n",
    "    ## import pandas as pd: confirmed with versions 0.25.3 and 1.3.1\n",
    "    # Input:\n",
    "    ## rankDF: pd.DataFrame obtained from the above network_ranking() function\n",
    "    ## phenotypeS: pd.Series containing phenotypes (k; 1-K) with sample (n; 1-N) indices (i.e., 1-on-1 long-format)\n",
    "    # Output:\n",
    "    ## pd.DataFrame containing the expected binary values of network ranking comparison for each phenotype (T_mk)\n",
    "    ## -> row: comparison_id (indicated by network (m) and ordering (g_i < g_j) columns)\n",
    "    ## -> column: phenotype (k) (with exception of the NetworkID and Ordering columns)\n",
    "    # Note:\n",
    "    ## True rate = 0.5 is assigned to 0 in this code.\n",
    "    ## If an item in phenotype k is 'NetworkID' or 'Ordering', this code would produce error or unexpected output.\n",
    "    \n",
    "    #Calculate the expected binary values of network ranking comparison for each phenotype (T_mk)\n",
    "    templateDF = rankDF[['NetworkID', 'Ordering']]\n",
    "    phenotypeL = phenotypeS.unique().tolist()\n",
    "    for k in phenotypeL:\n",
    "        sampleL_k = phenotypeS.loc[phenotypeS==k].index.tolist()\n",
    "        tempDF = rankDF[sampleL_k]\n",
    "        tempS = tempDF.mean(axis=1)#True (=1) rate\n",
    "        tempS = (tempS>0.5).astype('int64')#If true rate > 0.5, 1; otherwise (<= 0.5), 0\n",
    "        templateDF[k] = tempS\n",
    "    return templateDF\n",
    "\n",
    "def rank_matching_score(rankDF, templateDF):\n",
    "    # This function calculates the rank matching score (R) of a sample.\n",
    "    ## Ref. Eddy, J. A. et al. PLoS Comput. Biol. 2010 (Figure 1 at a galance)\n",
    "    # Requirements:\n",
    "    ## import numpy as np: confirmed with versions 1.17.5 and 1.21.1\n",
    "    ## import pandas as pd: confirmed with versions 0.25.3 and 1.3.1\n",
    "    # Input:\n",
    "    ## rankDF: pd.DataFrame obtained from the above network_ranking() function\n",
    "    ## templateDF: pd.DataFrame obtained from the above rank_template() function\n",
    "    # Output:\n",
    "    ## pd.DataFrame containing the rates of gene pairs matching to a rank template in a sample (R_mkn)\n",
    "    ## -> row: rank template (T_mk) (indicated by network (m) and template phenotype (k) columns)\n",
    "    ## -> column: sample (n) (with exception of the NetworkID and Template columns)\n",
    "    # Note:\n",
    "    ## True rate = 0.5 was assigned to 0 in the above rank_template() function.\n",
    "    ## -> 'Match (1)' and 'Mismatch (0)' are evenly assigned to samples in the phenotype.\n",
    "    ##    (i.e., 'Match' for (X_i < X_j) = 0 and 'Mismatch' for (X_i < X_j) = 1 in the tie case)\n",
    "    ## If items in network m and phenotype k contain ' : ', this code would produce error or unexpected output.\n",
    "    \n",
    "    #Calculate the rates of gene pairs matching to a rank template in a sample (R_mkn)\n",
    "    scoreDF = pd.DataFrame()\n",
    "    sampleL = rankDF.drop(columns=['NetworkID', 'Ordering']).columns.tolist()\n",
    "    phenotypeL = templateDF.drop(columns=['NetworkID', 'Ordering']).columns.tolist()\n",
    "    for n in sampleL:\n",
    "        scoreDF_n = pd.DataFrame()\n",
    "        for k_template in phenotypeL:\n",
    "            tempDF = rankDF[['NetworkID', 'Ordering']]\n",
    "            tempS = (rankDF[n]==templateDF[k_template]).astype('int64')#If matching, 1; otherwise, 0\n",
    "            tempDF['Match'] = tempS\n",
    "            #Calculate the rank matching score\n",
    "            scoreDF_k = tempDF.groupby(by='NetworkID', as_index=False, sort=False).mean()\n",
    "            #Update the rank matching score dataframe of sample n\n",
    "            scoreDF_k['Template'] = k_template\n",
    "            scoreDF_k = scoreDF_k[['NetworkID', 'Template', 'Match']]\n",
    "            scoreDF_n = pd.concat([scoreDF_n, scoreDF_k], axis=0)\n",
    "        #Update the rank matching score dataframe of all samples\n",
    "        scoreDF_n['Sample'] = n\n",
    "        scoreDF = pd.concat([scoreDF, scoreDF_n], axis=0)\n",
    "    ##Prepare dummy index and clean dataframe\n",
    "    scoreDF['RMSmkID'] = scoreDF['NetworkID'].str.cat(scoreDF['Template'], sep=' : ')\n",
    "    scoreDF = scoreDF.pivot(index='RMSmkID', columns='Sample', values='Match')#Sorted by index during this\n",
    "    scoreDF = scoreDF.reset_index()#Index becomes row number here\n",
    "    tempDF = scoreDF['RMSmkID'].str.split(pat=' : ', expand=True)\n",
    "    tempDF = tempDF.rename(columns={0:'NetworkID', 1:'Template'})\n",
    "    scoreDF = pd.concat([tempDF, scoreDF], axis=1)#Dropping columns name 'Sample' during this\n",
    "    scoreDF = scoreDF.drop(columns='RMSmkID')\n",
    "    return scoreDF\n",
    "\n",
    "def rank_conservation_index(scoreDF, phenotypeS):\n",
    "    # This function calculates the rank conservation index (muR) of a phenotype.\n",
    "    ## Ref. Eddy, J. A. et al. PLoS Comput. Biol. 2010 (Figure 1 at a galance)\n",
    "    # Requirements:\n",
    "    ## import numpy as np: confirmed with versions 1.17.5 and 1.21.1\n",
    "    ## import pandas as pd: confirmed with versions 0.25.3 and 1.3.1\n",
    "    # Input:\n",
    "    ## scoreDF: pd.DataFrame obtained from the above rank_matching_score() function\n",
    "    ## phenotypeS: pd.Series containing phenotypes (k; 1-K) with sample (n; 1-N) indices (i.e., 1-on-1 long-format)\n",
    "    # Output:\n",
    "    ## pd.DataFrame containing the mean values of rank matching scores (RMSs) in a phenotype (muR_mkk)\n",
    "    ## -> row: rank template (T_mk) (indicated by network (m) and template phenotype (k) columns)\n",
    "    ## -> column: phenotype (k) (with exception of the NetworkID and Template columns)\n",
    "    # Note:\n",
    "    ## Rank conservation index (RCI) is used with a broader stance here, but following a strict interpretation,\n",
    "    ## the term RCI is used for the mean of RMSs only when template phenotype is same with the sample phenotype.\n",
    "    ## If an item in phenotype k is 'NetworkID' or 'Template', this code would produce error or unexpected output.\n",
    "    \n",
    "    #Calculate the mean values of RMSs in a phenotype (muR_mkk)\n",
    "    conservationDF = scoreDF[['NetworkID', 'Template']]\n",
    "    phenotypeL = phenotypeS.unique().tolist()\n",
    "    for k in phenotypeL:\n",
    "        sampleL_k = phenotypeS.loc[phenotypeS==k].index.tolist()\n",
    "        tempS = scoreDF[sampleL_k].mean(axis=1)\n",
    "        conservationDF[k] = tempS\n",
    "    return conservationDF"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. Prepare dataset and metadata"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1-1. Analyte data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Import analyte data\n",
    "fileDir = './ExportData/'\n",
    "ipynbName = '201221_LC-M001-related-transcriptomics-DIRAC_'\n",
    "fileName = 'cleaned-robustZscored.tsv'\n",
    "tempDF = pd.read_csv(fileDir+ipynbName+fileName, sep='\\t').set_index('EnsemblID')\n",
    "display(tempDF)\n",
    "\n",
    "analyteDF = tempDF"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1-2. Module–analyte metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Import module-analyte metadata\n",
    "fileDir = './ExportData/'\n",
    "ipynbName = '220516_LC-M001-related-transcriptomics-DIRAC_Preprocessing_ver2-2_'\n",
    "fileName = 'module-metadata_orgMmegdb-GOBP_min-n4-cov50.tsv'\n",
    "tempDF = pd.read_csv(fileDir+ipynbName+fileName, sep='\\t').set_index('ModuleID')\n",
    "print(' - Unique analytes with module:', len(tempDF['EnsemblID'].unique()))\n",
    "print(' - Unique modules with analytes:', len(tempDF.index.unique()))\n",
    "\n",
    "#Prepare moduleS\n",
    "moduleS = tempDF['EnsemblID']\n",
    "display(moduleS)\n",
    "\n",
    "#Retrieve module metadata\n",
    "tempDF = tempDF[['ModuleName', 'nAnalytes', 'nBackgrounds', 'Coverage']]\n",
    "moduleDF = tempDF.reset_index().drop_duplicates(keep='first').set_index('ModuleID')\n",
    "display(moduleDF)\n",
    "display(moduleDF.describe(include='all'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> –> Retrieve module name drectly from go-basic OBO file using GOATOOLS.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Download the latest file\n",
    "!wget -O ./ImportData/go-basic.obo http://purl.obolibrary.org/obo/go/go-basic.obo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tempDF = moduleDF\n",
    "tempDF1 = tempDF.loc[tempDF['ModuleName']=='Dummy: to be retrieved later']\n",
    "print('No name:', len(tempDF1))\n",
    "\n",
    "#Initialize GODag object\n",
    "fileDir = './ImportData/'\n",
    "fileName = 'go-basic.obo'\n",
    "godag_gobasic = GODag(fileDir+fileName,\n",
    "                      optional_attrs=['def','consider', 'replaced_by'], load_obsolete=True)\n",
    "\n",
    "#Update GO name\n",
    "tempL = [godag_gobasic[goid].name for goid in tempDF.index.tolist()]\n",
    "tempDF['ModuleName'] = tempL\n",
    "\n",
    "#Check\n",
    "tempDF1 = tempDF.loc[tempDF['ModuleName']=='Dummy: to be retrieved later']\n",
    "print('-> After update:', len(tempDF1))\n",
    "tempDF1 = tempDF.loc[tempDF['ModuleName'].str.contains('obsolete')]\n",
    "print('-> Currently obsolete:', len(tempDF1))\n",
    "display(tempDF1)\n",
    "\n",
    "#Rename the obsolete for now\n",
    "tempDF['ModuleName'] = tempDF['ModuleName'].str.replace('^obsolete ', '')\n",
    "display(tempDF)\n",
    "\n",
    "moduleDF = tempDF"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> –> Add the mapped analytes to module metadata.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Import analyte metadata\n",
    "fileDir = './ExportData/'\n",
    "ipynbName = '220516_LC-M001-related-transcriptomics-DIRAC_Preprocessing_ver2-2_'\n",
    "fileName = 'analyte-metadata_EnsDb.tsv'\n",
    "tempDF = pd.read_csv(fileDir+ipynbName+fileName, sep='\\t')\n",
    "tempDF = pd.merge(moduleS.reset_index(), tempDF, on='EnsemblID', how='left').set_index('ModuleID')\n",
    "display(tempDF)\n",
    "print(' - Unique analytes:', len(tempDF['EnsemblID'].unique()))\n",
    "print(' - Unique labels:', len(tempDF['GeneLabel'].unique()))\n",
    "\n",
    "#Concatenate labels\n",
    "t_start = time.time()\n",
    "moduleDF['MappedAnalyteIDs'] = ''#Initialize\n",
    "moduleDF['MappedAnalyteGeneLabels'] = ''#Initialize\n",
    "for module in moduleDF.index.tolist():\n",
    "    tempS = tempDF['EnsemblID'].loc[tempDF.index.isin([module])]#Retrieve as pd.Series\n",
    "    label = tempS.str.cat(sep=';')\n",
    "    moduleDF.loc[module, 'MappedAnalyteIDs'] = label\n",
    "    tempS = tempDF['GeneLabel'].loc[tempDF.index.isin([module])]#Retrieve as pd.Series\n",
    "    label = tempS.str.cat(sep=';')\n",
    "    moduleDF.loc[module, 'MappedAnalyteGeneLabels'] = label\n",
    "t_elapsed = time.time() - t_start\n",
    "print('Elapsed time:', round(t_elapsed//60), 'min', round(t_elapsed%60, 1), 'sec')\n",
    "\n",
    "#Clean\n",
    "moduleDF['ModuleType'] = 'Gene Ontology (Biological Process)'\n",
    "moduleDF['Source'] = 'R org.Mm.eg.db package'\n",
    "moduleDF = moduleDF[['ModuleName', 'ModuleType', 'MappedAnalyteIDs', 'MappedAnalyteGeneLabels',\n",
    "                     'nAnalytes', 'nBackgrounds', 'Coverage', 'Source']]\n",
    "moduleDF = moduleDF.sort_index(ascending=True)\n",
    "\n",
    "#Save\n",
    "fileDir = './ExportData/'\n",
    "ipynbName = '220522_LC-M001-related-transcriptomics-DIRAC_DIRAC-GOBP_ver2-4_'\n",
    "fileName = 'module-metadata.tsv'\n",
    "moduleDF.to_csv(fileDir+ipynbName+fileName, index=True, sep='\\t')\n",
    "\n",
    "display(moduleDF)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1-3. Sample–mouse metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Import sample-mouse metadata\n",
    "fileDir = './ExportData/'\n",
    "ipynbName = '201221_LC-M001-related-transcriptomics-DIRAC_'\n",
    "fileName = 'metadata.tsv'\n",
    "tempDF = pd.read_csv(fileDir+ipynbName+fileName, sep='\\t')\n",
    "tempDF = tempDF.rename(columns={'ID':'SampleID'})\n",
    "tempDF = tempDF.loc[tempDF['SampleID'].isin(analyteDF.columns)]\n",
    "tempDF = tempDF.set_index('SampleID')\n",
    "tempDF = tempDF[['Intervention', 'Sex', 'Age']]\n",
    "tempD = {'-':'Control', 'Acarbose':'Acarbose', 'Rapamycin':'Rapamycin', 'CR':'Calorie restriction'}\n",
    "tempDF['Intervention'] = tempDF['Intervention'].map(tempD)\n",
    "\n",
    "#Prepare phenotypeS\n",
    "tempDF['Phenotype'] = tempDF['Intervention']\n",
    "display(tempDF)\n",
    "display(tempDF['Phenotype'].value_counts())\n",
    "\n",
    "sampleDF = tempDF"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. Perform DIRAC with sex-pooled rank consensus"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> When applied to overall dataset, the DIRAC code often stops runnning due to memory limit (e.g., transcriptomics).  \n",
    "> –> Divide moduleS into subsets while considering the number of comparisons, and compute DIRAC in parallel.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Divide moduleS\n",
    "cutoff = 500#The maximum number of analytes per module\n",
    "tempDF = moduleDF.sort_values(by='nAnalytes', ascending=True)\n",
    "nwsubL = []\n",
    "tempL = []#Initialize\n",
    "count = 0#Initialize\n",
    "for module in tempDF.index.tolist():\n",
    "    nanalytes = tempDF.loc[module, 'nAnalytes']\n",
    "    if nanalytes>cutoff:\n",
    "        nwsubL.append([module])\n",
    "    else:\n",
    "        tempL.append(module)\n",
    "        count += nanalytes\n",
    "        if count>cutoff:\n",
    "            nwsubL.append(tempL)\n",
    "            tempL = []#Initialize\n",
    "            count = 0#Initialize\n",
    "if len(tempL)>0:#The last one but still count<=cutoff\n",
    "    nwsubL.append(tempL)\n",
    "nSub = len(nwsubL)\n",
    "print('nSublists: ', nSub)\n",
    "print('nModules per sublist:', [len(sublist) for sublist in nwsubL])\n",
    "\n",
    "#Check nAnalytes distribution\n",
    "tempS = pd.Series(name='Subset')\n",
    "for list_i in range(nSub):\n",
    "    tempL = moduleS.loc[nwsubL[list_i]].index.unique().tolist()\n",
    "    tempS1 = pd.Series(np.repeat('Subset '+str(list_i+1).zfill(3), len(tempL)), index=tempL, name='Subset')\n",
    "    tempS = pd.concat([tempS, tempS1], axis=0)\n",
    "tempS.index.set_names('ModuleID', inplace=True)\n",
    "tempDF = pd.merge(moduleDF, tempS, left_index=True, right_index=True, how='left')\n",
    "tempDF = tempDF.sort_values(by='nAnalytes', ascending=False)\n",
    "display(tempDF)\n",
    "tempDF = tempDF.groupby('Subset').agg({'nAnalytes':[len, sum, max, np.median]})\n",
    "tempDF = tempDF.sort_values(by=('nAnalytes', 'sum'), ascending=False)\n",
    "print('Subset summary:')\n",
    "display(tempDF.describe())\n",
    "print(' -> Check subset which will need high computational cost:')\n",
    "display(tempDF.loc[tempDF[('nAnalytes', 'max')]>cutoff])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "nprocessors = 25\n",
    "fileDir = './ExportData/'\n",
    "ipynbName = '210429_LC-M001-related-transcriptomics-DIRAC_DIRAC-GOBP_ver2_'\n",
    "\n",
    "#Wrap as a single function\n",
    "def parallel_dirac(list_i):\n",
    "    tempS = moduleS.loc[nwsubL[list_i]]\n",
    "    \n",
    "    #Calculate the pairwise ordering of network genes (i.e., network ranking)\n",
    "    t_start1 = time.time()\n",
    "    rankingDF = network_ranking(analyteDF, tempS)\n",
    "    t_elapsed1 = time.time() - t_start1\n",
    "    \n",
    "    #Generate the rank template (T) presenting the expected network ranking in a phenotype\n",
    "    t_start2 = time.time()\n",
    "    templateDF = rank_template(rankingDF, sampleDF['Phenotype'])\n",
    "    t_elapsed2 = time.time() - t_start2\n",
    "    \n",
    "    #Calculate the rank matching score (R) of a sample\n",
    "    t_start3 = time.time()\n",
    "    rmsDF = rank_matching_score(rankingDF, templateDF)\n",
    "    t_elapsed3 = time.time() - t_start3\n",
    "    \n",
    "    #Calculate the rank conservation index (muR) of a phenotype\n",
    "    t_start4 = time.time()\n",
    "    rciDF = rank_conservation_index(rmsDF, sampleDF['Phenotype'])\n",
    "    t_elapsed4 = time.time() - t_start4\n",
    "    \n",
    "    #Save\n",
    "    fileName = 'NetworkRanking-'+str(list_i+1).zfill(3)+'.tsv'\n",
    "    rankingDF.to_csv(fileDir+ipynbName+fileName, index=False, sep='\\t')\n",
    "    fileName = 'RankTemplate-BS-'+str(list_i+1).zfill(3)+'.tsv'\n",
    "    templateDF.to_csv(fileDir+ipynbName+fileName, index=False, sep='\\t')\n",
    "    fileName = 'RankMatchingScore-BS-'+str(list_i+1).zfill(3)+'.tsv'\n",
    "    rmsDF.to_csv(fileDir+ipynbName+fileName, index=False, sep='\\t')\n",
    "    fileName = 'RankConservationIndex-BS-'+str(list_i+1).zfill(3)+'.tsv'\n",
    "    rciDF.to_csv(fileDir+ipynbName+fileName, index=False, sep='\\t')\n",
    "    \n",
    "    #Check results\n",
    "    print('Subset '+str(list_i+1).zfill(3)+':',\n",
    "          len(tempS.index.unique()), 'modules,', len(tempS.unique()), 'analytes')\n",
    "    print('• Network ranking dataframe:')\n",
    "    print('  - DF shape:', rankingDF.shape)\n",
    "    print('  - Elapsed time:', round(t_elapsed1//60), 'min', round(t_elapsed1%60, 1), 'sec')\n",
    "    print('• Rank template dataframe:')\n",
    "    print('  - DF shape:', templateDF.shape)\n",
    "    print('  - Elapsed time:', round(t_elapsed2//60), 'min', round(t_elapsed2%60, 1), 'sec')\n",
    "    print('• Rank matching score dataframe:')\n",
    "    print('  - DF shape:', rmsDF.shape)\n",
    "    print('  - Elapsed time:', round(t_elapsed3//60), 'min', round(t_elapsed3%60, 1), 'sec')\n",
    "    print('• Rank conservation index dataframe:')\n",
    "    print('  - DF shape:', rciDF.shape)\n",
    "    print('  - Elapsed time:', round(t_elapsed4//60), 'min', round(t_elapsed4%60, 1), 'sec')\n",
    "    print('')\n",
    "\n",
    "#Parallel computing\n",
    "if __name__=='__main__':\n",
    "    t_start = time.time()\n",
    "    p = Pool(nprocessors)\n",
    "    p.map(parallel_dirac, range(nSub))\n",
    "    t_finish = time.time()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Record as reference\n",
    "print(nSub, 'sublists with', nprocessors, 'processors')\n",
    "print(' - Start:', time.ctime(t_start))\n",
    "print(' - Finish:', time.ctime(t_finish))\n",
    "t_elapsed = (t_finish - t_start)\n",
    "print(' - Elapsed time:', round(t_elapsed//60), 'min', round(t_elapsed%60, 1), 'sec')\n",
    "t_elapsed = (t_finish - t_start) * nprocessors\n",
    "print(' - Total (approximate) elapsed time:', round(t_elapsed//60), 'min', round(t_elapsed%60, 1), 'sec')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Combine each result\n",
    "rmsDF = pd.DataFrame()\n",
    "rciDF = pd.DataFrame()\n",
    "for list_i in range(nSub):\n",
    "    fileName = 'RankMatchingScore-BS-'+str(list_i+1).zfill(3)+'.tsv'\n",
    "    tempDF = pd.read_csv(fileDir+ipynbName+fileName, sep='\\t')\n",
    "    tempDF = tempDF.rename(columns={'NetworkID':'ModuleID'})\n",
    "    rmsDF = pd.concat([rmsDF, tempDF], axis=0, ignore_index=True)\n",
    "    \n",
    "    fileName = 'RankConservationIndex-BS-'+str(list_i+1).zfill(3)+'.tsv'\n",
    "    tempDF = pd.read_csv(fileDir+ipynbName+fileName, sep='\\t')\n",
    "    tempDF = tempDF.rename(columns={'NetworkID':'ModuleID'})\n",
    "    rciDF = pd.concat([rciDF, tempDF], axis=0, ignore_index=True)\n",
    "\n",
    "print('• Rank matching score dataframe:')\n",
    "display(rmsDF)\n",
    "print('')\n",
    "print('• Rank conservation index dataframe:')\n",
    "display(rciDF)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3. Rank conservation index: general pattern"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3-1. Extract RCI (the mean of RMSs under the own phenotype consensus)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Extract RCI whose template phenotype corresponds to the own phenotype\n",
    "phenotypeL = rciDF.drop(columns=['ModuleID', 'Template']).columns.tolist()\n",
    "rciDF_kk = pd.DataFrame(index=pd.Index(rciDF['ModuleID'].unique(), name='ModuleID'))\n",
    "tempDF = rciDF.set_index('ModuleID')\n",
    "for k in phenotypeL:\n",
    "    tempS = tempDF[k].loc[tempDF['Template']==k]\n",
    "    rciDF_kk = pd.merge(rciDF_kk, tempS, left_index=True, right_index=True, how='left')\n",
    "\n",
    "#Order and re-label\n",
    "tempD = {'Control':'Control', 'Acarbose':'Acarbose',\n",
    "         'Rapamycin':'Rapamycin', 'Calorie restriction':'CR diet'}\n",
    "rciDF_kk = rciDF_kk[list(tempD.keys())]\n",
    "rciDF_kk = rciDF_kk.rename(columns=tempD)\n",
    "display(rciDF_kk)\n",
    "display(rciDF_kk.describe())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# — †1. Go to the top of the R sub-notebook —  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3-2. Dunnett's test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Import the summary table\n",
    "fileDir = './ExportData/'\n",
    "ipynbName = '220522_LC-M001-related-transcriptomics-DIRAC_StatisticalTest-GOBP_ver2-4_'\n",
    "fileName = 'inter-group-comparison.xlsx'\n",
    "sheetName = 'RCImean'\n",
    "tempDF = pd.read_excel(fileDir+ipynbName+fileName, sheet_name=sheetName, engine='openpyxl')\n",
    "tempDF = tempDF.set_index('ModuleID')\n",
    "display(tempDF)\n",
    "\n",
    "statDF = tempDF"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3-3. Visualization: boxplot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#Prepare DF and color\n",
    "tempDF = rciDF_kk.reset_index().melt(var_name='Group', value_name='RCI', id_vars='ModuleID')\n",
    "tempD = {'Control':'tab:blue', 'Acarbose':'tab:red',\n",
    "         'Rapamycin':'tab:purple', 'CR diet':'tab:olive'}\n",
    "\n",
    "#Prepare significance labels\n",
    "##Retrieve statistical significance\n",
    "module = 'All'\n",
    "tempS = statDF.loc[module, statDF.columns.str.contains('AdjPval')]\n",
    "tempS.index = tempS.index.str.replace('_AdjPval', '')\n",
    "tempS.name = 'AdjPval'\n",
    "##Clean\n",
    "tempDF1 = tempS.index.to_series().str.split(pat='-vs-', expand=True)\n",
    "tempDF1 = tempDF1.rename(columns={0:'Contrast', 1:'Baseline'})\n",
    "tempDF1 = pd.merge(tempDF1, tempS, left_index=True, right_index=True, how='left')\n",
    "tempD0 = {'Cont':'Control', 'Acar':'Acarbose',\n",
    "          'Rapa':'Rapamycin', 'Calo':'CR diet'}\n",
    "tempDF1['Contrast'] = tempDF1['Contrast'].map(tempD0)\n",
    "tempDF1['Baseline'] = tempDF1['Baseline'].map(tempD0)\n",
    "##Convert p-value to label\n",
    "tempL = []\n",
    "for row_i in range(len(tempDF1)):\n",
    "    pval = tempDF1['AdjPval'].iloc[row_i]\n",
    "    if pval<0.001:\n",
    "        tempL.append('***')\n",
    "    elif pval<0.01:\n",
    "        tempL.append('**')\n",
    "    elif pval<0.05:\n",
    "        tempL.append('*')\n",
    "    else:\n",
    "        pval_text = Decimal(str(pval)).quantize(Decimal('0.01'), rounding=ROUND_HALF_UP)\n",
    "        tempL.append(r'$P$ = '+str(pval_text))\n",
    "tempDF1['SignifLabel'] = tempL\n",
    "display(tempDF1)\n",
    "\n",
    "#Visualization\n",
    "ymax = 1.0\n",
    "ymin = 0.5\n",
    "yinter = 0.1\n",
    "ymargin_t = 0.06\n",
    "ymargin_b = 0.01\n",
    "aline_ymin = 0.9\n",
    "aline_ymargin = 0.05\n",
    "sns.set(style='ticks', font='Arial', context='talk')\n",
    "plt.figure(figsize=(2, 4))\n",
    "p = sns.boxplot(data=tempDF, y='RCI', x='Group', order=list(tempD.keys()), palette=tempD, dodge=False,\n",
    "                showfliers=True, flierprops={'marker':'o', 'markerfacecolor':'gray', 'alpha':0.4},\n",
    "                showcaps=True, notch=True)\n",
    "p.set(ylim=(ymin-ymargin_b, ymax+ymargin_t), yticks=np.arange(ymin, ymax + yinter/10, yinter))\n",
    "##Add significance labels\n",
    "lines = p.axes.get_lines()#Line2D: [[Q1, Q1-1.5IQR], [Q3, Q3+1.5IQR], [Q1, Q1], [Q3, Q3], [Med, Med], [flier]]\n",
    "lines_unit = 5 + int(True)#showfliers=True\n",
    "for row_i in range(len(tempDF1)):\n",
    "    #Baseline\n",
    "    group_0 = tempDF1['Baseline'].iloc[row_i]\n",
    "    index_0 = list(tempD.keys()).index(group_0)\n",
    "    whisker_0 = lines[index_0*lines_unit + 1]\n",
    "    xcoord_0 = whisker_0._x[1]#Q3+1.5IQR\n",
    "    #ycoord_0 = whisker_0._y[1]#Q3+1.5IQR\n",
    "    #Contrast\n",
    "    group_1 = tempDF1['Contrast'].iloc[row_i]\n",
    "    index_1 = list(tempD.keys()).index(group_1)\n",
    "    whisker_1 = lines[index_1*lines_unit + 1]\n",
    "    xcoord_1 = whisker_1._x[1]#Q3+1.5IQR\n",
    "    #ycoord_1 = whisker_1._y[1]#Q3+1.5IQR\n",
    "    #Standard point of marker\n",
    "    xcoord = (xcoord_0+xcoord_1)/2\n",
    "    #ycoord = max(ycoord_0, ycoord_1)\n",
    "    ycoord = aline_ymin + aline_ymargin*row_i\n",
    "    label = tempDF1['SignifLabel'].iloc[row_i]\n",
    "    #Add annotation lines\n",
    "    aline_offset = yinter/10\n",
    "    aline_length = yinter/10 + aline_offset\n",
    "    plt.plot([xcoord_0, xcoord_0, xcoord_1, xcoord_1],\n",
    "             [ycoord+aline_offset, ycoord+aline_length, ycoord+aline_length, ycoord+aline_offset],\n",
    "             lw=1.5, c='k')\n",
    "    #Add annotation text\n",
    "    if label in ['***', '**', '*']:\n",
    "        text_offset = yinter/25\n",
    "        p.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                   horizontalalignment='center', verticalalignment='bottom',\n",
    "                   fontsize='medium', color='k')\n",
    "    else:\n",
    "        text_offset = yinter/5\n",
    "        p.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                   horizontalalignment='center', verticalalignment='bottom',\n",
    "                   fontsize='x-small', color='k')\n",
    "sns.despine()\n",
    "plt.xlabel('')\n",
    "plt.ylabel('Module RCI')\n",
    "plt.xticks(rotation=70, horizontalalignment='right', verticalalignment='center', rotation_mode='anchor')\n",
    "##Save\n",
    "fileDir = './ExportFigures/'\n",
    "ipynbName = '220522_LC-M001-related-transcriptomics-DIRAC_DIRAC-GOBP_ver2-4_'\n",
    "fileName = 'RCI-boxplot.tif'\n",
    "plt.gcf().savefig(fileDir+ipynbName+fileName, dpi=300, bbox_inches='tight', pad_inches=0.04,\n",
    "                  pil_kwargs={'compression':'tiff_lzw'})\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4. Rank conservation index: inter-group module comparison"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> Test specific hypothesis: control RCI == intervention RCI (i.e., inter-group module comparison).  \n",
    "> 1. Testing the main effect of intervention on rank mathing scores (RMSs) for each module using ANOVA model  \n",
    "> 2. Then, performing post-hoc comparisons of RMSs between control vs. each intervention using Dunnett's test  \n",
    ">  \n",
    "> Since RCI is the mean of RMSs under the own phenotype consensus, not RCI but RMS is practically used to obtain variance around group consensus. As well as using sex-pooled rank consensus, sex is NOT included in the ANOVA model, which is a change from version 2-2. Note that ANOVA can shrink the variance utilizing all samples (per module), whose statistical power is better than the repeated Welch's t-tests (version 2-3) in the case of small sample size. Although tricky, the P-value adjustment in (1) is performed across modules under the assumption that modules are independent, which would be more conservative and less likely raise referees' eyebrows for venn diagram-type summary than using nominal P-value cutoff. Because the post-hoc comparisons (2) are to address the effect of each intervention within a specific module, the P-values are adjusted across interventions only within the module (not across modules) using Dunnett's test.  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 4-1. Extract RMS under the own phenotype consensus"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Extract RMS whose template phenotype corresponds to the own phenotype\n",
    "phenotypeL = rciDF.drop(columns=['ModuleID', 'Template']).columns.tolist()\n",
    "rmsDF_kk = pd.DataFrame(index=pd.Index(rmsDF['ModuleID'].unique(), name='ModuleID'))\n",
    "tempDF = rmsDF.set_index('ModuleID')\n",
    "for k in phenotypeL:\n",
    "    tempL = sampleDF.loc[sampleDF['Phenotype']==k].index.tolist()\n",
    "    tempDF1 = tempDF[tempL].loc[tempDF['Template']==k]\n",
    "    rmsDF_kk = pd.merge(rmsDF_kk, tempDF1, left_index=True, right_index=True, how='left')\n",
    "\n",
    "display(rmsDF_kk)\n",
    "display(rmsDF_kk.describe())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# — †2. Go to †1 of the R sub-notebook —  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 4-2. ANOVA test (RMS ~ Intervention), followed by Dunnett's test (Intervention)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 4-2-1/2. Import the summary table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Import the summary table\n",
    "fileDir = './ExportData/'\n",
    "ipynbName = '220522_LC-M001-related-transcriptomics-DIRAC_StatisticalTest-GOBP_ver2-4_'\n",
    "fileName = 'inter-group-comparison.xlsx'\n",
    "sheetName = 'RCI'\n",
    "tempDF = pd.read_excel(fileDir+ipynbName+fileName, sheet_name=sheetName, engine='openpyxl')\n",
    "tempDF = tempDF.set_index('ModuleID')\n",
    "display(tempDF)\n",
    "\n",
    "statDF = tempDF"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 4-2-3. Changed modules (ANOVA)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Prepare variables in the model\n",
    "tempS = statDF.loc[:, statDF.columns.str.contains('_Fstat')].columns.to_series()\n",
    "variableL = tempS.str.replace('_Fstat', '').tolist()\n",
    "\n",
    "#Changed modules\n",
    "for variable in variableL:\n",
    "    tempDF = statDF.loc[statDF[variable+'_AdjPval']<0.05]\n",
    "    tempDF = tempDF.sort_values(by=variable+'_AdjPval', ascending=True)\n",
    "    tempL1 = tempDF.loc[:, tempDF.columns.str.contains('_RMSmean')].columns.tolist()\n",
    "    tempL2 = tempDF.loc[:, tempDF.columns.str.contains('^'+variable+'_')].columns.tolist()\n",
    "    tempDF = tempDF[[col_n for subL in [['ModuleName'], tempL1, tempL2] for col_n in subL]]\n",
    "    print(variable+' (adjusted P < 0.05):', len(tempDF))\n",
    "    display(tempDF)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 4-2-4. Changed modules by each intervention (Dunnett's test)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Extract only the changed modules\n",
    "variable = 'Intervention'\n",
    "tempDF = statDF.loc[statDF[variable+'_AdjPval']<0.05]\n",
    "tempDF = tempDF.sort_values(by=variable+'_AdjPval', ascending=True)\n",
    "print(variable+' (adjusted P < 0.05):', len(tempDF))\n",
    "\n",
    "#Take adjusted P-value\n",
    "tempDF1 = tempDF.loc[:, tempDF.columns.str.contains('-vs-.*_AdjPval$')]\n",
    "tempDF1.columns = tempDF1.columns.str.replace('_AdjPval$', '')\n",
    "tempDF1 = pd.merge(tempDF[['ModuleName', variable+'_AdjPval']], tempDF1,\n",
    "                   left_index=True, right_index=True, how='right')\n",
    "print('Adjusted P-value:')\n",
    "display(tempDF1)\n",
    "display(tempDF1.describe())\n",
    "\n",
    "#Take effect size\n",
    "tempDF2 = tempDF.loc[:, tempDF.columns.str.contains('-vs-.*_Coef$')]\n",
    "tempDF2.columns = tempDF2.columns.str.replace('_Coef$', '')\n",
    "tempDF2 = pd.merge(tempDF[['ModuleName', variable+'_AdjPval']], tempDF2,\n",
    "                   left_index=True, right_index=True, how='right')\n",
    "print('Changed direction (effect size):')\n",
    "display(tempDF2)\n",
    "display(tempDF2.describe())\n",
    "\n",
    "pvalDF = tempDF1\n",
    "diffDF = tempDF2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> Check the changed modules (nominal P-value < 0.05) as reference.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Extract only the changed modules\n",
    "variable = 'Intervention'\n",
    "tempDF = statDF.loc[statDF[variable+'_Pval']<0.05]\n",
    "tempDF = tempDF.sort_values(by=variable+'_Pval', ascending=True)\n",
    "print(variable+' (nominal P < 0.05):', len(tempDF))\n",
    "\n",
    "#Take adjusted P-value\n",
    "tempDF1 = tempDF.loc[:, tempDF.columns.str.contains('-vs-.*_AdjPval$')]\n",
    "tempDF1.columns = tempDF1.columns.str.replace('_AdjPval$', '')\n",
    "tempDF1 = pd.merge(tempDF[['ModuleName', variable+'_AdjPval']], tempDF1,\n",
    "                   left_index=True, right_index=True, how='right')\n",
    "print('Adjusted P-value:')\n",
    "display(tempDF1)\n",
    "display(tempDF1.describe())\n",
    "\n",
    "#Take effect size\n",
    "tempDF2 = tempDF.loc[:, tempDF.columns.str.contains('-vs-.*_Coef$')]\n",
    "tempDF2.columns = tempDF2.columns.str.replace('_Coef$', '')\n",
    "tempDF2 = pd.merge(tempDF[['ModuleName', variable+'_AdjPval']], tempDF2,\n",
    "                   left_index=True, right_index=True, how='right')\n",
    "#print('Changed direction (effect size):')\n",
    "#display(tempDF2)\n",
    "#display(tempDF2.describe())\n",
    "\n",
    "pvalDF_ref = tempDF1\n",
    "diffDF_ref = tempDF2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "### 4-3. Visualization: clustermap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Prepare color labels for tightened module set\n",
    "regulation = 'Tightened'\n",
    "tempD0 = {'Cont':'Control', 'Acar':'Acarbose',\n",
    "          'Rapa':'Rapamycin', 'Calo':'CR diet'}\n",
    "tempD1 = {'Acarbose':'tab:red',\n",
    "          'Rapamycin':'tab:purple', 'CR diet':'tab:olive'}\n",
    "tempD2 = {'Acarbose':plt.get_cmap('tab20')(7),\n",
    "          'Rapamycin':plt.get_cmap('tab20')(9),\n",
    "          'CR diet':plt.get_cmap('tab20')(17)}\n",
    "tempDF = pvalDF_ref.loc[:, pvalDF_ref.columns.str.contains('-vs-')]#Nominal P-value\n",
    "for col_n in tempDF.columns.tolist():\n",
    "    tempS1 = pvalDF_ref[col_n]\n",
    "    tempS2 = diffDF_ref[col_n]\n",
    "    if regulation=='Changed':\n",
    "        tempS2 = tempS2.loc[(tempS1<0.05)]\n",
    "    elif regulation=='Tightened':\n",
    "        tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2>0)]\n",
    "    elif regulation=='Loosened':\n",
    "        tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2<0)]\n",
    "    label = tempD0[re.sub('-vs-.*', '', col_n)]\n",
    "    tempL = []\n",
    "    count = 0#just for checking\n",
    "    for module in tempDF.index.tolist():\n",
    "        if module in tempS2.index.tolist():\n",
    "            if module in pvalDF.index.tolist():#Adjusted P-value\n",
    "                tempL.append(tempD1[label])\n",
    "                count += 1#Just for checking\n",
    "            else:#Nominal P-value\n",
    "                tempL.append(tempD2[label])\n",
    "        else:\n",
    "            tempL.append('white')\n",
    "    tempDF[label] = tempL\n",
    "    print(regulation+' module in '+col_n)\n",
    "    print(' -> adjusted P < 0.05:', count)\n",
    "    print(' -> nominal P < 0.05:', len(tempS2))\n",
    "tempDF = tempDF[list(tempD1.keys())]\n",
    "\n",
    "#Clustermap\n",
    "sns.set(style='ticks', font='Arial', context='talk')\n",
    "cm = sns.clustermap(rciDF_kk.T, method='ward', metric='euclidean', cmap='afmhot',\n",
    "                    row_cluster=True, col_cluster=True, row_linkage=None, col_linkage=None,\n",
    "                    row_colors=None, col_colors=tempDF, xticklabels=False, yticklabels=True,\n",
    "                    dendrogram_ratio=(0.025, 0.2), colors_ratio=(0.01, 0.1),\n",
    "                    cbar_pos=(0.05, -0.15, 0.3, 0.075), cbar_kws={'orientation': 'horizontal'},\n",
    "                    figsize=(12, 3), **{'vmin':0.5, 'vmax':1})\n",
    "cm.cax.set_title('Module RCI', size='medium',\n",
    "                 verticalalignment='bottom', horizontalalignment='center')\n",
    "cm.cax.tick_params(labelsize='small')\n",
    "bottom, top = cm.ax_heatmap.get_ylim()\n",
    "#cm.ax_heatmap.set_ylim(bottom + 0.5, top - 0.5)##To avoid half cut of first and last rows\n",
    "hm = cm.ax_heatmap.get_position()\n",
    "rd = cm.ax_row_dendrogram.get_position()\n",
    "cd = cm.ax_col_dendrogram.get_position()\n",
    "cm.ax_heatmap.set_position([hm.x0, hm.y0, hm.width, hm.height])\n",
    "cm.ax_row_dendrogram.set_position([rd.x0, rd.y0, rd.width, rd.height])\n",
    "cm.ax_col_dendrogram.set_position([cd.x0, cd.y0, cd.width, cd.height])\n",
    "cm.ax_heatmap.set_xlabel('GOBP module')\n",
    "cm.ax_heatmap.set_ylabel('')\n",
    "##row/column color bar legend (axis is same with cm.cax!)\n",
    "tempL = []\n",
    "for group in tempD1.keys():\n",
    "    tempL.append(mpatches.Patch(color=tempD1[group],\n",
    "                                label='by '+group+' (adjusted '+r'$P$'+' < 0.05)'))\n",
    "for group in tempD2.keys():\n",
    "    tempL.append(mpatches.Patch(color=tempD2[group], label='(nominal '+r'$P$'+' < 0.05)'))\n",
    "legend1 = plt.legend(handles=tempL, fontsize='small', labelspacing=0.2, ncol=2,\n",
    "                     title='Tightened module (vs. Control)', title_fontsize='medium',\n",
    "                     bbox_to_anchor=(1, 0.5), loc='center left', borderaxespad=3.5, frameon=False)\n",
    "plt.gca().add_artist(legend1)\n",
    "##Save\n",
    "fileDir = './ExportFigures/'\n",
    "ipynbName = '220522_LC-M001-related-transcriptomics-DIRAC_DIRAC-GOBP_ver2-4_'\n",
    "fileName = 'RCI-clustermap.tif'\n",
    "plt.gcf().savefig(fileDir+ipynbName+fileName, dpi=300, bbox_inches='tight', pad_inches=0.04,\n",
    "                  pil_kwargs={'compression':'tiff_lzw'})\n",
    "plt.show()\n",
    "\n",
    "#Save label order\n",
    "tempDF = moduleDF.loc[rciDF_kk.index[cm.dendrogram_col.reordered_ind]]\n",
    "tempD = {'Control':'Cont', 'Acarbose':'Acar',\n",
    "         'Rapa':'Rapamycin', 'CR diet':'Calo'}\n",
    "tempDF1 = rciDF_kk.copy()\n",
    "tempDF1.columns = tempDF1.columns.map(tempD)\n",
    "tempDF = pd.merge(tempDF['ModuleName'], tempDF1, left_index=True, right_index=True, how='left')\n",
    "tempDF1 = pvalDF.loc[:, pvalDF.columns.str.contains('-vs-')]#Adjusted P-value\n",
    "tempDF = pd.merge(tempDF, tempDF1, left_index=True, right_index=True, how='left')\n",
    "tempDF = tempDF.reset_index()\n",
    "tempDF.index.name = 'Xcoord'\n",
    "display(tempDF)\n",
    "fileDir = './ExportData/'\n",
    "ipynbName = '220522_LC-M001-related-transcriptomics-DIRAC_DIRAC-GOBP_ver2-4_'\n",
    "fileName = 'RCI-clustermap-xticks.tsv'\n",
    "tempDF.to_csv(fileDir+ipynbName+fileName, index=True, sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 4-4. Visualization: venn diagram"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Prepare label and color\n",
    "tempD0 = {'Cont':'Control', 'Acar':'Acarbose',\n",
    "          'Rapa':'Rapamycin', 'Calo':'CR diet'}\n",
    "tempD1 = {'Acarbose':'tab:red',\n",
    "          'Rapamycin':'tab:purple', 'CR diet':'tab:olive'}\n",
    "\n",
    "#Visualization per direction\n",
    "for regulation in ['Changed', 'Tightened', 'Loosened']:\n",
    "    #Prepare module sets\n",
    "    tempD2 = {}\n",
    "    tempDF = pvalDF.loc[:, pvalDF.columns.str.contains('-vs-')]#Adjusted P-value\n",
    "    for col_n in tempDF.columns.tolist():\n",
    "        tempS1 = pvalDF[col_n]\n",
    "        tempS2 = diffDF[col_n]\n",
    "        if regulation=='Changed':\n",
    "            tempS2 = tempS2.loc[(tempS1<0.05)]\n",
    "        elif regulation=='Tightened':\n",
    "            tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2>0)]\n",
    "        elif regulation=='Loosened':\n",
    "            tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2<0)]\n",
    "        label = tempD0[re.sub('-vs-.*', '', col_n)]\n",
    "        tempD2[label] = set(tempS2.index.tolist())\n",
    "    ##Sort to make consistent order in manual legend generation\n",
    "    tempD = {}\n",
    "    for label in tempD1.keys():\n",
    "        tempD[label] = tempD2[label]\n",
    "    \n",
    "    #Not significant in all contrasts\n",
    "    tempDF = pvalDF.copy()\n",
    "    for moduleS in tempD.values():\n",
    "        tempDF = tempDF.loc[~tempDF.index.isin(moduleS)]\n",
    "    count = len(tempDF)\n",
    "    print(regulation+' modules (vs. Control):')\n",
    "    print(' -> Not significant in all contrasts:', count)\n",
    "    \n",
    "    #Skip the followings if no significant module\n",
    "    if count==len(pvalDF):\n",
    "        continue\n",
    "    \n",
    "    #Venn diagram\n",
    "    sns.set(style='ticks', font='Arial', context='talk')\n",
    "    fig, ax = plt.subplots(figsize=(5, 5))\n",
    "    venn(tempD, fmt='{size:,}', cmap=list(tempD1.values()), legend_loc=None, ax=ax)\n",
    "    plt.setp(ax, ylim=(0.0, 0.95))#Otherwise, weird space...\n",
    "    ##Add legend annotation\n",
    "    x_coord = [0.1, 0.9, 0.8]\n",
    "    y_coord = [0.8, 0.8, 0.25]\n",
    "    h_align = ['right', 'left', 'left']\n",
    "    v_align = ['bottom', 'bottom', 'top']\n",
    "    for i in range(len(tempD1)):\n",
    "        key = list(tempD1.keys())[i]\n",
    "        total = f'{len(tempD[key]):,}'\n",
    "        ax.text(x_coord[i], y_coord[i], key+'\\n('+total+' modules)',\n",
    "                fontsize='small', multialignment='center',\n",
    "                horizontalalignment=h_align[i], verticalalignment=v_align[i],\n",
    "                bbox={'boxstyle':'round', 'facecolor':tempD1[key], 'pad':0.2, 'alpha':0.5})\n",
    "    title = regulation+' modules (vs. Control)'\n",
    "    ax.set_title(title, fontsize='medium')\n",
    "    ##Save\n",
    "    if regulation!='Changed':\n",
    "        fileDir = './ExportFigures/'\n",
    "        ipynbName = '220522_LC-M001-related-transcriptomics-DIRAC_DIRAC-GOBP_ver2-4_'\n",
    "        fileName = 'RCI-inter-group-comparison_venn-'+regulation.lower()+'.tif'\n",
    "        plt.gcf().savefig(fileDir+ipynbName+fileName, dpi=300, bbox_inches='tight', pad_inches=0.04,\n",
    "                          pil_kwargs={'compression':'tiff_lzw'})\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Export module list in each subset in the venn diagram\n",
    "for regulation in ['Tightened', 'Loosened']:\n",
    "    #Prepare module sets\n",
    "    tempD = {}\n",
    "    tempDF = pvalDF.loc[:, pvalDF.columns.str.contains('-vs-')]#Adjusted P-value\n",
    "    for col_n in tempDF.columns.tolist():\n",
    "        tempS1 = pvalDF[col_n]\n",
    "        tempS2 = diffDF[col_n]\n",
    "        if regulation=='Changed':\n",
    "            tempS2 = tempS2.loc[(tempS1<0.05)]\n",
    "        elif regulation=='Tightened':\n",
    "            tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2>0)]\n",
    "        elif regulation=='Loosened':\n",
    "            tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2<0)]\n",
    "        tempD[col_n] = set(tempS2.index.tolist())\n",
    "    \n",
    "    #Not significant in all contrasts\n",
    "    tempDF = pvalDF.copy()\n",
    "    for moduleS in tempD.values():\n",
    "        tempDF = tempDF.loc[~tempDF.index.isin(moduleS)]\n",
    "    count = len(tempDF)\n",
    "    print(regulation+' modules (vs. Control):')\n",
    "    print(' -> Not significant in all contrasts:', count)\n",
    "    \n",
    "    #Skip the followings if no significant module\n",
    "    if count==len(pvalDF):\n",
    "        continue\n",
    "    \n",
    "    #Prepare a new .xlsx file (dummy README)\n",
    "    tempL1 = [len(tempD[key]) for key in tempD.keys()]\n",
    "    tempDF = pd.DataFrame({'Group':tempD.keys(), 'nModules':tempL1})\n",
    "    tempDF = tempDF.reset_index().rename(columns={'index':'VennOrder'})\n",
    "    tempDF['VennOrder'] = tempDF['VennOrder'] + 1\n",
    "    fileDir = './ExportData/'\n",
    "    ipynbName = '220522_LC-M001-related-transcriptomics-DIRAC_DIRAC-GOBP_ver2-4_'\n",
    "    fileName = 'RCI-inter-group-comparison_venn-'+regulation.lower()+'.xlsx'\n",
    "    tempDF.to_excel(fileDir+ipynbName+fileName, sheet_name='README', header=True, index=False)\n",
    "    display(tempDF)#Check\n",
    "    \n",
    "    t_start = time.time()\n",
    "    #Extract overall set\n",
    "    for key_i in range(len(tempD)):\n",
    "        key = list(tempD.keys())[key_i]\n",
    "        tempS = tempD[key]\n",
    "        tempDF = statDF.loc[statDF.index.isin(tempS)]\n",
    "        #Save summary table by appended to the above .xlsx file\n",
    "        ##Prepare sheet name\n",
    "        tempL1 = ['NA' for i in range(len(tempD))]\n",
    "        tempL1[key_i] = '1'\n",
    "        setName = '('+','.join(tempL1)+')'\n",
    "        with pd.ExcelWriter(fileDir+ipynbName+fileName, mode='a', engine='openpyxl') as writer:\n",
    "            tempDF.to_excel(writer, sheet_name=setName, header=True, index=True)\n",
    "        print(' - '+setName+':', len(tempDF))\n",
    "    \n",
    "    #Extract subset\n",
    "    tempL1 = ['1', '0']\n",
    "    tempL2 = [[k1, k2, k3] for k1 in tempL1 for k2 in tempL1 for k3 in tempL1]\n",
    "    #tempL2.remove(['0', '0', '0'])\n",
    "    for tempL1 in tempL2:\n",
    "        #Positive module set\n",
    "        tempL3 = [list(tempD.values())[key_i] for key_i, binary in enumerate(tempL1) if binary=='1']\n",
    "        tempS1 = set(pvalDF.index.tolist())#Initialize\n",
    "        for tempS in tempL3:\n",
    "            tempS1 = tempS1 & tempS\n",
    "        #Negative module set\n",
    "        tempL3 = [list(tempD.values())[key_i] for key_i, binary in enumerate(tempL1) if binary=='0']\n",
    "        tempS2 = set()#Initialize\n",
    "        for tempS in tempL3:\n",
    "            tempS2 = tempS2 | tempS\n",
    "        #Extract subset\n",
    "        tempS = tempS1 - tempS2\n",
    "        tempDF = statDF.loc[statDF.index.isin(tempS)]\n",
    "        #Save summary table by appended to the above .xlsx file\n",
    "        ##Prepare sheet name\n",
    "        setName = '('+','.join(tempL1)+')'\n",
    "        with pd.ExcelWriter(fileDir+ipynbName+fileName, mode='a', engine='openpyxl') as writer:\n",
    "            tempDF.to_excel(writer, sheet_name=setName, header=True, index=True)\n",
    "        print(' - '+setName+':', len(tempDF))\n",
    "    \n",
    "    t_elapsed = time.time() - t_start\n",
    "    print(' - Elapsed time:', round(t_elapsed//60), 'min', round(t_elapsed%60, 1), 'sec')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 4-5. Visualization: pointplot"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 4-5-1. Modules tightened by all interventions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Prepare the target module set\n",
    "posL = ['Acar-vs-Cont', 'Rapa-vs-Cont', 'Calo-vs-Cont']\n",
    "negL = ['']\n",
    "regulation = 'Tightened'\n",
    "tempS = pd.Series(np.repeat(True, len(pvalDF)), index=pvalDF.index)#Initialize\n",
    "tempDF = pvalDF.loc[:, pvalDF.columns.str.contains('-vs-')]#Adjusted P-value\n",
    "for col_n in tempDF.columns.tolist():\n",
    "    tempS1 = pvalDF[col_n]\n",
    "    tempS2 = diffDF[col_n]\n",
    "    if col_n in posL:\n",
    "        if regulation=='Tightened':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2>0)\n",
    "        elif regulation=='Loosened':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2<0)\n",
    "    elif col_n in negL:\n",
    "        tempS3 = (tempS1>=0.05)\n",
    "        #Significance for inverse regulation\n",
    "        if regulation=='Tightened':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2<0)\n",
    "        elif regulation=='Loosened':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2>0)\n",
    "        tempS1 = tempS3 | tempS1\n",
    "    else:\n",
    "        tempS1 = (tempS1>=0.0)\n",
    "    #Update True\n",
    "    tempS = tempS & tempS1\n",
    "tempL = tempS.loc[tempS.tolist()].index.tolist()\n",
    "print(len(tempL), regulation.lower()+' modules with significance in', posL, 'but not in', negL)\n",
    "\n",
    "#Select representatives\n",
    "topX = np.min([30, len(tempL)])\n",
    "topX_plot = np.min([10, len(tempL)])\n",
    "tempDF = statDF.loc[:, statDF.columns.str.contains('Pval$')]\n",
    "tempDF = pd.merge(statDF['ModuleName'], tempDF, left_index=True, right_index=True, how='left')\n",
    "tempDF = tempDF.loc[tempL].sort_values(by=variable+'_AdjPval', ascending=True)\n",
    "print('Top', topX, 'modules (sort by the main effect of '+variable+'):')\n",
    "display(tempDF.iloc[:topX])\n",
    "plotL = tempDF.index.tolist()[:topX_plot]\n",
    "\n",
    "#Prepare DF for plot\n",
    "tempDF = rmsDF_kk.reset_index().melt(var_name='SampleID', value_name='RMS', id_vars='ModuleID')\n",
    "tempDF1 = sampleDF.reset_index()[['SampleID', 'Phenotype']]\n",
    "tempDF = pd.merge(tempDF, tempDF1, on='SampleID', how='left')\n",
    "\n",
    "#Prepare label and color\n",
    "tempD = {'Control':'Control', 'Acarbose':'Acarbose',\n",
    "         'Rapamycin':'Rapamycin', 'Calorie restriction':'CR diet'}\n",
    "tempDF['Group'] = tempDF['Phenotype'].map(tempD)\n",
    "tempD0 = {'Cont':'Control', 'Acar':'Acarbose',\n",
    "          'Rapa':'Rapamycin', 'Calo':'CR diet'}\n",
    "tempD1 = {'Control':'tab:blue', 'Acarbose':'tab:red',\n",
    "          'Rapamycin':'tab:purple', 'CR diet':'tab:olive'}\n",
    "\n",
    "#Visualize each representative\n",
    "for rank_i in range(len(plotL)):\n",
    "    print(' - Rank '+str(rank_i+1)+' (sort by the main effect of '+variable+'):')\n",
    "    module = plotL[rank_i]\n",
    "    #Check module summary\n",
    "    tempDF1 = pd.DataFrame(moduleDF.loc[module]).T\n",
    "    display(tempDF1)\n",
    "    \n",
    "    #Select RMS\n",
    "    tempDF1 = tempDF.loc[tempDF['ModuleID']==module]\n",
    "    \n",
    "    #Check RMS summary\n",
    "    tempDF2 = tempDF1.groupby(['Group'])['RMS'].agg(['count', 'mean', 'std'])\n",
    "    tempL1 = []\n",
    "    tempL2 = []\n",
    "    for row_n in tempDF2.index.tolist():\n",
    "        count, mean, std = tempDF2.loc[row_n]\n",
    "        tempL1.append(mean - 1.96*std/np.sqrt(count))\n",
    "        tempL2.append(mean + 1.96*std/np.sqrt(count))\n",
    "    tempDF2['0.025'] = tempL1\n",
    "    tempDF2['0.975'] = tempL2\n",
    "    tempDF2 = tempDF2.loc[list(tempD1.keys())]#Sort\n",
    "    display(tempDF2)\n",
    "    \n",
    "    #Prepare significance labels\n",
    "    ##Retrieve statistical significance\n",
    "    tempS = pvalDF.loc[module, pvalDF.columns.str.contains('-vs-')]#Adjusted P-value\n",
    "    tempS.name = 'AdjPval'\n",
    "    ##Clean\n",
    "    tempDF2 = tempS.index.to_series().str.split(pat='-vs-', expand=True)\n",
    "    tempDF2 = tempDF2.rename(columns={0:'Contrast', 1:'Baseline'})\n",
    "    tempDF2 = pd.merge(tempDF2, tempS, left_index=True, right_index=True, how='left')\n",
    "    tempDF2['Contrast'] = tempDF2['Contrast'].map(tempD0)\n",
    "    tempDF2['Baseline'] = tempDF2['Baseline'].map(tempD0)\n",
    "    ##Convert p-value to label\n",
    "    tempL = []\n",
    "    for row_i in range(len(tempDF2)):\n",
    "        pval = tempDF2['AdjPval'].iloc[row_i]\n",
    "        if pval<0.001:\n",
    "            tempL.append('***')\n",
    "        elif pval<0.01:\n",
    "            tempL.append('**')\n",
    "        elif pval<0.05:\n",
    "            tempL.append('*')\n",
    "        else:\n",
    "            pval_text = Decimal(str(pval)).quantize(Decimal('0.01'), rounding=ROUND_HALF_UP)\n",
    "            tempL.append(r'$P$ = '+str(pval_text))\n",
    "    tempDF2['SignifLabel'] = tempL\n",
    "    display(tempDF2)\n",
    "    \n",
    "    #Visualization\n",
    "    ymax = 1.0\n",
    "    ymin = 0.0\n",
    "    yinter = 0.2\n",
    "    ymargin_t = 0.4\n",
    "    ymargin_b = 0.05\n",
    "    aline_ymin = 1.0\n",
    "    aline_ymargin = 0.125\n",
    "    sns.set(style='ticks', font='Arial', context='talk')\n",
    "    plt.figure(figsize=(2, 4))\n",
    "    sns.pointplot(data=tempDF1, x='Group', y='RMS', order=list(tempD1.keys()), palette=tempD1,\n",
    "                  markers='o', dodge=False, join=False, capsize=0.6, estimator=np.mean, ci=95)\n",
    "    p = sns.stripplot(data=tempDF1, x='Group', y='RMS',\n",
    "                      order=list(tempD1.keys()), palette=tempD1, dodge=False, jitter=0.15,\n",
    "                      size=5, edgecolor='black', linewidth=1, **{'marker':'o', 'alpha':0.5})\n",
    "    ##Set axis\n",
    "    sns.despine()\n",
    "    p.set(ylim=(ymin-ymargin_b, ymax+ymargin_t), yticks=np.arange(ymin, ymax + yinter/10, yinter))\n",
    "    plt.setp(p.get_xticklabels(), rotation=70, horizontalalignment='right',\n",
    "             verticalalignment='center', rotation_mode='anchor')\n",
    "    ##Add significance labels\n",
    "    for row_i in range(len(tempDF2)):\n",
    "        #Baseline\n",
    "        group_0 = tempDF2['Baseline'].iloc[row_i]\n",
    "        index_0 = list(tempD1.keys()).index(group_0)\n",
    "        xcoord_0 = index_0\n",
    "        #Contrast\n",
    "        group_1 = tempDF2['Contrast'].iloc[row_i]\n",
    "        index_1 = list(tempD1.keys()).index(group_1)\n",
    "        xcoord_1 = index_1\n",
    "        #Standard point of marker\n",
    "        xcoord = (xcoord_0+xcoord_1)/2\n",
    "        ycoord = aline_ymin + aline_ymargin*row_i\n",
    "        label = tempDF2['SignifLabel'].iloc[row_i]\n",
    "        #Add annotation lines\n",
    "        aline_offset = yinter/5\n",
    "        aline_length = yinter/5 + aline_offset/2\n",
    "        plt.plot([xcoord_0, xcoord_0, xcoord_1, xcoord_1],\n",
    "                 [ycoord+aline_offset, ycoord+aline_length, ycoord+aline_length, ycoord+aline_offset],\n",
    "                 lw=1.5, c='k')\n",
    "        #Add annotation text\n",
    "        if label in ['***', '**', '*']:\n",
    "            text_offset = yinter/21\n",
    "            p.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                       horizontalalignment='center', verticalalignment='bottom',\n",
    "                       fontsize='medium', color='k')\n",
    "        else:\n",
    "            text_offset = yinter/3.5\n",
    "            p.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                       horizontalalignment='center', verticalalignment='bottom',\n",
    "                       fontsize='x-small', color='k')\n",
    "    ##Add annotation\n",
    "    p.set_title('Consensus:\\nown group', {'fontsize':'small'})\n",
    "    xoff = 0.015\n",
    "    yoff = 0.01\n",
    "    rect = plt.Rectangle((xoff, 1), 1-xoff, 0.155,#Manual adjustment\n",
    "                         transform=p.axes.transAxes, facecolor=plt.get_cmap('tab20')(15),\n",
    "                         clip_on=False, linewidth=0, zorder=0.5)\n",
    "    p.add_patch(rect)\n",
    "    ##Set axis label and title\n",
    "    plt.setp(p, xlabel='', ylabel='Sample RMS\\n(Mean = Module RCI)')\n",
    "    plt.suptitle(module, size='small',\n",
    "                 verticalalignment='bottom', horizontalalignment='center', wrap=True, y=1.0)\n",
    "    ##Save\n",
    "    fileDir = './ExportFigures/'\n",
    "    ipynbName = '220522_LC-M001-related-transcriptomics-DIRAC_DIRAC-GOBP_ver2-4_'\n",
    "    fileName = 'RCI-inter-group-comparison_RMS-pointplot-'+module.replace('GO:', 'GO')+'.tif'\n",
    "    plt.gcf().savefig(fileDir+ipynbName+fileName, dpi=300, bbox_inches='tight', pad_inches=0.04,\n",
    "                      pil_kwargs={'compression':'tiff_lzw'})\n",
    "    plt.show()\n",
    "    print('')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 4-5-2. Modules tightened specifically by Aca"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Prepare the target module set\n",
    "posL = ['Acar-vs-Cont']\n",
    "negL = ['Rapa-vs-Cont', 'Calo-vs-Cont']\n",
    "regulation = 'Tightened'\n",
    "tempS = pd.Series(np.repeat(True, len(pvalDF)), index=pvalDF.index)#Initialize\n",
    "tempDF = pvalDF.loc[:, pvalDF.columns.str.contains('-vs-')]#Adjusted P-value\n",
    "for col_n in tempDF.columns.tolist():\n",
    "    tempS1 = pvalDF[col_n]\n",
    "    tempS2 = diffDF[col_n]\n",
    "    if col_n in posL:\n",
    "        if regulation=='Tightened':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2>0)\n",
    "        elif regulation=='Loosened':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2<0)\n",
    "    elif col_n in negL:\n",
    "        tempS3 = (tempS1>=0.05)\n",
    "        #Significance for inverse regulation\n",
    "        if regulation=='Tightened':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2<0)\n",
    "        elif regulation=='Loosened':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2>0)\n",
    "        tempS1 = tempS3 | tempS1\n",
    "    else:\n",
    "        tempS1 = (tempS1>=0.0)\n",
    "    #Update True\n",
    "    tempS = tempS & tempS1\n",
    "tempL = tempS.loc[tempS.tolist()].index.tolist()\n",
    "print(len(tempL), regulation.lower()+' modules with significance in', posL, 'but not in', negL)\n",
    "\n",
    "#Select representatives\n",
    "topX = np.min([30, len(tempL)])\n",
    "topX_plot = np.min([3, len(tempL)])\n",
    "tempDF = statDF.loc[:, statDF.columns.str.contains('Pval$')]\n",
    "tempDF = pd.merge(statDF['ModuleName'], tempDF, left_index=True, right_index=True, how='left')\n",
    "tempDF = tempDF.loc[tempL].sort_values(by=variable+'_AdjPval', ascending=True)\n",
    "print('Top', topX, 'modules (sort by the main effect of '+variable+'):')\n",
    "display(tempDF.iloc[:topX])\n",
    "plotL = tempDF.index.tolist()[:topX_plot]\n",
    "\n",
    "#Prepare DF for plot\n",
    "tempDF = rmsDF_kk.reset_index().melt(var_name='SampleID', value_name='RMS', id_vars='ModuleID')\n",
    "tempDF1 = sampleDF.reset_index()[['SampleID', 'Phenotype']]\n",
    "tempDF = pd.merge(tempDF, tempDF1, on='SampleID', how='left')\n",
    "\n",
    "#Prepare label and color\n",
    "tempD = {'Control':'Control', 'Acarbose':'Acarbose',\n",
    "         'Rapamycin':'Rapamycin', 'Calorie restriction':'CR diet'}\n",
    "tempDF['Group'] = tempDF['Phenotype'].map(tempD)\n",
    "tempD0 = {'Cont':'Control', 'Acar':'Acarbose',\n",
    "          'Rapa':'Rapamycin', 'Calo':'CR diet'}\n",
    "tempD1 = {'Control':'tab:blue', 'Acarbose':'tab:red',\n",
    "          'Rapamycin':'tab:purple', 'CR diet':'tab:olive'}\n",
    "\n",
    "#Visualize each representative\n",
    "for rank_i in range(len(plotL)):\n",
    "    print(' - Rank '+str(rank_i+1)+' (sort by the main effect of '+variable+'):')\n",
    "    module = plotL[rank_i]\n",
    "    #Check module summary\n",
    "    tempDF1 = pd.DataFrame(moduleDF.loc[module]).T\n",
    "    display(tempDF1)\n",
    "    \n",
    "    #Select RMS\n",
    "    tempDF1 = tempDF.loc[tempDF['ModuleID']==module]\n",
    "    \n",
    "    #Check RMS summary\n",
    "    tempDF2 = tempDF1.groupby(['Group'])['RMS'].agg(['count', 'mean', 'std'])\n",
    "    tempL1 = []\n",
    "    tempL2 = []\n",
    "    for row_n in tempDF2.index.tolist():\n",
    "        count, mean, std = tempDF2.loc[row_n]\n",
    "        tempL1.append(mean - 1.96*std/np.sqrt(count))\n",
    "        tempL2.append(mean + 1.96*std/np.sqrt(count))\n",
    "    tempDF2['0.025'] = tempL1\n",
    "    tempDF2['0.975'] = tempL2\n",
    "    tempDF2 = tempDF2.loc[list(tempD1.keys())]#Sort\n",
    "    display(tempDF2)\n",
    "    \n",
    "    #Prepare significance labels\n",
    "    ##Retrieve statistical significance\n",
    "    tempS = pvalDF.loc[module, pvalDF.columns.str.contains('-vs-')]#Adjusted P-value\n",
    "    tempS.name = 'AdjPval'\n",
    "    ##Clean\n",
    "    tempDF2 = tempS.index.to_series().str.split(pat='-vs-', expand=True)\n",
    "    tempDF2 = tempDF2.rename(columns={0:'Contrast', 1:'Baseline'})\n",
    "    tempDF2 = pd.merge(tempDF2, tempS, left_index=True, right_index=True, how='left')\n",
    "    tempDF2['Contrast'] = tempDF2['Contrast'].map(tempD0)\n",
    "    tempDF2['Baseline'] = tempDF2['Baseline'].map(tempD0)\n",
    "    ##Convert p-value to label\n",
    "    tempL = []\n",
    "    for row_i in range(len(tempDF2)):\n",
    "        pval = tempDF2['AdjPval'].iloc[row_i]\n",
    "        if pval<0.001:\n",
    "            tempL.append('***')\n",
    "        elif pval<0.01:\n",
    "            tempL.append('**')\n",
    "        elif pval<0.05:\n",
    "            tempL.append('*')\n",
    "        else:\n",
    "            pval_text = Decimal(str(pval)).quantize(Decimal('0.01'), rounding=ROUND_HALF_UP)\n",
    "            tempL.append(r'$P$ = '+str(pval_text))\n",
    "    tempDF2['SignifLabel'] = tempL\n",
    "    display(tempDF2)\n",
    "    \n",
    "    #Visualization\n",
    "    ymax = 1.0\n",
    "    ymin = 0.0\n",
    "    yinter = 0.2\n",
    "    ymargin_t = 0.4\n",
    "    ymargin_b = 0.05\n",
    "    aline_ymin = 1.0\n",
    "    aline_ymargin = 0.125\n",
    "    sns.set(style='ticks', font='Arial', context='talk')\n",
    "    plt.figure(figsize=(2, 4))\n",
    "    sns.pointplot(data=tempDF1, x='Group', y='RMS', order=list(tempD1.keys()), palette=tempD1,\n",
    "                  markers='o', dodge=False, join=False, capsize=0.6, estimator=np.mean, ci=95)\n",
    "    p = sns.stripplot(data=tempDF1, x='Group', y='RMS',\n",
    "                      order=list(tempD1.keys()), palette=tempD1, dodge=False, jitter=0.15,\n",
    "                      size=5, edgecolor='black', linewidth=1, **{'marker':'o', 'alpha':0.5})\n",
    "    ##Set axis\n",
    "    sns.despine()\n",
    "    p.set(ylim=(ymin-ymargin_b, ymax+ymargin_t), yticks=np.arange(ymin, ymax + yinter/10, yinter))\n",
    "    plt.setp(p.get_xticklabels(), rotation=70, horizontalalignment='right',\n",
    "             verticalalignment='center', rotation_mode='anchor')\n",
    "    ##Add significance labels\n",
    "    for row_i in range(len(tempDF2)):\n",
    "        #Baseline\n",
    "        group_0 = tempDF2['Baseline'].iloc[row_i]\n",
    "        index_0 = list(tempD1.keys()).index(group_0)\n",
    "        xcoord_0 = index_0\n",
    "        #Contrast\n",
    "        group_1 = tempDF2['Contrast'].iloc[row_i]\n",
    "        index_1 = list(tempD1.keys()).index(group_1)\n",
    "        xcoord_1 = index_1\n",
    "        #Standard point of marker\n",
    "        xcoord = (xcoord_0+xcoord_1)/2\n",
    "        ycoord = aline_ymin + aline_ymargin*row_i\n",
    "        label = tempDF2['SignifLabel'].iloc[row_i]\n",
    "        #Add annotation lines\n",
    "        aline_offset = yinter/5\n",
    "        aline_length = yinter/5 + aline_offset/2\n",
    "        plt.plot([xcoord_0, xcoord_0, xcoord_1, xcoord_1],\n",
    "                 [ycoord+aline_offset, ycoord+aline_length, ycoord+aline_length, ycoord+aline_offset],\n",
    "                 lw=1.5, c='k')\n",
    "        #Add annotation text\n",
    "        if label in ['***', '**', '*']:\n",
    "            text_offset = yinter/21\n",
    "            p.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                       horizontalalignment='center', verticalalignment='bottom',\n",
    "                       fontsize='medium', color='k')\n",
    "        else:\n",
    "            text_offset = yinter/3.5\n",
    "            p.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                       horizontalalignment='center', verticalalignment='bottom',\n",
    "                       fontsize='x-small', color='k')\n",
    "    ##Add annotation\n",
    "    p.set_title('Consensus:\\nown group', {'fontsize':'small'})\n",
    "    xoff = 0.015\n",
    "    yoff = 0.01\n",
    "    rect = plt.Rectangle((xoff, 1), 1-xoff, 0.155,#Manual adjustment\n",
    "                         transform=p.axes.transAxes, facecolor=plt.get_cmap('tab20')(15),\n",
    "                         clip_on=False, linewidth=0, zorder=0.5)\n",
    "    p.add_patch(rect)\n",
    "    ##Set axis label and title\n",
    "    plt.setp(p, xlabel='', ylabel='Sample RMS\\n(Mean = Module RCI)')\n",
    "    plt.suptitle(module, size='small',\n",
    "                 verticalalignment='bottom', horizontalalignment='center', wrap=True, y=1.0)\n",
    "    ##Save\n",
    "    fileDir = './ExportFigures/'\n",
    "    ipynbName = '220522_LC-M001-related-transcriptomics-DIRAC_DIRAC-GOBP_ver2-4_'\n",
    "    fileName = 'RCI-inter-group-comparison_RMS-pointplot-'+module.replace('GO:', 'GO')+'.tif'\n",
    "    plt.gcf().savefig(fileDir+ipynbName+fileName, dpi=300, bbox_inches='tight', pad_inches=0.04,\n",
    "                      pil_kwargs={'compression':'tiff_lzw'})\n",
    "    plt.show()\n",
    "    print('')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 4-5-3. Modules tightened specifically by Rapa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Prepare the target module set\n",
    "posL = ['Rapa-vs-Cont']\n",
    "negL = ['Acar-vs-Cont', 'Calo-vs-Cont']\n",
    "regulation = 'Tightened'\n",
    "tempS = pd.Series(np.repeat(True, len(pvalDF)), index=pvalDF.index)#Initialize\n",
    "tempDF = pvalDF.loc[:, pvalDF.columns.str.contains('-vs-')]#Adjusted P-value\n",
    "for col_n in tempDF.columns.tolist():\n",
    "    tempS1 = pvalDF[col_n]\n",
    "    tempS2 = diffDF[col_n]\n",
    "    if col_n in posL:\n",
    "        if regulation=='Tightened':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2>0)\n",
    "        elif regulation=='Loosened':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2<0)\n",
    "    elif col_n in negL:\n",
    "        tempS3 = (tempS1>=0.05)\n",
    "        #Significance for inverse regulation\n",
    "        if regulation=='Tightened':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2<0)\n",
    "        elif regulation=='Loosened':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2>0)\n",
    "        tempS1 = tempS3 | tempS1\n",
    "    else:\n",
    "        tempS1 = (tempS1>=0.0)\n",
    "    #Update True\n",
    "    tempS = tempS & tempS1\n",
    "tempL = tempS.loc[tempS.tolist()].index.tolist()\n",
    "print(len(tempL), regulation.lower()+' modules with significance in', posL, 'but not in', negL)\n",
    "\n",
    "#Select representatives\n",
    "topX = np.min([30, len(tempL)])\n",
    "topX_plot = np.min([3, len(tempL)])\n",
    "tempDF = statDF.loc[:, statDF.columns.str.contains('Pval$')]\n",
    "tempDF = pd.merge(statDF['ModuleName'], tempDF, left_index=True, right_index=True, how='left')\n",
    "tempDF = tempDF.loc[tempL].sort_values(by=variable+'_AdjPval', ascending=True)\n",
    "print('Top', topX, 'modules (sort by the main effect of '+variable+'):')\n",
    "display(tempDF.iloc[:topX])\n",
    "plotL = tempDF.index.tolist()[:topX_plot]\n",
    "\n",
    "#Prepare DF for plot\n",
    "tempDF = rmsDF_kk.reset_index().melt(var_name='SampleID', value_name='RMS', id_vars='ModuleID')\n",
    "tempDF1 = sampleDF.reset_index()[['SampleID', 'Phenotype']]\n",
    "tempDF = pd.merge(tempDF, tempDF1, on='SampleID', how='left')\n",
    "\n",
    "#Prepare label and color\n",
    "tempD = {'Control':'Control', 'Acarbose':'Acarbose',\n",
    "         'Rapamycin':'Rapamycin', 'Calorie restriction':'CR diet'}\n",
    "tempDF['Group'] = tempDF['Phenotype'].map(tempD)\n",
    "tempD0 = {'Cont':'Control', 'Acar':'Acarbose',\n",
    "          'Rapa':'Rapamycin', 'Calo':'CR diet'}\n",
    "tempD1 = {'Control':'tab:blue', 'Acarbose':'tab:red',\n",
    "          'Rapamycin':'tab:purple', 'CR diet':'tab:olive'}\n",
    "\n",
    "#Visualize each representative\n",
    "for rank_i in range(len(plotL)):\n",
    "    print(' - Rank '+str(rank_i+1)+' (sort by the main effect of '+variable+'):')\n",
    "    module = plotL[rank_i]\n",
    "    #Check module summary\n",
    "    tempDF1 = pd.DataFrame(moduleDF.loc[module]).T\n",
    "    display(tempDF1)\n",
    "    \n",
    "    #Select RMS\n",
    "    tempDF1 = tempDF.loc[tempDF['ModuleID']==module]\n",
    "    \n",
    "    #Check RMS summary\n",
    "    tempDF2 = tempDF1.groupby(['Group'])['RMS'].agg(['count', 'mean', 'std'])\n",
    "    tempL1 = []\n",
    "    tempL2 = []\n",
    "    for row_n in tempDF2.index.tolist():\n",
    "        count, mean, std = tempDF2.loc[row_n]\n",
    "        tempL1.append(mean - 1.96*std/np.sqrt(count))\n",
    "        tempL2.append(mean + 1.96*std/np.sqrt(count))\n",
    "    tempDF2['0.025'] = tempL1\n",
    "    tempDF2['0.975'] = tempL2\n",
    "    tempDF2 = tempDF2.loc[list(tempD1.keys())]#Sort\n",
    "    display(tempDF2)\n",
    "    \n",
    "    #Prepare significance labels\n",
    "    ##Retrieve statistical significance\n",
    "    tempS = pvalDF.loc[module, pvalDF.columns.str.contains('-vs-')]#Adjusted P-value\n",
    "    tempS.name = 'AdjPval'\n",
    "    ##Clean\n",
    "    tempDF2 = tempS.index.to_series().str.split(pat='-vs-', expand=True)\n",
    "    tempDF2 = tempDF2.rename(columns={0:'Contrast', 1:'Baseline'})\n",
    "    tempDF2 = pd.merge(tempDF2, tempS, left_index=True, right_index=True, how='left')\n",
    "    tempDF2['Contrast'] = tempDF2['Contrast'].map(tempD0)\n",
    "    tempDF2['Baseline'] = tempDF2['Baseline'].map(tempD0)\n",
    "    ##Convert p-value to label\n",
    "    tempL = []\n",
    "    for row_i in range(len(tempDF2)):\n",
    "        pval = tempDF2['AdjPval'].iloc[row_i]\n",
    "        if pval<0.001:\n",
    "            tempL.append('***')\n",
    "        elif pval<0.01:\n",
    "            tempL.append('**')\n",
    "        elif pval<0.05:\n",
    "            tempL.append('*')\n",
    "        else:\n",
    "            pval_text = Decimal(str(pval)).quantize(Decimal('0.01'), rounding=ROUND_HALF_UP)\n",
    "            tempL.append(r'$P$ = '+str(pval_text))\n",
    "    tempDF2['SignifLabel'] = tempL\n",
    "    display(tempDF2)\n",
    "    \n",
    "    #Visualization\n",
    "    ymax = 1.0\n",
    "    ymin = 0.0\n",
    "    yinter = 0.2\n",
    "    ymargin_t = 0.4\n",
    "    ymargin_b = 0.05\n",
    "    aline_ymin = 1.0\n",
    "    aline_ymargin = 0.125\n",
    "    sns.set(style='ticks', font='Arial', context='talk')\n",
    "    plt.figure(figsize=(2, 4))\n",
    "    sns.pointplot(data=tempDF1, x='Group', y='RMS', order=list(tempD1.keys()), palette=tempD1,\n",
    "                  markers='o', dodge=False, join=False, capsize=0.6, estimator=np.mean, ci=95)\n",
    "    p = sns.stripplot(data=tempDF1, x='Group', y='RMS',\n",
    "                      order=list(tempD1.keys()), palette=tempD1, dodge=False, jitter=0.15,\n",
    "                      size=5, edgecolor='black', linewidth=1, **{'marker':'o', 'alpha':0.5})\n",
    "    ##Set axis\n",
    "    sns.despine()\n",
    "    p.set(ylim=(ymin-ymargin_b, ymax+ymargin_t), yticks=np.arange(ymin, ymax + yinter/10, yinter))\n",
    "    plt.setp(p.get_xticklabels(), rotation=70, horizontalalignment='right',\n",
    "             verticalalignment='center', rotation_mode='anchor')\n",
    "    ##Add significance labels\n",
    "    for row_i in range(len(tempDF2)):\n",
    "        #Baseline\n",
    "        group_0 = tempDF2['Baseline'].iloc[row_i]\n",
    "        index_0 = list(tempD1.keys()).index(group_0)\n",
    "        xcoord_0 = index_0\n",
    "        #Contrast\n",
    "        group_1 = tempDF2['Contrast'].iloc[row_i]\n",
    "        index_1 = list(tempD1.keys()).index(group_1)\n",
    "        xcoord_1 = index_1\n",
    "        #Standard point of marker\n",
    "        xcoord = (xcoord_0+xcoord_1)/2\n",
    "        ycoord = aline_ymin + aline_ymargin*row_i\n",
    "        label = tempDF2['SignifLabel'].iloc[row_i]\n",
    "        #Add annotation lines\n",
    "        aline_offset = yinter/5\n",
    "        aline_length = yinter/5 + aline_offset/2\n",
    "        plt.plot([xcoord_0, xcoord_0, xcoord_1, xcoord_1],\n",
    "                 [ycoord+aline_offset, ycoord+aline_length, ycoord+aline_length, ycoord+aline_offset],\n",
    "                 lw=1.5, c='k')\n",
    "        #Add annotation text\n",
    "        if label in ['***', '**', '*']:\n",
    "            text_offset = yinter/21\n",
    "            p.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                       horizontalalignment='center', verticalalignment='bottom',\n",
    "                       fontsize='medium', color='k')\n",
    "        else:\n",
    "            text_offset = yinter/3.5\n",
    "            p.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                       horizontalalignment='center', verticalalignment='bottom',\n",
    "                       fontsize='x-small', color='k')\n",
    "    ##Add annotation\n",
    "    p.set_title('Consensus:\\nown group', {'fontsize':'small'})\n",
    "    xoff = 0.015\n",
    "    yoff = 0.01\n",
    "    rect = plt.Rectangle((xoff, 1), 1-xoff, 0.155,#Manual adjustment\n",
    "                         transform=p.axes.transAxes, facecolor=plt.get_cmap('tab20')(15),\n",
    "                         clip_on=False, linewidth=0, zorder=0.5)\n",
    "    p.add_patch(rect)\n",
    "    ##Set axis label and title\n",
    "    plt.setp(p, xlabel='', ylabel='Sample RMS\\n(Mean = Module RCI)')\n",
    "    plt.suptitle(module, size='small',\n",
    "                 verticalalignment='bottom', horizontalalignment='center', wrap=True, y=1.0)\n",
    "    ##Save\n",
    "    fileDir = './ExportFigures/'\n",
    "    ipynbName = '220522_LC-M001-related-transcriptomics-DIRAC_DIRAC-GOBP_ver2-4_'\n",
    "    fileName = 'RCI-inter-group-comparison_RMS-pointplot-'+module.replace('GO:', 'GO')+'.tif'\n",
    "    plt.gcf().savefig(fileDir+ipynbName+fileName, dpi=300, bbox_inches='tight', pad_inches=0.04,\n",
    "                      pil_kwargs={'compression':'tiff_lzw'})\n",
    "    plt.show()\n",
    "    print('')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 4-5-4. Modules tightened specifically by CR"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Prepare the target module set\n",
    "posL = ['Calo-vs-Cont']\n",
    "negL = ['Acar-vs-Cont', 'Rapa-vs-Cont']\n",
    "regulation = 'Tightened'\n",
    "tempS = pd.Series(np.repeat(True, len(pvalDF)), index=pvalDF.index)#Initialize\n",
    "tempDF = pvalDF.loc[:, pvalDF.columns.str.contains('-vs-')]#Adjusted P-value\n",
    "for col_n in tempDF.columns.tolist():\n",
    "    tempS1 = pvalDF[col_n]\n",
    "    tempS2 = diffDF[col_n]\n",
    "    if col_n in posL:\n",
    "        if regulation=='Tightened':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2>0)\n",
    "        elif regulation=='Loosened':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2<0)\n",
    "    elif col_n in negL:\n",
    "        tempS3 = (tempS1>=0.05)\n",
    "        #Significance for inverse regulation\n",
    "        if regulation=='Tightened':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2<0)\n",
    "        elif regulation=='Loosened':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2>0)\n",
    "        tempS1 = tempS3 | tempS1\n",
    "    else:\n",
    "        tempS1 = (tempS1>=0.0)\n",
    "    #Update True\n",
    "    tempS = tempS & tempS1\n",
    "tempL = tempS.loc[tempS.tolist()].index.tolist()\n",
    "print(len(tempL), regulation.lower()+' modules with significance in', posL, 'but not in', negL)\n",
    "\n",
    "#Select representatives\n",
    "topX = np.min([30, len(tempL)])\n",
    "topX_plot = np.min([3, len(tempL)])\n",
    "tempDF = statDF.loc[:, statDF.columns.str.contains('Pval$')]\n",
    "tempDF = pd.merge(statDF['ModuleName'], tempDF, left_index=True, right_index=True, how='left')\n",
    "tempDF = tempDF.loc[tempL].sort_values(by=variable+'_AdjPval', ascending=True)\n",
    "print('Top', topX, 'modules (sort by the main effect of '+variable+'):')\n",
    "display(tempDF.iloc[:topX])\n",
    "plotL = tempDF.index.tolist()[:topX_plot]\n",
    "\n",
    "#Prepare DF for plot\n",
    "tempDF = rmsDF_kk.reset_index().melt(var_name='SampleID', value_name='RMS', id_vars='ModuleID')\n",
    "tempDF1 = sampleDF.reset_index()[['SampleID', 'Phenotype']]\n",
    "tempDF = pd.merge(tempDF, tempDF1, on='SampleID', how='left')\n",
    "\n",
    "#Prepare label and color\n",
    "tempD = {'Control':'Control', 'Acarbose':'Acarbose',\n",
    "         'Rapamycin':'Rapamycin', 'Calorie restriction':'CR diet'}\n",
    "tempDF['Group'] = tempDF['Phenotype'].map(tempD)\n",
    "tempD0 = {'Cont':'Control', 'Acar':'Acarbose',\n",
    "          'Rapa':'Rapamycin', 'Calo':'CR diet'}\n",
    "tempD1 = {'Control':'tab:blue', 'Acarbose':'tab:red',\n",
    "          'Rapamycin':'tab:purple', 'CR diet':'tab:olive'}\n",
    "\n",
    "#Visualize each representative\n",
    "for rank_i in range(len(plotL)):\n",
    "    print(' - Rank '+str(rank_i+1)+' (sort by the main effect of '+variable+'):')\n",
    "    module = plotL[rank_i]\n",
    "    #Check module summary\n",
    "    tempDF1 = pd.DataFrame(moduleDF.loc[module]).T\n",
    "    display(tempDF1)\n",
    "    \n",
    "    #Select RMS\n",
    "    tempDF1 = tempDF.loc[tempDF['ModuleID']==module]\n",
    "    \n",
    "    #Check RMS summary\n",
    "    tempDF2 = tempDF1.groupby(['Group'])['RMS'].agg(['count', 'mean', 'std'])\n",
    "    tempL1 = []\n",
    "    tempL2 = []\n",
    "    for row_n in tempDF2.index.tolist():\n",
    "        count, mean, std = tempDF2.loc[row_n]\n",
    "        tempL1.append(mean - 1.96*std/np.sqrt(count))\n",
    "        tempL2.append(mean + 1.96*std/np.sqrt(count))\n",
    "    tempDF2['0.025'] = tempL1\n",
    "    tempDF2['0.975'] = tempL2\n",
    "    tempDF2 = tempDF2.loc[list(tempD1.keys())]#Sort\n",
    "    display(tempDF2)\n",
    "    \n",
    "    #Prepare significance labels\n",
    "    ##Retrieve statistical significance\n",
    "    tempS = pvalDF.loc[module, pvalDF.columns.str.contains('-vs-')]#Adjusted P-value\n",
    "    tempS.name = 'AdjPval'\n",
    "    ##Clean\n",
    "    tempDF2 = tempS.index.to_series().str.split(pat='-vs-', expand=True)\n",
    "    tempDF2 = tempDF2.rename(columns={0:'Contrast', 1:'Baseline'})\n",
    "    tempDF2 = pd.merge(tempDF2, tempS, left_index=True, right_index=True, how='left')\n",
    "    tempDF2['Contrast'] = tempDF2['Contrast'].map(tempD0)\n",
    "    tempDF2['Baseline'] = tempDF2['Baseline'].map(tempD0)\n",
    "    ##Convert p-value to label\n",
    "    tempL = []\n",
    "    for row_i in range(len(tempDF2)):\n",
    "        pval = tempDF2['AdjPval'].iloc[row_i]\n",
    "        if pval<0.001:\n",
    "            tempL.append('***')\n",
    "        elif pval<0.01:\n",
    "            tempL.append('**')\n",
    "        elif pval<0.05:\n",
    "            tempL.append('*')\n",
    "        else:\n",
    "            pval_text = Decimal(str(pval)).quantize(Decimal('0.01'), rounding=ROUND_HALF_UP)\n",
    "            tempL.append(r'$P$ = '+str(pval_text))\n",
    "    tempDF2['SignifLabel'] = tempL\n",
    "    display(tempDF2)\n",
    "    \n",
    "    #Visualization\n",
    "    ymax = 1.0\n",
    "    ymin = 0.0\n",
    "    yinter = 0.2\n",
    "    ymargin_t = 0.4\n",
    "    ymargin_b = 0.05\n",
    "    aline_ymin = 1.0\n",
    "    aline_ymargin = 0.125\n",
    "    sns.set(style='ticks', font='Arial', context='talk')\n",
    "    plt.figure(figsize=(2, 4))\n",
    "    sns.pointplot(data=tempDF1, x='Group', y='RMS', order=list(tempD1.keys()), palette=tempD1,\n",
    "                  markers='o', dodge=False, join=False, capsize=0.6, estimator=np.mean, ci=95)\n",
    "    p = sns.stripplot(data=tempDF1, x='Group', y='RMS',\n",
    "                      order=list(tempD1.keys()), palette=tempD1, dodge=False, jitter=0.15,\n",
    "                      size=5, edgecolor='black', linewidth=1, **{'marker':'o', 'alpha':0.5})\n",
    "    ##Set axis\n",
    "    sns.despine()\n",
    "    p.set(ylim=(ymin-ymargin_b, ymax+ymargin_t), yticks=np.arange(ymin, ymax + yinter/10, yinter))\n",
    "    plt.setp(p.get_xticklabels(), rotation=70, horizontalalignment='right',\n",
    "             verticalalignment='center', rotation_mode='anchor')\n",
    "    ##Add significance labels\n",
    "    for row_i in range(len(tempDF2)):\n",
    "        #Baseline\n",
    "        group_0 = tempDF2['Baseline'].iloc[row_i]\n",
    "        index_0 = list(tempD1.keys()).index(group_0)\n",
    "        xcoord_0 = index_0\n",
    "        #Contrast\n",
    "        group_1 = tempDF2['Contrast'].iloc[row_i]\n",
    "        index_1 = list(tempD1.keys()).index(group_1)\n",
    "        xcoord_1 = index_1\n",
    "        #Standard point of marker\n",
    "        xcoord = (xcoord_0+xcoord_1)/2\n",
    "        ycoord = aline_ymin + aline_ymargin*row_i\n",
    "        label = tempDF2['SignifLabel'].iloc[row_i]\n",
    "        #Add annotation lines\n",
    "        aline_offset = yinter/5\n",
    "        aline_length = yinter/5 + aline_offset/2\n",
    "        plt.plot([xcoord_0, xcoord_0, xcoord_1, xcoord_1],\n",
    "                 [ycoord+aline_offset, ycoord+aline_length, ycoord+aline_length, ycoord+aline_offset],\n",
    "                 lw=1.5, c='k')\n",
    "        #Add annotation text\n",
    "        if label in ['***', '**', '*']:\n",
    "            text_offset = yinter/21\n",
    "            p.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                       horizontalalignment='center', verticalalignment='bottom',\n",
    "                       fontsize='medium', color='k')\n",
    "        else:\n",
    "            text_offset = yinter/3.5\n",
    "            p.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                       horizontalalignment='center', verticalalignment='bottom',\n",
    "                       fontsize='x-small', color='k')\n",
    "    ##Add annotation\n",
    "    p.set_title('Consensus:\\nown group', {'fontsize':'small'})\n",
    "    xoff = 0.015\n",
    "    yoff = 0.01\n",
    "    rect = plt.Rectangle((xoff, 1), 1-xoff, 0.155,#Manual adjustment\n",
    "                         transform=p.axes.transAxes, facecolor=plt.get_cmap('tab20')(15),\n",
    "                         clip_on=False, linewidth=0, zorder=0.5)\n",
    "    p.add_patch(rect)\n",
    "    ##Set axis label and title\n",
    "    plt.setp(p, xlabel='', ylabel='Sample RMS\\n(Mean = Module RCI)')\n",
    "    plt.suptitle(module, size='small',\n",
    "                 verticalalignment='bottom', horizontalalignment='center', wrap=True, y=1.0)\n",
    "    ##Save\n",
    "    fileDir = './ExportFigures/'\n",
    "    ipynbName = '220522_LC-M001-related-transcriptomics-DIRAC_DIRAC-GOBP_ver2-4_'\n",
    "    fileName = 'RCI-inter-group-comparison_RMS-pointplot-'+module.replace('GO:', 'GO')+'.tif'\n",
    "    plt.gcf().savefig(fileDir+ipynbName+fileName, dpi=300, bbox_inches='tight', pad_inches=0.04,\n",
    "                      pil_kwargs={'compression':'tiff_lzw'})\n",
    "    plt.show()\n",
    "    print('')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 4-5-5. Top changed modules by Aca"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Prepare the target module set\n",
    "target = 'Acar-vs-Cont'\n",
    "tempL = pvalDF.loc[pvalDF[target]<0.05].index.tolist()\n",
    "print(len(tempL), 'changed modules with significance in '+target)\n",
    "\n",
    "#Select representatives\n",
    "topX = np.min([30, len(tempL)])\n",
    "topX_plot = np.min([3, len(tempL)])\n",
    "tempDF = statDF.loc[:, statDF.columns.str.contains('Pval$')]\n",
    "tempDF = pd.merge(statDF['ModuleName'], tempDF, left_index=True, right_index=True, how='left')\n",
    "tempDF = tempDF.loc[tempL].sort_values(by=target+'_AdjPval', ascending=True)\n",
    "print('Top', topX, 'modules (sort by significance in '+target+'):')\n",
    "display(tempDF.iloc[:topX])\n",
    "plotL = tempDF.index.tolist()[:topX_plot]\n",
    "\n",
    "#Prepare DF for plot\n",
    "tempDF = rmsDF_kk.reset_index().melt(var_name='SampleID', value_name='RMS', id_vars='ModuleID')\n",
    "tempDF1 = sampleDF.reset_index()[['SampleID', 'Phenotype']]\n",
    "tempDF = pd.merge(tempDF, tempDF1, on='SampleID', how='left')\n",
    "\n",
    "#Prepare label and color\n",
    "tempD = {'Control':'Control', 'Acarbose':'Acarbose',\n",
    "         'Rapamycin':'Rapamycin', 'Calorie restriction':'CR diet'}\n",
    "tempDF['Group'] = tempDF['Phenotype'].map(tempD)\n",
    "tempD0 = {'Cont':'Control', 'Acar':'Acarbose',\n",
    "          'Rapa':'Rapamycin', 'Calo':'CR diet'}\n",
    "tempD1 = {'Control':'tab:blue', 'Acarbose':'tab:red',\n",
    "          'Rapamycin':'tab:purple', 'CR diet':'tab:olive'}\n",
    "\n",
    "#Visualize each representative\n",
    "for rank_i in range(len(plotL)):\n",
    "    print(' - Rank '+str(rank_i+1)+' (sort by significance in '+target+'):')\n",
    "    module = plotL[rank_i]\n",
    "    #Check module summary\n",
    "    tempDF1 = pd.DataFrame(moduleDF.loc[module]).T\n",
    "    display(tempDF1)\n",
    "    \n",
    "    #Select RMS\n",
    "    tempDF1 = tempDF.loc[tempDF['ModuleID']==module]\n",
    "    \n",
    "    #Check RMS summary\n",
    "    tempDF2 = tempDF1.groupby(['Group'])['RMS'].agg(['count', 'mean', 'std'])\n",
    "    tempL1 = []\n",
    "    tempL2 = []\n",
    "    for row_n in tempDF2.index.tolist():\n",
    "        count, mean, std = tempDF2.loc[row_n]\n",
    "        tempL1.append(mean - 1.96*std/np.sqrt(count))\n",
    "        tempL2.append(mean + 1.96*std/np.sqrt(count))\n",
    "    tempDF2['0.025'] = tempL1\n",
    "    tempDF2['0.975'] = tempL2\n",
    "    tempDF2 = tempDF2.loc[list(tempD1.keys())]#Sort\n",
    "    display(tempDF2)\n",
    "    \n",
    "    #Prepare significance labels\n",
    "    ##Retrieve statistical significance\n",
    "    tempS = pvalDF.loc[module, pvalDF.columns.str.contains('-vs-')]#Adjusted P-value\n",
    "    tempS.name = 'AdjPval'\n",
    "    ##Clean\n",
    "    tempDF2 = tempS.index.to_series().str.split(pat='-vs-', expand=True)\n",
    "    tempDF2 = tempDF2.rename(columns={0:'Contrast', 1:'Baseline'})\n",
    "    tempDF2 = pd.merge(tempDF2, tempS, left_index=True, right_index=True, how='left')\n",
    "    tempDF2['Contrast'] = tempDF2['Contrast'].map(tempD0)\n",
    "    tempDF2['Baseline'] = tempDF2['Baseline'].map(tempD0)\n",
    "    ##Convert p-value to label\n",
    "    tempL = []\n",
    "    for row_i in range(len(tempDF2)):\n",
    "        pval = tempDF2['AdjPval'].iloc[row_i]\n",
    "        if pval<0.001:\n",
    "            tempL.append('***')\n",
    "        elif pval<0.01:\n",
    "            tempL.append('**')\n",
    "        elif pval<0.05:\n",
    "            tempL.append('*')\n",
    "        else:\n",
    "            pval_text = Decimal(str(pval)).quantize(Decimal('0.01'), rounding=ROUND_HALF_UP)\n",
    "            tempL.append(r'$P$ = '+str(pval_text))\n",
    "    tempDF2['SignifLabel'] = tempL\n",
    "    display(tempDF2)\n",
    "    \n",
    "    #Visualization\n",
    "    ymax = 1.0\n",
    "    ymin = 0.0\n",
    "    yinter = 0.2\n",
    "    ymargin_t = 0.4\n",
    "    ymargin_b = 0.05\n",
    "    aline_ymin = 1.0\n",
    "    aline_ymargin = 0.125\n",
    "    sns.set(style='ticks', font='Arial', context='talk')\n",
    "    plt.figure(figsize=(2, 4))\n",
    "    sns.pointplot(data=tempDF1, x='Group', y='RMS', order=list(tempD1.keys()), palette=tempD1,\n",
    "                  markers='o', dodge=False, join=False, capsize=0.6, estimator=np.mean, ci=95)\n",
    "    p = sns.stripplot(data=tempDF1, x='Group', y='RMS',\n",
    "                      order=list(tempD1.keys()), palette=tempD1, dodge=False, jitter=0.15,\n",
    "                      size=5, edgecolor='black', linewidth=1, **{'marker':'o', 'alpha':0.5})\n",
    "    ##Set axis\n",
    "    sns.despine()\n",
    "    p.set(ylim=(ymin-ymargin_b, ymax+ymargin_t), yticks=np.arange(ymin, ymax + yinter/10, yinter))\n",
    "    plt.setp(p.get_xticklabels(), rotation=70, horizontalalignment='right',\n",
    "             verticalalignment='center', rotation_mode='anchor')\n",
    "    ##Add significance labels\n",
    "    for row_i in range(len(tempDF2)):\n",
    "        #Baseline\n",
    "        group_0 = tempDF2['Baseline'].iloc[row_i]\n",
    "        index_0 = list(tempD1.keys()).index(group_0)\n",
    "        xcoord_0 = index_0\n",
    "        #Contrast\n",
    "        group_1 = tempDF2['Contrast'].iloc[row_i]\n",
    "        index_1 = list(tempD1.keys()).index(group_1)\n",
    "        xcoord_1 = index_1\n",
    "        #Standard point of marker\n",
    "        xcoord = (xcoord_0+xcoord_1)/2\n",
    "        ycoord = aline_ymin + aline_ymargin*row_i\n",
    "        label = tempDF2['SignifLabel'].iloc[row_i]\n",
    "        #Add annotation lines\n",
    "        aline_offset = yinter/5\n",
    "        aline_length = yinter/5 + aline_offset/2\n",
    "        plt.plot([xcoord_0, xcoord_0, xcoord_1, xcoord_1],\n",
    "                 [ycoord+aline_offset, ycoord+aline_length, ycoord+aline_length, ycoord+aline_offset],\n",
    "                 lw=1.5, c='k')\n",
    "        #Add annotation text\n",
    "        if label in ['***', '**', '*']:\n",
    "            text_offset = yinter/21\n",
    "            p.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                       horizontalalignment='center', verticalalignment='bottom',\n",
    "                       fontsize='medium', color='k')\n",
    "        else:\n",
    "            text_offset = yinter/3.5\n",
    "            p.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                       horizontalalignment='center', verticalalignment='bottom',\n",
    "                       fontsize='x-small', color='k')\n",
    "    ##Add annotation\n",
    "    p.set_title('Consensus:\\nown group', {'fontsize':'small'})\n",
    "    xoff = 0.015\n",
    "    yoff = 0.01\n",
    "    rect = plt.Rectangle((xoff, 1), 1-xoff, 0.155,#Manual adjustment\n",
    "                         transform=p.axes.transAxes, facecolor=plt.get_cmap('tab20')(15),\n",
    "                         clip_on=False, linewidth=0, zorder=0.5)\n",
    "    p.add_patch(rect)\n",
    "    ##Set axis label and title\n",
    "    plt.setp(p, xlabel='', ylabel='Sample RMS\\n(Mean = Module RCI)')\n",
    "    plt.suptitle(module, size='small',\n",
    "                 verticalalignment='bottom', horizontalalignment='center', wrap=True, y=1.0)\n",
    "    ##Save\n",
    "    fileDir = './ExportFigures/'\n",
    "    ipynbName = '220522_LC-M001-related-transcriptomics-DIRAC_DIRAC-GOBP_ver2-4_'\n",
    "    fileName = 'RCI-inter-group-comparison_RMS-pointplot-'+module.replace('GO:', 'GO')+'.tif'\n",
    "    plt.gcf().savefig(fileDir+ipynbName+fileName, dpi=300, bbox_inches='tight', pad_inches=0.04,\n",
    "                      pil_kwargs={'compression':'tiff_lzw'})\n",
    "    plt.show()\n",
    "    print('')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 4-5-6. Top changed modules by Rapa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Prepare the target module set\n",
    "target = 'Rapa-vs-Cont'\n",
    "tempL = pvalDF.loc[pvalDF[target]<0.05].index.tolist()\n",
    "print(len(tempL), 'changed modules with significance in '+target)\n",
    "\n",
    "#Select representatives\n",
    "topX = np.min([30, len(tempL)])\n",
    "topX_plot = np.min([3, len(tempL)])\n",
    "tempDF = statDF.loc[:, statDF.columns.str.contains('Pval$')]\n",
    "tempDF = pd.merge(statDF['ModuleName'], tempDF, left_index=True, right_index=True, how='left')\n",
    "tempDF = tempDF.loc[tempL].sort_values(by=target+'_AdjPval', ascending=True)\n",
    "print('Top', topX, 'modules (sort by significance in '+target+'):')\n",
    "display(tempDF.iloc[:topX])\n",
    "plotL = tempDF.index.tolist()[:topX_plot]\n",
    "\n",
    "#Prepare DF for plot\n",
    "tempDF = rmsDF_kk.reset_index().melt(var_name='SampleID', value_name='RMS', id_vars='ModuleID')\n",
    "tempDF1 = sampleDF.reset_index()[['SampleID', 'Phenotype']]\n",
    "tempDF = pd.merge(tempDF, tempDF1, on='SampleID', how='left')\n",
    "\n",
    "#Prepare label and color\n",
    "tempD = {'Control':'Control', 'Acarbose':'Acarbose',\n",
    "         'Rapamycin':'Rapamycin', 'Calorie restriction':'CR diet'}\n",
    "tempDF['Group'] = tempDF['Phenotype'].map(tempD)\n",
    "tempD0 = {'Cont':'Control', 'Acar':'Acarbose',\n",
    "          'Rapa':'Rapamycin', 'Calo':'CR diet'}\n",
    "tempD1 = {'Control':'tab:blue', 'Acarbose':'tab:red',\n",
    "          'Rapamycin':'tab:purple', 'CR diet':'tab:olive'}\n",
    "\n",
    "#Visualize each representative\n",
    "for rank_i in range(len(plotL)):\n",
    "    print(' - Rank '+str(rank_i+1)+' (sort by significance in '+target+'):')\n",
    "    module = plotL[rank_i]\n",
    "    #Check module summary\n",
    "    tempDF1 = pd.DataFrame(moduleDF.loc[module]).T\n",
    "    display(tempDF1)\n",
    "    \n",
    "    #Select RMS\n",
    "    tempDF1 = tempDF.loc[tempDF['ModuleID']==module]\n",
    "    \n",
    "    #Check RMS summary\n",
    "    tempDF2 = tempDF1.groupby(['Group'])['RMS'].agg(['count', 'mean', 'std'])\n",
    "    tempL1 = []\n",
    "    tempL2 = []\n",
    "    for row_n in tempDF2.index.tolist():\n",
    "        count, mean, std = tempDF2.loc[row_n]\n",
    "        tempL1.append(mean - 1.96*std/np.sqrt(count))\n",
    "        tempL2.append(mean + 1.96*std/np.sqrt(count))\n",
    "    tempDF2['0.025'] = tempL1\n",
    "    tempDF2['0.975'] = tempL2\n",
    "    tempDF2 = tempDF2.loc[list(tempD1.keys())]#Sort\n",
    "    display(tempDF2)\n",
    "    \n",
    "    #Prepare significance labels\n",
    "    ##Retrieve statistical significance\n",
    "    tempS = pvalDF.loc[module, pvalDF.columns.str.contains('-vs-')]#Adjusted P-value\n",
    "    tempS.name = 'AdjPval'\n",
    "    ##Clean\n",
    "    tempDF2 = tempS.index.to_series().str.split(pat='-vs-', expand=True)\n",
    "    tempDF2 = tempDF2.rename(columns={0:'Contrast', 1:'Baseline'})\n",
    "    tempDF2 = pd.merge(tempDF2, tempS, left_index=True, right_index=True, how='left')\n",
    "    tempDF2['Contrast'] = tempDF2['Contrast'].map(tempD0)\n",
    "    tempDF2['Baseline'] = tempDF2['Baseline'].map(tempD0)\n",
    "    ##Convert p-value to label\n",
    "    tempL = []\n",
    "    for row_i in range(len(tempDF2)):\n",
    "        pval = tempDF2['AdjPval'].iloc[row_i]\n",
    "        if pval<0.001:\n",
    "            tempL.append('***')\n",
    "        elif pval<0.01:\n",
    "            tempL.append('**')\n",
    "        elif pval<0.05:\n",
    "            tempL.append('*')\n",
    "        else:\n",
    "            pval_text = Decimal(str(pval)).quantize(Decimal('0.01'), rounding=ROUND_HALF_UP)\n",
    "            tempL.append(r'$P$ = '+str(pval_text))\n",
    "    tempDF2['SignifLabel'] = tempL\n",
    "    display(tempDF2)\n",
    "    \n",
    "    #Visualization\n",
    "    ymax = 1.0\n",
    "    ymin = 0.0\n",
    "    yinter = 0.2\n",
    "    ymargin_t = 0.4\n",
    "    ymargin_b = 0.05\n",
    "    aline_ymin = 1.0\n",
    "    aline_ymargin = 0.125\n",
    "    sns.set(style='ticks', font='Arial', context='talk')\n",
    "    plt.figure(figsize=(2, 4))\n",
    "    sns.pointplot(data=tempDF1, x='Group', y='RMS', order=list(tempD1.keys()), palette=tempD1,\n",
    "                  markers='o', dodge=False, join=False, capsize=0.6, estimator=np.mean, ci=95)\n",
    "    p = sns.stripplot(data=tempDF1, x='Group', y='RMS',\n",
    "                      order=list(tempD1.keys()), palette=tempD1, dodge=False, jitter=0.15,\n",
    "                      size=5, edgecolor='black', linewidth=1, **{'marker':'o', 'alpha':0.5})\n",
    "    ##Set axis\n",
    "    sns.despine()\n",
    "    p.set(ylim=(ymin-ymargin_b, ymax+ymargin_t), yticks=np.arange(ymin, ymax + yinter/10, yinter))\n",
    "    plt.setp(p.get_xticklabels(), rotation=70, horizontalalignment='right',\n",
    "             verticalalignment='center', rotation_mode='anchor')\n",
    "    ##Add significance labels\n",
    "    for row_i in range(len(tempDF2)):\n",
    "        #Baseline\n",
    "        group_0 = tempDF2['Baseline'].iloc[row_i]\n",
    "        index_0 = list(tempD1.keys()).index(group_0)\n",
    "        xcoord_0 = index_0\n",
    "        #Contrast\n",
    "        group_1 = tempDF2['Contrast'].iloc[row_i]\n",
    "        index_1 = list(tempD1.keys()).index(group_1)\n",
    "        xcoord_1 = index_1\n",
    "        #Standard point of marker\n",
    "        xcoord = (xcoord_0+xcoord_1)/2\n",
    "        ycoord = aline_ymin + aline_ymargin*row_i\n",
    "        label = tempDF2['SignifLabel'].iloc[row_i]\n",
    "        #Add annotation lines\n",
    "        aline_offset = yinter/5\n",
    "        aline_length = yinter/5 + aline_offset/2\n",
    "        plt.plot([xcoord_0, xcoord_0, xcoord_1, xcoord_1],\n",
    "                 [ycoord+aline_offset, ycoord+aline_length, ycoord+aline_length, ycoord+aline_offset],\n",
    "                 lw=1.5, c='k')\n",
    "        #Add annotation text\n",
    "        if label in ['***', '**', '*']:\n",
    "            text_offset = yinter/21\n",
    "            p.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                       horizontalalignment='center', verticalalignment='bottom',\n",
    "                       fontsize='medium', color='k')\n",
    "        else:\n",
    "            text_offset = yinter/3.5\n",
    "            p.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                       horizontalalignment='center', verticalalignment='bottom',\n",
    "                       fontsize='x-small', color='k')\n",
    "    ##Add annotation\n",
    "    p.set_title('Consensus:\\nown group', {'fontsize':'small'})\n",
    "    xoff = 0.015\n",
    "    yoff = 0.01\n",
    "    rect = plt.Rectangle((xoff, 1), 1-xoff, 0.155,#Manual adjustment\n",
    "                         transform=p.axes.transAxes, facecolor=plt.get_cmap('tab20')(15),\n",
    "                         clip_on=False, linewidth=0, zorder=0.5)\n",
    "    p.add_patch(rect)\n",
    "    ##Set axis label and title\n",
    "    plt.setp(p, xlabel='', ylabel='Sample RMS\\n(Mean = Module RCI)')\n",
    "    plt.suptitle(module, size='small',\n",
    "                 verticalalignment='bottom', horizontalalignment='center', wrap=True, y=1.0)\n",
    "    ##Save\n",
    "    fileDir = './ExportFigures/'\n",
    "    ipynbName = '220522_LC-M001-related-transcriptomics-DIRAC_DIRAC-GOBP_ver2-4_'\n",
    "    fileName = 'RCI-inter-group-comparison_RMS-pointplot-'+module.replace('GO:', 'GO')+'.tif'\n",
    "    plt.gcf().savefig(fileDir+ipynbName+fileName, dpi=300, bbox_inches='tight', pad_inches=0.04,\n",
    "                      pil_kwargs={'compression':'tiff_lzw'})\n",
    "    plt.show()\n",
    "    print('')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 4-5-7. Top changed modules by CR"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Prepare the target module set\n",
    "target = 'Calo-vs-Cont'\n",
    "tempL = pvalDF.loc[pvalDF[target]<0.05].index.tolist()\n",
    "print(len(tempL), 'changed modules with significance in '+target)\n",
    "\n",
    "#Select representatives\n",
    "topX = np.min([30, len(tempL)])\n",
    "topX_plot = np.min([3, len(tempL)])\n",
    "tempDF = statDF.loc[:, statDF.columns.str.contains('Pval$')]\n",
    "tempDF = pd.merge(statDF['ModuleName'], tempDF, left_index=True, right_index=True, how='left')\n",
    "tempDF = tempDF.loc[tempL].sort_values(by=target+'_AdjPval', ascending=True)\n",
    "print('Top', topX, 'modules (sort by significance in '+target+'):')\n",
    "display(tempDF.iloc[:topX])\n",
    "plotL = tempDF.index.tolist()[:topX_plot]\n",
    "\n",
    "#Prepare DF for plot\n",
    "tempDF = rmsDF_kk.reset_index().melt(var_name='SampleID', value_name='RMS', id_vars='ModuleID')\n",
    "tempDF1 = sampleDF.reset_index()[['SampleID', 'Phenotype']]\n",
    "tempDF = pd.merge(tempDF, tempDF1, on='SampleID', how='left')\n",
    "\n",
    "#Prepare label and color\n",
    "tempD = {'Control':'Control', 'Acarbose':'Acarbose',\n",
    "         'Rapamycin':'Rapamycin', 'Calorie restriction':'CR diet'}\n",
    "tempDF['Group'] = tempDF['Phenotype'].map(tempD)\n",
    "tempD0 = {'Cont':'Control', 'Acar':'Acarbose',\n",
    "          'Rapa':'Rapamycin', 'Calo':'CR diet'}\n",
    "tempD1 = {'Control':'tab:blue', 'Acarbose':'tab:red',\n",
    "          'Rapamycin':'tab:purple', 'CR diet':'tab:olive'}\n",
    "\n",
    "#Visualize each representative\n",
    "for rank_i in range(len(plotL)):\n",
    "    print(' - Rank '+str(rank_i+1)+' (sort by significance in '+target+'):')\n",
    "    module = plotL[rank_i]\n",
    "    #Check module summary\n",
    "    tempDF1 = pd.DataFrame(moduleDF.loc[module]).T\n",
    "    display(tempDF1)\n",
    "    \n",
    "    #Select RMS\n",
    "    tempDF1 = tempDF.loc[tempDF['ModuleID']==module]\n",
    "    \n",
    "    #Check RMS summary\n",
    "    tempDF2 = tempDF1.groupby(['Group'])['RMS'].agg(['count', 'mean', 'std'])\n",
    "    tempL1 = []\n",
    "    tempL2 = []\n",
    "    for row_n in tempDF2.index.tolist():\n",
    "        count, mean, std = tempDF2.loc[row_n]\n",
    "        tempL1.append(mean - 1.96*std/np.sqrt(count))\n",
    "        tempL2.append(mean + 1.96*std/np.sqrt(count))\n",
    "    tempDF2['0.025'] = tempL1\n",
    "    tempDF2['0.975'] = tempL2\n",
    "    tempDF2 = tempDF2.loc[list(tempD1.keys())]#Sort\n",
    "    display(tempDF2)\n",
    "    \n",
    "    #Prepare significance labels\n",
    "    ##Retrieve statistical significance\n",
    "    tempS = pvalDF.loc[module, pvalDF.columns.str.contains('-vs-')]#Adjusted P-value\n",
    "    tempS.name = 'AdjPval'\n",
    "    ##Clean\n",
    "    tempDF2 = tempS.index.to_series().str.split(pat='-vs-', expand=True)\n",
    "    tempDF2 = tempDF2.rename(columns={0:'Contrast', 1:'Baseline'})\n",
    "    tempDF2 = pd.merge(tempDF2, tempS, left_index=True, right_index=True, how='left')\n",
    "    tempDF2['Contrast'] = tempDF2['Contrast'].map(tempD0)\n",
    "    tempDF2['Baseline'] = tempDF2['Baseline'].map(tempD0)\n",
    "    ##Convert p-value to label\n",
    "    tempL = []\n",
    "    for row_i in range(len(tempDF2)):\n",
    "        pval = tempDF2['AdjPval'].iloc[row_i]\n",
    "        if pval<0.001:\n",
    "            tempL.append('***')\n",
    "        elif pval<0.01:\n",
    "            tempL.append('**')\n",
    "        elif pval<0.05:\n",
    "            tempL.append('*')\n",
    "        else:\n",
    "            pval_text = Decimal(str(pval)).quantize(Decimal('0.01'), rounding=ROUND_HALF_UP)\n",
    "            tempL.append(r'$P$ = '+str(pval_text))\n",
    "    tempDF2['SignifLabel'] = tempL\n",
    "    display(tempDF2)\n",
    "    \n",
    "    #Visualization\n",
    "    ymax = 1.0\n",
    "    ymin = 0.0\n",
    "    yinter = 0.2\n",
    "    ymargin_t = 0.4\n",
    "    ymargin_b = 0.05\n",
    "    aline_ymin = 1.0\n",
    "    aline_ymargin = 0.125\n",
    "    sns.set(style='ticks', font='Arial', context='talk')\n",
    "    plt.figure(figsize=(2, 4))\n",
    "    sns.pointplot(data=tempDF1, x='Group', y='RMS', order=list(tempD1.keys()), palette=tempD1,\n",
    "                  markers='o', dodge=False, join=False, capsize=0.6, estimator=np.mean, ci=95)\n",
    "    p = sns.stripplot(data=tempDF1, x='Group', y='RMS',\n",
    "                      order=list(tempD1.keys()), palette=tempD1, dodge=False, jitter=0.15,\n",
    "                      size=5, edgecolor='black', linewidth=1, **{'marker':'o', 'alpha':0.5})\n",
    "    ##Set axis\n",
    "    sns.despine()\n",
    "    p.set(ylim=(ymin-ymargin_b, ymax+ymargin_t), yticks=np.arange(ymin, ymax + yinter/10, yinter))\n",
    "    plt.setp(p.get_xticklabels(), rotation=70, horizontalalignment='right',\n",
    "             verticalalignment='center', rotation_mode='anchor')\n",
    "    ##Add significance labels\n",
    "    for row_i in range(len(tempDF2)):\n",
    "        #Baseline\n",
    "        group_0 = tempDF2['Baseline'].iloc[row_i]\n",
    "        index_0 = list(tempD1.keys()).index(group_0)\n",
    "        xcoord_0 = index_0\n",
    "        #Contrast\n",
    "        group_1 = tempDF2['Contrast'].iloc[row_i]\n",
    "        index_1 = list(tempD1.keys()).index(group_1)\n",
    "        xcoord_1 = index_1\n",
    "        #Standard point of marker\n",
    "        xcoord = (xcoord_0+xcoord_1)/2\n",
    "        ycoord = aline_ymin + aline_ymargin*row_i\n",
    "        label = tempDF2['SignifLabel'].iloc[row_i]\n",
    "        #Add annotation lines\n",
    "        aline_offset = yinter/5\n",
    "        aline_length = yinter/5 + aline_offset/2\n",
    "        plt.plot([xcoord_0, xcoord_0, xcoord_1, xcoord_1],\n",
    "                 [ycoord+aline_offset, ycoord+aline_length, ycoord+aline_length, ycoord+aline_offset],\n",
    "                 lw=1.5, c='k')\n",
    "        #Add annotation text\n",
    "        if label in ['***', '**', '*']:\n",
    "            text_offset = yinter/21\n",
    "            p.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                       horizontalalignment='center', verticalalignment='bottom',\n",
    "                       fontsize='medium', color='k')\n",
    "        else:\n",
    "            text_offset = yinter/3.5\n",
    "            p.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                       horizontalalignment='center', verticalalignment='bottom',\n",
    "                       fontsize='x-small', color='k')\n",
    "    ##Add annotation\n",
    "    p.set_title('Consensus:\\nown group', {'fontsize':'small'})\n",
    "    xoff = 0.015\n",
    "    yoff = 0.01\n",
    "    rect = plt.Rectangle((xoff, 1), 1-xoff, 0.155,#Manual adjustment\n",
    "                         transform=p.axes.transAxes, facecolor=plt.get_cmap('tab20')(15),\n",
    "                         clip_on=False, linewidth=0, zorder=0.5)\n",
    "    p.add_patch(rect)\n",
    "    ##Set axis label and title\n",
    "    plt.setp(p, xlabel='', ylabel='Sample RMS\\n(Mean = Module RCI)')\n",
    "    plt.suptitle(module, size='small',\n",
    "                 verticalalignment='bottom', horizontalalignment='center', wrap=True, y=1.0)\n",
    "    ##Save\n",
    "    fileDir = './ExportFigures/'\n",
    "    ipynbName = '220522_LC-M001-related-transcriptomics-DIRAC_DIRAC-GOBP_ver2-4_'\n",
    "    fileName = 'RCI-inter-group-comparison_RMS-pointplot-'+module.replace('GO:', 'GO')+'.tif'\n",
    "    plt.gcf().savefig(fileDir+ipynbName+fileName, dpi=300, bbox_inches='tight', pad_inches=0.04,\n",
    "                      pil_kwargs={'compression':'tiff_lzw'})\n",
    "    plt.show()\n",
    "    print('')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 4-5-8. Top RCI modules in Ctrl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "target = 'Control'\n",
    "\n",
    "#Select representatives\n",
    "topX = 30\n",
    "topX_plot = 3\n",
    "tempDF = statDF.loc[:, statDF.columns.str.contains('Pval$')]\n",
    "tempDF = pd.merge(statDF['ModuleName'], tempDF, left_index=True, right_index=True, how='left')\n",
    "tempDF = pd.merge(tempDF, rciDF_kk, left_index=True, right_index=True, how='left')\n",
    "tempDF = tempDF.sort_values(by=target, ascending=False)\n",
    "print('Top', topX, 'modules (sort by RCI of '+target+'):')\n",
    "display(tempDF.iloc[:topX])\n",
    "plotL = tempDF.index.tolist()[:topX_plot]\n",
    "\n",
    "#Prepare DF for plot\n",
    "tempDF = rmsDF_kk.reset_index().melt(var_name='SampleID', value_name='RMS', id_vars='ModuleID')\n",
    "tempDF1 = sampleDF.reset_index()[['SampleID', 'Phenotype']]\n",
    "tempDF = pd.merge(tempDF, tempDF1, on='SampleID', how='left')\n",
    "\n",
    "#Prepare label and color\n",
    "tempD = {'Control':'Control', 'Acarbose':'Acarbose',\n",
    "         'Rapamycin':'Rapamycin', 'Calorie restriction':'CR diet'}\n",
    "tempDF['Group'] = tempDF['Phenotype'].map(tempD)\n",
    "tempD0 = {'Cont':'Control', 'Acar':'Acarbose',\n",
    "          'Rapa':'Rapamycin', 'Calo':'CR diet'}\n",
    "tempD1 = {'Control':'tab:blue', 'Acarbose':'tab:red',\n",
    "          'Rapamycin':'tab:purple', 'CR diet':'tab:olive'}\n",
    "\n",
    "#Visualize each representative\n",
    "for rank_i in range(len(plotL)):\n",
    "    print(' - Rank '+str(rank_i+1)+' (sort by RCI of '+target+'):')\n",
    "    module = plotL[rank_i]\n",
    "    #Check module summary\n",
    "    tempDF1 = pd.DataFrame(moduleDF.loc[module]).T\n",
    "    display(tempDF1)\n",
    "    \n",
    "    #Select RMS\n",
    "    tempDF1 = tempDF.loc[tempDF['ModuleID']==module]\n",
    "    \n",
    "    #Check RMS summary\n",
    "    tempDF2 = tempDF1.groupby(['Group'])['RMS'].agg(['count', 'mean', 'std'])\n",
    "    tempL1 = []\n",
    "    tempL2 = []\n",
    "    for row_n in tempDF2.index.tolist():\n",
    "        count, mean, std = tempDF2.loc[row_n]\n",
    "        tempL1.append(mean - 1.96*std/np.sqrt(count))\n",
    "        tempL2.append(mean + 1.96*std/np.sqrt(count))\n",
    "    tempDF2['0.025'] = tempL1\n",
    "    tempDF2['0.975'] = tempL2\n",
    "    tempDF2 = tempDF2.loc[list(tempD1.keys())]#Sort\n",
    "    display(tempDF2)\n",
    "    \n",
    "    #Prepare significance labels\n",
    "    ##Re-prepare p-values because these modules can be filtered out in pvalDF\n",
    "    tempDF2 = statDF.loc[:, statDF.columns.str.contains('-vs-.*_AdjPval$')]\n",
    "    tempDF2.columns = tempDF2.columns.str.replace('_AdjPval$', '')\n",
    "    ##Retrieve statistical significance\n",
    "    tempS = tempDF2.loc[module]\n",
    "    tempS.name = 'AdjPval'\n",
    "    ##Clean\n",
    "    tempDF2 = tempS.index.to_series().str.split(pat='-vs-', expand=True)\n",
    "    tempDF2 = tempDF2.rename(columns={0:'Contrast', 1:'Baseline'})\n",
    "    tempDF2 = pd.merge(tempDF2, tempS, left_index=True, right_index=True, how='left')\n",
    "    tempDF2['Contrast'] = tempDF2['Contrast'].map(tempD0)\n",
    "    tempDF2['Baseline'] = tempDF2['Baseline'].map(tempD0)\n",
    "    ##Convert p-value to label\n",
    "    tempL = []\n",
    "    for row_i in range(len(tempDF2)):\n",
    "        pval = tempDF2['AdjPval'].iloc[row_i]\n",
    "        if pval<0.001:\n",
    "            tempL.append('***')\n",
    "        elif pval<0.01:\n",
    "            tempL.append('**')\n",
    "        elif pval<0.05:\n",
    "            tempL.append('*')\n",
    "        else:\n",
    "            pval_text = Decimal(str(pval)).quantize(Decimal('0.01'), rounding=ROUND_HALF_UP)\n",
    "            tempL.append(r'$P$ = '+str(pval_text))\n",
    "    tempDF2['SignifLabel'] = tempL\n",
    "    display(tempDF2)\n",
    "    \n",
    "    #Visualization\n",
    "    ymax = 1.0\n",
    "    ymin = 0.0\n",
    "    yinter = 0.2\n",
    "    ymargin_t = 0.4\n",
    "    ymargin_b = 0.05\n",
    "    aline_ymin = 1.0\n",
    "    aline_ymargin = 0.125\n",
    "    sns.set(style='ticks', font='Arial', context='talk')\n",
    "    plt.figure(figsize=(2, 4))\n",
    "    sns.pointplot(data=tempDF1, x='Group', y='RMS', order=list(tempD1.keys()), palette=tempD1,\n",
    "                  markers='o', dodge=False, join=False, capsize=0.6, estimator=np.mean, ci=95)\n",
    "    p = sns.stripplot(data=tempDF1, x='Group', y='RMS',\n",
    "                      order=list(tempD1.keys()), palette=tempD1, dodge=False, jitter=0.15,\n",
    "                      size=5, edgecolor='black', linewidth=1, **{'marker':'o', 'alpha':0.5})\n",
    "    ##Set axis\n",
    "    sns.despine()\n",
    "    p.set(ylim=(ymin-ymargin_b, ymax+ymargin_t), yticks=np.arange(ymin, ymax + yinter/10, yinter))\n",
    "    plt.setp(p.get_xticklabels(), rotation=70, horizontalalignment='right',\n",
    "             verticalalignment='center', rotation_mode='anchor')\n",
    "    ##Add significance labels\n",
    "    for row_i in range(len(tempDF2)):\n",
    "        #Baseline\n",
    "        group_0 = tempDF2['Baseline'].iloc[row_i]\n",
    "        index_0 = list(tempD1.keys()).index(group_0)\n",
    "        xcoord_0 = index_0\n",
    "        #Contrast\n",
    "        group_1 = tempDF2['Contrast'].iloc[row_i]\n",
    "        index_1 = list(tempD1.keys()).index(group_1)\n",
    "        xcoord_1 = index_1\n",
    "        #Standard point of marker\n",
    "        xcoord = (xcoord_0+xcoord_1)/2\n",
    "        ycoord = aline_ymin + aline_ymargin*row_i\n",
    "        label = tempDF2['SignifLabel'].iloc[row_i]\n",
    "        #Add annotation lines\n",
    "        aline_offset = yinter/5\n",
    "        aline_length = yinter/5 + aline_offset/2\n",
    "        plt.plot([xcoord_0, xcoord_0, xcoord_1, xcoord_1],\n",
    "                 [ycoord+aline_offset, ycoord+aline_length, ycoord+aline_length, ycoord+aline_offset],\n",
    "                 lw=1.5, c='k')\n",
    "        #Add annotation text\n",
    "        if label in ['***', '**', '*']:\n",
    "            text_offset = yinter/21\n",
    "            p.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                       horizontalalignment='center', verticalalignment='bottom',\n",
    "                       fontsize='medium', color='k')\n",
    "        else:\n",
    "            text_offset = yinter/3.5\n",
    "            p.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                       horizontalalignment='center', verticalalignment='bottom',\n",
    "                       fontsize='x-small', color='k')\n",
    "    ##Add annotation\n",
    "    p.set_title('Consensus:\\nown group', {'fontsize':'small'})\n",
    "    xoff = 0.015\n",
    "    yoff = 0.01\n",
    "    rect = plt.Rectangle((xoff, 1), 1-xoff, 0.155,#Manual adjustment\n",
    "                         transform=p.axes.transAxes, facecolor=plt.get_cmap('tab20')(15),\n",
    "                         clip_on=False, linewidth=0, zorder=0.5)\n",
    "    p.add_patch(rect)\n",
    "    ##Set axis label and title\n",
    "    plt.setp(p, xlabel='', ylabel='Sample RMS\\n(Mean = Module RCI)')\n",
    "    plt.suptitle(module, size='small',\n",
    "                 verticalalignment='bottom', horizontalalignment='center', wrap=True, y=1.0)\n",
    "    ##Save\n",
    "    fileDir = './ExportFigures/'\n",
    "    ipynbName = '220522_LC-M001-related-transcriptomics-DIRAC_DIRAC-GOBP_ver2-4_'\n",
    "    fileName = 'RCI-inter-group-comparison_RMS-pointplot-'+module.replace('GO:', 'GO')+'.tif'\n",
    "    plt.gcf().savefig(fileDir+ipynbName+fileName, dpi=300, bbox_inches='tight', pad_inches=0.04,\n",
    "                      pil_kwargs={'compression':'tiff_lzw'})\n",
    "    plt.show()\n",
    "    print('')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 4-5-9. Modules of special interest"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plotL = ['GO:0006635', 'GO:0031998', 'GO:0016558', 'GO:0006625',#Common in M001\n",
    "         'GO:0006637', 'GO:0006734', 'GO:0001561',#Related to fatty acid oxidation (mentioned in previous manuscript)\n",
    "         'GO:0023035', 'GO:0039529',#Immune response (mentioned in previous manuscript)\n",
    "         'GO:0010638', 'GO:0002181', 'GO:0045899',#Mentioned in previous manuscript\n",
    "         'GO:0006635', 'GO:1990126', 'GO:0098761',#Common in proteomics and transcriptomics\n",
    "         'GO:0061732', 'GO:0006086', 'GO:0019441',#Specific to proteomics (Aca)\n",
    "         'GO:0031998', 'GO:0044794', 'GO:0006625', 'GO:0016558', 'GO:0019441', 'GO:0033572',#Specific to proteomics (Rapa)\n",
    "         'GO:0006635', 'GO:0002181', 'GO:0000028', 'GO:0016558', 'GO:0006734', 'GO:0098761',#Common in M001 + M004\n",
    "         'GO:0015986', 'GO:0042776', 'GO:0070934', 'GO:0006703', 'GO:0034354', 'GO:0045899']#Similar to 4EGI-1\n",
    "plotL = list(set(plotL))\n",
    "plotL.sort()\n",
    "for module in plotL:\n",
    "    if module not in moduleDF.index.tolist():\n",
    "        print(module+' was NOT included in this analysis.')\n",
    "        plotL.remove(module)\n",
    "\n",
    "#Prepare the target module set\n",
    "tempDF = statDF.loc[:, statDF.columns.str.contains('Pval$')]\n",
    "tempDF = pd.merge(statDF['ModuleName'], tempDF, left_index=True, right_index=True, how='left')\n",
    "tempDF = tempDF.loc[plotL]\n",
    "display(tempDF)\n",
    "\n",
    "#Prepare DF for plot\n",
    "tempDF = rmsDF_kk.reset_index().melt(var_name='SampleID', value_name='RMS', id_vars='ModuleID')\n",
    "tempDF1 = sampleDF.reset_index()[['SampleID', 'Phenotype']]\n",
    "tempDF = pd.merge(tempDF, tempDF1, on='SampleID', how='left')\n",
    "\n",
    "#Prepare label and color\n",
    "tempD = {'Control':'Control', 'Acarbose':'Acarbose',\n",
    "         'Rapamycin':'Rapamycin', 'Calorie restriction':'CR diet'}\n",
    "tempDF['Group'] = tempDF['Phenotype'].map(tempD)\n",
    "tempD0 = {'Cont':'Control', 'Acar':'Acarbose',\n",
    "          'Rapa':'Rapamycin', 'Calo':'CR diet'}\n",
    "tempD1 = {'Control':'tab:blue', 'Acarbose':'tab:red',\n",
    "          'Rapamycin':'tab:purple', 'CR diet':'tab:olive'}\n",
    "\n",
    "#Visualize each representative\n",
    "for rank_i in range(len(plotL)):\n",
    "    print(' - Module '+str(rank_i+1)+':')\n",
    "    module = plotL[rank_i]\n",
    "    #Check module summary\n",
    "    tempDF1 = pd.DataFrame(moduleDF.loc[module]).T\n",
    "    display(tempDF1)\n",
    "    \n",
    "    #Select RMS\n",
    "    tempDF1 = tempDF.loc[tempDF['ModuleID']==module]\n",
    "    \n",
    "    #Check RMS summary\n",
    "    tempDF2 = tempDF1.groupby(['Group'])['RMS'].agg(['count', 'mean', 'std'])\n",
    "    tempL1 = []\n",
    "    tempL2 = []\n",
    "    for row_n in tempDF2.index.tolist():\n",
    "        count, mean, std = tempDF2.loc[row_n]\n",
    "        tempL1.append(mean - 1.96*std/np.sqrt(count))\n",
    "        tempL2.append(mean + 1.96*std/np.sqrt(count))\n",
    "    tempDF2['0.025'] = tempL1\n",
    "    tempDF2['0.975'] = tempL2\n",
    "    tempDF2 = tempDF2.loc[list(tempD1.keys())]#Sort\n",
    "    display(tempDF2)\n",
    "    \n",
    "    #Prepare significance labels\n",
    "    ##Re-prepare p-values because these modules can be filtered out in pvalDF\n",
    "    tempDF2 = statDF.loc[:, statDF.columns.str.contains('-vs-.*_AdjPval$')]\n",
    "    tempDF2.columns = tempDF2.columns.str.replace('_AdjPval$', '')\n",
    "    ##Retrieve statistical significance\n",
    "    tempS = tempDF2.loc[module]\n",
    "    tempS.name = 'AdjPval'\n",
    "    ##Clean\n",
    "    tempDF2 = tempS.index.to_series().str.split(pat='-vs-', expand=True)\n",
    "    tempDF2 = tempDF2.rename(columns={0:'Contrast', 1:'Baseline'})\n",
    "    tempDF2 = pd.merge(tempDF2, tempS, left_index=True, right_index=True, how='left')\n",
    "    tempDF2['Contrast'] = tempDF2['Contrast'].map(tempD0)\n",
    "    tempDF2['Baseline'] = tempDF2['Baseline'].map(tempD0)\n",
    "    ##Convert p-value to label\n",
    "    tempL = []\n",
    "    for row_i in range(len(tempDF2)):\n",
    "        pval = tempDF2['AdjPval'].iloc[row_i]\n",
    "        if pval<0.001:\n",
    "            tempL.append('***')\n",
    "        elif pval<0.01:\n",
    "            tempL.append('**')\n",
    "        elif pval<0.05:\n",
    "            tempL.append('*')\n",
    "        else:\n",
    "            pval_text = Decimal(str(pval)).quantize(Decimal('0.01'), rounding=ROUND_HALF_UP)\n",
    "            tempL.append(r'$P$ = '+str(pval_text))\n",
    "    tempDF2['SignifLabel'] = tempL\n",
    "    display(tempDF2)\n",
    "    \n",
    "    #Visualization\n",
    "    ymax = 1.0\n",
    "    ymin = 0.0\n",
    "    yinter = 0.2\n",
    "    ymargin_t = 0.4\n",
    "    ymargin_b = 0.05\n",
    "    aline_ymin = 1.0\n",
    "    aline_ymargin = 0.125\n",
    "    sns.set(style='ticks', font='Arial', context='talk')\n",
    "    plt.figure(figsize=(2, 4))\n",
    "    sns.pointplot(data=tempDF1, x='Group', y='RMS', order=list(tempD1.keys()), palette=tempD1,\n",
    "                  markers='o', dodge=False, join=False, capsize=0.6, estimator=np.mean, ci=95)\n",
    "    p = sns.stripplot(data=tempDF1, x='Group', y='RMS',\n",
    "                      order=list(tempD1.keys()), palette=tempD1, dodge=False, jitter=0.15,\n",
    "                      size=5, edgecolor='black', linewidth=1, **{'marker':'o', 'alpha':0.5})\n",
    "    ##Set axis\n",
    "    sns.despine()\n",
    "    p.set(ylim=(ymin-ymargin_b, ymax+ymargin_t), yticks=np.arange(ymin, ymax + yinter/10, yinter))\n",
    "    plt.setp(p.get_xticklabels(), rotation=70, horizontalalignment='right',\n",
    "             verticalalignment='center', rotation_mode='anchor')\n",
    "    ##Add significance labels\n",
    "    for row_i in range(len(tempDF2)):\n",
    "        #Baseline\n",
    "        group_0 = tempDF2['Baseline'].iloc[row_i]\n",
    "        index_0 = list(tempD1.keys()).index(group_0)\n",
    "        xcoord_0 = index_0\n",
    "        #Contrast\n",
    "        group_1 = tempDF2['Contrast'].iloc[row_i]\n",
    "        index_1 = list(tempD1.keys()).index(group_1)\n",
    "        xcoord_1 = index_1\n",
    "        #Standard point of marker\n",
    "        xcoord = (xcoord_0+xcoord_1)/2\n",
    "        ycoord = aline_ymin + aline_ymargin*row_i\n",
    "        label = tempDF2['SignifLabel'].iloc[row_i]\n",
    "        #Add annotation lines\n",
    "        aline_offset = yinter/5\n",
    "        aline_length = yinter/5 + aline_offset/2\n",
    "        plt.plot([xcoord_0, xcoord_0, xcoord_1, xcoord_1],\n",
    "                 [ycoord+aline_offset, ycoord+aline_length, ycoord+aline_length, ycoord+aline_offset],\n",
    "                 lw=1.5, c='k')\n",
    "        #Add annotation text\n",
    "        if label in ['***', '**', '*']:\n",
    "            text_offset = yinter/21\n",
    "            p.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                       horizontalalignment='center', verticalalignment='bottom',\n",
    "                       fontsize='medium', color='k')\n",
    "        else:\n",
    "            text_offset = yinter/3.5\n",
    "            p.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                       horizontalalignment='center', verticalalignment='bottom',\n",
    "                       fontsize='x-small', color='k')\n",
    "    ##Add annotation\n",
    "    p.set_title('Consensus:\\nown group', {'fontsize':'small'})\n",
    "    xoff = 0.015\n",
    "    yoff = 0.01\n",
    "    rect = plt.Rectangle((xoff, 1), 1-xoff, 0.155,#Manual adjustment\n",
    "                         transform=p.axes.transAxes, facecolor=plt.get_cmap('tab20')(15),\n",
    "                         clip_on=False, linewidth=0, zorder=0.5)\n",
    "    p.add_patch(rect)\n",
    "    ##Set axis label and title\n",
    "    plt.setp(p, xlabel='', ylabel='Sample RMS\\n(Mean = Module RCI)')\n",
    "    plt.suptitle(module, size='small',\n",
    "                 verticalalignment='bottom', horizontalalignment='center', wrap=True, y=1.0)\n",
    "    ##Save\n",
    "    fileDir = './ExportFigures/'\n",
    "    ipynbName = '220522_LC-M001-related-transcriptomics-DIRAC_DIRAC-GOBP_ver2-4_'\n",
    "    fileName = 'RCI-inter-group-comparison_RMS-pointplot-'+module.replace('GO:', 'GO')+'.tif'\n",
    "    plt.gcf().savefig(fileDir+ipynbName+fileName, dpi=300, bbox_inches='tight', pad_inches=0.04,\n",
    "                      pil_kwargs={'compression':'tiff_lzw'})\n",
    "    plt.show()\n",
    "    print('')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 5. Rank matching score under a fixed consensus: inter-group module comparison"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> Test specific hypotheses: control RMS mean == intervention RMS mean per a fixed rank consensus for each module (i.e., inter-group module comparison).  \n",
    ">\n",
    "> 3. performing Dunnett's tests per rank consensus for each module.  \n",
    ">  \n",
    "> Because the main interests are whether the changed patterns are similar or not in a changed module (i.e., this is a downstream analysis of RCI), ANOVA and the p-value adjustment across modules are unnecessary. In all the (3) tests for each rank consensus, the Dunnett's test p-values are NOT further adjusted across all tests because the hypotheses across rank consensus are not independent. In each Dunnett's test, the rank consensus group is eliminated because its mean of RMS is RCI (i.e., expected mean and variance is different) and because the number of hypotheses can be reduced.  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# — †3. Go to †2 of the R sub-notebook —  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 5-1. RMS under Ctrl consensus"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> Skip because template is Control.  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 5-2. RMS under Aca consensus"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "template = 'Acar-fixed'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 5-2-1/2/3. Dunnett's test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Import the summary table\n",
    "fileDir = './ExportData/'\n",
    "ipynbName = '220522_LC-M001-related-transcriptomics-DIRAC_StatisticalTest-GOBP_ver2-4_'\n",
    "fileName = 'inter-group-comparison.xlsx'\n",
    "sheetName = template+'-RMSmean'\n",
    "tempDF = pd.read_excel(fileDir+ipynbName+fileName, sheet_name=sheetName, engine='openpyxl')\n",
    "tempDF = tempDF.set_index('ModuleID')\n",
    "display(tempDF)\n",
    "\n",
    "statDF = tempDF"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 5-2-4. Changed modules"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> Focus only the changed modules based on RCI.  \n",
    "> –> Utilize and update the above pvalDF and diffDF (4-2-4 section).  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Extract only the changed modules\n",
    "tempDF = statDF.loc[pvalDF.index.tolist()]\n",
    "print('Changed modules (adjusted P < 0.05):', len(tempDF))\n",
    "\n",
    "#Take adjusted P-value\n",
    "tempDF1 = tempDF.loc[:, tempDF.columns.str.contains('-vs-.*_AdjPval$')]\n",
    "tempDF1.columns = tempDF1.columns.str.replace('_AdjPval$', '')\n",
    "tempDF1.columns = template+'_'+tempDF1.columns\n",
    "tempDF1 = pd.merge(pvalDF, tempDF1, left_index=True, right_index=True, how='left')\n",
    "print('Adjusted P-value:')\n",
    "display(tempDF1)\n",
    "display(tempDF1.describe())\n",
    "\n",
    "#Take effect size\n",
    "tempDF2 = tempDF.loc[:, tempDF.columns.str.contains('-vs-.*_Coef$')]\n",
    "tempDF2.columns = tempDF2.columns.str.replace('_Coef$', '')\n",
    "tempDF2.columns = template+'_'+tempDF2.columns\n",
    "tempDF2 = pd.merge(diffDF, tempDF2, left_index=True, right_index=True, how='left')\n",
    "print('Changed direction (effect size):')\n",
    "display(tempDF2)\n",
    "display(tempDF2.describe())\n",
    "\n",
    "pvalDF = tempDF1\n",
    "diffDF = tempDF2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 5-2-5. Visualization: venn diagram"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Prepare label and color\n",
    "tempD0 = {'Cont':'Control', 'Acar':'Acarbose',\n",
    "          'Rapa':'Rapamycin', 'Calo':'CR diet'}\n",
    "tempD1 = {'Dissimilarly changed\\nby Rapamycin':plt.get_cmap('tab20')(9),\n",
    "          'Similarly changed\\nby Rapamycin':'tab:purple',\n",
    "          'Similarly changed\\nby CR diet':'tab:olive',\n",
    "          'Dissimilarly changed\\nby CR diet':plt.get_cmap('tab20')(17)}\n",
    "\n",
    "#Visualization per direction\n",
    "for regulation in ['Tightened', 'Loosened']:\n",
    "    #Prepare the changed module set\n",
    "    contrast = template.replace('-fixed', '-vs-Cont')\n",
    "    tempS1 = pvalDF[contrast]\n",
    "    tempS2 = diffDF[contrast]\n",
    "    if regulation=='Changed':\n",
    "        tempS2 = tempS2.loc[(tempS1<0.05)]\n",
    "    elif regulation=='Tightened':\n",
    "        tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2>0)]\n",
    "    elif regulation=='Loosened':\n",
    "        tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2<0)]\n",
    "    tempL = tempS2.index.tolist()\n",
    "    print(regulation+' modules (vs. Control):', len(tempL))\n",
    "    \n",
    "    #Skip the followings if no changed module\n",
    "    if len(tempL)==0:\n",
    "        continue\n",
    "    \n",
    "    #Prepare similar/dissimilar module set\n",
    "    tempD2 = {}\n",
    "    tempL1 = pvalDF.loc[:, pvalDF.columns.str.contains(template)].columns.tolist()\n",
    "    for col_n in tempL1:\n",
    "        tempS1 = pvalDF.loc[tempL, col_n]\n",
    "        tempS2 = diffDF.loc[tempL, col_n]\n",
    "        for similarity in ['Similar', 'Dissimilar']:\n",
    "            if similarity=='Similar':\n",
    "                tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2>0)]\n",
    "            elif similarity=='Dissimilar':\n",
    "                tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2<0)]\n",
    "            contrast = col_n.replace(template+'_', '')\n",
    "            label = similarity+'ly changed\\nby '+tempD0[re.sub('-vs-.*', '', contrast)]\n",
    "            tempD2[label] = set(tempS2.index.tolist())\n",
    "    ##Sort to make consistent order in manual legend generation\n",
    "    tempD = {}\n",
    "    for label in tempD1.keys():\n",
    "        tempD[label] = tempD2[label]\n",
    "    \n",
    "    #Not significant in all contrasts\n",
    "    tempDF = pvalDF.loc[tempL]\n",
    "    for moduleS in tempD.values():\n",
    "        tempDF = tempDF.loc[~tempDF.index.isin(moduleS)]\n",
    "    count = len(tempDF)\n",
    "    print(' -> Not similarly/dissimilarly changed in the other interventions:', count)\n",
    "    \n",
    "    #Skip the followings if no similarly/dissimilarly changed module\n",
    "    if count==len(tempL):\n",
    "        continue\n",
    "    \n",
    "    #Venn diagram\n",
    "    sns.set(style='ticks', font='Arial', context='talk')\n",
    "    fig, ax = plt.subplots(figsize=(3, 3))\n",
    "    venn(tempD, fmt='{size:,}', cmap=list(tempD1.values()), legend_loc=None, ax=ax)\n",
    "    plt.setp(ax, ylim=(0.1, 1.05))#Otherwise, weird space...\n",
    "    #Replace text for the n/a regions\n",
    "    for text_i, text in enumerate(ax.texts):\n",
    "        if text_i in [2, 6, 10, 11, 12, 13, 14]:\n",
    "            text.set_text('n/a')\n",
    "    #Add legend annotation\n",
    "    x_coord = [0.1, 0.1, 0.9, 0.9]\n",
    "    y_coord = [0.25, 0.7, 0.7, 0.25]\n",
    "    h_align = ['right', 'right', 'left', 'left']\n",
    "    v_align = ['top', 'bottom', 'bottom', 'top']\n",
    "    for i in range(len(tempD1)):\n",
    "        key = list(tempD1.keys())[i]\n",
    "        total = f'{len(tempD[key]):,}'\n",
    "        ax.text(x_coord[i], y_coord[i], key+'\\n('+total+' modules)',\n",
    "                fontsize='small', multialignment='center',\n",
    "                horizontalalignment=h_align[i], verticalalignment=v_align[i],\n",
    "                bbox={'boxstyle':'round', 'facecolor':tempD1[key], 'pad':0.2, 'alpha':0.5})\n",
    "    title = 'Similarity to '+tempD0[template.replace('-fixed', '')]+' consensus'\n",
    "    ax.set_title(title, fontsize='medium')\n",
    "    ##Save\n",
    "    if regulation!='Changed':\n",
    "        fileDir = './ExportFigures/'\n",
    "        ipynbName = '220522_LC-M001-related-transcriptomics-DIRAC_DIRAC-GOBP_ver2-4_'\n",
    "        fileName = template+'-RMSmean-inter-group-comparison_venn-'+regulation.lower()+'.tif'\n",
    "        plt.gcf().savefig(fileDir+ipynbName+fileName, dpi=300, bbox_inches='tight', pad_inches=0.04,\n",
    "                          pil_kwargs={'compression':'tiff_lzw'})\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Export module list in each subset in the venn diagram\n",
    "for regulation in ['Tightened', 'Loosened']:\n",
    "    #Prepare the changed module set\n",
    "    contrast = template.replace('-fixed', '-vs-Cont')\n",
    "    tempS1 = pvalDF[contrast]\n",
    "    tempS2 = diffDF[contrast]\n",
    "    if regulation=='Changed':\n",
    "        tempS2 = tempS2.loc[(tempS1<0.05)]\n",
    "    elif regulation=='Tightened':\n",
    "        tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2>0)]\n",
    "    elif regulation=='Loosened':\n",
    "        tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2<0)]\n",
    "    tempL = tempS2.index.tolist()\n",
    "    print(regulation+' modules (vs. Control):', len(tempL))\n",
    "    \n",
    "    #Skip the followings if no changed module\n",
    "    if len(tempL)==0:\n",
    "        continue\n",
    "    \n",
    "    #Prepare similar/dissimilar module set\n",
    "    tempD = {}\n",
    "    tempL1 = pvalDF.loc[:, pvalDF.columns.str.contains(template)].columns.tolist()\n",
    "    for col_n in tempL1:\n",
    "        tempS1 = pvalDF.loc[tempL, col_n]\n",
    "        tempS2 = diffDF.loc[tempL, col_n]\n",
    "        for similarity in ['Similar', 'Dissimilar']:\n",
    "            if similarity=='Similar':\n",
    "                tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2>0)]\n",
    "            elif similarity=='Dissimilar':\n",
    "                tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2<0)]\n",
    "            label = col_n+'_'+similarity+'lyChanged'\n",
    "            tempD[label] = set(tempS2.index.tolist())\n",
    "    \n",
    "    #Not significant in all contrasts\n",
    "    tempDF = pvalDF.loc[tempL]\n",
    "    for moduleS in tempD.values():\n",
    "        tempDF = tempDF.loc[~tempDF.index.isin(moduleS)]\n",
    "    count = len(tempDF)\n",
    "    print(' -> Not similarly/dissimilarly changed in the other interventions:', count)\n",
    "    \n",
    "    #Skip the followings if no similarly/dissimilarly changed module\n",
    "    if count==len(tempL):\n",
    "        continue\n",
    "    \n",
    "    #Prepare a new .xlsx file (dummy README)\n",
    "    tempL1 = [len(tempD[key]) for key in tempD.keys()]\n",
    "    tempDF = pd.DataFrame({'Group':tempD.keys(), 'nModules':tempL1})\n",
    "    tempDF = tempDF.reset_index().rename(columns={'index':'VennOrder'})\n",
    "    tempDF['VennOrder'] = tempDF['VennOrder'] + 1\n",
    "    fileDir = './ExportData/'\n",
    "    ipynbName = '220522_LC-M001-related-transcriptomics-DIRAC_DIRAC-GOBP_ver2-4_'\n",
    "    fileName = template+'-RMSmean-inter-group-comparison_venn-'+regulation.lower()+'.xlsx'\n",
    "    tempDF.to_excel(fileDir+ipynbName+fileName, sheet_name='README', header=True, index=False)\n",
    "    display(tempDF)#Check\n",
    "    \n",
    "    t_start = time.time()\n",
    "    #Extract overall set\n",
    "    for key_i in range(len(tempD)):\n",
    "        key = list(tempD.keys())[key_i]\n",
    "        tempS = tempD[key]\n",
    "        tempDF = statDF.loc[statDF.index.isin(tempS)]\n",
    "        #Save summary table by appended to the above .xlsx file\n",
    "        ##Prepare sheet name\n",
    "        tempL1 = ['NA' for i in range(len(tempD))]\n",
    "        tempL1[key_i] = '1'\n",
    "        setName = '('+','.join(tempL1)+')'\n",
    "        with pd.ExcelWriter(fileDir+ipynbName+fileName, mode='a', engine='openpyxl') as writer:\n",
    "            tempDF.to_excel(writer, sheet_name=setName, header=True, index=True)\n",
    "        print(' - '+setName+':', len(tempDF))\n",
    "    \n",
    "    #Extract subset\n",
    "    tempL1 = ['1', '0']\n",
    "    tempL2 = [[k1, k2, k3, k4] for k1 in tempL1 for k2 in tempL1 for k3 in tempL1 for k4 in tempL1]\n",
    "    #tempL2.remove(['0', '0', '0', '0'])\n",
    "    for tempL1 in tempL2:\n",
    "        #Positive module set\n",
    "        tempL3 = [list(tempD.values())[key_i] for key_i, binary in enumerate(tempL1) if binary=='1']\n",
    "        tempS1 = set(pvalDF.index.tolist())#Initialize\n",
    "        for tempS in tempL3:\n",
    "            tempS1 = tempS1 & tempS\n",
    "        #Negative module set\n",
    "        tempL3 = [list(tempD.values())[key_i] for key_i, binary in enumerate(tempL1) if binary=='0']\n",
    "        tempS2 = set()#Initialize\n",
    "        for tempS in tempL3:\n",
    "            tempS2 = tempS2 | tempS\n",
    "        #Extract subset\n",
    "        tempS = tempS1 - tempS2\n",
    "        tempDF = statDF.loc[statDF.index.isin(tempS)]\n",
    "        #Save summary table by appended to the above .xlsx file\n",
    "        ##Prepare sheet name\n",
    "        setName = '('+','.join(tempL1)+')'\n",
    "        with pd.ExcelWriter(fileDir+ipynbName+fileName, mode='a', engine='openpyxl') as writer:\n",
    "            tempDF.to_excel(writer, sheet_name=setName, header=True, index=True)\n",
    "        print(' - '+setName+':', len(tempDF))\n",
    "    \n",
    "    t_elapsed = time.time() - t_start\n",
    "    print(' - Elapsed time:', round(t_elapsed//60), 'min', round(t_elapsed%60, 1), 'sec')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 5-3. RMS under Rapa consensus"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "template = 'Rapa-fixed'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 5-3-1/2/3. Dunnett's test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Import the summary table\n",
    "fileDir = './ExportData/'\n",
    "ipynbName = '220522_LC-M001-related-transcriptomics-DIRAC_StatisticalTest-GOBP_ver2-4_'\n",
    "fileName = 'inter-group-comparison.xlsx'\n",
    "sheetName = template+'-RMSmean'\n",
    "tempDF = pd.read_excel(fileDir+ipynbName+fileName, sheet_name=sheetName, engine='openpyxl')\n",
    "tempDF = tempDF.set_index('ModuleID')\n",
    "display(tempDF)\n",
    "\n",
    "statDF = tempDF"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 5-3-4. Changed modules"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> Focus only the changed modules based on RCI.  \n",
    "> –> Utilize and update the above pvalDF and diffDF (4-2-4 section).  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Extract only the changed modules\n",
    "tempDF = statDF.loc[pvalDF.index.tolist()]\n",
    "print('Changed modules (adjusted P < 0.05):', len(tempDF))\n",
    "\n",
    "#Take adjusted P-value\n",
    "tempDF1 = tempDF.loc[:, tempDF.columns.str.contains('-vs-.*_AdjPval$')]\n",
    "tempDF1.columns = tempDF1.columns.str.replace('_AdjPval$', '')\n",
    "tempDF1.columns = template+'_'+tempDF1.columns\n",
    "tempDF1 = pd.merge(pvalDF, tempDF1, left_index=True, right_index=True, how='left')\n",
    "print('Adjusted P-value:')\n",
    "display(tempDF1)\n",
    "display(tempDF1.describe())\n",
    "\n",
    "#Take effect size\n",
    "tempDF2 = tempDF.loc[:, tempDF.columns.str.contains('-vs-.*_Coef$')]\n",
    "tempDF2.columns = tempDF2.columns.str.replace('_Coef$', '')\n",
    "tempDF2.columns = template+'_'+tempDF2.columns\n",
    "tempDF2 = pd.merge(diffDF, tempDF2, left_index=True, right_index=True, how='left')\n",
    "print('Changed direction (effect size):')\n",
    "display(tempDF2)\n",
    "display(tempDF2.describe())\n",
    "\n",
    "pvalDF = tempDF1\n",
    "diffDF = tempDF2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 5-3-5. Visualization: venn diagram"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Prepare label and color\n",
    "tempD0 = {'Cont':'Control', 'Acar':'Acarbose',\n",
    "          'Rapa':'Rapamycin', 'Calo':'CR diet'}\n",
    "tempD1 = {'Dissimilarly changed\\nby CR diet':plt.get_cmap('tab20')(17),\n",
    "          'Similarly changed\\nby CR diet':'tab:olive',\n",
    "          'Similarly changed\\nby Acarbose':'tab:red',\n",
    "          'Dissimilarly changed\\nby Acarbose':plt.get_cmap('tab20')(7)}\n",
    "\n",
    "#Visualization per direction\n",
    "for regulation in ['Tightened', 'Loosened']:\n",
    "    #Prepare the changed module set\n",
    "    contrast = template.replace('-fixed', '-vs-Cont')\n",
    "    tempS1 = pvalDF[contrast]\n",
    "    tempS2 = diffDF[contrast]\n",
    "    if regulation=='Changed':\n",
    "        tempS2 = tempS2.loc[(tempS1<0.05)]\n",
    "    elif regulation=='Tightened':\n",
    "        tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2>0)]\n",
    "    elif regulation=='Loosened':\n",
    "        tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2<0)]\n",
    "    tempL = tempS2.index.tolist()\n",
    "    print(regulation+' modules (vs. Control):', len(tempL))\n",
    "    \n",
    "    #Skip the followings if no changed module\n",
    "    if len(tempL)==0:\n",
    "        continue\n",
    "    \n",
    "    #Prepare similar/dissimilar module set\n",
    "    tempD2 = {}\n",
    "    tempL1 = pvalDF.loc[:, pvalDF.columns.str.contains(template)].columns.tolist()\n",
    "    for col_n in tempL1:\n",
    "        tempS1 = pvalDF.loc[tempL, col_n]\n",
    "        tempS2 = diffDF.loc[tempL, col_n]\n",
    "        for similarity in ['Similar', 'Dissimilar']:\n",
    "            if similarity=='Similar':\n",
    "                tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2>0)]\n",
    "            elif similarity=='Dissimilar':\n",
    "                tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2<0)]\n",
    "            contrast = col_n.replace(template+'_', '')\n",
    "            label = similarity+'ly changed\\nby '+tempD0[re.sub('-vs-.*', '', contrast)]\n",
    "            tempD2[label] = set(tempS2.index.tolist())\n",
    "    ##Sort to make consistent order in manual legend generation\n",
    "    tempD = {}\n",
    "    for label in tempD1.keys():\n",
    "        tempD[label] = tempD2[label]\n",
    "    \n",
    "    #Not significant in all contrasts\n",
    "    tempDF = pvalDF.loc[tempL]\n",
    "    for moduleS in tempD.values():\n",
    "        tempDF = tempDF.loc[~tempDF.index.isin(moduleS)]\n",
    "    count = len(tempDF)\n",
    "    print(' -> Not similarly/dissimilarly changed in the other interventions:', count)\n",
    "    \n",
    "    #Skip the followings if no similarly/dissimilarly changed module\n",
    "    if count==len(tempL):\n",
    "        continue\n",
    "    \n",
    "    #Venn diagram\n",
    "    sns.set(style='ticks', font='Arial', context='talk')\n",
    "    fig, ax = plt.subplots(figsize=(3, 3))\n",
    "    venn(tempD, fmt='{size:,}', cmap=list(tempD1.values()), legend_loc=None, ax=ax)\n",
    "    plt.setp(ax, ylim=(0.1, 1.05))#Otherwise, weird space...\n",
    "    #Replace text for the n/a regions\n",
    "    for text_i, text in enumerate(ax.texts):\n",
    "        if text_i in [2, 6, 10, 11, 12, 13, 14]:\n",
    "            text.set_text('n/a')\n",
    "    #Add legend annotation\n",
    "    x_coord = [0.1, 0.1, 0.9, 0.9]\n",
    "    y_coord = [0.25, 0.7, 0.7, 0.25]\n",
    "    h_align = ['right', 'right', 'left', 'left']\n",
    "    v_align = ['top', 'bottom', 'bottom', 'top']\n",
    "    for i in range(len(tempD1)):\n",
    "        key = list(tempD1.keys())[i]\n",
    "        total = f'{len(tempD[key]):,}'\n",
    "        ax.text(x_coord[i], y_coord[i], key+'\\n('+total+' modules)',\n",
    "                fontsize='small', multialignment='center',\n",
    "                horizontalalignment=h_align[i], verticalalignment=v_align[i],\n",
    "                bbox={'boxstyle':'round', 'facecolor':tempD1[key], 'pad':0.2, 'alpha':0.5})\n",
    "    title = 'Similarity to '+tempD0[template.replace('-fixed', '')]+' consensus'\n",
    "    ax.set_title(title, fontsize='medium')\n",
    "    ##Save\n",
    "    if regulation!='Changed':\n",
    "        fileDir = './ExportFigures/'\n",
    "        ipynbName = '220522_LC-M001-related-transcriptomics-DIRAC_DIRAC-GOBP_ver2-4_'\n",
    "        fileName = template+'-RMSmean-inter-group-comparison_venn-'+regulation.lower()+'.tif'\n",
    "        plt.gcf().savefig(fileDir+ipynbName+fileName, dpi=300, bbox_inches='tight', pad_inches=0.04,\n",
    "                          pil_kwargs={'compression':'tiff_lzw'})\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Export module list in each subset in the venn diagram\n",
    "for regulation in ['Tightened', 'Loosened']:\n",
    "    #Prepare the changed module set\n",
    "    contrast = template.replace('-fixed', '-vs-Cont')\n",
    "    tempS1 = pvalDF[contrast]\n",
    "    tempS2 = diffDF[contrast]\n",
    "    if regulation=='Changed':\n",
    "        tempS2 = tempS2.loc[(tempS1<0.05)]\n",
    "    elif regulation=='Tightened':\n",
    "        tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2>0)]\n",
    "    elif regulation=='Loosened':\n",
    "        tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2<0)]\n",
    "    tempL = tempS2.index.tolist()\n",
    "    print(regulation+' modules (vs. Control):', len(tempL))\n",
    "    \n",
    "    #Skip the followings if no changed module\n",
    "    if len(tempL)==0:\n",
    "        continue\n",
    "    \n",
    "    #Prepare similar/dissimilar module set\n",
    "    tempD = {}\n",
    "    tempL1 = pvalDF.loc[:, pvalDF.columns.str.contains(template)].columns.tolist()\n",
    "    for col_n in tempL1:\n",
    "        tempS1 = pvalDF.loc[tempL, col_n]\n",
    "        tempS2 = diffDF.loc[tempL, col_n]\n",
    "        for similarity in ['Similar', 'Dissimilar']:\n",
    "            if similarity=='Similar':\n",
    "                tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2>0)]\n",
    "            elif similarity=='Dissimilar':\n",
    "                tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2<0)]\n",
    "            label = col_n+'_'+similarity+'lyChanged'\n",
    "            tempD[label] = set(tempS2.index.tolist())\n",
    "    \n",
    "    #Not significant in all contrasts\n",
    "    tempDF = pvalDF.loc[tempL]\n",
    "    for moduleS in tempD.values():\n",
    "        tempDF = tempDF.loc[~tempDF.index.isin(moduleS)]\n",
    "    count = len(tempDF)\n",
    "    print(' -> Not similarly/dissimilarly changed in the other interventions:', count)\n",
    "    \n",
    "    #Skip the followings if no similarly/dissimilarly changed module\n",
    "    if count==len(tempL):\n",
    "        continue\n",
    "    \n",
    "    #Prepare a new .xlsx file (dummy README)\n",
    "    tempL1 = [len(tempD[key]) for key in tempD.keys()]\n",
    "    tempDF = pd.DataFrame({'Group':tempD.keys(), 'nModules':tempL1})\n",
    "    tempDF = tempDF.reset_index().rename(columns={'index':'VennOrder'})\n",
    "    tempDF['VennOrder'] = tempDF['VennOrder'] + 1\n",
    "    fileDir = './ExportData/'\n",
    "    ipynbName = '220522_LC-M001-related-transcriptomics-DIRAC_DIRAC-GOBP_ver2-4_'\n",
    "    fileName = template+'-RMSmean-inter-group-comparison_venn-'+regulation.lower()+'.xlsx'\n",
    "    tempDF.to_excel(fileDir+ipynbName+fileName, sheet_name='README', header=True, index=False)\n",
    "    display(tempDF)#Check\n",
    "    \n",
    "    t_start = time.time()\n",
    "    #Extract overall set\n",
    "    for key_i in range(len(tempD)):\n",
    "        key = list(tempD.keys())[key_i]\n",
    "        tempS = tempD[key]\n",
    "        tempDF = statDF.loc[statDF.index.isin(tempS)]\n",
    "        #Save summary table by appended to the above .xlsx file\n",
    "        ##Prepare sheet name\n",
    "        tempL1 = ['NA' for i in range(len(tempD))]\n",
    "        tempL1[key_i] = '1'\n",
    "        setName = '('+','.join(tempL1)+')'\n",
    "        with pd.ExcelWriter(fileDir+ipynbName+fileName, mode='a', engine='openpyxl') as writer:\n",
    "            tempDF.to_excel(writer, sheet_name=setName, header=True, index=True)\n",
    "        print(' - '+setName+':', len(tempDF))\n",
    "    \n",
    "    #Extract subset\n",
    "    tempL1 = ['1', '0']\n",
    "    tempL2 = [[k1, k2, k3, k4] for k1 in tempL1 for k2 in tempL1 for k3 in tempL1 for k4 in tempL1]\n",
    "    #tempL2.remove(['0', '0', '0', '0'])\n",
    "    for tempL1 in tempL2:\n",
    "        #Positive module set\n",
    "        tempL3 = [list(tempD.values())[key_i] for key_i, binary in enumerate(tempL1) if binary=='1']\n",
    "        tempS1 = set(pvalDF.index.tolist())#Initialize\n",
    "        for tempS in tempL3:\n",
    "            tempS1 = tempS1 & tempS\n",
    "        #Negative module set\n",
    "        tempL3 = [list(tempD.values())[key_i] for key_i, binary in enumerate(tempL1) if binary=='0']\n",
    "        tempS2 = set()#Initialize\n",
    "        for tempS in tempL3:\n",
    "            tempS2 = tempS2 | tempS\n",
    "        #Extract subset\n",
    "        tempS = tempS1 - tempS2\n",
    "        tempDF = statDF.loc[statDF.index.isin(tempS)]\n",
    "        #Save summary table by appended to the above .xlsx file\n",
    "        ##Prepare sheet name\n",
    "        setName = '('+','.join(tempL1)+')'\n",
    "        with pd.ExcelWriter(fileDir+ipynbName+fileName, mode='a', engine='openpyxl') as writer:\n",
    "            tempDF.to_excel(writer, sheet_name=setName, header=True, index=True)\n",
    "        print(' - '+setName+':', len(tempDF))\n",
    "    \n",
    "    t_elapsed = time.time() - t_start\n",
    "    print(' - Elapsed time:', round(t_elapsed//60), 'min', round(t_elapsed%60, 1), 'sec')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 5-4. RMS under CR consensus"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "template = 'Calo-fixed'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 5-4-1/2/3. Dunnett's test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Import the summary table\n",
    "fileDir = './ExportData/'\n",
    "ipynbName = '220522_LC-M001-related-transcriptomics-DIRAC_StatisticalTest-GOBP_ver2-4_'\n",
    "fileName = 'inter-group-comparison.xlsx'\n",
    "sheetName = template+'-RMSmean'\n",
    "tempDF = pd.read_excel(fileDir+ipynbName+fileName, sheet_name=sheetName, engine='openpyxl')\n",
    "tempDF = tempDF.set_index('ModuleID')\n",
    "display(tempDF)\n",
    "\n",
    "statDF = tempDF"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 5-4-4. Changed modules"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> Focus only the changed modules based on RCI.  \n",
    "> –> Utilize and update the above pvalDF and diffDF (4-2-4 section).  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Extract only the changed modules\n",
    "tempDF = statDF.loc[pvalDF.index.tolist()]\n",
    "print('Changed modules (adjusted P < 0.05):', len(tempDF))\n",
    "\n",
    "#Take adjusted P-value\n",
    "tempDF1 = tempDF.loc[:, tempDF.columns.str.contains('-vs-.*_AdjPval$')]\n",
    "tempDF1.columns = tempDF1.columns.str.replace('_AdjPval$', '')\n",
    "tempDF1.columns = template+'_'+tempDF1.columns\n",
    "tempDF1 = pd.merge(pvalDF, tempDF1, left_index=True, right_index=True, how='left')\n",
    "print('Adjusted P-value:')\n",
    "display(tempDF1)\n",
    "display(tempDF1.describe())\n",
    "\n",
    "#Take effect size\n",
    "tempDF2 = tempDF.loc[:, tempDF.columns.str.contains('-vs-.*_Coef$')]\n",
    "tempDF2.columns = tempDF2.columns.str.replace('_Coef$', '')\n",
    "tempDF2.columns = template+'_'+tempDF2.columns\n",
    "tempDF2 = pd.merge(diffDF, tempDF2, left_index=True, right_index=True, how='left')\n",
    "print('Changed direction (effect size):')\n",
    "display(tempDF2)\n",
    "display(tempDF2.describe())\n",
    "\n",
    "pvalDF = tempDF1\n",
    "diffDF = tempDF2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 5-4-5. Visualization: venn diagram"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Prepare label and color\n",
    "tempD0 = {'Cont':'Control', 'Acar':'Acarbose',\n",
    "          'Rapa':'Rapamycin', 'Calo':'CR diet'}\n",
    "tempD1 = {'Dissimilarly changed\\nby Acarbose':plt.get_cmap('tab20')(7),\n",
    "          'Similarly changed\\nby Acarbose':'tab:red',\n",
    "          'Similarly changed\\nby Rapamycin':'tab:purple',\n",
    "          'Dissimilarly changed\\nby Rapamycin':plt.get_cmap('tab20')(9)}\n",
    "\n",
    "#Visualization per direction\n",
    "for regulation in ['Tightened', 'Loosened']:\n",
    "    #Prepare the changed module set\n",
    "    contrast = template.replace('-fixed', '-vs-Cont')\n",
    "    tempS1 = pvalDF[contrast]\n",
    "    tempS2 = diffDF[contrast]\n",
    "    if regulation=='Changed':\n",
    "        tempS2 = tempS2.loc[(tempS1<0.05)]\n",
    "    elif regulation=='Tightened':\n",
    "        tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2>0)]\n",
    "    elif regulation=='Loosened':\n",
    "        tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2<0)]\n",
    "    tempL = tempS2.index.tolist()\n",
    "    print(regulation+' modules (vs. Control):', len(tempL))\n",
    "    \n",
    "    #Skip the followings if no changed module\n",
    "    if len(tempL)==0:\n",
    "        continue\n",
    "    \n",
    "    #Prepare similar/dissimilar module set\n",
    "    tempD2 = {}\n",
    "    tempL1 = pvalDF.loc[:, pvalDF.columns.str.contains(template)].columns.tolist()\n",
    "    for col_n in tempL1:\n",
    "        tempS1 = pvalDF.loc[tempL, col_n]\n",
    "        tempS2 = diffDF.loc[tempL, col_n]\n",
    "        for similarity in ['Similar', 'Dissimilar']:\n",
    "            if similarity=='Similar':\n",
    "                tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2>0)]\n",
    "            elif similarity=='Dissimilar':\n",
    "                tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2<0)]\n",
    "            contrast = col_n.replace(template+'_', '')\n",
    "            label = similarity+'ly changed\\nby '+tempD0[re.sub('-vs-.*', '', contrast)]\n",
    "            tempD2[label] = set(tempS2.index.tolist())\n",
    "    ##Sort to make consistent order in manual legend generation\n",
    "    tempD = {}\n",
    "    for label in tempD1.keys():\n",
    "        tempD[label] = tempD2[label]\n",
    "    \n",
    "    #Not significant in all contrasts\n",
    "    tempDF = pvalDF.loc[tempL]\n",
    "    for moduleS in tempD.values():\n",
    "        tempDF = tempDF.loc[~tempDF.index.isin(moduleS)]\n",
    "    count = len(tempDF)\n",
    "    print(' -> Not similarly/dissimilarly changed in the other interventions:', count)\n",
    "    \n",
    "    #Skip the followings if no similarly/dissimilarly changed module\n",
    "    if count==len(tempL):\n",
    "        continue\n",
    "    \n",
    "    #Venn diagram\n",
    "    sns.set(style='ticks', font='Arial', context='talk')\n",
    "    fig, ax = plt.subplots(figsize=(3, 3))\n",
    "    venn(tempD, fmt='{size:,}', cmap=list(tempD1.values()), legend_loc=None, ax=ax)\n",
    "    plt.setp(ax, ylim=(0.1, 1.05))#Otherwise, weird space...\n",
    "    #Replace text for the n/a regions\n",
    "    for text_i, text in enumerate(ax.texts):\n",
    "        if text_i in [2, 6, 10, 11, 12, 13, 14]:\n",
    "            text.set_text('n/a')\n",
    "    #Add legend annotation\n",
    "    x_coord = [0.1, 0.1, 0.9, 0.9]\n",
    "    y_coord = [0.25, 0.7, 0.7, 0.25]\n",
    "    h_align = ['right', 'right', 'left', 'left']\n",
    "    v_align = ['top', 'bottom', 'bottom', 'top']\n",
    "    for i in range(len(tempD1)):\n",
    "        key = list(tempD1.keys())[i]\n",
    "        total = f'{len(tempD[key]):,}'\n",
    "        ax.text(x_coord[i], y_coord[i], key+'\\n('+total+' modules)',\n",
    "                fontsize='small', multialignment='center',\n",
    "                horizontalalignment=h_align[i], verticalalignment=v_align[i],\n",
    "                bbox={'boxstyle':'round', 'facecolor':tempD1[key], 'pad':0.2, 'alpha':0.5})\n",
    "    title = 'Similarity to '+tempD0[template.replace('-fixed', '')]+' consensus'\n",
    "    ax.set_title(title, fontsize='medium')\n",
    "    ##Save\n",
    "    if regulation!='Changed':\n",
    "        fileDir = './ExportFigures/'\n",
    "        ipynbName = '220522_LC-M001-related-transcriptomics-DIRAC_DIRAC-GOBP_ver2-4_'\n",
    "        fileName = template+'-RMSmean-inter-group-comparison_venn-'+regulation.lower()+'.tif'\n",
    "        plt.gcf().savefig(fileDir+ipynbName+fileName, dpi=300, bbox_inches='tight', pad_inches=0.04,\n",
    "                          pil_kwargs={'compression':'tiff_lzw'})\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Export module list in each subset in the venn diagram\n",
    "for regulation in ['Tightened', 'Loosened']:\n",
    "    #Prepare the changed module set\n",
    "    contrast = template.replace('-fixed', '-vs-Cont')\n",
    "    tempS1 = pvalDF[contrast]\n",
    "    tempS2 = diffDF[contrast]\n",
    "    if regulation=='Changed':\n",
    "        tempS2 = tempS2.loc[(tempS1<0.05)]\n",
    "    elif regulation=='Tightened':\n",
    "        tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2>0)]\n",
    "    elif regulation=='Loosened':\n",
    "        tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2<0)]\n",
    "    tempL = tempS2.index.tolist()\n",
    "    print(regulation+' modules (vs. Control):', len(tempL))\n",
    "    \n",
    "    #Skip the followings if no changed module\n",
    "    if len(tempL)==0:\n",
    "        continue\n",
    "    \n",
    "    #Prepare similar/dissimilar module set\n",
    "    tempD = {}\n",
    "    tempL1 = pvalDF.loc[:, pvalDF.columns.str.contains(template)].columns.tolist()\n",
    "    for col_n in tempL1:\n",
    "        tempS1 = pvalDF.loc[tempL, col_n]\n",
    "        tempS2 = diffDF.loc[tempL, col_n]\n",
    "        for similarity in ['Similar', 'Dissimilar']:\n",
    "            if similarity=='Similar':\n",
    "                tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2>0)]\n",
    "            elif similarity=='Dissimilar':\n",
    "                tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2<0)]\n",
    "            label = col_n+'_'+similarity+'lyChanged'\n",
    "            tempD[label] = set(tempS2.index.tolist())\n",
    "    \n",
    "    #Not significant in all contrasts\n",
    "    tempDF = pvalDF.loc[tempL]\n",
    "    for moduleS in tempD.values():\n",
    "        tempDF = tempDF.loc[~tempDF.index.isin(moduleS)]\n",
    "    count = len(tempDF)\n",
    "    print(' -> Not similarly/dissimilarly changed in the other interventions:', count)\n",
    "    \n",
    "    #Skip the followings if no similarly/dissimilarly changed module\n",
    "    if count==len(tempL):\n",
    "        continue\n",
    "    \n",
    "    #Prepare a new .xlsx file (dummy README)\n",
    "    tempL1 = [len(tempD[key]) for key in tempD.keys()]\n",
    "    tempDF = pd.DataFrame({'Group':tempD.keys(), 'nModules':tempL1})\n",
    "    tempDF = tempDF.reset_index().rename(columns={'index':'VennOrder'})\n",
    "    tempDF['VennOrder'] = tempDF['VennOrder'] + 1\n",
    "    fileDir = './ExportData/'\n",
    "    ipynbName = '220522_LC-M001-related-transcriptomics-DIRAC_DIRAC-GOBP_ver2-4_'\n",
    "    fileName = template+'-RMSmean-inter-group-comparison_venn-'+regulation.lower()+'.xlsx'\n",
    "    tempDF.to_excel(fileDir+ipynbName+fileName, sheet_name='README', header=True, index=False)\n",
    "    display(tempDF)#Check\n",
    "    \n",
    "    t_start = time.time()\n",
    "    #Extract overall set\n",
    "    for key_i in range(len(tempD)):\n",
    "        key = list(tempD.keys())[key_i]\n",
    "        tempS = tempD[key]\n",
    "        tempDF = statDF.loc[statDF.index.isin(tempS)]\n",
    "        #Save summary table by appended to the above .xlsx file\n",
    "        ##Prepare sheet name\n",
    "        tempL1 = ['NA' for i in range(len(tempD))]\n",
    "        tempL1[key_i] = '1'\n",
    "        setName = '('+','.join(tempL1)+')'\n",
    "        with pd.ExcelWriter(fileDir+ipynbName+fileName, mode='a', engine='openpyxl') as writer:\n",
    "            tempDF.to_excel(writer, sheet_name=setName, header=True, index=True)\n",
    "        print(' - '+setName+':', len(tempDF))\n",
    "    \n",
    "    #Extract subset\n",
    "    tempL1 = ['1', '0']\n",
    "    tempL2 = [[k1, k2, k3, k4] for k1 in tempL1 for k2 in tempL1 for k3 in tempL1 for k4 in tempL1]\n",
    "    #tempL2.remove(['0', '0', '0', '0'])\n",
    "    for tempL1 in tempL2:\n",
    "        #Positive module set\n",
    "        tempL3 = [list(tempD.values())[key_i] for key_i, binary in enumerate(tempL1) if binary=='1']\n",
    "        tempS1 = set(pvalDF.index.tolist())#Initialize\n",
    "        for tempS in tempL3:\n",
    "            tempS1 = tempS1 & tempS\n",
    "        #Negative module set\n",
    "        tempL3 = [list(tempD.values())[key_i] for key_i, binary in enumerate(tempL1) if binary=='0']\n",
    "        tempS2 = set()#Initialize\n",
    "        for tempS in tempL3:\n",
    "            tempS2 = tempS2 | tempS\n",
    "        #Extract subset\n",
    "        tempS = tempS1 - tempS2\n",
    "        tempDF = statDF.loc[statDF.index.isin(tempS)]\n",
    "        #Save summary table by appended to the above .xlsx file\n",
    "        ##Prepare sheet name\n",
    "        setName = '('+','.join(tempL1)+')'\n",
    "        with pd.ExcelWriter(fileDir+ipynbName+fileName, mode='a', engine='openpyxl') as writer:\n",
    "            tempDF.to_excel(writer, sheet_name=setName, header=True, index=True)\n",
    "        print(' - '+setName+':', len(tempDF))\n",
    "    \n",
    "    t_elapsed = time.time() - t_start\n",
    "    print(' - Elapsed time:', round(t_elapsed//60), 'min', round(t_elapsed%60, 1), 'sec')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 5-5. Visualization: pointplot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Re-import the ANOVA summary table\n",
    "fileDir = './ExportData/'\n",
    "ipynbName = '220522_LC-M001-related-transcriptomics-DIRAC_StatisticalTest-GOBP_ver2-4_'\n",
    "fileName = 'inter-group-comparison.xlsx'\n",
    "sheetName = 'RCI'\n",
    "tempDF = pd.read_excel(fileDir+ipynbName+fileName, sheet_name=sheetName, engine='openpyxl')\n",
    "tempDF = tempDF.set_index('ModuleID')\n",
    "display(tempDF)\n",
    "\n",
    "statDF = tempDF"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 5-5-1. Modules tightened by all interventions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Prepare the target module set\n",
    "posL = ['Acar-vs-Cont', 'Rapa-vs-Cont', 'Calo-vs-Cont']\n",
    "negL = ['']\n",
    "regulation = 'Tightened'\n",
    "tempS = pd.Series(np.repeat(True, len(pvalDF)), index=pvalDF.index)#Initialize\n",
    "tempDF = pvalDF.loc[:, pvalDF.columns.str.contains('-vs-')]#Adjusted P-value\n",
    "tempDF = tempDF.loc[:, ~tempDF.columns.str.contains('-fixed')]\n",
    "for col_n in tempDF.columns.tolist():\n",
    "    tempS1 = pvalDF[col_n]\n",
    "    tempS2 = diffDF[col_n]\n",
    "    if col_n in posL:\n",
    "        if regulation=='Tightened':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2>0)\n",
    "        elif regulation=='Loosened':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2<0)\n",
    "    elif col_n in negL:\n",
    "        tempS3 = (tempS1>=0.05)\n",
    "        #Significance for inverse regulation\n",
    "        if regulation=='Tightened':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2<0)\n",
    "        elif regulation=='Loosened':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2>0)\n",
    "        tempS1 = tempS3 | tempS1\n",
    "    else:\n",
    "        tempS1 = (tempS1>=0.0)\n",
    "    #Update True\n",
    "    tempS = tempS & tempS1\n",
    "tempL = tempS.loc[tempS.tolist()].index.tolist()\n",
    "print(len(tempL), regulation.lower()+' modules with significance in', posL, 'but not in', negL)\n",
    "\n",
    "#Select representatives\n",
    "topX = np.min([30, len(tempL)])\n",
    "topX_plot = np.min([10, len(tempL)])\n",
    "tempDF = statDF.loc[:, statDF.columns.str.contains('Pval$')]\n",
    "tempDF = pd.merge(statDF['ModuleName'], tempDF, left_index=True, right_index=True, how='left')\n",
    "tempDF = tempDF.loc[tempL].sort_values(by=variable+'_AdjPval', ascending=True)\n",
    "print('Top', topX, 'modules (sort by the main effect of '+variable+'):')\n",
    "display(tempDF.iloc[:topX])\n",
    "plotL = tempDF.index.tolist()[:topX_plot]\n",
    "\n",
    "#Prepare DF for plot\n",
    "tempDF = rmsDF.reset_index().melt(var_name='SampleID', value_name='RMS', id_vars=['ModuleID', 'Template'])\n",
    "tempDF1 = sampleDF.reset_index()[['SampleID', 'Phenotype']]\n",
    "tempDF = pd.merge(tempDF, tempDF1, on='SampleID', how='left')\n",
    "\n",
    "#Prepare label and color\n",
    "tempD = {'Control':'Control', 'Acarbose':'Acarbose',\n",
    "         'Rapamycin':'Rapamycin', 'Calorie restriction':'CR diet'}\n",
    "tempDF['Template'] = tempDF['Template'].map(tempD)\n",
    "tempDF['Group'] = tempDF['Phenotype'].map(tempD)\n",
    "tempD0 = {'Cont':'Control', 'Acar':'Acarbose',\n",
    "          'Rapa':'Rapamycin', 'Calo':'CR diet'}\n",
    "tempD1 = {'Control':'tab:blue', 'Acarbose':'tab:red',\n",
    "          'Rapamycin':'tab:purple', 'CR diet':'tab:olive'}\n",
    "tempD2 = {}\n",
    "for label in tempD1.keys():\n",
    "    if tempD1[label]=='tab:blue':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(1)\n",
    "    elif tempD1[label]=='tab:orange':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(3)\n",
    "    elif tempD1[label]=='tab:green':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(5)\n",
    "    elif tempD1[label]=='tab:red':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(7)\n",
    "    elif tempD1[label]=='tab:purple':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(9)\n",
    "    elif tempD1[label]=='tab:olive':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(17)\n",
    "    else:\n",
    "        tempD2[label] = 'Error?'\n",
    "\n",
    "#Visualize each representative\n",
    "for rank_i in range(len(plotL)):\n",
    "    print(' - Rank '+str(rank_i+1)+' (sort by the main effect of '+variable+'):')\n",
    "    module = plotL[rank_i]\n",
    "    #Check module summary\n",
    "    tempDF1 = pd.DataFrame(moduleDF.loc[module]).T\n",
    "    display(tempDF1)\n",
    "    \n",
    "    #Select RMS\n",
    "    tempDF1 = tempDF.loc[tempDF['ModuleID']==module]\n",
    "    \n",
    "    #Check RMS summary\n",
    "    tempDF2 = tempDF1.groupby(['Template', 'Group'])['RMS'].agg(['count', 'mean', 'std'])\n",
    "    tempL1 = []\n",
    "    tempL2 = []\n",
    "    for row_n in tempDF2.index.tolist():\n",
    "        count, mean, std = tempDF2.loc[row_n]\n",
    "        tempL1.append(mean - 1.96*std/np.sqrt(count))\n",
    "        tempL2.append(mean + 1.96*std/np.sqrt(count))\n",
    "    tempDF2['0.025'] = tempL1\n",
    "    tempDF2['0.975'] = tempL2\n",
    "    ##Multiindex sort\n",
    "    tempDF2 = tempDF2.reset_index()\n",
    "    tempDF2['Template'] = pd.Categorical(tempDF2['Template'], categories=list(tempD2.keys()))\n",
    "    tempDF2['Group'] = pd.Categorical(tempDF2['Group'], categories=list(tempD1.keys()))\n",
    "    tempDF2 = tempDF2.sort_values(by=['Template', 'Group']).set_index(['Template', 'Group'])\n",
    "    display(tempDF2)\n",
    "    \n",
    "    #Prepare significance labels\n",
    "    ##Retrieve statistical significance\n",
    "    tempS = pvalDF.loc[module, pvalDF.columns.str.contains('-vs-')]#Adjusted P-value\n",
    "    tempS = tempS.loc[tempS.index.str.contains('-fixed')]\n",
    "    tempS.name = 'AdjPval'\n",
    "    ##Clean\n",
    "    tempDF2 = tempS.index.to_series().str.split(pat='-fixed_', expand=True)\n",
    "    tempDF2 = tempDF2.rename(columns={0:'Template', 1:'Comparison'})\n",
    "    tempS1 = tempDF2['Template']\n",
    "    tempDF2 = tempDF2['Comparison'].str.split(pat='-vs-', expand=True)\n",
    "    tempDF2 = tempDF2.rename(columns={0:'Contrast', 1:'Baseline'})\n",
    "    tempDF2 = pd.merge(tempS1, tempDF2, left_index=True, right_index=True, how='left')\n",
    "    tempDF2 = pd.merge(tempDF2, tempS, left_index=True, right_index=True, how='left')\n",
    "    tempDF2['Template'] = tempDF2['Template'].map(tempD0)\n",
    "    tempDF2['Contrast'] = tempDF2['Contrast'].map(tempD0)\n",
    "    tempDF2['Baseline'] = tempDF2['Baseline'].map(tempD0)\n",
    "    ##Convert p-value to label\n",
    "    tempL = []\n",
    "    for row_i in range(len(tempDF2)):\n",
    "        pval = tempDF2['AdjPval'].iloc[row_i]\n",
    "        if pval<0.001:\n",
    "            tempL.append('***')\n",
    "        elif pval<0.01:\n",
    "            tempL.append('**')\n",
    "        elif pval<0.05:\n",
    "            tempL.append('*')\n",
    "        else:\n",
    "            pval_text = Decimal(str(pval)).quantize(Decimal('0.01'), rounding=ROUND_HALF_UP)\n",
    "            tempL.append(r'$P$ = '+str(pval_text))\n",
    "    tempDF2['SignifLabel'] = tempL\n",
    "    display(tempDF2)\n",
    "    \n",
    "    #Visualization\n",
    "    ymax = 1.0\n",
    "    ymin = 0.0\n",
    "    yinter = 0.2\n",
    "    ymargin_t = 0.275\n",
    "    ymargin_b = 0.05\n",
    "    aline_ymin = 1.0\n",
    "    aline_ymargin = 0.125\n",
    "    sns.set(style='ticks', font='Arial', context='talk')\n",
    "    fig, axes = plt.subplots(nrows=1, ncols=4, figsize=(8.8, 5.25), sharex=True, sharey=True,\n",
    "                             gridspec_kw={'width_ratios':[2.2, 2.2, 2.2, 2.2]})\n",
    "    for ax_i, ax in enumerate(axes.flat):\n",
    "        template = list(tempD2.keys())[ax_i]\n",
    "        tempDF3 = tempDF1.loc[tempDF1['Template']==template]\n",
    "        sns.pointplot(data=tempDF3, x='Group', y='RMS', order=list(tempD1.keys()), palette=tempD1,\n",
    "                      markers='o', dodge=False, join=False, capsize=0.6, estimator=np.mean, ci=95, ax=ax)\n",
    "        sns.stripplot(data=tempDF3, x='Group', y='RMS',\n",
    "                      order=list(tempD1.keys()), palette=tempD1, dodge=False, jitter=0.15,\n",
    "                      size=5, edgecolor='black', linewidth=1, **{'marker':'o', 'alpha':0.5}, ax=ax)\n",
    "        #Add RCI line\n",
    "        rci = tempDF3['RMS'].loc[tempDF3['Group']==template].mean()\n",
    "        ax.axhline(y=rci, **{'linestyle':'--', 'color':tempD1[template], 'zorder':0})\n",
    "        #Set axis\n",
    "        sns.despine()\n",
    "        ax.set(ylim=(ymin-ymargin_b, ymax+ymargin_t), yticks=np.arange(ymin, ymax + yinter/10, yinter))\n",
    "        plt.setp(ax.get_xticklabels(), rotation=70, horizontalalignment='right',\n",
    "                 verticalalignment='center', rotation_mode='anchor')\n",
    "        if ax_i==0:\n",
    "            plt.setp(ax, xlabel='', ylabel='Sample RMS')\n",
    "        else:\n",
    "            plt.setp(ax.get_yticklabels(), visible=False)\n",
    "            plt.setp(ax, xlabel='', ylabel='')\n",
    "        #Add significance labels\n",
    "        tempDF3 = tempDF2.loc[tempDF2['Template']==template]\n",
    "        for row_i in range(len(tempDF3)):\n",
    "            #Baseline\n",
    "            group_0 = tempDF3['Baseline'].iloc[row_i]\n",
    "            index_0 = list(tempD1.keys()).index(group_0)\n",
    "            xcoord_0 = index_0\n",
    "            #Contrast\n",
    "            group_1 = tempDF3['Contrast'].iloc[row_i]\n",
    "            index_1 = list(tempD1.keys()).index(group_1)\n",
    "            xcoord_1 = index_1\n",
    "            #Standard point of marker\n",
    "            xcoord = (xcoord_0+xcoord_1)/2\n",
    "            ycoord = aline_ymin + aline_ymargin*row_i\n",
    "            label = tempDF3['SignifLabel'].iloc[row_i]\n",
    "            #Add annotation lines\n",
    "            aline_offset = yinter/5\n",
    "            aline_length = yinter/5 + aline_offset/2\n",
    "            ax.plot([xcoord_0, xcoord_0, xcoord_1, xcoord_1],\n",
    "                    [ycoord+aline_offset, ycoord+aline_length, ycoord+aline_length, ycoord+aline_offset],\n",
    "                    lw=1.5, c='k')\n",
    "            #Add annotation text\n",
    "            if label in ['***', '**', '*']:\n",
    "                text_offset = yinter/21\n",
    "                ax.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                            horizontalalignment='center', verticalalignment='bottom',\n",
    "                            fontsize='medium', color='k')\n",
    "            else:\n",
    "                text_offset = yinter/3.5\n",
    "                ax.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                            horizontalalignment='center', verticalalignment='bottom',\n",
    "                            fontsize='x-small', color='k')\n",
    "        #Add annotation\n",
    "        ax.set_title('Consensus:\\n'+template, {'fontsize':'small'})\n",
    "        xoff = 0.015\n",
    "        yoff = 0.01\n",
    "        rect = plt.Rectangle((xoff, 1), 1-xoff, 0.18,#Manual adjustment\n",
    "                             transform=ax.transAxes, facecolor=tempD2[template],\n",
    "                             clip_on=False, linewidth=0, zorder=0.5)\n",
    "        ax.add_patch(rect)\n",
    "    fig.tight_layout()\n",
    "    #Set title\n",
    "    modulename = moduleDF.loc[module, 'ModuleName']\n",
    "    initial = modulename[0].capitalize()\n",
    "    title = re.sub('^.', initial, modulename)+' ('+module+')'\n",
    "    fig.suptitle(title, size='small',\n",
    "                 verticalalignment='top', horizontalalignment='center', wrap=True, y=0.95)\n",
    "    ##Save\n",
    "    fileDir = './ExportFigures/'\n",
    "    ipynbName = '220522_LC-M001-related-transcriptomics-DIRAC_DIRAC-GOBP_ver2-4_'\n",
    "    fileName = 'fixed-RMSmean-inter-group-comparison_RMS-pointplot-'+module.replace('GO:', 'GO')+'.tif'\n",
    "    plt.gcf().savefig(fileDir+ipynbName+fileName, dpi=300, bbox_inches='tight', pad_inches=0.04,\n",
    "                      pil_kwargs={'compression':'tiff_lzw'})\n",
    "    plt.show()\n",
    "    print('')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 5-5-2. Top changed modules by Aca"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Prepare the target module set\n",
    "target = 'Acar-vs-Cont'\n",
    "tempL = pvalDF.loc[pvalDF[target]<0.05].index.tolist()\n",
    "print(len(tempL), 'changed modules with significance in '+target)\n",
    "\n",
    "#Select representatives\n",
    "topX = np.min([30, len(tempL)])\n",
    "topX_plot = np.min([3, len(tempL)])\n",
    "tempDF = statDF.loc[:, statDF.columns.str.contains('Pval$')]\n",
    "tempDF = pd.merge(statDF['ModuleName'], tempDF, left_index=True, right_index=True, how='left')\n",
    "tempDF = tempDF.loc[tempL].sort_values(by=target+'_AdjPval', ascending=True)\n",
    "print('Top', topX, 'modules (sort by significance in '+target+'):')\n",
    "display(tempDF.iloc[:topX])\n",
    "plotL = tempDF.index.tolist()[:topX_plot]\n",
    "\n",
    "#Prepare DF for plot\n",
    "tempDF = rmsDF.reset_index().melt(var_name='SampleID', value_name='RMS', id_vars=['ModuleID', 'Template'])\n",
    "tempDF1 = sampleDF.reset_index()[['SampleID', 'Phenotype']]\n",
    "tempDF = pd.merge(tempDF, tempDF1, on='SampleID', how='left')\n",
    "\n",
    "#Prepare label and color\n",
    "tempD = {'Control':'Control', 'Acarbose':'Acarbose',\n",
    "         'Rapamycin':'Rapamycin', 'Calorie restriction':'CR diet'}\n",
    "tempDF['Template'] = tempDF['Template'].map(tempD)\n",
    "tempDF['Group'] = tempDF['Phenotype'].map(tempD)\n",
    "tempD0 = {'Cont':'Control', 'Acar':'Acarbose',\n",
    "          'Rapa':'Rapamycin', 'Calo':'CR diet'}\n",
    "tempD1 = {'Control':'tab:blue', 'Acarbose':'tab:red',\n",
    "          'Rapamycin':'tab:purple', 'CR diet':'tab:olive'}\n",
    "tempD2 = {}\n",
    "for label in tempD1.keys():\n",
    "    if tempD1[label]=='tab:blue':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(1)\n",
    "    elif tempD1[label]=='tab:orange':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(3)\n",
    "    elif tempD1[label]=='tab:green':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(5)\n",
    "    elif tempD1[label]=='tab:red':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(7)\n",
    "    elif tempD1[label]=='tab:purple':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(9)\n",
    "    elif tempD1[label]=='tab:olive':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(17)\n",
    "    else:\n",
    "        tempD2[label] = 'Error?'\n",
    "\n",
    "#Visualize each representative\n",
    "for rank_i in range(len(plotL)):\n",
    "    print(' - Rank '+str(rank_i+1)+' (sort by significance in '+target+'):')\n",
    "    module = plotL[rank_i]\n",
    "    #Check module summary\n",
    "    tempDF1 = pd.DataFrame(moduleDF.loc[module]).T\n",
    "    display(tempDF1)\n",
    "    \n",
    "    #Select RMS\n",
    "    tempDF1 = tempDF.loc[tempDF['ModuleID']==module]\n",
    "    \n",
    "    #Check RMS summary\n",
    "    tempDF2 = tempDF1.groupby(['Template', 'Group'])['RMS'].agg(['count', 'mean', 'std'])\n",
    "    tempL1 = []\n",
    "    tempL2 = []\n",
    "    for row_n in tempDF2.index.tolist():\n",
    "        count, mean, std = tempDF2.loc[row_n]\n",
    "        tempL1.append(mean - 1.96*std/np.sqrt(count))\n",
    "        tempL2.append(mean + 1.96*std/np.sqrt(count))\n",
    "    tempDF2['0.025'] = tempL1\n",
    "    tempDF2['0.975'] = tempL2\n",
    "    ##Multiindex sort\n",
    "    tempDF2 = tempDF2.reset_index()\n",
    "    tempDF2['Template'] = pd.Categorical(tempDF2['Template'], categories=list(tempD2.keys()))\n",
    "    tempDF2['Group'] = pd.Categorical(tempDF2['Group'], categories=list(tempD1.keys()))\n",
    "    tempDF2 = tempDF2.sort_values(by=['Template', 'Group']).set_index(['Template', 'Group'])\n",
    "    display(tempDF2)\n",
    "    \n",
    "    #Prepare significance labels\n",
    "    ##Retrieve statistical significance\n",
    "    tempS = pvalDF.loc[module, pvalDF.columns.str.contains('-vs-')]#Adjusted P-value\n",
    "    tempS = tempS.loc[tempS.index.str.contains('-fixed')]\n",
    "    tempS.name = 'AdjPval'\n",
    "    ##Clean\n",
    "    tempDF2 = tempS.index.to_series().str.split(pat='-fixed_', expand=True)\n",
    "    tempDF2 = tempDF2.rename(columns={0:'Template', 1:'Comparison'})\n",
    "    tempS1 = tempDF2['Template']\n",
    "    tempDF2 = tempDF2['Comparison'].str.split(pat='-vs-', expand=True)\n",
    "    tempDF2 = tempDF2.rename(columns={0:'Contrast', 1:'Baseline'})\n",
    "    tempDF2 = pd.merge(tempS1, tempDF2, left_index=True, right_index=True, how='left')\n",
    "    tempDF2 = pd.merge(tempDF2, tempS, left_index=True, right_index=True, how='left')\n",
    "    tempDF2['Template'] = tempDF2['Template'].map(tempD0)\n",
    "    tempDF2['Contrast'] = tempDF2['Contrast'].map(tempD0)\n",
    "    tempDF2['Baseline'] = tempDF2['Baseline'].map(tempD0)\n",
    "    ##Convert p-value to label\n",
    "    tempL = []\n",
    "    for row_i in range(len(tempDF2)):\n",
    "        pval = tempDF2['AdjPval'].iloc[row_i]\n",
    "        if pval<0.001:\n",
    "            tempL.append('***')\n",
    "        elif pval<0.01:\n",
    "            tempL.append('**')\n",
    "        elif pval<0.05:\n",
    "            tempL.append('*')\n",
    "        else:\n",
    "            pval_text = Decimal(str(pval)).quantize(Decimal('0.01'), rounding=ROUND_HALF_UP)\n",
    "            tempL.append(r'$P$ = '+str(pval_text))\n",
    "    tempDF2['SignifLabel'] = tempL\n",
    "    display(tempDF2)\n",
    "    \n",
    "    #Visualization\n",
    "    ymax = 1.0\n",
    "    ymin = 0.0\n",
    "    yinter = 0.2\n",
    "    ymargin_t = 0.275\n",
    "    ymargin_b = 0.05\n",
    "    aline_ymin = 1.0\n",
    "    aline_ymargin = 0.125\n",
    "    sns.set(style='ticks', font='Arial', context='talk')\n",
    "    fig, axes = plt.subplots(nrows=1, ncols=4, figsize=(8.8, 5.25), sharex=True, sharey=True,\n",
    "                             gridspec_kw={'width_ratios':[2.2, 2.2, 2.2, 2.2]})\n",
    "    for ax_i, ax in enumerate(axes.flat):\n",
    "        template = list(tempD2.keys())[ax_i]\n",
    "        tempDF3 = tempDF1.loc[tempDF1['Template']==template]\n",
    "        sns.pointplot(data=tempDF3, x='Group', y='RMS', order=list(tempD1.keys()), palette=tempD1,\n",
    "                      markers='o', dodge=False, join=False, capsize=0.6, estimator=np.mean, ci=95, ax=ax)\n",
    "        sns.stripplot(data=tempDF3, x='Group', y='RMS',\n",
    "                      order=list(tempD1.keys()), palette=tempD1, dodge=False, jitter=0.15,\n",
    "                      size=5, edgecolor='black', linewidth=1, **{'marker':'o', 'alpha':0.5}, ax=ax)\n",
    "        #Add RCI line\n",
    "        rci = tempDF3['RMS'].loc[tempDF3['Group']==template].mean()\n",
    "        ax.axhline(y=rci, **{'linestyle':'--', 'color':tempD1[template], 'zorder':0})\n",
    "        #Set axis\n",
    "        sns.despine()\n",
    "        ax.set(ylim=(ymin-ymargin_b, ymax+ymargin_t), yticks=np.arange(ymin, ymax + yinter/10, yinter))\n",
    "        plt.setp(ax.get_xticklabels(), rotation=70, horizontalalignment='right',\n",
    "                 verticalalignment='center', rotation_mode='anchor')\n",
    "        if ax_i==0:\n",
    "            plt.setp(ax, xlabel='', ylabel='Sample RMS')\n",
    "        else:\n",
    "            plt.setp(ax.get_yticklabels(), visible=False)\n",
    "            plt.setp(ax, xlabel='', ylabel='')\n",
    "        #Add significance labels\n",
    "        tempDF3 = tempDF2.loc[tempDF2['Template']==template]\n",
    "        for row_i in range(len(tempDF3)):\n",
    "            #Baseline\n",
    "            group_0 = tempDF3['Baseline'].iloc[row_i]\n",
    "            index_0 = list(tempD1.keys()).index(group_0)\n",
    "            xcoord_0 = index_0\n",
    "            #Contrast\n",
    "            group_1 = tempDF3['Contrast'].iloc[row_i]\n",
    "            index_1 = list(tempD1.keys()).index(group_1)\n",
    "            xcoord_1 = index_1\n",
    "            #Standard point of marker\n",
    "            xcoord = (xcoord_0+xcoord_1)/2\n",
    "            ycoord = aline_ymin + aline_ymargin*row_i\n",
    "            label = tempDF3['SignifLabel'].iloc[row_i]\n",
    "            #Add annotation lines\n",
    "            aline_offset = yinter/5\n",
    "            aline_length = yinter/5 + aline_offset/2\n",
    "            ax.plot([xcoord_0, xcoord_0, xcoord_1, xcoord_1],\n",
    "                    [ycoord+aline_offset, ycoord+aline_length, ycoord+aline_length, ycoord+aline_offset],\n",
    "                    lw=1.5, c='k')\n",
    "            #Add annotation text\n",
    "            if label in ['***', '**', '*']:\n",
    "                text_offset = yinter/21\n",
    "                ax.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                            horizontalalignment='center', verticalalignment='bottom',\n",
    "                            fontsize='medium', color='k')\n",
    "            else:\n",
    "                text_offset = yinter/3.5\n",
    "                ax.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                            horizontalalignment='center', verticalalignment='bottom',\n",
    "                            fontsize='x-small', color='k')\n",
    "        #Add annotation\n",
    "        ax.set_title('Consensus:\\n'+template, {'fontsize':'small'})\n",
    "        xoff = 0.015\n",
    "        yoff = 0.01\n",
    "        rect = plt.Rectangle((xoff, 1), 1-xoff, 0.18,#Manual adjustment\n",
    "                             transform=ax.transAxes, facecolor=tempD2[template],\n",
    "                             clip_on=False, linewidth=0, zorder=0.5)\n",
    "        ax.add_patch(rect)\n",
    "    fig.tight_layout()\n",
    "    #Set title\n",
    "    modulename = moduleDF.loc[module, 'ModuleName']\n",
    "    initial = modulename[0].capitalize()\n",
    "    title = re.sub('^.', initial, modulename)+' ('+module+')'\n",
    "    fig.suptitle(title, size='small',\n",
    "                 verticalalignment='top', horizontalalignment='center', wrap=True, y=0.95)\n",
    "    ##Save\n",
    "    fileDir = './ExportFigures/'\n",
    "    ipynbName = '220522_LC-M001-related-transcriptomics-DIRAC_DIRAC-GOBP_ver2-4_'\n",
    "    fileName = 'fixed-RMSmean-inter-group-comparison_RMS-pointplot-'+module.replace('GO:', 'GO')+'.tif'\n",
    "    plt.gcf().savefig(fileDir+ipynbName+fileName, dpi=300, bbox_inches='tight', pad_inches=0.04,\n",
    "                      pil_kwargs={'compression':'tiff_lzw'})\n",
    "    plt.show()\n",
    "    print('')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 5-5-3. Top changed modules by Rapa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Prepare the target module set\n",
    "target = 'Rapa-vs-Cont'\n",
    "tempL = pvalDF.loc[pvalDF[target]<0.05].index.tolist()\n",
    "print(len(tempL), 'changed modules with significance in '+target)\n",
    "\n",
    "#Select representatives\n",
    "topX = np.min([30, len(tempL)])\n",
    "topX_plot = np.min([3, len(tempL)])\n",
    "tempDF = statDF.loc[:, statDF.columns.str.contains('Pval$')]\n",
    "tempDF = pd.merge(statDF['ModuleName'], tempDF, left_index=True, right_index=True, how='left')\n",
    "tempDF = tempDF.loc[tempL].sort_values(by=target+'_AdjPval', ascending=True)\n",
    "print('Top', topX, 'modules (sort by significance in '+target+'):')\n",
    "display(tempDF.iloc[:topX])\n",
    "plotL = tempDF.index.tolist()[:topX_plot]\n",
    "\n",
    "#Prepare DF for plot\n",
    "tempDF = rmsDF.reset_index().melt(var_name='SampleID', value_name='RMS', id_vars=['ModuleID', 'Template'])\n",
    "tempDF1 = sampleDF.reset_index()[['SampleID', 'Phenotype']]\n",
    "tempDF = pd.merge(tempDF, tempDF1, on='SampleID', how='left')\n",
    "\n",
    "#Prepare label and color\n",
    "tempD = {'Control':'Control', 'Acarbose':'Acarbose',\n",
    "         'Rapamycin':'Rapamycin', 'Calorie restriction':'CR diet'}\n",
    "tempDF['Template'] = tempDF['Template'].map(tempD)\n",
    "tempDF['Group'] = tempDF['Phenotype'].map(tempD)\n",
    "tempD0 = {'Cont':'Control', 'Acar':'Acarbose',\n",
    "          'Rapa':'Rapamycin', 'Calo':'CR diet'}\n",
    "tempD1 = {'Control':'tab:blue', 'Acarbose':'tab:red',\n",
    "          'Rapamycin':'tab:purple', 'CR diet':'tab:olive'}\n",
    "tempD2 = {}\n",
    "for label in tempD1.keys():\n",
    "    if tempD1[label]=='tab:blue':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(1)\n",
    "    elif tempD1[label]=='tab:orange':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(3)\n",
    "    elif tempD1[label]=='tab:green':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(5)\n",
    "    elif tempD1[label]=='tab:red':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(7)\n",
    "    elif tempD1[label]=='tab:purple':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(9)\n",
    "    elif tempD1[label]=='tab:olive':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(17)\n",
    "    else:\n",
    "        tempD2[label] = 'Error?'\n",
    "\n",
    "#Visualize each representative\n",
    "for rank_i in range(len(plotL)):\n",
    "    print(' - Rank '+str(rank_i+1)+' (sort by significance in '+target+'):')\n",
    "    module = plotL[rank_i]\n",
    "    #Check module summary\n",
    "    tempDF1 = pd.DataFrame(moduleDF.loc[module]).T\n",
    "    display(tempDF1)\n",
    "    \n",
    "    #Select RMS\n",
    "    tempDF1 = tempDF.loc[tempDF['ModuleID']==module]\n",
    "    \n",
    "    #Check RMS summary\n",
    "    tempDF2 = tempDF1.groupby(['Template', 'Group'])['RMS'].agg(['count', 'mean', 'std'])\n",
    "    tempL1 = []\n",
    "    tempL2 = []\n",
    "    for row_n in tempDF2.index.tolist():\n",
    "        count, mean, std = tempDF2.loc[row_n]\n",
    "        tempL1.append(mean - 1.96*std/np.sqrt(count))\n",
    "        tempL2.append(mean + 1.96*std/np.sqrt(count))\n",
    "    tempDF2['0.025'] = tempL1\n",
    "    tempDF2['0.975'] = tempL2\n",
    "    ##Multiindex sort\n",
    "    tempDF2 = tempDF2.reset_index()\n",
    "    tempDF2['Template'] = pd.Categorical(tempDF2['Template'], categories=list(tempD2.keys()))\n",
    "    tempDF2['Group'] = pd.Categorical(tempDF2['Group'], categories=list(tempD1.keys()))\n",
    "    tempDF2 = tempDF2.sort_values(by=['Template', 'Group']).set_index(['Template', 'Group'])\n",
    "    display(tempDF2)\n",
    "    \n",
    "    #Prepare significance labels\n",
    "    ##Retrieve statistical significance\n",
    "    tempS = pvalDF.loc[module, pvalDF.columns.str.contains('-vs-')]#Adjusted P-value\n",
    "    tempS = tempS.loc[tempS.index.str.contains('-fixed')]\n",
    "    tempS.name = 'AdjPval'\n",
    "    ##Clean\n",
    "    tempDF2 = tempS.index.to_series().str.split(pat='-fixed_', expand=True)\n",
    "    tempDF2 = tempDF2.rename(columns={0:'Template', 1:'Comparison'})\n",
    "    tempS1 = tempDF2['Template']\n",
    "    tempDF2 = tempDF2['Comparison'].str.split(pat='-vs-', expand=True)\n",
    "    tempDF2 = tempDF2.rename(columns={0:'Contrast', 1:'Baseline'})\n",
    "    tempDF2 = pd.merge(tempS1, tempDF2, left_index=True, right_index=True, how='left')\n",
    "    tempDF2 = pd.merge(tempDF2, tempS, left_index=True, right_index=True, how='left')\n",
    "    tempDF2['Template'] = tempDF2['Template'].map(tempD0)\n",
    "    tempDF2['Contrast'] = tempDF2['Contrast'].map(tempD0)\n",
    "    tempDF2['Baseline'] = tempDF2['Baseline'].map(tempD0)\n",
    "    ##Convert p-value to label\n",
    "    tempL = []\n",
    "    for row_i in range(len(tempDF2)):\n",
    "        pval = tempDF2['AdjPval'].iloc[row_i]\n",
    "        if pval<0.001:\n",
    "            tempL.append('***')\n",
    "        elif pval<0.01:\n",
    "            tempL.append('**')\n",
    "        elif pval<0.05:\n",
    "            tempL.append('*')\n",
    "        else:\n",
    "            pval_text = Decimal(str(pval)).quantize(Decimal('0.01'), rounding=ROUND_HALF_UP)\n",
    "            tempL.append(r'$P$ = '+str(pval_text))\n",
    "    tempDF2['SignifLabel'] = tempL\n",
    "    display(tempDF2)\n",
    "    \n",
    "    #Visualization\n",
    "    ymax = 1.0\n",
    "    ymin = 0.0\n",
    "    yinter = 0.2\n",
    "    ymargin_t = 0.275\n",
    "    ymargin_b = 0.05\n",
    "    aline_ymin = 1.0\n",
    "    aline_ymargin = 0.125\n",
    "    sns.set(style='ticks', font='Arial', context='talk')\n",
    "    fig, axes = plt.subplots(nrows=1, ncols=4, figsize=(8.8, 5.25), sharex=True, sharey=True,\n",
    "                             gridspec_kw={'width_ratios':[2.2, 2.2, 2.2, 2.2]})\n",
    "    for ax_i, ax in enumerate(axes.flat):\n",
    "        template = list(tempD2.keys())[ax_i]\n",
    "        tempDF3 = tempDF1.loc[tempDF1['Template']==template]\n",
    "        sns.pointplot(data=tempDF3, x='Group', y='RMS', order=list(tempD1.keys()), palette=tempD1,\n",
    "                      markers='o', dodge=False, join=False, capsize=0.6, estimator=np.mean, ci=95, ax=ax)\n",
    "        sns.stripplot(data=tempDF3, x='Group', y='RMS',\n",
    "                      order=list(tempD1.keys()), palette=tempD1, dodge=False, jitter=0.15,\n",
    "                      size=5, edgecolor='black', linewidth=1, **{'marker':'o', 'alpha':0.5}, ax=ax)\n",
    "        #Add RCI line\n",
    "        rci = tempDF3['RMS'].loc[tempDF3['Group']==template].mean()\n",
    "        ax.axhline(y=rci, **{'linestyle':'--', 'color':tempD1[template], 'zorder':0})\n",
    "        #Set axis\n",
    "        sns.despine()\n",
    "        ax.set(ylim=(ymin-ymargin_b, ymax+ymargin_t), yticks=np.arange(ymin, ymax + yinter/10, yinter))\n",
    "        plt.setp(ax.get_xticklabels(), rotation=70, horizontalalignment='right',\n",
    "                 verticalalignment='center', rotation_mode='anchor')\n",
    "        if ax_i==0:\n",
    "            plt.setp(ax, xlabel='', ylabel='Sample RMS')\n",
    "        else:\n",
    "            plt.setp(ax.get_yticklabels(), visible=False)\n",
    "            plt.setp(ax, xlabel='', ylabel='')\n",
    "        #Add significance labels\n",
    "        tempDF3 = tempDF2.loc[tempDF2['Template']==template]\n",
    "        for row_i in range(len(tempDF3)):\n",
    "            #Baseline\n",
    "            group_0 = tempDF3['Baseline'].iloc[row_i]\n",
    "            index_0 = list(tempD1.keys()).index(group_0)\n",
    "            xcoord_0 = index_0\n",
    "            #Contrast\n",
    "            group_1 = tempDF3['Contrast'].iloc[row_i]\n",
    "            index_1 = list(tempD1.keys()).index(group_1)\n",
    "            xcoord_1 = index_1\n",
    "            #Standard point of marker\n",
    "            xcoord = (xcoord_0+xcoord_1)/2\n",
    "            ycoord = aline_ymin + aline_ymargin*row_i\n",
    "            label = tempDF3['SignifLabel'].iloc[row_i]\n",
    "            #Add annotation lines\n",
    "            aline_offset = yinter/5\n",
    "            aline_length = yinter/5 + aline_offset/2\n",
    "            ax.plot([xcoord_0, xcoord_0, xcoord_1, xcoord_1],\n",
    "                    [ycoord+aline_offset, ycoord+aline_length, ycoord+aline_length, ycoord+aline_offset],\n",
    "                    lw=1.5, c='k')\n",
    "            #Add annotation text\n",
    "            if label in ['***', '**', '*']:\n",
    "                text_offset = yinter/21\n",
    "                ax.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                            horizontalalignment='center', verticalalignment='bottom',\n",
    "                            fontsize='medium', color='k')\n",
    "            else:\n",
    "                text_offset = yinter/3.5\n",
    "                ax.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                            horizontalalignment='center', verticalalignment='bottom',\n",
    "                            fontsize='x-small', color='k')\n",
    "        #Add annotation\n",
    "        ax.set_title('Consensus:\\n'+template, {'fontsize':'small'})\n",
    "        xoff = 0.015\n",
    "        yoff = 0.01\n",
    "        rect = plt.Rectangle((xoff, 1), 1-xoff, 0.18,#Manual adjustment\n",
    "                             transform=ax.transAxes, facecolor=tempD2[template],\n",
    "                             clip_on=False, linewidth=0, zorder=0.5)\n",
    "        ax.add_patch(rect)\n",
    "    fig.tight_layout()\n",
    "    #Set title\n",
    "    modulename = moduleDF.loc[module, 'ModuleName']\n",
    "    initial = modulename[0].capitalize()\n",
    "    title = re.sub('^.', initial, modulename)+' ('+module+')'\n",
    "    fig.suptitle(title, size='small',\n",
    "                 verticalalignment='top', horizontalalignment='center', wrap=True, y=0.95)\n",
    "    ##Save\n",
    "    fileDir = './ExportFigures/'\n",
    "    ipynbName = '220522_LC-M001-related-transcriptomics-DIRAC_DIRAC-GOBP_ver2-4_'\n",
    "    fileName = 'fixed-RMSmean-inter-group-comparison_RMS-pointplot-'+module.replace('GO:', 'GO')+'.tif'\n",
    "    plt.gcf().savefig(fileDir+ipynbName+fileName, dpi=300, bbox_inches='tight', pad_inches=0.04,\n",
    "                      pil_kwargs={'compression':'tiff_lzw'})\n",
    "    plt.show()\n",
    "    print('')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 5-5-4. Top changed modules by CR"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Prepare the target module set\n",
    "target = 'Calo-vs-Cont'\n",
    "tempL = pvalDF.loc[pvalDF[target]<0.05].index.tolist()\n",
    "print(len(tempL), 'changed modules with significance in '+target)\n",
    "\n",
    "#Select representatives\n",
    "topX = np.min([30, len(tempL)])\n",
    "topX_plot = np.min([3, len(tempL)])\n",
    "tempDF = statDF.loc[:, statDF.columns.str.contains('Pval$')]\n",
    "tempDF = pd.merge(statDF['ModuleName'], tempDF, left_index=True, right_index=True, how='left')\n",
    "tempDF = tempDF.loc[tempL].sort_values(by=target+'_AdjPval', ascending=True)\n",
    "print('Top', topX, 'modules (sort by significance in '+target+'):')\n",
    "display(tempDF.iloc[:topX])\n",
    "plotL = tempDF.index.tolist()[:topX_plot]\n",
    "\n",
    "#Prepare DF for plot\n",
    "tempDF = rmsDF.reset_index().melt(var_name='SampleID', value_name='RMS', id_vars=['ModuleID', 'Template'])\n",
    "tempDF1 = sampleDF.reset_index()[['SampleID', 'Phenotype']]\n",
    "tempDF = pd.merge(tempDF, tempDF1, on='SampleID', how='left')\n",
    "\n",
    "#Prepare label and color\n",
    "tempD = {'Control':'Control', 'Acarbose':'Acarbose',\n",
    "         'Rapamycin':'Rapamycin', 'Calorie restriction':'CR diet'}\n",
    "tempDF['Template'] = tempDF['Template'].map(tempD)\n",
    "tempDF['Group'] = tempDF['Phenotype'].map(tempD)\n",
    "tempD0 = {'Cont':'Control', 'Acar':'Acarbose',\n",
    "          'Rapa':'Rapamycin', 'Calo':'CR diet'}\n",
    "tempD1 = {'Control':'tab:blue', 'Acarbose':'tab:red',\n",
    "          'Rapamycin':'tab:purple', 'CR diet':'tab:olive'}\n",
    "tempD2 = {}\n",
    "for label in tempD1.keys():\n",
    "    if tempD1[label]=='tab:blue':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(1)\n",
    "    elif tempD1[label]=='tab:orange':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(3)\n",
    "    elif tempD1[label]=='tab:green':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(5)\n",
    "    elif tempD1[label]=='tab:red':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(7)\n",
    "    elif tempD1[label]=='tab:purple':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(9)\n",
    "    elif tempD1[label]=='tab:olive':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(17)\n",
    "    else:\n",
    "        tempD2[label] = 'Error?'\n",
    "\n",
    "#Visualize each representative\n",
    "for rank_i in range(len(plotL)):\n",
    "    print(' - Rank '+str(rank_i+1)+' (sort by significance in '+target+'):')\n",
    "    module = plotL[rank_i]\n",
    "    #Check module summary\n",
    "    tempDF1 = pd.DataFrame(moduleDF.loc[module]).T\n",
    "    display(tempDF1)\n",
    "    \n",
    "    #Select RMS\n",
    "    tempDF1 = tempDF.loc[tempDF['ModuleID']==module]\n",
    "    \n",
    "    #Check RMS summary\n",
    "    tempDF2 = tempDF1.groupby(['Template', 'Group'])['RMS'].agg(['count', 'mean', 'std'])\n",
    "    tempL1 = []\n",
    "    tempL2 = []\n",
    "    for row_n in tempDF2.index.tolist():\n",
    "        count, mean, std = tempDF2.loc[row_n]\n",
    "        tempL1.append(mean - 1.96*std/np.sqrt(count))\n",
    "        tempL2.append(mean + 1.96*std/np.sqrt(count))\n",
    "    tempDF2['0.025'] = tempL1\n",
    "    tempDF2['0.975'] = tempL2\n",
    "    ##Multiindex sort\n",
    "    tempDF2 = tempDF2.reset_index()\n",
    "    tempDF2['Template'] = pd.Categorical(tempDF2['Template'], categories=list(tempD2.keys()))\n",
    "    tempDF2['Group'] = pd.Categorical(tempDF2['Group'], categories=list(tempD1.keys()))\n",
    "    tempDF2 = tempDF2.sort_values(by=['Template', 'Group']).set_index(['Template', 'Group'])\n",
    "    display(tempDF2)\n",
    "    \n",
    "    #Prepare significance labels\n",
    "    ##Retrieve statistical significance\n",
    "    tempS = pvalDF.loc[module, pvalDF.columns.str.contains('-vs-')]#Adjusted P-value\n",
    "    tempS = tempS.loc[tempS.index.str.contains('-fixed')]\n",
    "    tempS.name = 'AdjPval'\n",
    "    ##Clean\n",
    "    tempDF2 = tempS.index.to_series().str.split(pat='-fixed_', expand=True)\n",
    "    tempDF2 = tempDF2.rename(columns={0:'Template', 1:'Comparison'})\n",
    "    tempS1 = tempDF2['Template']\n",
    "    tempDF2 = tempDF2['Comparison'].str.split(pat='-vs-', expand=True)\n",
    "    tempDF2 = tempDF2.rename(columns={0:'Contrast', 1:'Baseline'})\n",
    "    tempDF2 = pd.merge(tempS1, tempDF2, left_index=True, right_index=True, how='left')\n",
    "    tempDF2 = pd.merge(tempDF2, tempS, left_index=True, right_index=True, how='left')\n",
    "    tempDF2['Template'] = tempDF2['Template'].map(tempD0)\n",
    "    tempDF2['Contrast'] = tempDF2['Contrast'].map(tempD0)\n",
    "    tempDF2['Baseline'] = tempDF2['Baseline'].map(tempD0)\n",
    "    ##Convert p-value to label\n",
    "    tempL = []\n",
    "    for row_i in range(len(tempDF2)):\n",
    "        pval = tempDF2['AdjPval'].iloc[row_i]\n",
    "        if pval<0.001:\n",
    "            tempL.append('***')\n",
    "        elif pval<0.01:\n",
    "            tempL.append('**')\n",
    "        elif pval<0.05:\n",
    "            tempL.append('*')\n",
    "        else:\n",
    "            pval_text = Decimal(str(pval)).quantize(Decimal('0.01'), rounding=ROUND_HALF_UP)\n",
    "            tempL.append(r'$P$ = '+str(pval_text))\n",
    "    tempDF2['SignifLabel'] = tempL\n",
    "    display(tempDF2)\n",
    "    \n",
    "    #Visualization\n",
    "    ymax = 1.0\n",
    "    ymin = 0.0\n",
    "    yinter = 0.2\n",
    "    ymargin_t = 0.275\n",
    "    ymargin_b = 0.05\n",
    "    aline_ymin = 1.0\n",
    "    aline_ymargin = 0.125\n",
    "    sns.set(style='ticks', font='Arial', context='talk')\n",
    "    fig, axes = plt.subplots(nrows=1, ncols=4, figsize=(8.8, 5.25), sharex=True, sharey=True,\n",
    "                             gridspec_kw={'width_ratios':[2.2, 2.2, 2.2, 2.2]})\n",
    "    for ax_i, ax in enumerate(axes.flat):\n",
    "        template = list(tempD2.keys())[ax_i]\n",
    "        tempDF3 = tempDF1.loc[tempDF1['Template']==template]\n",
    "        sns.pointplot(data=tempDF3, x='Group', y='RMS', order=list(tempD1.keys()), palette=tempD1,\n",
    "                      markers='o', dodge=False, join=False, capsize=0.6, estimator=np.mean, ci=95, ax=ax)\n",
    "        sns.stripplot(data=tempDF3, x='Group', y='RMS',\n",
    "                      order=list(tempD1.keys()), palette=tempD1, dodge=False, jitter=0.15,\n",
    "                      size=5, edgecolor='black', linewidth=1, **{'marker':'o', 'alpha':0.5}, ax=ax)\n",
    "        #Add RCI line\n",
    "        rci = tempDF3['RMS'].loc[tempDF3['Group']==template].mean()\n",
    "        ax.axhline(y=rci, **{'linestyle':'--', 'color':tempD1[template], 'zorder':0})\n",
    "        #Set axis\n",
    "        sns.despine()\n",
    "        ax.set(ylim=(ymin-ymargin_b, ymax+ymargin_t), yticks=np.arange(ymin, ymax + yinter/10, yinter))\n",
    "        plt.setp(ax.get_xticklabels(), rotation=70, horizontalalignment='right',\n",
    "                 verticalalignment='center', rotation_mode='anchor')\n",
    "        if ax_i==0:\n",
    "            plt.setp(ax, xlabel='', ylabel='Sample RMS')\n",
    "        else:\n",
    "            plt.setp(ax.get_yticklabels(), visible=False)\n",
    "            plt.setp(ax, xlabel='', ylabel='')\n",
    "        #Add significance labels\n",
    "        tempDF3 = tempDF2.loc[tempDF2['Template']==template]\n",
    "        for row_i in range(len(tempDF3)):\n",
    "            #Baseline\n",
    "            group_0 = tempDF3['Baseline'].iloc[row_i]\n",
    "            index_0 = list(tempD1.keys()).index(group_0)\n",
    "            xcoord_0 = index_0\n",
    "            #Contrast\n",
    "            group_1 = tempDF3['Contrast'].iloc[row_i]\n",
    "            index_1 = list(tempD1.keys()).index(group_1)\n",
    "            xcoord_1 = index_1\n",
    "            #Standard point of marker\n",
    "            xcoord = (xcoord_0+xcoord_1)/2\n",
    "            ycoord = aline_ymin + aline_ymargin*row_i\n",
    "            label = tempDF3['SignifLabel'].iloc[row_i]\n",
    "            #Add annotation lines\n",
    "            aline_offset = yinter/5\n",
    "            aline_length = yinter/5 + aline_offset/2\n",
    "            ax.plot([xcoord_0, xcoord_0, xcoord_1, xcoord_1],\n",
    "                    [ycoord+aline_offset, ycoord+aline_length, ycoord+aline_length, ycoord+aline_offset],\n",
    "                    lw=1.5, c='k')\n",
    "            #Add annotation text\n",
    "            if label in ['***', '**', '*']:\n",
    "                text_offset = yinter/21\n",
    "                ax.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                            horizontalalignment='center', verticalalignment='bottom',\n",
    "                            fontsize='medium', color='k')\n",
    "            else:\n",
    "                text_offset = yinter/3.5\n",
    "                ax.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                            horizontalalignment='center', verticalalignment='bottom',\n",
    "                            fontsize='x-small', color='k')\n",
    "        #Add annotation\n",
    "        ax.set_title('Consensus:\\n'+template, {'fontsize':'small'})\n",
    "        xoff = 0.015\n",
    "        yoff = 0.01\n",
    "        rect = plt.Rectangle((xoff, 1), 1-xoff, 0.18,#Manual adjustment\n",
    "                             transform=ax.transAxes, facecolor=tempD2[template],\n",
    "                             clip_on=False, linewidth=0, zorder=0.5)\n",
    "        ax.add_patch(rect)\n",
    "    fig.tight_layout()\n",
    "    #Set title\n",
    "    modulename = moduleDF.loc[module, 'ModuleName']\n",
    "    initial = modulename[0].capitalize()\n",
    "    title = re.sub('^.', initial, modulename)+' ('+module+')'\n",
    "    fig.suptitle(title, size='small',\n",
    "                 verticalalignment='top', horizontalalignment='center', wrap=True, y=0.95)\n",
    "    ##Save\n",
    "    fileDir = './ExportFigures/'\n",
    "    ipynbName = '220522_LC-M001-related-transcriptomics-DIRAC_DIRAC-GOBP_ver2-4_'\n",
    "    fileName = 'fixed-RMSmean-inter-group-comparison_RMS-pointplot-'+module.replace('GO:', 'GO')+'.tif'\n",
    "    plt.gcf().savefig(fileDir+ipynbName+fileName, dpi=300, bbox_inches='tight', pad_inches=0.04,\n",
    "                      pil_kwargs={'compression':'tiff_lzw'})\n",
    "    plt.show()\n",
    "    print('')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 5-5-5. Top RCI modules in Ctrl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "target = 'Control'\n",
    "\n",
    "#Select representatives\n",
    "topX = 30\n",
    "topX_plot = 3\n",
    "tempDF = statDF.loc[:, statDF.columns.str.contains('Pval$')]\n",
    "tempDF = pd.merge(statDF['ModuleName'], tempDF, left_index=True, right_index=True, how='left')\n",
    "tempDF = pd.merge(tempDF, rciDF_kk, left_index=True, right_index=True, how='left')\n",
    "tempDF = tempDF.sort_values(by=target, ascending=False)\n",
    "print('Top', topX, 'modules (sort by RCI of '+target+'):')\n",
    "display(tempDF.iloc[:topX])\n",
    "plotL = tempDF.index.tolist()[:topX_plot]\n",
    "\n",
    "#Prepare DF for plot\n",
    "tempDF = rmsDF.reset_index().melt(var_name='SampleID', value_name='RMS', id_vars=['ModuleID', 'Template'])\n",
    "tempDF1 = sampleDF.reset_index()[['SampleID', 'Phenotype']]\n",
    "tempDF = pd.merge(tempDF, tempDF1, on='SampleID', how='left')\n",
    "\n",
    "#Prepare label and color\n",
    "tempD = {'Control':'Control', 'Acarbose':'Acarbose',\n",
    "         'Rapamycin':'Rapamycin', 'Calorie restriction':'CR diet'}\n",
    "tempDF['Template'] = tempDF['Template'].map(tempD)\n",
    "tempDF['Group'] = tempDF['Phenotype'].map(tempD)\n",
    "tempD0 = {'Cont':'Control', 'Acar':'Acarbose',\n",
    "          'Rapa':'Rapamycin', 'Calo':'CR diet'}\n",
    "tempD1 = {'Control':'tab:blue', 'Acarbose':'tab:red',\n",
    "          'Rapamycin':'tab:purple', 'CR diet':'tab:olive'}\n",
    "tempD2 = {}\n",
    "for label in tempD1.keys():\n",
    "    if tempD1[label]=='tab:blue':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(1)\n",
    "    elif tempD1[label]=='tab:orange':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(3)\n",
    "    elif tempD1[label]=='tab:green':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(5)\n",
    "    elif tempD1[label]=='tab:red':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(7)\n",
    "    elif tempD1[label]=='tab:purple':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(9)\n",
    "    elif tempD1[label]=='tab:olive':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(17)\n",
    "    else:\n",
    "        tempD2[label] = 'Error?'\n",
    "\n",
    "#Re-prepare p-values because these modules can be filtered out in pvalDF\n",
    "tempDF4 = statDF.loc[:, statDF.columns.str.contains('-vs-.*_AdjPval$')]\n",
    "tempDF4.columns = tempDF4.columns.str.replace('_AdjPval$', '')\n",
    "for template in ['Acar-fixed', 'Rapa-fixed', 'Calo-fixed']:\n",
    "    #Re-import the summary table fixed consensus version\n",
    "    fileDir = './ExportData/'\n",
    "    ipynbName = '220522_LC-M001-related-transcriptomics-DIRAC_StatisticalTest-GOBP_ver2-4_'\n",
    "    fileName = 'inter-group-comparison.xlsx'\n",
    "    sheetName = template+'-RMSmean'\n",
    "    tempDF1 = pd.read_excel(fileDir+ipynbName+fileName, sheet_name=sheetName, engine='openpyxl')\n",
    "    tempDF1 = tempDF1.set_index('ModuleID')\n",
    "    #Take adjusted P-value\n",
    "    tempDF1 = tempDF1.loc[:, tempDF1.columns.str.contains('-vs-.*_AdjPval$')]\n",
    "    tempDF1.columns = tempDF1.columns.str.replace('_AdjPval$', '')\n",
    "    tempDF1.columns = template+'_'+tempDF1.columns\n",
    "    #Update\n",
    "    tempDF4 = pd.merge(tempDF4, tempDF1, left_index=True, right_index=True, how='left')\n",
    "\n",
    "#Visualize each representative\n",
    "for rank_i in range(len(plotL)):\n",
    "    print(' - Rank '+str(rank_i+1)+' (sort by RCI of '+target+'):')\n",
    "    module = plotL[rank_i]\n",
    "    #Check module summary\n",
    "    tempDF1 = pd.DataFrame(moduleDF.loc[module]).T\n",
    "    display(tempDF1)\n",
    "    \n",
    "    #Select RMS\n",
    "    tempDF1 = tempDF.loc[tempDF['ModuleID']==module]\n",
    "    \n",
    "    #Check RMS summary\n",
    "    tempDF2 = tempDF1.groupby(['Template', 'Group'])['RMS'].agg(['count', 'mean', 'std'])\n",
    "    tempL1 = []\n",
    "    tempL2 = []\n",
    "    for row_n in tempDF2.index.tolist():\n",
    "        count, mean, std = tempDF2.loc[row_n]\n",
    "        tempL1.append(mean - 1.96*std/np.sqrt(count))\n",
    "        tempL2.append(mean + 1.96*std/np.sqrt(count))\n",
    "    tempDF2['0.025'] = tempL1\n",
    "    tempDF2['0.975'] = tempL2\n",
    "    ##Multiindex sort\n",
    "    tempDF2 = tempDF2.reset_index()\n",
    "    tempDF2['Template'] = pd.Categorical(tempDF2['Template'], categories=list(tempD2.keys()))\n",
    "    tempDF2['Group'] = pd.Categorical(tempDF2['Group'], categories=list(tempD1.keys()))\n",
    "    tempDF2 = tempDF2.sort_values(by=['Template', 'Group']).set_index(['Template', 'Group'])\n",
    "    display(tempDF2)\n",
    "    \n",
    "    #Prepare significance labels\n",
    "    ##Re-prepare p-values because these modules can be filtered out in pvalDF\n",
    "    ##-> Moved to outside of for-loop\n",
    "    ##Retrieve statistical significance\n",
    "    tempS = tempDF4.loc[module]\n",
    "    tempS = tempS.loc[tempS.index.str.contains('-fixed')]\n",
    "    tempS.name = 'AdjPval'\n",
    "    ##Clean\n",
    "    tempDF2 = tempS.index.to_series().str.split(pat='-fixed_', expand=True)\n",
    "    tempDF2 = tempDF2.rename(columns={0:'Template', 1:'Comparison'})\n",
    "    tempS1 = tempDF2['Template']\n",
    "    tempDF2 = tempDF2['Comparison'].str.split(pat='-vs-', expand=True)\n",
    "    tempDF2 = tempDF2.rename(columns={0:'Contrast', 1:'Baseline'})\n",
    "    tempDF2 = pd.merge(tempS1, tempDF2, left_index=True, right_index=True, how='left')\n",
    "    tempDF2 = pd.merge(tempDF2, tempS, left_index=True, right_index=True, how='left')\n",
    "    tempDF2['Template'] = tempDF2['Template'].map(tempD0)\n",
    "    tempDF2['Contrast'] = tempDF2['Contrast'].map(tempD0)\n",
    "    tempDF2['Baseline'] = tempDF2['Baseline'].map(tempD0)\n",
    "    ##Convert p-value to label\n",
    "    tempL = []\n",
    "    for row_i in range(len(tempDF2)):\n",
    "        pval = tempDF2['AdjPval'].iloc[row_i]\n",
    "        if pval<0.001:\n",
    "            tempL.append('***')\n",
    "        elif pval<0.01:\n",
    "            tempL.append('**')\n",
    "        elif pval<0.05:\n",
    "            tempL.append('*')\n",
    "        else:\n",
    "            pval_text = Decimal(str(pval)).quantize(Decimal('0.01'), rounding=ROUND_HALF_UP)\n",
    "            tempL.append(r'$P$ = '+str(pval_text))\n",
    "    tempDF2['SignifLabel'] = tempL\n",
    "    display(tempDF2)\n",
    "    \n",
    "    #Visualization\n",
    "    ymax = 1.0\n",
    "    ymin = 0.0\n",
    "    yinter = 0.2\n",
    "    ymargin_t = 0.275\n",
    "    ymargin_b = 0.05\n",
    "    aline_ymin = 1.0\n",
    "    aline_ymargin = 0.125\n",
    "    sns.set(style='ticks', font='Arial', context='talk')\n",
    "    fig, axes = plt.subplots(nrows=1, ncols=4, figsize=(8.8, 5.25), sharex=True, sharey=True,\n",
    "                             gridspec_kw={'width_ratios':[2.2, 2.2, 2.2, 2.2]})\n",
    "    for ax_i, ax in enumerate(axes.flat):\n",
    "        template = list(tempD2.keys())[ax_i]\n",
    "        tempDF3 = tempDF1.loc[tempDF1['Template']==template]\n",
    "        sns.pointplot(data=tempDF3, x='Group', y='RMS', order=list(tempD1.keys()), palette=tempD1,\n",
    "                      markers='o', dodge=False, join=False, capsize=0.6, estimator=np.mean, ci=95, ax=ax)\n",
    "        sns.stripplot(data=tempDF3, x='Group', y='RMS',\n",
    "                      order=list(tempD1.keys()), palette=tempD1, dodge=False, jitter=0.15,\n",
    "                      size=5, edgecolor='black', linewidth=1, **{'marker':'o', 'alpha':0.5}, ax=ax)\n",
    "        #Add RCI line\n",
    "        rci = tempDF3['RMS'].loc[tempDF3['Group']==template].mean()\n",
    "        ax.axhline(y=rci, **{'linestyle':'--', 'color':tempD1[template], 'zorder':0})\n",
    "        #Set axis\n",
    "        sns.despine()\n",
    "        ax.set(ylim=(ymin-ymargin_b, ymax+ymargin_t), yticks=np.arange(ymin, ymax + yinter/10, yinter))\n",
    "        plt.setp(ax.get_xticklabels(), rotation=70, horizontalalignment='right',\n",
    "                 verticalalignment='center', rotation_mode='anchor')\n",
    "        if ax_i==0:\n",
    "            plt.setp(ax, xlabel='', ylabel='Sample RMS')\n",
    "        else:\n",
    "            plt.setp(ax.get_yticklabels(), visible=False)\n",
    "            plt.setp(ax, xlabel='', ylabel='')\n",
    "        #Add significance labels\n",
    "        tempDF3 = tempDF2.loc[tempDF2['Template']==template]\n",
    "        for row_i in range(len(tempDF3)):\n",
    "            #Baseline\n",
    "            group_0 = tempDF3['Baseline'].iloc[row_i]\n",
    "            index_0 = list(tempD1.keys()).index(group_0)\n",
    "            xcoord_0 = index_0\n",
    "            #Contrast\n",
    "            group_1 = tempDF3['Contrast'].iloc[row_i]\n",
    "            index_1 = list(tempD1.keys()).index(group_1)\n",
    "            xcoord_1 = index_1\n",
    "            #Standard point of marker\n",
    "            xcoord = (xcoord_0+xcoord_1)/2\n",
    "            ycoord = aline_ymin + aline_ymargin*row_i\n",
    "            label = tempDF3['SignifLabel'].iloc[row_i]\n",
    "            #Add annotation lines\n",
    "            aline_offset = yinter/5\n",
    "            aline_length = yinter/5 + aline_offset/2\n",
    "            ax.plot([xcoord_0, xcoord_0, xcoord_1, xcoord_1],\n",
    "                    [ycoord+aline_offset, ycoord+aline_length, ycoord+aline_length, ycoord+aline_offset],\n",
    "                    lw=1.5, c='k')\n",
    "            #Add annotation text\n",
    "            if label in ['***', '**', '*']:\n",
    "                text_offset = yinter/21\n",
    "                ax.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                            horizontalalignment='center', verticalalignment='bottom',\n",
    "                            fontsize='medium', color='k')\n",
    "            else:\n",
    "                text_offset = yinter/3.5\n",
    "                ax.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                            horizontalalignment='center', verticalalignment='bottom',\n",
    "                            fontsize='x-small', color='k')\n",
    "        #Add annotation\n",
    "        ax.set_title('Consensus:\\n'+template, {'fontsize':'small'})\n",
    "        xoff = 0.015\n",
    "        yoff = 0.01\n",
    "        rect = plt.Rectangle((xoff, 1), 1-xoff, 0.18,#Manual adjustment\n",
    "                             transform=ax.transAxes, facecolor=tempD2[template],\n",
    "                             clip_on=False, linewidth=0, zorder=0.5)\n",
    "        ax.add_patch(rect)\n",
    "    fig.tight_layout()\n",
    "    #Set title\n",
    "    modulename = moduleDF.loc[module, 'ModuleName']\n",
    "    initial = modulename[0].capitalize()\n",
    "    title = re.sub('^.', initial, modulename)+' ('+module+')'\n",
    "    fig.suptitle(title, size='small',\n",
    "                 verticalalignment='top', horizontalalignment='center', wrap=True, y=0.95)\n",
    "    ##Save\n",
    "    fileDir = './ExportFigures/'\n",
    "    ipynbName = '220522_LC-M001-related-transcriptomics-DIRAC_DIRAC-GOBP_ver2-4_'\n",
    "    fileName = 'fixed-RMSmean-inter-group-comparison_RMS-pointplot-'+module.replace('GO:', 'GO')+'.tif'\n",
    "    plt.gcf().savefig(fileDir+ipynbName+fileName, dpi=300, bbox_inches='tight', pad_inches=0.04,\n",
    "                      pil_kwargs={'compression':'tiff_lzw'})\n",
    "    plt.show()\n",
    "    print('')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 5-5-6. Modules of special interest"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plotL = ['GO:0006635', 'GO:0031998', 'GO:0016558', 'GO:0006625',#Common in M001\n",
    "         'GO:0006637', 'GO:0006734', 'GO:0001561',#Related to fatty acid oxidation (mentioned in previous manuscript)\n",
    "         'GO:0023035', 'GO:0039529',#Immune response (mentioned in previous manuscript)\n",
    "         'GO:0010638', 'GO:0002181', 'GO:0045899',#Mentioned in previous manuscript\n",
    "         'GO:0006635', 'GO:1990126', 'GO:0098761',#Common in proteomics and transcriptomics\n",
    "         'GO:0061732', 'GO:0006086', 'GO:0019441',#Specific to proteomics (Aca)\n",
    "         'GO:0031998', 'GO:0044794', 'GO:0006625', 'GO:0016558', 'GO:0019441', 'GO:0033572',#Specific to proteomics (Rapa)\n",
    "         'GO:0006635', 'GO:0002181', 'GO:0000028', 'GO:0016558', 'GO:0006734', 'GO:0098761',#Common in M001 + M004\n",
    "         'GO:0015986', 'GO:0042776', 'GO:0070934', 'GO:0006703', 'GO:0034354', 'GO:0045899']#Similar to 4EGI-1\n",
    "plotL = list(set(plotL))\n",
    "plotL.sort()\n",
    "for module in plotL:\n",
    "    if module not in moduleDF.index.tolist():\n",
    "        print(module+' was NOT included in this analysis.')\n",
    "        plotL.remove(module)\n",
    "\n",
    "#Prepare the target module set\n",
    "tempDF = statDF.loc[:, statDF.columns.str.contains('Pval$')]\n",
    "tempDF = pd.merge(statDF['ModuleName'], tempDF, left_index=True, right_index=True, how='left')\n",
    "tempDF = tempDF.loc[plotL]\n",
    "display(tempDF)\n",
    "\n",
    "#Prepare DF for plot\n",
    "tempDF = rmsDF.reset_index().melt(var_name='SampleID', value_name='RMS', id_vars=['ModuleID', 'Template'])\n",
    "tempDF1 = sampleDF.reset_index()[['SampleID', 'Phenotype']]\n",
    "tempDF = pd.merge(tempDF, tempDF1, on='SampleID', how='left')\n",
    "\n",
    "#Prepare label and color\n",
    "tempD = {'Control':'Control', 'Acarbose':'Acarbose',\n",
    "         'Rapamycin':'Rapamycin', 'Calorie restriction':'CR diet'}\n",
    "tempDF['Template'] = tempDF['Template'].map(tempD)\n",
    "tempDF['Group'] = tempDF['Phenotype'].map(tempD)\n",
    "tempD0 = {'Cont':'Control', 'Acar':'Acarbose',\n",
    "          'Rapa':'Rapamycin', 'Calo':'CR diet'}\n",
    "tempD1 = {'Control':'tab:blue', 'Acarbose':'tab:red',\n",
    "          'Rapamycin':'tab:purple', 'CR diet':'tab:olive'}\n",
    "tempD2 = {}\n",
    "for label in tempD1.keys():\n",
    "    if tempD1[label]=='tab:blue':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(1)\n",
    "    elif tempD1[label]=='tab:orange':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(3)\n",
    "    elif tempD1[label]=='tab:green':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(5)\n",
    "    elif tempD1[label]=='tab:red':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(7)\n",
    "    elif tempD1[label]=='tab:purple':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(9)\n",
    "    elif tempD1[label]=='tab:olive':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(17)\n",
    "    else:\n",
    "        tempD2[label] = 'Error?'\n",
    "\n",
    "#Re-prepare p-values because these modules can be filtered out in pvalDF\n",
    "tempDF4 = statDF.loc[:, statDF.columns.str.contains('-vs-.*_AdjPval$')]\n",
    "tempDF4.columns = tempDF4.columns.str.replace('_AdjPval$', '')\n",
    "for template in ['Acar-fixed', 'Rapa-fixed', 'Calo-fixed']:\n",
    "    #Re-import the summary table fixed consensus version\n",
    "    fileDir = './ExportData/'\n",
    "    ipynbName = '220522_LC-M001-related-transcriptomics-DIRAC_StatisticalTest-GOBP_ver2-4_'\n",
    "    fileName = 'inter-group-comparison.xlsx'\n",
    "    sheetName = template+'-RMSmean'\n",
    "    tempDF1 = pd.read_excel(fileDir+ipynbName+fileName, sheet_name=sheetName, engine='openpyxl')\n",
    "    tempDF1 = tempDF1.set_index('ModuleID')\n",
    "    #Take adjusted P-value\n",
    "    tempDF1 = tempDF1.loc[:, tempDF1.columns.str.contains('-vs-.*_AdjPval$')]\n",
    "    tempDF1.columns = tempDF1.columns.str.replace('_AdjPval$', '')\n",
    "    tempDF1.columns = template+'_'+tempDF1.columns\n",
    "    #Update\n",
    "    tempDF4 = pd.merge(tempDF4, tempDF1, left_index=True, right_index=True, how='left')\n",
    "\n",
    "#Visualize each representative\n",
    "for rank_i in range(len(plotL)):\n",
    "    print(' - Module '+str(rank_i+1)+':')\n",
    "    module = plotL[rank_i]\n",
    "    #Check module summary\n",
    "    tempDF1 = pd.DataFrame(moduleDF.loc[module]).T\n",
    "    display(tempDF1)\n",
    "    \n",
    "    #Select RMS\n",
    "    tempDF1 = tempDF.loc[tempDF['ModuleID']==module]\n",
    "    \n",
    "    #Check RMS summary\n",
    "    tempDF2 = tempDF1.groupby(['Template', 'Group'])['RMS'].agg(['count', 'mean', 'std'])\n",
    "    tempL1 = []\n",
    "    tempL2 = []\n",
    "    for row_n in tempDF2.index.tolist():\n",
    "        count, mean, std = tempDF2.loc[row_n]\n",
    "        tempL1.append(mean - 1.96*std/np.sqrt(count))\n",
    "        tempL2.append(mean + 1.96*std/np.sqrt(count))\n",
    "    tempDF2['0.025'] = tempL1\n",
    "    tempDF2['0.975'] = tempL2\n",
    "    ##Multiindex sort\n",
    "    tempDF2 = tempDF2.reset_index()\n",
    "    tempDF2['Template'] = pd.Categorical(tempDF2['Template'], categories=list(tempD2.keys()))\n",
    "    tempDF2['Group'] = pd.Categorical(tempDF2['Group'], categories=list(tempD1.keys()))\n",
    "    tempDF2 = tempDF2.sort_values(by=['Template', 'Group']).set_index(['Template', 'Group'])\n",
    "    display(tempDF2)\n",
    "    \n",
    "    #Prepare significance labels\n",
    "    ##Re-prepare p-values because these modules can be filtered out in pvalDF\n",
    "    ##-> Moved to outside of for-loop\n",
    "    ##Retrieve statistical significance\n",
    "    tempS = tempDF4.loc[module]\n",
    "    tempS = tempS.loc[tempS.index.str.contains('-fixed')]\n",
    "    tempS.name = 'AdjPval'\n",
    "    ##Clean\n",
    "    tempDF2 = tempS.index.to_series().str.split(pat='-fixed_', expand=True)\n",
    "    tempDF2 = tempDF2.rename(columns={0:'Template', 1:'Comparison'})\n",
    "    tempS1 = tempDF2['Template']\n",
    "    tempDF2 = tempDF2['Comparison'].str.split(pat='-vs-', expand=True)\n",
    "    tempDF2 = tempDF2.rename(columns={0:'Contrast', 1:'Baseline'})\n",
    "    tempDF2 = pd.merge(tempS1, tempDF2, left_index=True, right_index=True, how='left')\n",
    "    tempDF2 = pd.merge(tempDF2, tempS, left_index=True, right_index=True, how='left')\n",
    "    tempDF2['Template'] = tempDF2['Template'].map(tempD0)\n",
    "    tempDF2['Contrast'] = tempDF2['Contrast'].map(tempD0)\n",
    "    tempDF2['Baseline'] = tempDF2['Baseline'].map(tempD0)\n",
    "    ##Convert p-value to label\n",
    "    tempL = []\n",
    "    for row_i in range(len(tempDF2)):\n",
    "        pval = tempDF2['AdjPval'].iloc[row_i]\n",
    "        if pval<0.001:\n",
    "            tempL.append('***')\n",
    "        elif pval<0.01:\n",
    "            tempL.append('**')\n",
    "        elif pval<0.05:\n",
    "            tempL.append('*')\n",
    "        else:\n",
    "            pval_text = Decimal(str(pval)).quantize(Decimal('0.01'), rounding=ROUND_HALF_UP)\n",
    "            tempL.append(r'$P$ = '+str(pval_text))\n",
    "    tempDF2['SignifLabel'] = tempL\n",
    "    display(tempDF2)\n",
    "    \n",
    "    #Visualization\n",
    "    ymax = 1.0\n",
    "    ymin = 0.0\n",
    "    yinter = 0.2\n",
    "    ymargin_t = 0.275\n",
    "    ymargin_b = 0.05\n",
    "    aline_ymin = 1.0\n",
    "    aline_ymargin = 0.125\n",
    "    sns.set(style='ticks', font='Arial', context='talk')\n",
    "    fig, axes = plt.subplots(nrows=1, ncols=4, figsize=(8.8, 5.25), sharex=True, sharey=True,\n",
    "                             gridspec_kw={'width_ratios':[2.2, 2.2, 2.2, 2.2]})\n",
    "    for ax_i, ax in enumerate(axes.flat):\n",
    "        template = list(tempD2.keys())[ax_i]\n",
    "        tempDF3 = tempDF1.loc[tempDF1['Template']==template]\n",
    "        sns.pointplot(data=tempDF3, x='Group', y='RMS', order=list(tempD1.keys()), palette=tempD1,\n",
    "                      markers='o', dodge=False, join=False, capsize=0.6, estimator=np.mean, ci=95, ax=ax)\n",
    "        sns.stripplot(data=tempDF3, x='Group', y='RMS',\n",
    "                      order=list(tempD1.keys()), palette=tempD1, dodge=False, jitter=0.15,\n",
    "                      size=5, edgecolor='black', linewidth=1, **{'marker':'o', 'alpha':0.5}, ax=ax)\n",
    "        #Add RCI line\n",
    "        rci = tempDF3['RMS'].loc[tempDF3['Group']==template].mean()\n",
    "        ax.axhline(y=rci, **{'linestyle':'--', 'color':tempD1[template], 'zorder':0})\n",
    "        #Set axis\n",
    "        sns.despine()\n",
    "        ax.set(ylim=(ymin-ymargin_b, ymax+ymargin_t), yticks=np.arange(ymin, ymax + yinter/10, yinter))\n",
    "        plt.setp(ax.get_xticklabels(), rotation=70, horizontalalignment='right',\n",
    "                 verticalalignment='center', rotation_mode='anchor')\n",
    "        if ax_i==0:\n",
    "            plt.setp(ax, xlabel='', ylabel='Sample RMS')\n",
    "        else:\n",
    "            plt.setp(ax.get_yticklabels(), visible=False)\n",
    "            plt.setp(ax, xlabel='', ylabel='')\n",
    "        #Add significance labels\n",
    "        tempDF3 = tempDF2.loc[tempDF2['Template']==template]\n",
    "        for row_i in range(len(tempDF3)):\n",
    "            #Baseline\n",
    "            group_0 = tempDF3['Baseline'].iloc[row_i]\n",
    "            index_0 = list(tempD1.keys()).index(group_0)\n",
    "            xcoord_0 = index_0\n",
    "            #Contrast\n",
    "            group_1 = tempDF3['Contrast'].iloc[row_i]\n",
    "            index_1 = list(tempD1.keys()).index(group_1)\n",
    "            xcoord_1 = index_1\n",
    "            #Standard point of marker\n",
    "            xcoord = (xcoord_0+xcoord_1)/2\n",
    "            ycoord = aline_ymin + aline_ymargin*row_i\n",
    "            label = tempDF3['SignifLabel'].iloc[row_i]\n",
    "            #Add annotation lines\n",
    "            aline_offset = yinter/5\n",
    "            aline_length = yinter/5 + aline_offset/2\n",
    "            ax.plot([xcoord_0, xcoord_0, xcoord_1, xcoord_1],\n",
    "                    [ycoord+aline_offset, ycoord+aline_length, ycoord+aline_length, ycoord+aline_offset],\n",
    "                    lw=1.5, c='k')\n",
    "            #Add annotation text\n",
    "            if label in ['***', '**', '*']:\n",
    "                text_offset = yinter/21\n",
    "                ax.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                            horizontalalignment='center', verticalalignment='bottom',\n",
    "                            fontsize='medium', color='k')\n",
    "            else:\n",
    "                text_offset = yinter/3.5\n",
    "                ax.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                            horizontalalignment='center', verticalalignment='bottom',\n",
    "                            fontsize='x-small', color='k')\n",
    "        #Add annotation\n",
    "        ax.set_title('Consensus:\\n'+template, {'fontsize':'small'})\n",
    "        xoff = 0.015\n",
    "        yoff = 0.01\n",
    "        rect = plt.Rectangle((xoff, 1), 1-xoff, 0.18,#Manual adjustment\n",
    "                             transform=ax.transAxes, facecolor=tempD2[template],\n",
    "                             clip_on=False, linewidth=0, zorder=0.5)\n",
    "        ax.add_patch(rect)\n",
    "    fig.tight_layout()\n",
    "    #Set title\n",
    "    modulename = moduleDF.loc[module, 'ModuleName']\n",
    "    initial = modulename[0].capitalize()\n",
    "    title = re.sub('^.', initial, modulename)+' ('+module+')'\n",
    "    fig.suptitle(title, size='small',\n",
    "                 verticalalignment='top', horizontalalignment='center', wrap=True, y=0.95)\n",
    "    ##Save\n",
    "    fileDir = './ExportFigures/'\n",
    "    ipynbName = '220522_LC-M001-related-transcriptomics-DIRAC_DIRAC-GOBP_ver2-4_'\n",
    "    fileName = 'fixed-RMSmean-inter-group-comparison_RMS-pointplot-'+module.replace('GO:', 'GO')+'.tif'\n",
    "    plt.gcf().savefig(fileDir+ipynbName+fileName, dpi=300, bbox_inches='tight', pad_inches=0.04,\n",
    "                      pil_kwargs={'compression':'tiff_lzw'})\n",
    "    plt.show()\n",
    "    print('')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# — End of notebook —"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "arivale-py3 - Python",
   "language": "python",
   "name": "conda-env-arivale-py3-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  },
  "widgets": {
   "application/vnd.jupyter.widget-state+json": {
    "state": {},
    "version_major": 2,
    "version_minor": 0
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
