{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# DIRAC Analysis of LC M004 Proteomics with LC M001 Proteomics — DIRAC with GOBP Modules"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***by Kengo Watanabe***  \n",
    "\n",
    "In this notebook, the differential rank conservation (DIRAC; Eddy, J.A. et al. PLoS Comput. Biol. 2010) analysis is performed on the preprocessed Longevity Consortium (LC) M001 and M004 proteomics datasets (analytes detected in all samples; sample-based robust Z-score followed by analyte-based robust Z-score) using the retrieved a priori module set (Gene Ontology (Biological Process) derived by EMBL-EBI QuickGO API; ≥4 analytes and ≥50% coverage). To directly compare the DIRAC results between M001 and M004, each rank consensus is utilized across the datasets.  \n",
    "> To maintain the consistency with the other DIRAC analyses, statistical tests are performed in a different notebook with R kernel.  \n",
    "\n",
    "Input:  \n",
    "* Preprocessed analyte data (M001): 210126_LCprotomics-M001-DIRAC-ver6_preprocessing_cleaned-robustZscored-data.tsv  \n",
    "* Preprocessed analyte data (M004): 210125_LCprotomics-M004-DIRAC_preprocessing_cleaned-robustZscored-data.tsv  \n",
    "* Module–analyte metadata: 220525_LCproteomics-M004-DIRAC_Preprocessing-with-M001_ver2-2_module-metadata_QuickGO-GOBP-min-n4-cov50.tsv  \n",
    "* Analyte metadata: 220525_LCproteomics-M004-DIRAC_Preprocessing-with-M001_ver2-2_analyte-metadata_UniProt.tsv  \n",
    "* Sample–mouse metadata (M001): 210126_LCprotomics-M001-DIRAC-ver6_preprocessing_metadata-sample.tsv  \n",
    "* Sample–mouse metadata (M004): 210125_LCprotomics-M004-DIRAC_preprocessing_metadata-sample.tsv  \n",
    "* Statistical test summary: 220529_LCproteomics-M004-DIRAC_StatisticalTest-GOBP-with-M001_ver2-3_inter-group-comparison.xlsx (Supplementary Data 7)  \n",
    "\n",
    "Output:  \n",
    "* Cleaned module metadata, which is incorporated into Supplementary Data 7 in R sub-notebook  \n",
    "* Combined sample–mouse metadata, which is used in R sub-notebook  \n",
    "* DIRAC measures, which are used in statistical analysis (R sub-notebook)  \n",
    "* Figure 5e–g  \n",
    "* Supplementary Figure 3e, f, h, i  \n",
    "\n",
    "Original notebook (memo for my future tracing):  \n",
    "* dalek:[JupyterLab HOME]/210125_LCproteomics-M004-DIRAC/220529_LCproteomics-M004-DIRAC_DIRAC-GOBP-with-M001_ver2-3.ipynb  \n",
    "* dalek:[JupyterLab HOME]/210125_LCproteomics-M004-DIRAC/220606_LCproteomics-M004-DIRAC-ver2-3_Supplement.ipynb  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scipy.stats as stats\n",
    "import matplotlib.pyplot as plt\n",
    "%matplotlib inline\n",
    "import seaborn as sns\n",
    "#For Arial font\n",
    "#!conda install -c conda-forge -y mscorefonts\n",
    "##-> The below was also needed in matplotlib 3.4.2\n",
    "#import shutil\n",
    "#import matplotlib\n",
    "#shutil.rmtree(matplotlib.get_cachedir())\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')\n",
    "from IPython.display import display\n",
    "import time\n",
    "\n",
    "from itertools import combinations\n",
    "import math\n",
    "from multiprocessing import Pool\n",
    "from decimal import Decimal, ROUND_HALF_UP\n",
    "import re\n",
    "import matplotlib.patches as mpatches\n",
    "#!pip install venn\n",
    "from venn import venn\n",
    "#!conda install -c conda-forge -y matplotlib-venn\n",
    "from matplotlib_venn import venn3, venn3_circles, venn2, venn2_circles\n",
    "from textwrap import wrap\n",
    "\n",
    "!conda list"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 0. DIRAC code"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> The original code for DIRAC was written in MATLAB. Hence, it is re-written in Python 3 here.  \n",
    "> <– I don't care about computational cost here; rather, the code adheres to the story in the original paper.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def network_ranking(DF, networkS):\n",
    "    # This function calculates the pairwise ordering of network genes (i.e., network ranking).\n",
    "    ## Ref. Eddy, J. A. et al. PLoS Comput. Biol. 2010 (Figure 1 at a glance)\n",
    "    # Requirements:\n",
    "    ## import numpy as np: confirmed with versions 1.17.5 and 1.21.1\n",
    "    ## import pandas as pd: confirmed with versions 0.25.3 and 1.3.1\n",
    "    ## from itertools import combinations: confirmed with Python 3.7.6 and 3.9.6\n",
    "    # Input:\n",
    "    ## DF: pd.DataFrame containing expression values (X_gn) with gene (g; 1-G) indices and sample (n; 1-N) columns\n",
    "    ## networkS: pd.Series containing genes (g; 1-G) with network (m; 1-M) indices (i.e., 1-on-1 long-format)\n",
    "    # Output:\n",
    "    ## pd.DataFrame containing binary values of the networking ranking comparison for each sample\n",
    "    ## -> row: comparison_id (indicated by network (m) and ordering (g_i < g_j) columns)\n",
    "    ## -> column: sample (n) (with exception of the NetworkID and Ordering columns)\n",
    "    # Note:\n",
    "    ## If items in network m and gene g contain ' : ' or ' < ', this code would produce error or unexpected output.\n",
    "    ## If an item in sample n is 'NetworkID' or 'Ordering', this code would produce error or unexpected output.\n",
    "    \n",
    "    #Calculate binary values of the network ranking comparison for each sample\n",
    "    rankDF = pd.DataFrame()\n",
    "    sampleL = DF.columns.tolist()\n",
    "    networkL = networkS.index.unique()\n",
    "    for n in sampleL:\n",
    "        rankDF_n = pd.DataFrame()\n",
    "        for m in networkL:\n",
    "            #Pairs of genes (g_i, g_j) in the network m\n",
    "            networkS_m = networkS.loc[m]\n",
    "            pairL_m = list(combinations(range(0, len(networkS_m)), 2))\n",
    "            pairL_m_i, pairL_m_j = [[pair[x] for pair in pairL_m] for x in (0,1)]\n",
    "            rankDF_m = pd.DataFrame({'g_i':networkS_m.iloc[pairL_m_i],\n",
    "                                     'g_j':networkS_m.iloc[pairL_m_j]})#Hold network (m; 1-M) index\n",
    "            #Compare the expression values (X_gn) between pairwise genes (g_i vs. g_j)\n",
    "            tempL = []\n",
    "            for pair_i in range(0, len(rankDF_m)):\n",
    "                g_i = rankDF_m.iloc[pair_i, 0]\n",
    "                g_j = rankDF_m.iloc[pair_i, 1]\n",
    "                X_i = DF.loc[g_i, n]\n",
    "                X_j = DF.loc[g_j, n]\n",
    "                #If X_i < X_j is true, add 1; otherwise (X_i >= X_j), add 0\n",
    "                if X_i < X_j:\n",
    "                    tempL.append(1)\n",
    "                else:\n",
    "                    tempL.append(0)\n",
    "            rankDF_m['X_i<X_j'] = tempL\n",
    "            #Update the network ranking dataframe of sample n\n",
    "            rankDF_m.index.set_names('NetworkID', inplace=True)#Set/reset index name\n",
    "            rankDF_m = rankDF_m.reset_index()\n",
    "            rankDF_n = pd.concat([rankDF_n, rankDF_m], axis=0)\n",
    "        #Updata the network ranking dataframe of all samples\n",
    "        rankDF_n['Sample'] = n\n",
    "        rankDF = pd.concat([rankDF, rankDF_n], axis=0)\n",
    "    ##Prepare dummy index and clean dataframe\n",
    "    rankDF['ComparisonID'] = rankDF['NetworkID'] + ' : ' + rankDF['g_i'] + ' < ' + rankDF['g_j']\n",
    "    rankDF = rankDF.pivot(index='ComparisonID', columns='Sample', values='X_i<X_j')#Sorted by index during this\n",
    "    rankDF = rankDF.reset_index()#Index becomes row number here\n",
    "    tempDF = rankDF['ComparisonID'].str.split(pat=' : ', expand=True)\n",
    "    tempDF = tempDF.rename(columns={0:'NetworkID', 1:'Ordering'})\n",
    "    rankDF = pd.concat([tempDF, rankDF], axis=1)#Dropping columns name 'Sample' during this\n",
    "    rankDF = rankDF.drop(columns='ComparisonID')\n",
    "    return rankDF\n",
    "\n",
    "def rank_template(rankDF, phenotypeS):\n",
    "    # This function generates the rank template (T) presenting the expected network ranking in a phenotype.\n",
    "    ## Ref. Eddy, J. A. et al. PLoS Comput. Biol. 2010 (Figure 1 at a galance)\n",
    "    # Requirements:\n",
    "    ## import numpy as np: confirmed with versions 1.17.5 and 1.21.1\n",
    "    ## import pandas as pd: confirmed with versions 0.25.3 and 1.3.1\n",
    "    # Input:\n",
    "    ## rankDF: pd.DataFrame obtained from the above network_ranking() function\n",
    "    ## phenotypeS: pd.Series containing phenotypes (k; 1-K) with sample (n; 1-N) indices (i.e., 1-on-1 long-format)\n",
    "    # Output:\n",
    "    ## pd.DataFrame containing the expected binary values of network ranking comparison for each phenotype (T_mk)\n",
    "    ## -> row: comparison_id (indicated by network (m) and ordering (g_i < g_j) columns)\n",
    "    ## -> column: phenotype (k) (with exception of the NetworkID and Ordering columns)\n",
    "    # Note:\n",
    "    ## True rate = 0.5 is assigned to 0 in this code.\n",
    "    ## If an item in phenotype k is 'NetworkID' or 'Ordering', this code would produce error or unexpected output.\n",
    "    \n",
    "    #Calculate the expected binary values of network ranking comparison for each phenotype (T_mk)\n",
    "    templateDF = rankDF[['NetworkID', 'Ordering']]\n",
    "    phenotypeL = phenotypeS.unique().tolist()\n",
    "    for k in phenotypeL:\n",
    "        sampleL_k = phenotypeS.loc[phenotypeS==k].index.tolist()\n",
    "        tempDF = rankDF[sampleL_k]\n",
    "        tempS = tempDF.mean(axis=1)#True (=1) rate\n",
    "        tempS = (tempS>0.5).astype('int64')#If true rate > 0.5, 1; otherwise (<= 0.5), 0\n",
    "        templateDF[k] = tempS\n",
    "    return templateDF\n",
    "\n",
    "def rank_matching_score(rankDF, templateDF):\n",
    "    # This function calculates the rank matching score (R) of a sample.\n",
    "    ## Ref. Eddy, J. A. et al. PLoS Comput. Biol. 2010 (Figure 1 at a galance)\n",
    "    # Requirements:\n",
    "    ## import numpy as np: confirmed with versions 1.17.5 and 1.21.1\n",
    "    ## import pandas as pd: confirmed with versions 0.25.3 and 1.3.1\n",
    "    # Input:\n",
    "    ## rankDF: pd.DataFrame obtained from the above network_ranking() function\n",
    "    ## templateDF: pd.DataFrame obtained from the above rank_template() function\n",
    "    # Output:\n",
    "    ## pd.DataFrame containing the rates of gene pairs matching to a rank template in a sample (R_mkn)\n",
    "    ## -> row: rank template (T_mk) (indicated by network (m) and template phenotype (k) columns)\n",
    "    ## -> column: sample (n) (with exception of the NetworkID and Template columns)\n",
    "    # Note:\n",
    "    ## True rate = 0.5 was assigned to 0 in the above rank_template() function.\n",
    "    ## -> 'Match (1)' and 'Mismatch (0)' are evenly assigned to samples in the phenotype.\n",
    "    ##    (i.e., 'Match' for (X_i < X_j) = 0 and 'Mismatch' for (X_i < X_j) = 1 in the tie case)\n",
    "    ## If items in network m and phenotype k contain ' : ', this code would produce error or unexpected output.\n",
    "    \n",
    "    #Calculate the rates of gene pairs matching to a rank template in a sample (R_mkn)\n",
    "    scoreDF = pd.DataFrame()\n",
    "    sampleL = rankDF.drop(columns=['NetworkID', 'Ordering']).columns.tolist()\n",
    "    phenotypeL = templateDF.drop(columns=['NetworkID', 'Ordering']).columns.tolist()\n",
    "    for n in sampleL:\n",
    "        scoreDF_n = pd.DataFrame()\n",
    "        for k_template in phenotypeL:\n",
    "            tempDF = rankDF[['NetworkID', 'Ordering']]\n",
    "            tempS = (rankDF[n]==templateDF[k_template]).astype('int64')#If matching, 1; otherwise, 0\n",
    "            tempDF['Match'] = tempS\n",
    "            #Calculate the rank matching score\n",
    "            scoreDF_k = tempDF.groupby(by='NetworkID', as_index=False, sort=False).mean()\n",
    "            #Update the rank matching score dataframe of sample n\n",
    "            scoreDF_k['Template'] = k_template\n",
    "            scoreDF_k = scoreDF_k[['NetworkID', 'Template', 'Match']]\n",
    "            scoreDF_n = pd.concat([scoreDF_n, scoreDF_k], axis=0)\n",
    "        #Update the rank matching score dataframe of all samples\n",
    "        scoreDF_n['Sample'] = n\n",
    "        scoreDF = pd.concat([scoreDF, scoreDF_n], axis=0)\n",
    "    ##Prepare dummy index and clean dataframe\n",
    "    scoreDF['RMSmkID'] = scoreDF['NetworkID'].str.cat(scoreDF['Template'], sep=' : ')\n",
    "    scoreDF = scoreDF.pivot(index='RMSmkID', columns='Sample', values='Match')#Sorted by index during this\n",
    "    scoreDF = scoreDF.reset_index()#Index becomes row number here\n",
    "    tempDF = scoreDF['RMSmkID'].str.split(pat=' : ', expand=True)\n",
    "    tempDF = tempDF.rename(columns={0:'NetworkID', 1:'Template'})\n",
    "    scoreDF = pd.concat([tempDF, scoreDF], axis=1)#Dropping columns name 'Sample' during this\n",
    "    scoreDF = scoreDF.drop(columns='RMSmkID')\n",
    "    return scoreDF\n",
    "\n",
    "def rank_conservation_index(scoreDF, phenotypeS):\n",
    "    # This function calculates the rank conservation index (muR) of a phenotype.\n",
    "    ## Ref. Eddy, J. A. et al. PLoS Comput. Biol. 2010 (Figure 1 at a galance)\n",
    "    # Requirements:\n",
    "    ## import numpy as np: confirmed with versions 1.17.5 and 1.21.1\n",
    "    ## import pandas as pd: confirmed with versions 0.25.3 and 1.3.1\n",
    "    # Input:\n",
    "    ## scoreDF: pd.DataFrame obtained from the above rank_matching_score() function\n",
    "    ## phenotypeS: pd.Series containing phenotypes (k; 1-K) with sample (n; 1-N) indices (i.e., 1-on-1 long-format)\n",
    "    # Output:\n",
    "    ## pd.DataFrame containing the mean values of rank matching scores (RMSs) in a phenotype (muR_mkk)\n",
    "    ## -> row: rank template (T_mk) (indicated by network (m) and template phenotype (k) columns)\n",
    "    ## -> column: phenotype (k) (with exception of the NetworkID and Template columns)\n",
    "    # Note:\n",
    "    ## Rank conservation index (RCI) is used with a broader stance here, but following a strict interpretation,\n",
    "    ## the term RCI is used for the mean of RMSs only when template phenotype is same with the sample phenotype.\n",
    "    ## If an item in phenotype k is 'NetworkID' or 'Template', this code would produce error or unexpected output.\n",
    "    \n",
    "    #Calculate the mean values of RMSs in a phenotype (muR_mkk)\n",
    "    conservationDF = scoreDF[['NetworkID', 'Template']]\n",
    "    phenotypeL = phenotypeS.unique().tolist()\n",
    "    for k in phenotypeL:\n",
    "        sampleL_k = phenotypeS.loc[phenotypeS==k].index.tolist()\n",
    "        tempS = scoreDF[sampleL_k].mean(axis=1)\n",
    "        conservationDF[k] = tempS\n",
    "    return conservationDF"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. Prepare dataset and metadata"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1-1. Analyte data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Import M001 analyte data\n",
    "fileDir = '../210126_LCproteomics-M001-DIRAC-ver6/ExportData/'\n",
    "ipynbName = '210126_LCprotomics-M001-DIRAC-ver6_preprocessing_'\n",
    "fileName = 'cleaned-robustZscored-data.tsv'\n",
    "tempDF1 = pd.read_csv(fileDir+ipynbName+fileName, sep='\\t').set_index('UniProtID')\n",
    "print('Original shape of M001 DF:', tempDF1.shape)\n",
    "\n",
    "#Import M004 analyte data\n",
    "fileDir = './ExportData/'\n",
    "ipynbName = '210125_LCprotomics-M004-DIRAC_preprocessing_'\n",
    "fileName = 'cleaned-robustZscored-data.tsv'\n",
    "tempDF2 = pd.read_csv(fileDir+ipynbName+fileName, sep='\\t').set_index('UniProtID')\n",
    "print('Original shape of M004 DF:', tempDF2.shape)\n",
    "\n",
    "#Merge and extract common analytes\n",
    "tempDF = pd.merge(tempDF1, tempDF2, left_index=True, right_index=True, how='inner')\n",
    "display(tempDF)\n",
    "\n",
    "analyteDF = tempDF"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1-2. Module–analyte metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Import module-analyte metadata\n",
    "fileDir = './ExportData/'\n",
    "ipynbName = '220525_LCproteomics-M004-DIRAC_Preprocessing-with-M001_ver2-2_'\n",
    "fileName = 'module-metadata_QuickGO-GOBP-min-n4-cov50.tsv'\n",
    "tempDF = pd.read_csv(fileDir+ipynbName+fileName, sep='\\t').set_index('ModuleID')\n",
    "print(' - Unique analytes with module:', len(tempDF['UniProtID'].unique()))\n",
    "print(' - Unique modules with analytes:', len(tempDF.index.unique()))\n",
    "\n",
    "#Prepare moduleS\n",
    "moduleS = tempDF['UniProtID']\n",
    "display(moduleS)\n",
    "\n",
    "#Retrieve module metadata\n",
    "tempDF = tempDF[['ModuleName', 'nAnalytes', 'nBackgrounds', 'Coverage']]\n",
    "moduleDF = tempDF.reset_index().drop_duplicates(keep='first').set_index('ModuleID')\n",
    "display(moduleDF)\n",
    "display(moduleDF.describe(include='all'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> –> Add the mapped analytes to module metadata.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Import analyte metadata\n",
    "fileDir = './ExportData/'\n",
    "ipynbName = '220525_LCproteomics-M004-DIRAC_Preprocessing-with-M001_ver2-2_'\n",
    "fileName = 'analyte-metadata_UniProt.tsv'\n",
    "tempDF = pd.read_csv(fileDir+ipynbName+fileName, sep='\\t')\n",
    "tempDF = pd.merge(moduleS.reset_index(), tempDF, on='UniProtID', how='left').set_index('ModuleID')\n",
    "display(tempDF)\n",
    "print(' - Unique analytes:', len(tempDF['UniProtID'].unique()))\n",
    "print(' - Unique labels:', len(tempDF['GeneLabel'].unique()))\n",
    "\n",
    "#Concatenate labels\n",
    "t_start = time.time()\n",
    "moduleDF['MappedAnalyteIDs'] = ''#Initialize\n",
    "moduleDF['MappedAnalyteGeneLabels'] = ''#Initialize\n",
    "for module in moduleDF.index.tolist():\n",
    "    tempS = tempDF['UniProtID'].loc[tempDF.index.isin([module])]#Retrieve as pd.Series\n",
    "    label = tempS.str.cat(sep=';')\n",
    "    moduleDF.loc[module, 'MappedAnalyteIDs'] = label\n",
    "    tempS = tempDF['GeneLabel'].loc[tempDF.index.isin([module])]#Retrieve as pd.Series\n",
    "    label = tempS.str.cat(sep=';')\n",
    "    moduleDF.loc[module, 'MappedAnalyteGeneLabels'] = label\n",
    "t_elapsed = time.time() - t_start\n",
    "print('Elapsed time:', round(t_elapsed//60), 'min', round(t_elapsed%60, 1), 'sec')\n",
    "\n",
    "#Clean\n",
    "moduleDF['ModuleType'] = 'Gene Ontology (Biological Process)'\n",
    "moduleDF['Source'] = 'EMBL-EBI QuickGO API'\n",
    "moduleDF = moduleDF[['ModuleName', 'ModuleType', 'MappedAnalyteIDs', 'MappedAnalyteGeneLabels',\n",
    "                     'nAnalytes', 'nBackgrounds', 'Coverage', 'Source']]\n",
    "moduleDF = moduleDF.sort_index(ascending=True)\n",
    "\n",
    "#Save for using in the sub-notebook\n",
    "fileDir = './ExportData/'\n",
    "ipynbName = '220529_LCproteomics-M004-DIRAC_DIRAC-GOBP-with-M001_ver2-3_'\n",
    "fileName = 'module-metadata.tsv'\n",
    "moduleDF.to_csv(fileDir+ipynbName+fileName, index=True, sep='\\t')\n",
    "\n",
    "display(moduleDF)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1-3. Sample–mouse metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Import M001 sample-mouse metadata\n",
    "fileDir = '../210126_LCproteomics-M001-DIRAC-ver6/ExportData/'\n",
    "ipynbName = '210126_LCprotomics-M001-DIRAC-ver6_preprocessing_'\n",
    "fileName = 'metadata-sample.tsv'\n",
    "tempDF1 = pd.read_csv(fileDir+ipynbName+fileName, sep='\\t')\n",
    "tempDF1 = tempDF1.rename(columns={'MouseID':'SampleID', 'Treatment':'Intervention'})\n",
    "tempDF1 = tempDF1.loc[tempDF1['SampleID'].isin(analyteDF.columns)]\n",
    "tempDF1 = tempDF1.set_index('SampleID')\n",
    "tempDF1['Dataset'] = 'M001'\n",
    "\n",
    "#Import M004 sample-mouse metadata\n",
    "fileDir = './ExportData/'\n",
    "ipynbName = '210125_LCprotomics-M004-DIRAC_preprocessing_'\n",
    "fileName = 'metadata-sample.tsv'\n",
    "tempDF2 = pd.read_csv(fileDir+ipynbName+fileName, sep='\\t')\n",
    "tempDF2 = tempDF2.rename(columns={'MouseID':'SampleID', 'Treatment':'Intervention'})\n",
    "tempDF2 = tempDF2.loc[tempDF2['SampleID'].isin(analyteDF.columns)]\n",
    "tempDF2 = tempDF2.set_index('SampleID')\n",
    "tempDF2['Dataset'] = 'M004'\n",
    "\n",
    "#Merge\n",
    "tempDF = pd.concat([tempDF1, tempDF2], axis=0)\n",
    "tempDF = tempDF[['Dataset', 'Intervention', 'Sex']]\n",
    "\n",
    "#Prepare phenotypeS and re-label control group\n",
    "tempDF['Phenotype'] = tempDF['Intervention']\n",
    "for row_i in range(len(tempDF)):\n",
    "    phenotype = tempDF['Phenotype'].iloc[row_i]\n",
    "    if phenotype=='Control':\n",
    "        dataset = tempDF['Dataset'].iloc[row_i]\n",
    "        if dataset=='M001':\n",
    "            tempDF['Phenotype'].iloc[row_i] = 'Control (M001)'\n",
    "        elif dataset=='M004':\n",
    "            tempDF['Phenotype'].iloc[row_i] = 'Control (M004)'\n",
    "\n",
    "display(tempDF)\n",
    "display(tempDF['Phenotype'].value_counts())\n",
    "\n",
    "#Save for using in the sub-notebook\n",
    "fileDir = './ExportData/'\n",
    "ipynbName = '220529_LCproteomics-M004-DIRAC_DIRAC-GOBP-with-M001_ver2-3_'\n",
    "fileName = 'sample-metadata.tsv'\n",
    "tempDF.to_csv(fileDir+ipynbName+fileName, index=True, sep='\\t')\n",
    "\n",
    "sampleDF = tempDF"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. Perform DIRAC with sex-pooled rank consensus"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> When applied to overall dataset, the DIRAC code often stops runnning due to memory limit (e.g., transcriptomics).  \n",
    "> –> Divide moduleS into subsets while considering the number of comparisons, and compute DIRAC in parallel.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Divide moduleS\n",
    "cutoff = 250#The maximum number of analytes per module\n",
    "tempDF = moduleDF.sort_values(by='nAnalytes', ascending=True)\n",
    "nwsubL = []\n",
    "tempL = []#Initialize\n",
    "count = 0#Initialize\n",
    "for module in tempDF.index.tolist():\n",
    "    nanalytes = tempDF.loc[module, 'nAnalytes']\n",
    "    if nanalytes>cutoff:\n",
    "        nwsubL.append([module])\n",
    "    else:\n",
    "        tempL.append(module)\n",
    "        count += nanalytes\n",
    "        if count>cutoff:\n",
    "            nwsubL.append(tempL)\n",
    "            tempL = []#Initialize\n",
    "            count = 0#Initialize\n",
    "if len(tempL)>0:#The last one but still count<=cutoff\n",
    "    nwsubL.append(tempL)\n",
    "nSub = len(nwsubL)\n",
    "print('nSublists: ', nSub)\n",
    "print('nModules per sublist:', [len(sublist) for sublist in nwsubL])\n",
    "\n",
    "#Check nAnalytes distribution\n",
    "tempS = pd.Series(name='Subset')\n",
    "for list_i in range(nSub):\n",
    "    tempL = moduleS.loc[nwsubL[list_i]].index.unique().tolist()\n",
    "    tempS1 = pd.Series(np.repeat('Subset '+str(list_i+1).zfill(3), len(tempL)), index=tempL, name='Subset')\n",
    "    tempS = pd.concat([tempS, tempS1], axis=0)\n",
    "tempS.index.set_names('ModuleID', inplace=True)\n",
    "tempDF = pd.merge(moduleDF, tempS, left_index=True, right_index=True, how='left')\n",
    "tempDF = tempDF.sort_values(by='nAnalytes', ascending=False)\n",
    "display(tempDF)\n",
    "tempDF = tempDF.groupby('Subset').agg({'nAnalytes':[len, sum, max, np.median]})\n",
    "tempDF = tempDF.sort_values(by=('nAnalytes', 'sum'), ascending=False)\n",
    "print('Subset summary:')\n",
    "display(tempDF.describe())\n",
    "print(' -> Check subset which will need high computational cost:')\n",
    "display(tempDF.loc[tempDF[('nAnalytes', 'max')]>cutoff])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "nprocessors = 4\n",
    "fileDir = './ExportData/'\n",
    "ipynbName = '210128_LCproteomics-M004-DIRAC_M001-M004-common-DIRAC-GOBP_'\n",
    "\n",
    "#Wrap as a single function\n",
    "def parallel_dirac(list_i):\n",
    "    tempS = moduleS.loc[nwsubL[list_i]]\n",
    "    \n",
    "    #Calculate the pairwise ordering of network genes (i.e., network ranking)\n",
    "    t_start1 = time.time()\n",
    "    rankingDF = network_ranking(analyteDF, tempS)\n",
    "    t_elapsed1 = time.time() - t_start1\n",
    "    \n",
    "    #Generate the rank template (T) presenting the expected network ranking in a phenotype\n",
    "    t_start2 = time.time()\n",
    "    templateDF = rank_template(rankingDF, sampleDF['Phenotype'])\n",
    "    t_elapsed2 = time.time() - t_start2\n",
    "    \n",
    "    #Calculate the rank matching score (R) of a sample\n",
    "    t_start3 = time.time()\n",
    "    rmsDF = rank_matching_score(rankingDF, templateDF)\n",
    "    t_elapsed3 = time.time() - t_start3\n",
    "    \n",
    "    #Calculate the rank conservation index (muR) of a phenotype\n",
    "    t_start4 = time.time()\n",
    "    rciDF = rank_conservation_index(rmsDF, sampleDF['Phenotype'])\n",
    "    t_elapsed4 = time.time() - t_start4\n",
    "    \n",
    "    #Save\n",
    "    fileName = 'NetworkRanking-'+str(list_i+1).zfill(3)+'.tsv'\n",
    "    rankingDF.to_csv(fileDir+ipynbName+fileName, index=False, sep='\\t')\n",
    "    fileName = 'RankTemplate-BS-'+str(list_i+1).zfill(3)+'.tsv'\n",
    "    templateDF.to_csv(fileDir+ipynbName+fileName, index=False, sep='\\t')\n",
    "    fileName = 'RankMatchingScore-BS-'+str(list_i+1).zfill(3)+'.tsv'\n",
    "    rmsDF.to_csv(fileDir+ipynbName+fileName, index=False, sep='\\t')\n",
    "    fileName = 'RankConservationIndex-BS-'+str(list_i+1).zfill(3)+'.tsv'\n",
    "    rciDF.to_csv(fileDir+ipynbName+fileName, index=False, sep='\\t')\n",
    "    \n",
    "    #Check results\n",
    "    print('Subset '+str(list_i+1).zfill(3)+':',\n",
    "          len(tempS.index.unique()), 'modules,', len(tempS.unique()), 'analytes')\n",
    "    print('• Network ranking dataframe:')\n",
    "    print('  - DF shape:', rankingDF.shape)\n",
    "    print('  - Elapsed time:', round(t_elapsed1//60), 'min', round(t_elapsed1%60, 1), 'sec')\n",
    "    print('• Rank template dataframe:')\n",
    "    print('  - DF shape:', templateDF.shape)\n",
    "    print('  - Elapsed time:', round(t_elapsed2//60), 'min', round(t_elapsed2%60, 1), 'sec')\n",
    "    print('• Rank matching score dataframe:')\n",
    "    print('  - DF shape:', rmsDF.shape)\n",
    "    print('  - Elapsed time:', round(t_elapsed3//60), 'min', round(t_elapsed3%60, 1), 'sec')\n",
    "    print('• Rank conservation index dataframe:')\n",
    "    print('  - DF shape:', rciDF.shape)\n",
    "    print('  - Elapsed time:', round(t_elapsed4//60), 'min', round(t_elapsed4%60, 1), 'sec')\n",
    "    print('')\n",
    "\n",
    "#Parallel computing\n",
    "if __name__=='__main__':\n",
    "    t_start = time.time()\n",
    "    p = Pool(nprocessors)\n",
    "    p.map(parallel_dirac, range(nSub))\n",
    "    t_finish = time.time()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Record as reference\n",
    "print(nSub, 'sublists with', nprocessors, 'processors')\n",
    "print(' - Start:', time.ctime(t_start))\n",
    "print(' - Finish:', time.ctime(t_finish))\n",
    "t_elapsed = (t_finish - t_start)\n",
    "print(' - Elapsed time:', round(t_elapsed//60), 'min', round(t_elapsed%60, 1), 'sec')\n",
    "t_elapsed = (t_finish - t_start) * nprocessors\n",
    "print(' - Total (approximate) elapsed time:', round(t_elapsed//60), 'min', round(t_elapsed%60, 1), 'sec')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Combine each result\n",
    "rmsDF = pd.DataFrame()\n",
    "rciDF = pd.DataFrame()\n",
    "for list_i in range(nSub):\n",
    "    fileName = 'RankMatchingScore-BS-'+str(list_i+1).zfill(3)+'.tsv'\n",
    "    tempDF = pd.read_csv(fileDir+ipynbName+fileName, sep='\\t')\n",
    "    tempDF = tempDF.rename(columns={'NetworkID':'ModuleID'})\n",
    "    rmsDF = pd.concat([rmsDF, tempDF], axis=0, ignore_index=True)\n",
    "    \n",
    "    fileName = 'RankConservationIndex-BS-'+str(list_i+1).zfill(3)+'.tsv'\n",
    "    tempDF = pd.read_csv(fileDir+ipynbName+fileName, sep='\\t')\n",
    "    tempDF = tempDF.rename(columns={'NetworkID':'ModuleID'})\n",
    "    rciDF = pd.concat([rciDF, tempDF], axis=0, ignore_index=True)\n",
    "\n",
    "print('• Rank matching score dataframe:')\n",
    "display(rmsDF)\n",
    "print('')\n",
    "print('• Rank conservation index dataframe:')\n",
    "display(rciDF)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3. Rank conservation index: general pattern"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3-1. Extract RCI (the mean of RMSs under the own phenotype consensus)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Extract RCI whose template phenotype corresponds to the own phenotype\n",
    "phenotypeL = rciDF.drop(columns=['ModuleID', 'Template']).columns.tolist()\n",
    "rciDF_kk = pd.DataFrame(index=pd.Index(rciDF['ModuleID'].unique(), name='ModuleID'))\n",
    "tempDF = rciDF.set_index('ModuleID')\n",
    "for k in phenotypeL:\n",
    "    tempS = tempDF[k].loc[tempDF['Template']==k]\n",
    "    rciDF_kk = pd.merge(rciDF_kk, tempS, left_index=True, right_index=True, how='left')\n",
    "\n",
    "#Order and re-label\n",
    "tempD = {'Control (M001)':'Control-1', 'Acarbose':'Acarbose',\n",
    "         'Estradiol':'17'+r'$\\alpha$'+'-Estradiol', 'Rapamycin':'Rapamycin',\n",
    "         'Control (M004)':'Control-2', '4EGI-1':'4EGI-1'}\n",
    "rciDF_kk = rciDF_kk[list(tempD.keys())]\n",
    "rciDF_kk = rciDF_kk.rename(columns=tempD)\n",
    "display(rciDF_kk)\n",
    "display(rciDF_kk.describe())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# — †1. Go to the top of the R sub-notebook —  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3-2. Repeated Student's t-tests"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> In this study, RCI was not normalized (i.e., the expected mean and variance were different between M001 and M004 datasets due to different sample size).  \n",
    "> –> Therefore, not Dunnett's test but Student's t-test (i.e., t-test with pooled variance) is used because RCI is NOT comparable across datasets (i.e., Control group is different between datasets).  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Import the summary table\n",
    "fileDir = './ExportData/'\n",
    "ipynbName = '220529_LCproteomics-M004-DIRAC_StatisticalTest-GOBP-with-M001_ver2-3_'\n",
    "fileName = 'inter-group-comparison.xlsx'\n",
    "sheetName = 'RCImean'\n",
    "tempDF = pd.read_excel(fileDir+ipynbName+fileName, sheet_name=sheetName, engine='openpyxl')\n",
    "tempDF = tempDF.set_index('ModuleID')\n",
    "display(tempDF)\n",
    "\n",
    "statDF = tempDF"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3-3. Visualization: boxplot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#Prepare DF and color\n",
    "tempDF = rciDF_kk.reset_index().melt(var_name='Group', value_name='RCI', id_vars='ModuleID')\n",
    "tempD = {'Control-1':'tab:blue', 'Acarbose':'tab:red',\n",
    "         '17'+r'$\\alpha$'+'-Estradiol':'tab:green', 'Rapamycin':'tab:purple',\n",
    "         'Control-2':'tab:blue', '4EGI-1':'tab:orange'}\n",
    "\n",
    "#Prepare significance labels\n",
    "##Retrieve statistical significance\n",
    "module = 'All'\n",
    "tempS = statDF.loc[module, statDF.columns.str.contains('AdjPval')]\n",
    "tempS.index = tempS.index.str.replace('_AdjPval', '')\n",
    "tempS.name = 'AdjPval'\n",
    "##Clean\n",
    "tempDF1 = tempS.index.to_series().str.split(pat='-vs-', expand=True)\n",
    "tempDF1 = tempDF1.rename(columns={0:'Contrast', 1:'Baseline'})\n",
    "tempDF1 = pd.merge(tempDF1, tempS, left_index=True, right_index=True, how='left')\n",
    "tempD0 = {'M001:Cont':'Control-1', 'M001:Acar':'Acarbose',\n",
    "          'M001:17aE':'17'+r'$\\alpha$'+'-Estradiol', 'M001:Rapa':'Rapamycin',\n",
    "          'M004:Cont':'Control-2', 'M004:4EGI':'4EGI-1'}\n",
    "tempDF1['Contrast'] = tempDF1['Contrast'].map(tempD0)\n",
    "tempDF1['Baseline'] = tempDF1['Baseline'].map(tempD0)\n",
    "##Convert p-value to label\n",
    "tempL = []\n",
    "for row_i in range(len(tempDF1)):\n",
    "    pval = tempDF1['AdjPval'].iloc[row_i]\n",
    "    if pval<0.001:\n",
    "        tempL.append('***')\n",
    "    elif pval<0.01:\n",
    "        tempL.append('**')\n",
    "    elif pval<0.05:\n",
    "        tempL.append('*')\n",
    "    else:\n",
    "        pval_text = Decimal(str(pval)).quantize(Decimal('0.01'), rounding=ROUND_HALF_UP)\n",
    "        tempL.append(r'$P$ = '+str(pval_text))\n",
    "tempDF1['SignifLabel'] = tempL\n",
    "##Add the y-position level in figure\n",
    "tempDF1['YposLevel'] = [0, 1, 2, 2]\n",
    "display(tempDF1)\n",
    "\n",
    "#Visualization\n",
    "ymax = 1.0\n",
    "ymin = 0.5\n",
    "yinter = 0.1\n",
    "ymargin_t = 0.06\n",
    "ymargin_b = 0.01\n",
    "aline_ymin = 0.9\n",
    "aline_ymargin = 0.05\n",
    "sns.set(style='ticks', font='Arial', context='talk')\n",
    "plt.figure(figsize=(3, 4))\n",
    "p = sns.boxplot(data=tempDF, y='RCI', x='Group', order=list(tempD.keys()), palette=tempD, dodge=False,\n",
    "                showfliers=True, flierprops={'marker':'o', 'markerfacecolor':'gray', 'alpha':0.4},\n",
    "                showcaps=True, notch=True)\n",
    "p.set(ylim=(ymin-ymargin_b, ymax+ymargin_t), yticks=np.arange(ymin, ymax + yinter/10, yinter))\n",
    "##Add border line\n",
    "p.axvline(x=3.5, **{'linestyle':'dotted', 'color':'black', 'zorder':0})\n",
    "##Add significance labels\n",
    "lines = p.axes.get_lines()#Line2D: [[Q1, Q1-1.5IQR], [Q3, Q3+1.5IQR], [Q1, Q1], [Q3, Q3], [Med, Med], [flier]]\n",
    "lines_unit = 5 + int(True)#showfliers=True\n",
    "for row_i in range(len(tempDF1)):\n",
    "    #Baseline\n",
    "    group_0 = tempDF1['Baseline'].iloc[row_i]\n",
    "    index_0 = list(tempD.keys()).index(group_0)\n",
    "    whisker_0 = lines[index_0*lines_unit + 1]\n",
    "    xcoord_0 = whisker_0._x[1]#Q3+1.5IQR\n",
    "    #ycoord_0 = whisker_0._y[1]#Q3+1.5IQR\n",
    "    #Contrast\n",
    "    group_1 = tempDF1['Contrast'].iloc[row_i]\n",
    "    index_1 = list(tempD.keys()).index(group_1)\n",
    "    whisker_1 = lines[index_1*lines_unit + 1]\n",
    "    xcoord_1 = whisker_1._x[1]#Q3+1.5IQR\n",
    "    #ycoord_1 = whisker_1._y[1]#Q3+1.5IQR\n",
    "    #Standard point of marker\n",
    "    xcoord = (xcoord_0+xcoord_1)/2\n",
    "    #ycoord = max(ycoord_0, ycoord_1)\n",
    "    ycoord = aline_ymin + aline_ymargin*tempDF1['YposLevel'].iloc[row_i]\n",
    "    label = tempDF1['SignifLabel'].iloc[row_i]\n",
    "    #Add annotation lines\n",
    "    aline_offset = yinter/10\n",
    "    aline_length = yinter/10 + aline_offset\n",
    "    plt.plot([xcoord_0, xcoord_0, xcoord_1, xcoord_1],\n",
    "             [ycoord+aline_offset, ycoord+aline_length, ycoord+aline_length, ycoord+aline_offset],\n",
    "             lw=1.5, c='k')\n",
    "    #Add annotation text\n",
    "    if label in ['***', '**', '*']:\n",
    "        text_offset = yinter/25\n",
    "        p.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                   horizontalalignment='center', verticalalignment='bottom',\n",
    "                   fontsize='medium', color='k')\n",
    "    else:\n",
    "        text_offset = yinter/5\n",
    "        p.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                   horizontalalignment='center', verticalalignment='bottom',\n",
    "                   fontsize='x-small', color='k')\n",
    "sns.despine()\n",
    "plt.xlabel('')\n",
    "plt.ylabel('Module RCI')\n",
    "plt.xticks(rotation=70, horizontalalignment='right', verticalalignment='center', rotation_mode='anchor')\n",
    "##Save\n",
    "fileDir = './ExportFigures/'\n",
    "ipynbName = '220529_LCproteomics-M004-DIRAC_DIRAC-GOBP-with-M001_ver2-3_'\n",
    "fileName = 'RCI-boxplot.tif'\n",
    "plt.gcf().savefig(fileDir+ipynbName+fileName, dpi=300, bbox_inches='tight', pad_inches=0.04,\n",
    "                  pil_kwargs={'compression':'tiff_lzw'})\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4. Rank matching score under the M004 consensus: inter-group module comparison"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> Test specific hypotheses: control RMS mean == intervention RMS mean per the M004-derived rank consensus for each module (i.e., inter-group module comparison).  \n",
    ">\n",
    "> 1. Testing the main effect of intervention on rank mathing scores (RMSs) for each module using ANOVA models  \n",
    "> 2. Then, performing post-hoc comparisons of RMSs between control vs. each intervention using the repeated Student's t-tests  \n",
    ">\n",
    "> As well as using sex-pooled rank consensus, sex is NOT included in the ANOVA model. Instead of adding the rank consensus group (i.e., M004 group) and its interaction term to the ANOVA model, stratified ANOVA model is generated for each rank consensus group. Note that ANOVA can shrink the variance utilizing all samples (per module), whose statistical power is better than the repeated Welch's t-tests in the case of small sample size. Although tricky, the P-value adjustment in (1) is performed across all models (= modules x rank consensus groups) under the assumption that modules are independent, which would be more conservative and less likely raise referees' eyebrows for venn diagram-type summary than using nominal P-value cutoff. Because the post-hoc comparisons (2) are to address the similarity of each intervention within a specific module, the P-values are adjusted across interventions only within the module (not across modules). At the same time, these P-values are simultaneously adjusted across rank consensus groups within the module, because the hypotheses across rank consensus are independent in contrast to the previous M001-alone DIRAC analysis. Note that not Dunnett's test but Student's t-test (i.e., t-test with pooled variance) is used as the post-hoc test because further adjustment of the Dunnett's test p-values with the Holm-Bonferroni method is too much (incorrect) adjustment for family-wise error rate (FWER)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# — †2. Go to †1 of the R sub-notebook —  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 4-1. ANOVA test (RMS ~ Intervention), followed by repeated Student's t-tests (Intervention)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 4-1-1/2/3. Import the summary table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Import the summary table\n",
    "fileDir = './ExportData/'\n",
    "ipynbName = '220529_LCproteomics-M004-DIRAC_StatisticalTest-GOBP-with-M001_ver2-3_'\n",
    "fileName = 'inter-group-comparison.xlsx'\n",
    "tempDF = pd.DataFrame()\n",
    "for template in ['M004:Cont', 'M004:4EGI']:\n",
    "    sheetName = template.replace(':', '-')+'-fixed-RMSmean'#Colon not allowed in sheet name in Excel...\n",
    "    tempDF1 = pd.read_excel(fileDir+ipynbName+fileName, sheet_name=sheetName, engine='openpyxl')\n",
    "    tempDF1['Template'] = template\n",
    "    tempDF1 = tempDF1.set_index(['Template', 'ModuleID'])\n",
    "    tempDF = pd.concat([tempDF, tempDF1], axis=0)\n",
    "tempDF = tempDF.sort_values(by='Intervention_Pval', ascending=True)\n",
    "display(tempDF)\n",
    "\n",
    "statDF = tempDF"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 4-1-4. Changed modules (ANOVA)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Prepare variables in the model\n",
    "tempS = statDF.loc[:, statDF.columns.str.contains('_Fstat')].columns.to_series()\n",
    "variableL = tempS.str.replace('_Fstat', '').tolist()\n",
    "\n",
    "#Changed modules\n",
    "for variable in variableL:\n",
    "    tempDF = statDF.loc[statDF[variable+'_AdjPval']<0.05]\n",
    "    tempDF = tempDF.sort_values(by=variable+'_AdjPval', ascending=True)\n",
    "    tempL1 = tempDF.loc[:, tempDF.columns.str.contains('_RMSmean')].columns.tolist()\n",
    "    tempL2 = tempDF.loc[:, tempDF.columns.str.contains('^'+variable+'_')].columns.tolist()\n",
    "    tempDF = tempDF[[col_n for subL in [['ModuleName'], tempL1, tempL2] for col_n in subL]]\n",
    "    print(variable+' (adjusted P < 0.05):', len(tempDF))\n",
    "    tempL = tempDF.index.to_frame()['ModuleID'].unique().tolist()\n",
    "    print(' -> Unique module:', len(tempL))\n",
    "    tempL = tempDF.index.to_frame()['Template'].unique().tolist()\n",
    "    for template in tempL:\n",
    "        tempDF1 = tempDF.loc[template]\n",
    "        print(' -> '+template+' rank consensus:', len(tempDF1))\n",
    "    display(tempDF)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 4-1-5. Changed modules by each intervention (Student's t-tests)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Spread statDF by template to flatten multi-index\n",
    "tempDF = moduleDF['ModuleName']#pd.Series() for now\n",
    "for template in ['M004:Cont', 'M004:4EGI']:\n",
    "    tempDF1 = statDF.loc[template].drop(columns=['ModuleName'])\n",
    "    tempDF1.columns = template+'-fixed_'+tempDF1.columns\n",
    "    tempDF = pd.merge(tempDF, tempDF1, left_index=True, right_index=True, how='left')\n",
    "\n",
    "#Extract only the changed modules\n",
    "variable = 'Intervention'\n",
    "tempS1 = tempDF['M004:Cont-fixed_'+variable+'_AdjPval']<0.05\n",
    "tempS2 = tempDF['M004:4EGI-fixed_'+variable+'_AdjPval']<0.05\n",
    "tempDF = tempDF.loc[tempS1|tempS2]\n",
    "tempDF = tempDF.sort_values(by=['M004:4EGI-fixed_'+variable+'_AdjPval',\n",
    "                                'M004:Cont-fixed_'+variable+'_AdjPval'], ascending=True)\n",
    "print(variable+' (adjusted P < 0.05):', len(tempDF))\n",
    "print(' -> M004:Cont rank consensus:', sum(tempS1))\n",
    "print(' -> M004:4EGI rank consensus:', sum(tempS2))\n",
    "\n",
    "#Take adjusted P-value\n",
    "tempDF1 = tempDF.loc[:, tempDF.columns.str.contains('-vs-.*_AdjPval$')]\n",
    "tempDF1.columns = tempDF1.columns.str.replace('_AdjPval$', '')\n",
    "tempDF1 = pd.merge(tempDF[['ModuleName', 'M004:4EGI-fixed_'+variable+'_AdjPval',\n",
    "                           'M004:Cont-fixed_'+variable+'_AdjPval']],\n",
    "                   tempDF1, left_index=True, right_index=True, how='left')\n",
    "print('Adjusted P-value:')\n",
    "display(tempDF1)\n",
    "display(tempDF1.describe())\n",
    "\n",
    "#Take effect size\n",
    "tempDF2 = tempDF.loc[:, tempDF.columns.str.contains('-vs-.*_Coef$')]\n",
    "tempDF2.columns = tempDF2.columns.str.replace('_Coef$', '')\n",
    "tempDF2 = pd.merge(tempDF[['ModuleName', 'M004:4EGI-fixed_'+variable+'_AdjPval',\n",
    "                           'M004:Cont-fixed_'+variable+'_AdjPval']],\n",
    "                   tempDF2, left_index=True, right_index=True, how='left')\n",
    "print('Changed direction (effect size):')\n",
    "display(tempDF2)\n",
    "display(tempDF2.describe())\n",
    "\n",
    "pvalDF = tempDF1\n",
    "diffDF = tempDF2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> Check the changed modules (based on the nominal P-value for the main effect and the adjusted P-values for the post-hoc tests) as reference.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Spread statDF by template to flatten multi-index\n",
    "tempDF = moduleDF['ModuleName']#pd.Series() for now\n",
    "for template in ['M004:Cont', 'M004:4EGI']:\n",
    "    tempDF1 = statDF.loc[template].drop(columns=['ModuleName'])\n",
    "    tempDF1.columns = template+'-fixed_'+tempDF1.columns\n",
    "    tempDF = pd.merge(tempDF, tempDF1, left_index=True, right_index=True, how='left')\n",
    "\n",
    "#Extract only the changed modules\n",
    "variable = 'Intervention'\n",
    "tempS1 = tempDF['M004:Cont-fixed_'+variable+'_Pval']<0.05\n",
    "tempS2 = tempDF['M004:4EGI-fixed_'+variable+'_Pval']<0.05\n",
    "tempDF = tempDF.loc[tempS1|tempS2]\n",
    "tempDF = tempDF.sort_values(by=['M004:4EGI-fixed_'+variable+'_Pval',\n",
    "                                'M004:Cont-fixed_'+variable+'_Pval'], ascending=True)\n",
    "print(variable+' (nominal P < 0.05):', len(tempDF))\n",
    "print(' -> M004:Cont rank consensus:', sum(tempS1))\n",
    "print(' -> M004:4EGI rank consensus:', sum(tempS2))\n",
    "\n",
    "#Take adjusted P-value\n",
    "tempDF1 = tempDF.loc[:, tempDF.columns.str.contains('-vs-.*_AdjPval$')]\n",
    "tempDF1.columns = tempDF1.columns.str.replace('_AdjPval$', '')\n",
    "tempDF1 = pd.merge(tempDF[['ModuleName', 'M004:4EGI-fixed_'+variable+'_AdjPval',\n",
    "                           'M004:Cont-fixed_'+variable+'_AdjPval']],\n",
    "                   tempDF1, left_index=True, right_index=True, how='left')\n",
    "print('Adjusted P-value:')\n",
    "display(tempDF1)\n",
    "display(tempDF1.describe())\n",
    "\n",
    "#Take effect size\n",
    "tempDF2 = tempDF.loc[:, tempDF.columns.str.contains('-vs-.*_Coef$')]\n",
    "tempDF2.columns = tempDF2.columns.str.replace('_Coef$', '')\n",
    "tempDF2 = pd.merge(tempDF[['ModuleName', 'M004:4EGI-fixed_'+variable+'_Pval',\n",
    "                           'M004:Cont-fixed_'+variable+'_Pval']],\n",
    "                   tempDF2, left_index=True, right_index=True, how='left')\n",
    "#print('Changed direction (effect size):')\n",
    "#display(tempDF2)\n",
    "#display(tempDF2.describe())\n",
    "\n",
    "pvalDF_ref = tempDF1\n",
    "diffDF_ref = tempDF2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 4-2. Visualization: venn diagram"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 4-2-1. RMS under 4EGI-1 consensus"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "template_label = 'M004:4EGI-fixed'\n",
    "\n",
    "#Prepare label and color\n",
    "tempD0 = {'M001:Cont':'Control-1', 'M001:Acar':'Acarbose',\n",
    "          'M001:17aE':'17'+r'$\\alpha$'+'-Estradiol', 'M001:Rapa':'Rapamycin',\n",
    "          'M004:Cont':'Control-2', 'M004:4EGI':'4EGI-1'}\n",
    "tempD1 = {'Acarbose':'tab:red', '17'+r'$\\alpha$'+'-Estradiol':'tab:green',\n",
    "          'Rapamycin':'tab:purple'}\n",
    "\n",
    "#Visualization per direction\n",
    "for similarity in ['Similar', 'Dissimilar']:\n",
    "    #Prepare similar/dissimilar module set\n",
    "    tempD2 = {}\n",
    "    tempL = pvalDF.loc[:, pvalDF.columns.str.contains(template_label+'_.*-vs-')].columns.tolist()#Based on adjusted ANOVA P-values\n",
    "    for col_n in tempL:\n",
    "        tempS1 = pvalDF[col_n]\n",
    "        tempS2 = diffDF[col_n]\n",
    "        if similarity=='Similar':\n",
    "            tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2>0)]\n",
    "        elif similarity=='Dissimilar':\n",
    "            tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2<0)]\n",
    "        comparison = col_n.replace(template_label+'_', '')\n",
    "        contrast = re.sub('-vs-.*', '', comparison)\n",
    "        label = tempD0[contrast]\n",
    "        tempD2[label] = set(tempS2.index.tolist())\n",
    "    ##Sort to make consistent order in manual legend generation\n",
    "    tempD = {}\n",
    "    for label in tempD1.keys():\n",
    "        tempD[label] = tempD2[label]\n",
    "    \n",
    "    #Not significant in all contrasts\n",
    "    tempDF = pvalDF.copy()\n",
    "    for moduleS in tempD.values():\n",
    "        tempDF = tempDF.loc[~tempDF.index.isin(moduleS)]\n",
    "    count = len(pvalDF) - len(tempDF)\n",
    "    print('Changed modules (adjusted ANOVA P < 0.05):', len(pvalDF))\n",
    "    print(' -> '+similarity+'ly changed in any of interventions:', count)\n",
    "    \n",
    "    #Skip the followings if no similarly/dissimilarly changed module\n",
    "    if count==0:\n",
    "        continue\n",
    "    \n",
    "    #Venn diagram\n",
    "    sns.set(style='ticks', font='Arial', context='talk')\n",
    "    fig, ax = plt.subplots(figsize=(4, 4))\n",
    "    venn(tempD, fmt='{size:,}', cmap=list(tempD1.values()), legend_loc=None, ax=ax)\n",
    "    plt.setp(ax, ylim=(0.05, 0.975))#Otherwise, weird space...\n",
    "    ##Add legend annotation\n",
    "    x_coord = [0.1, 0.9, 0.8]\n",
    "    y_coord = [0.8, 0.8, 0.25]\n",
    "    h_align = ['right', 'left', 'left']\n",
    "    v_align = ['bottom', 'bottom', 'top']\n",
    "    for i in range(len(tempD1)):\n",
    "        key = list(tempD1.keys())[i]\n",
    "        total = f'{len(tempD[key]):,}'\n",
    "        ax.text(x_coord[i], y_coord[i], key+'\\n('+total+' modules)',\n",
    "                fontsize='small', multialignment='center',\n",
    "                horizontalalignment=h_align[i], verticalalignment=v_align[i],\n",
    "                bbox={'boxstyle':'round', 'facecolor':tempD1[key], 'pad':0.2, 'alpha':0.5})\n",
    "    title = similarity+'ly changed modules (vs. Control-1)\\nto '+tempD0[template_label.replace('-fixed', '')]+' consensus'\n",
    "    ax.set_title(title, fontsize='medium')\n",
    "    ##Save\n",
    "    fileDir = './ExportFigures/'\n",
    "    ipynbName = '220529_LCproteomics-M004-DIRAC_DIRAC-GOBP-with-M001_ver2-3_'\n",
    "    fileName = template_label.replace(':', '-')+'-RMSmean-inter-group-comparison_venn-'+similarity.lower()+'ly-changed.tif'\n",
    "    plt.gcf().savefig(fileDir+ipynbName+fileName, dpi=300, bbox_inches='tight', pad_inches=0.04,\n",
    "                      pil_kwargs={'compression':'tiff_lzw'})\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Export module list in each subset in the venn diagram\n",
    "for similarity in ['Similar', 'Dissimilar']:\n",
    "    #Prepare similar/dissimilar module set\n",
    "    tempD = {}\n",
    "    tempL = pvalDF.loc[:, pvalDF.columns.str.contains(template_label+'_.*-vs-')].columns.tolist()#Based on adjusted ANOVA P-values\n",
    "    for col_n in tempL:\n",
    "        tempS1 = pvalDF[col_n]\n",
    "        tempS2 = diffDF[col_n]\n",
    "        if similarity=='Similar':\n",
    "            tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2>0)]\n",
    "        elif similarity=='Dissimilar':\n",
    "            tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2<0)]\n",
    "        tempD[col_n] = set(tempS2.index.tolist())\n",
    "    \n",
    "    #Not significant in all contrasts\n",
    "    tempDF = pvalDF.copy()\n",
    "    for moduleS in tempD.values():\n",
    "        tempDF = tempDF.loc[~tempDF.index.isin(moduleS)]\n",
    "    count = len(pvalDF) - len(tempDF)\n",
    "    print('Changed modules (adjusted ANOVA P < 0.05):', len(pvalDF))\n",
    "    print(' -> '+similarity+'ly changed in any of interventions:', count)\n",
    "    \n",
    "    #Skip the followings if no similarly/dissimilarly changed module\n",
    "    if count==0:\n",
    "        continue\n",
    "    \n",
    "    #Prepare a new .xlsx file (dummy README)\n",
    "    tempL1 = [len(tempD[key]) for key in tempD.keys()]\n",
    "    tempDF = pd.DataFrame({'Group':tempD.keys(), 'nModules':tempL1})\n",
    "    tempDF = tempDF.reset_index().rename(columns={'index':'VennOrder'})\n",
    "    tempDF['VennOrder'] = tempDF['VennOrder'] + 1\n",
    "    fileDir = './ExportData/'\n",
    "    ipynbName = '220529_LCproteomics-M004-DIRAC_DIRAC-GOBP-with-M001_ver2-3_'\n",
    "    fileName = template_label.replace(':', '-')+'-RMSmean-inter-group-comparison_venn-'+similarity.lower()+'ly-changed.xlsx'\n",
    "    tempDF.to_excel(fileDir+ipynbName+fileName, sheet_name='README', header=True, index=False)\n",
    "    display(tempDF)#Check\n",
    "    \n",
    "    #Spread statDF by template to flatten multi-index\n",
    "    tempDF = moduleDF['ModuleName']#pd.Series() for now\n",
    "    for template in ['M004:Cont', 'M004:4EGI']:\n",
    "        tempDF1 = statDF.loc[template].drop(columns=['ModuleName'])\n",
    "        tempDF1.columns = template+'-fixed_'+tempDF1.columns\n",
    "        tempDF = pd.merge(tempDF, tempDF1, left_index=True, right_index=True, how='left')\n",
    "    tempDF = tempDF.sort_index(ascending=True)\n",
    "    \n",
    "    t_start = time.time()\n",
    "    #Extract overall set\n",
    "    for key_i in range(len(tempD)):\n",
    "        key = list(tempD.keys())[key_i]\n",
    "        tempS = tempD[key]\n",
    "        tempDF1 = tempDF.loc[tempDF.index.isin(tempS)]\n",
    "        #Save summary table by appended to the above .xlsx file\n",
    "        ##Prepare sheet name\n",
    "        tempL1 = ['NA' for i in range(len(tempD))]\n",
    "        tempL1[key_i] = '1'\n",
    "        setName = '('+','.join(tempL1)+')'\n",
    "        with pd.ExcelWriter(fileDir+ipynbName+fileName, mode='a', engine='openpyxl') as writer:\n",
    "            tempDF1.to_excel(writer, sheet_name=setName, header=True, index=True)\n",
    "        print(' - '+setName+':', len(tempDF1))\n",
    "    \n",
    "    #Extract subset\n",
    "    tempL1 = ['1', '0']\n",
    "    tempL2 = [[k1, k2, k3] for k1 in tempL1 for k2 in tempL1 for k3 in tempL1]\n",
    "    #tempL2.remove(['0', '0', '0'])\n",
    "    for tempL1 in tempL2:\n",
    "        #Positive module set\n",
    "        tempL3 = [list(tempD.values())[key_i] for key_i, binary in enumerate(tempL1) if binary=='1']\n",
    "        tempS1 = set(pvalDF.index.tolist())#Initialize\n",
    "        for tempS in tempL3:\n",
    "            tempS1 = tempS1 & tempS\n",
    "        #Negative module set\n",
    "        tempL3 = [list(tempD.values())[key_i] for key_i, binary in enumerate(tempL1) if binary=='0']\n",
    "        tempS2 = set()#Initialize\n",
    "        for tempS in tempL3:\n",
    "            tempS2 = tempS2 | tempS\n",
    "        #Extract subset\n",
    "        tempS = tempS1 - tempS2\n",
    "        tempDF1 = tempDF.loc[tempDF.index.isin(tempS)]\n",
    "        #Save summary table by appended to the above .xlsx file\n",
    "        ##Prepare sheet name\n",
    "        setName = '('+','.join(tempL1)+')'\n",
    "        with pd.ExcelWriter(fileDir+ipynbName+fileName, mode='a', engine='openpyxl') as writer:\n",
    "            tempDF1.to_excel(writer, sheet_name=setName, header=True, index=True)\n",
    "        print(' - '+setName+':', len(tempDF1))\n",
    "    \n",
    "    t_elapsed = time.time() - t_start\n",
    "    print(' - Elapsed time:', round(t_elapsed//60), 'min', round(t_elapsed%60, 1), 'sec')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> Check the changed modules (based on the nominal P-value for the main effect and the adjusted P-values for the post-hoc tests) as reference.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "template_label = 'M004:4EGI-fixed'\n",
    "\n",
    "#Prepare label and color\n",
    "tempD0 = {'M001:Cont':'Control-1', 'M001:Acar':'Acarbose',\n",
    "          'M001:17aE':'17'+r'$\\alpha$'+'-Estradiol', 'M001:Rapa':'Rapamycin',\n",
    "          'M004:Cont':'Control-2', 'M004:4EGI':'4EGI-1'}\n",
    "tempD1 = {'Acarbose':'tab:red', '17'+r'$\\alpha$'+'-Estradiol':'tab:green',\n",
    "          'Rapamycin':'tab:purple'}\n",
    "\n",
    "#Visualization per direction\n",
    "for similarity in ['Similar', 'Dissimilar']:\n",
    "    #Prepare similar/dissimilar module set\n",
    "    tempD2 = {}\n",
    "    tempL = pvalDF_ref.loc[:, pvalDF_ref.columns.str.contains(template_label+'_.*-vs-')].columns.tolist()#Based on nominal ANOVA P-values\n",
    "    for col_n in tempL:\n",
    "        tempS1 = pvalDF_ref[col_n]\n",
    "        tempS2 = diffDF_ref[col_n]\n",
    "        if similarity=='Similar':\n",
    "            tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2>0)]\n",
    "        elif similarity=='Dissimilar':\n",
    "            tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2<0)]\n",
    "        comparison = col_n.replace(template_label+'_', '')\n",
    "        contrast = re.sub('-vs-.*', '', comparison)\n",
    "        label = tempD0[contrast]\n",
    "        tempD2[label] = set(tempS2.index.tolist())\n",
    "    ##Sort to make consistent order in manual legend generation\n",
    "    tempD = {}\n",
    "    for label in tempD1.keys():\n",
    "        tempD[label] = tempD2[label]\n",
    "    \n",
    "    #Not significant in all contrasts\n",
    "    tempDF = pvalDF_ref.copy()\n",
    "    for moduleS in tempD.values():\n",
    "        tempDF = tempDF.loc[~tempDF.index.isin(moduleS)]\n",
    "    count = len(pvalDF_ref) - len(tempDF)\n",
    "    print('Changed modules (nominal ANOVA P < 0.05):', len(pvalDF_ref))\n",
    "    print(' -> '+similarity+'ly changed in any of interventions:', count)\n",
    "    \n",
    "    #Skip the followings if no similarly/dissimilarly changed module\n",
    "    if count==0:\n",
    "        continue\n",
    "    \n",
    "    #Venn diagram\n",
    "    sns.set(style='ticks', font='Arial', context='talk')\n",
    "    fig, ax = plt.subplots(figsize=(4, 4))\n",
    "    venn(tempD, fmt='{size:,}', cmap=list(tempD1.values()), legend_loc=None, ax=ax)\n",
    "    plt.setp(ax, ylim=(0.05, 0.975))#Otherwise, weird space...\n",
    "    ##Add legend annotation\n",
    "    x_coord = [0.1, 0.9, 0.8]\n",
    "    y_coord = [0.8, 0.8, 0.25]\n",
    "    h_align = ['right', 'left', 'left']\n",
    "    v_align = ['bottom', 'bottom', 'top']\n",
    "    for i in range(len(tempD1)):\n",
    "        key = list(tempD1.keys())[i]\n",
    "        total = f'{len(tempD[key]):,}'\n",
    "        ax.text(x_coord[i], y_coord[i], key+'\\n('+total+' modules)',\n",
    "                fontsize='small', multialignment='center',\n",
    "                horizontalalignment=h_align[i], verticalalignment=v_align[i],\n",
    "                bbox={'boxstyle':'round', 'facecolor':tempD1[key], 'pad':0.2, 'alpha':0.5})\n",
    "    title = similarity+'ly changed modules (vs. Control-1)\\nto '+tempD0[template_label.replace('-fixed', '')]+' consensus'\n",
    "    ax.set_title(title, fontsize='medium')\n",
    "    ##Save\n",
    "    fileDir = './ExportFigures/'\n",
    "    ipynbName = '220529_LCproteomics-M004-DIRAC_DIRAC-GOBP-with-M001_ver2-3_'\n",
    "    fileName = template_label.replace(':', '-')+'-RMSmean-inter-group-comparison_venn-'+similarity.lower()+'ly-changed(nominalPval).tif'\n",
    "    plt.gcf().savefig(fileDir+ipynbName+fileName, dpi=300, bbox_inches='tight', pad_inches=0.04,\n",
    "                      pil_kwargs={'compression':'tiff_lzw'})\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Export module list in each subset in the venn diagram\n",
    "for similarity in ['Similar', 'Dissimilar']:\n",
    "    #Prepare similar/dissimilar module set\n",
    "    tempD = {}\n",
    "    tempL = pvalDF_ref.loc[:, pvalDF_ref.columns.str.contains(template_label+'_.*-vs-')].columns.tolist()#Based on nominal ANOVA P-values\n",
    "    for col_n in tempL:\n",
    "        tempS1 = pvalDF_ref[col_n]\n",
    "        tempS2 = diffDF_ref[col_n]\n",
    "        if similarity=='Similar':\n",
    "            tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2>0)]\n",
    "        elif similarity=='Dissimilar':\n",
    "            tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2<0)]\n",
    "        tempD[col_n] = set(tempS2.index.tolist())\n",
    "    \n",
    "    #Not significant in all contrasts\n",
    "    tempDF = pvalDF_ref.copy()\n",
    "    for moduleS in tempD.values():\n",
    "        tempDF = tempDF.loc[~tempDF.index.isin(moduleS)]\n",
    "    count = len(pvalDF_ref) - len(tempDF)\n",
    "    print('Changed modules (nominal ANOVA P < 0.05):', len(pvalDF_ref))\n",
    "    print(' -> '+similarity+'ly changed in any of interventions:', count)\n",
    "    \n",
    "    #Skip the followings if no similarly/dissimilarly changed module\n",
    "    if count==0:\n",
    "        continue\n",
    "    \n",
    "    #Prepare a new .xlsx file (dummy README)\n",
    "    tempL1 = [len(tempD[key]) for key in tempD.keys()]\n",
    "    tempDF = pd.DataFrame({'Group':tempD.keys(), 'nModules':tempL1})\n",
    "    tempDF = tempDF.reset_index().rename(columns={'index':'VennOrder'})\n",
    "    tempDF['VennOrder'] = tempDF['VennOrder'] + 1\n",
    "    fileDir = './ExportData/'\n",
    "    ipynbName = '220529_LCproteomics-M004-DIRAC_DIRAC-GOBP-with-M001_ver2-3_'\n",
    "    fileName = template_label.replace(':', '-')+'-RMSmean-inter-group-comparison_venn-'+similarity.lower()+'ly-changed(nominalPval).xlsx'\n",
    "    tempDF.to_excel(fileDir+ipynbName+fileName, sheet_name='README', header=True, index=False)\n",
    "    display(tempDF)#Check\n",
    "    \n",
    "    #Spread statDF by template to flatten multi-index\n",
    "    tempDF = moduleDF['ModuleName']#pd.Series() for now\n",
    "    for template in ['M004:Cont', 'M004:4EGI']:\n",
    "        tempDF1 = statDF.loc[template].drop(columns=['ModuleName'])\n",
    "        tempDF1.columns = template+'-fixed_'+tempDF1.columns\n",
    "        tempDF = pd.merge(tempDF, tempDF1, left_index=True, right_index=True, how='left')\n",
    "    tempDF = tempDF.sort_index(ascending=True)\n",
    "    \n",
    "    t_start = time.time()\n",
    "    #Extract overall set\n",
    "    for key_i in range(len(tempD)):\n",
    "        key = list(tempD.keys())[key_i]\n",
    "        tempS = tempD[key]\n",
    "        tempDF1 = tempDF.loc[tempDF.index.isin(tempS)]\n",
    "        #Save summary table by appended to the above .xlsx file\n",
    "        ##Prepare sheet name\n",
    "        tempL1 = ['NA' for i in range(len(tempD))]\n",
    "        tempL1[key_i] = '1'\n",
    "        setName = '('+','.join(tempL1)+')'\n",
    "        with pd.ExcelWriter(fileDir+ipynbName+fileName, mode='a', engine='openpyxl') as writer:\n",
    "            tempDF1.to_excel(writer, sheet_name=setName, header=True, index=True)\n",
    "        print(' - '+setName+':', len(tempDF1))\n",
    "    \n",
    "    #Extract subset\n",
    "    tempL1 = ['1', '0']\n",
    "    tempL2 = [[k1, k2, k3] for k1 in tempL1 for k2 in tempL1 for k3 in tempL1]\n",
    "    #tempL2.remove(['0', '0', '0'])\n",
    "    for tempL1 in tempL2:\n",
    "        #Positive module set\n",
    "        tempL3 = [list(tempD.values())[key_i] for key_i, binary in enumerate(tempL1) if binary=='1']\n",
    "        tempS1 = set(pvalDF.index.tolist())#Initialize\n",
    "        for tempS in tempL3:\n",
    "            tempS1 = tempS1 & tempS\n",
    "        #Negative module set\n",
    "        tempL3 = [list(tempD.values())[key_i] for key_i, binary in enumerate(tempL1) if binary=='0']\n",
    "        tempS2 = set()#Initialize\n",
    "        for tempS in tempL3:\n",
    "            tempS2 = tempS2 | tempS\n",
    "        #Extract subset\n",
    "        tempS = tempS1 - tempS2\n",
    "        tempDF1 = tempDF.loc[tempDF.index.isin(tempS)]\n",
    "        #Save summary table by appended to the above .xlsx file\n",
    "        ##Prepare sheet name\n",
    "        setName = '('+','.join(tempL1)+')'\n",
    "        with pd.ExcelWriter(fileDir+ipynbName+fileName, mode='a', engine='openpyxl') as writer:\n",
    "            tempDF1.to_excel(writer, sheet_name=setName, header=True, index=True)\n",
    "        print(' - '+setName+':', len(tempDF1))\n",
    "    \n",
    "    t_elapsed = time.time() - t_start\n",
    "    print(' - Elapsed time:', round(t_elapsed//60), 'min', round(t_elapsed%60, 1), 'sec')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> –> Add the nominal P-value-based counts to the adjusted P-value-based venn diagrams.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "countD_sub = {}\n",
    "countD_total = {}\n",
    "\n",
    "#Prepare label and color\n",
    "tempD0 = {'M001:Cont':'Control-1', 'M001:Acar':'Acarbose',\n",
    "          'M001:17aE':'17'+r'$\\alpha$'+'-Estradiol', 'M001:Rapa':'Rapamycin',\n",
    "          'M004:Cont':'Control-2', 'M004:4EGI':'4EGI-1'}\n",
    "tempD1 = {'Acarbose':'tab:red', '17'+r'$\\alpha$'+'-Estradiol':'tab:green',\n",
    "          'Rapamycin':'tab:purple'}\n",
    "\n",
    "#Visualization per direction\n",
    "for similarity in ['Similar', 'Dissimilar']:\n",
    "    #Prepare similar/dissimilar module set\n",
    "    tempD2 = {}\n",
    "    tempL = pvalDF_ref.loc[:, pvalDF_ref.columns.str.contains(template_label+'_.*-vs-')].columns.tolist()#Based on nominal ANOVA P-values\n",
    "    for col_n in tempL:\n",
    "        tempS1 = pvalDF_ref[col_n]\n",
    "        tempS2 = diffDF_ref[col_n]\n",
    "        if similarity=='Similar':\n",
    "            tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2>0)]\n",
    "        elif similarity=='Dissimilar':\n",
    "            tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2<0)]\n",
    "        comparison = col_n.replace(template_label+'_', '')\n",
    "        contrast = re.sub('-vs-.*', '', comparison)\n",
    "        label = tempD0[contrast]\n",
    "        tempD2[label] = set(tempS2.index.tolist())\n",
    "    ##Sort to make consistent order in manual legend generation\n",
    "    tempD = {}\n",
    "    for label in tempD1.keys():\n",
    "        tempD[label] = tempD2[label]\n",
    "    \n",
    "    #Not significant in all contrasts\n",
    "    tempDF = pvalDF_ref.copy()\n",
    "    for moduleS in tempD.values():\n",
    "        tempDF = tempDF.loc[~tempDF.index.isin(moduleS)]\n",
    "    count = len(pvalDF_ref) - len(tempDF)\n",
    "    print('Changed modules (nominal ANOVA P < 0.05):', len(pvalDF_ref))\n",
    "    print(' -> '+similarity+'ly changed in any of interventions:', count)\n",
    "    \n",
    "    #Skip the followings if no similarly/dissimilarly changed module\n",
    "    if count==0:\n",
    "        continue\n",
    "    \n",
    "    #Venn diagram\n",
    "    sns.set(style='ticks', font='Arial', context='talk')\n",
    "    fig, ax = plt.subplots(figsize=(4, 4))\n",
    "    venn(tempD, fmt='{size:,}', cmap=list(tempD1.values()), legend_loc=None, ax=ax)\n",
    "    plt.setp(ax, ylim=(0.05, 0.975))#Otherwise, weird space...\n",
    "    ##Save the count texts to use later\n",
    "    tempL = []\n",
    "    for text in ax.texts:\n",
    "        tempL.append(text.get_text())\n",
    "    countD_sub[similarity] = tempL\n",
    "    ##Add legend annotation\n",
    "    x_coord = [0.1, 0.9, 0.8]\n",
    "    y_coord = [0.8, 0.8, 0.25]\n",
    "    h_align = ['right', 'left', 'left']\n",
    "    v_align = ['bottom', 'bottom', 'top']\n",
    "    tempL = []\n",
    "    for i in range(len(tempD1)):\n",
    "        key = list(tempD1.keys())[i]\n",
    "        total = f'{len(tempD[key]):,}'\n",
    "        ax.text(x_coord[i], y_coord[i], key+'\\n('+total+' modules)',\n",
    "                fontsize='small', multialignment='center',\n",
    "                horizontalalignment=h_align[i], verticalalignment=v_align[i],\n",
    "                bbox={'boxstyle':'round', 'facecolor':tempD1[key], 'pad':0.2, 'alpha':0.5})\n",
    "        ##Save the count texts to use later\n",
    "        tempL.append(total)\n",
    "    countD_total[similarity] = tempL\n",
    "    title = similarity+'ly changed modules (vs. Control-1)\\nto '+tempD0[template_label.replace('-fixed', '')]+' consensus'\n",
    "    ax.set_title(title, fontsize='medium')\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Check\n",
    "display(countD_sub)\n",
    "display(countD_total)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Prepare label and color\n",
    "tempD0 = {'M001:Cont':'Control-1', 'M001:Acar':'Acarbose',\n",
    "          'M001:17aE':'17'+r'$\\alpha$'+'-Estradiol', 'M001:Rapa':'Rapamycin',\n",
    "          'M004:Cont':'Control-2', 'M004:4EGI':'4EGI-1'}\n",
    "tempD1 = {'Acarbose':'tab:red', '17'+r'$\\alpha$'+'-Estradiol':'tab:green',\n",
    "          'Rapamycin':'tab:purple'}\n",
    "\n",
    "#Visualization per direction\n",
    "for similarity in ['Similar', 'Dissimilar']:\n",
    "    #Prepare similar/dissimilar module set\n",
    "    tempD2 = {}\n",
    "    tempL = pvalDF.loc[:, pvalDF.columns.str.contains(template_label+'_.*-vs-')].columns.tolist()#Based on adjusted ANOVA P-values\n",
    "    for col_n in tempL:\n",
    "        tempS1 = pvalDF[col_n]\n",
    "        tempS2 = diffDF[col_n]\n",
    "        if similarity=='Similar':\n",
    "            tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2>0)]\n",
    "        elif similarity=='Dissimilar':\n",
    "            tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2<0)]\n",
    "        comparison = col_n.replace(template_label+'_', '')\n",
    "        contrast = re.sub('-vs-.*', '', comparison)\n",
    "        label = tempD0[contrast]\n",
    "        tempD2[label] = set(tempS2.index.tolist())\n",
    "    ##Sort to make consistent order in manual legend generation\n",
    "    tempD = {}\n",
    "    for label in tempD1.keys():\n",
    "        tempD[label] = tempD2[label]\n",
    "    \n",
    "    #Not significant in all contrasts\n",
    "    tempDF = pvalDF.copy()\n",
    "    for moduleS in tempD.values():\n",
    "        tempDF = tempDF.loc[~tempDF.index.isin(moduleS)]\n",
    "    count = len(pvalDF) - len(tempDF)\n",
    "    print('Changed modules (adjusted ANOVA P < 0.05):', len(pvalDF))\n",
    "    print(' -> '+similarity+'ly changed in any of interventions:', count)\n",
    "    \n",
    "    #Skip the followings if no similarly/dissimilarly changed module\n",
    "    if count==0:\n",
    "        continue\n",
    "    \n",
    "    #Venn diagram\n",
    "    sns.set(style='ticks', font='Arial', context='talk')\n",
    "    fig, ax = plt.subplots(figsize=(4, 4))\n",
    "    venn(tempD, fmt='{size:,}', cmap=list(tempD1.values()), legend_loc=None, ax=ax)\n",
    "    plt.setp(ax, ylim=(0.05, 0.975))#Otherwise, weird space...\n",
    "    ##Add the nominal P-value-based counts\n",
    "    for text_i, text in enumerate(ax.texts):\n",
    "        count_adj = text.get_text()\n",
    "        count_nom = countD_sub[similarity][text_i]\n",
    "        text.set_text(count_adj+' ['+count_nom+']')\n",
    "    ##Add legend annotation\n",
    "    x_coord = [0.1, 0.9, 0.8]\n",
    "    y_coord = [0.8, 0.8, 0.25]\n",
    "    h_align = ['right', 'left', 'left']\n",
    "    v_align = ['bottom', 'bottom', 'top']\n",
    "    for i in range(len(tempD1)):\n",
    "        key = list(tempD1.keys())[i]\n",
    "        total_adj = f'{len(tempD[key]):,}'\n",
    "        total_nom = countD_total[similarity][i]\n",
    "        ax.text(x_coord[i], y_coord[i], key+'\\n('+total_adj+' ['+total_nom+'] modules)',\n",
    "                fontsize='small', multialignment='center',\n",
    "                horizontalalignment=h_align[i], verticalalignment=v_align[i],\n",
    "                bbox={'boxstyle':'round', 'facecolor':tempD1[key], 'pad':0.2, 'alpha':0.5})\n",
    "    title = similarity+'ly changed modules (vs. Control-1)\\nto '+tempD0[template_label.replace('-fixed', '')]+' consensus'\n",
    "    ax.set_title(title, fontsize='medium')\n",
    "    ##Save\n",
    "    fileDir = './ExportFigures/'\n",
    "    ipynbName = '220606_LCproteomics-M004-DIRAC-ver2-3_Supplement_'\n",
    "    fileName = template_label.replace(':', '-')+'-RMSmean-inter-group-comparison_venn-'+similarity.lower()+'ly-changed.tif'\n",
    "    plt.gcf().savefig(fileDir+ipynbName+fileName, dpi=300, bbox_inches='tight', pad_inches=0.04,\n",
    "                      pil_kwargs={'compression':'tiff_lzw'})\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 4-2-2. RMS under Ctrl (M004) consensus"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "template_label = 'M004:Cont-fixed'\n",
    "\n",
    "#Prepare label and color\n",
    "tempD0 = {'M001:Cont':'Control-1', 'M001:Acar':'Acarbose',\n",
    "          'M001:17aE':'17'+r'$\\alpha$'+'-Estradiol', 'M001:Rapa':'Rapamycin',\n",
    "          'M004:Cont':'Control-2', 'M004:4EGI':'4EGI-1'}\n",
    "tempD1 = {'Acarbose':'tab:red', '17'+r'$\\alpha$'+'-Estradiol':'tab:green',\n",
    "          'Rapamycin':'tab:purple'}\n",
    "\n",
    "#Visualization per direction\n",
    "for similarity in ['Similar', 'Dissimilar']:\n",
    "    #Prepare similar/dissimilar module set\n",
    "    tempD2 = {}\n",
    "    tempL = pvalDF.loc[:, pvalDF.columns.str.contains(template_label+'_.*-vs-')].columns.tolist()#Based on adjusted ANOVA P-values\n",
    "    for col_n in tempL:\n",
    "        tempS1 = pvalDF[col_n]\n",
    "        tempS2 = diffDF[col_n]\n",
    "        if similarity=='Similar':\n",
    "            tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2>0)]\n",
    "        elif similarity=='Dissimilar':\n",
    "            tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2<0)]\n",
    "        comparison = col_n.replace(template_label+'_', '')\n",
    "        contrast = re.sub('-vs-.*', '', comparison)\n",
    "        label = tempD0[contrast]\n",
    "        tempD2[label] = set(tempS2.index.tolist())\n",
    "    ##Sort to make consistent order in manual legend generation\n",
    "    tempD = {}\n",
    "    for label in tempD1.keys():\n",
    "        tempD[label] = tempD2[label]\n",
    "    \n",
    "    #Not significant in all contrasts\n",
    "    tempDF = pvalDF.copy()\n",
    "    for moduleS in tempD.values():\n",
    "        tempDF = tempDF.loc[~tempDF.index.isin(moduleS)]\n",
    "    count = len(pvalDF) - len(tempDF)\n",
    "    print('Changed modules (adjusted ANOVA P < 0.05):', len(pvalDF))\n",
    "    print(' -> '+similarity+'ly changed in any of interventions:', count)\n",
    "    \n",
    "    #Skip the followings if no similarly/dissimilarly changed module\n",
    "    if count==0:\n",
    "        continue\n",
    "    \n",
    "    #Venn diagram\n",
    "    sns.set(style='ticks', font='Arial', context='talk')\n",
    "    fig, ax = plt.subplots(figsize=(4, 4))\n",
    "    venn(tempD, fmt='{size:,}', cmap=list(tempD1.values()), legend_loc=None, ax=ax)\n",
    "    plt.setp(ax, ylim=(0.05, 0.975))#Otherwise, weird space...\n",
    "    ##Add legend annotation\n",
    "    x_coord = [0.1, 0.9, 0.8]\n",
    "    y_coord = [0.8, 0.8, 0.25]\n",
    "    h_align = ['right', 'left', 'left']\n",
    "    v_align = ['bottom', 'bottom', 'top']\n",
    "    for i in range(len(tempD1)):\n",
    "        key = list(tempD1.keys())[i]\n",
    "        total = f'{len(tempD[key]):,}'\n",
    "        ax.text(x_coord[i], y_coord[i], key+'\\n('+total+' modules)',\n",
    "                fontsize='small', multialignment='center',\n",
    "                horizontalalignment=h_align[i], verticalalignment=v_align[i],\n",
    "                bbox={'boxstyle':'round', 'facecolor':tempD1[key], 'pad':0.2, 'alpha':0.5})\n",
    "    title = similarity+'ly changed modules (vs. Control-1)\\nto '+tempD0[template_label.replace('-fixed', '')]+' consensus'\n",
    "    ax.set_title(title, fontsize='medium')\n",
    "    ##Save\n",
    "    fileDir = './ExportFigures/'\n",
    "    ipynbName = '220529_LCproteomics-M004-DIRAC_DIRAC-GOBP-with-M001_ver2-3_'\n",
    "    fileName = template_label.replace(':', '-')+'-RMSmean-inter-group-comparison_venn-'+similarity.lower()+'ly-changed.tif'\n",
    "    plt.gcf().savefig(fileDir+ipynbName+fileName, dpi=300, bbox_inches='tight', pad_inches=0.04,\n",
    "                      pil_kwargs={'compression':'tiff_lzw'})\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Export module list in each subset in the venn diagram\n",
    "for similarity in ['Similar', 'Dissimilar']:\n",
    "    #Prepare similar/dissimilar module set\n",
    "    tempD = {}\n",
    "    tempL = pvalDF.loc[:, pvalDF.columns.str.contains(template_label+'_.*-vs-')].columns.tolist()#Based on adjusted ANOVA P-values\n",
    "    for col_n in tempL:\n",
    "        tempS1 = pvalDF[col_n]\n",
    "        tempS2 = diffDF[col_n]\n",
    "        if similarity=='Similar':\n",
    "            tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2>0)]\n",
    "        elif similarity=='Dissimilar':\n",
    "            tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2<0)]\n",
    "        tempD[col_n] = set(tempS2.index.tolist())\n",
    "    \n",
    "    #Not significant in all contrasts\n",
    "    tempDF = pvalDF.copy()\n",
    "    for moduleS in tempD.values():\n",
    "        tempDF = tempDF.loc[~tempDF.index.isin(moduleS)]\n",
    "    count = len(pvalDF) - len(tempDF)\n",
    "    print('Changed modules (adjusted ANOVA P < 0.05):', len(pvalDF))\n",
    "    print(' -> '+similarity+'ly changed in any of interventions:', count)\n",
    "    \n",
    "    #Skip the followings if no similarly/dissimilarly changed module\n",
    "    if count==0:\n",
    "        continue\n",
    "    \n",
    "    #Prepare a new .xlsx file (dummy README)\n",
    "    tempL1 = [len(tempD[key]) for key in tempD.keys()]\n",
    "    tempDF = pd.DataFrame({'Group':tempD.keys(), 'nModules':tempL1})\n",
    "    tempDF = tempDF.reset_index().rename(columns={'index':'VennOrder'})\n",
    "    tempDF['VennOrder'] = tempDF['VennOrder'] + 1\n",
    "    fileDir = './ExportData/'\n",
    "    ipynbName = '220529_LCproteomics-M004-DIRAC_DIRAC-GOBP-with-M001_ver2-3_'\n",
    "    fileName = template_label.replace(':', '-')+'-RMSmean-inter-group-comparison_venn-'+similarity.lower()+'ly-changed.xlsx'\n",
    "    tempDF.to_excel(fileDir+ipynbName+fileName, sheet_name='README', header=True, index=False)\n",
    "    display(tempDF)#Check\n",
    "    \n",
    "    #Spread statDF by template to flatten multi-index\n",
    "    tempDF = moduleDF['ModuleName']#pd.Series() for now\n",
    "    for template in ['M004:Cont', 'M004:4EGI']:\n",
    "        tempDF1 = statDF.loc[template].drop(columns=['ModuleName'])\n",
    "        tempDF1.columns = template+'-fixed_'+tempDF1.columns\n",
    "        tempDF = pd.merge(tempDF, tempDF1, left_index=True, right_index=True, how='left')\n",
    "    tempDF = tempDF.sort_index(ascending=True)\n",
    "    \n",
    "    t_start = time.time()\n",
    "    #Extract overall set\n",
    "    for key_i in range(len(tempD)):\n",
    "        key = list(tempD.keys())[key_i]\n",
    "        tempS = tempD[key]\n",
    "        tempDF1 = tempDF.loc[tempDF.index.isin(tempS)]\n",
    "        #Save summary table by appended to the above .xlsx file\n",
    "        ##Prepare sheet name\n",
    "        tempL1 = ['NA' for i in range(len(tempD))]\n",
    "        tempL1[key_i] = '1'\n",
    "        setName = '('+','.join(tempL1)+')'\n",
    "        with pd.ExcelWriter(fileDir+ipynbName+fileName, mode='a', engine='openpyxl') as writer:\n",
    "            tempDF1.to_excel(writer, sheet_name=setName, header=True, index=True)\n",
    "        print(' - '+setName+':', len(tempDF1))\n",
    "    \n",
    "    #Extract subset\n",
    "    tempL1 = ['1', '0']\n",
    "    tempL2 = [[k1, k2, k3] for k1 in tempL1 for k2 in tempL1 for k3 in tempL1]\n",
    "    #tempL2.remove(['0', '0', '0'])\n",
    "    for tempL1 in tempL2:\n",
    "        #Positive module set\n",
    "        tempL3 = [list(tempD.values())[key_i] for key_i, binary in enumerate(tempL1) if binary=='1']\n",
    "        tempS1 = set(pvalDF.index.tolist())#Initialize\n",
    "        for tempS in tempL3:\n",
    "            tempS1 = tempS1 & tempS\n",
    "        #Negative module set\n",
    "        tempL3 = [list(tempD.values())[key_i] for key_i, binary in enumerate(tempL1) if binary=='0']\n",
    "        tempS2 = set()#Initialize\n",
    "        for tempS in tempL3:\n",
    "            tempS2 = tempS2 | tempS\n",
    "        #Extract subset\n",
    "        tempS = tempS1 - tempS2\n",
    "        tempDF1 = tempDF.loc[tempDF.index.isin(tempS)]\n",
    "        #Save summary table by appended to the above .xlsx file\n",
    "        ##Prepare sheet name\n",
    "        setName = '('+','.join(tempL1)+')'\n",
    "        with pd.ExcelWriter(fileDir+ipynbName+fileName, mode='a', engine='openpyxl') as writer:\n",
    "            tempDF1.to_excel(writer, sheet_name=setName, header=True, index=True)\n",
    "        print(' - '+setName+':', len(tempDF1))\n",
    "    \n",
    "    t_elapsed = time.time() - t_start\n",
    "    print(' - Elapsed time:', round(t_elapsed//60), 'min', round(t_elapsed%60, 1), 'sec')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> Check the changed modules (based on the nominal P-value for the main effect and the adjusted P-values for the post-hoc tests) as reference.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "template_label = 'M004:Cont-fixed'\n",
    "\n",
    "#Prepare label and color\n",
    "tempD0 = {'M001:Cont':'Control-1', 'M001:Acar':'Acarbose',\n",
    "          'M001:17aE':'17'+r'$\\alpha$'+'-Estradiol', 'M001:Rapa':'Rapamycin',\n",
    "          'M004:Cont':'Control-2', 'M004:4EGI':'4EGI-1'}\n",
    "tempD1 = {'Acarbose':'tab:red', '17'+r'$\\alpha$'+'-Estradiol':'tab:green',\n",
    "          'Rapamycin':'tab:purple'}\n",
    "\n",
    "#Visualization per direction\n",
    "for similarity in ['Similar', 'Dissimilar']:\n",
    "    #Prepare similar/dissimilar module set\n",
    "    tempD2 = {}\n",
    "    tempL = pvalDF_ref.loc[:, pvalDF_ref.columns.str.contains(template_label+'_.*-vs-')].columns.tolist()#Based on nominal ANOVA P-values\n",
    "    for col_n in tempL:\n",
    "        tempS1 = pvalDF_ref[col_n]\n",
    "        tempS2 = diffDF_ref[col_n]\n",
    "        if similarity=='Similar':\n",
    "            tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2>0)]\n",
    "        elif similarity=='Dissimilar':\n",
    "            tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2<0)]\n",
    "        comparison = col_n.replace(template_label+'_', '')\n",
    "        contrast = re.sub('-vs-.*', '', comparison)\n",
    "        label = tempD0[contrast]\n",
    "        tempD2[label] = set(tempS2.index.tolist())\n",
    "    ##Sort to make consistent order in manual legend generation\n",
    "    tempD = {}\n",
    "    for label in tempD1.keys():\n",
    "        tempD[label] = tempD2[label]\n",
    "    \n",
    "    #Not significant in all contrasts\n",
    "    tempDF = pvalDF_ref.copy()\n",
    "    for moduleS in tempD.values():\n",
    "        tempDF = tempDF.loc[~tempDF.index.isin(moduleS)]\n",
    "    count = len(pvalDF_ref) - len(tempDF)\n",
    "    print('Changed modules (nominal ANOVA P < 0.05):', len(pvalDF_ref))\n",
    "    print(' -> '+similarity+'ly changed in any of interventions:', count)\n",
    "    \n",
    "    #Skip the followings if no similarly/dissimilarly changed module\n",
    "    if count==0:\n",
    "        continue\n",
    "    \n",
    "    #Venn diagram\n",
    "    sns.set(style='ticks', font='Arial', context='talk')\n",
    "    fig, ax = plt.subplots(figsize=(4, 4))\n",
    "    venn(tempD, fmt='{size:,}', cmap=list(tempD1.values()), legend_loc=None, ax=ax)\n",
    "    plt.setp(ax, ylim=(0.05, 0.975))#Otherwise, weird space...\n",
    "    ##Add legend annotation\n",
    "    x_coord = [0.1, 0.9, 0.8]\n",
    "    y_coord = [0.8, 0.8, 0.25]\n",
    "    h_align = ['right', 'left', 'left']\n",
    "    v_align = ['bottom', 'bottom', 'top']\n",
    "    for i in range(len(tempD1)):\n",
    "        key = list(tempD1.keys())[i]\n",
    "        total = f'{len(tempD[key]):,}'\n",
    "        ax.text(x_coord[i], y_coord[i], key+'\\n('+total+' modules)',\n",
    "                fontsize='small', multialignment='center',\n",
    "                horizontalalignment=h_align[i], verticalalignment=v_align[i],\n",
    "                bbox={'boxstyle':'round', 'facecolor':tempD1[key], 'pad':0.2, 'alpha':0.5})\n",
    "    title = similarity+'ly changed modules (vs. Control-1)\\nto '+tempD0[template_label.replace('-fixed', '')]+' consensus'\n",
    "    ax.set_title(title, fontsize='medium')\n",
    "    ##Save\n",
    "    fileDir = './ExportFigures/'\n",
    "    ipynbName = '220529_LCproteomics-M004-DIRAC_DIRAC-GOBP-with-M001_ver2-3_'\n",
    "    fileName = template_label.replace(':', '-')+'-RMSmean-inter-group-comparison_venn-'+similarity.lower()+'ly-changed(nominalPval).tif'\n",
    "    plt.gcf().savefig(fileDir+ipynbName+fileName, dpi=300, bbox_inches='tight', pad_inches=0.04,\n",
    "                      pil_kwargs={'compression':'tiff_lzw'})\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Export module list in each subset in the venn diagram\n",
    "for similarity in ['Similar', 'Dissimilar']:\n",
    "    #Prepare similar/dissimilar module set\n",
    "    tempD = {}\n",
    "    tempL = pvalDF_ref.loc[:, pvalDF_ref.columns.str.contains(template_label+'_.*-vs-')].columns.tolist()#Based on nominal ANOVA P-values\n",
    "    for col_n in tempL:\n",
    "        tempS1 = pvalDF_ref[col_n]\n",
    "        tempS2 = diffDF_ref[col_n]\n",
    "        if similarity=='Similar':\n",
    "            tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2>0)]\n",
    "        elif similarity=='Dissimilar':\n",
    "            tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2<0)]\n",
    "        tempD[col_n] = set(tempS2.index.tolist())\n",
    "    \n",
    "    #Not significant in all contrasts\n",
    "    tempDF = pvalDF_ref.copy()\n",
    "    for moduleS in tempD.values():\n",
    "        tempDF = tempDF.loc[~tempDF.index.isin(moduleS)]\n",
    "    count = len(pvalDF_ref) - len(tempDF)\n",
    "    print('Changed modules (nominal ANOVA P < 0.05):', len(pvalDF_ref))\n",
    "    print(' -> '+similarity+'ly changed in any of interventions:', count)\n",
    "    \n",
    "    #Skip the followings if no similarly/dissimilarly changed module\n",
    "    if count==0:\n",
    "        continue\n",
    "    \n",
    "    #Prepare a new .xlsx file (dummy README)\n",
    "    tempL1 = [len(tempD[key]) for key in tempD.keys()]\n",
    "    tempDF = pd.DataFrame({'Group':tempD.keys(), 'nModules':tempL1})\n",
    "    tempDF = tempDF.reset_index().rename(columns={'index':'VennOrder'})\n",
    "    tempDF['VennOrder'] = tempDF['VennOrder'] + 1\n",
    "    fileDir = './ExportData/'\n",
    "    ipynbName = '220529_LCproteomics-M004-DIRAC_DIRAC-GOBP-with-M001_ver2-3_'\n",
    "    fileName = template_label.replace(':', '-')+'-RMSmean-inter-group-comparison_venn-'+similarity.lower()+'ly-changed(nominalPval).xlsx'\n",
    "    tempDF.to_excel(fileDir+ipynbName+fileName, sheet_name='README', header=True, index=False)\n",
    "    display(tempDF)#Check\n",
    "    \n",
    "    #Spread statDF by template to flatten multi-index\n",
    "    tempDF = moduleDF['ModuleName']#pd.Series() for now\n",
    "    for template in ['M004:Cont', 'M004:4EGI']:\n",
    "        tempDF1 = statDF.loc[template].drop(columns=['ModuleName'])\n",
    "        tempDF1.columns = template+'-fixed_'+tempDF1.columns\n",
    "        tempDF = pd.merge(tempDF, tempDF1, left_index=True, right_index=True, how='left')\n",
    "    tempDF = tempDF.sort_index(ascending=True)\n",
    "    \n",
    "    t_start = time.time()\n",
    "    #Extract overall set\n",
    "    for key_i in range(len(tempD)):\n",
    "        key = list(tempD.keys())[key_i]\n",
    "        tempS = tempD[key]\n",
    "        tempDF1 = tempDF.loc[tempDF.index.isin(tempS)]\n",
    "        #Save summary table by appended to the above .xlsx file\n",
    "        ##Prepare sheet name\n",
    "        tempL1 = ['NA' for i in range(len(tempD))]\n",
    "        tempL1[key_i] = '1'\n",
    "        setName = '('+','.join(tempL1)+')'\n",
    "        with pd.ExcelWriter(fileDir+ipynbName+fileName, mode='a', engine='openpyxl') as writer:\n",
    "            tempDF1.to_excel(writer, sheet_name=setName, header=True, index=True)\n",
    "        print(' - '+setName+':', len(tempDF1))\n",
    "    \n",
    "    #Extract subset\n",
    "    tempL1 = ['1', '0']\n",
    "    tempL2 = [[k1, k2, k3] for k1 in tempL1 for k2 in tempL1 for k3 in tempL1]\n",
    "    #tempL2.remove(['0', '0', '0'])\n",
    "    for tempL1 in tempL2:\n",
    "        #Positive module set\n",
    "        tempL3 = [list(tempD.values())[key_i] for key_i, binary in enumerate(tempL1) if binary=='1']\n",
    "        tempS1 = set(pvalDF.index.tolist())#Initialize\n",
    "        for tempS in tempL3:\n",
    "            tempS1 = tempS1 & tempS\n",
    "        #Negative module set\n",
    "        tempL3 = [list(tempD.values())[key_i] for key_i, binary in enumerate(tempL1) if binary=='0']\n",
    "        tempS2 = set()#Initialize\n",
    "        for tempS in tempL3:\n",
    "            tempS2 = tempS2 | tempS\n",
    "        #Extract subset\n",
    "        tempS = tempS1 - tempS2\n",
    "        tempDF1 = tempDF.loc[tempDF.index.isin(tempS)]\n",
    "        #Save summary table by appended to the above .xlsx file\n",
    "        ##Prepare sheet name\n",
    "        setName = '('+','.join(tempL1)+')'\n",
    "        with pd.ExcelWriter(fileDir+ipynbName+fileName, mode='a', engine='openpyxl') as writer:\n",
    "            tempDF1.to_excel(writer, sheet_name=setName, header=True, index=True)\n",
    "        print(' - '+setName+':', len(tempDF1))\n",
    "    \n",
    "    t_elapsed = time.time() - t_start\n",
    "    print(' - Elapsed time:', round(t_elapsed//60), 'min', round(t_elapsed%60, 1), 'sec')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> –> Add the nominal P-value-based counts to the adjusted P-value-based venn diagrams.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "countD_sub = {}\n",
    "countD_total = {}\n",
    "\n",
    "#Prepare label and color\n",
    "tempD0 = {'M001:Cont':'Control-1', 'M001:Acar':'Acarbose',\n",
    "          'M001:17aE':'17'+r'$\\alpha$'+'-Estradiol', 'M001:Rapa':'Rapamycin',\n",
    "          'M004:Cont':'Control-2', 'M004:4EGI':'4EGI-1'}\n",
    "tempD1 = {'Acarbose':'tab:red', '17'+r'$\\alpha$'+'-Estradiol':'tab:green',\n",
    "          'Rapamycin':'tab:purple'}\n",
    "\n",
    "#Visualization per direction\n",
    "for similarity in ['Similar', 'Dissimilar']:\n",
    "    #Prepare similar/dissimilar module set\n",
    "    tempD2 = {}\n",
    "    tempL = pvalDF_ref.loc[:, pvalDF_ref.columns.str.contains(template_label+'_.*-vs-')].columns.tolist()#Based on nominal ANOVA P-values\n",
    "    for col_n in tempL:\n",
    "        tempS1 = pvalDF_ref[col_n]\n",
    "        tempS2 = diffDF_ref[col_n]\n",
    "        if similarity=='Similar':\n",
    "            tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2>0)]\n",
    "        elif similarity=='Dissimilar':\n",
    "            tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2<0)]\n",
    "        comparison = col_n.replace(template_label+'_', '')\n",
    "        contrast = re.sub('-vs-.*', '', comparison)\n",
    "        label = tempD0[contrast]\n",
    "        tempD2[label] = set(tempS2.index.tolist())\n",
    "    ##Sort to make consistent order in manual legend generation\n",
    "    tempD = {}\n",
    "    for label in tempD1.keys():\n",
    "        tempD[label] = tempD2[label]\n",
    "    \n",
    "    #Not significant in all contrasts\n",
    "    tempDF = pvalDF_ref.copy()\n",
    "    for moduleS in tempD.values():\n",
    "        tempDF = tempDF.loc[~tempDF.index.isin(moduleS)]\n",
    "    count = len(pvalDF_ref) - len(tempDF)\n",
    "    print('Changed modules (nominal ANOVA P < 0.05):', len(pvalDF_ref))\n",
    "    print(' -> '+similarity+'ly changed in any of interventions:', count)\n",
    "    \n",
    "    #Skip the followings if no similarly/dissimilarly changed module\n",
    "    if count==0:\n",
    "        continue\n",
    "    \n",
    "    #Venn diagram\n",
    "    sns.set(style='ticks', font='Arial', context='talk')\n",
    "    fig, ax = plt.subplots(figsize=(4, 4))\n",
    "    venn(tempD, fmt='{size:,}', cmap=list(tempD1.values()), legend_loc=None, ax=ax)\n",
    "    plt.setp(ax, ylim=(0.05, 0.975))#Otherwise, weird space...\n",
    "    ##Save the count texts to use later\n",
    "    tempL = []\n",
    "    for text in ax.texts:\n",
    "        tempL.append(text.get_text())\n",
    "    countD_sub[similarity] = tempL\n",
    "    ##Add legend annotation\n",
    "    x_coord = [0.1, 0.9, 0.8]\n",
    "    y_coord = [0.8, 0.8, 0.25]\n",
    "    h_align = ['right', 'left', 'left']\n",
    "    v_align = ['bottom', 'bottom', 'top']\n",
    "    tempL = []\n",
    "    for i in range(len(tempD1)):\n",
    "        key = list(tempD1.keys())[i]\n",
    "        total = f'{len(tempD[key]):,}'\n",
    "        ax.text(x_coord[i], y_coord[i], key+'\\n('+total+' modules)',\n",
    "                fontsize='small', multialignment='center',\n",
    "                horizontalalignment=h_align[i], verticalalignment=v_align[i],\n",
    "                bbox={'boxstyle':'round', 'facecolor':tempD1[key], 'pad':0.2, 'alpha':0.5})\n",
    "        ##Save the count texts to use later\n",
    "        tempL.append(total)\n",
    "    countD_total[similarity] = tempL\n",
    "    title = similarity+'ly changed modules (vs. Control-1)\\nto '+tempD0[template_label.replace('-fixed', '')]+' consensus'\n",
    "    ax.set_title(title, fontsize='medium')\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Check\n",
    "display(countD_sub)\n",
    "display(countD_total)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Prepare label and color\n",
    "tempD0 = {'M001:Cont':'Control-1', 'M001:Acar':'Acarbose',\n",
    "          'M001:17aE':'17'+r'$\\alpha$'+'-Estradiol', 'M001:Rapa':'Rapamycin',\n",
    "          'M004:Cont':'Control-2', 'M004:4EGI':'4EGI-1'}\n",
    "tempD1 = {'Acarbose':'tab:red', '17'+r'$\\alpha$'+'-Estradiol':'tab:green',\n",
    "          'Rapamycin':'tab:purple'}\n",
    "\n",
    "#Visualization per direction\n",
    "for similarity in ['Similar', 'Dissimilar']:\n",
    "    #Prepare similar/dissimilar module set\n",
    "    tempD2 = {}\n",
    "    tempL = pvalDF.loc[:, pvalDF.columns.str.contains(template_label+'_.*-vs-')].columns.tolist()#Based on adjusted ANOVA P-values\n",
    "    for col_n in tempL:\n",
    "        tempS1 = pvalDF[col_n]\n",
    "        tempS2 = diffDF[col_n]\n",
    "        if similarity=='Similar':\n",
    "            tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2>0)]\n",
    "        elif similarity=='Dissimilar':\n",
    "            tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2<0)]\n",
    "        comparison = col_n.replace(template_label+'_', '')\n",
    "        contrast = re.sub('-vs-.*', '', comparison)\n",
    "        label = tempD0[contrast]\n",
    "        tempD2[label] = set(tempS2.index.tolist())\n",
    "    ##Sort to make consistent order in manual legend generation\n",
    "    tempD = {}\n",
    "    for label in tempD1.keys():\n",
    "        tempD[label] = tempD2[label]\n",
    "    \n",
    "    #Not significant in all contrasts\n",
    "    tempDF = pvalDF.copy()\n",
    "    for moduleS in tempD.values():\n",
    "        tempDF = tempDF.loc[~tempDF.index.isin(moduleS)]\n",
    "    count = len(pvalDF) - len(tempDF)\n",
    "    print('Changed modules (adjusted ANOVA P < 0.05):', len(pvalDF))\n",
    "    print(' -> '+similarity+'ly changed in any of interventions:', count)\n",
    "    \n",
    "    #Skip the followings if no similarly/dissimilarly changed module\n",
    "    if count==0:\n",
    "        continue\n",
    "    \n",
    "    #Venn diagram\n",
    "    sns.set(style='ticks', font='Arial', context='talk')\n",
    "    fig, ax = plt.subplots(figsize=(4, 4))\n",
    "    venn(tempD, fmt='{size:,}', cmap=list(tempD1.values()), legend_loc=None, ax=ax)\n",
    "    plt.setp(ax, ylim=(0.05, 0.975))#Otherwise, weird space...\n",
    "    ##Add the nominal P-value-based counts\n",
    "    for text_i, text in enumerate(ax.texts):\n",
    "        count_adj = text.get_text()\n",
    "        count_nom = countD_sub[similarity][text_i]\n",
    "        text.set_text(count_adj+' ['+count_nom+']')\n",
    "    ##Add legend annotation\n",
    "    x_coord = [0.1, 0.9, 0.8]\n",
    "    y_coord = [0.8, 0.8, 0.25]\n",
    "    h_align = ['right', 'left', 'left']\n",
    "    v_align = ['bottom', 'bottom', 'top']\n",
    "    for i in range(len(tempD1)):\n",
    "        key = list(tempD1.keys())[i]\n",
    "        total_adj = f'{len(tempD[key]):,}'\n",
    "        total_nom = countD_total[similarity][i]\n",
    "        ax.text(x_coord[i], y_coord[i], key+'\\n('+total_adj+' ['+total_nom+'] modules)',\n",
    "                fontsize='small', multialignment='center',\n",
    "                horizontalalignment=h_align[i], verticalalignment=v_align[i],\n",
    "                bbox={'boxstyle':'round', 'facecolor':tempD1[key], 'pad':0.2, 'alpha':0.5})\n",
    "    title = similarity+'ly changed modules (vs. Control-1)\\nto '+tempD0[template_label.replace('-fixed', '')]+' consensus'\n",
    "    ax.set_title(title, fontsize='medium')\n",
    "    ##Save\n",
    "    fileDir = './ExportFigures/'\n",
    "    ipynbName = '220606_LCproteomics-M004-DIRAC-ver2-3_Supplement_'\n",
    "    fileName = template_label.replace(':', '-')+'-RMSmean-inter-group-comparison_venn-'+similarity.lower()+'ly-changed.tif'\n",
    "    plt.gcf().savefig(fileDir+ipynbName+fileName, dpi=300, bbox_inches='tight', pad_inches=0.04,\n",
    "                      pil_kwargs={'compression':'tiff_lzw'})\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 4-3. Visualization: pointplot"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 4-3-1. Similarly changed modules to 4EGI-1 consensus by all interventions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Prepare the target module set\n",
    "posL = ['M004:4EGI-fixed_M001:Acar-vs-M001:Cont',\n",
    "        'M004:4EGI-fixed_M001:17aE-vs-M001:Cont',\n",
    "        'M004:4EGI-fixed_M001:Rapa-vs-M001:Cont']\n",
    "negL = ['']\n",
    "template_label = 'M004:4EGI-fixed'\n",
    "similarity = 'Similar'\n",
    "tempS = pd.Series(np.repeat(True, len(pvalDF_ref)), index=pvalDF_ref.index)#Initialize based on nominal ANOVA P-values\n",
    "tempL = pvalDF_ref.loc[:, pvalDF_ref.columns.str.contains(template_label+'_.*-vs-')].columns.tolist()\n",
    "for col_n in tempL:\n",
    "    tempS1 = pvalDF_ref[col_n]\n",
    "    tempS2 = diffDF_ref[col_n]\n",
    "    if col_n in posL:\n",
    "        if similarity=='Similar':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2>0)\n",
    "        elif similarity=='Dissimilar':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2<0)\n",
    "    elif col_n in negL:\n",
    "        tempS3 = (tempS1>=0.05)\n",
    "        #Significance for inverse regulation\n",
    "        if similarity=='Similar':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2<0)\n",
    "        elif similarity=='Dissimilar':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2>0)\n",
    "        tempS1 = tempS3 | tempS1\n",
    "    else:\n",
    "        tempS1 = (tempS1>=0.0)\n",
    "    #Update True\n",
    "    tempS = tempS & tempS1\n",
    "tempL = tempS.loc[tempS.tolist()].index.tolist()\n",
    "print(len(tempL), similarity.lower()+'ly changed modules to '+template_label+' consensus with significance in', posL, 'but not in', negL)\n",
    "\n",
    "#Select representatives\n",
    "topX = np.min([30, len(tempL)])\n",
    "topX_plot = np.min([10, len(tempL)])\n",
    "#Spread statDF by template to flatten multi-index\n",
    "tempDF = moduleDF['ModuleName']#pd.Series() for now\n",
    "for template in ['M004:Cont', 'M004:4EGI']:\n",
    "    tempDF1 = statDF.loc[template].drop(columns=['ModuleName'])\n",
    "    tempDF1.columns = template+'-fixed_'+tempDF1.columns\n",
    "    tempDF = pd.merge(tempDF, tempDF1, left_index=True, right_index=True, how='left')\n",
    "tempDF = tempDF.loc[:, tempDF.columns.str.contains('Pval$')]\n",
    "tempDF = pd.merge(moduleDF['ModuleName'], tempDF, left_index=True, right_index=True, how='left')\n",
    "tempDF = tempDF.loc[tempL]\n",
    "tempDF = tempDF.sort_values(by=['M004:4EGI-fixed_'+variable+'_AdjPval',\n",
    "                                'M004:Cont-fixed_'+variable+'_AdjPval'], ascending=True)\n",
    "print('Top', topX, 'modules (sort by the main effect of '+variable+'):')\n",
    "display(tempDF.iloc[:topX])\n",
    "plotL = tempDF.index.tolist()[:topX_plot]\n",
    "\n",
    "#Prepare DF for plot\n",
    "tempDF = rmsDF.reset_index().melt(var_name='SampleID', value_name='RMS', id_vars=['ModuleID', 'Template'])\n",
    "tempDF1 = sampleDF.reset_index()[['SampleID', 'Phenotype']]\n",
    "tempDF = pd.merge(tempDF, tempDF1, on='SampleID', how='left')\n",
    "\n",
    "#Prepare label and color\n",
    "tempD = {'Control (M001)':'Control-1', 'Acarbose':'Acarbose',\n",
    "         'Estradiol':'17'+r'$\\alpha$'+'-Estradiol', 'Rapamycin':'Rapamycin',\n",
    "         'Control (M004)':'Control-2', '4EGI-1':'4EGI-1'}\n",
    "tempDF['Template'] = tempDF['Template'].map(tempD)\n",
    "tempDF['Group'] = tempDF['Phenotype'].map(tempD)\n",
    "tempD0 = {'M001:Cont':'Control-1', 'M001:Acar':'Acarbose',\n",
    "          'M001:17aE':'17'+r'$\\alpha$'+'-Estradiol', 'M001:Rapa':'Rapamycin',\n",
    "          'M004:Cont':'Control-2', 'M004:4EGI':'4EGI-1'}\n",
    "tempD1 = {'Control-1':'tab:blue', 'Acarbose':'tab:red',\n",
    "          '17'+r'$\\alpha$'+'-Estradiol':'tab:green', 'Rapamycin':'tab:purple',\n",
    "          'Control-2':'tab:blue', '4EGI-1':'tab:orange'}\n",
    "tempD2 = {}\n",
    "for label in ['Control-2', '4EGI-1']:\n",
    "    if tempD1[label]=='tab:blue':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(1)\n",
    "    elif tempD1[label]=='tab:orange':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(3)\n",
    "    elif tempD1[label]=='tab:green':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(5)\n",
    "    elif tempD1[label]=='tab:red':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(7)\n",
    "    elif tempD1[label]=='tab:purple':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(9)\n",
    "    elif tempD1[label]=='tab:olive':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(17)\n",
    "    else:\n",
    "        tempD2[label] = 'Error?'\n",
    "\n",
    "#Visualize each representative\n",
    "for rank_i in range(len(plotL)):\n",
    "    print(' - Rank '+str(rank_i+1)+' (sort by the main effect of '+variable+'):')\n",
    "    module = plotL[rank_i]\n",
    "    #Check module summary\n",
    "    tempDF1 = pd.DataFrame(moduleDF.loc[module]).T\n",
    "    display(tempDF1)\n",
    "    \n",
    "    #Select RMS\n",
    "    tempDF1 = tempDF.loc[tempDF['ModuleID']==module]\n",
    "    tempDF1 = tempDF1.loc[tempDF1['Template'].isin(tempD2.keys())]\n",
    "    \n",
    "    #Check RMS summary\n",
    "    tempDF2 = tempDF1.groupby(['Template', 'Group'])['RMS'].agg(['count', 'mean', 'std'])\n",
    "    tempL1 = []\n",
    "    tempL2 = []\n",
    "    for row_n in tempDF2.index.tolist():\n",
    "        count, mean, std = tempDF2.loc[row_n]\n",
    "        tempL1.append(mean - 1.96*std/np.sqrt(count))\n",
    "        tempL2.append(mean + 1.96*std/np.sqrt(count))\n",
    "    tempDF2['0.025'] = tempL1\n",
    "    tempDF2['0.975'] = tempL2\n",
    "    ##Multiindex sort\n",
    "    tempDF2 = tempDF2.reset_index()\n",
    "    tempDF2['Template'] = pd.Categorical(tempDF2['Template'], categories=list(tempD2.keys()))\n",
    "    tempDF2['Group'] = pd.Categorical(tempDF2['Group'], categories=list(tempD1.keys()))\n",
    "    tempDF2 = tempDF2.sort_values(by=['Template', 'Group']).set_index(['Template', 'Group'])\n",
    "    display(tempDF2)\n",
    "    \n",
    "    #Prepare significance labels\n",
    "    ##Retrieve statistical significance\n",
    "    tempS = pvalDF_ref.loc[module, pvalDF_ref.columns.str.contains('-fixed_.*-vs-')]#Adjusted P-value\n",
    "    tempS.name = 'AdjPval'\n",
    "    ##Clean\n",
    "    tempDF2 = tempS.index.to_series().str.split(pat='-fixed_', expand=True)\n",
    "    tempDF2 = tempDF2.rename(columns={0:'Template', 1:'Comparison'})\n",
    "    tempS1 = tempDF2['Template']\n",
    "    tempDF2 = tempDF2['Comparison'].str.split(pat='-vs-', expand=True)\n",
    "    tempDF2 = tempDF2.rename(columns={0:'Contrast', 1:'Baseline'})\n",
    "    tempDF2 = pd.merge(tempS1, tempDF2, left_index=True, right_index=True, how='left')\n",
    "    tempDF2 = pd.merge(tempDF2, tempS, left_index=True, right_index=True, how='left')\n",
    "    tempDF2['Template'] = tempDF2['Template'].map(tempD0)\n",
    "    tempDF2['Contrast'] = tempDF2['Contrast'].map(tempD0)\n",
    "    tempDF2['Baseline'] = tempDF2['Baseline'].map(tempD0)\n",
    "    ##Convert p-value to label\n",
    "    tempL = []\n",
    "    for row_i in range(len(tempDF2)):\n",
    "        pval = tempDF2['AdjPval'].iloc[row_i]\n",
    "        if pval<0.001:\n",
    "            tempL.append('***')\n",
    "        elif pval<0.01:\n",
    "            tempL.append('**')\n",
    "        elif pval<0.05:\n",
    "            tempL.append('*')\n",
    "        else:\n",
    "            pval_text = Decimal(str(pval)).quantize(Decimal('0.01'), rounding=ROUND_HALF_UP)\n",
    "            tempL.append(r'$P$ = '+str(pval_text))\n",
    "    tempDF2['SignifLabel'] = tempL\n",
    "    display(tempDF2)\n",
    "    \n",
    "    #Visualization\n",
    "    ymax = 1.0\n",
    "    ymin = 0.0\n",
    "    yinter = 0.2\n",
    "    ymargin_t = 0.4\n",
    "    ymargin_b = 0.05\n",
    "    aline_ymin = 1.0\n",
    "    aline_ymargin = 0.125\n",
    "    sns.set(style='ticks', font='Arial', context='talk')\n",
    "    fig, axes = plt.subplots(nrows=1, ncols=2, figsize=(6.4, 5.875), sharex=True, sharey=True,\n",
    "                             gridspec_kw={'width_ratios':[3.2, 3.2]})\n",
    "    for ax_i, ax in enumerate(axes.flat):\n",
    "        template = list(tempD2.keys())[ax_i]\n",
    "        tempDF3 = tempDF1.loc[tempDF1['Template']==template]\n",
    "        sns.pointplot(data=tempDF3, x='Group', y='RMS', order=list(tempD1.keys()), palette=tempD1,\n",
    "                      markers='o', dodge=False, join=False, capsize=0.6, estimator=np.mean, ci=95, ax=ax)\n",
    "        sns.stripplot(data=tempDF3, x='Group', y='RMS',\n",
    "                      order=list(tempD1.keys()), palette=tempD1, dodge=False, jitter=0.15,\n",
    "                      size=5, edgecolor='black', linewidth=1, **{'marker':'o', 'alpha':0.5}, ax=ax)\n",
    "        #Add border line\n",
    "        ax.axvline(x=3.5, **{'linestyle':'dotted', 'color':'black', 'zorder':0})\n",
    "        #Add RCI line\n",
    "        rci = tempDF3['RMS'].loc[tempDF3['Group']==template].mean()\n",
    "        ax.axhline(y=rci, **{'linestyle':'--', 'color':tempD1[template], 'zorder':0})\n",
    "        #Set axis\n",
    "        sns.despine()\n",
    "        ax.set(ylim=(ymin-ymargin_b, ymax+ymargin_t), yticks=np.arange(ymin, ymax + yinter/10, yinter))\n",
    "        plt.setp(ax.get_xticklabels(), rotation=70, horizontalalignment='right',\n",
    "                 verticalalignment='center', rotation_mode='anchor')\n",
    "        if ax_i==0:\n",
    "            plt.setp(ax, xlabel='', ylabel='Sample RMS')\n",
    "        else:\n",
    "            plt.setp(ax.get_yticklabels(), visible=False)\n",
    "            plt.setp(ax, xlabel='', ylabel='')\n",
    "        #Add significance labels\n",
    "        tempDF3 = tempDF2.loc[tempDF2['Template']==template]\n",
    "        for row_i in range(len(tempDF3)):\n",
    "            #Baseline\n",
    "            group_0 = tempDF3['Baseline'].iloc[row_i]\n",
    "            index_0 = list(tempD1.keys()).index(group_0)\n",
    "            xcoord_0 = index_0\n",
    "            #Contrast\n",
    "            group_1 = tempDF3['Contrast'].iloc[row_i]\n",
    "            index_1 = list(tempD1.keys()).index(group_1)\n",
    "            xcoord_1 = index_1\n",
    "            #Standard point of marker\n",
    "            xcoord = (xcoord_0+xcoord_1)/2\n",
    "            ycoord = aline_ymin + aline_ymargin*row_i\n",
    "            label = tempDF3['SignifLabel'].iloc[row_i]\n",
    "            #Add annotation lines\n",
    "            aline_offset = yinter/5\n",
    "            aline_length = yinter/5 + aline_offset/2\n",
    "            ax.plot([xcoord_0, xcoord_0, xcoord_1, xcoord_1],\n",
    "                    [ycoord+aline_offset, ycoord+aline_length, ycoord+aline_length, ycoord+aline_offset],\n",
    "                    lw=1.5, c='k')\n",
    "            #Add annotation text\n",
    "            if label in ['***', '**', '*']:\n",
    "                text_offset = yinter/21\n",
    "                ax.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                            horizontalalignment='center', verticalalignment='bottom',\n",
    "                            fontsize='medium', color='k')\n",
    "            else:\n",
    "                text_offset = yinter/3.5\n",
    "                ax.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                            horizontalalignment='center', verticalalignment='bottom',\n",
    "                            fontsize='x-small', color='k')\n",
    "        #Add annotation\n",
    "        ax.set_title('Consensus:\\n'+template, {'fontsize':'small'})\n",
    "        xoff = 0.015\n",
    "        yoff = 0.01\n",
    "        rect = plt.Rectangle((xoff, 1), 1-xoff, 0.16,#Manual adjustment\n",
    "                             transform=ax.transAxes, facecolor=tempD2[template],\n",
    "                             clip_on=False, linewidth=0, zorder=0.5)\n",
    "        ax.add_patch(rect)\n",
    "    fig.tight_layout()\n",
    "    #Set title\n",
    "    modulename = moduleDF.loc[module, 'ModuleName']\n",
    "    initial = modulename[0].capitalize()\n",
    "    title = re.sub('^.', initial, modulename)+' ('+module+')'\n",
    "    title = '\\n'.join(wrap(title, 60))#Because the below wrap=True didn't work\n",
    "    fig.suptitle(title, size='small',\n",
    "                 verticalalignment='baseline', horizontalalignment='center', wrap=True, y=0.935)\n",
    "    ##Save\n",
    "    fileDir = './ExportFigures/'\n",
    "    ipynbName = '220529_LCproteomics-M004-DIRAC_DIRAC-GOBP-with-M001_ver2-3_'\n",
    "    fileName = 'fixed-RMSmean-inter-group-comparison_RMS-pointplot-'+module.replace('GO:', 'GO')+'.tif'\n",
    "    plt.gcf().savefig(fileDir+ipynbName+fileName, dpi=300, bbox_inches='tight', pad_inches=0.04,\n",
    "                      pil_kwargs={'compression':'tiff_lzw'})\n",
    "    plt.show()\n",
    "    print('')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 4-3-2. Similarly changed modules to 4EGI-1 consensus by any of interventions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Prepare the target module set\n",
    "posL = ['M004:4EGI-fixed_M001:Acar-vs-M001:Cont',\n",
    "        'M004:4EGI-fixed_M001:17aE-vs-M001:Cont',\n",
    "        'M004:4EGI-fixed_M001:Rapa-vs-M001:Cont']\n",
    "negL = ['']\n",
    "template_label = 'M004:4EGI-fixed'\n",
    "similarity = 'Similar'\n",
    "tempS = pd.Series(np.repeat(False, len(pvalDF_ref)), index=pvalDF_ref.index)#Initialize based on nominal ANOVA P-values\n",
    "tempL = pvalDF_ref.loc[:, pvalDF_ref.columns.str.contains(template_label+'_.*-vs-')].columns.tolist()\n",
    "for col_n in tempL:\n",
    "    tempS1 = pvalDF_ref[col_n]\n",
    "    tempS2 = diffDF_ref[col_n]\n",
    "    if col_n in posL:\n",
    "        if similarity=='Similar':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2>0)\n",
    "        elif similarity=='Dissimilar':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2<0)\n",
    "    elif col_n in negL:\n",
    "        tempS3 = (tempS1>=0.05)\n",
    "        #Significance for inverse regulation\n",
    "        if similarity=='Similar':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2<0)\n",
    "        elif similarity=='Dissimilar':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2>0)\n",
    "        tempS1 = tempS3 | tempS1\n",
    "    else:\n",
    "        tempS1 = (tempS1>=0.0)\n",
    "    #Update True\n",
    "    tempS = tempS | tempS1\n",
    "tempL = tempS.loc[tempS.tolist()].index.tolist()\n",
    "print(len(tempL), similarity.lower()+'ly changed modules to '+template_label+' consensus with significance in', posL, 'but not in', negL)\n",
    "\n",
    "#Select representatives\n",
    "topX = np.min([30, len(tempL)])\n",
    "topX_plot = np.min([10, len(tempL)])\n",
    "#Spread statDF by template to flatten multi-index\n",
    "tempDF = moduleDF['ModuleName']#pd.Series() for now\n",
    "for template in ['M004:Cont', 'M004:4EGI']:\n",
    "    tempDF1 = statDF.loc[template].drop(columns=['ModuleName'])\n",
    "    tempDF1.columns = template+'-fixed_'+tempDF1.columns\n",
    "    tempDF = pd.merge(tempDF, tempDF1, left_index=True, right_index=True, how='left')\n",
    "tempDF = tempDF.loc[:, tempDF.columns.str.contains('Pval$')]\n",
    "tempDF = pd.merge(moduleDF['ModuleName'], tempDF, left_index=True, right_index=True, how='left')\n",
    "tempDF = tempDF.loc[tempL]\n",
    "tempDF = tempDF.sort_values(by=['M004:4EGI-fixed_'+variable+'_AdjPval',\n",
    "                                'M004:Cont-fixed_'+variable+'_AdjPval'], ascending=True)\n",
    "print('Top', topX, 'modules (sort by the main effect of '+variable+'):')\n",
    "display(tempDF.iloc[:topX])\n",
    "plotL = tempDF.index.tolist()[:topX_plot]\n",
    "\n",
    "#Prepare DF for plot\n",
    "tempDF = rmsDF.reset_index().melt(var_name='SampleID', value_name='RMS', id_vars=['ModuleID', 'Template'])\n",
    "tempDF1 = sampleDF.reset_index()[['SampleID', 'Phenotype']]\n",
    "tempDF = pd.merge(tempDF, tempDF1, on='SampleID', how='left')\n",
    "\n",
    "#Prepare label and color\n",
    "tempD = {'Control (M001)':'Control-1', 'Acarbose':'Acarbose',\n",
    "         'Estradiol':'17'+r'$\\alpha$'+'-Estradiol', 'Rapamycin':'Rapamycin',\n",
    "         'Control (M004)':'Control-2', '4EGI-1':'4EGI-1'}\n",
    "tempDF['Template'] = tempDF['Template'].map(tempD)\n",
    "tempDF['Group'] = tempDF['Phenotype'].map(tempD)\n",
    "tempD0 = {'M001:Cont':'Control-1', 'M001:Acar':'Acarbose',\n",
    "          'M001:17aE':'17'+r'$\\alpha$'+'-Estradiol', 'M001:Rapa':'Rapamycin',\n",
    "          'M004:Cont':'Control-2', 'M004:4EGI':'4EGI-1'}\n",
    "tempD1 = {'Control-1':'tab:blue', 'Acarbose':'tab:red',\n",
    "          '17'+r'$\\alpha$'+'-Estradiol':'tab:green', 'Rapamycin':'tab:purple',\n",
    "          'Control-2':'tab:blue', '4EGI-1':'tab:orange'}\n",
    "tempD2 = {}\n",
    "for label in ['Control-2', '4EGI-1']:\n",
    "    if tempD1[label]=='tab:blue':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(1)\n",
    "    elif tempD1[label]=='tab:orange':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(3)\n",
    "    elif tempD1[label]=='tab:green':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(5)\n",
    "    elif tempD1[label]=='tab:red':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(7)\n",
    "    elif tempD1[label]=='tab:purple':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(9)\n",
    "    elif tempD1[label]=='tab:olive':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(17)\n",
    "    else:\n",
    "        tempD2[label] = 'Error?'\n",
    "\n",
    "#Visualize each representative\n",
    "for rank_i in range(len(plotL)):\n",
    "    print(' - Rank '+str(rank_i+1)+' (sort by the main effect of '+variable+'):')\n",
    "    module = plotL[rank_i]\n",
    "    #Check module summary\n",
    "    tempDF1 = pd.DataFrame(moduleDF.loc[module]).T\n",
    "    display(tempDF1)\n",
    "    \n",
    "    #Select RMS\n",
    "    tempDF1 = tempDF.loc[tempDF['ModuleID']==module]\n",
    "    tempDF1 = tempDF1.loc[tempDF1['Template'].isin(tempD2.keys())]\n",
    "    \n",
    "    #Check RMS summary\n",
    "    tempDF2 = tempDF1.groupby(['Template', 'Group'])['RMS'].agg(['count', 'mean', 'std'])\n",
    "    tempL1 = []\n",
    "    tempL2 = []\n",
    "    for row_n in tempDF2.index.tolist():\n",
    "        count, mean, std = tempDF2.loc[row_n]\n",
    "        tempL1.append(mean - 1.96*std/np.sqrt(count))\n",
    "        tempL2.append(mean + 1.96*std/np.sqrt(count))\n",
    "    tempDF2['0.025'] = tempL1\n",
    "    tempDF2['0.975'] = tempL2\n",
    "    ##Multiindex sort\n",
    "    tempDF2 = tempDF2.reset_index()\n",
    "    tempDF2['Template'] = pd.Categorical(tempDF2['Template'], categories=list(tempD2.keys()))\n",
    "    tempDF2['Group'] = pd.Categorical(tempDF2['Group'], categories=list(tempD1.keys()))\n",
    "    tempDF2 = tempDF2.sort_values(by=['Template', 'Group']).set_index(['Template', 'Group'])\n",
    "    display(tempDF2)\n",
    "    \n",
    "    #Prepare significance labels\n",
    "    ##Retrieve statistical significance\n",
    "    tempS = pvalDF_ref.loc[module, pvalDF_ref.columns.str.contains('-fixed_.*-vs-')]#Adjusted P-value\n",
    "    tempS.name = 'AdjPval'\n",
    "    ##Clean\n",
    "    tempDF2 = tempS.index.to_series().str.split(pat='-fixed_', expand=True)\n",
    "    tempDF2 = tempDF2.rename(columns={0:'Template', 1:'Comparison'})\n",
    "    tempS1 = tempDF2['Template']\n",
    "    tempDF2 = tempDF2['Comparison'].str.split(pat='-vs-', expand=True)\n",
    "    tempDF2 = tempDF2.rename(columns={0:'Contrast', 1:'Baseline'})\n",
    "    tempDF2 = pd.merge(tempS1, tempDF2, left_index=True, right_index=True, how='left')\n",
    "    tempDF2 = pd.merge(tempDF2, tempS, left_index=True, right_index=True, how='left')\n",
    "    tempDF2['Template'] = tempDF2['Template'].map(tempD0)\n",
    "    tempDF2['Contrast'] = tempDF2['Contrast'].map(tempD0)\n",
    "    tempDF2['Baseline'] = tempDF2['Baseline'].map(tempD0)\n",
    "    ##Convert p-value to label\n",
    "    tempL = []\n",
    "    for row_i in range(len(tempDF2)):\n",
    "        pval = tempDF2['AdjPval'].iloc[row_i]\n",
    "        if pval<0.001:\n",
    "            tempL.append('***')\n",
    "        elif pval<0.01:\n",
    "            tempL.append('**')\n",
    "        elif pval<0.05:\n",
    "            tempL.append('*')\n",
    "        else:\n",
    "            pval_text = Decimal(str(pval)).quantize(Decimal('0.01'), rounding=ROUND_HALF_UP)\n",
    "            tempL.append(r'$P$ = '+str(pval_text))\n",
    "    tempDF2['SignifLabel'] = tempL\n",
    "    display(tempDF2)\n",
    "    \n",
    "    #Visualization\n",
    "    ymax = 1.0\n",
    "    ymin = 0.0\n",
    "    yinter = 0.2\n",
    "    ymargin_t = 0.4\n",
    "    ymargin_b = 0.05\n",
    "    aline_ymin = 1.0\n",
    "    aline_ymargin = 0.125\n",
    "    sns.set(style='ticks', font='Arial', context='talk')\n",
    "    fig, axes = plt.subplots(nrows=1, ncols=2, figsize=(6.4, 5.875), sharex=True, sharey=True,\n",
    "                             gridspec_kw={'width_ratios':[3.2, 3.2]})\n",
    "    for ax_i, ax in enumerate(axes.flat):\n",
    "        template = list(tempD2.keys())[ax_i]\n",
    "        tempDF3 = tempDF1.loc[tempDF1['Template']==template]\n",
    "        sns.pointplot(data=tempDF3, x='Group', y='RMS', order=list(tempD1.keys()), palette=tempD1,\n",
    "                      markers='o', dodge=False, join=False, capsize=0.6, estimator=np.mean, ci=95, ax=ax)\n",
    "        sns.stripplot(data=tempDF3, x='Group', y='RMS',\n",
    "                      order=list(tempD1.keys()), palette=tempD1, dodge=False, jitter=0.15,\n",
    "                      size=5, edgecolor='black', linewidth=1, **{'marker':'o', 'alpha':0.5}, ax=ax)\n",
    "        #Add border line\n",
    "        ax.axvline(x=3.5, **{'linestyle':'dotted', 'color':'black', 'zorder':0})\n",
    "        #Add RCI line\n",
    "        rci = tempDF3['RMS'].loc[tempDF3['Group']==template].mean()\n",
    "        ax.axhline(y=rci, **{'linestyle':'--', 'color':tempD1[template], 'zorder':0})\n",
    "        #Set axis\n",
    "        sns.despine()\n",
    "        ax.set(ylim=(ymin-ymargin_b, ymax+ymargin_t), yticks=np.arange(ymin, ymax + yinter/10, yinter))\n",
    "        plt.setp(ax.get_xticklabels(), rotation=70, horizontalalignment='right',\n",
    "                 verticalalignment='center', rotation_mode='anchor')\n",
    "        if ax_i==0:\n",
    "            plt.setp(ax, xlabel='', ylabel='Sample RMS')\n",
    "        else:\n",
    "            plt.setp(ax.get_yticklabels(), visible=False)\n",
    "            plt.setp(ax, xlabel='', ylabel='')\n",
    "        #Add significance labels\n",
    "        tempDF3 = tempDF2.loc[tempDF2['Template']==template]\n",
    "        for row_i in range(len(tempDF3)):\n",
    "            #Baseline\n",
    "            group_0 = tempDF3['Baseline'].iloc[row_i]\n",
    "            index_0 = list(tempD1.keys()).index(group_0)\n",
    "            xcoord_0 = index_0\n",
    "            #Contrast\n",
    "            group_1 = tempDF3['Contrast'].iloc[row_i]\n",
    "            index_1 = list(tempD1.keys()).index(group_1)\n",
    "            xcoord_1 = index_1\n",
    "            #Standard point of marker\n",
    "            xcoord = (xcoord_0+xcoord_1)/2\n",
    "            ycoord = aline_ymin + aline_ymargin*row_i\n",
    "            label = tempDF3['SignifLabel'].iloc[row_i]\n",
    "            #Add annotation lines\n",
    "            aline_offset = yinter/5\n",
    "            aline_length = yinter/5 + aline_offset/2\n",
    "            ax.plot([xcoord_0, xcoord_0, xcoord_1, xcoord_1],\n",
    "                    [ycoord+aline_offset, ycoord+aline_length, ycoord+aline_length, ycoord+aline_offset],\n",
    "                    lw=1.5, c='k')\n",
    "            #Add annotation text\n",
    "            if label in ['***', '**', '*']:\n",
    "                text_offset = yinter/21\n",
    "                ax.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                            horizontalalignment='center', verticalalignment='bottom',\n",
    "                            fontsize='medium', color='k')\n",
    "            else:\n",
    "                text_offset = yinter/3.5\n",
    "                ax.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                            horizontalalignment='center', verticalalignment='bottom',\n",
    "                            fontsize='x-small', color='k')\n",
    "        #Add annotation\n",
    "        ax.set_title('Consensus:\\n'+template, {'fontsize':'small'})\n",
    "        xoff = 0.015\n",
    "        yoff = 0.01\n",
    "        rect = plt.Rectangle((xoff, 1), 1-xoff, 0.16,#Manual adjustment\n",
    "                             transform=ax.transAxes, facecolor=tempD2[template],\n",
    "                             clip_on=False, linewidth=0, zorder=0.5)\n",
    "        ax.add_patch(rect)\n",
    "    fig.tight_layout()\n",
    "    #Set title\n",
    "    modulename = moduleDF.loc[module, 'ModuleName']\n",
    "    initial = modulename[0].capitalize()\n",
    "    title = re.sub('^.', initial, modulename)+' ('+module+')'\n",
    "    title = '\\n'.join(wrap(title, 60))#Because the below wrap=True didn't work\n",
    "    fig.suptitle(title, size='small',\n",
    "                 verticalalignment='baseline', horizontalalignment='center', wrap=True, y=0.935)\n",
    "    ##Save\n",
    "    fileDir = './ExportFigures/'\n",
    "    ipynbName = '220529_LCproteomics-M004-DIRAC_DIRAC-GOBP-with-M001_ver2-3_'\n",
    "    fileName = 'fixed-RMSmean-inter-group-comparison_RMS-pointplot-'+module.replace('GO:', 'GO')+'.tif'\n",
    "    plt.gcf().savefig(fileDir+ipynbName+fileName, dpi=300, bbox_inches='tight', pad_inches=0.04,\n",
    "                      pil_kwargs={'compression':'tiff_lzw'})\n",
    "    plt.show()\n",
    "    print('')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 4-3-3. Dissimilarly changed modules to 4EGI-1 consensus by any of interventions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Prepare the target module set\n",
    "posL = ['M004:4EGI-fixed_M001:Acar-vs-M001:Cont',\n",
    "        'M004:4EGI-fixed_M001:17aE-vs-M001:Cont',\n",
    "        'M004:4EGI-fixed_M001:Rapa-vs-M001:Cont']\n",
    "negL = ['']\n",
    "template_label = 'M004:4EGI-fixed'\n",
    "similarity = 'Dissimilar'\n",
    "tempS = pd.Series(np.repeat(False, len(pvalDF_ref)), index=pvalDF_ref.index)#Initialize based on nominal ANOVA P-values\n",
    "tempL = pvalDF_ref.loc[:, pvalDF_ref.columns.str.contains(template_label+'_.*-vs-')].columns.tolist()\n",
    "for col_n in tempL:\n",
    "    tempS1 = pvalDF_ref[col_n]\n",
    "    tempS2 = diffDF_ref[col_n]\n",
    "    if col_n in posL:\n",
    "        if similarity=='Similar':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2>0)\n",
    "        elif similarity=='Dissimilar':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2<0)\n",
    "    elif col_n in negL:\n",
    "        tempS3 = (tempS1>=0.05)\n",
    "        #Significance for inverse regulation\n",
    "        if similarity=='Similar':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2<0)\n",
    "        elif similarity=='Dissimilar':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2>0)\n",
    "        tempS1 = tempS3 | tempS1\n",
    "    else:\n",
    "        tempS1 = (tempS1>=0.0)\n",
    "    #Update True\n",
    "    tempS = tempS | tempS1\n",
    "tempL = tempS.loc[tempS.tolist()].index.tolist()\n",
    "print(len(tempL), similarity.lower()+'ly changed modules to '+template_label+' consensus with significance in', posL, 'but not in', negL)\n",
    "\n",
    "#Select representatives\n",
    "topX = np.min([30, len(tempL)])\n",
    "topX_plot = np.min([10, len(tempL)])\n",
    "#Spread statDF by template to flatten multi-index\n",
    "tempDF = moduleDF['ModuleName']#pd.Series() for now\n",
    "for template in ['M004:Cont', 'M004:4EGI']:\n",
    "    tempDF1 = statDF.loc[template].drop(columns=['ModuleName'])\n",
    "    tempDF1.columns = template+'-fixed_'+tempDF1.columns\n",
    "    tempDF = pd.merge(tempDF, tempDF1, left_index=True, right_index=True, how='left')\n",
    "tempDF = tempDF.loc[:, tempDF.columns.str.contains('Pval$')]\n",
    "tempDF = pd.merge(moduleDF['ModuleName'], tempDF, left_index=True, right_index=True, how='left')\n",
    "tempDF = tempDF.loc[tempL]\n",
    "tempDF = tempDF.sort_values(by=['M004:4EGI-fixed_'+variable+'_AdjPval',\n",
    "                                'M004:Cont-fixed_'+variable+'_AdjPval'], ascending=True)\n",
    "print('Top', topX, 'modules (sort by the main effect of '+variable+'):')\n",
    "display(tempDF.iloc[:topX])\n",
    "plotL = tempDF.index.tolist()[:topX_plot]\n",
    "\n",
    "#Prepare DF for plot\n",
    "tempDF = rmsDF.reset_index().melt(var_name='SampleID', value_name='RMS', id_vars=['ModuleID', 'Template'])\n",
    "tempDF1 = sampleDF.reset_index()[['SampleID', 'Phenotype']]\n",
    "tempDF = pd.merge(tempDF, tempDF1, on='SampleID', how='left')\n",
    "\n",
    "#Prepare label and color\n",
    "tempD = {'Control (M001)':'Control-1', 'Acarbose':'Acarbose',\n",
    "         'Estradiol':'17'+r'$\\alpha$'+'-Estradiol', 'Rapamycin':'Rapamycin',\n",
    "         'Control (M004)':'Control-2', '4EGI-1':'4EGI-1'}\n",
    "tempDF['Template'] = tempDF['Template'].map(tempD)\n",
    "tempDF['Group'] = tempDF['Phenotype'].map(tempD)\n",
    "tempD0 = {'M001:Cont':'Control-1', 'M001:Acar':'Acarbose',\n",
    "          'M001:17aE':'17'+r'$\\alpha$'+'-Estradiol', 'M001:Rapa':'Rapamycin',\n",
    "          'M004:Cont':'Control-2', 'M004:4EGI':'4EGI-1'}\n",
    "tempD1 = {'Control-1':'tab:blue', 'Acarbose':'tab:red',\n",
    "          '17'+r'$\\alpha$'+'-Estradiol':'tab:green', 'Rapamycin':'tab:purple',\n",
    "          'Control-2':'tab:blue', '4EGI-1':'tab:orange'}\n",
    "tempD2 = {}\n",
    "for label in ['Control-2', '4EGI-1']:\n",
    "    if tempD1[label]=='tab:blue':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(1)\n",
    "    elif tempD1[label]=='tab:orange':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(3)\n",
    "    elif tempD1[label]=='tab:green':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(5)\n",
    "    elif tempD1[label]=='tab:red':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(7)\n",
    "    elif tempD1[label]=='tab:purple':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(9)\n",
    "    elif tempD1[label]=='tab:olive':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(17)\n",
    "    else:\n",
    "        tempD2[label] = 'Error?'\n",
    "\n",
    "#Visualize each representative\n",
    "for rank_i in range(len(plotL)):\n",
    "    print(' - Rank '+str(rank_i+1)+' (sort by the main effect of '+variable+'):')\n",
    "    module = plotL[rank_i]\n",
    "    #Check module summary\n",
    "    tempDF1 = pd.DataFrame(moduleDF.loc[module]).T\n",
    "    display(tempDF1)\n",
    "    \n",
    "    #Select RMS\n",
    "    tempDF1 = tempDF.loc[tempDF['ModuleID']==module]\n",
    "    tempDF1 = tempDF1.loc[tempDF1['Template'].isin(tempD2.keys())]\n",
    "    \n",
    "    #Check RMS summary\n",
    "    tempDF2 = tempDF1.groupby(['Template', 'Group'])['RMS'].agg(['count', 'mean', 'std'])\n",
    "    tempL1 = []\n",
    "    tempL2 = []\n",
    "    for row_n in tempDF2.index.tolist():\n",
    "        count, mean, std = tempDF2.loc[row_n]\n",
    "        tempL1.append(mean - 1.96*std/np.sqrt(count))\n",
    "        tempL2.append(mean + 1.96*std/np.sqrt(count))\n",
    "    tempDF2['0.025'] = tempL1\n",
    "    tempDF2['0.975'] = tempL2\n",
    "    ##Multiindex sort\n",
    "    tempDF2 = tempDF2.reset_index()\n",
    "    tempDF2['Template'] = pd.Categorical(tempDF2['Template'], categories=list(tempD2.keys()))\n",
    "    tempDF2['Group'] = pd.Categorical(tempDF2['Group'], categories=list(tempD1.keys()))\n",
    "    tempDF2 = tempDF2.sort_values(by=['Template', 'Group']).set_index(['Template', 'Group'])\n",
    "    display(tempDF2)\n",
    "    \n",
    "    #Prepare significance labels\n",
    "    ##Retrieve statistical significance\n",
    "    tempS = pvalDF_ref.loc[module, pvalDF_ref.columns.str.contains('-fixed_.*-vs-')]#Adjusted P-value\n",
    "    tempS.name = 'AdjPval'\n",
    "    ##Clean\n",
    "    tempDF2 = tempS.index.to_series().str.split(pat='-fixed_', expand=True)\n",
    "    tempDF2 = tempDF2.rename(columns={0:'Template', 1:'Comparison'})\n",
    "    tempS1 = tempDF2['Template']\n",
    "    tempDF2 = tempDF2['Comparison'].str.split(pat='-vs-', expand=True)\n",
    "    tempDF2 = tempDF2.rename(columns={0:'Contrast', 1:'Baseline'})\n",
    "    tempDF2 = pd.merge(tempS1, tempDF2, left_index=True, right_index=True, how='left')\n",
    "    tempDF2 = pd.merge(tempDF2, tempS, left_index=True, right_index=True, how='left')\n",
    "    tempDF2['Template'] = tempDF2['Template'].map(tempD0)\n",
    "    tempDF2['Contrast'] = tempDF2['Contrast'].map(tempD0)\n",
    "    tempDF2['Baseline'] = tempDF2['Baseline'].map(tempD0)\n",
    "    ##Convert p-value to label\n",
    "    tempL = []\n",
    "    for row_i in range(len(tempDF2)):\n",
    "        pval = tempDF2['AdjPval'].iloc[row_i]\n",
    "        if pval<0.001:\n",
    "            tempL.append('***')\n",
    "        elif pval<0.01:\n",
    "            tempL.append('**')\n",
    "        elif pval<0.05:\n",
    "            tempL.append('*')\n",
    "        else:\n",
    "            pval_text = Decimal(str(pval)).quantize(Decimal('0.01'), rounding=ROUND_HALF_UP)\n",
    "            tempL.append(r'$P$ = '+str(pval_text))\n",
    "    tempDF2['SignifLabel'] = tempL\n",
    "    display(tempDF2)\n",
    "    \n",
    "    #Visualization\n",
    "    ymax = 1.0\n",
    "    ymin = 0.0\n",
    "    yinter = 0.2\n",
    "    ymargin_t = 0.4\n",
    "    ymargin_b = 0.05\n",
    "    aline_ymin = 1.0\n",
    "    aline_ymargin = 0.125\n",
    "    sns.set(style='ticks', font='Arial', context='talk')\n",
    "    fig, axes = plt.subplots(nrows=1, ncols=2, figsize=(6.4, 5.875), sharex=True, sharey=True,\n",
    "                             gridspec_kw={'width_ratios':[3.2, 3.2]})\n",
    "    for ax_i, ax in enumerate(axes.flat):\n",
    "        template = list(tempD2.keys())[ax_i]\n",
    "        tempDF3 = tempDF1.loc[tempDF1['Template']==template]\n",
    "        sns.pointplot(data=tempDF3, x='Group', y='RMS', order=list(tempD1.keys()), palette=tempD1,\n",
    "                      markers='o', dodge=False, join=False, capsize=0.6, estimator=np.mean, ci=95, ax=ax)\n",
    "        sns.stripplot(data=tempDF3, x='Group', y='RMS',\n",
    "                      order=list(tempD1.keys()), palette=tempD1, dodge=False, jitter=0.15,\n",
    "                      size=5, edgecolor='black', linewidth=1, **{'marker':'o', 'alpha':0.5}, ax=ax)\n",
    "        #Add border line\n",
    "        ax.axvline(x=3.5, **{'linestyle':'dotted', 'color':'black', 'zorder':0})\n",
    "        #Add RCI line\n",
    "        rci = tempDF3['RMS'].loc[tempDF3['Group']==template].mean()\n",
    "        ax.axhline(y=rci, **{'linestyle':'--', 'color':tempD1[template], 'zorder':0})\n",
    "        #Set axis\n",
    "        sns.despine()\n",
    "        ax.set(ylim=(ymin-ymargin_b, ymax+ymargin_t), yticks=np.arange(ymin, ymax + yinter/10, yinter))\n",
    "        plt.setp(ax.get_xticklabels(), rotation=70, horizontalalignment='right',\n",
    "                 verticalalignment='center', rotation_mode='anchor')\n",
    "        if ax_i==0:\n",
    "            plt.setp(ax, xlabel='', ylabel='Sample RMS')\n",
    "        else:\n",
    "            plt.setp(ax.get_yticklabels(), visible=False)\n",
    "            plt.setp(ax, xlabel='', ylabel='')\n",
    "        #Add significance labels\n",
    "        tempDF3 = tempDF2.loc[tempDF2['Template']==template]\n",
    "        for row_i in range(len(tempDF3)):\n",
    "            #Baseline\n",
    "            group_0 = tempDF3['Baseline'].iloc[row_i]\n",
    "            index_0 = list(tempD1.keys()).index(group_0)\n",
    "            xcoord_0 = index_0\n",
    "            #Contrast\n",
    "            group_1 = tempDF3['Contrast'].iloc[row_i]\n",
    "            index_1 = list(tempD1.keys()).index(group_1)\n",
    "            xcoord_1 = index_1\n",
    "            #Standard point of marker\n",
    "            xcoord = (xcoord_0+xcoord_1)/2\n",
    "            ycoord = aline_ymin + aline_ymargin*row_i\n",
    "            label = tempDF3['SignifLabel'].iloc[row_i]\n",
    "            #Add annotation lines\n",
    "            aline_offset = yinter/5\n",
    "            aline_length = yinter/5 + aline_offset/2\n",
    "            ax.plot([xcoord_0, xcoord_0, xcoord_1, xcoord_1],\n",
    "                    [ycoord+aline_offset, ycoord+aline_length, ycoord+aline_length, ycoord+aline_offset],\n",
    "                    lw=1.5, c='k')\n",
    "            #Add annotation text\n",
    "            if label in ['***', '**', '*']:\n",
    "                text_offset = yinter/21\n",
    "                ax.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                            horizontalalignment='center', verticalalignment='bottom',\n",
    "                            fontsize='medium', color='k')\n",
    "            else:\n",
    "                text_offset = yinter/3.5\n",
    "                ax.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                            horizontalalignment='center', verticalalignment='bottom',\n",
    "                            fontsize='x-small', color='k')\n",
    "        #Add annotation\n",
    "        ax.set_title('Consensus:\\n'+template, {'fontsize':'small'})\n",
    "        xoff = 0.015\n",
    "        yoff = 0.01\n",
    "        rect = plt.Rectangle((xoff, 1), 1-xoff, 0.16,#Manual adjustment\n",
    "                             transform=ax.transAxes, facecolor=tempD2[template],\n",
    "                             clip_on=False, linewidth=0, zorder=0.5)\n",
    "        ax.add_patch(rect)\n",
    "    fig.tight_layout()\n",
    "    #Set title\n",
    "    modulename = moduleDF.loc[module, 'ModuleName']\n",
    "    initial = modulename[0].capitalize()\n",
    "    title = re.sub('^.', initial, modulename)+' ('+module+')'\n",
    "    title = '\\n'.join(wrap(title, 60))#Because the below wrap=True didn't work\n",
    "    fig.suptitle(title, size='small',\n",
    "                 verticalalignment='baseline', horizontalalignment='center', wrap=True, y=0.935)\n",
    "    ##Save\n",
    "    fileDir = './ExportFigures/'\n",
    "    ipynbName = '220529_LCproteomics-M004-DIRAC_DIRAC-GOBP-with-M001_ver2-3_'\n",
    "    fileName = 'fixed-RMSmean-inter-group-comparison_RMS-pointplot-'+module.replace('GO:', 'GO')+'.tif'\n",
    "    plt.gcf().savefig(fileDir+ipynbName+fileName, dpi=300, bbox_inches='tight', pad_inches=0.04,\n",
    "                      pil_kwargs={'compression':'tiff_lzw'})\n",
    "    plt.show()\n",
    "    print('')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 4-3-4. Modules of special interest"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plotL = ['GO:0006635', 'GO:0031998', 'GO:0016558', 'GO:0006625',#Common in M001\n",
    "         'GO:0006637', 'GO:0006734', 'GO:0001561',#Related to fatty acid oxidation (mentioned in previous manuscript)\n",
    "         'GO:0023035', 'GO:0039529',#Immune response (mentioned in previous manuscript)\n",
    "         'GO:0010638', 'GO:0002181', 'GO:0045899',#Mentioned in previous manuscript\n",
    "         'GO:0006635', 'GO:1990126', 'GO:0098761',#Common in proteomics and transcriptomics\n",
    "         'GO:0061732', 'GO:0006086', 'GO:0019441',#Specific to proteomics (Aca)\n",
    "         'GO:0031998', 'GO:0044794', 'GO:0006625', 'GO:0016558', 'GO:0019441', 'GO:0033572',#Specific to proteomics (Rapa)\n",
    "         'GO:0006635', 'GO:0002181', 'GO:0000028', 'GO:0016558', 'GO:0006734', 'GO:0098761',#Common in M001 + M004\n",
    "         'GO:0015986', 'GO:0042776', 'GO:0070934', 'GO:0006703', 'GO:0034354', 'GO:0045899']#Similar to 4EGI-1\n",
    "plotL = list(set(plotL))\n",
    "plotL.sort()\n",
    "for module in plotL:\n",
    "    if module not in moduleDF.index.tolist():\n",
    "        print(module+' was NOT included in this analysis.')\n",
    "        plotL.remove(module)\n",
    "\n",
    "#Prepare the target module set\n",
    "##Spread statDF by template to flatten multi-index\n",
    "tempDF = moduleDF['ModuleName']#pd.Series() for now\n",
    "for template in ['M004:Cont', 'M004:4EGI']:\n",
    "    tempDF1 = statDF.loc[template].drop(columns=['ModuleName'])\n",
    "    tempDF1.columns = template+'-fixed_'+tempDF1.columns\n",
    "    tempDF = pd.merge(tempDF, tempDF1, left_index=True, right_index=True, how='left')\n",
    "tempDF = tempDF.loc[:, tempDF.columns.str.contains('Pval$')]\n",
    "tempDF = pd.merge(moduleDF['ModuleName'], tempDF, left_index=True, right_index=True, how='left')\n",
    "tempDF = tempDF.loc[plotL]\n",
    "display(tempDF)\n",
    "\n",
    "#Prepare DF for plot\n",
    "tempDF = rmsDF.reset_index().melt(var_name='SampleID', value_name='RMS', id_vars=['ModuleID', 'Template'])\n",
    "tempDF1 = sampleDF.reset_index()[['SampleID', 'Phenotype']]\n",
    "tempDF = pd.merge(tempDF, tempDF1, on='SampleID', how='left')\n",
    "\n",
    "#Prepare label and color\n",
    "tempD = {'Control (M001)':'Control-1', 'Acarbose':'Acarbose',\n",
    "         'Estradiol':'17'+r'$\\alpha$'+'-Estradiol', 'Rapamycin':'Rapamycin',\n",
    "         'Control (M004)':'Control-2', '4EGI-1':'4EGI-1'}\n",
    "tempDF['Template'] = tempDF['Template'].map(tempD)\n",
    "tempDF['Group'] = tempDF['Phenotype'].map(tempD)\n",
    "tempD0 = {'M001:Cont':'Control-1', 'M001:Acar':'Acarbose',\n",
    "          'M001:17aE':'17'+r'$\\alpha$'+'-Estradiol', 'M001:Rapa':'Rapamycin',\n",
    "          'M004:Cont':'Control-2', 'M004:4EGI':'4EGI-1'}\n",
    "tempD1 = {'Control-1':'tab:blue', 'Acarbose':'tab:red',\n",
    "          '17'+r'$\\alpha$'+'-Estradiol':'tab:green', 'Rapamycin':'tab:purple',\n",
    "          'Control-2':'tab:blue', '4EGI-1':'tab:orange'}\n",
    "tempD2 = {}\n",
    "for label in ['Control-2', '4EGI-1']:\n",
    "    if tempD1[label]=='tab:blue':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(1)\n",
    "    elif tempD1[label]=='tab:orange':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(3)\n",
    "    elif tempD1[label]=='tab:green':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(5)\n",
    "    elif tempD1[label]=='tab:red':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(7)\n",
    "    elif tempD1[label]=='tab:purple':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(9)\n",
    "    elif tempD1[label]=='tab:olive':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(17)\n",
    "    else:\n",
    "        tempD2[label] = 'Error?'\n",
    "\n",
    "#Visualize each representative\n",
    "for rank_i in range(len(plotL)):\n",
    "    print(' - Rank '+str(rank_i+1)+' (sort by the main effect of '+variable+'):')\n",
    "    module = plotL[rank_i]\n",
    "    #Check module summary\n",
    "    tempDF1 = pd.DataFrame(moduleDF.loc[module]).T\n",
    "    display(tempDF1)\n",
    "    \n",
    "    #Select RMS\n",
    "    tempDF1 = tempDF.loc[tempDF['ModuleID']==module]\n",
    "    tempDF1 = tempDF1.loc[tempDF1['Template'].isin(tempD2.keys())]\n",
    "    \n",
    "    #Check RMS summary\n",
    "    tempDF2 = tempDF1.groupby(['Template', 'Group'])['RMS'].agg(['count', 'mean', 'std'])\n",
    "    tempL1 = []\n",
    "    tempL2 = []\n",
    "    for row_n in tempDF2.index.tolist():\n",
    "        count, mean, std = tempDF2.loc[row_n]\n",
    "        tempL1.append(mean - 1.96*std/np.sqrt(count))\n",
    "        tempL2.append(mean + 1.96*std/np.sqrt(count))\n",
    "    tempDF2['0.025'] = tempL1\n",
    "    tempDF2['0.975'] = tempL2\n",
    "    ##Multiindex sort\n",
    "    tempDF2 = tempDF2.reset_index()\n",
    "    tempDF2['Template'] = pd.Categorical(tempDF2['Template'], categories=list(tempD2.keys()))\n",
    "    tempDF2['Group'] = pd.Categorical(tempDF2['Group'], categories=list(tempD1.keys()))\n",
    "    tempDF2 = tempDF2.sort_values(by=['Template', 'Group']).set_index(['Template', 'Group'])\n",
    "    display(tempDF2)\n",
    "    \n",
    "    #Prepare significance labels\n",
    "    ##Re-prepare p-values because these modules can be filtered out in pvalDF\n",
    "    tempDF2 = moduleDF['ModuleName']#pd.Series() for now\n",
    "    for template in ['M004:Cont', 'M004:4EGI']:\n",
    "        tempDF3 = statDF.loc[template].drop(columns=['ModuleName'])\n",
    "        tempDF3.columns = template+'-fixed_'+tempDF3.columns\n",
    "        tempDF2 = pd.merge(tempDF2, tempDF3, left_index=True, right_index=True, how='left')\n",
    "    tempDF2 = tempDF2.loc[:, tempDF2.columns.str.contains('-vs-.*_AdjPval$')]\n",
    "    tempDF2.columns = tempDF2.columns.str.replace('_AdjPval$', '')\n",
    "    ##Retrieve statistical significance\n",
    "    tempS = tempDF2.loc[module]\n",
    "    tempS.name = 'AdjPval'\n",
    "    ##Clean\n",
    "    tempDF2 = tempS.index.to_series().str.split(pat='-fixed_', expand=True)\n",
    "    tempDF2 = tempDF2.rename(columns={0:'Template', 1:'Comparison'})\n",
    "    tempS1 = tempDF2['Template']\n",
    "    tempDF2 = tempDF2['Comparison'].str.split(pat='-vs-', expand=True)\n",
    "    tempDF2 = tempDF2.rename(columns={0:'Contrast', 1:'Baseline'})\n",
    "    tempDF2 = pd.merge(tempS1, tempDF2, left_index=True, right_index=True, how='left')\n",
    "    tempDF2 = pd.merge(tempDF2, tempS, left_index=True, right_index=True, how='left')\n",
    "    tempDF2['Template'] = tempDF2['Template'].map(tempD0)\n",
    "    tempDF2['Contrast'] = tempDF2['Contrast'].map(tempD0)\n",
    "    tempDF2['Baseline'] = tempDF2['Baseline'].map(tempD0)\n",
    "    ##Convert p-value to label\n",
    "    tempL = []\n",
    "    for row_i in range(len(tempDF2)):\n",
    "        pval = tempDF2['AdjPval'].iloc[row_i]\n",
    "        if pval<0.001:\n",
    "            tempL.append('***')\n",
    "        elif pval<0.01:\n",
    "            tempL.append('**')\n",
    "        elif pval<0.05:\n",
    "            tempL.append('*')\n",
    "        else:\n",
    "            pval_text = Decimal(str(pval)).quantize(Decimal('0.01'), rounding=ROUND_HALF_UP)\n",
    "            tempL.append(r'$P$ = '+str(pval_text))\n",
    "    tempDF2['SignifLabel'] = tempL\n",
    "    display(tempDF2)\n",
    "    \n",
    "    #Visualization\n",
    "    ymax = 1.0\n",
    "    ymin = 0.0\n",
    "    yinter = 0.2\n",
    "    ymargin_t = 0.4\n",
    "    ymargin_b = 0.05\n",
    "    aline_ymin = 1.0\n",
    "    aline_ymargin = 0.125\n",
    "    sns.set(style='ticks', font='Arial', context='talk')\n",
    "    fig, axes = plt.subplots(nrows=1, ncols=2, figsize=(6.4, 5.875), sharex=True, sharey=True,\n",
    "                             gridspec_kw={'width_ratios':[3.2, 3.2]})\n",
    "    for ax_i, ax in enumerate(axes.flat):\n",
    "        template = list(tempD2.keys())[ax_i]\n",
    "        tempDF3 = tempDF1.loc[tempDF1['Template']==template]\n",
    "        sns.pointplot(data=tempDF3, x='Group', y='RMS', order=list(tempD1.keys()), palette=tempD1,\n",
    "                      markers='o', dodge=False, join=False, capsize=0.6, estimator=np.mean, ci=95, ax=ax)\n",
    "        sns.stripplot(data=tempDF3, x='Group', y='RMS',\n",
    "                      order=list(tempD1.keys()), palette=tempD1, dodge=False, jitter=0.15,\n",
    "                      size=5, edgecolor='black', linewidth=1, **{'marker':'o', 'alpha':0.5}, ax=ax)\n",
    "        #Add border line\n",
    "        ax.axvline(x=3.5, **{'linestyle':'dotted', 'color':'black', 'zorder':0})\n",
    "        #Add RCI line\n",
    "        rci = tempDF3['RMS'].loc[tempDF3['Group']==template].mean()\n",
    "        ax.axhline(y=rci, **{'linestyle':'--', 'color':tempD1[template], 'zorder':0})\n",
    "        #Set axis\n",
    "        sns.despine()\n",
    "        ax.set(ylim=(ymin-ymargin_b, ymax+ymargin_t), yticks=np.arange(ymin, ymax + yinter/10, yinter))\n",
    "        plt.setp(ax.get_xticklabels(), rotation=70, horizontalalignment='right',\n",
    "                 verticalalignment='center', rotation_mode='anchor')\n",
    "        if ax_i==0:\n",
    "            plt.setp(ax, xlabel='', ylabel='Sample RMS')\n",
    "        else:\n",
    "            plt.setp(ax.get_yticklabels(), visible=False)\n",
    "            plt.setp(ax, xlabel='', ylabel='')\n",
    "        #Add significance labels\n",
    "        tempDF3 = tempDF2.loc[tempDF2['Template']==template]\n",
    "        for row_i in range(len(tempDF3)):\n",
    "            #Baseline\n",
    "            group_0 = tempDF3['Baseline'].iloc[row_i]\n",
    "            index_0 = list(tempD1.keys()).index(group_0)\n",
    "            xcoord_0 = index_0\n",
    "            #Contrast\n",
    "            group_1 = tempDF3['Contrast'].iloc[row_i]\n",
    "            index_1 = list(tempD1.keys()).index(group_1)\n",
    "            xcoord_1 = index_1\n",
    "            #Standard point of marker\n",
    "            xcoord = (xcoord_0+xcoord_1)/2\n",
    "            ycoord = aline_ymin + aline_ymargin*row_i\n",
    "            label = tempDF3['SignifLabel'].iloc[row_i]\n",
    "            #Add annotation lines\n",
    "            aline_offset = yinter/5\n",
    "            aline_length = yinter/5 + aline_offset/2\n",
    "            ax.plot([xcoord_0, xcoord_0, xcoord_1, xcoord_1],\n",
    "                    [ycoord+aline_offset, ycoord+aline_length, ycoord+aline_length, ycoord+aline_offset],\n",
    "                    lw=1.5, c='k')\n",
    "            #Add annotation text\n",
    "            if label in ['***', '**', '*']:\n",
    "                text_offset = yinter/21\n",
    "                ax.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                            horizontalalignment='center', verticalalignment='bottom',\n",
    "                            fontsize='medium', color='k')\n",
    "            else:\n",
    "                text_offset = yinter/3.5\n",
    "                ax.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                            horizontalalignment='center', verticalalignment='bottom',\n",
    "                            fontsize='x-small', color='k')\n",
    "        #Add annotation\n",
    "        ax.set_title('Consensus:\\n'+template, {'fontsize':'small'})\n",
    "        xoff = 0.015\n",
    "        yoff = 0.01\n",
    "        rect = plt.Rectangle((xoff, 1), 1-xoff, 0.16,#Manual adjustment\n",
    "                             transform=ax.transAxes, facecolor=tempD2[template],\n",
    "                             clip_on=False, linewidth=0, zorder=0.5)\n",
    "        ax.add_patch(rect)\n",
    "    fig.tight_layout()\n",
    "    #Set title\n",
    "    modulename = moduleDF.loc[module, 'ModuleName']\n",
    "    initial = modulename[0].capitalize()\n",
    "    title = re.sub('^.', initial, modulename)+' ('+module+')'\n",
    "    title = '\\n'.join(wrap(title, 60))#Because the below wrap=True didn't work\n",
    "    fig.suptitle(title, size='small',\n",
    "                 verticalalignment='baseline', horizontalalignment='center', wrap=True, y=0.935)\n",
    "    ##Save\n",
    "    fileDir = './ExportFigures/'\n",
    "    ipynbName = '220529_LCproteomics-M004-DIRAC_DIRAC-GOBP-with-M001_ver2-3_'\n",
    "    fileName = 'fixed-RMSmean-inter-group-comparison_RMS-pointplot-'+module.replace('GO:', 'GO')+'.tif'\n",
    "    plt.gcf().savefig(fileDir+ipynbName+fileName, dpi=300, bbox_inches='tight', pad_inches=0.04,\n",
    "                      pil_kwargs={'compression':'tiff_lzw'})\n",
    "    plt.show()\n",
    "    print('')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# — End of notebook —"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "arivale-py3 - Python",
   "language": "python",
   "name": "conda-env-arivale-py3-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  },
  "widgets": {
   "application/vnd.jupyter.widget-state+json": {
    "state": {},
    "version_major": 2,
    "version_minor": 0
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
