{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# DIRAC Analyses of LC M001 Proteomics and Related Transcriptomics — Comparisons in Common GOBP Modules"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***by Kengo Watanabe***  \n",
    "\n",
    "The differential rank conservation (DIRAC; Eddy, J.A. et al. PLoS Comput. Biol. 2010) analyses were performed on (1) the Longevity Consortium (LC) M001 proteomics dataset using a priori module set (Gene Ontology Biological Process (GOBP) derived by EMBL-EBI QuickGO API; ≥4 analytes and ≥50% coverage) and (2) the LC M001-related transcriptomics dataset (Tyshkovskiy, A. et al. Cell Metab. 2019) using a priori module set (GOBP derived by R org.Mm.eg.db package; ≥4 analytes and ≥50% coverage).  \n",
    "–> In this notebook, these results are further compared between the datasets.  \n",
    "> To maintain the consistency with the other DIRAC analyses, statistical tests are performed in a different notebook with R kernel.  \n",
    "\n",
    "Input:  \n",
    "* Cleaned module metadata (proteomics): 220520_LCproteomics-M001-DIRAC-ver6-4_DIRAC-GOBP_module-metadata.tsv  \n",
    "* Cleaned module metadata (transcriptomics): 220522_LC-M001-related-transcriptomics-DIRAC_DIRAC-GOBP_ver2-4_module-metadata.tsv  \n",
    "* Sample–mouse metadata (proteomics): 210126_LCprotomics-M001-DIRAC-ver6_preprocessing_metadata-sample.tsv  \n",
    "* Sample–mouse metadata (transcriptomics): 201221_LC-M001-related-transcriptomics-DIRAC_metadata.tsv  \n",
    "* Combined table of DIRAC RMSs (proteomics): 210127_LCproteomics-M001-DIRAC-ver6_DIRAC-GOBP_QuickGO-GOBP_min-n4-cov50_RankMatchingScore-BS.tsv  \n",
    "* Combined table of DIRAC RCIs (proteomics): 210127_LCproteomics-M001-DIRAC-ver6_DIRAC-GOBP_QuickGO-GOBP_min-n4-cov50_RankConservationIndex-BS.tsv  \n",
    "* Tables of DIRAC RMSs (transcriptomics): 210429_LC-M001-related-transcriptomics-DIRAC_DIRAC-GOBP_ver2_orgMmegdb-GOBP_min-n4-cov50_RankMatchingScore-BS-[digit].tsv  \n",
    "* Tables of DIRAC RCIs (transcriptomics): 210429_LC-M001-related-transcriptomics-DIRAC_DIRAC-GOBP_ver2_orgMmegdb-GOBP_min-n4-cov50_RankConservationIndex-BS-[digit].tsv  \n",
    "* Statistical test summary: 220525_LC-M001-DIRAC-prot-vs-txn_StatisticalTest-GOBP_ver2-2_inter-group-comparison.xlsx (Supplementary Data 6)  \n",
    "\n",
    "Output:  \n",
    "* Combined module metadata, which is incorporated into Supplementary Data 6 in R sub-notebook  \n",
    "* Combined sample–mouse metadata, which is used in R sub-notebook  \n",
    "* Cleaned tables of DIRAC measures, which are used in statistical analysis (R sub-notebook)  \n",
    "* Figure 5b–d  \n",
    "\n",
    "Original notebook (memo for my future tracing):  \n",
    "* dalek:[JupyterLab HOME]/220523_LC-M001-DIRAC-prot-vs-txn/220525_LC-M001-DIRAC-prot-vs-txn_Comparison-GOBP_ver2-2.ipynb  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scipy.stats as stats\n",
    "import matplotlib.pyplot as plt\n",
    "%matplotlib inline\n",
    "import seaborn as sns\n",
    "#For Arial font\n",
    "#!conda install -c conda-forge -y mscorefonts\n",
    "##-> The below was also needed in matplotlib 3.4.2\n",
    "#import shutil\n",
    "#import matplotlib\n",
    "#shutil.rmtree(matplotlib.get_cachedir())\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')\n",
    "from IPython.display import display\n",
    "import time\n",
    "\n",
    "from decimal import Decimal, ROUND_HALF_UP\n",
    "import re\n",
    "import matplotlib.patches as mpatches\n",
    "#!pip install venn\n",
    "from venn import venn\n",
    "#!conda install -c conda-forge -y matplotlib-venn\n",
    "from matplotlib_venn import venn3, venn3_circles, venn2, venn2_circles\n",
    "from mpl_toolkits.axes_grid1 import make_axes_locatable\n",
    "from matplotlib.offsetbox import AnchoredText\n",
    "from textwrap import wrap\n",
    "\n",
    "!conda list"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. Prepare metadata and DIRAC results"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1-1. Module metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Import module metadata\n",
    "fileDir = '../210126_LCproteomics-M001-DIRAC-ver6/ExportData/'\n",
    "ipynbName = '220520_LCproteomics-M001-DIRAC-ver6-4_DIRAC-GOBP_'\n",
    "fileName = 'module-metadata.tsv'\n",
    "tempDF1 = pd.read_csv(fileDir+ipynbName+fileName, sep='\\t').set_index('ModuleID')\n",
    "print('Proteomics:', len(tempDF1))\n",
    "\n",
    "fileDir = '../201221_LC-M001-related-transcriptomics-DIRAC/ExportData/'\n",
    "ipynbName = '220522_LC-M001-related-transcriptomics-DIRAC_DIRAC-GOBP_ver2-4_'\n",
    "fileName = 'module-metadata.tsv'\n",
    "tempDF2 = pd.read_csv(fileDir+ipynbName+fileName, sep='\\t').set_index('ModuleID')\n",
    "print('Transcriptomics:', len(tempDF2))\n",
    "\n",
    "#Retrieve common modules\n",
    "tempL = ['ModuleName', 'ModuleType']\n",
    "tempDF = tempDF1.loc[tempDF1.index.isin(tempDF2.index.tolist()), tempL]\n",
    "tempDF1 = tempDF1.drop(columns=tempL)\n",
    "tempDF1.columns = 'Prot_'+tempDF1.columns\n",
    "tempDF = pd.merge(tempDF, tempDF1, left_index=True, right_index=True, how='left')\n",
    "tempDF2 = tempDF2.drop(columns=tempL)\n",
    "tempDF2.columns = 'Tran_'+tempDF2.columns\n",
    "tempDF = pd.merge(tempDF, tempDF2, left_index=True, right_index=True, how='left')\n",
    "print('Common:', len(tempDF))\n",
    "display(tempDF)\n",
    "\n",
    "#Save for using in the sub-notebook\n",
    "fileDir = './ExportData/'\n",
    "ipynbName = '220525_LC-M001-DIRAC-prot-vs-txn_Comparison-GOBP_ver2-2_'\n",
    "fileName = 'module-metadata.tsv'\n",
    "tempDF.to_csv(fileDir+ipynbName+fileName, index=True, sep='\\t')\n",
    "\n",
    "moduleDF = tempDF"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1-2. Sample–mouse metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Import sample-mouse metadata\n",
    "fileDir = '../210126_LCproteomics-M001-DIRAC-ver6/ExportData/'\n",
    "ipynbName = '210126_LCprotomics-M001-DIRAC-ver6_preprocessing_'\n",
    "fileName = 'metadata-sample.tsv'\n",
    "tempDF1 = pd.read_csv(fileDir+ipynbName+fileName, sep='\\t')\n",
    "tempDF1 = tempDF1.rename(columns={'MouseID':'SampleID', 'Treatment':'Intervention'})\n",
    "tempDF1 = tempDF1.set_index('SampleID')\n",
    "tempDF1['Age'] = '12m'\n",
    "tempDF1['Dataset'] = 'Proteomics'\n",
    "\n",
    "fileDir = '../201221_LC-M001-related-transcriptomics-DIRAC/ExportData/'\n",
    "ipynbName = '201221_LC-M001-related-transcriptomics-DIRAC_'\n",
    "fileName = 'metadata.tsv'\n",
    "tempDF2 = pd.read_csv(fileDir+ipynbName+fileName, sep='\\t')\n",
    "tempDF2 = tempDF2.rename(columns={'ID':'SampleID'})\n",
    "tempDF2 = tempDF2.set_index('SampleID')\n",
    "tempD = {'-':'Control', 'Acarbose':'Acarbose', 'Rapamycin':'Rapamycin', 'CR':'Calorie restriction'}\n",
    "tempDF2['Intervention'] = tempDF2['Intervention'].map(tempD)\n",
    "tempD = {6:'6m', 12:'12m'}\n",
    "tempDF2['Age'] = tempDF2['Age'].map(tempD)\n",
    "tempDF2['Dataset'] = 'Transcriptomics'\n",
    "\n",
    "#Take only the groups to be comparaed\n",
    "tempL1 = ['Control', 'Acarbose', 'Rapamycin']\n",
    "tempL2 = ['Dataset', 'Intervention', 'Sex', 'Age']\n",
    "tempDF1 = tempDF1.loc[tempDF1['Intervention'].isin(tempL1), tempL2]\n",
    "tempDF2 = tempDF2.loc[tempDF2['Intervention'].isin(tempL1), tempL2]\n",
    "tempDF = pd.concat([tempDF1, tempDF2], axis=0)\n",
    "tempDF['Phenotype'] = tempDF['Intervention']#For DIRAC template (rank consensus)\n",
    "tempDF['Group'] = tempDF['Dataset'].str.slice(start=0, stop=1)+'-'+tempDF['Phenotype']\n",
    "\n",
    "display(tempDF)\n",
    "display(tempDF['Group'].value_counts())\n",
    "\n",
    "#Save for using in the sub-notebook\n",
    "fileDir = './ExportData/'\n",
    "ipynbName = '220525_LC-M001-DIRAC-prot-vs-txn_Comparison-GOBP_ver2-2_'\n",
    "fileName = 'sample-metadata.tsv'\n",
    "tempDF.to_csv(fileDir+ipynbName+fileName, index=True, sep='\\t')\n",
    "\n",
    "sampleDF = tempDF"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1-3. DIRAC results (sex/age-pooled rank consensus)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Import DIRAC results\n",
    "fileDir = '../210126_LCproteomics-M001-DIRAC-ver6/ExportData/'\n",
    "ipynbName = '210127_LCproteomics-M001-DIRAC-ver6_DIRAC-GOBP_'\n",
    "fileName = 'QuickGO-GOBP_min-n4-cov50_RankMatchingScore-BS.tsv'\n",
    "tempDF1 = pd.read_csv(fileDir+ipynbName+fileName, sep='\\t')\n",
    "tempDF1 = tempDF1.rename(columns={'NetworkID':'ModuleID'})\n",
    "fileName = 'QuickGO-GOBP_min-n4-cov50_RankConservationIndex-BS.tsv'\n",
    "tempDF2 = pd.read_csv(fileDir+ipynbName+fileName, sep='\\t')\n",
    "tempDF2 = tempDF2.rename(columns={'NetworkID':'ModuleID'})\n",
    "\n",
    "#Take only the modules and samples/groups to be compared\n",
    "print('Original RMS:', tempDF1.shape)\n",
    "tempDF1 = tempDF1.loc[tempDF1['ModuleID'].isin(moduleDF.index.tolist())]\n",
    "tempDF1 = tempDF1.loc[tempDF1['Template'].isin(sampleDF['Phenotype'].unique())]\n",
    "tempDF1 = tempDF1.set_index(['ModuleID', 'Template'])\n",
    "tempDF1 = tempDF1.loc[:, tempDF1.columns.isin(sampleDF.index.tolist())]\n",
    "tempDF1 = tempDF1.reset_index()\n",
    "print(' -> After:', tempDF1.shape)\n",
    "print('Original RCI:', tempDF2.shape)\n",
    "tempDF2 = tempDF2.loc[tempDF2['ModuleID'].isin(moduleDF.index.tolist())]\n",
    "tempDF2 = tempDF2.loc[tempDF2['Template'].isin(sampleDF['Phenotype'].unique())]\n",
    "tempDF2 = tempDF2.set_index(['ModuleID', 'Template'])\n",
    "tempDF2 = tempDF2.loc[:, tempDF2.columns.isin(sampleDF['Phenotype'].unique())]\n",
    "tempDF2 = tempDF2.reset_index()\n",
    "print(' -> After:', tempDF2.shape)\n",
    "\n",
    "#Clean to handle across data\n",
    "dataset = 'Proteomics'\n",
    "tempDF = sampleDF.loc[sampleDF['Dataset']==dataset]\n",
    "tempDF = tempDF.reset_index()[['Phenotype', 'Group']].drop_duplicates(keep='first')\n",
    "tempD = {}\n",
    "for row_i in range(len(tempDF)):\n",
    "    key = tempDF['Phenotype'].iloc[row_i]\n",
    "    value = tempDF['Group'].iloc[row_i]\n",
    "    tempD[key] = value\n",
    "tempDF1['Template'] = tempDF1['Template'].map(tempD)\n",
    "tempDF2['Template'] = tempDF2['Template'].map(tempD)\n",
    "tempDF2 = tempDF2.set_index(['ModuleID', 'Template'])\n",
    "tempDF2.columns = tempDF2.columns.map(tempD)\n",
    "tempDF2 = tempDF2.reset_index()\n",
    "\n",
    "display(tempDF1)\n",
    "display(tempDF2)\n",
    "\n",
    "#Save for using in the sub-notebook\n",
    "fileDir = './ExportData/'\n",
    "ipynbName = '220525_LC-M001-DIRAC-prot-vs-txn_Comparison-GOBP_ver2-2_'\n",
    "fileName = dataset.lower()+'-RMS.tsv'\n",
    "tempDF1.to_csv(fileDir+ipynbName+fileName, index=False, sep='\\t')\n",
    "fileName = dataset.lower()+'-RCI.tsv'\n",
    "tempDF2.to_csv(fileDir+ipynbName+fileName, index=False, sep='\\t')\n",
    "\n",
    "rmsDF_prot = tempDF1\n",
    "rciDF_prot = tempDF2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Import DIRAC results\n",
    "fileDir = '../201221_LC-M001-related-transcriptomics-DIRAC/ExportData/'\n",
    "ipynbName = '210429_LC-M001-related-transcriptomics-DIRAC_DIRAC-GOBP_ver2_'\n",
    "nSub = 25\n",
    "tempDF1 = pd.DataFrame()\n",
    "tempDF2 = pd.DataFrame()\n",
    "for list_i in range(nSub):\n",
    "    fileName = 'orgMmegdb-GOBP_min-n4-cov50_RankMatchingScore-BS-'+str(list_i+1).zfill(2)+'.tsv'\n",
    "    tempDF = pd.read_csv(fileDir+ipynbName+fileName, sep='\\t')\n",
    "    tempDF = tempDF.rename(columns={'NetworkID':'ModuleID'})\n",
    "    tempDF1 = pd.concat([tempDF1, tempDF], axis=0, ignore_index=True)\n",
    "    \n",
    "    fileName = 'orgMmegdb-GOBP_min-n4-cov50_RankConservationIndex-BS-'+str(list_i+1).zfill(2)+'.tsv'\n",
    "    tempDF = pd.read_csv(fileDir+ipynbName+fileName, sep='\\t')\n",
    "    tempDF = tempDF.rename(columns={'NetworkID':'ModuleID'})\n",
    "    tempDF2 = pd.concat([tempDF2, tempDF], axis=0, ignore_index=True)\n",
    "\n",
    "#Take only the modules and samples/groups to be compared\n",
    "print('Original RMS:', tempDF1.shape)\n",
    "tempDF1 = tempDF1.loc[tempDF1['ModuleID'].isin(moduleDF.index.tolist())]\n",
    "tempDF1 = tempDF1.loc[tempDF1['Template'].isin(sampleDF['Phenotype'].unique())]\n",
    "tempDF1 = tempDF1.set_index(['ModuleID', 'Template'])\n",
    "tempDF1 = tempDF1.loc[:, tempDF1.columns.isin(sampleDF.index.tolist())]\n",
    "tempDF1 = tempDF1.reset_index()\n",
    "print(' -> After:', tempDF1.shape)\n",
    "print('Original RCI:', tempDF2.shape)\n",
    "tempDF2 = tempDF2.loc[tempDF2['ModuleID'].isin(moduleDF.index.tolist())]\n",
    "tempDF2 = tempDF2.loc[tempDF2['Template'].isin(sampleDF['Phenotype'].unique())]\n",
    "tempDF2 = tempDF2.set_index(['ModuleID', 'Template'])\n",
    "tempDF2 = tempDF2.loc[:, tempDF2.columns.isin(sampleDF['Phenotype'].unique())]\n",
    "tempDF2 = tempDF2.reset_index()\n",
    "print(' -> After:', tempDF2.shape)\n",
    "\n",
    "#Clean to handle across data\n",
    "dataset = 'Transcriptomics'\n",
    "tempDF = sampleDF.loc[sampleDF['Dataset']==dataset]\n",
    "tempDF = tempDF.reset_index()[['Phenotype', 'Group']].drop_duplicates(keep='first')\n",
    "tempD = {}\n",
    "for row_i in range(len(tempDF)):\n",
    "    key = tempDF['Phenotype'].iloc[row_i]\n",
    "    value = tempDF['Group'].iloc[row_i]\n",
    "    tempD[key] = value\n",
    "tempDF1['Template'] = tempDF1['Template'].map(tempD)\n",
    "tempDF2['Template'] = tempDF2['Template'].map(tempD)\n",
    "tempDF2 = tempDF2.set_index(['ModuleID', 'Template'])\n",
    "tempDF2.columns = tempDF2.columns.map(tempD)\n",
    "tempDF2 = tempDF2.reset_index()\n",
    "\n",
    "display(tempDF1)\n",
    "display(tempDF2)\n",
    "\n",
    "#Save for using in the sub-notebook\n",
    "fileDir = './ExportData/'\n",
    "ipynbName = '220525_LC-M001-DIRAC-prot-vs-txn_Comparison-GOBP_ver2-2_'\n",
    "fileName = dataset.lower()+'-RMS.tsv'\n",
    "tempDF1.to_csv(fileDir+ipynbName+fileName, index=False, sep='\\t')\n",
    "fileName = dataset.lower()+'-RCI.tsv'\n",
    "tempDF2.to_csv(fileDir+ipynbName+fileName, index=False, sep='\\t')\n",
    "\n",
    "rmsDF_txn = tempDF1\n",
    "rciDF_txn = tempDF2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. Rank conservation index: general pattern"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> This would NOT be used for the manuscript; hence, statistical test is skipped in this version.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Extract RCI whose template phenotype corresponds to the own phenotype\n",
    "rciDF_kk = pd.DataFrame(index=moduleDF.index)\n",
    "for tempDF in [rciDF_prot, rciDF_txn]:\n",
    "    phenotypeL = tempDF.drop(columns=['ModuleID', 'Template']).columns.tolist()\n",
    "    tempDF = tempDF.set_index('ModuleID')\n",
    "    for k in phenotypeL:\n",
    "        tempS = tempDF[k].loc[tempDF['Template']==k]\n",
    "        rciDF_kk = pd.merge(rciDF_kk, tempS, left_index=True, right_index=True, how='left')\n",
    "\n",
    "#Order\n",
    "tempL1 = ['Control', 'Acarbose', 'Rapamycin']\n",
    "tempL2 = ['Proteomics', 'Transcriptomics']\n",
    "tempL = [dataset[0]+'-'+intervention for dataset in tempL2 for intervention in tempL1]\n",
    "rciDF_kk = rciDF_kk[tempL]\n",
    "display(rciDF_kk)\n",
    "display(rciDF_kk.describe())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Visualization (simple version for checking)\n",
    "tempDF = rciDF_kk.reset_index().melt(var_name='Group', value_name='RCI', id_vars='ModuleID')\n",
    "tempD = {'P-Control':'tab:blue', 'P-Acarbose':'tab:red', 'P-Rapamycin':'tab:purple',\n",
    "         'T-Control':plt.get_cmap('tab20')(1), 'T-Acarbose':plt.get_cmap('tab20')(7),\n",
    "         'T-Rapamycin':plt.get_cmap('tab20')(9)}\n",
    "sns.set(style='ticks', font='Arial', context='talk')\n",
    "plt.figure(figsize=(3, 4))\n",
    "p = sns.boxplot(data=tempDF, y='RCI', x='Group', order=list(tempD.keys()), palette=tempD, dodge=False,\n",
    "                showfliers=True, flierprops={'marker':'o', 'markerfacecolor':'gray', 'alpha':0.4},\n",
    "                showcaps=True, notch=True)\n",
    "p.set(ylim=(0.49, 1.01), yticks=np.arange(0.5, 1.01, 0.1))\n",
    "sns.despine()\n",
    "plt.xlabel('')\n",
    "plt.ylabel('Module RCI')\n",
    "plt.xticks(rotation=70, horizontalalignment='right', verticalalignment='center', rotation_mode='anchor')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3. Rank conservation index: inter-group module comparison"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> Test specific hypothesis: control RCI == intervention RCI (i.e., inter-group module comparison).  \n",
    "> 1. Testing the main effect of intervention on RMSs for each module using ANOVA model  \n",
    "> 2. Then, performing post-hoc comparisons of RMSs between control vs. each intervention using the repeated Student's t-tests  \n",
    ">  \n",
    "> Basically, statistical strategy is same with the one used in each dataset analysis. Because RMS/RCI was not normalized (i.e., the expected mean and variance could be different between datasets due to different number of mapped analytes), dataset and its interaction term are NOT included in ANOVA model; instead, ANOVA model is generated per dataset. The p-value adjustment is performed in a conservative manner: the P-values in ANOVA tests are adjusted across all models (= modules x datasets), and those in post-hoc tests are adjusted across datasets only within the module (not across modules). Not Dunnett's test but Student's t-test (i.e., t-test with pooled variance) is used as the post-hoc test because further adjustment of the Dunnett's test p-values with the Holm-Bonferroni method is too much (incorrect) adjustment for family-wise error rate (FWER).  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3-1. Extract RMS under the own phenotype consensus"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Extract RMS whose template phenotype corresponds to the own phenotype\n",
    "rmsDF_kk = pd.DataFrame(index=moduleDF.index)\n",
    "for tempDF1, tempDF2 in [[rmsDF_prot, rciDF_prot], [rmsDF_txn, rciDF_txn]]:\n",
    "    phenotypeL = tempDF2.drop(columns=['ModuleID', 'Template']).columns.tolist()\n",
    "    tempDF1 = tempDF1.set_index('ModuleID')\n",
    "    for k in phenotypeL:\n",
    "        tempL = sampleDF.loc[sampleDF['Group']==k].index.tolist()\n",
    "        tempDF = tempDF1[tempL].loc[tempDF1['Template']==k]\n",
    "        rmsDF_kk = pd.merge(rmsDF_kk, tempDF, left_index=True, right_index=True, how='left')\n",
    "\n",
    "display(rmsDF_kk)\n",
    "display(rmsDF_kk.describe())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# — Go to the R sub-notebook —  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3-2. ANOVA test (RMS ~ Intervention), followed by repeated Student's t-tests (Intervention)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 3-2-1/2/3. Import the summary table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Import the summary table\n",
    "fileDir = './ExportData/'\n",
    "ipynbName = '220525_LC-M001-DIRAC-prot-vs-txn_StatisticalTest-GOBP_ver2-2_'\n",
    "fileName = 'inter-group-comparison.xlsx'\n",
    "tempDF = pd.DataFrame()\n",
    "for dataset in ['Proteomics', 'Transcriptomics']:\n",
    "    sheetName = 'RCI-'+dataset\n",
    "    tempDF1 = pd.read_excel(fileDir+ipynbName+fileName, sheet_name=sheetName, engine='openpyxl')\n",
    "    tempDF1['Dataset'] = dataset\n",
    "    tempDF1 = tempDF1.set_index(['Dataset', 'ModuleID'])\n",
    "    tempDF = pd.concat([tempDF, tempDF1], axis=0)\n",
    "tempDF = tempDF.sort_values(by='Intervention_Pval', ascending=True)\n",
    "display(tempDF)\n",
    "\n",
    "statDF = tempDF"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 3-2-4. Changed modules (ANOVA)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Prepare variables in the model\n",
    "tempS = statDF.loc[:, statDF.columns.str.contains('_Fstat')].columns.to_series()\n",
    "variableL = tempS.str.replace('_Fstat', '').tolist()\n",
    "\n",
    "#Changed modules\n",
    "for variable in variableL:\n",
    "    tempDF = statDF.loc[statDF[variable+'_AdjPval']<0.05]\n",
    "    tempDF = tempDF.sort_values(by=variable+'_AdjPval', ascending=True)\n",
    "    tempL1 = tempDF.loc[:, tempDF.columns.str.contains('_RMSmean')].columns.tolist()\n",
    "    tempL2 = tempDF.loc[:, tempDF.columns.str.contains('^'+variable+'_')].columns.tolist()\n",
    "    tempDF = tempDF[[col_n for subL in [['ModuleName'], tempL1, tempL2] for col_n in subL]]\n",
    "    print(variable+' (adjusted P < 0.05):', len(tempDF))\n",
    "    tempL = tempDF.index.to_frame()['ModuleID'].unique().tolist()\n",
    "    print(' -> Unique module:', len(tempL))\n",
    "    tempL = tempDF.index.to_frame()['Dataset'].unique().tolist()\n",
    "    for dataset in tempL:\n",
    "        tempDF1 = tempDF.loc[dataset]\n",
    "        print(' -> '+dataset+':', len(tempDF1))\n",
    "    display(tempDF)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 3-2-5. Changed modules by each intervention (Student's t-tests)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Spread statDF by dataset to flatten multi-index\n",
    "tempDF = moduleDF['ModuleName']#pd.Series() for now\n",
    "for dataset in ['Proteomics', 'Transcriptomics']:\n",
    "    tempDF1 = statDF.loc[dataset].drop(columns=['ModuleName'])\n",
    "    tempDF1.columns = dataset[0:4]+'_'+tempDF1.columns\n",
    "    tempDF = pd.merge(tempDF, tempDF1, left_index=True, right_index=True, how='left')\n",
    "\n",
    "#Extract only the changed modules\n",
    "variable = 'Intervention'\n",
    "tempS1 = tempDF['Prot_'+variable+'_AdjPval']<0.05\n",
    "tempS2 = tempDF['Tran_'+variable+'_AdjPval']<0.05\n",
    "tempDF = tempDF.loc[tempS1|tempS2]\n",
    "tempDF = tempDF.sort_values(by=['Prot_'+variable+'_AdjPval', 'Tran_'+variable+'_AdjPval'], ascending=True)\n",
    "print(variable+' (adjusted P < 0.05):', len(tempDF))\n",
    "print(' -> Proteomics:', sum(tempS1))\n",
    "print(' -> Transcriptomics:', sum(tempS2))\n",
    "\n",
    "#Take adjusted P-value\n",
    "tempDF1 = tempDF.loc[:, tempDF.columns.str.contains('-vs-.*_AdjPval$')]\n",
    "tempDF1.columns = tempDF1.columns.str.replace('_AdjPval$', '')\n",
    "tempDF1 = pd.merge(tempDF[['ModuleName', 'Prot_'+variable+'_AdjPval', 'Tran_'+variable+'_AdjPval']], tempDF1,\n",
    "                   left_index=True, right_index=True, how='left')\n",
    "print('Adjusted P-value:')\n",
    "display(tempDF1)\n",
    "display(tempDF1.describe())\n",
    "\n",
    "#Take effect size\n",
    "tempDF2 = tempDF.loc[:, tempDF.columns.str.contains('-vs-.*_Coef$')]\n",
    "tempDF2.columns = tempDF2.columns.str.replace('_Coef$', '')\n",
    "tempDF2 = pd.merge(tempDF[['ModuleName', 'Prot_'+variable+'_AdjPval', 'Tran_'+variable+'_AdjPval']], tempDF2,\n",
    "                   left_index=True, right_index=True, how='left')\n",
    "print('Changed direction (effect size):')\n",
    "display(tempDF2)\n",
    "display(tempDF2.describe())\n",
    "\n",
    "pvalDF = tempDF1\n",
    "diffDF = tempDF2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> Check the changed modules (based on the nominal P-value for the main effect and the adjusted P-values for the post-hoc tests) as reference.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Spread statDF by dataset to flatten multi-index\n",
    "tempDF = moduleDF['ModuleName']#pd.Series() for now\n",
    "for dataset in ['Proteomics', 'Transcriptomics']:\n",
    "    tempDF1 = statDF.loc[dataset].drop(columns=['ModuleName'])\n",
    "    tempDF1.columns = dataset[0:4]+'_'+tempDF1.columns\n",
    "    tempDF = pd.merge(tempDF, tempDF1, left_index=True, right_index=True, how='left')\n",
    "\n",
    "#Extract only the changed modules\n",
    "variable = 'Intervention'\n",
    "tempS1 = tempDF['Prot_'+variable+'_Pval']<0.05\n",
    "tempS2 = tempDF['Tran_'+variable+'_Pval']<0.05\n",
    "tempDF = tempDF.loc[tempS1|tempS2]\n",
    "tempDF = tempDF.sort_values(by=['Prot_'+variable+'_Pval', 'Tran_'+variable+'_Pval'], ascending=True)\n",
    "print(variable+' (nominal P < 0.05):', len(tempDF))\n",
    "print(' -> Proteomics:', sum(tempS1))\n",
    "print(' -> Transcriptomics:', sum(tempS2))\n",
    "\n",
    "#Take adjusted P-value\n",
    "tempDF1 = tempDF.loc[:, tempDF.columns.str.contains('-vs-.*_AdjPval$')]\n",
    "tempDF1.columns = tempDF1.columns.str.replace('_AdjPval$', '')\n",
    "tempDF1 = pd.merge(tempDF[['ModuleName', 'Prot_'+variable+'_Pval', 'Tran_'+variable+'_Pval']], tempDF1,\n",
    "                   left_index=True, right_index=True, how='left')\n",
    "print('Adjusted P-value:')\n",
    "display(tempDF1)\n",
    "display(tempDF1.describe())\n",
    "\n",
    "#Take effect size\n",
    "tempDF2 = tempDF.loc[:, tempDF.columns.str.contains('-vs-.*_Coef$')]\n",
    "tempDF2.columns = tempDF2.columns.str.replace('_Coef$', '')\n",
    "tempDF2 = pd.merge(tempDF[['ModuleName', 'Prot_'+variable+'_Pval', 'Tran_'+variable+'_Pval']], tempDF2,\n",
    "                   left_index=True, right_index=True, how='left')\n",
    "#print('Changed direction (effect size):')\n",
    "#display(tempDF2)\n",
    "#display(tempDF2.describe())\n",
    "\n",
    "pvalDF_ref = tempDF1\n",
    "diffDF_ref = tempDF2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "### 3-3. Visualization: clustermap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Prepare color labels for tightened module set\n",
    "regulation = 'Tightened'\n",
    "tempD0 = {'Cont':'Control', 'Acar':'Acarbose', 'Rapa':'Rapamycin'}\n",
    "tempD1 = {'Control':'tab:blue', 'Acarbose':'tab:red', 'Rapamycin':'tab:purple'}\n",
    "tempD2 = {'Control':plt.get_cmap('tab20')(1), 'Acarbose':plt.get_cmap('tab20')(7),\n",
    "          'Rapamycin':plt.get_cmap('tab20')(9)}\n",
    "tempDF = pvalDF_ref.loc[:, pvalDF_ref.columns.str.contains('-vs-')]#Based on nominal ANOVA P-values\n",
    "for col_n in tempDF.columns.tolist():\n",
    "    tempS1 = pvalDF_ref[col_n]#Adjusted P-values in post-hoc tests\n",
    "    tempS2 = diffDF_ref[col_n]\n",
    "    tempS3 = pvalDF[col_n]#Adjusted P-values in post-hoc tests\n",
    "    tempS4 = diffDF[col_n]\n",
    "    if regulation=='Changed':\n",
    "        tempS2 = tempS2.loc[(tempS1<0.05)]\n",
    "        tempS4 = tempS4.loc[(tempS3<0.05)]\n",
    "    elif regulation=='Tightened':\n",
    "        tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2>0)]\n",
    "        tempS4 = tempS4.loc[(tempS3<0.05)&(tempS4>0)]\n",
    "    elif regulation=='Loosened':\n",
    "        tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2<0)]\n",
    "        tempS4 = tempS4.loc[(tempS3<0.05)&(tempS4<0)]\n",
    "    label = re.sub('-vs-.*', '', col_n)\n",
    "    group = tempD0[re.sub('.*_', '', label)]\n",
    "    dataset = re.sub('_.*', '', label)\n",
    "    tempL = []\n",
    "    for module in tempDF.index.tolist():\n",
    "        if module in tempS2.index.tolist():\n",
    "            if module in tempS4.index.tolist():#Based on adjusted ANOVA P-values\n",
    "                tempL.append(tempD1[group])\n",
    "            else:#Based on nominal ANOVA P-values\n",
    "                tempL.append(tempD2[group])\n",
    "        else:\n",
    "            tempL.append('white')\n",
    "    tempDF[dataset[0]+'-'+group] = tempL\n",
    "    print(regulation+' module in '+col_n)\n",
    "    print(' -> in adjusted ANOVA P < 0.05:', len(tempS4))\n",
    "    print(' -> in nominal ANOVA P < 0.05:', len(tempS2))\n",
    "tempDF = tempDF[['P-Acarbose', 'P-Rapamycin', 'T-Acarbose', 'T-Rapamycin']]\n",
    "\n",
    "#Prepare color labels for samples\n",
    "tempL = [dataset[0]+'-'+group for dataset in ['Proteomics', 'Transcriptomics'] for group in list(tempD1.keys())]\n",
    "tempA = np.repeat(['tab:pink', 'tab:cyan'], 3)\n",
    "tempDF1 = pd.DataFrame({'Data':tempA, 'Group':np.tile(list(tempD1.values()), 2)}, index=tempL)\n",
    "\n",
    "#Clustermap\n",
    "sns.set(style='ticks', font='Arial', context='talk')\n",
    "cm = sns.clustermap(rciDF_kk.T, method='ward', metric='euclidean', cmap='afmhot',\n",
    "                    row_cluster=True, col_cluster=True, row_linkage=None, col_linkage=None,\n",
    "                    row_colors=tempDF1, col_colors=tempDF, xticklabels=False, yticklabels=True,\n",
    "                    dendrogram_ratio=(0.025, 0.2), colors_ratio=(0.025, 0.075),\n",
    "                    cbar_pos=(0.05, -0.05, 0.3, 0.075), cbar_kws={'orientation': 'horizontal'},\n",
    "                    figsize=(12, 4), **{'vmin':0.5, 'vmax':1})\n",
    "cm.cax.set_title('Module RCI', size='medium',\n",
    "                 verticalalignment='bottom', horizontalalignment='center')\n",
    "cm.cax.tick_params(labelsize='small')\n",
    "bottom, top = cm.ax_heatmap.get_ylim()\n",
    "#cm.ax_heatmap.set_ylim(bottom + 0.5, top - 0.5)##To avoid half cut of first and last rows\n",
    "hm = cm.ax_heatmap.get_position()\n",
    "rd = cm.ax_row_dendrogram.get_position()\n",
    "cd = cm.ax_col_dendrogram.get_position()\n",
    "cm.ax_heatmap.set_position([hm.x0, hm.y0, hm.width, hm.height])\n",
    "cm.ax_row_dendrogram.set_position([rd.x0, rd.y0, rd.width, rd.height])\n",
    "cm.ax_col_dendrogram.set_position([cd.x0, cd.y0, cd.width, cd.height])\n",
    "cm.ax_heatmap.set_xlabel('GOBP module')\n",
    "cm.ax_heatmap.set_ylabel('')\n",
    "##row/column color bar legend (axis is same with cm.cax!)\n",
    "tempL = []\n",
    "for group in tempD1.keys():\n",
    "    if group!='Control':\n",
    "        tempL.append(mpatches.Patch(color=tempD1[group],\n",
    "                                    label='by '+group+' (adjusted '+r'$P$'+' < 0.05)'))\n",
    "for group in tempD2.keys():\n",
    "    if group!='Control':\n",
    "        tempL.append(mpatches.Patch(color=tempD2[group], label='(nominal '+r'$P$'+' < 0.05)'))\n",
    "legend1 = plt.legend(handles=tempL, fontsize='small', labelspacing=0.2, ncol=2,\n",
    "                     title=regulation+' module (vs. Control)', title_fontsize='medium',\n",
    "                     bbox_to_anchor=(1, 0.5), loc='center left', borderaxespad=3.5, frameon=False)\n",
    "plt.gca().add_artist(legend1)\n",
    "##Save\n",
    "fileDir = './ExportFigures/'\n",
    "ipynbName = '220525_LC-M001-DIRAC-prot-vs-txn_Comparison-GOBP_ver2-2_'\n",
    "fileName = 'RCI-clustermap.tif'\n",
    "plt.gcf().savefig(fileDir+ipynbName+fileName, dpi=300, bbox_inches='tight', pad_inches=0.04,\n",
    "                  pil_kwargs={'compression':'tiff_lzw'})\n",
    "plt.show()\n",
    "\n",
    "#Save label order\n",
    "tempDF = moduleDF.loc[rciDF_kk.index[cm.dendrogram_col.reordered_ind]]\n",
    "tempD = {'P-Control':'Prot-Cont', 'P-Acarbose':'Prot-Acar', 'P-Rapamycin':'Prot-Rapa',\n",
    "         'T-Control':'Tran-Cont', 'T-Acarbose':'Tran-Acar', 'T-Rapamycin':'Tran-Rapa'}\n",
    "tempDF1 = rciDF_kk.copy()\n",
    "tempDF1.columns = tempDF1.columns.map(tempD)\n",
    "tempDF = pd.merge(tempDF['ModuleName'], tempDF1, left_index=True, right_index=True, how='left')\n",
    "tempDF = tempDF.reset_index()\n",
    "tempDF.index.name = 'Xcoord'\n",
    "display(tempDF)\n",
    "fileDir = './ExportData/'\n",
    "ipynbName = '220525_LC-M001-DIRAC-prot-vs-txn_Comparison-GOBP_ver2-2_'\n",
    "fileName = 'RCI-clustermap-xticks.tsv'\n",
    "tempDF.to_csv(fileDir+ipynbName+fileName, index=True, sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3-4. Visualization: venn diagram"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Prepare label and color\n",
    "tempD0 = {'Cont':'Control', 'Acar':'Acarbose', 'Rapa':'Rapamycin'}\n",
    "tempD1 = {'T-Acarbose':plt.get_cmap('tab20')(7), 'P-Acarbose':'tab:red',\n",
    "          'P-Rapamycin':'tab:purple', 'T-Rapamycin':plt.get_cmap('tab20')(9)}\n",
    "\n",
    "#Visualization per direction\n",
    "for regulation in ['Changed', 'Tightened', 'Loosened']:\n",
    "    #Prepare module sets\n",
    "    tempD2 = {}\n",
    "    tempDF = pvalDF.loc[:, pvalDF.columns.str.contains('-vs-')]#Based on adjusted ANOVA P-values\n",
    "    for col_n in tempDF.columns.tolist():\n",
    "        tempS1 = pvalDF[col_n]\n",
    "        tempS2 = diffDF[col_n]\n",
    "        if regulation=='Changed':\n",
    "            tempS2 = tempS2.loc[(tempS1<0.05)]\n",
    "        elif regulation=='Tightened':\n",
    "            tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2>0)]\n",
    "        elif regulation=='Loosened':\n",
    "            tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2<0)]\n",
    "        label = re.sub('-vs-.*', '', col_n)\n",
    "        group = tempD0[re.sub('.*_', '', label)]\n",
    "        dataset = re.sub('_.*', '', label)\n",
    "        tempD2[dataset[0]+'-'+group] = set(tempS2.index.tolist())\n",
    "    ##Sort to make consistent order in manual legend generation\n",
    "    tempD = {}\n",
    "    for label in tempD1.keys():\n",
    "        tempD[label] = tempD2[label]\n",
    "    \n",
    "    #Not significant in all contrasts\n",
    "    tempDF = pvalDF.copy()#Based on adjusted ANOVA P-values\n",
    "    for moduleS in tempD.values():\n",
    "        tempDF = tempDF.loc[~tempDF.index.isin(moduleS)]\n",
    "    count = len(tempDF)\n",
    "    print(regulation+' modules (vs. Control):')\n",
    "    print(' -> Not significant in all contrasts:', count)\n",
    "    \n",
    "    #Skip the followings if no significant module\n",
    "    if count==len(pvalDF):\n",
    "        continue\n",
    "    \n",
    "    #Venn diagram\n",
    "    sns.set(style='ticks', font='Arial', context='talk')\n",
    "    fig, ax = plt.subplots(figsize=(4, 4))\n",
    "    venn(tempD, fmt='{size:,}', cmap=list(tempD1.values()), legend_loc=None, ax=ax)\n",
    "    plt.setp(ax, ylim=(0.1, 0.875))#Otherwise, weird space...\n",
    "    ##Add legend annotation\n",
    "    x_coord = [0.1, 0.1, 0.9, 0.9]\n",
    "    y_coord = [0.25, 0.7, 0.7, 0.25]\n",
    "    h_align = ['right', 'right', 'left', 'left']\n",
    "    v_align = ['top', 'bottom', 'bottom', 'top']\n",
    "    for i in range(len(tempD1)):\n",
    "        key = list(tempD1.keys())[i]\n",
    "        total = f'{len(tempD[key]):,}'\n",
    "        ax.text(x_coord[i], y_coord[i], key+'\\n('+total+' modules)',\n",
    "                fontsize='small', multialignment='center',\n",
    "                horizontalalignment=h_align[i], verticalalignment=v_align[i],\n",
    "                bbox={'boxstyle':'round', 'facecolor':tempD1[key], 'pad':0.2, 'alpha':0.5})\n",
    "    title = regulation+' modules (vs. Control)'\n",
    "    ax.set_title(title, fontsize='medium')\n",
    "    ##Save\n",
    "    if regulation!='Changed':\n",
    "        fileDir = './ExportFigures/'\n",
    "        ipynbName = '220525_LC-M001-DIRAC-prot-vs-txn_Comparison-GOBP_ver2-2_'\n",
    "        fileName = 'RCI-inter-group-comparison_venn-'+regulation.lower()+'.tif'\n",
    "        plt.gcf().savefig(fileDir+ipynbName+fileName, dpi=300, bbox_inches='tight', pad_inches=0.04,\n",
    "                          pil_kwargs={'compression':'tiff_lzw'})\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Export module list in each subset in the venn diagram\n",
    "for regulation in ['Tightened', 'Loosened']:\n",
    "    #Prepare module sets\n",
    "    tempD = {}\n",
    "    tempDF = pvalDF.loc[:, pvalDF.columns.str.contains('-vs-')]#Based on adjusted ANOVA P-values\n",
    "    for col_n in tempDF.columns.tolist():\n",
    "        tempS1 = pvalDF[col_n]\n",
    "        tempS2 = diffDF[col_n]\n",
    "        if regulation=='Changed':\n",
    "            tempS2 = tempS2.loc[(tempS1<0.05)]\n",
    "        elif regulation=='Tightened':\n",
    "            tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2>0)]\n",
    "        elif regulation=='Loosened':\n",
    "            tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2<0)]\n",
    "        tempD[col_n] = set(tempS2.index.tolist())\n",
    "    \n",
    "    #Not significant in all contrasts\n",
    "    tempDF = pvalDF.copy()#Based on adjusted ANOVA P-values\n",
    "    for moduleS in tempD.values():\n",
    "        tempDF = tempDF.loc[~tempDF.index.isin(moduleS)]\n",
    "    count = len(tempDF)\n",
    "    print(regulation+' modules (vs. Control):')\n",
    "    print(' -> Not significant in all contrasts:', count)\n",
    "    \n",
    "    #Skip the followings if no significant module\n",
    "    if count==len(pvalDF):\n",
    "        continue\n",
    "    \n",
    "    #Prepare a new .xlsx file (dummy README)\n",
    "    tempL1 = [len(tempD[key]) for key in tempD.keys()]\n",
    "    tempDF = pd.DataFrame({'Group':tempD.keys(), 'nModules':tempL1})\n",
    "    tempDF = tempDF.reset_index().rename(columns={'index':'VennOrder'})\n",
    "    tempDF['VennOrder'] = tempDF['VennOrder'] + 1\n",
    "    fileDir = './ExportData/'\n",
    "    ipynbName = '220525_LC-M001-DIRAC-prot-vs-txn_Comparison-GOBP_ver2-2_'\n",
    "    fileName = 'RCI-inter-group-comparison_venn-'+regulation.lower()+'.xlsx'\n",
    "    tempDF.to_excel(fileDir+ipynbName+fileName, sheet_name='README', header=True, index=False)\n",
    "    display(tempDF)#Check\n",
    "    \n",
    "    #Prepare saving data (statDF spread by dataset)\n",
    "    tempDF = moduleDF['ModuleName']#pd.Series() for now\n",
    "    for dataset in ['Proteomics', 'Transcriptomics']:\n",
    "        tempDF1 = statDF.loc[dataset].drop(columns=['ModuleName'])\n",
    "        tempDF1.columns = dataset[0:4]+'_'+tempDF1.columns\n",
    "        tempDF = pd.merge(tempDF, tempDF1, left_index=True, right_index=True, how='left')\n",
    "    \n",
    "    t_start = time.time()\n",
    "    #Extract overall set\n",
    "    for key_i in range(len(tempD)):\n",
    "        key = list(tempD.keys())[key_i]\n",
    "        tempS = tempD[key]\n",
    "        tempDF1 = tempDF.loc[tempDF.index.isin(tempS)].sort_index(ascending=True)\n",
    "        #Save summary table by appended to the above .xlsx file\n",
    "        ##Prepare sheet name\n",
    "        tempL1 = ['NA' for i in range(len(tempD))]\n",
    "        tempL1[key_i] = '1'\n",
    "        setName = '('+','.join(tempL1)+')'\n",
    "        with pd.ExcelWriter(fileDir+ipynbName+fileName, mode='a', engine='openpyxl') as writer:\n",
    "            tempDF1.to_excel(writer, sheet_name=setName, header=True, index=True)\n",
    "        print(' - '+setName+':', len(tempDF1))\n",
    "    \n",
    "    #Extract subset\n",
    "    tempL1 = ['1', '0']\n",
    "    tempL2 = [[k1, k2, k3, k4] for k1 in tempL1 for k2 in tempL1 for k3 in tempL1 for k4 in tempL1]\n",
    "    #tempL2.remove(['0', '0', '0', '0'])\n",
    "    for tempL1 in tempL2:\n",
    "        #Positive module set\n",
    "        tempL3 = [list(tempD.values())[key_i] for key_i, binary in enumerate(tempL1) if binary=='1']\n",
    "        tempS1 = set(pvalDF.index.tolist())#Initialize\n",
    "        for tempS in tempL3:\n",
    "            tempS1 = tempS1 & tempS\n",
    "        #Negative module set\n",
    "        tempL3 = [list(tempD.values())[key_i] for key_i, binary in enumerate(tempL1) if binary=='0']\n",
    "        tempS2 = set()#Initialize\n",
    "        for tempS in tempL3:\n",
    "            tempS2 = tempS2 | tempS\n",
    "        #Extract subset\n",
    "        tempS = tempS1 - tempS2\n",
    "        tempDF1 = tempDF.loc[tempDF.index.isin(tempS)].sort_index(ascending=True)\n",
    "        #Save summary table by appended to the above .xlsx file\n",
    "        ##Prepare sheet name\n",
    "        setName = '('+','.join(tempL1)+')'\n",
    "        with pd.ExcelWriter(fileDir+ipynbName+fileName, mode='a', engine='openpyxl') as writer:\n",
    "            tempDF1.to_excel(writer, sheet_name=setName, header=True, index=True)\n",
    "        print(' - '+setName+':', len(tempDF1))\n",
    "    \n",
    "    t_elapsed = time.time() - t_start\n",
    "    print(' - Elapsed time:', round(t_elapsed//60), 'min', round(t_elapsed%60, 1), 'sec')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3-5. Visualization: pointplot"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 3-5-1. Modules tightened by all interventions and datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Prepare the target module set\n",
    "posL = ['Prot_Acar-vs-Cont', 'Prot_Rapa-vs-Cont', 'Tran_Acar-vs-Cont', 'Tran_Rapa-vs-Cont']\n",
    "negL = ['']\n",
    "regulation = 'Tightened'\n",
    "tempS = pd.Series(np.repeat(True, len(pvalDF)), index=pvalDF.index)#Initialize\n",
    "tempDF = pvalDF.loc[:, pvalDF.columns.str.contains('-vs-')]#Based on adjusted ANOVA P-values\n",
    "for col_n in tempDF.columns.tolist():\n",
    "    tempS1 = pvalDF[col_n]\n",
    "    tempS2 = diffDF[col_n]\n",
    "    if col_n in posL:\n",
    "        if regulation=='Tightened':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2>0)\n",
    "        elif regulation=='Loosened':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2<0)\n",
    "    elif col_n in negL:\n",
    "        tempS3 = (tempS1>=0.05)\n",
    "        #Significance for inverse regulation\n",
    "        if regulation=='Tightened':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2<0)\n",
    "        elif regulation=='Loosened':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2>0)\n",
    "        tempS1 = tempS3 | tempS1\n",
    "    else:\n",
    "        tempS1 = (tempS1>=0.0)\n",
    "    #Update True\n",
    "    tempS = tempS & tempS1\n",
    "tempL = tempS.loc[tempS.tolist()].index.tolist()\n",
    "print(len(tempL), regulation.lower()+' modules with significance in', posL, 'but not in', negL)\n",
    "\n",
    "#Select representatives\n",
    "topX = np.min([30, len(tempL)])\n",
    "topX_plot = np.min([10, len(tempL)])\n",
    "##Spread statDF by dataset to flatten multi-index\n",
    "tempDF = moduleDF['ModuleName']#pd.Series() for now\n",
    "for dataset in ['Proteomics', 'Transcriptomics']:\n",
    "    tempDF1 = statDF.loc[dataset].drop(columns=['ModuleName'])\n",
    "    tempDF1.columns = dataset[0:4]+'_'+tempDF1.columns\n",
    "    tempDF = pd.merge(tempDF, tempDF1, left_index=True, right_index=True, how='left')\n",
    "tempDF = tempDF.loc[:, tempDF.columns.str.contains('Pval')]\n",
    "tempDF = pd.merge(moduleDF['ModuleName'], tempDF, left_index=True, right_index=True, how='left')\n",
    "tempDF = tempDF.loc[tempL]\n",
    "tempDF = tempDF.sort_values(by=['Prot_'+variable+'_AdjPval', 'Tran_'+variable+'_AdjPval'], ascending=True)\n",
    "print('Top', topX, 'modules (sort by the main effect of '+variable+'):')\n",
    "display(tempDF.iloc[:topX])\n",
    "plotL = tempDF.index.tolist()[:topX_plot]\n",
    "\n",
    "#Prepare DF for plot\n",
    "tempDF = rmsDF_kk.reset_index().melt(var_name='SampleID', value_name='RMS', id_vars='ModuleID')\n",
    "tempDF1 = sampleDF.reset_index()[['SampleID', 'Dataset', 'Phenotype']]\n",
    "tempDF = pd.merge(tempDF, tempDF1, on='SampleID', how='left')\n",
    "\n",
    "#Prepare label and color\n",
    "tempD = {'Proteomics':'Proteins', 'Transcriptomics':'Transcripts'}\n",
    "tempDF['Dataset'] = tempDF['Dataset'].map(tempD)\n",
    "tempDF['Group'] = tempDF['Phenotype']\n",
    "tempD0 = {'Cont':'Control', 'Acar':'Acarbose', 'Rapa':'Rapamycin', 'Prot':'Proteins', 'Tran':'Transcripts'}\n",
    "tempD1 = {'Control':'tab:blue', 'Acarbose':'tab:red', 'Rapamycin':'tab:purple'}\n",
    "tempD2 = {'Proteins':plt.get_cmap('tab20')(13), 'Transcripts':plt.get_cmap('tab20')(19)}\n",
    "\n",
    "#Visualize each representative\n",
    "for rank_i in range(len(plotL)):\n",
    "    print(' - Rank '+str(rank_i+1)+' (sort by the main effect of '+variable+'):')\n",
    "    module = plotL[rank_i]\n",
    "    #Check module summary\n",
    "    tempDF1 = pd.DataFrame(moduleDF.loc[module]).T\n",
    "    display(tempDF1)\n",
    "    \n",
    "    #Select RMS\n",
    "    tempDF1 = tempDF.loc[tempDF['ModuleID']==module]\n",
    "    \n",
    "    #Check RMS summary\n",
    "    tempDF2 = tempDF1.groupby(['Dataset', 'Group'])['RMS'].agg(['count', 'mean', 'std'])\n",
    "    tempL1 = []\n",
    "    tempL2 = []\n",
    "    for row_n in tempDF2.index.tolist():\n",
    "        count, mean, std = tempDF2.loc[row_n]\n",
    "        tempL1.append(mean - 1.96*std/np.sqrt(count))\n",
    "        tempL2.append(mean + 1.96*std/np.sqrt(count))\n",
    "    tempDF2['0.025'] = tempL1\n",
    "    tempDF2['0.975'] = tempL2\n",
    "    ##Multiindex sort\n",
    "    tempDF2 = tempDF2.reset_index()\n",
    "    tempDF2['Dataset'] = pd.Categorical(tempDF2['Dataset'], categories=list(tempD2.keys()))\n",
    "    tempDF2['Group'] = pd.Categorical(tempDF2['Group'], categories=list(tempD1.keys()))\n",
    "    tempDF2 = tempDF2.sort_values(by=['Dataset', 'Group']).set_index(['Dataset', 'Group'])\n",
    "    display(tempDF2)\n",
    "    \n",
    "    #Prepare significance labels\n",
    "    ##Retrieve statistical significance\n",
    "    tempS = pvalDF.loc[module, pvalDF.columns.str.contains('-vs-')]\n",
    "    tempS.name = 'AdjPval'\n",
    "    ##Clean\n",
    "    tempDF2 = tempS.index.to_series().str.split(pat='_', expand=True)\n",
    "    tempDF2 = tempDF2.rename(columns={0:'Dataset', 1:'Comparison'})\n",
    "    tempS1 = tempDF2['Dataset']\n",
    "    tempDF2 = tempDF2['Comparison'].str.split(pat='-vs-', expand=True)\n",
    "    tempDF2 = tempDF2.rename(columns={0:'Contrast', 1:'Baseline'})\n",
    "    tempDF2 = pd.merge(tempS1, tempDF2, left_index=True, right_index=True, how='left')\n",
    "    tempDF2 = pd.merge(tempDF2, tempS, left_index=True, right_index=True, how='left')\n",
    "    tempDF2['Dataset'] = tempDF2['Dataset'].map(tempD0)\n",
    "    tempDF2['Contrast'] = tempDF2['Contrast'].map(tempD0)\n",
    "    tempDF2['Baseline'] = tempDF2['Baseline'].map(tempD0)\n",
    "    ##Convert p-value to label\n",
    "    tempL = []\n",
    "    for row_i in range(len(tempDF2)):\n",
    "        pval = tempDF2['AdjPval'].iloc[row_i]\n",
    "        if pval<0.001:\n",
    "            tempL.append('***')\n",
    "        elif pval<0.01:\n",
    "            tempL.append('**')\n",
    "        elif pval<0.05:\n",
    "            tempL.append('*')\n",
    "        else:\n",
    "            pval_text = Decimal(str(pval)).quantize(Decimal('0.01'), rounding=ROUND_HALF_UP)\n",
    "            tempL.append(r'$P$ = '+str(pval_text))\n",
    "    tempDF2['SignifLabel'] = tempL\n",
    "    display(tempDF2)\n",
    "    \n",
    "    #Visualization\n",
    "    ymax = 1.0\n",
    "    ymin = 0.0\n",
    "    yinter = 0.2\n",
    "    ymargin_t = 0.275\n",
    "    ymargin_b = 0.05\n",
    "    aline_ymin = 1.0\n",
    "    aline_ymargin = 0.125\n",
    "    sns.set(style='ticks', font='Arial', context='talk')\n",
    "    fig, axes = plt.subplots(nrows=1, ncols=2, figsize=(4.75, 5), sharex=True, sharey=True,\n",
    "                             gridspec_kw={'width_ratios':[2, 2]})\n",
    "    for ax_i, ax in enumerate(axes.flat):\n",
    "        dataset = list(tempD2.keys())[ax_i]\n",
    "        tempDF3 = tempDF1.loc[tempDF1['Dataset']==dataset]\n",
    "        sns.pointplot(data=tempDF3, x='Group', y='RMS', order=list(tempD1.keys()), palette=tempD1,\n",
    "                      markers='o', dodge=False, join=False, capsize=0.6, estimator=np.mean, ci=95, ax=ax)\n",
    "        sns.stripplot(data=tempDF3, x='Group', y='RMS',\n",
    "                      order=list(tempD1.keys()), palette=tempD1, dodge=False, jitter=0.15,\n",
    "                      size=5, edgecolor='black', linewidth=1, **{'marker':'o', 'alpha':0.5}, ax=ax)\n",
    "        #Set axis\n",
    "        sns.despine()\n",
    "        ax.set(ylim=(ymin-ymargin_b, ymax+ymargin_t), yticks=np.arange(ymin, ymax + yinter/10, yinter))\n",
    "        plt.setp(ax.get_xticklabels(), rotation=70, horizontalalignment='right',\n",
    "                 verticalalignment='center', rotation_mode='anchor')\n",
    "        if ax_i==0:\n",
    "            plt.setp(ax, xlabel='', ylabel='Sample RMS\\n(Mean = Module RCI)')\n",
    "        else:\n",
    "            plt.setp(ax.get_yticklabels(), visible=False)\n",
    "            plt.setp(ax, xlabel='', ylabel='')\n",
    "        #Add significance labels\n",
    "        tempDF3 = tempDF2.loc[tempDF2['Dataset']==dataset]\n",
    "        for row_i in range(len(tempDF3)):\n",
    "            #Baseline\n",
    "            group_0 = tempDF3['Baseline'].iloc[row_i]\n",
    "            index_0 = list(tempD1.keys()).index(group_0)\n",
    "            xcoord_0 = index_0\n",
    "            #Contrast\n",
    "            group_1 = tempDF3['Contrast'].iloc[row_i]\n",
    "            index_1 = list(tempD1.keys()).index(group_1)\n",
    "            xcoord_1 = index_1\n",
    "            #Standard point of marker\n",
    "            xcoord = (xcoord_0+xcoord_1)/2\n",
    "            ycoord = aline_ymin + aline_ymargin*row_i\n",
    "            label = tempDF3['SignifLabel'].iloc[row_i]\n",
    "            #Add annotation lines\n",
    "            aline_offset = yinter/5\n",
    "            aline_length = yinter/5 + aline_offset/2\n",
    "            ax.plot([xcoord_0, xcoord_0, xcoord_1, xcoord_1],\n",
    "                    [ycoord+aline_offset, ycoord+aline_length, ycoord+aline_length, ycoord+aline_offset],\n",
    "                    lw=1.5, c='k')\n",
    "            #Add annotation text\n",
    "            if label in ['***', '**', '*']:\n",
    "                text_offset = yinter/21\n",
    "                ax.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                            horizontalalignment='center', verticalalignment='bottom',\n",
    "                            fontsize='medium', color='k')\n",
    "            else:\n",
    "                text_offset = yinter/3.5\n",
    "                ax.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                            horizontalalignment='center', verticalalignment='bottom',\n",
    "                            fontsize='x-small', color='k')\n",
    "        #Add annotation\n",
    "        ax.set_title(dataset, {'fontsize':'small'})\n",
    "        xoff = 0.025\n",
    "        yoff = 0.01\n",
    "        rect = plt.Rectangle((xoff, 1), 1-xoff, 0.1,#Manual adjustment\n",
    "                             transform=ax.transAxes, facecolor=tempD2[dataset],\n",
    "                             clip_on=False, linewidth=0, zorder=0.5)\n",
    "        ax.add_patch(rect)\n",
    "        if ax_i==1:\n",
    "            at = AnchoredText('Consensus: own group', loc='center',\n",
    "                              bbox_to_anchor=(1.15, 0.505), bbox_transform=ax.transAxes,#Minor manual adjustment\n",
    "                              frameon=False, prop={'size':'small', 'rotation':90})\n",
    "            ax.add_artist(at)\n",
    "            rect = plt.Rectangle((1+xoff, yoff), 0.225, 1-2*yoff,#Manual adjustment\n",
    "                                 transform=ax.transAxes, facecolor=plt.get_cmap('tab20')(15),\n",
    "                                 clip_on=False, linewidth=0, zorder=0.5)\n",
    "            ax.add_patch(rect)\n",
    "    fig.tight_layout()\n",
    "    #Set title\n",
    "    modulename = moduleDF.loc[module, 'ModuleName']\n",
    "    initial = modulename[0].capitalize()\n",
    "    title = re.sub('^.', initial, modulename)+' ('+module+')'\n",
    "    title = '\\n'.join(wrap(title, 40))#Because the below wrap=True didn't work\n",
    "    fig.suptitle(title, size='small',\n",
    "                 verticalalignment='bottom', horizontalalignment='center', wrap=True, y=0.93)\n",
    "    ##Save\n",
    "    fileDir = './ExportFigures/'\n",
    "    ipynbName = '220525_LC-M001-DIRAC-prot-vs-txn_Comparison-GOBP_ver2-2_'\n",
    "    fileName = 'RCI-inter-group-comparison_RMS-pointplot-'+module.replace('GO:', 'GO')+'.tif'\n",
    "    plt.gcf().savefig(fileDir+ipynbName+fileName, dpi=300, bbox_inches='tight', pad_inches=0.04,\n",
    "                      pil_kwargs={'compression':'tiff_lzw'})\n",
    "    plt.show()\n",
    "    print('')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 3-5-2. Modules tightened by all interventions specifically in proteomics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Prepare the target module set\n",
    "posL = ['Prot_Acar-vs-Cont', 'Prot_Rapa-vs-Cont']\n",
    "negL = ['Tran_Acar-vs-Cont', 'Tran_Rapa-vs-Cont']\n",
    "regulation = 'Tightened'\n",
    "tempS = pd.Series(np.repeat(True, len(pvalDF)), index=pvalDF.index)#Initialize\n",
    "tempDF = pvalDF.loc[:, pvalDF.columns.str.contains('-vs-')]#Based on adjusted ANOVA P-values\n",
    "for col_n in tempDF.columns.tolist():\n",
    "    tempS1 = pvalDF[col_n]\n",
    "    tempS2 = diffDF[col_n]\n",
    "    if col_n in posL:\n",
    "        if regulation=='Tightened':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2>0)\n",
    "        elif regulation=='Loosened':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2<0)\n",
    "    elif col_n in negL:\n",
    "        tempS3 = (tempS1>=0.05)\n",
    "        #Significance for inverse regulation\n",
    "        if regulation=='Tightened':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2<0)\n",
    "        elif regulation=='Loosened':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2>0)\n",
    "        tempS1 = tempS3 | tempS1\n",
    "    else:\n",
    "        tempS1 = (tempS1>=0.0)\n",
    "    #Update True\n",
    "    tempS = tempS & tempS1\n",
    "tempL = tempS.loc[tempS.tolist()].index.tolist()\n",
    "print(len(tempL), regulation.lower()+' modules with significance in', posL, 'but not in', negL)\n",
    "\n",
    "#Select representatives\n",
    "topX = np.min([30, len(tempL)])\n",
    "topX_plot = np.min([10, len(tempL)])\n",
    "##Spread statDF by dataset to flatten multi-index\n",
    "tempDF = moduleDF['ModuleName']#pd.Series() for now\n",
    "for dataset in ['Proteomics', 'Transcriptomics']:\n",
    "    tempDF1 = statDF.loc[dataset].drop(columns=['ModuleName'])\n",
    "    tempDF1.columns = dataset[0:4]+'_'+tempDF1.columns\n",
    "    tempDF = pd.merge(tempDF, tempDF1, left_index=True, right_index=True, how='left')\n",
    "tempDF = tempDF.loc[:, tempDF.columns.str.contains('Pval')]\n",
    "tempDF = pd.merge(moduleDF['ModuleName'], tempDF, left_index=True, right_index=True, how='left')\n",
    "tempDF = tempDF.loc[tempL]\n",
    "tempDF = tempDF.sort_values(by=['Prot_'+variable+'_AdjPval', 'Tran_'+variable+'_AdjPval'], ascending=True)\n",
    "print('Top', topX, 'modules (sort by the main effect of '+variable+'):')\n",
    "display(tempDF.iloc[:topX])\n",
    "plotL = tempDF.index.tolist()[:topX_plot]\n",
    "\n",
    "#Prepare DF for plot\n",
    "tempDF = rmsDF_kk.reset_index().melt(var_name='SampleID', value_name='RMS', id_vars='ModuleID')\n",
    "tempDF1 = sampleDF.reset_index()[['SampleID', 'Dataset', 'Phenotype']]\n",
    "tempDF = pd.merge(tempDF, tempDF1, on='SampleID', how='left')\n",
    "\n",
    "#Prepare label and color\n",
    "tempD = {'Proteomics':'Proteins', 'Transcriptomics':'Transcripts'}\n",
    "tempDF['Dataset'] = tempDF['Dataset'].map(tempD)\n",
    "tempDF['Group'] = tempDF['Phenotype']\n",
    "tempD0 = {'Cont':'Control', 'Acar':'Acarbose', 'Rapa':'Rapamycin', 'Prot':'Proteins', 'Tran':'Transcripts'}\n",
    "tempD1 = {'Control':'tab:blue', 'Acarbose':'tab:red', 'Rapamycin':'tab:purple'}\n",
    "tempD2 = {'Proteins':plt.get_cmap('tab20')(13), 'Transcripts':plt.get_cmap('tab20')(19)}\n",
    "\n",
    "#Visualize each representative\n",
    "for rank_i in range(len(plotL)):\n",
    "    print(' - Rank '+str(rank_i+1)+' (sort by the main effect of '+variable+'):')\n",
    "    module = plotL[rank_i]\n",
    "    #Check module summary\n",
    "    tempDF1 = pd.DataFrame(moduleDF.loc[module]).T\n",
    "    display(tempDF1)\n",
    "    \n",
    "    #Select RMS\n",
    "    tempDF1 = tempDF.loc[tempDF['ModuleID']==module]\n",
    "    \n",
    "    #Check RMS summary\n",
    "    tempDF2 = tempDF1.groupby(['Dataset', 'Group'])['RMS'].agg(['count', 'mean', 'std'])\n",
    "    tempL1 = []\n",
    "    tempL2 = []\n",
    "    for row_n in tempDF2.index.tolist():\n",
    "        count, mean, std = tempDF2.loc[row_n]\n",
    "        tempL1.append(mean - 1.96*std/np.sqrt(count))\n",
    "        tempL2.append(mean + 1.96*std/np.sqrt(count))\n",
    "    tempDF2['0.025'] = tempL1\n",
    "    tempDF2['0.975'] = tempL2\n",
    "    ##Multiindex sort\n",
    "    tempDF2 = tempDF2.reset_index()\n",
    "    tempDF2['Dataset'] = pd.Categorical(tempDF2['Dataset'], categories=list(tempD2.keys()))\n",
    "    tempDF2['Group'] = pd.Categorical(tempDF2['Group'], categories=list(tempD1.keys()))\n",
    "    tempDF2 = tempDF2.sort_values(by=['Dataset', 'Group']).set_index(['Dataset', 'Group'])\n",
    "    display(tempDF2)\n",
    "    \n",
    "    #Prepare significance labels\n",
    "    ##Retrieve statistical significance\n",
    "    tempS = pvalDF.loc[module, pvalDF.columns.str.contains('-vs-')]\n",
    "    tempS.name = 'AdjPval'\n",
    "    ##Clean\n",
    "    tempDF2 = tempS.index.to_series().str.split(pat='_', expand=True)\n",
    "    tempDF2 = tempDF2.rename(columns={0:'Dataset', 1:'Comparison'})\n",
    "    tempS1 = tempDF2['Dataset']\n",
    "    tempDF2 = tempDF2['Comparison'].str.split(pat='-vs-', expand=True)\n",
    "    tempDF2 = tempDF2.rename(columns={0:'Contrast', 1:'Baseline'})\n",
    "    tempDF2 = pd.merge(tempS1, tempDF2, left_index=True, right_index=True, how='left')\n",
    "    tempDF2 = pd.merge(tempDF2, tempS, left_index=True, right_index=True, how='left')\n",
    "    tempDF2['Dataset'] = tempDF2['Dataset'].map(tempD0)\n",
    "    tempDF2['Contrast'] = tempDF2['Contrast'].map(tempD0)\n",
    "    tempDF2['Baseline'] = tempDF2['Baseline'].map(tempD0)\n",
    "    ##Convert p-value to label\n",
    "    tempL = []\n",
    "    for row_i in range(len(tempDF2)):\n",
    "        pval = tempDF2['AdjPval'].iloc[row_i]\n",
    "        if pval<0.001:\n",
    "            tempL.append('***')\n",
    "        elif pval<0.01:\n",
    "            tempL.append('**')\n",
    "        elif pval<0.05:\n",
    "            tempL.append('*')\n",
    "        else:\n",
    "            pval_text = Decimal(str(pval)).quantize(Decimal('0.01'), rounding=ROUND_HALF_UP)\n",
    "            tempL.append(r'$P$ = '+str(pval_text))\n",
    "    tempDF2['SignifLabel'] = tempL\n",
    "    display(tempDF2)\n",
    "    \n",
    "    #Visualization\n",
    "    ymax = 1.0\n",
    "    ymin = 0.0\n",
    "    yinter = 0.2\n",
    "    ymargin_t = 0.275\n",
    "    ymargin_b = 0.05\n",
    "    aline_ymin = 1.0\n",
    "    aline_ymargin = 0.125\n",
    "    sns.set(style='ticks', font='Arial', context='talk')\n",
    "    fig, axes = plt.subplots(nrows=1, ncols=2, figsize=(4.75, 5), sharex=True, sharey=True,\n",
    "                             gridspec_kw={'width_ratios':[2, 2]})\n",
    "    for ax_i, ax in enumerate(axes.flat):\n",
    "        dataset = list(tempD2.keys())[ax_i]\n",
    "        tempDF3 = tempDF1.loc[tempDF1['Dataset']==dataset]\n",
    "        sns.pointplot(data=tempDF3, x='Group', y='RMS', order=list(tempD1.keys()), palette=tempD1,\n",
    "                      markers='o', dodge=False, join=False, capsize=0.6, estimator=np.mean, ci=95, ax=ax)\n",
    "        sns.stripplot(data=tempDF3, x='Group', y='RMS',\n",
    "                      order=list(tempD1.keys()), palette=tempD1, dodge=False, jitter=0.15,\n",
    "                      size=5, edgecolor='black', linewidth=1, **{'marker':'o', 'alpha':0.5}, ax=ax)\n",
    "        #Set axis\n",
    "        sns.despine()\n",
    "        ax.set(ylim=(ymin-ymargin_b, ymax+ymargin_t), yticks=np.arange(ymin, ymax + yinter/10, yinter))\n",
    "        plt.setp(ax.get_xticklabels(), rotation=70, horizontalalignment='right',\n",
    "                 verticalalignment='center', rotation_mode='anchor')\n",
    "        if ax_i==0:\n",
    "            plt.setp(ax, xlabel='', ylabel='Sample RMS\\n(Mean = Module RCI)')\n",
    "        else:\n",
    "            plt.setp(ax.get_yticklabels(), visible=False)\n",
    "            plt.setp(ax, xlabel='', ylabel='')\n",
    "        #Add significance labels\n",
    "        tempDF3 = tempDF2.loc[tempDF2['Dataset']==dataset]\n",
    "        for row_i in range(len(tempDF3)):\n",
    "            #Baseline\n",
    "            group_0 = tempDF3['Baseline'].iloc[row_i]\n",
    "            index_0 = list(tempD1.keys()).index(group_0)\n",
    "            xcoord_0 = index_0\n",
    "            #Contrast\n",
    "            group_1 = tempDF3['Contrast'].iloc[row_i]\n",
    "            index_1 = list(tempD1.keys()).index(group_1)\n",
    "            xcoord_1 = index_1\n",
    "            #Standard point of marker\n",
    "            xcoord = (xcoord_0+xcoord_1)/2\n",
    "            ycoord = aline_ymin + aline_ymargin*row_i\n",
    "            label = tempDF3['SignifLabel'].iloc[row_i]\n",
    "            #Add annotation lines\n",
    "            aline_offset = yinter/5\n",
    "            aline_length = yinter/5 + aline_offset/2\n",
    "            ax.plot([xcoord_0, xcoord_0, xcoord_1, xcoord_1],\n",
    "                    [ycoord+aline_offset, ycoord+aline_length, ycoord+aline_length, ycoord+aline_offset],\n",
    "                    lw=1.5, c='k')\n",
    "            #Add annotation text\n",
    "            if label in ['***', '**', '*']:\n",
    "                text_offset = yinter/21\n",
    "                ax.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                            horizontalalignment='center', verticalalignment='bottom',\n",
    "                            fontsize='medium', color='k')\n",
    "            else:\n",
    "                text_offset = yinter/3.5\n",
    "                ax.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                            horizontalalignment='center', verticalalignment='bottom',\n",
    "                            fontsize='x-small', color='k')\n",
    "        #Add annotation\n",
    "        ax.set_title(dataset, {'fontsize':'small'})\n",
    "        xoff = 0.025\n",
    "        yoff = 0.01\n",
    "        rect = plt.Rectangle((xoff, 1), 1-xoff, 0.1,#Manual adjustment\n",
    "                             transform=ax.transAxes, facecolor=tempD2[dataset],\n",
    "                             clip_on=False, linewidth=0, zorder=0.5)\n",
    "        ax.add_patch(rect)\n",
    "        if ax_i==1:\n",
    "            at = AnchoredText('Consensus: own group', loc='center',\n",
    "                              bbox_to_anchor=(1.15, 0.505), bbox_transform=ax.transAxes,#Minor manual adjustment\n",
    "                              frameon=False, prop={'size':'small', 'rotation':90})\n",
    "            ax.add_artist(at)\n",
    "            rect = plt.Rectangle((1+xoff, yoff), 0.225, 1-2*yoff,#Manual adjustment\n",
    "                                 transform=ax.transAxes, facecolor=plt.get_cmap('tab20')(15),\n",
    "                                 clip_on=False, linewidth=0, zorder=0.5)\n",
    "            ax.add_patch(rect)\n",
    "    fig.tight_layout()\n",
    "    #Set title\n",
    "    modulename = moduleDF.loc[module, 'ModuleName']\n",
    "    initial = modulename[0].capitalize()\n",
    "    title = re.sub('^.', initial, modulename)+' ('+module+')'\n",
    "    title = '\\n'.join(wrap(title, 40))#Because the below wrap=True didn't work\n",
    "    fig.suptitle(title, size='small',\n",
    "                 verticalalignment='bottom', horizontalalignment='center', wrap=True, y=0.93)\n",
    "    ##Save\n",
    "    fileDir = './ExportFigures/'\n",
    "    ipynbName = '220525_LC-M001-DIRAC-prot-vs-txn_Comparison-GOBP_ver2-2_'\n",
    "    fileName = 'RCI-inter-group-comparison_RMS-pointplot-'+module.replace('GO:', 'GO')+'.tif'\n",
    "    plt.gcf().savefig(fileDir+ipynbName+fileName, dpi=300, bbox_inches='tight', pad_inches=0.04,\n",
    "                      pil_kwargs={'compression':'tiff_lzw'})\n",
    "    plt.show()\n",
    "    print('')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 3-5-3. Modules tightened by all interventions specifically in transcriptomics"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> Because proteins are the players for cellular functions, this module is not important even if positive, though.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Prepare the target module set\n",
    "posL = ['Tran_Acar-vs-Cont', 'Tran_Rapa-vs-Cont']\n",
    "negL = ['Prot_Acar-vs-Cont', 'Prot_Rapa-vs-Cont']\n",
    "regulation = 'Tightened'\n",
    "tempS = pd.Series(np.repeat(True, len(pvalDF)), index=pvalDF.index)#Initialize\n",
    "tempDF = pvalDF.loc[:, pvalDF.columns.str.contains('-vs-')]#Based on adjusted ANOVA P-values\n",
    "for col_n in tempDF.columns.tolist():\n",
    "    tempS1 = pvalDF[col_n]\n",
    "    tempS2 = diffDF[col_n]\n",
    "    if col_n in posL:\n",
    "        if regulation=='Tightened':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2>0)\n",
    "        elif regulation=='Loosened':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2<0)\n",
    "    elif col_n in negL:\n",
    "        tempS3 = (tempS1>=0.05)\n",
    "        #Significance for inverse regulation\n",
    "        if regulation=='Tightened':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2<0)\n",
    "        elif regulation=='Loosened':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2>0)\n",
    "        tempS1 = tempS3 | tempS1\n",
    "    else:\n",
    "        tempS1 = (tempS1>=0.0)\n",
    "    #Update True\n",
    "    tempS = tempS & tempS1\n",
    "tempL = tempS.loc[tempS.tolist()].index.tolist()\n",
    "print(len(tempL), regulation.lower()+' modules with significance in', posL, 'but not in', negL)\n",
    "\n",
    "#Select representatives\n",
    "topX = np.min([30, len(tempL)])\n",
    "topX_plot = np.min([10, len(tempL)])\n",
    "##Spread statDF by dataset to flatten multi-index\n",
    "tempDF = moduleDF['ModuleName']#pd.Series() for now\n",
    "for dataset in ['Proteomics', 'Transcriptomics']:\n",
    "    tempDF1 = statDF.loc[dataset].drop(columns=['ModuleName'])\n",
    "    tempDF1.columns = dataset[0:4]+'_'+tempDF1.columns\n",
    "    tempDF = pd.merge(tempDF, tempDF1, left_index=True, right_index=True, how='left')\n",
    "tempDF = tempDF.loc[:, tempDF.columns.str.contains('Pval')]\n",
    "tempDF = pd.merge(moduleDF['ModuleName'], tempDF, left_index=True, right_index=True, how='left')\n",
    "tempDF = tempDF.loc[tempL]\n",
    "tempDF = tempDF.sort_values(by=['Prot_'+variable+'_AdjPval', 'Tran_'+variable+'_AdjPval'], ascending=True)\n",
    "print('Top', topX, 'modules (sort by the main effect of '+variable+'):')\n",
    "display(tempDF.iloc[:topX])\n",
    "plotL = tempDF.index.tolist()[:topX_plot]\n",
    "\n",
    "#Prepare DF for plot\n",
    "tempDF = rmsDF_kk.reset_index().melt(var_name='SampleID', value_name='RMS', id_vars='ModuleID')\n",
    "tempDF1 = sampleDF.reset_index()[['SampleID', 'Dataset', 'Phenotype']]\n",
    "tempDF = pd.merge(tempDF, tempDF1, on='SampleID', how='left')\n",
    "\n",
    "#Prepare label and color\n",
    "tempD = {'Proteomics':'Proteins', 'Transcriptomics':'Transcripts'}\n",
    "tempDF['Dataset'] = tempDF['Dataset'].map(tempD)\n",
    "tempDF['Group'] = tempDF['Phenotype']\n",
    "tempD0 = {'Cont':'Control', 'Acar':'Acarbose', 'Rapa':'Rapamycin', 'Prot':'Proteins', 'Tran':'Transcripts'}\n",
    "tempD1 = {'Control':'tab:blue', 'Acarbose':'tab:red', 'Rapamycin':'tab:purple'}\n",
    "tempD2 = {'Proteins':plt.get_cmap('tab20')(13), 'Transcripts':plt.get_cmap('tab20')(19)}\n",
    "\n",
    "#Visualize each representative\n",
    "for rank_i in range(len(plotL)):\n",
    "    print(' - Rank '+str(rank_i+1)+' (sort by the main effect of '+variable+'):')\n",
    "    module = plotL[rank_i]\n",
    "    #Check module summary\n",
    "    tempDF1 = pd.DataFrame(moduleDF.loc[module]).T\n",
    "    display(tempDF1)\n",
    "    \n",
    "    #Select RMS\n",
    "    tempDF1 = tempDF.loc[tempDF['ModuleID']==module]\n",
    "    \n",
    "    #Check RMS summary\n",
    "    tempDF2 = tempDF1.groupby(['Dataset', 'Group'])['RMS'].agg(['count', 'mean', 'std'])\n",
    "    tempL1 = []\n",
    "    tempL2 = []\n",
    "    for row_n in tempDF2.index.tolist():\n",
    "        count, mean, std = tempDF2.loc[row_n]\n",
    "        tempL1.append(mean - 1.96*std/np.sqrt(count))\n",
    "        tempL2.append(mean + 1.96*std/np.sqrt(count))\n",
    "    tempDF2['0.025'] = tempL1\n",
    "    tempDF2['0.975'] = tempL2\n",
    "    ##Multiindex sort\n",
    "    tempDF2 = tempDF2.reset_index()\n",
    "    tempDF2['Dataset'] = pd.Categorical(tempDF2['Dataset'], categories=list(tempD2.keys()))\n",
    "    tempDF2['Group'] = pd.Categorical(tempDF2['Group'], categories=list(tempD1.keys()))\n",
    "    tempDF2 = tempDF2.sort_values(by=['Dataset', 'Group']).set_index(['Dataset', 'Group'])\n",
    "    display(tempDF2)\n",
    "    \n",
    "    #Prepare significance labels\n",
    "    ##Retrieve statistical significance\n",
    "    tempS = pvalDF.loc[module, pvalDF.columns.str.contains('-vs-')]\n",
    "    tempS.name = 'AdjPval'\n",
    "    ##Clean\n",
    "    tempDF2 = tempS.index.to_series().str.split(pat='_', expand=True)\n",
    "    tempDF2 = tempDF2.rename(columns={0:'Dataset', 1:'Comparison'})\n",
    "    tempS1 = tempDF2['Dataset']\n",
    "    tempDF2 = tempDF2['Comparison'].str.split(pat='-vs-', expand=True)\n",
    "    tempDF2 = tempDF2.rename(columns={0:'Contrast', 1:'Baseline'})\n",
    "    tempDF2 = pd.merge(tempS1, tempDF2, left_index=True, right_index=True, how='left')\n",
    "    tempDF2 = pd.merge(tempDF2, tempS, left_index=True, right_index=True, how='left')\n",
    "    tempDF2['Dataset'] = tempDF2['Dataset'].map(tempD0)\n",
    "    tempDF2['Contrast'] = tempDF2['Contrast'].map(tempD0)\n",
    "    tempDF2['Baseline'] = tempDF2['Baseline'].map(tempD0)\n",
    "    ##Convert p-value to label\n",
    "    tempL = []\n",
    "    for row_i in range(len(tempDF2)):\n",
    "        pval = tempDF2['AdjPval'].iloc[row_i]\n",
    "        if pval<0.001:\n",
    "            tempL.append('***')\n",
    "        elif pval<0.01:\n",
    "            tempL.append('**')\n",
    "        elif pval<0.05:\n",
    "            tempL.append('*')\n",
    "        else:\n",
    "            pval_text = Decimal(str(pval)).quantize(Decimal('0.01'), rounding=ROUND_HALF_UP)\n",
    "            tempL.append(r'$P$ = '+str(pval_text))\n",
    "    tempDF2['SignifLabel'] = tempL\n",
    "    display(tempDF2)\n",
    "    \n",
    "    #Visualization\n",
    "    ymax = 1.0\n",
    "    ymin = 0.0\n",
    "    yinter = 0.2\n",
    "    ymargin_t = 0.275\n",
    "    ymargin_b = 0.05\n",
    "    aline_ymin = 1.0\n",
    "    aline_ymargin = 0.125\n",
    "    sns.set(style='ticks', font='Arial', context='talk')\n",
    "    fig, axes = plt.subplots(nrows=1, ncols=2, figsize=(4.75, 5), sharex=True, sharey=True,\n",
    "                             gridspec_kw={'width_ratios':[2, 2]})\n",
    "    for ax_i, ax in enumerate(axes.flat):\n",
    "        dataset = list(tempD2.keys())[ax_i]\n",
    "        tempDF3 = tempDF1.loc[tempDF1['Dataset']==dataset]\n",
    "        sns.pointplot(data=tempDF3, x='Group', y='RMS', order=list(tempD1.keys()), palette=tempD1,\n",
    "                      markers='o', dodge=False, join=False, capsize=0.6, estimator=np.mean, ci=95, ax=ax)\n",
    "        sns.stripplot(data=tempDF3, x='Group', y='RMS',\n",
    "                      order=list(tempD1.keys()), palette=tempD1, dodge=False, jitter=0.15,\n",
    "                      size=5, edgecolor='black', linewidth=1, **{'marker':'o', 'alpha':0.5}, ax=ax)\n",
    "        #Set axis\n",
    "        sns.despine()\n",
    "        ax.set(ylim=(ymin-ymargin_b, ymax+ymargin_t), yticks=np.arange(ymin, ymax + yinter/10, yinter))\n",
    "        plt.setp(ax.get_xticklabels(), rotation=70, horizontalalignment='right',\n",
    "                 verticalalignment='center', rotation_mode='anchor')\n",
    "        if ax_i==0:\n",
    "            plt.setp(ax, xlabel='', ylabel='Sample RMS\\n(Mean = Module RCI)')\n",
    "        else:\n",
    "            plt.setp(ax.get_yticklabels(), visible=False)\n",
    "            plt.setp(ax, xlabel='', ylabel='')\n",
    "        #Add significance labels\n",
    "        tempDF3 = tempDF2.loc[tempDF2['Dataset']==dataset]\n",
    "        for row_i in range(len(tempDF3)):\n",
    "            #Baseline\n",
    "            group_0 = tempDF3['Baseline'].iloc[row_i]\n",
    "            index_0 = list(tempD1.keys()).index(group_0)\n",
    "            xcoord_0 = index_0\n",
    "            #Contrast\n",
    "            group_1 = tempDF3['Contrast'].iloc[row_i]\n",
    "            index_1 = list(tempD1.keys()).index(group_1)\n",
    "            xcoord_1 = index_1\n",
    "            #Standard point of marker\n",
    "            xcoord = (xcoord_0+xcoord_1)/2\n",
    "            ycoord = aline_ymin + aline_ymargin*row_i\n",
    "            label = tempDF3['SignifLabel'].iloc[row_i]\n",
    "            #Add annotation lines\n",
    "            aline_offset = yinter/5\n",
    "            aline_length = yinter/5 + aline_offset/2\n",
    "            ax.plot([xcoord_0, xcoord_0, xcoord_1, xcoord_1],\n",
    "                    [ycoord+aline_offset, ycoord+aline_length, ycoord+aline_length, ycoord+aline_offset],\n",
    "                    lw=1.5, c='k')\n",
    "            #Add annotation text\n",
    "            if label in ['***', '**', '*']:\n",
    "                text_offset = yinter/21\n",
    "                ax.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                            horizontalalignment='center', verticalalignment='bottom',\n",
    "                            fontsize='medium', color='k')\n",
    "            else:\n",
    "                text_offset = yinter/3.5\n",
    "                ax.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                            horizontalalignment='center', verticalalignment='bottom',\n",
    "                            fontsize='x-small', color='k')\n",
    "        #Add annotation\n",
    "        ax.set_title(dataset, {'fontsize':'small'})\n",
    "        xoff = 0.025\n",
    "        yoff = 0.01\n",
    "        rect = plt.Rectangle((xoff, 1), 1-xoff, 0.1,#Manual adjustment\n",
    "                             transform=ax.transAxes, facecolor=tempD2[dataset],\n",
    "                             clip_on=False, linewidth=0, zorder=0.5)\n",
    "        ax.add_patch(rect)\n",
    "        if ax_i==1:\n",
    "            at = AnchoredText('Consensus: own group', loc='center',\n",
    "                              bbox_to_anchor=(1.15, 0.505), bbox_transform=ax.transAxes,#Minor manual adjustment\n",
    "                              frameon=False, prop={'size':'small', 'rotation':90})\n",
    "            ax.add_artist(at)\n",
    "            rect = plt.Rectangle((1+xoff, yoff), 0.225, 1-2*yoff,#Manual adjustment\n",
    "                                 transform=ax.transAxes, facecolor=plt.get_cmap('tab20')(15),\n",
    "                                 clip_on=False, linewidth=0, zorder=0.5)\n",
    "            ax.add_patch(rect)\n",
    "    fig.tight_layout()\n",
    "    #Set title\n",
    "    modulename = moduleDF.loc[module, 'ModuleName']\n",
    "    initial = modulename[0].capitalize()\n",
    "    title = re.sub('^.', initial, modulename)+' ('+module+')'\n",
    "    title = '\\n'.join(wrap(title, 40))#Because the below wrap=True didn't work\n",
    "    fig.suptitle(title, size='small',\n",
    "                 verticalalignment='bottom', horizontalalignment='center', wrap=True, y=0.93)\n",
    "    ##Save\n",
    "    fileDir = './ExportFigures/'\n",
    "    ipynbName = '220525_LC-M001-DIRAC-prot-vs-txn_Comparison-GOBP_ver2-2_'\n",
    "    fileName = 'RCI-inter-group-comparison_RMS-pointplot-'+module.replace('GO:', 'GO')+'.tif'\n",
    "    plt.gcf().savefig(fileDir+ipynbName+fileName, dpi=300, bbox_inches='tight', pad_inches=0.04,\n",
    "                      pil_kwargs={'compression':'tiff_lzw'})\n",
    "    plt.show()\n",
    "    print('')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 3-5-4. Modules tightened by Aca in both datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Prepare the target module set\n",
    "posL = ['Prot_Acar-vs-Cont', 'Tran_Acar-vs-Cont']\n",
    "negL = []\n",
    "regulation = 'Tightened'\n",
    "tempS = pd.Series(np.repeat(True, len(pvalDF)), index=pvalDF.index)#Initialize\n",
    "tempDF = pvalDF.loc[:, pvalDF.columns.str.contains('-vs-')]#Based on adjusted ANOVA P-values\n",
    "for col_n in tempDF.columns.tolist():\n",
    "    tempS1 = pvalDF[col_n]\n",
    "    tempS2 = diffDF[col_n]\n",
    "    if col_n in posL:\n",
    "        if regulation=='Tightened':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2>0)\n",
    "        elif regulation=='Loosened':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2<0)\n",
    "    elif col_n in negL:\n",
    "        tempS3 = (tempS1>=0.05)\n",
    "        #Significance for inverse regulation\n",
    "        if regulation=='Tightened':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2<0)\n",
    "        elif regulation=='Loosened':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2>0)\n",
    "        tempS1 = tempS3 | tempS1\n",
    "    else:\n",
    "        tempS1 = (tempS1>=0.0)\n",
    "    #Update True\n",
    "    tempS = tempS & tempS1\n",
    "tempL = tempS.loc[tempS.tolist()].index.tolist()\n",
    "print(len(tempL), regulation.lower()+' modules with significance in', posL, 'but not in', negL)\n",
    "\n",
    "#Select representatives\n",
    "topX = np.min([30, len(tempL)])\n",
    "topX_plot = np.min([10, len(tempL)])\n",
    "##Spread statDF by dataset to flatten multi-index\n",
    "tempDF = moduleDF['ModuleName']#pd.Series() for now\n",
    "for dataset in ['Proteomics', 'Transcriptomics']:\n",
    "    tempDF1 = statDF.loc[dataset].drop(columns=['ModuleName'])\n",
    "    tempDF1.columns = dataset[0:4]+'_'+tempDF1.columns\n",
    "    tempDF = pd.merge(tempDF, tempDF1, left_index=True, right_index=True, how='left')\n",
    "tempDF = tempDF.loc[:, tempDF.columns.str.contains('Pval')]\n",
    "tempDF = pd.merge(moduleDF['ModuleName'], tempDF, left_index=True, right_index=True, how='left')\n",
    "tempDF = tempDF.loc[tempL]\n",
    "tempDF = tempDF.sort_values(by=['Prot_'+variable+'_AdjPval', 'Tran_'+variable+'_AdjPval'], ascending=True)\n",
    "print('Top', topX, 'modules (sort by the main effect of '+variable+'):')\n",
    "display(tempDF.iloc[:topX])\n",
    "plotL = tempDF.index.tolist()[:topX_plot]\n",
    "\n",
    "#Prepare DF for plot\n",
    "tempDF = rmsDF_kk.reset_index().melt(var_name='SampleID', value_name='RMS', id_vars='ModuleID')\n",
    "tempDF1 = sampleDF.reset_index()[['SampleID', 'Dataset', 'Phenotype']]\n",
    "tempDF = pd.merge(tempDF, tempDF1, on='SampleID', how='left')\n",
    "\n",
    "#Prepare label and color\n",
    "tempD = {'Proteomics':'Proteins', 'Transcriptomics':'Transcripts'}\n",
    "tempDF['Dataset'] = tempDF['Dataset'].map(tempD)\n",
    "tempDF['Group'] = tempDF['Phenotype']\n",
    "tempD0 = {'Cont':'Control', 'Acar':'Acarbose', 'Rapa':'Rapamycin', 'Prot':'Proteins', 'Tran':'Transcripts'}\n",
    "tempD1 = {'Control':'tab:blue', 'Acarbose':'tab:red', 'Rapamycin':'tab:purple'}\n",
    "tempD2 = {'Proteins':plt.get_cmap('tab20')(13), 'Transcripts':plt.get_cmap('tab20')(19)}\n",
    "\n",
    "#Visualize each representative\n",
    "for rank_i in range(len(plotL)):\n",
    "    print(' - Rank '+str(rank_i+1)+' (sort by the main effect of '+variable+'):')\n",
    "    module = plotL[rank_i]\n",
    "    #Check module summary\n",
    "    tempDF1 = pd.DataFrame(moduleDF.loc[module]).T\n",
    "    display(tempDF1)\n",
    "    \n",
    "    #Select RMS\n",
    "    tempDF1 = tempDF.loc[tempDF['ModuleID']==module]\n",
    "    \n",
    "    #Check RMS summary\n",
    "    tempDF2 = tempDF1.groupby(['Dataset', 'Group'])['RMS'].agg(['count', 'mean', 'std'])\n",
    "    tempL1 = []\n",
    "    tempL2 = []\n",
    "    for row_n in tempDF2.index.tolist():\n",
    "        count, mean, std = tempDF2.loc[row_n]\n",
    "        tempL1.append(mean - 1.96*std/np.sqrt(count))\n",
    "        tempL2.append(mean + 1.96*std/np.sqrt(count))\n",
    "    tempDF2['0.025'] = tempL1\n",
    "    tempDF2['0.975'] = tempL2\n",
    "    ##Multiindex sort\n",
    "    tempDF2 = tempDF2.reset_index()\n",
    "    tempDF2['Dataset'] = pd.Categorical(tempDF2['Dataset'], categories=list(tempD2.keys()))\n",
    "    tempDF2['Group'] = pd.Categorical(tempDF2['Group'], categories=list(tempD1.keys()))\n",
    "    tempDF2 = tempDF2.sort_values(by=['Dataset', 'Group']).set_index(['Dataset', 'Group'])\n",
    "    display(tempDF2)\n",
    "    \n",
    "    #Prepare significance labels\n",
    "    ##Retrieve statistical significance\n",
    "    tempS = pvalDF.loc[module, pvalDF.columns.str.contains('-vs-')]\n",
    "    tempS.name = 'AdjPval'\n",
    "    ##Clean\n",
    "    tempDF2 = tempS.index.to_series().str.split(pat='_', expand=True)\n",
    "    tempDF2 = tempDF2.rename(columns={0:'Dataset', 1:'Comparison'})\n",
    "    tempS1 = tempDF2['Dataset']\n",
    "    tempDF2 = tempDF2['Comparison'].str.split(pat='-vs-', expand=True)\n",
    "    tempDF2 = tempDF2.rename(columns={0:'Contrast', 1:'Baseline'})\n",
    "    tempDF2 = pd.merge(tempS1, tempDF2, left_index=True, right_index=True, how='left')\n",
    "    tempDF2 = pd.merge(tempDF2, tempS, left_index=True, right_index=True, how='left')\n",
    "    tempDF2['Dataset'] = tempDF2['Dataset'].map(tempD0)\n",
    "    tempDF2['Contrast'] = tempDF2['Contrast'].map(tempD0)\n",
    "    tempDF2['Baseline'] = tempDF2['Baseline'].map(tempD0)\n",
    "    ##Convert p-value to label\n",
    "    tempL = []\n",
    "    for row_i in range(len(tempDF2)):\n",
    "        pval = tempDF2['AdjPval'].iloc[row_i]\n",
    "        if pval<0.001:\n",
    "            tempL.append('***')\n",
    "        elif pval<0.01:\n",
    "            tempL.append('**')\n",
    "        elif pval<0.05:\n",
    "            tempL.append('*')\n",
    "        else:\n",
    "            pval_text = Decimal(str(pval)).quantize(Decimal('0.01'), rounding=ROUND_HALF_UP)\n",
    "            tempL.append(r'$P$ = '+str(pval_text))\n",
    "    tempDF2['SignifLabel'] = tempL\n",
    "    display(tempDF2)\n",
    "    \n",
    "    #Visualization\n",
    "    ymax = 1.0\n",
    "    ymin = 0.0\n",
    "    yinter = 0.2\n",
    "    ymargin_t = 0.275\n",
    "    ymargin_b = 0.05\n",
    "    aline_ymin = 1.0\n",
    "    aline_ymargin = 0.125\n",
    "    sns.set(style='ticks', font='Arial', context='talk')\n",
    "    fig, axes = plt.subplots(nrows=1, ncols=2, figsize=(4.75, 5), sharex=True, sharey=True,\n",
    "                             gridspec_kw={'width_ratios':[2, 2]})\n",
    "    for ax_i, ax in enumerate(axes.flat):\n",
    "        dataset = list(tempD2.keys())[ax_i]\n",
    "        tempDF3 = tempDF1.loc[tempDF1['Dataset']==dataset]\n",
    "        sns.pointplot(data=tempDF3, x='Group', y='RMS', order=list(tempD1.keys()), palette=tempD1,\n",
    "                      markers='o', dodge=False, join=False, capsize=0.6, estimator=np.mean, ci=95, ax=ax)\n",
    "        sns.stripplot(data=tempDF3, x='Group', y='RMS',\n",
    "                      order=list(tempD1.keys()), palette=tempD1, dodge=False, jitter=0.15,\n",
    "                      size=5, edgecolor='black', linewidth=1, **{'marker':'o', 'alpha':0.5}, ax=ax)\n",
    "        #Set axis\n",
    "        sns.despine()\n",
    "        ax.set(ylim=(ymin-ymargin_b, ymax+ymargin_t), yticks=np.arange(ymin, ymax + yinter/10, yinter))\n",
    "        plt.setp(ax.get_xticklabels(), rotation=70, horizontalalignment='right',\n",
    "                 verticalalignment='center', rotation_mode='anchor')\n",
    "        if ax_i==0:\n",
    "            plt.setp(ax, xlabel='', ylabel='Sample RMS\\n(Mean = Module RCI)')\n",
    "        else:\n",
    "            plt.setp(ax.get_yticklabels(), visible=False)\n",
    "            plt.setp(ax, xlabel='', ylabel='')\n",
    "        #Add significance labels\n",
    "        tempDF3 = tempDF2.loc[tempDF2['Dataset']==dataset]\n",
    "        for row_i in range(len(tempDF3)):\n",
    "            #Baseline\n",
    "            group_0 = tempDF3['Baseline'].iloc[row_i]\n",
    "            index_0 = list(tempD1.keys()).index(group_0)\n",
    "            xcoord_0 = index_0\n",
    "            #Contrast\n",
    "            group_1 = tempDF3['Contrast'].iloc[row_i]\n",
    "            index_1 = list(tempD1.keys()).index(group_1)\n",
    "            xcoord_1 = index_1\n",
    "            #Standard point of marker\n",
    "            xcoord = (xcoord_0+xcoord_1)/2\n",
    "            ycoord = aline_ymin + aline_ymargin*row_i\n",
    "            label = tempDF3['SignifLabel'].iloc[row_i]\n",
    "            #Add annotation lines\n",
    "            aline_offset = yinter/5\n",
    "            aline_length = yinter/5 + aline_offset/2\n",
    "            ax.plot([xcoord_0, xcoord_0, xcoord_1, xcoord_1],\n",
    "                    [ycoord+aline_offset, ycoord+aline_length, ycoord+aline_length, ycoord+aline_offset],\n",
    "                    lw=1.5, c='k')\n",
    "            #Add annotation text\n",
    "            if label in ['***', '**', '*']:\n",
    "                text_offset = yinter/21\n",
    "                ax.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                            horizontalalignment='center', verticalalignment='bottom',\n",
    "                            fontsize='medium', color='k')\n",
    "            else:\n",
    "                text_offset = yinter/3.5\n",
    "                ax.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                            horizontalalignment='center', verticalalignment='bottom',\n",
    "                            fontsize='x-small', color='k')\n",
    "        #Add annotation\n",
    "        ax.set_title(dataset, {'fontsize':'small'})\n",
    "        xoff = 0.025\n",
    "        yoff = 0.01\n",
    "        rect = plt.Rectangle((xoff, 1), 1-xoff, 0.1,#Manual adjustment\n",
    "                             transform=ax.transAxes, facecolor=tempD2[dataset],\n",
    "                             clip_on=False, linewidth=0, zorder=0.5)\n",
    "        ax.add_patch(rect)\n",
    "        if ax_i==1:\n",
    "            at = AnchoredText('Consensus: own group', loc='center',\n",
    "                              bbox_to_anchor=(1.15, 0.505), bbox_transform=ax.transAxes,#Minor manual adjustment\n",
    "                              frameon=False, prop={'size':'small', 'rotation':90})\n",
    "            ax.add_artist(at)\n",
    "            rect = plt.Rectangle((1+xoff, yoff), 0.225, 1-2*yoff,#Manual adjustment\n",
    "                                 transform=ax.transAxes, facecolor=plt.get_cmap('tab20')(15),\n",
    "                                 clip_on=False, linewidth=0, zorder=0.5)\n",
    "            ax.add_patch(rect)\n",
    "    fig.tight_layout()\n",
    "    #Set title\n",
    "    modulename = moduleDF.loc[module, 'ModuleName']\n",
    "    initial = modulename[0].capitalize()\n",
    "    title = re.sub('^.', initial, modulename)+' ('+module+')'\n",
    "    title = '\\n'.join(wrap(title, 40))#Because the below wrap=True didn't work\n",
    "    fig.suptitle(title, size='small',\n",
    "                 verticalalignment='bottom', horizontalalignment='center', wrap=True, y=0.93)\n",
    "    ##Save\n",
    "    fileDir = './ExportFigures/'\n",
    "    ipynbName = '220525_LC-M001-DIRAC-prot-vs-txn_Comparison-GOBP_ver2-2_'\n",
    "    fileName = 'RCI-inter-group-comparison_RMS-pointplot-'+module.replace('GO:', 'GO')+'.tif'\n",
    "    plt.gcf().savefig(fileDir+ipynbName+fileName, dpi=300, bbox_inches='tight', pad_inches=0.04,\n",
    "                      pil_kwargs={'compression':'tiff_lzw'})\n",
    "    plt.show()\n",
    "    print('')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 3-5-5. Modules tightened by Aca specifically in proteomics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Prepare the target module set\n",
    "posL = ['Prot_Acar-vs-Cont']\n",
    "negL = ['Tran_Acar-vs-Cont']\n",
    "regulation = 'Tightened'\n",
    "tempS = pd.Series(np.repeat(True, len(pvalDF)), index=pvalDF.index)#Initialize\n",
    "tempDF = pvalDF.loc[:, pvalDF.columns.str.contains('-vs-')]#Based on adjusted ANOVA P-values\n",
    "for col_n in tempDF.columns.tolist():\n",
    "    tempS1 = pvalDF[col_n]\n",
    "    tempS2 = diffDF[col_n]\n",
    "    if col_n in posL:\n",
    "        if regulation=='Tightened':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2>0)\n",
    "        elif regulation=='Loosened':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2<0)\n",
    "    elif col_n in negL:\n",
    "        tempS3 = (tempS1>=0.05)\n",
    "        #Significance for inverse regulation\n",
    "        if regulation=='Tightened':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2<0)\n",
    "        elif regulation=='Loosened':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2>0)\n",
    "        tempS1 = tempS3 | tempS1\n",
    "    else:\n",
    "        tempS1 = (tempS1>=0.0)\n",
    "    #Update True\n",
    "    tempS = tempS & tempS1\n",
    "tempL = tempS.loc[tempS.tolist()].index.tolist()\n",
    "print(len(tempL), regulation.lower()+' modules with significance in', posL, 'but not in', negL)\n",
    "\n",
    "#Select representatives\n",
    "topX = np.min([30, len(tempL)])\n",
    "topX_plot = np.min([10, len(tempL)])\n",
    "##Spread statDF by dataset to flatten multi-index\n",
    "tempDF = moduleDF['ModuleName']#pd.Series() for now\n",
    "for dataset in ['Proteomics', 'Transcriptomics']:\n",
    "    tempDF1 = statDF.loc[dataset].drop(columns=['ModuleName'])\n",
    "    tempDF1.columns = dataset[0:4]+'_'+tempDF1.columns\n",
    "    tempDF = pd.merge(tempDF, tempDF1, left_index=True, right_index=True, how='left')\n",
    "tempDF = tempDF.loc[:, tempDF.columns.str.contains('Pval')]\n",
    "tempDF = pd.merge(moduleDF['ModuleName'], tempDF, left_index=True, right_index=True, how='left')\n",
    "tempDF = tempDF.loc[tempL]\n",
    "tempDF = tempDF.sort_values(by=['Prot_'+variable+'_AdjPval', 'Tran_'+variable+'_AdjPval'], ascending=True)\n",
    "print('Top', topX, 'modules (sort by the main effect of '+variable+'):')\n",
    "display(tempDF.iloc[:topX])\n",
    "plotL = tempDF.index.tolist()[:topX_plot]\n",
    "\n",
    "#Prepare DF for plot\n",
    "tempDF = rmsDF_kk.reset_index().melt(var_name='SampleID', value_name='RMS', id_vars='ModuleID')\n",
    "tempDF1 = sampleDF.reset_index()[['SampleID', 'Dataset', 'Phenotype']]\n",
    "tempDF = pd.merge(tempDF, tempDF1, on='SampleID', how='left')\n",
    "\n",
    "#Prepare label and color\n",
    "tempD = {'Proteomics':'Proteins', 'Transcriptomics':'Transcripts'}\n",
    "tempDF['Dataset'] = tempDF['Dataset'].map(tempD)\n",
    "tempDF['Group'] = tempDF['Phenotype']\n",
    "tempD0 = {'Cont':'Control', 'Acar':'Acarbose', 'Rapa':'Rapamycin', 'Prot':'Proteins', 'Tran':'Transcripts'}\n",
    "tempD1 = {'Control':'tab:blue', 'Acarbose':'tab:red', 'Rapamycin':'tab:purple'}\n",
    "tempD2 = {'Proteins':plt.get_cmap('tab20')(13), 'Transcripts':plt.get_cmap('tab20')(19)}\n",
    "\n",
    "#Visualize each representative\n",
    "for rank_i in range(len(plotL)):\n",
    "    print(' - Rank '+str(rank_i+1)+' (sort by the main effect of '+variable+'):')\n",
    "    module = plotL[rank_i]\n",
    "    #Check module summary\n",
    "    tempDF1 = pd.DataFrame(moduleDF.loc[module]).T\n",
    "    display(tempDF1)\n",
    "    \n",
    "    #Select RMS\n",
    "    tempDF1 = tempDF.loc[tempDF['ModuleID']==module]\n",
    "    \n",
    "    #Check RMS summary\n",
    "    tempDF2 = tempDF1.groupby(['Dataset', 'Group'])['RMS'].agg(['count', 'mean', 'std'])\n",
    "    tempL1 = []\n",
    "    tempL2 = []\n",
    "    for row_n in tempDF2.index.tolist():\n",
    "        count, mean, std = tempDF2.loc[row_n]\n",
    "        tempL1.append(mean - 1.96*std/np.sqrt(count))\n",
    "        tempL2.append(mean + 1.96*std/np.sqrt(count))\n",
    "    tempDF2['0.025'] = tempL1\n",
    "    tempDF2['0.975'] = tempL2\n",
    "    ##Multiindex sort\n",
    "    tempDF2 = tempDF2.reset_index()\n",
    "    tempDF2['Dataset'] = pd.Categorical(tempDF2['Dataset'], categories=list(tempD2.keys()))\n",
    "    tempDF2['Group'] = pd.Categorical(tempDF2['Group'], categories=list(tempD1.keys()))\n",
    "    tempDF2 = tempDF2.sort_values(by=['Dataset', 'Group']).set_index(['Dataset', 'Group'])\n",
    "    display(tempDF2)\n",
    "    \n",
    "    #Prepare significance labels\n",
    "    ##Retrieve statistical significance\n",
    "    tempS = pvalDF.loc[module, pvalDF.columns.str.contains('-vs-')]\n",
    "    tempS.name = 'AdjPval'\n",
    "    ##Clean\n",
    "    tempDF2 = tempS.index.to_series().str.split(pat='_', expand=True)\n",
    "    tempDF2 = tempDF2.rename(columns={0:'Dataset', 1:'Comparison'})\n",
    "    tempS1 = tempDF2['Dataset']\n",
    "    tempDF2 = tempDF2['Comparison'].str.split(pat='-vs-', expand=True)\n",
    "    tempDF2 = tempDF2.rename(columns={0:'Contrast', 1:'Baseline'})\n",
    "    tempDF2 = pd.merge(tempS1, tempDF2, left_index=True, right_index=True, how='left')\n",
    "    tempDF2 = pd.merge(tempDF2, tempS, left_index=True, right_index=True, how='left')\n",
    "    tempDF2['Dataset'] = tempDF2['Dataset'].map(tempD0)\n",
    "    tempDF2['Contrast'] = tempDF2['Contrast'].map(tempD0)\n",
    "    tempDF2['Baseline'] = tempDF2['Baseline'].map(tempD0)\n",
    "    ##Convert p-value to label\n",
    "    tempL = []\n",
    "    for row_i in range(len(tempDF2)):\n",
    "        pval = tempDF2['AdjPval'].iloc[row_i]\n",
    "        if pval<0.001:\n",
    "            tempL.append('***')\n",
    "        elif pval<0.01:\n",
    "            tempL.append('**')\n",
    "        elif pval<0.05:\n",
    "            tempL.append('*')\n",
    "        else:\n",
    "            pval_text = Decimal(str(pval)).quantize(Decimal('0.01'), rounding=ROUND_HALF_UP)\n",
    "            tempL.append(r'$P$ = '+str(pval_text))\n",
    "    tempDF2['SignifLabel'] = tempL\n",
    "    display(tempDF2)\n",
    "    \n",
    "    #Visualization\n",
    "    ymax = 1.0\n",
    "    ymin = 0.0\n",
    "    yinter = 0.2\n",
    "    ymargin_t = 0.275\n",
    "    ymargin_b = 0.05\n",
    "    aline_ymin = 1.0\n",
    "    aline_ymargin = 0.125\n",
    "    sns.set(style='ticks', font='Arial', context='talk')\n",
    "    fig, axes = plt.subplots(nrows=1, ncols=2, figsize=(4.75, 5), sharex=True, sharey=True,\n",
    "                             gridspec_kw={'width_ratios':[2, 2]})\n",
    "    for ax_i, ax in enumerate(axes.flat):\n",
    "        dataset = list(tempD2.keys())[ax_i]\n",
    "        tempDF3 = tempDF1.loc[tempDF1['Dataset']==dataset]\n",
    "        sns.pointplot(data=tempDF3, x='Group', y='RMS', order=list(tempD1.keys()), palette=tempD1,\n",
    "                      markers='o', dodge=False, join=False, capsize=0.6, estimator=np.mean, ci=95, ax=ax)\n",
    "        sns.stripplot(data=tempDF3, x='Group', y='RMS',\n",
    "                      order=list(tempD1.keys()), palette=tempD1, dodge=False, jitter=0.15,\n",
    "                      size=5, edgecolor='black', linewidth=1, **{'marker':'o', 'alpha':0.5}, ax=ax)\n",
    "        #Set axis\n",
    "        sns.despine()\n",
    "        ax.set(ylim=(ymin-ymargin_b, ymax+ymargin_t), yticks=np.arange(ymin, ymax + yinter/10, yinter))\n",
    "        plt.setp(ax.get_xticklabels(), rotation=70, horizontalalignment='right',\n",
    "                 verticalalignment='center', rotation_mode='anchor')\n",
    "        if ax_i==0:\n",
    "            plt.setp(ax, xlabel='', ylabel='Sample RMS\\n(Mean = Module RCI)')\n",
    "        else:\n",
    "            plt.setp(ax.get_yticklabels(), visible=False)\n",
    "            plt.setp(ax, xlabel='', ylabel='')\n",
    "        #Add significance labels\n",
    "        tempDF3 = tempDF2.loc[tempDF2['Dataset']==dataset]\n",
    "        for row_i in range(len(tempDF3)):\n",
    "            #Baseline\n",
    "            group_0 = tempDF3['Baseline'].iloc[row_i]\n",
    "            index_0 = list(tempD1.keys()).index(group_0)\n",
    "            xcoord_0 = index_0\n",
    "            #Contrast\n",
    "            group_1 = tempDF3['Contrast'].iloc[row_i]\n",
    "            index_1 = list(tempD1.keys()).index(group_1)\n",
    "            xcoord_1 = index_1\n",
    "            #Standard point of marker\n",
    "            xcoord = (xcoord_0+xcoord_1)/2\n",
    "            ycoord = aline_ymin + aline_ymargin*row_i\n",
    "            label = tempDF3['SignifLabel'].iloc[row_i]\n",
    "            #Add annotation lines\n",
    "            aline_offset = yinter/5\n",
    "            aline_length = yinter/5 + aline_offset/2\n",
    "            ax.plot([xcoord_0, xcoord_0, xcoord_1, xcoord_1],\n",
    "                    [ycoord+aline_offset, ycoord+aline_length, ycoord+aline_length, ycoord+aline_offset],\n",
    "                    lw=1.5, c='k')\n",
    "            #Add annotation text\n",
    "            if label in ['***', '**', '*']:\n",
    "                text_offset = yinter/21\n",
    "                ax.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                            horizontalalignment='center', verticalalignment='bottom',\n",
    "                            fontsize='medium', color='k')\n",
    "            else:\n",
    "                text_offset = yinter/3.5\n",
    "                ax.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                            horizontalalignment='center', verticalalignment='bottom',\n",
    "                            fontsize='x-small', color='k')\n",
    "        #Add annotation\n",
    "        ax.set_title(dataset, {'fontsize':'small'})\n",
    "        xoff = 0.025\n",
    "        yoff = 0.01\n",
    "        rect = plt.Rectangle((xoff, 1), 1-xoff, 0.1,#Manual adjustment\n",
    "                             transform=ax.transAxes, facecolor=tempD2[dataset],\n",
    "                             clip_on=False, linewidth=0, zorder=0.5)\n",
    "        ax.add_patch(rect)\n",
    "        if ax_i==1:\n",
    "            at = AnchoredText('Consensus: own group', loc='center',\n",
    "                              bbox_to_anchor=(1.15, 0.505), bbox_transform=ax.transAxes,#Minor manual adjustment\n",
    "                              frameon=False, prop={'size':'small', 'rotation':90})\n",
    "            ax.add_artist(at)\n",
    "            rect = plt.Rectangle((1+xoff, yoff), 0.225, 1-2*yoff,#Manual adjustment\n",
    "                                 transform=ax.transAxes, facecolor=plt.get_cmap('tab20')(15),\n",
    "                                 clip_on=False, linewidth=0, zorder=0.5)\n",
    "            ax.add_patch(rect)\n",
    "    fig.tight_layout()\n",
    "    #Set title\n",
    "    modulename = moduleDF.loc[module, 'ModuleName']\n",
    "    initial = modulename[0].capitalize()\n",
    "    title = re.sub('^.', initial, modulename)+' ('+module+')'\n",
    "    title = '\\n'.join(wrap(title, 40))#Because the below wrap=True didn't work\n",
    "    fig.suptitle(title, size='small',\n",
    "                 verticalalignment='bottom', horizontalalignment='center', wrap=True, y=0.93)\n",
    "    ##Save\n",
    "    fileDir = './ExportFigures/'\n",
    "    ipynbName = '220525_LC-M001-DIRAC-prot-vs-txn_Comparison-GOBP_ver2-2_'\n",
    "    fileName = 'RCI-inter-group-comparison_RMS-pointplot-'+module.replace('GO:', 'GO')+'.tif'\n",
    "    plt.gcf().savefig(fileDir+ipynbName+fileName, dpi=300, bbox_inches='tight', pad_inches=0.04,\n",
    "                      pil_kwargs={'compression':'tiff_lzw'})\n",
    "    plt.show()\n",
    "    print('')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 3-5-6. Modules tightened by Rapa in both datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Prepare the target module set\n",
    "posL = ['Prot_Rapa-vs-Cont', 'Tran_Rapa-vs-Cont']\n",
    "negL = []\n",
    "regulation = 'Tightened'\n",
    "tempS = pd.Series(np.repeat(True, len(pvalDF)), index=pvalDF.index)#Initialize\n",
    "tempDF = pvalDF.loc[:, pvalDF.columns.str.contains('-vs-')]#Based on adjusted ANOVA P-values\n",
    "for col_n in tempDF.columns.tolist():\n",
    "    tempS1 = pvalDF[col_n]\n",
    "    tempS2 = diffDF[col_n]\n",
    "    if col_n in posL:\n",
    "        if regulation=='Tightened':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2>0)\n",
    "        elif regulation=='Loosened':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2<0)\n",
    "    elif col_n in negL:\n",
    "        tempS3 = (tempS1>=0.05)\n",
    "        #Significance for inverse regulation\n",
    "        if regulation=='Tightened':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2<0)\n",
    "        elif regulation=='Loosened':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2>0)\n",
    "        tempS1 = tempS3 | tempS1\n",
    "    else:\n",
    "        tempS1 = (tempS1>=0.0)\n",
    "    #Update True\n",
    "    tempS = tempS & tempS1\n",
    "tempL = tempS.loc[tempS.tolist()].index.tolist()\n",
    "print(len(tempL), regulation.lower()+' modules with significance in', posL, 'but not in', negL)\n",
    "\n",
    "#Select representatives\n",
    "topX = np.min([30, len(tempL)])\n",
    "topX_plot = np.min([10, len(tempL)])\n",
    "##Spread statDF by dataset to flatten multi-index\n",
    "tempDF = moduleDF['ModuleName']#pd.Series() for now\n",
    "for dataset in ['Proteomics', 'Transcriptomics']:\n",
    "    tempDF1 = statDF.loc[dataset].drop(columns=['ModuleName'])\n",
    "    tempDF1.columns = dataset[0:4]+'_'+tempDF1.columns\n",
    "    tempDF = pd.merge(tempDF, tempDF1, left_index=True, right_index=True, how='left')\n",
    "tempDF = tempDF.loc[:, tempDF.columns.str.contains('Pval')]\n",
    "tempDF = pd.merge(moduleDF['ModuleName'], tempDF, left_index=True, right_index=True, how='left')\n",
    "tempDF = tempDF.loc[tempL]\n",
    "tempDF = tempDF.sort_values(by=['Prot_'+variable+'_AdjPval', 'Tran_'+variable+'_AdjPval'], ascending=True)\n",
    "print('Top', topX, 'modules (sort by the main effect of '+variable+'):')\n",
    "display(tempDF.iloc[:topX])\n",
    "plotL = tempDF.index.tolist()[:topX_plot]\n",
    "\n",
    "#Prepare DF for plot\n",
    "tempDF = rmsDF_kk.reset_index().melt(var_name='SampleID', value_name='RMS', id_vars='ModuleID')\n",
    "tempDF1 = sampleDF.reset_index()[['SampleID', 'Dataset', 'Phenotype']]\n",
    "tempDF = pd.merge(tempDF, tempDF1, on='SampleID', how='left')\n",
    "\n",
    "#Prepare label and color\n",
    "tempD = {'Proteomics':'Proteins', 'Transcriptomics':'Transcripts'}\n",
    "tempDF['Dataset'] = tempDF['Dataset'].map(tempD)\n",
    "tempDF['Group'] = tempDF['Phenotype']\n",
    "tempD0 = {'Cont':'Control', 'Acar':'Acarbose', 'Rapa':'Rapamycin', 'Prot':'Proteins', 'Tran':'Transcripts'}\n",
    "tempD1 = {'Control':'tab:blue', 'Acarbose':'tab:red', 'Rapamycin':'tab:purple'}\n",
    "tempD2 = {'Proteins':plt.get_cmap('tab20')(13), 'Transcripts':plt.get_cmap('tab20')(19)}\n",
    "\n",
    "#Visualize each representative\n",
    "for rank_i in range(len(plotL)):\n",
    "    print(' - Rank '+str(rank_i+1)+' (sort by the main effect of '+variable+'):')\n",
    "    module = plotL[rank_i]\n",
    "    #Check module summary\n",
    "    tempDF1 = pd.DataFrame(moduleDF.loc[module]).T\n",
    "    display(tempDF1)\n",
    "    \n",
    "    #Select RMS\n",
    "    tempDF1 = tempDF.loc[tempDF['ModuleID']==module]\n",
    "    \n",
    "    #Check RMS summary\n",
    "    tempDF2 = tempDF1.groupby(['Dataset', 'Group'])['RMS'].agg(['count', 'mean', 'std'])\n",
    "    tempL1 = []\n",
    "    tempL2 = []\n",
    "    for row_n in tempDF2.index.tolist():\n",
    "        count, mean, std = tempDF2.loc[row_n]\n",
    "        tempL1.append(mean - 1.96*std/np.sqrt(count))\n",
    "        tempL2.append(mean + 1.96*std/np.sqrt(count))\n",
    "    tempDF2['0.025'] = tempL1\n",
    "    tempDF2['0.975'] = tempL2\n",
    "    ##Multiindex sort\n",
    "    tempDF2 = tempDF2.reset_index()\n",
    "    tempDF2['Dataset'] = pd.Categorical(tempDF2['Dataset'], categories=list(tempD2.keys()))\n",
    "    tempDF2['Group'] = pd.Categorical(tempDF2['Group'], categories=list(tempD1.keys()))\n",
    "    tempDF2 = tempDF2.sort_values(by=['Dataset', 'Group']).set_index(['Dataset', 'Group'])\n",
    "    display(tempDF2)\n",
    "    \n",
    "    #Prepare significance labels\n",
    "    ##Retrieve statistical significance\n",
    "    tempS = pvalDF.loc[module, pvalDF.columns.str.contains('-vs-')]\n",
    "    tempS.name = 'AdjPval'\n",
    "    ##Clean\n",
    "    tempDF2 = tempS.index.to_series().str.split(pat='_', expand=True)\n",
    "    tempDF2 = tempDF2.rename(columns={0:'Dataset', 1:'Comparison'})\n",
    "    tempS1 = tempDF2['Dataset']\n",
    "    tempDF2 = tempDF2['Comparison'].str.split(pat='-vs-', expand=True)\n",
    "    tempDF2 = tempDF2.rename(columns={0:'Contrast', 1:'Baseline'})\n",
    "    tempDF2 = pd.merge(tempS1, tempDF2, left_index=True, right_index=True, how='left')\n",
    "    tempDF2 = pd.merge(tempDF2, tempS, left_index=True, right_index=True, how='left')\n",
    "    tempDF2['Dataset'] = tempDF2['Dataset'].map(tempD0)\n",
    "    tempDF2['Contrast'] = tempDF2['Contrast'].map(tempD0)\n",
    "    tempDF2['Baseline'] = tempDF2['Baseline'].map(tempD0)\n",
    "    ##Convert p-value to label\n",
    "    tempL = []\n",
    "    for row_i in range(len(tempDF2)):\n",
    "        pval = tempDF2['AdjPval'].iloc[row_i]\n",
    "        if pval<0.001:\n",
    "            tempL.append('***')\n",
    "        elif pval<0.01:\n",
    "            tempL.append('**')\n",
    "        elif pval<0.05:\n",
    "            tempL.append('*')\n",
    "        else:\n",
    "            pval_text = Decimal(str(pval)).quantize(Decimal('0.01'), rounding=ROUND_HALF_UP)\n",
    "            tempL.append(r'$P$ = '+str(pval_text))\n",
    "    tempDF2['SignifLabel'] = tempL\n",
    "    display(tempDF2)\n",
    "    \n",
    "    #Visualization\n",
    "    ymax = 1.0\n",
    "    ymin = 0.0\n",
    "    yinter = 0.2\n",
    "    ymargin_t = 0.275\n",
    "    ymargin_b = 0.05\n",
    "    aline_ymin = 1.0\n",
    "    aline_ymargin = 0.125\n",
    "    sns.set(style='ticks', font='Arial', context='talk')\n",
    "    fig, axes = plt.subplots(nrows=1, ncols=2, figsize=(4.75, 5), sharex=True, sharey=True,\n",
    "                             gridspec_kw={'width_ratios':[2, 2]})\n",
    "    for ax_i, ax in enumerate(axes.flat):\n",
    "        dataset = list(tempD2.keys())[ax_i]\n",
    "        tempDF3 = tempDF1.loc[tempDF1['Dataset']==dataset]\n",
    "        sns.pointplot(data=tempDF3, x='Group', y='RMS', order=list(tempD1.keys()), palette=tempD1,\n",
    "                      markers='o', dodge=False, join=False, capsize=0.6, estimator=np.mean, ci=95, ax=ax)\n",
    "        sns.stripplot(data=tempDF3, x='Group', y='RMS',\n",
    "                      order=list(tempD1.keys()), palette=tempD1, dodge=False, jitter=0.15,\n",
    "                      size=5, edgecolor='black', linewidth=1, **{'marker':'o', 'alpha':0.5}, ax=ax)\n",
    "        #Set axis\n",
    "        sns.despine()\n",
    "        ax.set(ylim=(ymin-ymargin_b, ymax+ymargin_t), yticks=np.arange(ymin, ymax + yinter/10, yinter))\n",
    "        plt.setp(ax.get_xticklabels(), rotation=70, horizontalalignment='right',\n",
    "                 verticalalignment='center', rotation_mode='anchor')\n",
    "        if ax_i==0:\n",
    "            plt.setp(ax, xlabel='', ylabel='Sample RMS\\n(Mean = Module RCI)')\n",
    "        else:\n",
    "            plt.setp(ax.get_yticklabels(), visible=False)\n",
    "            plt.setp(ax, xlabel='', ylabel='')\n",
    "        #Add significance labels\n",
    "        tempDF3 = tempDF2.loc[tempDF2['Dataset']==dataset]\n",
    "        for row_i in range(len(tempDF3)):\n",
    "            #Baseline\n",
    "            group_0 = tempDF3['Baseline'].iloc[row_i]\n",
    "            index_0 = list(tempD1.keys()).index(group_0)\n",
    "            xcoord_0 = index_0\n",
    "            #Contrast\n",
    "            group_1 = tempDF3['Contrast'].iloc[row_i]\n",
    "            index_1 = list(tempD1.keys()).index(group_1)\n",
    "            xcoord_1 = index_1\n",
    "            #Standard point of marker\n",
    "            xcoord = (xcoord_0+xcoord_1)/2\n",
    "            ycoord = aline_ymin + aline_ymargin*row_i\n",
    "            label = tempDF3['SignifLabel'].iloc[row_i]\n",
    "            #Add annotation lines\n",
    "            aline_offset = yinter/5\n",
    "            aline_length = yinter/5 + aline_offset/2\n",
    "            ax.plot([xcoord_0, xcoord_0, xcoord_1, xcoord_1],\n",
    "                    [ycoord+aline_offset, ycoord+aline_length, ycoord+aline_length, ycoord+aline_offset],\n",
    "                    lw=1.5, c='k')\n",
    "            #Add annotation text\n",
    "            if label in ['***', '**', '*']:\n",
    "                text_offset = yinter/21\n",
    "                ax.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                            horizontalalignment='center', verticalalignment='bottom',\n",
    "                            fontsize='medium', color='k')\n",
    "            else:\n",
    "                text_offset = yinter/3.5\n",
    "                ax.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                            horizontalalignment='center', verticalalignment='bottom',\n",
    "                            fontsize='x-small', color='k')\n",
    "        #Add annotation\n",
    "        ax.set_title(dataset, {'fontsize':'small'})\n",
    "        xoff = 0.025\n",
    "        yoff = 0.01\n",
    "        rect = plt.Rectangle((xoff, 1), 1-xoff, 0.1,#Manual adjustment\n",
    "                             transform=ax.transAxes, facecolor=tempD2[dataset],\n",
    "                             clip_on=False, linewidth=0, zorder=0.5)\n",
    "        ax.add_patch(rect)\n",
    "        if ax_i==1:\n",
    "            at = AnchoredText('Consensus: own group', loc='center',\n",
    "                              bbox_to_anchor=(1.15, 0.505), bbox_transform=ax.transAxes,#Minor manual adjustment\n",
    "                              frameon=False, prop={'size':'small', 'rotation':90})\n",
    "            ax.add_artist(at)\n",
    "            rect = plt.Rectangle((1+xoff, yoff), 0.225, 1-2*yoff,#Manual adjustment\n",
    "                                 transform=ax.transAxes, facecolor=plt.get_cmap('tab20')(15),\n",
    "                                 clip_on=False, linewidth=0, zorder=0.5)\n",
    "            ax.add_patch(rect)\n",
    "    fig.tight_layout()\n",
    "    #Set title\n",
    "    modulename = moduleDF.loc[module, 'ModuleName']\n",
    "    initial = modulename[0].capitalize()\n",
    "    title = re.sub('^.', initial, modulename)+' ('+module+')'\n",
    "    title = '\\n'.join(wrap(title, 40))#Because the below wrap=True didn't work\n",
    "    fig.suptitle(title, size='small',\n",
    "                 verticalalignment='bottom', horizontalalignment='center', wrap=True, y=0.93)\n",
    "    ##Save\n",
    "    fileDir = './ExportFigures/'\n",
    "    ipynbName = '220525_LC-M001-DIRAC-prot-vs-txn_Comparison-GOBP_ver2-2_'\n",
    "    fileName = 'RCI-inter-group-comparison_RMS-pointplot-'+module.replace('GO:', 'GO')+'.tif'\n",
    "    plt.gcf().savefig(fileDir+ipynbName+fileName, dpi=300, bbox_inches='tight', pad_inches=0.04,\n",
    "                      pil_kwargs={'compression':'tiff_lzw'})\n",
    "    plt.show()\n",
    "    print('')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 3-5-7. Modules tightened by Rapa specifically in proteomics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Prepare the target module set\n",
    "posL = ['Prot_Rapa-vs-Cont']\n",
    "negL = ['Tran_Rapa-vs-Cont']\n",
    "regulation = 'Tightened'\n",
    "tempS = pd.Series(np.repeat(True, len(pvalDF)), index=pvalDF.index)#Initialize\n",
    "tempDF = pvalDF.loc[:, pvalDF.columns.str.contains('-vs-')]#Based on adjusted ANOVA P-values\n",
    "for col_n in tempDF.columns.tolist():\n",
    "    tempS1 = pvalDF[col_n]\n",
    "    tempS2 = diffDF[col_n]\n",
    "    if col_n in posL:\n",
    "        if regulation=='Tightened':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2>0)\n",
    "        elif regulation=='Loosened':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2<0)\n",
    "    elif col_n in negL:\n",
    "        tempS3 = (tempS1>=0.05)\n",
    "        #Significance for inverse regulation\n",
    "        if regulation=='Tightened':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2<0)\n",
    "        elif regulation=='Loosened':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2>0)\n",
    "        tempS1 = tempS3 | tempS1\n",
    "    else:\n",
    "        tempS1 = (tempS1>=0.0)\n",
    "    #Update True\n",
    "    tempS = tempS & tempS1\n",
    "tempL = tempS.loc[tempS.tolist()].index.tolist()\n",
    "print(len(tempL), regulation.lower()+' modules with significance in', posL, 'but not in', negL)\n",
    "\n",
    "#Select representatives\n",
    "topX = np.min([30, len(tempL)])\n",
    "topX_plot = np.min([10, len(tempL)])\n",
    "##Spread statDF by dataset to flatten multi-index\n",
    "tempDF = moduleDF['ModuleName']#pd.Series() for now\n",
    "for dataset in ['Proteomics', 'Transcriptomics']:\n",
    "    tempDF1 = statDF.loc[dataset].drop(columns=['ModuleName'])\n",
    "    tempDF1.columns = dataset[0:4]+'_'+tempDF1.columns\n",
    "    tempDF = pd.merge(tempDF, tempDF1, left_index=True, right_index=True, how='left')\n",
    "tempDF = tempDF.loc[:, tempDF.columns.str.contains('Pval')]\n",
    "tempDF = pd.merge(moduleDF['ModuleName'], tempDF, left_index=True, right_index=True, how='left')\n",
    "tempDF = tempDF.loc[tempL]\n",
    "tempDF = tempDF.sort_values(by=['Prot_'+variable+'_AdjPval', 'Tran_'+variable+'_AdjPval'], ascending=True)\n",
    "print('Top', topX, 'modules (sort by the main effect of '+variable+'):')\n",
    "display(tempDF.iloc[:topX])\n",
    "plotL = tempDF.index.tolist()[:topX_plot]\n",
    "\n",
    "#Prepare DF for plot\n",
    "tempDF = rmsDF_kk.reset_index().melt(var_name='SampleID', value_name='RMS', id_vars='ModuleID')\n",
    "tempDF1 = sampleDF.reset_index()[['SampleID', 'Dataset', 'Phenotype']]\n",
    "tempDF = pd.merge(tempDF, tempDF1, on='SampleID', how='left')\n",
    "\n",
    "#Prepare label and color\n",
    "tempD = {'Proteomics':'Proteins', 'Transcriptomics':'Transcripts'}\n",
    "tempDF['Dataset'] = tempDF['Dataset'].map(tempD)\n",
    "tempDF['Group'] = tempDF['Phenotype']\n",
    "tempD0 = {'Cont':'Control', 'Acar':'Acarbose', 'Rapa':'Rapamycin', 'Prot':'Proteins', 'Tran':'Transcripts'}\n",
    "tempD1 = {'Control':'tab:blue', 'Acarbose':'tab:red', 'Rapamycin':'tab:purple'}\n",
    "tempD2 = {'Proteins':plt.get_cmap('tab20')(13), 'Transcripts':plt.get_cmap('tab20')(19)}\n",
    "\n",
    "#Visualize each representative\n",
    "for rank_i in range(len(plotL)):\n",
    "    print(' - Rank '+str(rank_i+1)+' (sort by the main effect of '+variable+'):')\n",
    "    module = plotL[rank_i]\n",
    "    #Check module summary\n",
    "    tempDF1 = pd.DataFrame(moduleDF.loc[module]).T\n",
    "    display(tempDF1)\n",
    "    \n",
    "    #Select RMS\n",
    "    tempDF1 = tempDF.loc[tempDF['ModuleID']==module]\n",
    "    \n",
    "    #Check RMS summary\n",
    "    tempDF2 = tempDF1.groupby(['Dataset', 'Group'])['RMS'].agg(['count', 'mean', 'std'])\n",
    "    tempL1 = []\n",
    "    tempL2 = []\n",
    "    for row_n in tempDF2.index.tolist():\n",
    "        count, mean, std = tempDF2.loc[row_n]\n",
    "        tempL1.append(mean - 1.96*std/np.sqrt(count))\n",
    "        tempL2.append(mean + 1.96*std/np.sqrt(count))\n",
    "    tempDF2['0.025'] = tempL1\n",
    "    tempDF2['0.975'] = tempL2\n",
    "    ##Multiindex sort\n",
    "    tempDF2 = tempDF2.reset_index()\n",
    "    tempDF2['Dataset'] = pd.Categorical(tempDF2['Dataset'], categories=list(tempD2.keys()))\n",
    "    tempDF2['Group'] = pd.Categorical(tempDF2['Group'], categories=list(tempD1.keys()))\n",
    "    tempDF2 = tempDF2.sort_values(by=['Dataset', 'Group']).set_index(['Dataset', 'Group'])\n",
    "    display(tempDF2)\n",
    "    \n",
    "    #Prepare significance labels\n",
    "    ##Retrieve statistical significance\n",
    "    tempS = pvalDF.loc[module, pvalDF.columns.str.contains('-vs-')]\n",
    "    tempS.name = 'AdjPval'\n",
    "    ##Clean\n",
    "    tempDF2 = tempS.index.to_series().str.split(pat='_', expand=True)\n",
    "    tempDF2 = tempDF2.rename(columns={0:'Dataset', 1:'Comparison'})\n",
    "    tempS1 = tempDF2['Dataset']\n",
    "    tempDF2 = tempDF2['Comparison'].str.split(pat='-vs-', expand=True)\n",
    "    tempDF2 = tempDF2.rename(columns={0:'Contrast', 1:'Baseline'})\n",
    "    tempDF2 = pd.merge(tempS1, tempDF2, left_index=True, right_index=True, how='left')\n",
    "    tempDF2 = pd.merge(tempDF2, tempS, left_index=True, right_index=True, how='left')\n",
    "    tempDF2['Dataset'] = tempDF2['Dataset'].map(tempD0)\n",
    "    tempDF2['Contrast'] = tempDF2['Contrast'].map(tempD0)\n",
    "    tempDF2['Baseline'] = tempDF2['Baseline'].map(tempD0)\n",
    "    ##Convert p-value to label\n",
    "    tempL = []\n",
    "    for row_i in range(len(tempDF2)):\n",
    "        pval = tempDF2['AdjPval'].iloc[row_i]\n",
    "        if pval<0.001:\n",
    "            tempL.append('***')\n",
    "        elif pval<0.01:\n",
    "            tempL.append('**')\n",
    "        elif pval<0.05:\n",
    "            tempL.append('*')\n",
    "        else:\n",
    "            pval_text = Decimal(str(pval)).quantize(Decimal('0.01'), rounding=ROUND_HALF_UP)\n",
    "            tempL.append(r'$P$ = '+str(pval_text))\n",
    "    tempDF2['SignifLabel'] = tempL\n",
    "    display(tempDF2)\n",
    "    \n",
    "    #Visualization\n",
    "    ymax = 1.0\n",
    "    ymin = 0.0\n",
    "    yinter = 0.2\n",
    "    ymargin_t = 0.275\n",
    "    ymargin_b = 0.05\n",
    "    aline_ymin = 1.0\n",
    "    aline_ymargin = 0.125\n",
    "    sns.set(style='ticks', font='Arial', context='talk')\n",
    "    fig, axes = plt.subplots(nrows=1, ncols=2, figsize=(4.75, 5), sharex=True, sharey=True,\n",
    "                             gridspec_kw={'width_ratios':[2, 2]})\n",
    "    for ax_i, ax in enumerate(axes.flat):\n",
    "        dataset = list(tempD2.keys())[ax_i]\n",
    "        tempDF3 = tempDF1.loc[tempDF1['Dataset']==dataset]\n",
    "        sns.pointplot(data=tempDF3, x='Group', y='RMS', order=list(tempD1.keys()), palette=tempD1,\n",
    "                      markers='o', dodge=False, join=False, capsize=0.6, estimator=np.mean, ci=95, ax=ax)\n",
    "        sns.stripplot(data=tempDF3, x='Group', y='RMS',\n",
    "                      order=list(tempD1.keys()), palette=tempD1, dodge=False, jitter=0.15,\n",
    "                      size=5, edgecolor='black', linewidth=1, **{'marker':'o', 'alpha':0.5}, ax=ax)\n",
    "        #Set axis\n",
    "        sns.despine()\n",
    "        ax.set(ylim=(ymin-ymargin_b, ymax+ymargin_t), yticks=np.arange(ymin, ymax + yinter/10, yinter))\n",
    "        plt.setp(ax.get_xticklabels(), rotation=70, horizontalalignment='right',\n",
    "                 verticalalignment='center', rotation_mode='anchor')\n",
    "        if ax_i==0:\n",
    "            plt.setp(ax, xlabel='', ylabel='Sample RMS\\n(Mean = Module RCI)')\n",
    "        else:\n",
    "            plt.setp(ax.get_yticklabels(), visible=False)\n",
    "            plt.setp(ax, xlabel='', ylabel='')\n",
    "        #Add significance labels\n",
    "        tempDF3 = tempDF2.loc[tempDF2['Dataset']==dataset]\n",
    "        for row_i in range(len(tempDF3)):\n",
    "            #Baseline\n",
    "            group_0 = tempDF3['Baseline'].iloc[row_i]\n",
    "            index_0 = list(tempD1.keys()).index(group_0)\n",
    "            xcoord_0 = index_0\n",
    "            #Contrast\n",
    "            group_1 = tempDF3['Contrast'].iloc[row_i]\n",
    "            index_1 = list(tempD1.keys()).index(group_1)\n",
    "            xcoord_1 = index_1\n",
    "            #Standard point of marker\n",
    "            xcoord = (xcoord_0+xcoord_1)/2\n",
    "            ycoord = aline_ymin + aline_ymargin*row_i\n",
    "            label = tempDF3['SignifLabel'].iloc[row_i]\n",
    "            #Add annotation lines\n",
    "            aline_offset = yinter/5\n",
    "            aline_length = yinter/5 + aline_offset/2\n",
    "            ax.plot([xcoord_0, xcoord_0, xcoord_1, xcoord_1],\n",
    "                    [ycoord+aline_offset, ycoord+aline_length, ycoord+aline_length, ycoord+aline_offset],\n",
    "                    lw=1.5, c='k')\n",
    "            #Add annotation text\n",
    "            if label in ['***', '**', '*']:\n",
    "                text_offset = yinter/21\n",
    "                ax.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                            horizontalalignment='center', verticalalignment='bottom',\n",
    "                            fontsize='medium', color='k')\n",
    "            else:\n",
    "                text_offset = yinter/3.5\n",
    "                ax.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                            horizontalalignment='center', verticalalignment='bottom',\n",
    "                            fontsize='x-small', color='k')\n",
    "        #Add annotation\n",
    "        ax.set_title(dataset, {'fontsize':'small'})\n",
    "        xoff = 0.025\n",
    "        yoff = 0.01\n",
    "        rect = plt.Rectangle((xoff, 1), 1-xoff, 0.1,#Manual adjustment\n",
    "                             transform=ax.transAxes, facecolor=tempD2[dataset],\n",
    "                             clip_on=False, linewidth=0, zorder=0.5)\n",
    "        ax.add_patch(rect)\n",
    "        if ax_i==1:\n",
    "            at = AnchoredText('Consensus: own group', loc='center',\n",
    "                              bbox_to_anchor=(1.15, 0.505), bbox_transform=ax.transAxes,#Minor manual adjustment\n",
    "                              frameon=False, prop={'size':'small', 'rotation':90})\n",
    "            ax.add_artist(at)\n",
    "            rect = plt.Rectangle((1+xoff, yoff), 0.225, 1-2*yoff,#Manual adjustment\n",
    "                                 transform=ax.transAxes, facecolor=plt.get_cmap('tab20')(15),\n",
    "                                 clip_on=False, linewidth=0, zorder=0.5)\n",
    "            ax.add_patch(rect)\n",
    "    fig.tight_layout()\n",
    "    #Set title\n",
    "    modulename = moduleDF.loc[module, 'ModuleName']\n",
    "    initial = modulename[0].capitalize()\n",
    "    title = re.sub('^.', initial, modulename)+' ('+module+')'\n",
    "    title = '\\n'.join(wrap(title, 40))#Because the below wrap=True didn't work\n",
    "    fig.suptitle(title, size='small',\n",
    "                 verticalalignment='bottom', horizontalalignment='center', wrap=True, y=0.93)\n",
    "    ##Save\n",
    "    fileDir = './ExportFigures/'\n",
    "    ipynbName = '220525_LC-M001-DIRAC-prot-vs-txn_Comparison-GOBP_ver2-2_'\n",
    "    fileName = 'RCI-inter-group-comparison_RMS-pointplot-'+module.replace('GO:', 'GO')+'.tif'\n",
    "    plt.gcf().savefig(fileDir+ipynbName+fileName, dpi=300, bbox_inches='tight', pad_inches=0.04,\n",
    "                      pil_kwargs={'compression':'tiff_lzw'})\n",
    "    plt.show()\n",
    "    print('')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 3-5-8. Modules of special interest"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plotL = ['GO:0006635', 'GO:0031998', 'GO:0016558', 'GO:0006625',#Common in M001\n",
    "         'GO:0006637', 'GO:0006734', 'GO:0001561',#Related to fatty acid oxidation (mentioned in previous manuscript)\n",
    "         'GO:0023035', 'GO:0039529',#Immune response (mentioned in previous manuscript)\n",
    "         'GO:0010638', 'GO:0002181', 'GO:0045899',#Mentioned in previous manuscript\n",
    "         'GO:0006635', 'GO:1990126', 'GO:0098761',#Common in proteomics and transcriptomics\n",
    "         'GO:0061732', 'GO:0006086', 'GO:0019441',#Specific to proteomics (Aca)\n",
    "         'GO:0031998', 'GO:0044794', 'GO:0006625', 'GO:0016558', 'GO:0019441', 'GO:0033572',#Specific to proteomics (Rapa)\n",
    "         'GO:0006635', 'GO:0002181', 'GO:0000028', 'GO:0016558', 'GO:0006734', 'GO:0098761',#Common in M001 + M004\n",
    "         'GO:0015986', 'GO:0042776', 'GO:0070934', 'GO:0006703', 'GO:0034354', 'GO:0045899']#Similar to 4EGI-1\n",
    "plotL = list(set(plotL))\n",
    "plotL.sort()\n",
    "for module in plotL:\n",
    "    if module not in moduleDF.index.tolist():\n",
    "        print(module+' was NOT included in this analysis.')\n",
    "        plotL.remove(module)\n",
    "\n",
    "#Prepare the target module set\n",
    "##Spread statDF by dataset to flatten multi-index\n",
    "tempDF = moduleDF['ModuleName']#pd.Series() for now\n",
    "for dataset in ['Proteomics', 'Transcriptomics']:\n",
    "    tempDF1 = statDF.loc[dataset].drop(columns=['ModuleName'])\n",
    "    tempDF1.columns = dataset[0:4]+'_'+tempDF1.columns\n",
    "    tempDF = pd.merge(tempDF, tempDF1, left_index=True, right_index=True, how='left')\n",
    "tempDF = tempDF.loc[:, tempDF.columns.str.contains('Pval')]\n",
    "tempDF = pd.merge(moduleDF['ModuleName'], tempDF, left_index=True, right_index=True, how='left')\n",
    "tempDF = tempDF.loc[plotL]\n",
    "display(tempDF)\n",
    "\n",
    "#Prepare DF for plot\n",
    "tempDF = rmsDF_kk.reset_index().melt(var_name='SampleID', value_name='RMS', id_vars='ModuleID')\n",
    "tempDF1 = sampleDF.reset_index()[['SampleID', 'Dataset', 'Phenotype']]\n",
    "tempDF = pd.merge(tempDF, tempDF1, on='SampleID', how='left')\n",
    "\n",
    "#Prepare label and color\n",
    "tempD = {'Proteomics':'Proteins', 'Transcriptomics':'Transcripts'}\n",
    "tempDF['Dataset'] = tempDF['Dataset'].map(tempD)\n",
    "tempDF['Group'] = tempDF['Phenotype']\n",
    "tempD0 = {'Cont':'Control', 'Acar':'Acarbose', 'Rapa':'Rapamycin', 'Prot':'Proteins', 'Tran':'Transcripts'}\n",
    "tempD1 = {'Control':'tab:blue', 'Acarbose':'tab:red', 'Rapamycin':'tab:purple'}\n",
    "tempD2 = {'Proteins':plt.get_cmap('tab20')(13), 'Transcripts':plt.get_cmap('tab20')(19)}\n",
    "\n",
    "#Visualize each representative\n",
    "for rank_i in range(len(plotL)):\n",
    "    print(' - Module '+str(rank_i+1)+':')\n",
    "    module = plotL[rank_i]\n",
    "    #Check module summary\n",
    "    tempDF1 = pd.DataFrame(moduleDF.loc[module]).T\n",
    "    display(tempDF1)\n",
    "    \n",
    "    #Select RMS\n",
    "    tempDF1 = tempDF.loc[tempDF['ModuleID']==module]\n",
    "    \n",
    "    #Check RMS summary\n",
    "    tempDF2 = tempDF1.groupby(['Dataset', 'Group'])['RMS'].agg(['count', 'mean', 'std'])\n",
    "    tempL1 = []\n",
    "    tempL2 = []\n",
    "    for row_n in tempDF2.index.tolist():\n",
    "        count, mean, std = tempDF2.loc[row_n]\n",
    "        tempL1.append(mean - 1.96*std/np.sqrt(count))\n",
    "        tempL2.append(mean + 1.96*std/np.sqrt(count))\n",
    "    tempDF2['0.025'] = tempL1\n",
    "    tempDF2['0.975'] = tempL2\n",
    "    ##Multiindex sort\n",
    "    tempDF2 = tempDF2.reset_index()\n",
    "    tempDF2['Dataset'] = pd.Categorical(tempDF2['Dataset'], categories=list(tempD2.keys()))\n",
    "    tempDF2['Group'] = pd.Categorical(tempDF2['Group'], categories=list(tempD1.keys()))\n",
    "    tempDF2 = tempDF2.sort_values(by=['Dataset', 'Group']).set_index(['Dataset', 'Group'])\n",
    "    display(tempDF2)\n",
    "    \n",
    "    #Prepare significance labels\n",
    "    ##Re-prepare p-values because these modules can be filtered out in pvalDF\n",
    "    tempDF2 = moduleDF['ModuleName']#pd.Series() for now\n",
    "    for dataset in ['Proteomics', 'Transcriptomics']:\n",
    "        tempDF3 = statDF.loc[dataset].drop(columns=['ModuleName'])\n",
    "        tempDF3.columns = dataset[0:4]+'_'+tempDF3.columns\n",
    "        tempDF2 = pd.merge(tempDF2, tempDF3, left_index=True, right_index=True, how='left')\n",
    "    tempDF2 = tempDF2.loc[:, tempDF2.columns.str.contains('-vs-.*_AdjPval$')]\n",
    "    tempDF2.columns = tempDF2.columns.str.replace('_AdjPval$', '')\n",
    "    ##Retrieve statistical significance\n",
    "    tempS = tempDF2.loc[module]\n",
    "    tempS.name = 'AdjPval'\n",
    "    ##Clean\n",
    "    tempDF2 = tempS.index.to_series().str.split(pat='_', expand=True)\n",
    "    tempDF2 = tempDF2.rename(columns={0:'Dataset', 1:'Comparison'})\n",
    "    tempS1 = tempDF2['Dataset']\n",
    "    tempDF2 = tempDF2['Comparison'].str.split(pat='-vs-', expand=True)\n",
    "    tempDF2 = tempDF2.rename(columns={0:'Contrast', 1:'Baseline'})\n",
    "    tempDF2 = pd.merge(tempS1, tempDF2, left_index=True, right_index=True, how='left')\n",
    "    tempDF2 = pd.merge(tempDF2, tempS, left_index=True, right_index=True, how='left')\n",
    "    tempDF2['Dataset'] = tempDF2['Dataset'].map(tempD0)\n",
    "    tempDF2['Contrast'] = tempDF2['Contrast'].map(tempD0)\n",
    "    tempDF2['Baseline'] = tempDF2['Baseline'].map(tempD0)\n",
    "    ##Convert p-value to label\n",
    "    tempL = []\n",
    "    for row_i in range(len(tempDF2)):\n",
    "        pval = tempDF2['AdjPval'].iloc[row_i]\n",
    "        if pval<0.001:\n",
    "            tempL.append('***')\n",
    "        elif pval<0.01:\n",
    "            tempL.append('**')\n",
    "        elif pval<0.05:\n",
    "            tempL.append('*')\n",
    "        else:\n",
    "            pval_text = Decimal(str(pval)).quantize(Decimal('0.01'), rounding=ROUND_HALF_UP)\n",
    "            tempL.append(r'$P$ = '+str(pval_text))\n",
    "    tempDF2['SignifLabel'] = tempL\n",
    "    display(tempDF2)\n",
    "    \n",
    "    #Visualization\n",
    "    ymax = 1.0\n",
    "    ymin = 0.0\n",
    "    yinter = 0.2\n",
    "    ymargin_t = 0.275\n",
    "    ymargin_b = 0.05\n",
    "    aline_ymin = 1.0\n",
    "    aline_ymargin = 0.125\n",
    "    sns.set(style='ticks', font='Arial', context='talk')\n",
    "    fig, axes = plt.subplots(nrows=1, ncols=2, figsize=(4.75, 5), sharex=True, sharey=True,\n",
    "                             gridspec_kw={'width_ratios':[2, 2]})\n",
    "    for ax_i, ax in enumerate(axes.flat):\n",
    "        dataset = list(tempD2.keys())[ax_i]\n",
    "        tempDF3 = tempDF1.loc[tempDF1['Dataset']==dataset]\n",
    "        sns.pointplot(data=tempDF3, x='Group', y='RMS', order=list(tempD1.keys()), palette=tempD1,\n",
    "                      markers='o', dodge=False, join=False, capsize=0.6, estimator=np.mean, ci=95, ax=ax)\n",
    "        sns.stripplot(data=tempDF3, x='Group', y='RMS',\n",
    "                      order=list(tempD1.keys()), palette=tempD1, dodge=False, jitter=0.15,\n",
    "                      size=5, edgecolor='black', linewidth=1, **{'marker':'o', 'alpha':0.5}, ax=ax)\n",
    "        #Set axis\n",
    "        sns.despine()\n",
    "        ax.set(ylim=(ymin-ymargin_b, ymax+ymargin_t), yticks=np.arange(ymin, ymax + yinter/10, yinter))\n",
    "        plt.setp(ax.get_xticklabels(), rotation=70, horizontalalignment='right',\n",
    "                 verticalalignment='center', rotation_mode='anchor')\n",
    "        if ax_i==0:\n",
    "            plt.setp(ax, xlabel='', ylabel='Sample RMS\\n(Mean = Module RCI)')\n",
    "        else:\n",
    "            plt.setp(ax.get_yticklabels(), visible=False)\n",
    "            plt.setp(ax, xlabel='', ylabel='')\n",
    "        #Add significance labels\n",
    "        tempDF3 = tempDF2.loc[tempDF2['Dataset']==dataset]\n",
    "        for row_i in range(len(tempDF3)):\n",
    "            #Baseline\n",
    "            group_0 = tempDF3['Baseline'].iloc[row_i]\n",
    "            index_0 = list(tempD1.keys()).index(group_0)\n",
    "            xcoord_0 = index_0\n",
    "            #Contrast\n",
    "            group_1 = tempDF3['Contrast'].iloc[row_i]\n",
    "            index_1 = list(tempD1.keys()).index(group_1)\n",
    "            xcoord_1 = index_1\n",
    "            #Standard point of marker\n",
    "            xcoord = (xcoord_0+xcoord_1)/2\n",
    "            ycoord = aline_ymin + aline_ymargin*row_i\n",
    "            label = tempDF3['SignifLabel'].iloc[row_i]\n",
    "            #Add annotation lines\n",
    "            aline_offset = yinter/5\n",
    "            aline_length = yinter/5 + aline_offset/2\n",
    "            ax.plot([xcoord_0, xcoord_0, xcoord_1, xcoord_1],\n",
    "                    [ycoord+aline_offset, ycoord+aline_length, ycoord+aline_length, ycoord+aline_offset],\n",
    "                    lw=1.5, c='k')\n",
    "            #Add annotation text\n",
    "            if label in ['***', '**', '*']:\n",
    "                text_offset = yinter/21\n",
    "                ax.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                            horizontalalignment='center', verticalalignment='bottom',\n",
    "                            fontsize='medium', color='k')\n",
    "            else:\n",
    "                text_offset = yinter/3.5\n",
    "                ax.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                            horizontalalignment='center', verticalalignment='bottom',\n",
    "                            fontsize='x-small', color='k')\n",
    "        #Add annotation\n",
    "        ax.set_title(dataset, {'fontsize':'small'})\n",
    "        xoff = 0.025\n",
    "        yoff = 0.01\n",
    "        rect = plt.Rectangle((xoff, 1), 1-xoff, 0.1,#Manual adjustment\n",
    "                             transform=ax.transAxes, facecolor=tempD2[dataset],\n",
    "                             clip_on=False, linewidth=0, zorder=0.5)\n",
    "        ax.add_patch(rect)\n",
    "        if ax_i==1:\n",
    "            at = AnchoredText('Consensus: own group', loc='center',\n",
    "                              bbox_to_anchor=(1.15, 0.505), bbox_transform=ax.transAxes,#Minor manual adjustment\n",
    "                              frameon=False, prop={'size':'small', 'rotation':90})\n",
    "            ax.add_artist(at)\n",
    "            rect = plt.Rectangle((1+xoff, yoff), 0.225, 1-2*yoff,#Manual adjustment\n",
    "                                 transform=ax.transAxes, facecolor=plt.get_cmap('tab20')(15),\n",
    "                                 clip_on=False, linewidth=0, zorder=0.5)\n",
    "            ax.add_patch(rect)\n",
    "    fig.tight_layout()\n",
    "    #Set title\n",
    "    modulename = moduleDF.loc[module, 'ModuleName']\n",
    "    initial = modulename[0].capitalize()\n",
    "    title = re.sub('^.', initial, modulename)+' ('+module+')'\n",
    "    title = '\\n'.join(wrap(title, 40))#Because the below wrap=True didn't work\n",
    "    fig.suptitle(title, size='small',\n",
    "                 verticalalignment='bottom', horizontalalignment='center', wrap=True, y=0.93)\n",
    "    ##Save\n",
    "    fileDir = './ExportFigures/'\n",
    "    ipynbName = '220525_LC-M001-DIRAC-prot-vs-txn_Comparison-GOBP_ver2-2_'\n",
    "    fileName = 'RCI-inter-group-comparison_RMS-pointplot-'+module.replace('GO:', 'GO')+'.tif'\n",
    "    plt.gcf().savefig(fileDir+ipynbName+fileName, dpi=300, bbox_inches='tight', pad_inches=0.04,\n",
    "                      pil_kwargs={'compression':'tiff_lzw'})\n",
    "    plt.show()\n",
    "    print('')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4. Rank matching score under a fixed consensus: inter-group module comparison"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> The comparison of RMSs under a fixed rank consensus is skipped, since the pattern similarity across datasets is out of interest.  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# — End of notebook —"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "arivale-py3 - Python",
   "language": "python",
   "name": "conda-env-arivale-py3-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  },
  "widgets": {
   "application/vnd.jupyter.widget-state+json": {
    "state": {},
    "version_major": 2,
    "version_minor": 0
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
