{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "c0134e6b-01f7-4f62-a6b4-88a15459c575",
   "metadata": {},
   "source": [
    "# Weighted Gene Co-expression Network Analysis of LC M001 Liver Proteomics — WGCNA Module Detection"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "65b92a30-b425-4441-9937-5e5a16cb564f",
   "metadata": {},
   "source": [
    "***Originally written by Tomasz Wilmanski, and modified by Kengo Watanabe***  \n",
    "\n",
    "This Jupyter Notebook (with R kernel) performed the module detection part of the weighted gene co-expression network analysis (WGCNA; Langfelder, P. & Horvath, S. BMC Bioinform. 2008) on the preprocessed Longevity Consortium (LC) M001 proteomics dataset (adjusted with sex and age; analytes with 50% or less missingness), followed by the enrichment analysis on the hub proteins of WGCNA-identified modules using Gene Ontology Biological Process (GOBP) annotations.  \n",
    "\n",
    "Input files:  \n",
    "- Preprocessed analyte data: 230303_LC-M001-proteomics-WGCNA-ver2_Annotation_processed-data-for-WGCNA.tsv  \n",
    "- Analyte metadata: 230303_LC-M001-proteomics-WGCNA-ver2_Annotation_analyte-metadata_UniProt.tsv  \n",
    "- GOBP annotation metadata: 230303_LC-M001-proteomics-WGCNA-ver2_Annotation_module-metadata_QuickGO-GOBP-min-n4-cov50.tsv  \n",
    "\n",
    "Output figures and tables:  \n",
    "- Figure 3d  \n",
    "- Supplementary Figure 2c  \n",
    "- Supplementary Data 2, 3  \n",
    "\n",
    "Original notebook (memo for my future tracing):  \n",
    "- dalek:\\[JupyterLab HOME\\]/230303_LC-M001-proteomics-WGCNA-ver2/230303_LC-M001-proteomics-WGCNA-ver2_WGCNA-module.ipynb  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "c259db6e-6c00-4980-bfe0-87713bca52f5",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "── \u001b[1mAttaching packages\u001b[22m ─────────────────────────────────────── tidyverse 1.3.1 ──\n",
      "\n",
      "\u001b[32m✔\u001b[39m \u001b[34mggplot2\u001b[39m 3.3.6     \u001b[32m✔\u001b[39m \u001b[34mpurrr  \u001b[39m 0.3.4\n",
      "\u001b[32m✔\u001b[39m \u001b[34mtibble \u001b[39m 3.1.7     \u001b[32m✔\u001b[39m \u001b[34mdplyr  \u001b[39m 1.0.9\n",
      "\u001b[32m✔\u001b[39m \u001b[34mtidyr  \u001b[39m 1.2.0     \u001b[32m✔\u001b[39m \u001b[34mstringr\u001b[39m 1.4.0\n",
      "\u001b[32m✔\u001b[39m \u001b[34mreadr  \u001b[39m 2.1.2     \u001b[32m✔\u001b[39m \u001b[34mforcats\u001b[39m 0.5.1\n",
      "\n",
      "── \u001b[1mConflicts\u001b[22m ────────────────────────────────────────── tidyverse_conflicts() ──\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mfilter()\u001b[39m masks \u001b[34mstats\u001b[39m::filter()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mlag()\u001b[39m    masks \u001b[34mstats\u001b[39m::lag()\n",
      "\n",
      "Loading required package: dynamicTreeCut\n",
      "\n",
      "Loading required package: fastcluster\n",
      "\n",
      "\n",
      "Attaching package: ‘fastcluster’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:stats’:\n",
      "\n",
      "    hclust\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘WGCNA’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:stats’:\n",
      "\n",
      "    cor\n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"WGCNA: 1.71\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Registered S3 method overwritten by 'ggtree':\n",
      "  method      from \n",
      "  identify.gg ggfun\n",
      "\n",
      "clusterProfiler v4.2.2  For help: https://yulab-smu.top/biomedical-knowledge-mining-book/\n",
      "\n",
      "If you use clusterProfiler in published research, please cite:\n",
      "T Wu, E Hu, S Xu, M Chen, P Guo, Z Dai, T Feng, L Zhou, W Tang, L Zhan, X Fu, S Liu, X Bo, and G Yu. clusterProfiler 4.0: A universal enrichment tool for interpreting omics data. The Innovation. 2021, 2(3):100141\n",
      "\n",
      "\n",
      "Attaching package: ‘clusterProfiler’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:purrr’:\n",
      "\n",
      "    simplify\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:stats’:\n",
      "\n",
      "    filter\n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"clusterProfiler: 4.2.2\"\n",
      "[1] \"enrichplot: 1.14.2\"\n",
      "[1] \"readxl: 1.4.2\"\n",
      "[1] \"openxlsx: 4.2.5.2\"\n"
     ]
    }
   ],
   "source": [
    "library(\"tidyverse\")\n",
    "options(repr.plot.width=5, repr.plot.height=5)#Default=7x7\n",
    "\n",
    "#Bioconductor\n",
    "for (package in c(\"WGCNA\", \"clusterProfiler\", \"enrichplot\")) {\n",
    "    #if (!requireNamespace(\"BiocManager\", quietly=TRUE))\n",
    "    #    install.packages(\"BiocManager\")\n",
    "    #BiocManager::install(package)\n",
    "    eval(bquote(library(.(package))))\n",
    "    print(str_c(package, \": \", as.character(packageVersion(package))))\n",
    "}\n",
    "options(stringsAsFactors=FALSE)#Required for WGCNA\n",
    "#CRAN\n",
    "for (package in c(\"readxl\", \"openxlsx\")) {\n",
    "    #install.packages(package)\n",
    "    eval(bquote(library(.(package))))\n",
    "    print(str_c(package, \": \", as.character(packageVersion(package))))\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a650b434-d7ec-4672-b328-1c45f5386d58",
   "metadata": {},
   "source": [
    "## 1. Prepare dataset and metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "769bcfd4-0b17-4e45-88e6-8875109c7eba",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Import analyte data\n",
    "fileDir <- \"./ExportData/\"\n",
    "ipynbName <- \"230303_LC-M001-proteomics-WGCNA-ver2_Annotation_\"\n",
    "fileName <- \"processed-data-for-WGCNA.tsv\"\n",
    "data <- read_delim(str_c(fileDir,ipynbName,fileName), delim=\"\\t\") %>%\n",
    "    column_to_rownames(var=\"UniProtID\") %>%#For easily applying to the WGCNA functions\n",
    "    t()\n",
    "\n",
    "print(str_c(\"nrow: \", nrow(data)))\n",
    "head(data)\n",
    "data[, 1:5] %>% summary()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a4a6e54c-42f8-46fc-b17c-1e6e08c8d14c",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Import analyte metadata\n",
    "fileDir <- \"./ExportData/\"\n",
    "ipynbName <- \"230303_LC-M001-proteomics-WGCNA-ver2_Annotation_\"\n",
    "fileName <- \"analyte-metadata_UniProt.tsv\"\n",
    "meta_analyte <- read_delim(str_c(fileDir,ipynbName,fileName), delim=\"\\t\") %>%\n",
    "    dplyr::select(UniProtID, UniProtName, GeneSymbol)\n",
    "\n",
    "print(str_c(\"nrow: \", nrow(meta_analyte)))\n",
    "head(meta_analyte)\n",
    "print(str_c(\"Unique analytes: \", length(unique(meta_analyte$UniProtID))))\n",
    "print(str_c(\"Unique labels: \", length(unique(meta_analyte$GeneSymbol))))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c05b47ee-ba44-48cc-a69d-0d719ccf67cc",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Import annotation metadata\n",
    "fileDir <- \"./ExportData/\"\n",
    "ipynbName <- \"230303_LC-M001-proteomics-WGCNA-ver2_Annotation_\"\n",
    "fileName <- \"module-metadata_QuickGO-GOBP-min-n4-cov50.tsv\"\n",
    "meta_annot <- read_delim(str_c(fileDir,ipynbName,fileName), delim=\"\\t\")\n",
    "\n",
    "#Check\n",
    "print(str_c(\"nrow: \", nrow(meta_annot)))\n",
    "head(meta_annot)\n",
    "print(str_c(\"Unique analytes: \", length(unique(meta_annot$UniProtID))))\n",
    "print(str_c(\"Unique modules: \", length(unique(meta_annot$ModuleID))))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3295ee21-a227-434b-abf4-208f8f6bc398",
   "metadata": {},
   "source": [
    "## 2. WGCNA"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "613fd909-f7b5-4753-a0b1-2f57e899e3f1",
   "metadata": {},
   "source": [
    "### 2-1. Missingness filter"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "025e3d12-2f8b-40da-959d-499abe1b6666",
   "metadata": {},
   "source": [
    "> This preprocessing was already done.  "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "801199e8-649c-4aba-8f03-5e7443d97bfb",
   "metadata": {},
   "source": [
    "### 2-2. Network construction"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "06f9be1c-a2db-46f7-b6c0-732f99160994",
   "metadata": {},
   "source": [
    "#### 2-2-1. Choose the soft-thresholding power"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0f93e8f5-f122-4866-a573-b2af833e116a",
   "metadata": {},
   "source": [
    "> Choose a beta value to approximate a scale-free topology.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2aea2a38-bcbb-4817-92ff-3c85cfcbafd2",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Choose a set of soft-thresholding powers\n",
    "powers <- 1:20\n",
    "cutoff <- 0.8\n",
    "#Call the network topology analysis function\n",
    "sft <- pickSoftThreshold(data, powerVector=powers, verbose=5, RsquaredCut=cutoff,\n",
    "                         corOptions=c(use=\"p\", method=\"spearman\"), networkType=\"signed hybrid\")\n",
    "\n",
    "#Plot the result\n",
    "options(repr.plot.width=9, repr.plot.height=5)\n",
    "par(mfrow=c(1,2))\n",
    "cex1 <- 0.8\n",
    "##Scale-free topology fit index as a function of the soft-thresholding power\n",
    "plot(sft$fitIndices[,1], -sign(sft$fitIndices[,3])*sft$fitIndices[,2],\n",
    "     xlab=\"Soft Threshold (power)\", ylab=\"Scale Free Topology Model Fit, signed R^2\", type=\"n\",\n",
    "     main=paste(\"Scale independence\"))\n",
    "text(sft$fitIndices[,1], -sign(sft$fitIndices[,3])*sft$fitIndices[,2],\n",
    "     labels=powers, cex=cex1, col=\"black\")\n",
    "##Line corresponds to using an R^2 cut-off of h\n",
    "abline(h=cutoff, col=\"red\")\n",
    "##Mean connectivity as a function of the soft-thresholding power\n",
    "plot(sft$fitIndices[,1], sft$fitIndices[,5],\n",
    "     xlab=\"Soft Threshold (power)\", ylab=\"Mean Connectivity\", type=\"n\",\n",
    "     main=paste(\"Mean connectivity\"))\n",
    "text(sft$fitIndices[,1], sft$fitIndices[,5], labels=powers, cex=cex1, col=\"black\")\n",
    "\n",
    "print(str_c(\"Soft-thresholding power: \", sft$powerEstimate))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5a6393d7-603d-4a2b-a7ff-d6e6df5d0da0",
   "metadata": {},
   "source": [
    "#### 2-2-2. Co-expression similarity and adjacency"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ac24762d-4c20-4100-b833-7b6668b32376",
   "metadata": {},
   "source": [
    "> Calculate correlation network adjacency.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c07123fb-8475-4396-b7bd-01e8f2e89b60",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Generate the adjacency matrix using the chosen soft-thresholding power\n",
    "softPower <- sft$powerEstimate\n",
    "adjacency <- adjacency(data, power=softPower,\n",
    "                       corOptions=list(use=\"p\", method=\"spearman\"), type=\"signed hybrid\")\n",
    "\n",
    "print(str_c(\"nrow: \", nrow(adjacency)))\n",
    "head(adjacency)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "21e3aa67-d99c-4e69-9b75-fa282fdd6c03",
   "metadata": {},
   "source": [
    "### 2-3. Module detection"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7ab227bb-cca9-4cc9-b853-69f919f9f615",
   "metadata": {},
   "source": [
    "#### 2-3-1. Topological Overlap Matrix (TOM)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d2b6789d-bba5-46f0-b994-eba53aeb4375",
   "metadata": {},
   "source": [
    "> To minimize effects of noise and spurious associations, the adjacency is transformed into the topological overlap measure, and the corresponding dissimilarity is calculated.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4645f8aa-8c83-436b-8f06-6616f1411c60",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Turn adjacency into topological overlap\n",
    "TOM <- TOMsimilarity(adjacency, TOMType=\"signed\")\n",
    "dissTOM <- 1 - TOM\n",
    "\n",
    "print(str_c(\"nrow: \", nrow(dissTOM)))\n",
    "head(dissTOM)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "20f97089-26bf-447e-9bad-f5f17eec8449",
   "metadata": {},
   "source": [
    "#### 2-3-2. Hierarchical clustering using TOM dissimilarity"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "873ef9a6-7faa-4d0c-bf81-319dffae7d42",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Call the hierarchical clustering function\n",
    "geneTree <- hclust(as.dist(dissTOM), method=\"average\")\n",
    "\n",
    "#Plot the resulting clustering tree (dendrogram)\n",
    "options(repr.plot.width=12, repr.plot.height=6)\n",
    "plot(geneTree, xlab=\"\", sub=\"\", main=\"Gene clustering on TOM-based dissimilarity\",\n",
    "     labels=FALSE, hang=0.04)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "39a163df-b1e0-4c48-bc20-a0fad60843f0",
   "metadata": {},
   "source": [
    "> In the clustering tree (dendrogram), each leaf corresponds to a gene. Branches of the dendrogram group together densely interconnected, highly co-expressed genes. Module identification amounts to the identification of individual branches (”cutting the branches off the dendrogram”). There are several methods for branch cutting; the standard method of WGCNA package is Dynamic Tree Cut from the package dynamicTreeCut.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e2e2730d-a32b-4606-b1bb-c332eeece84a",
   "metadata": {},
   "outputs": [],
   "source": [
    "#We like large modules, so we set the minimum module size relatively high\n",
    "minModuleSize <- 50\n",
    "#Module identification using dynamic tree cut\n",
    "dynamicMods <- cutreeDynamic(dendro=geneTree, distM=dissTOM,\n",
    "                             deepSplit=4, pamStage=TRUE, pamRespectsDendro=FALSE,\n",
    "                             minClusterSize=minModuleSize)\n",
    "table(dynamicMods)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b3354778-3523-4d20-b4cf-5dbfbc63affd",
   "metadata": {},
   "source": [
    "> The above list shows each module size. Label 0 is reserved for unassigned genes.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d6622ef0-6ef8-4e20-8485-7b41a581ee40",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Convert numeric lables into colors\n",
    "dynamicColors <- labels2colors(dynamicMods)\n",
    "table(dynamicColors)\n",
    "\n",
    "#Plot the dendrogram and colors underneath\n",
    "options(repr.plot.width=12, repr.plot.height=6)\n",
    "plotDendroAndColors(geneTree, dynamicColors, \"Dynamic Tree Cut\",\n",
    "                    dendroLabels=FALSE, hang=0.03,\n",
    "                    addGuide=TRUE, guideHang=0.05,\n",
    "                    main=\"Gene dendrogram and module colors\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3484dfd7-7845-4b6d-8918-583e685d4b8f",
   "metadata": {},
   "source": [
    "#### 2-3-3. Merge similar modules based on eigengenes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "02f89498-769d-432e-8041-75a52365b5a1",
   "metadata": {},
   "source": [
    "> Dynamic Tree Cut may identify modules whose expression profiles are very similar. It would be prudent to merge such modules since their genes are highly co-expressed. To quantify co-expression similarity of entire modules, their eigengenes are calculated and clustered based on their correlation.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a855514f-e92d-4cd6-a62b-eeef33a9def0",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Calculate eigengenes\n",
    "MEList <- moduleEigengenes(data, colors=dynamicColors, impute=TRUE, nPC=2)\n",
    "MEs <- MEList$eigengenes\n",
    "print(str_c(\"nrow: \", nrow(MEs)))\n",
    "head(MEs)\n",
    "\n",
    "#Check the explained variance\n",
    "temp_df <- MEList$varExplained\n",
    "rownames(temp_df) <- c(\"PC1\", \"PC2\")\n",
    "colnames(temp_df) <- str_replace(names(MEList$eigengenes), \"^ME\", \"\")\n",
    "t(temp_df)\n",
    "\n",
    "#Calculate dissimilarity of module eigengenes\n",
    "MEDiss <- 1 - cor(MEs, use=\"pairwise.complete.obs\")\n",
    "\n",
    "#Cluster module eigengenes\n",
    "METree <- hclust(as.dist(MEDiss), method=\"average\")\n",
    "\n",
    "#Plot the result\n",
    "options(repr.plot.width=10, repr.plot.height=5)\n",
    "plot(METree, main=\"Clustering of module eigengenes\",\n",
    "     xlab=\"\", sub=\"\")\n",
    "MEDissThres <- 0.1\n",
    "abline(h=MEDissThres, col=\"red\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3af2c05f-14df-433e-b10d-ba8d807e8b77",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Call an automatic merging function\n",
    "merge <- mergeCloseModules(data, dynamicColors, cutHeight=MEDissThres, verbose=3)\n",
    "\n",
    "#Eigengenes of the new merged modules\n",
    "mergedMEs <- merge$newMEs\n",
    "print(str_c(\"nrow: \", nrow(mergedMEs)))\n",
    "head(mergedMEs)\n",
    "\n",
    "#Update the exparained variance\n",
    "MEList <- moduleEigengenes(data, colors=merge$colors, impute=TRUE, nPC=2)\n",
    "temp_df <- MEList$varExplained\n",
    "rownames(temp_df) <- c(\"PC1\", \"PC2\")\n",
    "colnames(temp_df) <- str_replace(names(MEList$eigengenes), \"^ME\", \"\")\n",
    "t(temp_df)\n",
    "\n",
    "#Update the module eigengene clustering\n",
    "##Calculate dissimilarity of module eigengenes\n",
    "MEDiss <- 1 - cor(mergedMEs, use=\"pairwise.complete.obs\")\n",
    "##Cluster module eigengenes\n",
    "METree <- hclust(as.dist(MEDiss), method=\"average\")\n",
    "##Plot the result\n",
    "options(repr.plot.width=10, repr.plot.height=5)\n",
    "plot(METree, main=\"Clustering of module eigengenes\",\n",
    "     xlab=\"\", sub=\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9466a887-36ae-45f1-883a-fb80bea841bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "#The merged module colors\n",
    "mergedColors <- merge$colors\n",
    "table(mergedColors)\n",
    "\n",
    "#Plot the dendrogram and module colors\n",
    "options(repr.plot.width=12, repr.plot.height=6)\n",
    "plotDendroAndColors(geneTree, cbind(dynamicColors, mergedColors),\n",
    "                    c(\"Dynamic Tree Cut\", \"Merged dynamic\"),\n",
    "                    dendroLabels=FALSE, hang=0.03,\n",
    "                    addGuide=TRUE, guideHang=0.05,\n",
    "                    main=\"Gene dendrogram and module colors\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8804ff05-f8e4-4f90-a5b7-4c4a9b8f9041",
   "metadata": {},
   "source": [
    "#### 2-3-4. Update/save"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8268911f-39c8-4daa-af96-ec6dc7467a78",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Rename\n",
    "moduleColors <- mergedColors\n",
    "MEs <- mergedMEs\n",
    "\n",
    "#Construct numerical labels corresponding to the colors\n",
    "colorOrder <- c(\"grey\", standardColors(50))\n",
    "moduleLabels <- match(moduleColors, colorOrder)-1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "84702b6b-0bd2-48a5-bf7e-65baf6d38b7a",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Prepare the module assignment table\n",
    "module_tbl <- tibble(UniProtID=colnames(data),\n",
    "                     ModuleID=str_to_title(moduleColors),\n",
    "                     ModuleNumber=moduleLabels) %>%\n",
    "    dplyr::left_join(meta_analyte, ., by=\"UniProtID\")\n",
    "print(str_c(\"nrow: \", nrow(module_tbl)))\n",
    "head(module_tbl)\n",
    "\n",
    "#Save module assignments\n",
    "#-> After adding intramodular connectivity to this table later"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "55841c24-a451-4c86-ab4d-4301de986118",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Clean the module eigengene table\n",
    "temp_df <- MEs %>%\n",
    "    rownames_to_column(var=\"SampleID\")\n",
    "names(temp_df)[2:ncol(temp_df)] <- names(temp_df)[2:ncol(temp_df)] %>%\n",
    "    str_replace(., \"^ME\", \"\") %>%\n",
    "    str_to_title(.)\n",
    "print(str_c(\"nrow: \", nrow(temp_df)))\n",
    "head(temp_df)\n",
    "    \n",
    "#Save module eigengene values\n",
    "fileDir <- \"./ExportData/\"\n",
    "ipynbName <- \"230303_LC-M001-proteomics-WGCNA-ver2_WGCNA-module_\"\n",
    "fileName <- \"module-eigengenes.tsv\"\n",
    "temp_df %>% write_tsv(str_c(fileDir,ipynbName,fileName))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "02efd6f2-58a3-42ee-8a77-05fef2a7c744",
   "metadata": {},
   "source": [
    "### 2-4. Relathionship between each analyte and module"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7c792d77-c1db-49c7-9fe5-cfd40db484c5",
   "metadata": {},
   "source": [
    "#### 2-4-1. Module membership"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f7a65791-c4d1-4968-999b-7db887b2ff64",
   "metadata": {},
   "source": [
    "> For each module and each gene, a quantitative measure of module membership (MM) is defined as the correlation between the module eigengene and the gene expression profile. This allows us to quantify the similarity of all input genes (irrespective of their original module membership) to every module.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "66ba4983-e54e-484e-a496-d33e862ce3f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Calculate the numbers of genes and samples\n",
    "nGenes <- ncol(data)\n",
    "nSamples <- nrow(data)\n",
    "\n",
    "#Names (colors) of the modules\n",
    "modNames = substring(names(MEs), 3)\n",
    "\n",
    "#Calculate module membership\n",
    "geneModuleMembership <- as.data.frame(cor(data, MEs, use=\"p\"))\n",
    "#names(geneModuleMembership) <- paste(\"MM\", modNames, sep=\"\")\n",
    "names(geneModuleMembership) <- str_to_title(modNames)\n",
    "print(str_c(\"nrow: \", nrow(geneModuleMembership)))\n",
    "head(geneModuleMembership)\n",
    "\n",
    "#Calculate statisitcal significance of module membership\n",
    "MMPvalue <- as.data.frame(corPvalueStudent(as.matrix(geneModuleMembership), nSamples))\n",
    "#names(MMPvalue) <- paste(\"p.MM\", modNames, sep=\"\")\n",
    "names(MMPvalue) <- str_to_title(modNames)\n",
    "print(str_c(\"nrow: \", nrow(MMPvalue)))\n",
    "head(MMPvalue)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8c55e607-f80c-4e38-af5b-c2ac128471c8",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Save module membership summary\n",
    "fileDir <- \"./ExportData/\"\n",
    "ipynbName <- \"230303_LC-M001-proteomics-WGCNA-ver2_WGCNA-module_\"\n",
    "fileName <- \"module-membership.tsv\"\n",
    "geneModuleMembership %>% rownames_to_column(var=\"UniProtID\") %>% write_tsv(str_c(fileDir,ipynbName,fileName))\n",
    "fileName <- \"module-membership_pvalue.tsv\"\n",
    "MMPvalue %>% rownames_to_column(var=\"UniProtID\") %>% write_tsv(str_c(fileDir,ipynbName,fileName))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f00bdc52-8a56-4522-bd17-6a21ddd3af7d",
   "metadata": {},
   "source": [
    "#### 2-4-2. Intramodular connectivity"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dd40cbd6-1d83-485f-9a54-83b036bf84d4",
   "metadata": {},
   "source": [
    "> This is not included in the official tutorial section, but prepared based on the last \"Export of networks to external software\" section and Tom's notebook.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7fb8e71d-51f1-4e76-bce2-d124b815e071",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Prepare target modules\n",
    "targets <- modNames[modNames!=\"grey\"]\n",
    "\n",
    "#Repeat for each module\n",
    "temp_tbl <- tibble()\n",
    "for (module in targets) {\n",
    "    #Select module probes\n",
    "    probes <- colnames(data)\n",
    "    inModule <- (moduleColors==module)\n",
    "    modProbes <- probes[inModule]\n",
    "    \n",
    "    #Select the corresponding Topological Overlap\n",
    "    modTOM <- TOM[inModule, inModule]\n",
    "    dimnames(modTOM) <- list(modProbes, modProbes)\n",
    "    \n",
    "    #Calculate intramodular connectivity\n",
    "    IMConn <- softConnectivity(data[, modProbes], power=softPower,\n",
    "                               corOptions=list(use=\"p\", method=\"spearman\"), type=\"signed hybrid\")\n",
    "    \n",
    "    #Summary table\n",
    "    connectivity <- tibble(ModuleID=str_to_title(module),\n",
    "                           UniProtID=modProbes,\n",
    "                           IntramodularConnectivity=IMConn,\n",
    "                           TOMsimilaritySum=rowSums(modTOM))\n",
    "    print(str_c(module,\" module: \", nrow(connectivity)))\n",
    "    print(head(connectivity))#Due to within for-loop\n",
    "    \n",
    "    #Add to the overall table\n",
    "    temp_tbl <- dplyr::bind_rows(temp_tbl, connectivity)\n",
    "}\n",
    "\n",
    "print(str_c(\"Total nrow: \", nrow(temp_tbl)))\n",
    "head(temp_tbl)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fe21a7b1-79c5-41ca-9c5e-5bd5df83fc50",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Update the module assignment table\n",
    "module_tbl <- dplyr::left_join(module_tbl, temp_tbl, by=c(\"UniProtID\", \"ModuleID\"))\n",
    "print(str_c(\"nrow: \", nrow(module_tbl)))\n",
    "head(module_tbl)\n",
    "module_tbl %>% dplyr::filter(is.na(IntramodularConnectivity))\n",
    "\n",
    "#Save module assignments\n",
    "fileDir <- \"./ExportData/\"\n",
    "ipynbName <- \"230303_LC-M001-proteomics-WGCNA-ver2_WGCNA-module_\"\n",
    "fileName <- \"module-assignments.tsv\"\n",
    "module_tbl %>% write_tsv(str_c(fileDir,ipynbName,fileName))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "20fa97c4-18b1-46e7-8878-71b795932605",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Check top15 analytes\n",
    "for (module in targets) {\n",
    "    print(module)\n",
    "    module_tbl %>%\n",
    "        select(-ModuleNumber) %>%\n",
    "        dplyr::filter(ModuleID==str_to_title(module)) %>%\n",
    "        dplyr::arrange(desc(IntramodularConnectivity)) %>%\n",
    "        head(., n=15) %>%\n",
    "        print()#Due to within for-loop\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "91030ba3-fe6f-4b29-afaa-f93b2a23a83a",
   "metadata": {},
   "source": [
    "### 2-5. Export module networks"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "29445d43-ba7a-4341-a28f-73304e4cfe4f",
   "metadata": {},
   "source": [
    "> Instead of VisANT, Cytoscape version is used here. Also, intra-module connectivity is calculated within this code.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0894456a-e472-4426-8be2-a08c12615fc1",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Repeat for each module\n",
    "for (module in targets) {\n",
    "    #Select module probes\n",
    "    probes <- colnames(data)\n",
    "    inModule <- (moduleColors==module)\n",
    "    modProbes <- probes[inModule]\n",
    "    print(str_c(module,\" module: \", length(modProbes)))\n",
    "    modGenes <- meta_analyte$GeneSymbol[match(modProbes, meta_analyte$UniProtID)]\n",
    "    \n",
    "    #Select the corresponding Topological Overlap\n",
    "    modTOM <- TOM[inModule, inModule]\n",
    "    dimnames(modTOM) <- list(modProbes, modProbes)\n",
    "    \n",
    "    #Export the network into edge and node list files Cytoscape can read\n",
    "    fileDir <- \"./ExportData/\"\n",
    "    ipynbName <- \"230303_LC-M001-proteomics-WGCNA-ver2_WGCNA-module_\"\n",
    "    fileName <- str_c(\"CytoscapeInput-\",module)\n",
    "    cyt <- exportNetworkToCytoscape(modTOM,\n",
    "                                    edgeFile=str_c(fileDir,ipynbName,fileName,\"-edges.txt\"),\n",
    "                                    nodeFile=str_c(fileDir,ipynbName,fileName,\"-nodes.txt\"),\n",
    "                                    weighted=TRUE,\n",
    "                                    threshold=0,\n",
    "                                    nodeNames=modProbes,\n",
    "                                    altNodeNames=modGenes,\n",
    "                                    nodeAttr=str_to_title(moduleColors[inModule]))\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4e56358d-e7db-4756-9759-57192004ec8c",
   "metadata": {},
   "source": [
    "## 3. Enrichment analysis for module members"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "13eccd02-9df5-497c-bf5c-7b4cc109755a",
   "metadata": {},
   "source": [
    "### 3-1. Save result objects"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7b6ad3df-1dd8-4906-b7dc-3e4b2a1c6755",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Create a workbook object to save as one single .xlsx file\n",
    "workbook <- createWorkbook()\n",
    "\n",
    "#Summarize results per WGCNA module\n",
    "nMembers_vec <- c()\n",
    "nMappedMembers_vec <- c()\n",
    "nMemberModules_vec <- c()\n",
    "res_list <- list()\n",
    "for (i in 1:length(targets)) {\n",
    "    #Prepare input for enricher()\n",
    "    module <- str_to_title(targets[i])#WGCNA module\n",
    "    members <- module_tbl %>%\n",
    "        dplyr::filter(ModuleID==module) %>%\n",
    "        .$GeneSymbol#Able to use instead of UniProtID because duplicate issues were resolved\n",
    "    backgrounds <- meta_annot %>%\n",
    "        dplyr::select(ModuleID, GeneSymbol)#Here, not WGCNA module but a priori module (functional annotation)\n",
    "    labels <- meta_annot %>%\n",
    "        dplyr::select(ModuleID, ModuleName) %>%#Here, not WGCNA module but a priori module (functional annotation)\n",
    "        dplyr::distinct()\n",
    "    print(str_c(module,\": \",length(members),\" members\"))\n",
    "    \n",
    "    #Save info\n",
    "    nMembers_vec <- c(nMembers_vec, length(members))\n",
    "    temp <- backgrounds %>% dplyr::filter(GeneSymbol %in% members)\n",
    "    nMappedMembers_vec <- c(nMappedMembers_vec, length(unique(temp$GeneSymbol)))\n",
    "    nMemberModules_vec <- c(nMemberModules_vec, length(unique(temp$ModuleID)))\n",
    "    \n",
    "    #Enrichment analysis\n",
    "    temp <- enricher(gene=members,\n",
    "                     pvalueCutoff=1.0,#To export all\n",
    "                     pAdjustMethod=\"BH\",\n",
    "                     #universe=backgrounds,#Already managed\n",
    "                     minGSSize=4,\n",
    "                     maxGSSize=10000,\n",
    "                     qvalueCutoff=1.0,#To export all\n",
    "                     TERM2GENE=backgrounds,\n",
    "                     TERM2NAME=labels)\n",
    "    \n",
    "    #Add the summary table to the workbook object as an independent sheet\n",
    "    if (is.data.frame(temp[])) {\n",
    "        temp1 <- tibble(temp[]) %>%\n",
    "            dplyr::rename(AnnotationID=ID, AnnotationName=Description, Ratio2Members=GeneRatio, Ratio2BGs=BgRatio,\n",
    "                          Pval=pvalue, AdjPval=p.adjust, Qval=qvalue, MappedMember=geneID, nMappedMembers=Count)\n",
    "    } else {\n",
    "        temp1 <- tibble(`n/a`=NA)\n",
    "    }\n",
    "    addWorksheet(workbook, sheetName=module)\n",
    "    writeData(workbook, module, temp1)\n",
    "    \n",
    "    #Add result object to list\n",
    "    res_list <- c(res_list, list(temp))\n",
    "}\n",
    "\n",
    "#Save the workbook as one single .xlsx file\n",
    "fileDir <- \"./ExportData/\"\n",
    "ipynbName <- \"230303_LC-M001-proteomics-WGCNA-ver2_WGCNA-module_\"\n",
    "fileName <- \"clusterProfiler-members.xlsx\"\n",
    "saveWorkbook(workbook, file=str_c(fileDir,ipynbName,fileName), overwrite=TRUE)\n",
    "\n",
    "print(str_c(\"nObjects: \", as.character(length(res_list))))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "49d517a4-2e7c-45bf-9383-dabf89f30f3f",
   "metadata": {},
   "source": [
    "### 3-2. Check results"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3476d97f-c0f3-464a-b4ce-27345613d86d",
   "metadata": {},
   "source": [
    "#### 3-2-1. Greenyellow"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d731eb49-d856-4671-a35d-bba477470b6b",
   "metadata": {},
   "outputs": [],
   "source": [
    "obj_i <- 1\n",
    "\n",
    "#Retreive results\n",
    "module <- str_to_title(targets[obj_i])#WGCNA module\n",
    "nMembers <- nMembers_vec[obj_i]\n",
    "nMappedMembers <- nMappedMembers_vec[obj_i]\n",
    "nMemberModules <- nMemberModules_vec[obj_i]\n",
    "res <- res_list[[obj_i]]\n",
    "\n",
    "#Check\n",
    "print(module)\n",
    "print(str_c(\" - # of members: \",as.character(nMembers)))\n",
    "print(str_c(\" - # of members mapped to any annotation: \",as.character(nMappedMembers)))\n",
    "print(str_c(\" - # of annotations having any member: \",as.character(nMemberModules)))\n",
    "res\n",
    "print(\" <- Note that the above 'X enriched terms found' is not correct. In this case, X indicates the number of all the tested terms.\")\n",
    "if (is.data.frame(res[])) {\n",
    "    tibble(res[]) %>%\n",
    "        dplyr::filter(pvalue<0.05) %>%#Display only nominal P-value < 0.05\n",
    "        dplyr::rename(AnnotationID=ID, AnnotationName=Description, Ratio2Members=GeneRatio, Ratio2BGs=BgRatio,\n",
    "                      Pval=pvalue, AdjPval=p.adjust, Qval=qvalue, MappedMember=geneID, nMappedMembers=Count)\n",
    "}\n",
    "\n",
    "#Visualization\n",
    "if (is.data.frame(res[])) {\n",
    "    display <- tibble(res[]) %>%\n",
    "        dplyr::filter(pvalue<0.05) %>%\n",
    "        nrow()\n",
    "} else {\n",
    "    display <- 0\n",
    "}\n",
    "if (display>0) {\n",
    "    temp <- res %>%\n",
    "        dplyr::filter(pvalue<0.05) %>%#Display only nominal P-value < 0.05\n",
    "        dplyr::mutate(PvalLabel=str_c(\"AdjPval = \",scales::scientific(p.adjust, digits=2)),\n",
    "                      AdjSignif=ifelse(p.adjust<0.05, \"True\", \"False\")) %>%\n",
    "        barplot(., x=\"Count\", color=\"p.adjust\", showCategory=display) +\n",
    "        geom_text(aes(label=PvalLabel, color=AdjSignif), nudge_x=0.5, hjust=0) +\n",
    "        coord_cartesian(clip=\"off\") +\n",
    "        scale_x_continuous(limits=c(0, 25), breaks=seq(0, 25, by=5), expand=c(0, 0)) +\n",
    "        scale_y_discrete(labels=function(x) {str_wrap(x, width=60)}) +\n",
    "        scale_fill_viridis_c(begin=0, end=1, direction=1, option=\"plasma\",\n",
    "                             limits=c(0, 0.2), breaks=seq(0, 0.2, by=0.05), oob=scales::squish,\n",
    "                             name=\"AdjPval\") +\n",
    "        scale_color_manual(values=c(\"True\"=\"#990000\", \"False\"=\"gray40\"), ) +\n",
    "        guides(fill=guide_colorbar(reverse=TRUE), color=\"none\") +\n",
    "        labs(x=\"Count of members\",\n",
    "             y=\"\", title=str_c(\"Enriched GOBP in \",module,\" module members\")) +\n",
    "        theme_classic(base_size=16, base_family=\"Helvetica\") +\n",
    "        theme(text=element_text(face=\"plain\", color=\"black\", family=\"Helvetica\")) +\n",
    "        theme(axis.text.x=element_text(face=\"plain\", color=\"black\", family=\"Helvetica\"),\n",
    "              axis.text.y=element_text(face=\"plain\", color=\"black\", family=\"Helvetica\", lineheight=0.75),\n",
    "              axis.title=element_text(face=\"plain\", color=\"black\", family=\"Helvetica\")) +\n",
    "        theme(plot.title=element_text(size=18, hjust=1.0)) +\n",
    "        theme(legend.direction=\"vertical\", legend.box=\"horizontal\",\n",
    "              legend.background=element_blank())\n",
    "    options(repr.plot.width=9, repr.plot.height=max(c(1+display*0.25, 2.5)))\n",
    "    plot(temp)\n",
    "    #Save\n",
    "    fileDir <- \"./ExportFigures/\"\n",
    "    ipynbName <- \"230303_LC-M001-proteomics-WGCNA-ver2_WGCNA-module_\"\n",
    "    fileName <- str_c(\"enrichment-\",module,\"-all.pdf\")\n",
    "    ggsave(file=str_c(fileDir,ipynbName,fileName), plot=temp,\n",
    "           width=9, height=max(c(1+display*0.25, 2.5)), units=\"in\")\n",
    "    #(Font family is not reflected in JupyterLab output, but correctly done in .pdf file.)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a14bb16e-4bf7-46d5-888f-c45b479f8b03",
   "metadata": {},
   "source": [
    "#### 3-2-2. Turquoise"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "49fd0cc0-676f-421e-a685-443f933573b9",
   "metadata": {},
   "outputs": [],
   "source": [
    "obj_i <- 2\n",
    "\n",
    "#Retreive results\n",
    "module <- str_to_title(targets[obj_i])#WGCNA module\n",
    "nMembers <- nMembers_vec[obj_i]\n",
    "nMappedMembers <- nMappedMembers_vec[obj_i]\n",
    "nMemberModules <- nMemberModules_vec[obj_i]\n",
    "res <- res_list[[obj_i]]\n",
    "\n",
    "#Check\n",
    "print(module)\n",
    "print(str_c(\" - # of members: \",as.character(nMembers)))\n",
    "print(str_c(\" - # of members mapped to any annotation: \",as.character(nMappedMembers)))\n",
    "print(str_c(\" - # of annotations having any member: \",as.character(nMemberModules)))\n",
    "res\n",
    "print(\" <- Note that the above 'X enriched terms found' is not correct. In this case, X indicates the number of all the tested terms.\")\n",
    "if (is.data.frame(res[])) {\n",
    "    tibble(res[]) %>%\n",
    "        dplyr::filter(pvalue<0.05) %>%#Display only nominal P-value < 0.05\n",
    "        dplyr::rename(AnnotationID=ID, AnnotationName=Description, Ratio2Members=GeneRatio, Ratio2BGs=BgRatio,\n",
    "                      Pval=pvalue, AdjPval=p.adjust, Qval=qvalue, MappedMember=geneID, nMappedMembers=Count)\n",
    "}\n",
    "\n",
    "#Visualization\n",
    "if (is.data.frame(res[])) {\n",
    "    display <- tibble(res[]) %>%\n",
    "        dplyr::filter(pvalue<0.05) %>%\n",
    "        nrow()\n",
    "} else {\n",
    "    display <- 0\n",
    "}\n",
    "if (display>0) {\n",
    "    temp <- res %>%\n",
    "        dplyr::filter(pvalue<0.05) %>%#Display only nominal P-value < 0.05\n",
    "        dplyr::mutate(PvalLabel=str_c(\"AdjPval = \",scales::scientific(p.adjust, digits=2)),\n",
    "                      AdjSignif=ifelse(p.adjust<0.05, \"True\", \"False\")) %>%\n",
    "        barplot(., x=\"Count\", color=\"p.adjust\", showCategory=display) +\n",
    "        geom_text(aes(label=PvalLabel, color=AdjSignif), nudge_x=0.5, hjust=0) +\n",
    "        coord_cartesian(clip=\"off\") +\n",
    "        scale_x_continuous(limits=c(0, 25), breaks=seq(0, 25, by=5), expand=c(0, 0)) +\n",
    "        scale_y_discrete(labels=function(x) {str_wrap(x, width=60)}) +\n",
    "        scale_fill_viridis_c(begin=0, end=1, direction=1, option=\"plasma\",\n",
    "                             limits=c(0, 0.2), breaks=seq(0, 0.2, by=0.05), oob=scales::squish,\n",
    "                             name=\"AdjPval\") +\n",
    "        scale_color_manual(values=c(\"True\"=\"#990000\", \"False\"=\"gray40\"), ) +\n",
    "        guides(fill=guide_colorbar(reverse=TRUE), color=\"none\") +\n",
    "        labs(x=\"Count of members\",\n",
    "             y=\"\", title=str_c(\"Enriched GOBP in \",module,\" module members\")) +\n",
    "        theme_classic(base_size=16, base_family=\"Helvetica\") +\n",
    "        theme(text=element_text(face=\"plain\", color=\"black\", family=\"Helvetica\")) +\n",
    "        theme(axis.text.x=element_text(face=\"plain\", color=\"black\", family=\"Helvetica\"),\n",
    "              axis.text.y=element_text(face=\"plain\", color=\"black\", family=\"Helvetica\", lineheight=0.75),\n",
    "              axis.title=element_text(face=\"plain\", color=\"black\", family=\"Helvetica\")) +\n",
    "        theme(plot.title=element_text(size=18, hjust=1.0)) +\n",
    "        theme(legend.direction=\"vertical\", legend.box=\"horizontal\",\n",
    "              legend.background=element_blank())\n",
    "    options(repr.plot.width=9, repr.plot.height=max(c(1+display*0.25, 2.5)))\n",
    "    plot(temp)\n",
    "    #Save\n",
    "    fileDir <- \"./ExportFigures/\"\n",
    "    ipynbName <- \"230303_LC-M001-proteomics-WGCNA-ver2_WGCNA-module_\"\n",
    "    fileName <- str_c(\"enrichment-\",module,\"-all.pdf\")\n",
    "    ggsave(file=str_c(fileDir,ipynbName,fileName), plot=temp,\n",
    "           width=9, height=max(c(1+display*0.25, 2.5)), units=\"in\")\n",
    "    #(Font family is not reflected in JupyterLab output, but correctly done in .pdf file.)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "11784582-d2fa-428e-ae1d-8556701ef734",
   "metadata": {},
   "source": [
    "#### 3-2-3. Yellow"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ce71a566-17e6-4636-99de-de7c32eea89f",
   "metadata": {},
   "outputs": [],
   "source": [
    "obj_i <- 3\n",
    "\n",
    "#Retreive results\n",
    "module <- str_to_title(targets[obj_i])#WGCNA module\n",
    "nMembers <- nMembers_vec[obj_i]\n",
    "nMappedMembers <- nMappedMembers_vec[obj_i]\n",
    "nMemberModules <- nMemberModules_vec[obj_i]\n",
    "res <- res_list[[obj_i]]\n",
    "\n",
    "#Check\n",
    "print(module)\n",
    "print(str_c(\" - # of members: \",as.character(nMembers)))\n",
    "print(str_c(\" - # of members mapped to any annotation: \",as.character(nMappedMembers)))\n",
    "print(str_c(\" - # of annotations having any member: \",as.character(nMemberModules)))\n",
    "res\n",
    "print(\" <- Note that the above 'X enriched terms found' is not correct. In this case, X indicates the number of all the tested terms.\")\n",
    "if (is.data.frame(res[])) {\n",
    "    tibble(res[]) %>%\n",
    "        dplyr::filter(pvalue<0.05) %>%#Display only nominal P-value < 0.05\n",
    "        dplyr::rename(AnnotationID=ID, AnnotationName=Description, Ratio2Members=GeneRatio, Ratio2BGs=BgRatio,\n",
    "                      Pval=pvalue, AdjPval=p.adjust, Qval=qvalue, MappedMember=geneID, nMappedMembers=Count)\n",
    "}\n",
    "\n",
    "#Visualization\n",
    "if (is.data.frame(res[])) {\n",
    "    display <- tibble(res[]) %>%\n",
    "        dplyr::filter(pvalue<0.05) %>%\n",
    "        nrow()\n",
    "} else {\n",
    "    display <- 0\n",
    "}\n",
    "if (display>0) {\n",
    "    temp <- res %>%\n",
    "        dplyr::filter(pvalue<0.05) %>%#Display only nominal P-value < 0.05\n",
    "        dplyr::mutate(PvalLabel=str_c(\"AdjPval = \",scales::scientific(p.adjust, digits=2)),\n",
    "                      AdjSignif=ifelse(p.adjust<0.05, \"True\", \"False\")) %>%\n",
    "        barplot(., x=\"Count\", color=\"p.adjust\", showCategory=display) +\n",
    "        geom_text(aes(label=PvalLabel, color=AdjSignif), nudge_x=0.5, hjust=0) +\n",
    "        coord_cartesian(clip=\"off\") +\n",
    "        scale_x_continuous(limits=c(0, 25), breaks=seq(0, 25, by=5), expand=c(0, 0)) +\n",
    "        scale_y_discrete(labels=function(x) {str_wrap(x, width=60)}) +\n",
    "        scale_fill_viridis_c(begin=0, end=1, direction=1, option=\"plasma\",\n",
    "                             limits=c(0, 0.2), breaks=seq(0, 0.2, by=0.05), oob=scales::squish,\n",
    "                             name=\"AdjPval\") +\n",
    "        scale_color_manual(values=c(\"True\"=\"#990000\", \"False\"=\"gray40\"), ) +\n",
    "        guides(fill=guide_colorbar(reverse=TRUE), color=\"none\") +\n",
    "        labs(x=\"Count of members\",\n",
    "             y=\"\", title=str_c(\"Enriched GOBP in \",module,\" module members\")) +\n",
    "        theme_classic(base_size=16, base_family=\"Helvetica\") +\n",
    "        theme(text=element_text(face=\"plain\", color=\"black\", family=\"Helvetica\")) +\n",
    "        theme(axis.text.x=element_text(face=\"plain\", color=\"black\", family=\"Helvetica\"),\n",
    "              axis.text.y=element_text(face=\"plain\", color=\"black\", family=\"Helvetica\", lineheight=0.75),\n",
    "              axis.title=element_text(face=\"plain\", color=\"black\", family=\"Helvetica\")) +\n",
    "        theme(plot.title=element_text(size=18, hjust=1.0)) +\n",
    "        theme(legend.direction=\"vertical\", legend.box=\"horizontal\",\n",
    "              legend.background=element_blank())\n",
    "    options(repr.plot.width=9, repr.plot.height=max(c(1+display*0.25, 2.5)))\n",
    "    plot(temp)\n",
    "    #Save\n",
    "    fileDir <- \"./ExportFigures/\"\n",
    "    ipynbName <- \"230303_LC-M001-proteomics-WGCNA-ver2_WGCNA-module_\"\n",
    "    fileName <- str_c(\"enrichment-\",module,\"-all.pdf\")\n",
    "    ggsave(file=str_c(fileDir,ipynbName,fileName), plot=temp,\n",
    "           width=9, height=max(c(1+display*0.25, 2.5)), units=\"in\")\n",
    "    #(Font family is not reflected in JupyterLab output, but correctly done in .pdf file.)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8a7fb33b-acb9-4875-8533-9a15521e37d1",
   "metadata": {},
   "source": [
    "#### 3-2-4. Black"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "62862ed0-e3b1-4af5-9436-37f3c690a8b9",
   "metadata": {},
   "outputs": [],
   "source": [
    "obj_i <- 4\n",
    "\n",
    "#Retreive results\n",
    "module <- str_to_title(targets[obj_i])#WGCNA module\n",
    "nMembers <- nMembers_vec[obj_i]\n",
    "nMappedMembers <- nMappedMembers_vec[obj_i]\n",
    "nMemberModules <- nMemberModules_vec[obj_i]\n",
    "res <- res_list[[obj_i]]\n",
    "\n",
    "#Check\n",
    "print(module)\n",
    "print(str_c(\" - # of members: \",as.character(nMembers)))\n",
    "print(str_c(\" - # of members mapped to any annotation: \",as.character(nMappedMembers)))\n",
    "print(str_c(\" - # of annotations having any member: \",as.character(nMemberModules)))\n",
    "res\n",
    "print(\" <- Note that the above 'X enriched terms found' is not correct. In this case, X indicates the number of all the tested terms.\")\n",
    "if (is.data.frame(res[])) {\n",
    "    tibble(res[]) %>%\n",
    "        dplyr::filter(pvalue<0.05) %>%#Display only nominal P-value < 0.05\n",
    "        dplyr::rename(AnnotationID=ID, AnnotationName=Description, Ratio2Members=GeneRatio, Ratio2BGs=BgRatio,\n",
    "                      Pval=pvalue, AdjPval=p.adjust, Qval=qvalue, MappedMember=geneID, nMappedMembers=Count)\n",
    "}\n",
    "\n",
    "#Visualization\n",
    "if (is.data.frame(res[])) {\n",
    "    display <- tibble(res[]) %>%\n",
    "        dplyr::filter(pvalue<0.05) %>%\n",
    "        nrow()\n",
    "} else {\n",
    "    display <- 0\n",
    "}\n",
    "if (display>0) {\n",
    "    temp <- res %>%\n",
    "        dplyr::filter(pvalue<0.05) %>%#Display only nominal P-value < 0.05\n",
    "        dplyr::mutate(PvalLabel=str_c(\"AdjPval = \",scales::scientific(p.adjust, digits=2)),\n",
    "                      AdjSignif=ifelse(p.adjust<0.05, \"True\", \"False\")) %>%\n",
    "        barplot(., x=\"Count\", color=\"p.adjust\", showCategory=display) +\n",
    "        geom_text(aes(label=PvalLabel, color=AdjSignif), nudge_x=0.5, hjust=0) +\n",
    "        coord_cartesian(clip=\"off\") +\n",
    "        scale_x_continuous(limits=c(0, 25), breaks=seq(0, 25, by=5), expand=c(0, 0)) +\n",
    "        scale_y_discrete(labels=function(x) {str_wrap(x, width=60)}) +\n",
    "        scale_fill_viridis_c(begin=0, end=1, direction=1, option=\"plasma\",\n",
    "                             limits=c(0, 0.2), breaks=seq(0, 0.2, by=0.05), oob=scales::squish,\n",
    "                             name=\"AdjPval\") +\n",
    "        scale_color_manual(values=c(\"True\"=\"#990000\", \"False\"=\"gray40\"), ) +\n",
    "        guides(fill=guide_colorbar(reverse=TRUE), color=\"none\") +\n",
    "        labs(x=\"Count of members\",\n",
    "             y=\"\", title=str_c(\"Enriched GOBP in \",module,\" module members\")) +\n",
    "        theme_classic(base_size=16, base_family=\"Helvetica\") +\n",
    "        theme(text=element_text(face=\"plain\", color=\"black\", family=\"Helvetica\")) +\n",
    "        theme(axis.text.x=element_text(face=\"plain\", color=\"black\", family=\"Helvetica\"),\n",
    "              axis.text.y=element_text(face=\"plain\", color=\"black\", family=\"Helvetica\", lineheight=0.75),\n",
    "              axis.title=element_text(face=\"plain\", color=\"black\", family=\"Helvetica\")) +\n",
    "        theme(plot.title=element_text(size=18, hjust=1.0)) +\n",
    "        theme(legend.direction=\"vertical\", legend.box=\"horizontal\",\n",
    "              legend.background=element_blank())\n",
    "    options(repr.plot.width=9, repr.plot.height=max(c(1+display*0.25, 2.5)))\n",
    "    plot(temp)\n",
    "    #Save\n",
    "    fileDir <- \"./ExportFigures/\"\n",
    "    ipynbName <- \"230303_LC-M001-proteomics-WGCNA-ver2_WGCNA-module_\"\n",
    "    fileName <- str_c(\"enrichment-\",module,\"-all.pdf\")\n",
    "    ggsave(file=str_c(fileDir,ipynbName,fileName), plot=temp,\n",
    "           width=9, height=max(c(1+display*0.25, 2.5)), units=\"in\")\n",
    "    #(Font family is not reflected in JupyterLab output, but correctly done in .pdf file.)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7ea1ca77-3cdf-437b-b95c-0006b70da99d",
   "metadata": {},
   "source": [
    "#### 3-2-5. Blue"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "93732237-7d89-4211-acdc-1bea82ce1695",
   "metadata": {},
   "outputs": [],
   "source": [
    "obj_i <- 5\n",
    "\n",
    "#Retreive results\n",
    "module <- str_to_title(targets[obj_i])#WGCNA module\n",
    "nMembers <- nMembers_vec[obj_i]\n",
    "nMappedMembers <- nMappedMembers_vec[obj_i]\n",
    "nMemberModules <- nMemberModules_vec[obj_i]\n",
    "res <- res_list[[obj_i]]\n",
    "\n",
    "#Check\n",
    "print(module)\n",
    "print(str_c(\" - # of members: \",as.character(nMembers)))\n",
    "print(str_c(\" - # of members mapped to any annotation: \",as.character(nMappedMembers)))\n",
    "print(str_c(\" - # of annotations having any member: \",as.character(nMemberModules)))\n",
    "res\n",
    "print(\" <- Note that the above 'X enriched terms found' is not correct. In this case, X indicates the number of all the tested terms.\")\n",
    "if (is.data.frame(res[])) {\n",
    "    tibble(res[]) %>%\n",
    "        dplyr::filter(pvalue<0.05) %>%#Display only nominal P-value < 0.05\n",
    "        dplyr::rename(AnnotationID=ID, AnnotationName=Description, Ratio2Members=GeneRatio, Ratio2BGs=BgRatio,\n",
    "                      Pval=pvalue, AdjPval=p.adjust, Qval=qvalue, MappedMember=geneID, nMappedMembers=Count)\n",
    "}\n",
    "\n",
    "#Visualization\n",
    "if (is.data.frame(res[])) {\n",
    "    display <- tibble(res[]) %>%\n",
    "        dplyr::filter(pvalue<0.05) %>%\n",
    "        nrow()\n",
    "} else {\n",
    "    display <- 0\n",
    "}\n",
    "if (display>0) {\n",
    "    temp <- res %>%\n",
    "        dplyr::filter(pvalue<0.05) %>%#Display only nominal P-value < 0.05\n",
    "        dplyr::mutate(PvalLabel=str_c(\"AdjPval = \",scales::scientific(p.adjust, digits=2)),\n",
    "                      AdjSignif=ifelse(p.adjust<0.05, \"True\", \"False\")) %>%\n",
    "        barplot(., x=\"Count\", color=\"p.adjust\", showCategory=display) +\n",
    "        geom_text(aes(label=PvalLabel, color=AdjSignif), nudge_x=0.5, hjust=0) +\n",
    "        coord_cartesian(clip=\"off\") +\n",
    "        scale_x_continuous(limits=c(0, 25), breaks=seq(0, 25, by=5), expand=c(0, 0)) +\n",
    "        scale_y_discrete(labels=function(x) {str_wrap(x, width=60)}) +\n",
    "        scale_fill_viridis_c(begin=0, end=1, direction=1, option=\"plasma\",\n",
    "                             limits=c(0, 0.2), breaks=seq(0, 0.2, by=0.05), oob=scales::squish,\n",
    "                             name=\"AdjPval\") +\n",
    "        scale_color_manual(values=c(\"True\"=\"#990000\", \"False\"=\"gray40\"), ) +\n",
    "        guides(fill=guide_colorbar(reverse=TRUE), color=\"none\") +\n",
    "        labs(x=\"Count of members\",\n",
    "             y=\"\", title=str_c(\"Enriched GOBP in \",module,\" module members\")) +\n",
    "        theme_classic(base_size=16, base_family=\"Helvetica\") +\n",
    "        theme(text=element_text(face=\"plain\", color=\"black\", family=\"Helvetica\")) +\n",
    "        theme(axis.text.x=element_text(face=\"plain\", color=\"black\", family=\"Helvetica\"),\n",
    "              axis.text.y=element_text(face=\"plain\", color=\"black\", family=\"Helvetica\", lineheight=0.75),\n",
    "              axis.title=element_text(face=\"plain\", color=\"black\", family=\"Helvetica\")) +\n",
    "        theme(plot.title=element_text(size=18, hjust=1.0)) +\n",
    "        theme(legend.direction=\"vertical\", legend.box=\"horizontal\",\n",
    "              legend.background=element_blank())\n",
    "    options(repr.plot.width=9, repr.plot.height=max(c(1+display*0.25, 2.5)))\n",
    "    plot(temp)\n",
    "    #Save\n",
    "    fileDir <- \"./ExportFigures/\"\n",
    "    ipynbName <- \"230303_LC-M001-proteomics-WGCNA-ver2_WGCNA-module_\"\n",
    "    fileName <- str_c(\"enrichment-\",module,\"-all.pdf\")\n",
    "    ggsave(file=str_c(fileDir,ipynbName,fileName), plot=temp,\n",
    "           width=9, height=max(c(1+display*0.25, 2.5)), units=\"in\")\n",
    "    #(Font family is not reflected in JupyterLab output, but correctly done in .pdf file.)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dbc90df1-c3db-4b41-a425-b10c3147c7a9",
   "metadata": {},
   "source": [
    "#### 3-2-6. Purple"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9ed0c753-5aea-4b24-a502-aeb6859392d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "obj_i <- 6\n",
    "\n",
    "#Retreive results\n",
    "module <- str_to_title(targets[obj_i])#WGCNA module\n",
    "nMembers <- nMembers_vec[obj_i]\n",
    "nMappedMembers <- nMappedMembers_vec[obj_i]\n",
    "nMemberModules <- nMemberModules_vec[obj_i]\n",
    "res <- res_list[[obj_i]]\n",
    "\n",
    "#Check\n",
    "print(module)\n",
    "print(str_c(\" - # of members: \",as.character(nMembers)))\n",
    "print(str_c(\" - # of members mapped to any annotation: \",as.character(nMappedMembers)))\n",
    "print(str_c(\" - # of annotations having any member: \",as.character(nMemberModules)))\n",
    "res\n",
    "print(\" <- Note that the above 'X enriched terms found' is not correct. In this case, X indicates the number of all the tested terms.\")\n",
    "if (is.data.frame(res[])) {\n",
    "    tibble(res[]) %>%\n",
    "        dplyr::filter(pvalue<0.05) %>%#Display only nominal P-value < 0.05\n",
    "        dplyr::rename(AnnotationID=ID, AnnotationName=Description, Ratio2Members=GeneRatio, Ratio2BGs=BgRatio,\n",
    "                      Pval=pvalue, AdjPval=p.adjust, Qval=qvalue, MappedMember=geneID, nMappedMembers=Count)\n",
    "}\n",
    "\n",
    "#Visualization\n",
    "if (is.data.frame(res[])) {\n",
    "    display <- tibble(res[]) %>%\n",
    "        dplyr::filter(pvalue<0.05) %>%\n",
    "        nrow()\n",
    "} else {\n",
    "    display <- 0\n",
    "}\n",
    "if (display>0) {\n",
    "    temp <- res %>%\n",
    "        dplyr::filter(pvalue<0.05) %>%#Display only nominal P-value < 0.05\n",
    "        dplyr::mutate(PvalLabel=str_c(\"AdjPval = \",scales::scientific(p.adjust, digits=2)),\n",
    "                      AdjSignif=ifelse(p.adjust<0.05, \"True\", \"False\")) %>%\n",
    "        barplot(., x=\"Count\", color=\"p.adjust\", showCategory=display) +\n",
    "        geom_text(aes(label=PvalLabel, color=AdjSignif), nudge_x=0.5, hjust=0) +\n",
    "        coord_cartesian(clip=\"off\") +\n",
    "        scale_x_continuous(limits=c(0, 25), breaks=seq(0, 25, by=5), expand=c(0, 0)) +\n",
    "        scale_y_discrete(labels=function(x) {str_wrap(x, width=60)}) +\n",
    "        scale_fill_viridis_c(begin=0, end=1, direction=1, option=\"plasma\",\n",
    "                             limits=c(0, 0.2), breaks=seq(0, 0.2, by=0.05), oob=scales::squish,\n",
    "                             name=\"AdjPval\") +\n",
    "        scale_color_manual(values=c(\"True\"=\"#990000\", \"False\"=\"gray40\"), ) +\n",
    "        guides(fill=guide_colorbar(reverse=TRUE), color=\"none\") +\n",
    "        labs(x=\"Count of members\",\n",
    "             y=\"\", title=str_c(\"Enriched GOBP in \",module,\" module members\")) +\n",
    "        theme_classic(base_size=16, base_family=\"Helvetica\") +\n",
    "        theme(text=element_text(face=\"plain\", color=\"black\", family=\"Helvetica\")) +\n",
    "        theme(axis.text.x=element_text(face=\"plain\", color=\"black\", family=\"Helvetica\"),\n",
    "              axis.text.y=element_text(face=\"plain\", color=\"black\", family=\"Helvetica\", lineheight=0.75),\n",
    "              axis.title=element_text(face=\"plain\", color=\"black\", family=\"Helvetica\")) +\n",
    "        theme(plot.title=element_text(size=18, hjust=1.0)) +\n",
    "        theme(legend.direction=\"vertical\", legend.box=\"horizontal\",\n",
    "              legend.background=element_blank())\n",
    "    options(repr.plot.width=9, repr.plot.height=max(c(1+display*0.25, 2.5)))\n",
    "    plot(temp)\n",
    "    #Save\n",
    "    fileDir <- \"./ExportFigures/\"\n",
    "    ipynbName <- \"230303_LC-M001-proteomics-WGCNA-ver2_WGCNA-module_\"\n",
    "    fileName <- str_c(\"enrichment-\",module,\"-all.pdf\")\n",
    "    ggsave(file=str_c(fileDir,ipynbName,fileName), plot=temp,\n",
    "           width=9, height=max(c(1+display*0.25, 2.5)), units=\"in\")\n",
    "    #(Font family is not reflected in JupyterLab output, but correctly done in .pdf file.)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e1ad095a-589c-4587-bf2a-083dd3725e13",
   "metadata": {},
   "source": [
    "#### 3-2-7. Salmon"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4c015f9f-d0e9-440f-bb47-de97c86a6c88",
   "metadata": {},
   "outputs": [],
   "source": [
    "obj_i <- 7\n",
    "\n",
    "#Retreive results\n",
    "module <- str_to_title(targets[obj_i])#WGCNA module\n",
    "nMembers <- nMembers_vec[obj_i]\n",
    "nMappedMembers <- nMappedMembers_vec[obj_i]\n",
    "nMemberModules <- nMemberModules_vec[obj_i]\n",
    "res <- res_list[[obj_i]]\n",
    "\n",
    "#Check\n",
    "print(module)\n",
    "print(str_c(\" - # of members: \",as.character(nMembers)))\n",
    "print(str_c(\" - # of members mapped to any annotation: \",as.character(nMappedMembers)))\n",
    "print(str_c(\" - # of annotations having any member: \",as.character(nMemberModules)))\n",
    "res\n",
    "print(\" <- Note that the above 'X enriched terms found' is not correct. In this case, X indicates the number of all the tested terms.\")\n",
    "if (is.data.frame(res[])) {\n",
    "    tibble(res[]) %>%\n",
    "        dplyr::filter(pvalue<0.05) %>%#Display only nominal P-value < 0.05\n",
    "        dplyr::rename(AnnotationID=ID, AnnotationName=Description, Ratio2Members=GeneRatio, Ratio2BGs=BgRatio,\n",
    "                      Pval=pvalue, AdjPval=p.adjust, Qval=qvalue, MappedMember=geneID, nMappedMembers=Count)\n",
    "}\n",
    "\n",
    "#Visualization\n",
    "if (is.data.frame(res[])) {\n",
    "    display <- tibble(res[]) %>%\n",
    "        dplyr::filter(pvalue<0.05) %>%\n",
    "        nrow()\n",
    "} else {\n",
    "    display <- 0\n",
    "}\n",
    "if (display>0) {\n",
    "    temp <- res %>%\n",
    "        dplyr::filter(pvalue<0.05) %>%#Display only nominal P-value < 0.05\n",
    "        dplyr::mutate(PvalLabel=str_c(\"AdjPval = \",scales::scientific(p.adjust, digits=2)),\n",
    "                      AdjSignif=ifelse(p.adjust<0.05, \"True\", \"False\")) %>%\n",
    "        barplot(., x=\"Count\", color=\"p.adjust\", showCategory=display) +\n",
    "        geom_text(aes(label=PvalLabel, color=AdjSignif), nudge_x=0.5, hjust=0) +\n",
    "        coord_cartesian(clip=\"off\") +\n",
    "        scale_x_continuous(limits=c(0, 25), breaks=seq(0, 25, by=5), expand=c(0, 0)) +\n",
    "        scale_y_discrete(labels=function(x) {str_wrap(x, width=60)}) +\n",
    "        scale_fill_viridis_c(begin=0, end=1, direction=1, option=\"plasma\",\n",
    "                             limits=c(0, 0.2), breaks=seq(0, 0.2, by=0.05), oob=scales::squish,\n",
    "                             name=\"AdjPval\") +\n",
    "        scale_color_manual(values=c(\"True\"=\"#990000\", \"False\"=\"gray40\"), ) +\n",
    "        guides(fill=guide_colorbar(reverse=TRUE), color=\"none\") +\n",
    "        labs(x=\"Count of members\",\n",
    "             y=\"\", title=str_c(\"Enriched GOBP in \",module,\" module members\")) +\n",
    "        theme_classic(base_size=16, base_family=\"Helvetica\") +\n",
    "        theme(text=element_text(face=\"plain\", color=\"black\", family=\"Helvetica\")) +\n",
    "        theme(axis.text.x=element_text(face=\"plain\", color=\"black\", family=\"Helvetica\"),\n",
    "              axis.text.y=element_text(face=\"plain\", color=\"black\", family=\"Helvetica\", lineheight=0.75),\n",
    "              axis.title=element_text(face=\"plain\", color=\"black\", family=\"Helvetica\")) +\n",
    "        theme(plot.title=element_text(size=18, hjust=1.0)) +\n",
    "        theme(legend.direction=\"vertical\", legend.box=\"horizontal\",\n",
    "              legend.background=element_blank())\n",
    "    options(repr.plot.width=9, repr.plot.height=max(c(1+display*0.25, 2.5)))\n",
    "    plot(temp)\n",
    "    #Save\n",
    "    fileDir <- \"./ExportFigures/\"\n",
    "    ipynbName <- \"230303_LC-M001-proteomics-WGCNA-ver2_WGCNA-module_\"\n",
    "    fileName <- str_c(\"enrichment-\",module,\"-all.pdf\")\n",
    "    ggsave(file=str_c(fileDir,ipynbName,fileName), plot=temp,\n",
    "           width=9, height=max(c(1+display*0.25, 2.5)), units=\"in\")\n",
    "    #(Font family is not reflected in JupyterLab output, but correctly done in .pdf file.)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4c82a37b-ef42-4899-8007-981198d0521c",
   "metadata": {},
   "source": [
    "#### 3-2-8. Tan"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "46f44596-f9fe-4122-9ea0-5b162989d391",
   "metadata": {},
   "outputs": [],
   "source": [
    "obj_i <- 8\n",
    "\n",
    "#Retreive results\n",
    "module <- str_to_title(targets[obj_i])#WGCNA module\n",
    "nMembers <- nMembers_vec[obj_i]\n",
    "nMappedMembers <- nMappedMembers_vec[obj_i]\n",
    "nMemberModules <- nMemberModules_vec[obj_i]\n",
    "res <- res_list[[obj_i]]\n",
    "\n",
    "#Check\n",
    "print(module)\n",
    "print(str_c(\" - # of members: \",as.character(nMembers)))\n",
    "print(str_c(\" - # of members mapped to any annotation: \",as.character(nMappedMembers)))\n",
    "print(str_c(\" - # of annotations having any member: \",as.character(nMemberModules)))\n",
    "res\n",
    "print(\" <- Note that the above 'X enriched terms found' is not correct. In this case, X indicates the number of all the tested terms.\")\n",
    "if (is.data.frame(res[])) {\n",
    "    tibble(res[]) %>%\n",
    "        dplyr::filter(pvalue<0.05) %>%#Display only nominal P-value < 0.05\n",
    "        dplyr::rename(AnnotationID=ID, AnnotationName=Description, Ratio2Members=GeneRatio, Ratio2BGs=BgRatio,\n",
    "                      Pval=pvalue, AdjPval=p.adjust, Qval=qvalue, MappedMember=geneID, nMappedMembers=Count)\n",
    "}\n",
    "\n",
    "#Visualization\n",
    "if (is.data.frame(res[])) {\n",
    "    display <- tibble(res[]) %>%\n",
    "        dplyr::filter(pvalue<0.05) %>%\n",
    "        nrow()\n",
    "} else {\n",
    "    display <- 0\n",
    "}\n",
    "if (display>0) {\n",
    "    temp <- res %>%\n",
    "        dplyr::filter(pvalue<0.05) %>%#Display only nominal P-value < 0.05\n",
    "        dplyr::mutate(PvalLabel=str_c(\"AdjPval = \",scales::scientific(p.adjust, digits=2)),\n",
    "                      AdjSignif=ifelse(p.adjust<0.05, \"True\", \"False\")) %>%\n",
    "        barplot(., x=\"Count\", color=\"p.adjust\", showCategory=display) +\n",
    "        geom_text(aes(label=PvalLabel, color=AdjSignif), nudge_x=0.5, hjust=0) +\n",
    "        coord_cartesian(clip=\"off\") +\n",
    "        scale_x_continuous(limits=c(0, 25), breaks=seq(0, 25, by=5), expand=c(0, 0)) +\n",
    "        scale_y_discrete(labels=function(x) {str_wrap(x, width=60)}) +\n",
    "        scale_fill_viridis_c(begin=0, end=1, direction=1, option=\"plasma\",\n",
    "                             limits=c(0, 0.2), breaks=seq(0, 0.2, by=0.05), oob=scales::squish,\n",
    "                             name=\"AdjPval\") +\n",
    "        scale_color_manual(values=c(\"True\"=\"#990000\", \"False\"=\"gray40\"), ) +\n",
    "        guides(fill=guide_colorbar(reverse=TRUE), color=\"none\") +\n",
    "        labs(x=\"Count of members\",\n",
    "             y=\"\", title=str_c(\"Enriched GOBP in \",module,\" module members\")) +\n",
    "        theme_classic(base_size=16, base_family=\"Helvetica\") +\n",
    "        theme(text=element_text(face=\"plain\", color=\"black\", family=\"Helvetica\")) +\n",
    "        theme(axis.text.x=element_text(face=\"plain\", color=\"black\", family=\"Helvetica\"),\n",
    "              axis.text.y=element_text(face=\"plain\", color=\"black\", family=\"Helvetica\", lineheight=0.75),\n",
    "              axis.title=element_text(face=\"plain\", color=\"black\", family=\"Helvetica\")) +\n",
    "        theme(plot.title=element_text(size=18, hjust=1.0)) +\n",
    "        theme(legend.direction=\"vertical\", legend.box=\"horizontal\",\n",
    "              legend.background=element_blank())\n",
    "    options(repr.plot.width=9, repr.plot.height=max(c(1+display*0.25, 2.5)))\n",
    "    plot(temp)\n",
    "    #Save\n",
    "    fileDir <- \"./ExportFigures/\"\n",
    "    ipynbName <- \"230303_LC-M001-proteomics-WGCNA-ver2_WGCNA-module_\"\n",
    "    fileName <- str_c(\"enrichment-\",module,\"-all.pdf\")\n",
    "    ggsave(file=str_c(fileDir,ipynbName,fileName), plot=temp,\n",
    "           width=9, height=max(c(1+display*0.25, 2.5)), units=\"in\")\n",
    "    #(Font family is not reflected in JupyterLab output, but correctly done in .pdf file.)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2bb2736a-e8a6-47ef-ab43-7e2a0fce0d3b",
   "metadata": {},
   "source": [
    "#### 3-2-9. Brown"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "56aa376e-8371-4673-a889-ea30244e4864",
   "metadata": {},
   "outputs": [],
   "source": [
    "obj_i <- 9\n",
    "\n",
    "#Retreive results\n",
    "module <- str_to_title(targets[obj_i])#WGCNA module\n",
    "nMembers <- nMembers_vec[obj_i]\n",
    "nMappedMembers <- nMappedMembers_vec[obj_i]\n",
    "nMemberModules <- nMemberModules_vec[obj_i]\n",
    "res <- res_list[[obj_i]]\n",
    "\n",
    "#Check\n",
    "print(module)\n",
    "print(str_c(\" - # of members: \",as.character(nMembers)))\n",
    "print(str_c(\" - # of members mapped to any annotation: \",as.character(nMappedMembers)))\n",
    "print(str_c(\" - # of annotations having any member: \",as.character(nMemberModules)))\n",
    "res\n",
    "print(\" <- Note that the above 'X enriched terms found' is not correct. In this case, X indicates the number of all the tested terms.\")\n",
    "if (is.data.frame(res[])) {\n",
    "    tibble(res[]) %>%\n",
    "        dplyr::filter(pvalue<0.05) %>%#Display only nominal P-value < 0.05\n",
    "        dplyr::rename(AnnotationID=ID, AnnotationName=Description, Ratio2Members=GeneRatio, Ratio2BGs=BgRatio,\n",
    "                      Pval=pvalue, AdjPval=p.adjust, Qval=qvalue, MappedMember=geneID, nMappedMembers=Count)\n",
    "}\n",
    "\n",
    "#Visualization\n",
    "if (is.data.frame(res[])) {\n",
    "    display <- tibble(res[]) %>%\n",
    "        dplyr::filter(pvalue<0.05) %>%\n",
    "        nrow()\n",
    "} else {\n",
    "    display <- 0\n",
    "}\n",
    "if (display>0) {\n",
    "    temp <- res %>%\n",
    "        dplyr::filter(pvalue<0.05) %>%#Display only nominal P-value < 0.05\n",
    "        dplyr::mutate(PvalLabel=str_c(\"AdjPval = \",scales::scientific(p.adjust, digits=2)),\n",
    "                      AdjSignif=ifelse(p.adjust<0.05, \"True\", \"False\")) %>%\n",
    "        barplot(., x=\"Count\", color=\"p.adjust\", showCategory=display) +\n",
    "        geom_text(aes(label=PvalLabel, color=AdjSignif), nudge_x=0.5, hjust=0) +\n",
    "        coord_cartesian(clip=\"off\") +\n",
    "        scale_x_continuous(limits=c(0, 25), breaks=seq(0, 25, by=5), expand=c(0, 0)) +\n",
    "        scale_y_discrete(labels=function(x) {str_wrap(x, width=60)}) +\n",
    "        scale_fill_viridis_c(begin=0, end=1, direction=1, option=\"plasma\",\n",
    "                             limits=c(0, 0.2), breaks=seq(0, 0.2, by=0.05), oob=scales::squish,\n",
    "                             name=\"AdjPval\") +\n",
    "        scale_color_manual(values=c(\"True\"=\"#990000\", \"False\"=\"gray40\"), ) +\n",
    "        guides(fill=guide_colorbar(reverse=TRUE), color=\"none\") +\n",
    "        labs(x=\"Count of members\",\n",
    "             y=\"\", title=str_c(\"Enriched GOBP in \",module,\" module members\")) +\n",
    "        theme_classic(base_size=16, base_family=\"Helvetica\") +\n",
    "        theme(text=element_text(face=\"plain\", color=\"black\", family=\"Helvetica\")) +\n",
    "        theme(axis.text.x=element_text(face=\"plain\", color=\"black\", family=\"Helvetica\"),\n",
    "              axis.text.y=element_text(face=\"plain\", color=\"black\", family=\"Helvetica\", lineheight=0.75),\n",
    "              axis.title=element_text(face=\"plain\", color=\"black\", family=\"Helvetica\")) +\n",
    "        theme(plot.title=element_text(size=18, hjust=1.0)) +\n",
    "        theme(legend.direction=\"vertical\", legend.box=\"horizontal\",\n",
    "              legend.background=element_blank())\n",
    "    options(repr.plot.width=9, repr.plot.height=max(c(1+display*0.25, 2.5)))\n",
    "    plot(temp)\n",
    "    #Save\n",
    "    fileDir <- \"./ExportFigures/\"\n",
    "    ipynbName <- \"230303_LC-M001-proteomics-WGCNA-ver2_WGCNA-module_\"\n",
    "    fileName <- str_c(\"enrichment-\",module,\"-all.pdf\")\n",
    "    ggsave(file=str_c(fileDir,ipynbName,fileName), plot=temp,\n",
    "           width=9, height=max(c(1+display*0.25, 2.5)), units=\"in\")\n",
    "    #(Font family is not reflected in JupyterLab output, but correctly done in .pdf file.)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "18960024-0774-4417-988f-992038c89d61",
   "metadata": {},
   "source": [
    "#### 3-2-10. Pink"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8ea13cd0-245b-4ff5-b435-c88ff058fe3f",
   "metadata": {},
   "outputs": [],
   "source": [
    "obj_i <- 10\n",
    "\n",
    "#Retreive results\n",
    "module <- str_to_title(targets[obj_i])#WGCNA module\n",
    "nMembers <- nMembers_vec[obj_i]\n",
    "nMappedMembers <- nMappedMembers_vec[obj_i]\n",
    "nMemberModules <- nMemberModules_vec[obj_i]\n",
    "res <- res_list[[obj_i]]\n",
    "\n",
    "#Check\n",
    "print(module)\n",
    "print(str_c(\" - # of members: \",as.character(nMembers)))\n",
    "print(str_c(\" - # of members mapped to any annotation: \",as.character(nMappedMembers)))\n",
    "print(str_c(\" - # of annotations having any member: \",as.character(nMemberModules)))\n",
    "res\n",
    "print(\" <- Note that the above 'X enriched terms found' is not correct. In this case, X indicates the number of all the tested terms.\")\n",
    "if (is.data.frame(res[])) {\n",
    "    tibble(res[]) %>%\n",
    "        dplyr::filter(pvalue<0.05) %>%#Display only nominal P-value < 0.05\n",
    "        dplyr::rename(AnnotationID=ID, AnnotationName=Description, Ratio2Members=GeneRatio, Ratio2BGs=BgRatio,\n",
    "                      Pval=pvalue, AdjPval=p.adjust, Qval=qvalue, MappedMember=geneID, nMappedMembers=Count)\n",
    "}\n",
    "\n",
    "#Visualization\n",
    "if (is.data.frame(res[])) {\n",
    "    display <- tibble(res[]) %>%\n",
    "        dplyr::filter(pvalue<0.05) %>%\n",
    "        nrow()\n",
    "} else {\n",
    "    display <- 0\n",
    "}\n",
    "if (display>0) {\n",
    "    temp <- res %>%\n",
    "        dplyr::filter(pvalue<0.05) %>%#Display only nominal P-value < 0.05\n",
    "        dplyr::mutate(PvalLabel=str_c(\"AdjPval = \",scales::scientific(p.adjust, digits=2)),\n",
    "                      AdjSignif=ifelse(p.adjust<0.05, \"True\", \"False\")) %>%\n",
    "        barplot(., x=\"Count\", color=\"p.adjust\", showCategory=display) +\n",
    "        geom_text(aes(label=PvalLabel, color=AdjSignif), nudge_x=0.5, hjust=0) +\n",
    "        coord_cartesian(clip=\"off\") +\n",
    "        scale_x_continuous(limits=c(0, 25), breaks=seq(0, 25, by=5), expand=c(0, 0)) +\n",
    "        scale_y_discrete(labels=function(x) {str_wrap(x, width=60)}) +\n",
    "        scale_fill_viridis_c(begin=0, end=1, direction=1, option=\"plasma\",\n",
    "                             limits=c(0, 0.2), breaks=seq(0, 0.2, by=0.05), oob=scales::squish,\n",
    "                             name=\"AdjPval\") +\n",
    "        scale_color_manual(values=c(\"True\"=\"#990000\", \"False\"=\"gray40\"), ) +\n",
    "        guides(fill=guide_colorbar(reverse=TRUE), color=\"none\") +\n",
    "        labs(x=\"Count of members\",\n",
    "             y=\"\", title=str_c(\"Enriched GOBP in \",module,\" module members\")) +\n",
    "        theme_classic(base_size=16, base_family=\"Helvetica\") +\n",
    "        theme(text=element_text(face=\"plain\", color=\"black\", family=\"Helvetica\")) +\n",
    "        theme(axis.text.x=element_text(face=\"plain\", color=\"black\", family=\"Helvetica\"),\n",
    "              axis.text.y=element_text(face=\"plain\", color=\"black\", family=\"Helvetica\", lineheight=0.75),\n",
    "              axis.title=element_text(face=\"plain\", color=\"black\", family=\"Helvetica\")) +\n",
    "        theme(plot.title=element_text(size=18, hjust=1.0)) +\n",
    "        theme(legend.direction=\"vertical\", legend.box=\"horizontal\",\n",
    "              legend.background=element_blank())\n",
    "    options(repr.plot.width=9, repr.plot.height=max(c(1+display*0.25, 2.5)))\n",
    "    plot(temp)\n",
    "    #Save\n",
    "    fileDir <- \"./ExportFigures/\"\n",
    "    ipynbName <- \"230303_LC-M001-proteomics-WGCNA-ver2_WGCNA-module_\"\n",
    "    fileName <- str_c(\"enrichment-\",module,\"-all.pdf\")\n",
    "    ggsave(file=str_c(fileDir,ipynbName,fileName), plot=temp,\n",
    "           width=9, height=max(c(1+display*0.25, 2.5)), units=\"in\")\n",
    "    #(Font family is not reflected in JupyterLab output, but correctly done in .pdf file.)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f88afbc8-577d-43dd-bc4e-63ac616cb715",
   "metadata": {},
   "source": [
    "## 4. Enrichment analysis for top hub proteins"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2c815910-a2ba-4c20-b075-93b27f5634b6",
   "metadata": {},
   "source": [
    "### 4-1. Save result objects"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eba85b20-3921-423e-9cb1-6e84ac7035b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "topQ <- 0.9#Focus on the top hubs\n",
    "\n",
    "#Create a workbook object to save as one single .xlsx file\n",
    "workbook <- createWorkbook()\n",
    "\n",
    "#Summarize results per WGCNA module\n",
    "nHubs_vec <- c()\n",
    "nMappedHubs_vec <- c()\n",
    "nHubModules_vec <- c()\n",
    "res_list <- list()\n",
    "for (i in 1:length(targets)) {\n",
    "    #Prepare input for enricher()\n",
    "    module <- str_to_title(targets[i])#WGCNA module\n",
    "    hubs <- module_tbl %>%\n",
    "        dplyr::filter(ModuleID==module) %>%\n",
    "        dplyr::filter(IntramodularConnectivity>=quantile(IntramodularConnectivity, topQ)) %>%\n",
    "        .$GeneSymbol#Able to use instead of UniProtID because duplicate issues were resolved\n",
    "    backgrounds <- meta_annot %>%\n",
    "        dplyr::select(ModuleID, GeneSymbol)#Here, not WGCNA module but a priori module (functional annotation)\n",
    "    labels <- meta_annot %>%\n",
    "        dplyr::select(ModuleID, ModuleName) %>%#Here, not WGCNA module but a priori module (functional annotation)\n",
    "        dplyr::distinct()\n",
    "    print(str_c(module,\": \",length(hubs),\" hubs\"))\n",
    "    \n",
    "    #Save info\n",
    "    nHubs_vec <- c(nHubs_vec, length(hubs))\n",
    "    temp <- backgrounds %>% dplyr::filter(GeneSymbol %in% hubs)\n",
    "    nMappedHubs_vec <- c(nMappedHubs_vec, length(unique(temp$GeneSymbol)))\n",
    "    nHubModules_vec <- c(nHubModules_vec, length(unique(temp$ModuleID)))\n",
    "    \n",
    "    #Enrichment analysis\n",
    "    temp <- enricher(gene=hubs,\n",
    "                     pvalueCutoff=1.0,#To export all\n",
    "                     pAdjustMethod=\"BH\",\n",
    "                     #universe=backgrounds,#Already managed\n",
    "                     minGSSize=4,\n",
    "                     maxGSSize=10000,\n",
    "                     qvalueCutoff=1.0,#To export all\n",
    "                     TERM2GENE=backgrounds,\n",
    "                     TERM2NAME=labels)\n",
    "    \n",
    "    #Add the summary table to the workbook object as an independent sheet\n",
    "    if (is.data.frame(temp[])) {\n",
    "        temp1 <- tibble(temp[]) %>%\n",
    "            dplyr::rename(AnnotationID=ID, AnnotationName=Description, Ratio2Hubs=GeneRatio, Ratio2BGs=BgRatio,\n",
    "                          Pval=pvalue, AdjPval=p.adjust, Qval=qvalue, MappedHub=geneID, nMappedHubs=Count)\n",
    "    } else {\n",
    "        temp1 <- tibble(`n/a`=NA)\n",
    "    }\n",
    "    addWorksheet(workbook, sheetName=module)\n",
    "    writeData(workbook, module, temp1)\n",
    "    \n",
    "    #Add result object to list\n",
    "    res_list <- c(res_list, list(temp))\n",
    "}\n",
    "\n",
    "#Save the workbook as one single .xlsx file\n",
    "fileDir <- \"./ExportData/\"\n",
    "ipynbName <- \"230303_LC-M001-proteomics-WGCNA-ver2_WGCNA-module_\"\n",
    "fileName <- \"clusterProfiler-hubs.xlsx\"\n",
    "saveWorkbook(workbook, file=str_c(fileDir,ipynbName,fileName), overwrite=TRUE)\n",
    "\n",
    "print(str_c(\"nObjects: \", as.character(length(res_list))))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "055f34dd-d1fc-4c4f-bc58-dfdb41d95ec2",
   "metadata": {},
   "source": [
    "### 4-2. Check results"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "82dae9ab-86e9-41c1-838f-065a4d6e6613",
   "metadata": {},
   "source": [
    "#### 4-2-1. Greenyellow"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "140e378c-f1fb-49d5-9e3f-b0343d9d2c6f",
   "metadata": {},
   "outputs": [],
   "source": [
    "obj_i <- 1\n",
    "\n",
    "#Retreive results\n",
    "module <- str_to_title(targets[obj_i])#WGCNA module\n",
    "nHubs <- nHubs_vec[obj_i]\n",
    "nMappedHubs <- nMappedHubs_vec[obj_i]\n",
    "nHubModules <- nHubModules_vec[obj_i]\n",
    "res <- res_list[[obj_i]]\n",
    "\n",
    "#Check\n",
    "print(module)\n",
    "print(str_c(\" - # of hubs: \",as.character(nHubs)))\n",
    "print(str_c(\" - # of hubs mapped to any annotation: \",as.character(nMappedHubs)))\n",
    "print(str_c(\" - # of annotations having any hub: \",as.character(nHubModules)))\n",
    "res\n",
    "print(\" <- Note that the above 'X enriched terms found' is not correct. In this case, X indicates the number of all the tested terms.\")\n",
    "if (is.data.frame(res[])) {\n",
    "    tibble(res[]) %>%\n",
    "        dplyr::filter(pvalue<0.05) %>%#Display only nominal P-value < 0.05\n",
    "        dplyr::rename(AnnotationID=ID, AnnotationName=Description, Ratio2Hubs=GeneRatio, Ratio2BGs=BgRatio,\n",
    "                      Pval=pvalue, AdjPval=p.adjust, Qval=qvalue, MappedHub=geneID, nMappedHubs=Count)\n",
    "}\n",
    "\n",
    "#Visualization\n",
    "if (is.data.frame(res[])) {\n",
    "    display <- tibble(res[]) %>%\n",
    "        dplyr::filter(pvalue<0.05) %>%\n",
    "        nrow()\n",
    "} else {\n",
    "    display <- 0\n",
    "}\n",
    "if (display>0) {\n",
    "    temp <- res %>%\n",
    "        dplyr::filter(pvalue<0.05) %>%#Display only nominal P-value < 0.05\n",
    "        dplyr::mutate(PvalLabel=str_c(\"AdjPval = \",scales::scientific(p.adjust, digits=2)),\n",
    "                      AdjSignif=ifelse(p.adjust<0.05, \"True\", \"False\")) %>%\n",
    "        barplot(., x=\"Count\", color=\"p.adjust\", showCategory=display) +\n",
    "        geom_text(aes(label=PvalLabel, color=AdjSignif), nudge_x=0.5, hjust=0) +\n",
    "        coord_cartesian(clip=\"off\") +\n",
    "        scale_x_continuous(limits=c(0, 15), breaks=seq(0, 15, by=5), expand=c(0, 0)) +\n",
    "        scale_y_discrete(labels=function(x) {str_wrap(x, width=60)}) +\n",
    "        scale_fill_viridis_c(begin=0, end=1, direction=1, option=\"plasma\",\n",
    "                             limits=c(0, 0.2), breaks=seq(0, 0.2, by=0.05), oob=scales::squish,\n",
    "                             name=\"AdjPval\") +\n",
    "        scale_color_manual(values=c(\"True\"=\"#990000\", \"False\"=\"gray40\"), ) +\n",
    "        guides(fill=guide_colorbar(reverse=TRUE), color=\"none\") +\n",
    "        labs(x=\"Count of hubs\",\n",
    "             y=\"\", title=str_c(\"Enriched GOBP in \",module,\" module hubs\")) +\n",
    "        theme_classic(base_size=16, base_family=\"Helvetica\") +\n",
    "        theme(text=element_text(face=\"plain\", color=\"black\", family=\"Helvetica\")) +\n",
    "        theme(axis.text.x=element_text(face=\"plain\", color=\"black\", family=\"Helvetica\"),\n",
    "              axis.text.y=element_text(face=\"plain\", color=\"black\", family=\"Helvetica\", lineheight=0.75),\n",
    "              axis.title=element_text(face=\"plain\", color=\"black\", family=\"Helvetica\")) +\n",
    "        theme(plot.title=element_text(size=18, hjust=1.0)) +\n",
    "        theme(legend.direction=\"vertical\", legend.box=\"horizontal\",\n",
    "              legend.background=element_blank())\n",
    "    options(repr.plot.width=9, repr.plot.height=max(c(1+display*0.25, 2.5)))\n",
    "    plot(temp)\n",
    "    #Save\n",
    "    fileDir <- \"./ExportFigures/\"\n",
    "    ipynbName <- \"230303_LC-M001-proteomics-WGCNA-ver2_WGCNA-module_\"\n",
    "    fileName <- str_c(\"enrichment-\",module,\"-hub.pdf\")\n",
    "    ggsave(file=str_c(fileDir,ipynbName,fileName), plot=temp,\n",
    "           width=9, height=max(c(1+display*0.25, 2.5)), units=\"in\")\n",
    "    #(Font family is not reflected in JupyterLab output, but correctly done in .pdf file.)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ec3ce3c8-e442-4baf-bb73-a7c03efb5ae0",
   "metadata": {},
   "source": [
    "#### 4-2-2. Turquoise"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1d829bbe-f994-4681-bbec-93bf335c19fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "obj_i <- 2\n",
    "\n",
    "#Retreive results\n",
    "module <- str_to_title(targets[obj_i])#WGCNA module\n",
    "nHubs <- nHubs_vec[obj_i]\n",
    "nMappedHubs <- nMappedHubs_vec[obj_i]\n",
    "nHubModules <- nHubModules_vec[obj_i]\n",
    "res <- res_list[[obj_i]]\n",
    "\n",
    "#Check\n",
    "print(module)\n",
    "print(str_c(\" - # of hubs: \",as.character(nHubs)))\n",
    "print(str_c(\" - # of hubs mapped to any annotation: \",as.character(nMappedHubs)))\n",
    "print(str_c(\" - # of annotations having any hub: \",as.character(nHubModules)))\n",
    "res\n",
    "print(\" <- Note that the above 'X enriched terms found' is not correct. In this case, X indicates the number of all the tested terms.\")\n",
    "if (is.data.frame(res[])) {\n",
    "    tibble(res[]) %>%\n",
    "        dplyr::filter(pvalue<0.05) %>%#Display only nominal P-value < 0.05\n",
    "        dplyr::rename(AnnotationID=ID, AnnotationName=Description, Ratio2Hubs=GeneRatio, Ratio2BGs=BgRatio,\n",
    "                      Pval=pvalue, AdjPval=p.adjust, Qval=qvalue, MappedHub=geneID, nMappedHubs=Count)\n",
    "}\n",
    "\n",
    "#Visualization\n",
    "if (is.data.frame(res[])) {\n",
    "    display <- tibble(res[]) %>%\n",
    "        dplyr::filter(pvalue<0.05) %>%\n",
    "        nrow()\n",
    "} else {\n",
    "    display <- 0\n",
    "}\n",
    "if (display>0) {\n",
    "    temp <- res %>%\n",
    "        dplyr::filter(pvalue<0.05) %>%#Display only nominal P-value < 0.05\n",
    "        dplyr::mutate(PvalLabel=str_c(\"AdjPval = \",scales::scientific(p.adjust, digits=2)),\n",
    "                      AdjSignif=ifelse(p.adjust<0.05, \"True\", \"False\")) %>%\n",
    "        barplot(., x=\"Count\", color=\"p.adjust\", showCategory=display) +\n",
    "        geom_text(aes(label=PvalLabel, color=AdjSignif), nudge_x=0.5, hjust=0) +\n",
    "        coord_cartesian(clip=\"off\") +\n",
    "        scale_x_continuous(limits=c(0, 15), breaks=seq(0, 15, by=5), expand=c(0, 0)) +\n",
    "        scale_y_discrete(labels=function(x) {str_wrap(x, width=60)}) +\n",
    "        scale_fill_viridis_c(begin=0, end=1, direction=1, option=\"plasma\",\n",
    "                             limits=c(0, 0.2), breaks=seq(0, 0.2, by=0.05), oob=scales::squish,\n",
    "                             name=\"AdjPval\") +\n",
    "        scale_color_manual(values=c(\"True\"=\"#990000\", \"False\"=\"gray40\"), ) +\n",
    "        guides(fill=guide_colorbar(reverse=TRUE), color=\"none\") +\n",
    "        labs(x=\"Count of hubs\",\n",
    "             y=\"\", title=str_c(\"Enriched GOBP in \",module,\" module hubs\")) +\n",
    "        theme_classic(base_size=16, base_family=\"Helvetica\") +\n",
    "        theme(text=element_text(face=\"plain\", color=\"black\", family=\"Helvetica\")) +\n",
    "        theme(axis.text.x=element_text(face=\"plain\", color=\"black\", family=\"Helvetica\"),\n",
    "              axis.text.y=element_text(face=\"plain\", color=\"black\", family=\"Helvetica\", lineheight=0.75),\n",
    "              axis.title=element_text(face=\"plain\", color=\"black\", family=\"Helvetica\")) +\n",
    "        theme(plot.title=element_text(size=18, hjust=1.0)) +\n",
    "        theme(legend.direction=\"vertical\", legend.box=\"horizontal\",\n",
    "              legend.background=element_blank())\n",
    "    options(repr.plot.width=9, repr.plot.height=max(c(1+display*0.25, 2.5)))\n",
    "    plot(temp)\n",
    "    #Save\n",
    "    fileDir <- \"./ExportFigures/\"\n",
    "    ipynbName <- \"230303_LC-M001-proteomics-WGCNA-ver2_WGCNA-module_\"\n",
    "    fileName <- str_c(\"enrichment-\",module,\"-hub.pdf\")\n",
    "    ggsave(file=str_c(fileDir,ipynbName,fileName), plot=temp,\n",
    "           width=9, height=max(c(1+display*0.25, 2.5)), units=\"in\")\n",
    "    #(Font family is not reflected in JupyterLab output, but correctly done in .pdf file.)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fd8ac67c-930d-4f77-adde-d94251b0ccce",
   "metadata": {},
   "source": [
    "#### 4-2-3. Yellow"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a043bfd9-5761-4a4b-a811-dee6ee55d7e0",
   "metadata": {},
   "outputs": [],
   "source": [
    "obj_i <- 3\n",
    "\n",
    "#Retreive results\n",
    "module <- str_to_title(targets[obj_i])#WGCNA module\n",
    "nHubs <- nHubs_vec[obj_i]\n",
    "nMappedHubs <- nMappedHubs_vec[obj_i]\n",
    "nHubModules <- nHubModules_vec[obj_i]\n",
    "res <- res_list[[obj_i]]\n",
    "\n",
    "#Check\n",
    "print(module)\n",
    "print(str_c(\" - # of hubs: \",as.character(nHubs)))\n",
    "print(str_c(\" - # of hubs mapped to any annotation: \",as.character(nMappedHubs)))\n",
    "print(str_c(\" - # of annotations having any hub: \",as.character(nHubModules)))\n",
    "res\n",
    "print(\" <- Note that the above 'X enriched terms found' is not correct. In this case, X indicates the number of all the tested terms.\")\n",
    "if (is.data.frame(res[])) {\n",
    "    tibble(res[]) %>%\n",
    "        dplyr::filter(pvalue<0.05) %>%#Display only nominal P-value < 0.05\n",
    "        dplyr::rename(AnnotationID=ID, AnnotationName=Description, Ratio2Hubs=GeneRatio, Ratio2BGs=BgRatio,\n",
    "                      Pval=pvalue, AdjPval=p.adjust, Qval=qvalue, MappedHub=geneID, nMappedHubs=Count)\n",
    "}\n",
    "\n",
    "#Visualization\n",
    "if (is.data.frame(res[])) {\n",
    "    display <- tibble(res[]) %>%\n",
    "        dplyr::filter(pvalue<0.05) %>%\n",
    "        nrow()\n",
    "} else {\n",
    "    display <- 0\n",
    "}\n",
    "if (display>0) {\n",
    "    temp <- res %>%\n",
    "        dplyr::filter(pvalue<0.05) %>%#Display only nominal P-value < 0.05\n",
    "        dplyr::mutate(PvalLabel=str_c(\"AdjPval = \",scales::scientific(p.adjust, digits=2)),\n",
    "                      AdjSignif=ifelse(p.adjust<0.05, \"True\", \"False\")) %>%\n",
    "        barplot(., x=\"Count\", color=\"p.adjust\", showCategory=display) +\n",
    "        geom_text(aes(label=PvalLabel, color=AdjSignif), nudge_x=0.5, hjust=0) +\n",
    "        coord_cartesian(clip=\"off\") +\n",
    "        scale_x_continuous(limits=c(0, 15), breaks=seq(0, 15, by=5), expand=c(0, 0)) +\n",
    "        scale_y_discrete(labels=function(x) {str_wrap(x, width=60)}) +\n",
    "        scale_fill_viridis_c(begin=0, end=1, direction=1, option=\"plasma\",\n",
    "                             limits=c(0, 0.2), breaks=seq(0, 0.2, by=0.05), oob=scales::squish,\n",
    "                             name=\"AdjPval\") +\n",
    "        scale_color_manual(values=c(\"True\"=\"#990000\", \"False\"=\"gray40\"), ) +\n",
    "        guides(fill=guide_colorbar(reverse=TRUE), color=\"none\") +\n",
    "        labs(x=\"Count of hubs\",\n",
    "             y=\"\", title=str_c(\"Enriched GOBP in \",module,\" module hubs\")) +\n",
    "        theme_classic(base_size=16, base_family=\"Helvetica\") +\n",
    "        theme(text=element_text(face=\"plain\", color=\"black\", family=\"Helvetica\")) +\n",
    "        theme(axis.text.x=element_text(face=\"plain\", color=\"black\", family=\"Helvetica\"),\n",
    "              axis.text.y=element_text(face=\"plain\", color=\"black\", family=\"Helvetica\", lineheight=0.75),\n",
    "              axis.title=element_text(face=\"plain\", color=\"black\", family=\"Helvetica\")) +\n",
    "        theme(plot.title=element_text(size=18, hjust=1.0)) +\n",
    "        theme(legend.direction=\"vertical\", legend.box=\"horizontal\",\n",
    "              legend.background=element_blank())\n",
    "    options(repr.plot.width=9, repr.plot.height=max(c(1+display*0.25, 2.5)))\n",
    "    plot(temp)\n",
    "    #Save\n",
    "    fileDir <- \"./ExportFigures/\"\n",
    "    ipynbName <- \"230303_LC-M001-proteomics-WGCNA-ver2_WGCNA-module_\"\n",
    "    fileName <- str_c(\"enrichment-\",module,\"-hub.pdf\")\n",
    "    ggsave(file=str_c(fileDir,ipynbName,fileName), plot=temp,\n",
    "           width=9, height=max(c(1+display*0.25, 2.5)), units=\"in\")\n",
    "    #(Font family is not reflected in JupyterLab output, but correctly done in .pdf file.)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2b4bec53-b70c-4261-821d-bbe1582edeae",
   "metadata": {},
   "source": [
    "#### 4-2-4. Black"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6659c87e-0af0-4879-9111-181cef511e9e",
   "metadata": {},
   "outputs": [],
   "source": [
    "obj_i <- 4\n",
    "\n",
    "#Retreive results\n",
    "module <- str_to_title(targets[obj_i])#WGCNA module\n",
    "nHubs <- nHubs_vec[obj_i]\n",
    "nMappedHubs <- nMappedHubs_vec[obj_i]\n",
    "nHubModules <- nHubModules_vec[obj_i]\n",
    "res <- res_list[[obj_i]]\n",
    "\n",
    "#Check\n",
    "print(module)\n",
    "print(str_c(\" - # of hubs: \",as.character(nHubs)))\n",
    "print(str_c(\" - # of hubs mapped to any annotation: \",as.character(nMappedHubs)))\n",
    "print(str_c(\" - # of annotations having any hub: \",as.character(nHubModules)))\n",
    "res\n",
    "print(\" <- Note that the above 'X enriched terms found' is not correct. In this case, X indicates the number of all the tested terms.\")\n",
    "if (is.data.frame(res[])) {\n",
    "    tibble(res[]) %>%\n",
    "        dplyr::filter(pvalue<0.05) %>%#Display only nominal P-value < 0.05\n",
    "        dplyr::rename(AnnotationID=ID, AnnotationName=Description, Ratio2Hubs=GeneRatio, Ratio2BGs=BgRatio,\n",
    "                      Pval=pvalue, AdjPval=p.adjust, Qval=qvalue, MappedHub=geneID, nMappedHubs=Count)\n",
    "}\n",
    "\n",
    "#Visualization\n",
    "if (is.data.frame(res[])) {\n",
    "    display <- tibble(res[]) %>%\n",
    "        dplyr::filter(pvalue<0.05) %>%\n",
    "        nrow()\n",
    "} else {\n",
    "    display <- 0\n",
    "}\n",
    "if (display>0) {\n",
    "    temp <- res %>%\n",
    "        dplyr::filter(pvalue<0.05) %>%#Display only nominal P-value < 0.05\n",
    "        dplyr::mutate(PvalLabel=str_c(\"AdjPval = \",scales::scientific(p.adjust, digits=2)),\n",
    "                      AdjSignif=ifelse(p.adjust<0.05, \"True\", \"False\")) %>%\n",
    "        barplot(., x=\"Count\", color=\"p.adjust\", showCategory=display) +\n",
    "        geom_text(aes(label=PvalLabel, color=AdjSignif), nudge_x=0.5, hjust=0) +\n",
    "        coord_cartesian(clip=\"off\") +\n",
    "        scale_x_continuous(limits=c(0, 15), breaks=seq(0, 15, by=5), expand=c(0, 0)) +\n",
    "        scale_y_discrete(labels=function(x) {str_wrap(x, width=60)}) +\n",
    "        scale_fill_viridis_c(begin=0, end=1, direction=1, option=\"plasma\",\n",
    "                             limits=c(0, 0.2), breaks=seq(0, 0.2, by=0.05), oob=scales::squish,\n",
    "                             name=\"AdjPval\") +\n",
    "        scale_color_manual(values=c(\"True\"=\"#990000\", \"False\"=\"gray40\"), ) +\n",
    "        guides(fill=guide_colorbar(reverse=TRUE), color=\"none\") +\n",
    "        labs(x=\"Count of hubs\",\n",
    "             y=\"\", title=str_c(\"Enriched GOBP in \",module,\" module hubs\")) +\n",
    "        theme_classic(base_size=16, base_family=\"Helvetica\") +\n",
    "        theme(text=element_text(face=\"plain\", color=\"black\", family=\"Helvetica\")) +\n",
    "        theme(axis.text.x=element_text(face=\"plain\", color=\"black\", family=\"Helvetica\"),\n",
    "              axis.text.y=element_text(face=\"plain\", color=\"black\", family=\"Helvetica\", lineheight=0.75),\n",
    "              axis.title=element_text(face=\"plain\", color=\"black\", family=\"Helvetica\")) +\n",
    "        theme(plot.title=element_text(size=18, hjust=1.0)) +\n",
    "        theme(legend.direction=\"vertical\", legend.box=\"horizontal\",\n",
    "              legend.background=element_blank())\n",
    "    options(repr.plot.width=9, repr.plot.height=max(c(1+display*0.25, 2.5)))\n",
    "    plot(temp)\n",
    "    #Save\n",
    "    fileDir <- \"./ExportFigures/\"\n",
    "    ipynbName <- \"230303_LC-M001-proteomics-WGCNA-ver2_WGCNA-module_\"\n",
    "    fileName <- str_c(\"enrichment-\",module,\"-hub.pdf\")\n",
    "    ggsave(file=str_c(fileDir,ipynbName,fileName), plot=temp,\n",
    "           width=9, height=max(c(1+display*0.25, 2.5)), units=\"in\")\n",
    "    #(Font family is not reflected in JupyterLab output, but correctly done in .pdf file.)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b5a4b7f0-2e64-48d1-8420-a96da3e49a54",
   "metadata": {},
   "source": [
    "#### 4-2-5. Blue"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1cd6c101-11e1-48c6-8c21-26ff435100dc",
   "metadata": {},
   "outputs": [],
   "source": [
    "obj_i <- 5\n",
    "\n",
    "#Retreive results\n",
    "module <- str_to_title(targets[obj_i])#WGCNA module\n",
    "nHubs <- nHubs_vec[obj_i]\n",
    "nMappedHubs <- nMappedHubs_vec[obj_i]\n",
    "nHubModules <- nHubModules_vec[obj_i]\n",
    "res <- res_list[[obj_i]]\n",
    "\n",
    "#Check\n",
    "print(module)\n",
    "print(str_c(\" - # of hubs: \",as.character(nHubs)))\n",
    "print(str_c(\" - # of hubs mapped to any annotation: \",as.character(nMappedHubs)))\n",
    "print(str_c(\" - # of annotations having any hub: \",as.character(nHubModules)))\n",
    "res\n",
    "print(\" <- Note that the above 'X enriched terms found' is not correct. In this case, X indicates the number of all the tested terms.\")\n",
    "if (is.data.frame(res[])) {\n",
    "    tibble(res[]) %>%\n",
    "        dplyr::filter(pvalue<0.05) %>%#Display only nominal P-value < 0.05\n",
    "        dplyr::rename(AnnotationID=ID, AnnotationName=Description, Ratio2Hubs=GeneRatio, Ratio2BGs=BgRatio,\n",
    "                      Pval=pvalue, AdjPval=p.adjust, Qval=qvalue, MappedHub=geneID, nMappedHubs=Count)\n",
    "}\n",
    "\n",
    "#Visualization\n",
    "if (is.data.frame(res[])) {\n",
    "    display <- tibble(res[]) %>%\n",
    "        dplyr::filter(pvalue<0.05) %>%\n",
    "        nrow()\n",
    "} else {\n",
    "    display <- 0\n",
    "}\n",
    "if (display>0) {\n",
    "    temp <- res %>%\n",
    "        dplyr::filter(pvalue<0.05) %>%#Display only nominal P-value < 0.05\n",
    "        dplyr::mutate(PvalLabel=str_c(\"AdjPval = \",scales::scientific(p.adjust, digits=2)),\n",
    "                      AdjSignif=ifelse(p.adjust<0.05, \"True\", \"False\")) %>%\n",
    "        barplot(., x=\"Count\", color=\"p.adjust\", showCategory=display) +\n",
    "        geom_text(aes(label=PvalLabel, color=AdjSignif), nudge_x=0.5, hjust=0) +\n",
    "        coord_cartesian(clip=\"off\") +\n",
    "        scale_x_continuous(limits=c(0, 15), breaks=seq(0, 15, by=5), expand=c(0, 0)) +\n",
    "        scale_y_discrete(labels=function(x) {str_wrap(x, width=60)}) +\n",
    "        scale_fill_viridis_c(begin=0, end=1, direction=1, option=\"plasma\",\n",
    "                             limits=c(0, 0.2), breaks=seq(0, 0.2, by=0.05), oob=scales::squish,\n",
    "                             name=\"AdjPval\") +\n",
    "        scale_color_manual(values=c(\"True\"=\"#990000\", \"False\"=\"gray40\"), ) +\n",
    "        guides(fill=guide_colorbar(reverse=TRUE), color=\"none\") +\n",
    "        labs(x=\"Count of hubs\",\n",
    "             y=\"\", title=str_c(\"Enriched GOBP in \",module,\" module hubs\")) +\n",
    "        theme_classic(base_size=16, base_family=\"Helvetica\") +\n",
    "        theme(text=element_text(face=\"plain\", color=\"black\", family=\"Helvetica\")) +\n",
    "        theme(axis.text.x=element_text(face=\"plain\", color=\"black\", family=\"Helvetica\"),\n",
    "              axis.text.y=element_text(face=\"plain\", color=\"black\", family=\"Helvetica\", lineheight=0.75),\n",
    "              axis.title=element_text(face=\"plain\", color=\"black\", family=\"Helvetica\")) +\n",
    "        theme(plot.title=element_text(size=18, hjust=1.0)) +\n",
    "        theme(legend.direction=\"vertical\", legend.box=\"horizontal\",\n",
    "              legend.background=element_blank())\n",
    "    options(repr.plot.width=9, repr.plot.height=max(c(1+display*0.25, 2.5)))\n",
    "    plot(temp)\n",
    "    #Save\n",
    "    fileDir <- \"./ExportFigures/\"\n",
    "    ipynbName <- \"230303_LC-M001-proteomics-WGCNA-ver2_WGCNA-module_\"\n",
    "    fileName <- str_c(\"enrichment-\",module,\"-hub.pdf\")\n",
    "    ggsave(file=str_c(fileDir,ipynbName,fileName), plot=temp,\n",
    "           width=9, height=max(c(1+display*0.25, 2.5)), units=\"in\")\n",
    "    #(Font family is not reflected in JupyterLab output, but correctly done in .pdf file.)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2856bf31-2da6-453c-9251-c149c2b39791",
   "metadata": {},
   "source": [
    "#### 4-2-6. Purple"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1c7d706b-7983-4064-a2fc-d2c23fd4356a",
   "metadata": {},
   "outputs": [],
   "source": [
    "obj_i <- 6\n",
    "\n",
    "#Retreive results\n",
    "module <- str_to_title(targets[obj_i])#WGCNA module\n",
    "nHubs <- nHubs_vec[obj_i]\n",
    "nMappedHubs <- nMappedHubs_vec[obj_i]\n",
    "nHubModules <- nHubModules_vec[obj_i]\n",
    "res <- res_list[[obj_i]]\n",
    "\n",
    "#Check\n",
    "print(module)\n",
    "print(str_c(\" - # of hubs: \",as.character(nHubs)))\n",
    "print(str_c(\" - # of hubs mapped to any annotation: \",as.character(nMappedHubs)))\n",
    "print(str_c(\" - # of annotations having any hub: \",as.character(nHubModules)))\n",
    "res\n",
    "print(\" <- Note that the above 'X enriched terms found' is not correct. In this case, X indicates the number of all the tested terms.\")\n",
    "if (is.data.frame(res[])) {\n",
    "    tibble(res[]) %>%\n",
    "        dplyr::filter(pvalue<0.05) %>%#Display only nominal P-value < 0.05\n",
    "        dplyr::rename(AnnotationID=ID, AnnotationName=Description, Ratio2Hubs=GeneRatio, Ratio2BGs=BgRatio,\n",
    "                      Pval=pvalue, AdjPval=p.adjust, Qval=qvalue, MappedHub=geneID, nMappedHubs=Count)\n",
    "}\n",
    "\n",
    "#Visualization\n",
    "if (is.data.frame(res[])) {\n",
    "    display <- tibble(res[]) %>%\n",
    "        dplyr::filter(pvalue<0.05) %>%\n",
    "        nrow()\n",
    "} else {\n",
    "    display <- 0\n",
    "}\n",
    "if (display>0) {\n",
    "    temp <- res %>%\n",
    "        dplyr::filter(pvalue<0.05) %>%#Display only nominal P-value < 0.05\n",
    "        dplyr::mutate(PvalLabel=str_c(\"AdjPval = \",scales::scientific(p.adjust, digits=2)),\n",
    "                      AdjSignif=ifelse(p.adjust<0.05, \"True\", \"False\")) %>%\n",
    "        barplot(., x=\"Count\", color=\"p.adjust\", showCategory=display) +\n",
    "        geom_text(aes(label=PvalLabel, color=AdjSignif), nudge_x=0.5, hjust=0) +\n",
    "        coord_cartesian(clip=\"off\") +\n",
    "        scale_x_continuous(limits=c(0, 15), breaks=seq(0, 15, by=5), expand=c(0, 0)) +\n",
    "        scale_y_discrete(labels=function(x) {str_wrap(x, width=60)}) +\n",
    "        scale_fill_viridis_c(begin=0, end=1, direction=1, option=\"plasma\",\n",
    "                             limits=c(0, 0.2), breaks=seq(0, 0.2, by=0.05), oob=scales::squish,\n",
    "                             name=\"AdjPval\") +\n",
    "        scale_color_manual(values=c(\"True\"=\"#990000\", \"False\"=\"gray40\"), ) +\n",
    "        guides(fill=guide_colorbar(reverse=TRUE), color=\"none\") +\n",
    "        labs(x=\"Count of hubs\",\n",
    "             y=\"\", title=str_c(\"Enriched GOBP in \",module,\" module hubs\")) +\n",
    "        theme_classic(base_size=16, base_family=\"Helvetica\") +\n",
    "        theme(text=element_text(face=\"plain\", color=\"black\", family=\"Helvetica\")) +\n",
    "        theme(axis.text.x=element_text(face=\"plain\", color=\"black\", family=\"Helvetica\"),\n",
    "              axis.text.y=element_text(face=\"plain\", color=\"black\", family=\"Helvetica\", lineheight=0.75),\n",
    "              axis.title=element_text(face=\"plain\", color=\"black\", family=\"Helvetica\")) +\n",
    "        theme(plot.title=element_text(size=18, hjust=1.0)) +\n",
    "        theme(legend.direction=\"vertical\", legend.box=\"horizontal\",\n",
    "              legend.background=element_blank())\n",
    "    options(repr.plot.width=9, repr.plot.height=max(c(1+display*0.25, 2.5)))\n",
    "    plot(temp)\n",
    "    #Save\n",
    "    fileDir <- \"./ExportFigures/\"\n",
    "    ipynbName <- \"230303_LC-M001-proteomics-WGCNA-ver2_WGCNA-module_\"\n",
    "    fileName <- str_c(\"enrichment-\",module,\"-hub.pdf\")\n",
    "    ggsave(file=str_c(fileDir,ipynbName,fileName), plot=temp,\n",
    "           width=9, height=max(c(1+display*0.25, 2.5)), units=\"in\")\n",
    "    #(Font family is not reflected in JupyterLab output, but correctly done in .pdf file.)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "70ea6e44-adbb-43fc-99b3-9aafaca96f6f",
   "metadata": {},
   "source": [
    "#### 4-2-7. Salmon"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b3664587-61a3-435f-bc76-1ec109486c14",
   "metadata": {},
   "outputs": [],
   "source": [
    "obj_i <- 7\n",
    "\n",
    "#Retreive results\n",
    "module <- str_to_title(targets[obj_i])#WGCNA module\n",
    "nHubs <- nHubs_vec[obj_i]\n",
    "nMappedHubs <- nMappedHubs_vec[obj_i]\n",
    "nHubModules <- nHubModules_vec[obj_i]\n",
    "res <- res_list[[obj_i]]\n",
    "\n",
    "#Check\n",
    "print(module)\n",
    "print(str_c(\" - # of hubs: \",as.character(nHubs)))\n",
    "print(str_c(\" - # of hubs mapped to any annotation: \",as.character(nMappedHubs)))\n",
    "print(str_c(\" - # of annotations having any hub: \",as.character(nHubModules)))\n",
    "res\n",
    "print(\" <- Note that the above 'X enriched terms found' is not correct. In this case, X indicates the number of all the tested terms.\")\n",
    "if (is.data.frame(res[])) {\n",
    "    tibble(res[]) %>%\n",
    "        dplyr::filter(pvalue<0.05) %>%#Display only nominal P-value < 0.05\n",
    "        dplyr::rename(AnnotationID=ID, AnnotationName=Description, Ratio2Hubs=GeneRatio, Ratio2BGs=BgRatio,\n",
    "                      Pval=pvalue, AdjPval=p.adjust, Qval=qvalue, MappedHub=geneID, nMappedHubs=Count)\n",
    "}\n",
    "\n",
    "#Visualization\n",
    "if (is.data.frame(res[])) {\n",
    "    display <- tibble(res[]) %>%\n",
    "        dplyr::filter(pvalue<0.05) %>%\n",
    "        nrow()\n",
    "} else {\n",
    "    display <- 0\n",
    "}\n",
    "if (display>0) {\n",
    "    temp <- res %>%\n",
    "        dplyr::filter(pvalue<0.05) %>%#Display only nominal P-value < 0.05\n",
    "        dplyr::mutate(PvalLabel=str_c(\"AdjPval = \",scales::scientific(p.adjust, digits=2)),\n",
    "                      AdjSignif=ifelse(p.adjust<0.05, \"True\", \"False\")) %>%\n",
    "        barplot(., x=\"Count\", color=\"p.adjust\", showCategory=display) +\n",
    "        geom_text(aes(label=PvalLabel, color=AdjSignif), nudge_x=0.5, hjust=0) +\n",
    "        coord_cartesian(clip=\"off\") +\n",
    "        scale_x_continuous(limits=c(0, 15), breaks=seq(0, 15, by=5), expand=c(0, 0)) +\n",
    "        scale_y_discrete(labels=function(x) {str_wrap(x, width=60)}) +\n",
    "        scale_fill_viridis_c(begin=0, end=1, direction=1, option=\"plasma\",\n",
    "                             limits=c(0, 0.2), breaks=seq(0, 0.2, by=0.05), oob=scales::squish,\n",
    "                             name=\"AdjPval\") +\n",
    "        scale_color_manual(values=c(\"True\"=\"#990000\", \"False\"=\"gray40\"), ) +\n",
    "        guides(fill=guide_colorbar(reverse=TRUE), color=\"none\") +\n",
    "        labs(x=\"Count of hubs\",\n",
    "             y=\"\", title=str_c(\"Enriched GOBP in \",module,\" module hubs\")) +\n",
    "        theme_classic(base_size=16, base_family=\"Helvetica\") +\n",
    "        theme(text=element_text(face=\"plain\", color=\"black\", family=\"Helvetica\")) +\n",
    "        theme(axis.text.x=element_text(face=\"plain\", color=\"black\", family=\"Helvetica\"),\n",
    "              axis.text.y=element_text(face=\"plain\", color=\"black\", family=\"Helvetica\", lineheight=0.75),\n",
    "              axis.title=element_text(face=\"plain\", color=\"black\", family=\"Helvetica\")) +\n",
    "        theme(plot.title=element_text(size=18, hjust=1.0)) +\n",
    "        theme(legend.direction=\"vertical\", legend.box=\"horizontal\",\n",
    "              legend.background=element_blank())\n",
    "    options(repr.plot.width=9, repr.plot.height=max(c(1+display*0.25, 2.5)))\n",
    "    plot(temp)\n",
    "    #Save\n",
    "    fileDir <- \"./ExportFigures/\"\n",
    "    ipynbName <- \"230303_LC-M001-proteomics-WGCNA-ver2_WGCNA-module_\"\n",
    "    fileName <- str_c(\"enrichment-\",module,\"-hub.pdf\")\n",
    "    ggsave(file=str_c(fileDir,ipynbName,fileName), plot=temp,\n",
    "           width=9, height=max(c(1+display*0.25, 2.5)), units=\"in\")\n",
    "    #(Font family is not reflected in JupyterLab output, but correctly done in .pdf file.)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f8710f1a-9406-417f-a97a-4f115ced5615",
   "metadata": {},
   "source": [
    "#### 4-2-8. Tan"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b56570eb-cd1b-4c2c-89e0-87a4ea6cb2e6",
   "metadata": {},
   "outputs": [],
   "source": [
    "obj_i <- 8\n",
    "\n",
    "#Retreive results\n",
    "module <- str_to_title(targets[obj_i])#WGCNA module\n",
    "nHubs <- nHubs_vec[obj_i]\n",
    "nMappedHubs <- nMappedHubs_vec[obj_i]\n",
    "nHubModules <- nHubModules_vec[obj_i]\n",
    "res <- res_list[[obj_i]]\n",
    "\n",
    "#Check\n",
    "print(module)\n",
    "print(str_c(\" - # of hubs: \",as.character(nHubs)))\n",
    "print(str_c(\" - # of hubs mapped to any annotation: \",as.character(nMappedHubs)))\n",
    "print(str_c(\" - # of annotations having any hub: \",as.character(nHubModules)))\n",
    "res\n",
    "print(\" <- Note that the above 'X enriched terms found' is not correct. In this case, X indicates the number of all the tested terms.\")\n",
    "if (is.data.frame(res[])) {\n",
    "    tibble(res[]) %>%\n",
    "        dplyr::filter(pvalue<0.05) %>%#Display only nominal P-value < 0.05\n",
    "        dplyr::rename(AnnotationID=ID, AnnotationName=Description, Ratio2Hubs=GeneRatio, Ratio2BGs=BgRatio,\n",
    "                      Pval=pvalue, AdjPval=p.adjust, Qval=qvalue, MappedHub=geneID, nMappedHubs=Count)\n",
    "}\n",
    "\n",
    "#Visualization\n",
    "if (is.data.frame(res[])) {\n",
    "    display <- tibble(res[]) %>%\n",
    "        dplyr::filter(pvalue<0.05) %>%\n",
    "        nrow()\n",
    "} else {\n",
    "    display <- 0\n",
    "}\n",
    "if (display>0) {\n",
    "    temp <- res %>%\n",
    "        dplyr::filter(pvalue<0.05) %>%#Display only nominal P-value < 0.05\n",
    "        dplyr::mutate(PvalLabel=str_c(\"AdjPval = \",scales::scientific(p.adjust, digits=2)),\n",
    "                      AdjSignif=ifelse(p.adjust<0.05, \"True\", \"False\")) %>%\n",
    "        barplot(., x=\"Count\", color=\"p.adjust\", showCategory=display) +\n",
    "        geom_text(aes(label=PvalLabel, color=AdjSignif), nudge_x=0.5, hjust=0) +\n",
    "        coord_cartesian(clip=\"off\") +\n",
    "        scale_x_continuous(limits=c(0, 15), breaks=seq(0, 15, by=5), expand=c(0, 0)) +\n",
    "        scale_y_discrete(labels=function(x) {str_wrap(x, width=60)}) +\n",
    "        scale_fill_viridis_c(begin=0, end=1, direction=1, option=\"plasma\",\n",
    "                             limits=c(0, 0.2), breaks=seq(0, 0.2, by=0.05), oob=scales::squish,\n",
    "                             name=\"AdjPval\") +\n",
    "        scale_color_manual(values=c(\"True\"=\"#990000\", \"False\"=\"gray40\"), ) +\n",
    "        guides(fill=guide_colorbar(reverse=TRUE), color=\"none\") +\n",
    "        labs(x=\"Count of hubs\",\n",
    "             y=\"\", title=str_c(\"Enriched GOBP in \",module,\" module hubs\")) +\n",
    "        theme_classic(base_size=16, base_family=\"Helvetica\") +\n",
    "        theme(text=element_text(face=\"plain\", color=\"black\", family=\"Helvetica\")) +\n",
    "        theme(axis.text.x=element_text(face=\"plain\", color=\"black\", family=\"Helvetica\"),\n",
    "              axis.text.y=element_text(face=\"plain\", color=\"black\", family=\"Helvetica\", lineheight=0.75),\n",
    "              axis.title=element_text(face=\"plain\", color=\"black\", family=\"Helvetica\")) +\n",
    "        theme(plot.title=element_text(size=18, hjust=1.0)) +\n",
    "        theme(legend.direction=\"vertical\", legend.box=\"horizontal\",\n",
    "              legend.background=element_blank())\n",
    "    options(repr.plot.width=9, repr.plot.height=max(c(1+display*0.25, 2.5)))\n",
    "    plot(temp)\n",
    "    #Save\n",
    "    fileDir <- \"./ExportFigures/\"\n",
    "    ipynbName <- \"230303_LC-M001-proteomics-WGCNA-ver2_WGCNA-module_\"\n",
    "    fileName <- str_c(\"enrichment-\",module,\"-hub.pdf\")\n",
    "    ggsave(file=str_c(fileDir,ipynbName,fileName), plot=temp,\n",
    "           width=9, height=max(c(1+display*0.25, 2.5)), units=\"in\")\n",
    "    #(Font family is not reflected in JupyterLab output, but correctly done in .pdf file.)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4dfdcfeb-4efd-4f95-b41e-957af54a6f3b",
   "metadata": {},
   "source": [
    "#### 4-2-9. Brown"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "94c214b9-d88f-4d36-987f-2639dc0964d2",
   "metadata": {},
   "outputs": [],
   "source": [
    "obj_i <- 9\n",
    "\n",
    "#Retreive results\n",
    "module <- str_to_title(targets[obj_i])#WGCNA module\n",
    "nHubs <- nHubs_vec[obj_i]\n",
    "nMappedHubs <- nMappedHubs_vec[obj_i]\n",
    "nHubModules <- nHubModules_vec[obj_i]\n",
    "res <- res_list[[obj_i]]\n",
    "\n",
    "#Check\n",
    "print(module)\n",
    "print(str_c(\" - # of hubs: \",as.character(nHubs)))\n",
    "print(str_c(\" - # of hubs mapped to any annotation: \",as.character(nMappedHubs)))\n",
    "print(str_c(\" - # of annotations having any hub: \",as.character(nHubModules)))\n",
    "res\n",
    "print(\" <- Note that the above 'X enriched terms found' is not correct. In this case, X indicates the number of all the tested terms.\")\n",
    "if (is.data.frame(res[])) {\n",
    "    tibble(res[]) %>%\n",
    "        dplyr::filter(pvalue<0.05) %>%#Display only nominal P-value < 0.05\n",
    "        dplyr::rename(AnnotationID=ID, AnnotationName=Description, Ratio2Hubs=GeneRatio, Ratio2BGs=BgRatio,\n",
    "                      Pval=pvalue, AdjPval=p.adjust, Qval=qvalue, MappedHub=geneID, nMappedHubs=Count)\n",
    "}\n",
    "\n",
    "#Visualization\n",
    "if (is.data.frame(res[])) {\n",
    "    display <- tibble(res[]) %>%\n",
    "        dplyr::filter(pvalue<0.05) %>%\n",
    "        nrow()\n",
    "} else {\n",
    "    display <- 0\n",
    "}\n",
    "if (display>0) {\n",
    "    temp <- res %>%\n",
    "        dplyr::filter(pvalue<0.05) %>%#Display only nominal P-value < 0.05\n",
    "        dplyr::mutate(PvalLabel=str_c(\"AdjPval = \",scales::scientific(p.adjust, digits=2)),\n",
    "                      AdjSignif=ifelse(p.adjust<0.05, \"True\", \"False\")) %>%\n",
    "        barplot(., x=\"Count\", color=\"p.adjust\", showCategory=display) +\n",
    "        geom_text(aes(label=PvalLabel, color=AdjSignif), nudge_x=0.5, hjust=0) +\n",
    "        coord_cartesian(clip=\"off\") +\n",
    "        scale_x_continuous(limits=c(0, 15), breaks=seq(0, 15, by=5), expand=c(0, 0)) +\n",
    "        scale_y_discrete(labels=function(x) {str_wrap(x, width=60)}) +\n",
    "        scale_fill_viridis_c(begin=0, end=1, direction=1, option=\"plasma\",\n",
    "                             limits=c(0, 0.2), breaks=seq(0, 0.2, by=0.05), oob=scales::squish,\n",
    "                             name=\"AdjPval\") +\n",
    "        scale_color_manual(values=c(\"True\"=\"#990000\", \"False\"=\"gray40\"), ) +\n",
    "        guides(fill=guide_colorbar(reverse=TRUE), color=\"none\") +\n",
    "        labs(x=\"Count of hubs\",\n",
    "             y=\"\", title=str_c(\"Enriched GOBP in \",module,\" module hubs\")) +\n",
    "        theme_classic(base_size=16, base_family=\"Helvetica\") +\n",
    "        theme(text=element_text(face=\"plain\", color=\"black\", family=\"Helvetica\")) +\n",
    "        theme(axis.text.x=element_text(face=\"plain\", color=\"black\", family=\"Helvetica\"),\n",
    "              axis.text.y=element_text(face=\"plain\", color=\"black\", family=\"Helvetica\", lineheight=0.75),\n",
    "              axis.title=element_text(face=\"plain\", color=\"black\", family=\"Helvetica\")) +\n",
    "        theme(plot.title=element_text(size=18, hjust=1.0)) +\n",
    "        theme(legend.direction=\"vertical\", legend.box=\"horizontal\",\n",
    "              legend.background=element_blank())\n",
    "    options(repr.plot.width=9, repr.plot.height=max(c(1+display*0.25, 2.5)))\n",
    "    plot(temp)\n",
    "    #Save\n",
    "    fileDir <- \"./ExportFigures/\"\n",
    "    ipynbName <- \"230303_LC-M001-proteomics-WGCNA-ver2_WGCNA-module_\"\n",
    "    fileName <- str_c(\"enrichment-\",module,\"-hub.pdf\")\n",
    "    ggsave(file=str_c(fileDir,ipynbName,fileName), plot=temp,\n",
    "           width=9, height=max(c(1+display*0.25, 2.5)), units=\"in\")\n",
    "    #(Font family is not reflected in JupyterLab output, but correctly done in .pdf file.)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2b583c25-63e1-4375-bb4a-cab309df081a",
   "metadata": {},
   "source": [
    "#### 4-2-10. Pink"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1398d1cf-d3d1-4564-b0ec-a9d8c69d79eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "obj_i <- 10\n",
    "\n",
    "#Retreive results\n",
    "module <- str_to_title(targets[obj_i])#WGCNA module\n",
    "nHubs <- nHubs_vec[obj_i]\n",
    "nMappedHubs <- nMappedHubs_vec[obj_i]\n",
    "nHubModules <- nHubModules_vec[obj_i]\n",
    "res <- res_list[[obj_i]]\n",
    "\n",
    "#Check\n",
    "print(module)\n",
    "print(str_c(\" - # of hubs: \",as.character(nHubs)))\n",
    "print(str_c(\" - # of hubs mapped to any annotation: \",as.character(nMappedHubs)))\n",
    "print(str_c(\" - # of annotations having any hub: \",as.character(nHubModules)))\n",
    "res\n",
    "print(\" <- Note that the above 'X enriched terms found' is not correct. In this case, X indicates the number of all the tested terms.\")\n",
    "if (is.data.frame(res[])) {\n",
    "    tibble(res[]) %>%\n",
    "        dplyr::filter(pvalue<0.05) %>%#Display only nominal P-value < 0.05\n",
    "        dplyr::rename(AnnotationID=ID, AnnotationName=Description, Ratio2Hubs=GeneRatio, Ratio2BGs=BgRatio,\n",
    "                      Pval=pvalue, AdjPval=p.adjust, Qval=qvalue, MappedHub=geneID, nMappedHubs=Count)\n",
    "}\n",
    "\n",
    "#Visualization\n",
    "if (is.data.frame(res[])) {\n",
    "    display <- tibble(res[]) %>%\n",
    "        dplyr::filter(pvalue<0.05) %>%\n",
    "        nrow()\n",
    "} else {\n",
    "    display <- 0\n",
    "}\n",
    "if (display>0) {\n",
    "    temp <- res %>%\n",
    "        dplyr::filter(pvalue<0.05) %>%#Display only nominal P-value < 0.05\n",
    "        dplyr::mutate(PvalLabel=str_c(\"AdjPval = \",scales::scientific(p.adjust, digits=2)),\n",
    "                      AdjSignif=ifelse(p.adjust<0.05, \"True\", \"False\")) %>%\n",
    "        barplot(., x=\"Count\", color=\"p.adjust\", showCategory=display) +\n",
    "        geom_text(aes(label=PvalLabel, color=AdjSignif), nudge_x=0.5, hjust=0) +\n",
    "        coord_cartesian(clip=\"off\") +\n",
    "        scale_x_continuous(limits=c(0, 15), breaks=seq(0, 15, by=5), expand=c(0, 0)) +\n",
    "        scale_y_discrete(labels=function(x) {str_wrap(x, width=60)}) +\n",
    "        scale_fill_viridis_c(begin=0, end=1, direction=1, option=\"plasma\",\n",
    "                             limits=c(0, 0.2), breaks=seq(0, 0.2, by=0.05), oob=scales::squish,\n",
    "                             name=\"AdjPval\") +\n",
    "        scale_color_manual(values=c(\"True\"=\"#990000\", \"False\"=\"gray40\"), ) +\n",
    "        guides(fill=guide_colorbar(reverse=TRUE), color=\"none\") +\n",
    "        labs(x=\"Count of hubs\",\n",
    "             y=\"\", title=str_c(\"Enriched GOBP in \",module,\" module hubs\")) +\n",
    "        theme_classic(base_size=16, base_family=\"Helvetica\") +\n",
    "        theme(text=element_text(face=\"plain\", color=\"black\", family=\"Helvetica\")) +\n",
    "        theme(axis.text.x=element_text(face=\"plain\", color=\"black\", family=\"Helvetica\"),\n",
    "              axis.text.y=element_text(face=\"plain\", color=\"black\", family=\"Helvetica\", lineheight=0.75),\n",
    "              axis.title=element_text(face=\"plain\", color=\"black\", family=\"Helvetica\")) +\n",
    "        theme(plot.title=element_text(size=18, hjust=1.0)) +\n",
    "        theme(legend.direction=\"vertical\", legend.box=\"horizontal\",\n",
    "              legend.background=element_blank())\n",
    "    options(repr.plot.width=9, repr.plot.height=max(c(1+display*0.25, 2.5)))\n",
    "    plot(temp)\n",
    "    #Save\n",
    "    fileDir <- \"./ExportFigures/\"\n",
    "    ipynbName <- \"230303_LC-M001-proteomics-WGCNA-ver2_WGCNA-module_\"\n",
    "    fileName <- str_c(\"enrichment-\",module,\"-hub.pdf\")\n",
    "    ggsave(file=str_c(fileDir,ipynbName,fileName), plot=temp,\n",
    "           width=9, height=max(c(1+display*0.25, 2.5)), units=\"in\")\n",
    "    #(Font family is not reflected in JupyterLab output, but correctly done in .pdf file.)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "71adff3d-628b-4d88-817e-1450a1d14d84",
   "metadata": {},
   "source": [
    "# — Session information —"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "5e55c69b-50c9-49fb-8216-6d54a0c9a7ea",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "R version 4.1.1 (2021-08-10)\n",
       "Platform: x86_64-conda-linux-gnu (64-bit)\n",
       "Running under: Ubuntu 20.04.3 LTS\n",
       "\n",
       "Matrix products: default\n",
       "BLAS/LAPACK: /opt/conda/envs/arivale-r/lib/libopenblasp-r0.3.18.so\n",
       "\n",
       "locale:\n",
       " [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              \n",
       " [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    \n",
       " [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   \n",
       " [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 \n",
       " [9] LC_ADDRESS=C               LC_TELEPHONE=C            \n",
       "[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       \n",
       "\n",
       "attached base packages:\n",
       "[1] stats     graphics  grDevices utils     datasets  methods   base     \n",
       "\n",
       "other attached packages:\n",
       " [1] openxlsx_4.2.5.2      readxl_1.4.2          enrichplot_1.14.2    \n",
       " [4] clusterProfiler_4.2.2 WGCNA_1.71            fastcluster_1.2.3    \n",
       " [7] dynamicTreeCut_1.63-1 forcats_0.5.1         stringr_1.4.0        \n",
       "[10] dplyr_1.0.9           purrr_0.3.4           readr_2.1.2          \n",
       "[13] tidyr_1.2.0           tibble_3.1.7          ggplot2_3.3.6        \n",
       "[16] tidyverse_1.3.1      \n",
       "\n",
       "loaded via a namespace (and not attached):\n",
       "  [1] shadowtext_0.1.2       uuid_1.1-0             backports_1.4.1       \n",
       "  [4] Hmisc_4.7-2            fastmatch_1.1-3        plyr_1.8.7            \n",
       "  [7] igraph_1.3.2           lazyeval_0.2.2         repr_1.1.4            \n",
       " [10] splines_4.1.1          BiocParallel_1.26.2    GenomeInfoDb_1.28.4   \n",
       " [13] digest_0.6.29          yulab.utils_0.0.4      foreach_1.5.2         \n",
       " [16] htmltools_0.5.2        GOSemSim_2.20.0        viridis_0.6.2         \n",
       " [19] GO.db_3.14.0           fansi_1.0.3            magrittr_2.0.3        \n",
       " [22] checkmate_2.1.0        memoise_2.0.1          cluster_2.1.3         \n",
       " [25] doParallel_1.0.17      tzdb_0.3.0             Biostrings_2.60.2     \n",
       " [28] graphlayouts_0.8.0     modelr_0.1.8           matrixStats_0.62.0    \n",
       " [31] vroom_1.5.7            jpeg_0.1-9             colorspace_2.0-3      \n",
       " [34] blob_1.2.3             rvest_1.0.2            ggrepel_0.9.1         \n",
       " [37] haven_2.5.0            xfun_0.31              crayon_1.5.1          \n",
       " [40] RCurl_1.98-1.7         jsonlite_1.8.0         scatterpie_0.1.7      \n",
       " [43] impute_1.68.0          ape_5.6-2              survival_3.3-1        \n",
       " [46] iterators_1.0.14       glue_1.6.2             polyclip_1.10-0       \n",
       " [49] gtable_0.3.0           zlibbioc_1.38.0        XVector_0.32.0        \n",
       " [52] BiocGenerics_0.38.0    scales_1.2.0           DOSE_3.20.1           \n",
       " [55] DBI_1.1.2              Rcpp_1.0.8.3           viridisLite_0.4.0     \n",
       " [58] htmlTable_2.4.1        tidytree_0.3.9         gridGraphics_0.5-1    \n",
       " [61] foreign_0.8-82         bit_4.0.4              preprocessCore_1.56.0 \n",
       " [64] Formula_1.2-4          stats4_4.1.1           htmlwidgets_1.5.4     \n",
       " [67] httr_1.4.3             fgsea_1.20.0           RColorBrewer_1.1-3    \n",
       " [70] ellipsis_0.3.2         pkgconfig_2.0.3        farver_2.1.0          \n",
       " [73] nnet_7.3-17            dbplyr_2.2.0           utf8_1.2.2            \n",
       " [76] ggplotify_0.1.0        tidyselect_1.1.2       rlang_1.0.2           \n",
       " [79] reshape2_1.4.4         AnnotationDbi_1.56.2   munsell_0.5.0         \n",
       " [82] cellranger_1.1.0       tools_4.1.1            cachem_1.0.6          \n",
       " [85] downloader_0.4         cli_3.3.0              generics_0.1.2        \n",
       " [88] RSQLite_2.2.14         broom_0.8.0            evaluate_0.15         \n",
       " [91] fastmap_1.1.0          ggtree_3.2.1           knitr_1.39            \n",
       " [94] bit64_4.0.5            fs_1.5.2               tidygraph_1.2.1       \n",
       " [97] zip_2.2.0              KEGGREST_1.32.0        ggraph_2.0.5          \n",
       "[100] nlme_3.1-158           aplot_0.1.6            DO.db_2.9             \n",
       "[103] xml2_1.3.3             compiler_4.1.1         rstudioapi_0.13       \n",
       "[106] png_0.1-7              treeio_1.18.1          reprex_2.0.1          \n",
       "[109] tweenr_1.0.2           stringi_1.7.6          lattice_0.20-45       \n",
       "[112] IRdisplay_1.1          Matrix_1.4-1           vctrs_0.4.1           \n",
       "[115] pillar_1.7.0           lifecycle_1.0.1        data.table_1.14.2     \n",
       "[118] bitops_1.0-7           patchwork_1.1.1        qvalue_2.26.0         \n",
       "[121] R6_2.5.1               latticeExtra_0.6-29    gridExtra_2.3         \n",
       "[124] IRanges_2.26.0         codetools_0.2-18       MASS_7.3-57           \n",
       "[127] assertthat_0.2.1       withr_2.5.0            S4Vectors_0.30.2      \n",
       "[130] GenomeInfoDbData_1.2.6 parallel_4.1.1         hms_1.1.1             \n",
       "[133] ggfun_0.0.6            grid_4.1.1             rpart_4.1.16          \n",
       "[136] IRkernel_1.3           pbdZMQ_0.3-7           ggforce_0.3.3         \n",
       "[139] Biobase_2.52.0         lubridate_1.8.0        base64enc_0.1-3       "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sessionInfo()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "arivale-r - R",
   "language": "R",
   "name": "conda-env-arivale-r-r"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.1.1"
  },
  "widgets": {
   "application/vnd.jupyter.widget-state+json": {
    "state": {},
    "version_major": 2,
    "version_minor": 0
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
