{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# GEM Reconstruction with LC M001-related Transcriptomics — Enrichment Analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***by Kengo Watanabe***  \n",
    "\n",
    "Priyanka Baloni reconstructed mouse genome-scale metabolic models (GEMs; Khodaee, S. et al. Sci. Rep. 2020) with the preprocessed Longevity Consortium (LC) M001-related transcriptomics dataset (Tyshkovskiy, A. et al. Cell Metab. 2019; adjusted with sex and age), and calculated maximum flux values using flux variability analysis (FVA).  \n",
    "–> This Jupyter Notebook (with R kernel) performed the enrichment analysis on the potentially changed reactions using GEM subsystem annotations.  \n",
    "\n",
    "Input files:  \n",
    "- Reaction metadata: 230502_LC-M001-related-TrOmics-GEM-ver3-15_FluxAnalysis_reaction-metadata.xlsx  \n",
    "- Flux data (assessed reactions): 230502_LC-M001-related-TrOmics-GEM-ver3-15_FluxAnalysis_average-flux-data_selected.tsv  \n",
    "- Flux comparison result: 230502_LC-M001-related-TrOmics-GEM-ver3-15_FluxAnalysis_flux-comparison_vs-each-control.tsv  \n",
    "\n",
    "Output figures and tables:  \n",
    "- Figure 6b, c  \n",
    "- Supplementary Data 8  \n",
    "\n",
    "Original notebook (memo for my future tracing):  \n",
    "- dalek:\\[JupyterLab HOME\\]/230315_LC-M001-related-TrOmics-GEM-ver3/230503_LC-M001-related-TrOmics-GEM-ver3-15_Enrichment.ipynb  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> I don't know exactly from when but at least after early Feb 2022, BiocManager::install(\"clusterProfiler\") falls into the non-zero exit status error at the sub-dependency: tidygraph package.  \n",
    "> –> After trials and errors, the following code to downgrade BiocManager to the 3.13 version worked! Of note, the currently instralled BiocManager was 3.14 version on Jun 16, 2022.  \n",
    "\n",
    ">> if (!require(\"BiocManager\", quietly = TRUE))  \n",
    ">>     install.packages(\"BiocManager\", version=\"3.13\")  \n",
    ">> BiocManager::install(version=\"3.13\", ask=FALSE)  \n",
    ">> BiocManager::install(\"clusterProfiler\")  \n",
    "\n",
    "> –> After this code, library() raises rlang error, but restarting r-kernel worked. Also, even though BiocManager was downgraded, clusterProfiler was updated from version 4.0.5 (Feb 2022) to version 4.2.2.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "── \u001b[1mAttaching packages\u001b[22m ─────────────────────────────────────── tidyverse 1.3.1 ──\n",
      "\n",
      "\u001b[32m✔\u001b[39m \u001b[34mggplot2\u001b[39m 3.3.6     \u001b[32m✔\u001b[39m \u001b[34mpurrr  \u001b[39m 0.3.4\n",
      "\u001b[32m✔\u001b[39m \u001b[34mtibble \u001b[39m 3.1.7     \u001b[32m✔\u001b[39m \u001b[34mdplyr  \u001b[39m 1.0.9\n",
      "\u001b[32m✔\u001b[39m \u001b[34mtidyr  \u001b[39m 1.2.0     \u001b[32m✔\u001b[39m \u001b[34mstringr\u001b[39m 1.4.0\n",
      "\u001b[32m✔\u001b[39m \u001b[34mreadr  \u001b[39m 2.1.2     \u001b[32m✔\u001b[39m \u001b[34mforcats\u001b[39m 0.5.1\n",
      "\n",
      "── \u001b[1mConflicts\u001b[22m ────────────────────────────────────────── tidyverse_conflicts() ──\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mfilter()\u001b[39m masks \u001b[34mstats\u001b[39m::filter()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mlag()\u001b[39m    masks \u001b[34mstats\u001b[39m::lag()\n",
      "\n",
      "\n",
      "\n",
      "Registered S3 method overwritten by 'ggtree':\n",
      "  method      from \n",
      "  identify.gg ggfun\n",
      "\n",
      "clusterProfiler v4.2.2  For help: https://yulab-smu.top/biomedical-knowledge-mining-book/\n",
      "\n",
      "If you use clusterProfiler in published research, please cite:\n",
      "T Wu, E Hu, S Xu, M Chen, P Guo, Z Dai, T Feng, L Zhou, W Tang, L Zhan, X Fu, S Liu, X Bo, and G Yu. clusterProfiler 4.0: A universal enrichment tool for interpreting omics data. The Innovation. 2021, 2(3):100141\n",
      "\n",
      "\n",
      "Attaching package: ‘clusterProfiler’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:purrr’:\n",
      "\n",
      "    simplify\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:stats’:\n",
      "\n",
      "    filter\n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"clusterProfiler: 4.2.2\"\n",
      "[1] \"enrichplot: 1.14.2\"\n",
      "[1] \"readxl: 1.4.2\"\n",
      "[1] \"openxlsx: 4.2.5.2\"\n"
     ]
    }
   ],
   "source": [
    "library(\"tidyverse\")\n",
    "options(repr.plot.width=5, repr.plot.height=5)#Default=7x7\n",
    "\n",
    "#Bioconductor\n",
    "for (package in c(\"clusterProfiler\", \"enrichplot\")) {\n",
    "    #if (!requireNamespace(\"BiocManager\", quietly=TRUE))\n",
    "    #    install.packages(\"BiocManager\")\n",
    "    #BiocManager::install(package)\n",
    "    eval(bquote(library(.(package))))\n",
    "    print(str_c(package, \": \", as.character(packageVersion(package))))\n",
    "}\n",
    "#CRAN\n",
    "for (package in c(\"readxl\", \"openxlsx\")) {\n",
    "    #install.packages(package)\n",
    "    eval(bquote(library(.(package))))\n",
    "    print(str_c(package, \": \", as.character(packageVersion(package))))\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. All the assessed reactions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Import reaction metadata\n",
    "fileDir <- \"./ExportData/\"\n",
    "ipynbName <- \"230502_LC-M001-related-TrOmics-GEM-ver3-15_FluxAnalysis_\"\n",
    "fileName <- \"reaction-metadata.xlsx\"\n",
    "sheetName <- \"Reaction\"\n",
    "temp <- read_excel(str_c(fileDir,ipynbName,fileName), sheet=sheetName)\n",
    "\n",
    "print(str_c(\"nrow: \",nrow(temp)))\n",
    "head(temp)\n",
    "print(str_c(\"Unique reaction: \", length(unique(temp$ReactionID))))\n",
    "print(str_c(\"Unique subsystem: \", length(unique(temp$Subsystem))))\n",
    "\n",
    "meta_tbl <- temp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Import the assessed reactions\n",
    "fileDir <- \"./ExportData/\"\n",
    "ipynbName <- \"230502_LC-M001-related-TrOmics-GEM-ver3-15_FluxAnalysis_\"\n",
    "fileName <- \"average-flux-data_selected.tsv\"\n",
    "temp <- read_delim(str_c(fileDir,ipynbName,fileName), delim=\"\\t\")\n",
    "\n",
    "#Take the successfully calculated reactions\n",
    "temp <- meta_tbl %>%\n",
    "    dplyr::filter(ReactionID %in% temp$ReactionID)\n",
    "\n",
    "print(str_c(\"nrow: \",nrow(temp)))\n",
    "head(temp)\n",
    "print(str_c(\"Unique reaction: \", length(unique(temp$ReactionID))))\n",
    "print(str_c(\"Unique subsystem: \", length(unique(temp$Subsystem))))\n",
    "\n",
    "bgd_tbl <- temp"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. Changed reactions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> To investigate as a system, changed reactions are selected based on not adjusted P-value but nominal P-value at this step.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Import the summary tables\n",
    "fileDir <- \"./ExportData/\"\n",
    "ipynbName <- \"230502_LC-M001-related-TrOmics-GEM-ver3-15_FluxAnalysis_\"\n",
    "fileName <- \"flux-comparison_vs-each-control.tsv\"\n",
    "temp <- read_delim(str_c(fileDir,ipynbName,fileName), delim=\"\\t\")\n",
    "print(str_c(\"Original nrow: \",nrow(temp)))\n",
    "\n",
    "#Clean\n",
    "temp1 <- str_subset(names(temp), \"_Pval\")\n",
    "temp <- temp %>%\n",
    "    dplyr::select(ReactionID, all_of(temp1))\n",
    "names(temp) <- str_replace(names(temp), \"_Pval\", \"\")\n",
    "\n",
    "print(str_c(\"nrow: \",nrow(temp)))\n",
    "head(temp)\n",
    "print(str_c(\"Unique reaction: \", length(unique(temp$ReactionID))))\n",
    "\n",
    "pval_tbl <- temp"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3. Enrichment Analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3-1. Intervention vs. each control"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 3-1-0. Save result objects"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "comparison_vec <- str_subset(names(pval_tbl), \"-vs-\")\n",
    "\n",
    "#Create a workbook object to save as one single .xlsx file\n",
    "workbook <- createWorkbook()\n",
    "\n",
    "#Summarize results per comparison\n",
    "nRxns_vec <- c()\n",
    "nMappedRxns_vec <- c()\n",
    "nRxnSystems_vec <- c()\n",
    "res_list <- list()\n",
    "for (i in 1:length(comparison_vec)) {\n",
    "    #Prepare input for enricher()\n",
    "    comparison <- comparison_vec[i]\n",
    "    rxns <- pval_tbl %>%\n",
    "        dplyr::filter(!!as.name(comparison)<0.05) %>%\n",
    "        .$ReactionID\n",
    "    print(str_c(comparison,\": \",length(rxns),\" changed reactions\"))\n",
    "    bgds <- bgd_tbl %>%\n",
    "        dplyr::select(Subsystem, ReactionID)\n",
    "    labels <- bgd_tbl %>%\n",
    "        dplyr::mutate(Label=Subsystem) %>%#Dummy in this case\n",
    "        dplyr::select(Subsystem, Label) %>%\n",
    "        dplyr::distinct()\n",
    "    \n",
    "    #Save info\n",
    "    nRxns_vec <- c(nRxns_vec, length(rxns))\n",
    "    temp <- bgds %>%\n",
    "        dplyr::filter(ReactionID %in% rxns)\n",
    "    nMappedRxns_vec <- c(nMappedRxns_vec, length(unique(temp$ReactionID)))\n",
    "    nRxnSystems_vec <- c(nRxnSystems_vec, length(unique(temp$Subsystem)))\n",
    "    \n",
    "    #Enrichment analysis\n",
    "    temp <- enricher(gene=rxns,\n",
    "                     pvalueCutoff=1.0,#To export all\n",
    "                     pAdjustMethod=\"BH\",\n",
    "                     #universe=backgrounds,#Already managed\n",
    "                     minGSSize=4,\n",
    "                     maxGSSize=10000,\n",
    "                     qvalueCutoff=1.0,#To export all\n",
    "                     TERM2GENE=bgds,\n",
    "                     TERM2NAME=labels)\n",
    "    \n",
    "    #Add the summary table to the workbook object as an independent sheet\n",
    "    if (is.data.frame(temp[])) {\n",
    "        temp1 <- tibble(temp[]) %>%\n",
    "            dplyr::select(-Description) %>%\n",
    "            dplyr::rename(Subsystem=ID, Ratio2ChangedRxns=GeneRatio, Ratio2BGs=BgRatio,\n",
    "                          Pval=pvalue, AdjPval=p.adjust, Qval=qvalue, MappedChangedRxn=geneID, nMappedChangedRxns=Count)\n",
    "    } else {\n",
    "        temp1 <- tibble(`n/a`=NA)\n",
    "    }\n",
    "    addWorksheet(workbook, sheetName=comparison)\n",
    "    writeData(workbook, comparison, temp1)\n",
    "    \n",
    "    #Add result object to list\n",
    "    res_list <- c(res_list, list(temp))\n",
    "}\n",
    "\n",
    "#Save the workbook as one single .xlsx file\n",
    "fileDir <- \"./ExportData/\"\n",
    "ipynbName <- \"230503_LC-M001-related-TrOmics-GEM-ver3-15_Enrichment_\"\n",
    "fileName <- \"clusterProfiler-results.xlsx\"\n",
    "saveWorkbook(workbook, file=str_c(fileDir,ipynbName,fileName), overwrite=TRUE)\n",
    "\n",
    "print(str_c(\"nObjects: \", as.character(length(res_list))))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 3-1-1. Aca"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "obj_i <- 1\n",
    "figtitle <- \"Acarbose vs. Control\"\n",
    "\n",
    "#Retreive results\n",
    "comparison <- comparison_vec[obj_i]\n",
    "nRxns <- nRxns_vec[obj_i]\n",
    "nMappedRxns <- nMappedRxns_vec[obj_i]\n",
    "nRxnSystems <- nRxnSystems_vec[obj_i]\n",
    "res <- res_list[[obj_i]]\n",
    "\n",
    "#Check\n",
    "print(comparison)\n",
    "print(str_c(\" - # of the changed reactions: \",as.character(nRxns)))\n",
    "print(str_c(\" - # of the changed reactions that were mapped to any subsystem: \",as.character(nMappedRxns)))\n",
    "print(str_c(\" - # of subsystems having any changed reactions as a member: \",as.character(nRxnSystems)))\n",
    "res\n",
    "print(\" <- Note that the above 'X enriched terms found' is not correct. In this case, X indicates the number of all the tested terms.\")\n",
    "if (is.data.frame(res[])) {\n",
    "    tibble(res[]) %>%\n",
    "        dplyr::filter(pvalue<0.05) %>%#Display only nominal P-value < 0.05\n",
    "        dplyr::select(-Description) %>%\n",
    "        dplyr::rename(Subsystem=ID, Ratio2ChangedRxns=GeneRatio, Ratio2BGs=BgRatio,\n",
    "                      Pval=pvalue, AdjPval=p.adjust, Qval=qvalue, MappedChangedRxn=geneID, nMappedChangedRxns=Count)\n",
    "}\n",
    "\n",
    "#Visualization\n",
    "if (length(tibble(res[]))>0) {\n",
    "    display <- tibble(res[]) %>%\n",
    "        dplyr::filter(pvalue<0.05) %>%\n",
    "        nrow()\n",
    "} else {\n",
    "    display <- 0\n",
    "}\n",
    "if (display>0) {\n",
    "    temp <- res %>%\n",
    "        dplyr::filter(pvalue<0.05) %>%#Display only nominal P-value < 0.05\n",
    "        dplyr::mutate(PvalLabel=str_c(\"AdjPval = \",scales::scientific(p.adjust, digits=2)),\n",
    "                      AdjSignif=ifelse(p.adjust<0.05, \"True\", \"False\")) %>%\n",
    "        barplot(., x=\"Count\", color=\"p.adjust\", showCategory=display) +\n",
    "        geom_text(aes(label=PvalLabel, color=AdjSignif), nudge_x=2.5, hjust=0) +\n",
    "        coord_cartesian(clip=\"off\") +\n",
    "        scale_x_continuous(limits=c(0, 110), breaks=seq(0, 100, by=25), expand=c(0, 0)) +\n",
    "        scale_y_discrete(labels=function(x) {str_wrap(x, width=50)}) +\n",
    "        scale_fill_viridis_c(begin=0, end=1, direction=1, option=\"plasma\",\n",
    "                             limits=c(0, 0.1), breaks=seq(0, 0.1, by=0.025), name=\"AdjPval\") +\n",
    "        scale_color_manual(values=c(\"True\"=\"#990000\", \"False\"=\"gray40\"), ) +\n",
    "        guides(fill=guide_colorbar(reverse=TRUE), color=\"none\") +\n",
    "        labs(x=\"Count of the changed reactions\",\n",
    "             y=\"\", title=str_c(\"Enriched subsystems: \",figtitle)) +\n",
    "        theme_classic(base_size=16, base_family=\"Helvetica\") +\n",
    "        theme(text=element_text(face=\"plain\", color=\"black\", family=\"Helvetica\")) +\n",
    "        theme(axis.text.x=element_text(face=\"plain\", color=\"black\", family=\"Helvetica\"),\n",
    "              axis.text.y=element_text(face=\"plain\", color=\"black\", family=\"Helvetica\"),\n",
    "              axis.title=element_text(face=\"plain\", color=\"black\", family=\"Helvetica\")) +\n",
    "        theme(plot.title=element_text(size=18, hjust=1.0)) +\n",
    "        theme(legend.direction=\"vertical\", legend.box=\"horizontal\",\n",
    "              legend.background=element_blank())\n",
    "    options(repr.plot.width=8.25, repr.plot.height=max(c(1+display*0.25, 2.5)))\n",
    "    plot(temp)\n",
    "    #Save\n",
    "    fileDir <- \"./ExportFigures/\"\n",
    "    ipynbName <- \"230503_LC-M001-related-TrOmics-GEM-ver3-15_Enrichment_\"\n",
    "    fileName <- str_c(comparison,\".pdf\")\n",
    "    ggsave(file=str_c(fileDir,ipynbName,fileName), plot=temp,\n",
    "           width=8.25, height=max(c(1+display*0.25, 2.5)), units=\"in\")\n",
    "    #(Font family is not reflected in JupyterLab output, but correctly done in .pdf file.)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 3-1-2. Rapa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "obj_i <- 2\n",
    "figtitle <- \"Rapamycin vs. Control\"\n",
    "\n",
    "#Retreive results\n",
    "comparison <- comparison_vec[obj_i]\n",
    "nRxns <- nRxns_vec[obj_i]\n",
    "nMappedRxns <- nMappedRxns_vec[obj_i]\n",
    "nRxnSystems <- nRxnSystems_vec[obj_i]\n",
    "res <- res_list[[obj_i]]\n",
    "\n",
    "#Check\n",
    "print(comparison)\n",
    "print(str_c(\" - # of the changed reactions: \",as.character(nRxns)))\n",
    "print(str_c(\" - # of the changed reactions that were mapped to any subsystem: \",as.character(nMappedRxns)))\n",
    "print(str_c(\" - # of subsystems having any changed reactions as a member: \",as.character(nRxnSystems)))\n",
    "res\n",
    "print(\" <- Note that the above 'X enriched terms found' is not correct. In this case, X indicates the number of all the tested terms.\")\n",
    "if (is.data.frame(res[])) {\n",
    "    tibble(res[]) %>%\n",
    "        dplyr::filter(pvalue<0.05) %>%#Display only nominal P-value < 0.05\n",
    "        dplyr::select(-Description) %>%\n",
    "        dplyr::rename(Subsystem=ID, Ratio2ChangedRxns=GeneRatio, Ratio2BGs=BgRatio,\n",
    "                      Pval=pvalue, AdjPval=p.adjust, Qval=qvalue, MappedChangedRxn=geneID, nMappedChangedRxns=Count)\n",
    "}\n",
    "\n",
    "#Visualization\n",
    "if (length(tibble(res[]))>0) {\n",
    "    display <- tibble(res[]) %>%\n",
    "        dplyr::filter(pvalue<0.05) %>%\n",
    "        nrow()\n",
    "} else {\n",
    "    display <- 0\n",
    "}\n",
    "if (display>0) {\n",
    "    temp <- res %>%\n",
    "        dplyr::filter(pvalue<0.05) %>%#Display only nominal P-value < 0.05\n",
    "        dplyr::mutate(PvalLabel=str_c(\"AdjPval = \",scales::scientific(p.adjust, digits=2)),\n",
    "                      AdjSignif=ifelse(p.adjust<0.05, \"True\", \"False\")) %>%\n",
    "        barplot(., x=\"Count\", color=\"p.adjust\", showCategory=display) +\n",
    "        geom_text(aes(label=PvalLabel, color=AdjSignif), nudge_x=2.5, hjust=0) +\n",
    "        coord_cartesian(clip=\"off\") +\n",
    "        scale_x_continuous(limits=c(0, 110), breaks=seq(0, 100, by=25), expand=c(0, 0)) +\n",
    "        scale_y_discrete(labels=function(x) {str_wrap(x, width=50)}) +\n",
    "        scale_fill_viridis_c(begin=0, end=1, direction=1, option=\"plasma\",\n",
    "                             limits=c(0, 0.1), breaks=seq(0, 0.1, by=0.025), name=\"AdjPval\") +\n",
    "        scale_color_manual(values=c(\"True\"=\"#990000\", \"False\"=\"gray40\"), ) +\n",
    "        guides(fill=guide_colorbar(reverse=TRUE), color=\"none\") +\n",
    "        labs(x=\"Count of the changed reactions\",\n",
    "             y=\"\", title=str_c(\"Enriched subsystems: \",figtitle)) +\n",
    "        theme_classic(base_size=16, base_family=\"Helvetica\") +\n",
    "        theme(text=element_text(face=\"plain\", color=\"black\", family=\"Helvetica\")) +\n",
    "        theme(axis.text.x=element_text(face=\"plain\", color=\"black\", family=\"Helvetica\"),\n",
    "              axis.text.y=element_text(face=\"plain\", color=\"black\", family=\"Helvetica\"),\n",
    "              axis.title=element_text(face=\"plain\", color=\"black\", family=\"Helvetica\")) +\n",
    "        theme(plot.title=element_text(size=18, hjust=1.0)) +\n",
    "        theme(legend.direction=\"vertical\", legend.box=\"horizontal\",\n",
    "              legend.background=element_blank())\n",
    "    options(repr.plot.width=8.25, repr.plot.height=max(c(1+display*0.25, 2.5)))\n",
    "    plot(temp)\n",
    "    #Save\n",
    "    fileDir <- \"./ExportFigures/\"\n",
    "    ipynbName <- \"230503_LC-M001-related-TrOmics-GEM-ver3-15_Enrichment_\"\n",
    "    fileName <- str_c(comparison,\".pdf\")\n",
    "    ggsave(file=str_c(fileDir,ipynbName,fileName), plot=temp,\n",
    "           width=8.25, height=max(c(1+display*0.25, 2.5)), units=\"in\")\n",
    "    #(Font family is not reflected in JupyterLab output, but correctly done in .pdf file.)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 3-1-3. CRdiet"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "obj_i <- 3\n",
    "figtitle <- \"CR diet vs. Control\"\n",
    "\n",
    "#Retreive results\n",
    "comparison <- comparison_vec[obj_i]\n",
    "nRxns <- nRxns_vec[obj_i]\n",
    "nMappedRxns <- nMappedRxns_vec[obj_i]\n",
    "nRxnSystems <- nRxnSystems_vec[obj_i]\n",
    "res <- res_list[[obj_i]]\n",
    "\n",
    "#Check\n",
    "print(comparison)\n",
    "print(str_c(\" - # of the changed reactions: \",as.character(nRxns)))\n",
    "print(str_c(\" - # of the changed reactions that were mapped to any subsystem: \",as.character(nMappedRxns)))\n",
    "print(str_c(\" - # of subsystems having any changed reactions as a member: \",as.character(nRxnSystems)))\n",
    "res\n",
    "print(\" <- Note that the above 'X enriched terms found' is not correct. In this case, X indicates the number of all the tested terms.\")\n",
    "if (is.data.frame(res[])) {\n",
    "    tibble(res[]) %>%\n",
    "        dplyr::filter(pvalue<0.05) %>%#Display only nominal P-value < 0.05\n",
    "        dplyr::select(-Description) %>%\n",
    "        dplyr::rename(Subsystem=ID, Ratio2ChangedRxns=GeneRatio, Ratio2BGs=BgRatio,\n",
    "                      Pval=pvalue, AdjPval=p.adjust, Qval=qvalue, MappedChangedRxn=geneID, nMappedChangedRxns=Count)\n",
    "}\n",
    "\n",
    "#Visualization\n",
    "if (length(tibble(res[]))>0) {\n",
    "    display <- tibble(res[]) %>%\n",
    "        dplyr::filter(pvalue<0.05) %>%\n",
    "        nrow()\n",
    "} else {\n",
    "    display <- 0\n",
    "}\n",
    "if (display>0) {\n",
    "    temp <- res %>%\n",
    "        dplyr::filter(pvalue<0.05) %>%#Display only nominal P-value < 0.05\n",
    "        dplyr::mutate(PvalLabel=str_c(\"AdjPval = \",scales::scientific(p.adjust, digits=2)),\n",
    "                      AdjSignif=ifelse(p.adjust<0.05, \"True\", \"False\")) %>%\n",
    "        barplot(., x=\"Count\", color=\"p.adjust\", showCategory=display) +\n",
    "        geom_text(aes(label=PvalLabel, color=AdjSignif), nudge_x=2.5, hjust=0) +\n",
    "        coord_cartesian(clip=\"off\") +\n",
    "        scale_x_continuous(limits=c(0, 110), breaks=seq(0, 100, by=25), expand=c(0, 0)) +\n",
    "        scale_y_discrete(labels=function(x) {str_wrap(x, width=50)}) +\n",
    "        scale_fill_viridis_c(begin=0, end=1, direction=1, option=\"plasma\",\n",
    "                             limits=c(0, 0.1), breaks=seq(0, 0.1, by=0.025), name=\"AdjPval\") +\n",
    "        scale_color_manual(values=c(\"True\"=\"#990000\", \"False\"=\"gray40\"), ) +\n",
    "        guides(fill=guide_colorbar(reverse=TRUE), color=\"none\") +\n",
    "        labs(x=\"Count of the changed reactions\",\n",
    "             y=\"\", title=str_c(\"Enriched subsystems: \",figtitle)) +\n",
    "        theme_classic(base_size=16, base_family=\"Helvetica\") +\n",
    "        theme(text=element_text(face=\"plain\", color=\"black\", family=\"Helvetica\")) +\n",
    "        theme(axis.text.x=element_text(face=\"plain\", color=\"black\", family=\"Helvetica\"),\n",
    "              axis.text.y=element_text(face=\"plain\", color=\"black\", family=\"Helvetica\"),\n",
    "              axis.title=element_text(face=\"plain\", color=\"black\", family=\"Helvetica\")) +\n",
    "        theme(plot.title=element_text(size=18, hjust=1.0)) +\n",
    "        theme(legend.direction=\"vertical\", legend.box=\"horizontal\",\n",
    "              legend.background=element_blank())\n",
    "    options(repr.plot.width=6.3, repr.plot.height=max(c(1+display*0.25, 2.5)))\n",
    "    plot(temp)\n",
    "    #Save\n",
    "    fileDir <- \"./ExportFigures/\"\n",
    "    ipynbName <- \"230503_LC-M001-related-TrOmics-GEM-ver3-15_Enrichment_\"\n",
    "    fileName <- str_c(comparison,\".pdf\")\n",
    "    ggsave(file=str_c(fileDir,ipynbName,fileName), plot=temp,\n",
    "           width=6.3, height=max(c(1+display*0.25, 2.5)), units=\"in\")\n",
    "    #(Font family is not reflected in JupyterLab output, but correctly done in .pdf file.)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# — Session information —"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "R version 4.1.1 (2021-08-10)\n",
       "Platform: x86_64-conda-linux-gnu (64-bit)\n",
       "Running under: Ubuntu 20.04.3 LTS\n",
       "\n",
       "Matrix products: default\n",
       "BLAS/LAPACK: /opt/conda/envs/arivale-r/lib/libopenblasp-r0.3.18.so\n",
       "\n",
       "locale:\n",
       " [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              \n",
       " [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    \n",
       " [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   \n",
       " [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 \n",
       " [9] LC_ADDRESS=C               LC_TELEPHONE=C            \n",
       "[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       \n",
       "\n",
       "attached base packages:\n",
       "[1] stats     graphics  grDevices utils     datasets  methods   base     \n",
       "\n",
       "other attached packages:\n",
       " [1] openxlsx_4.2.5.2      readxl_1.4.2          enrichplot_1.14.2    \n",
       " [4] clusterProfiler_4.2.2 forcats_0.5.1         stringr_1.4.0        \n",
       " [7] dplyr_1.0.9           purrr_0.3.4           readr_2.1.2          \n",
       "[10] tidyr_1.2.0           tibble_3.1.7          ggplot2_3.3.6        \n",
       "[13] tidyverse_1.3.1      \n",
       "\n",
       "loaded via a namespace (and not attached):\n",
       "  [1] ggtree_3.2.1           fgsea_1.20.0           colorspace_2.0-3      \n",
       "  [4] ellipsis_0.3.2         IRdisplay_1.1          qvalue_2.26.0         \n",
       "  [7] XVector_0.32.0         base64enc_0.1-3        fs_1.5.2              \n",
       " [10] aplot_0.1.6            rstudioapi_0.13        farver_2.1.0          \n",
       " [13] graphlayouts_0.8.0     ggrepel_0.9.1          bit64_4.0.5           \n",
       " [16] scatterpie_0.1.7       AnnotationDbi_1.56.2   fansi_1.0.3           \n",
       " [19] lubridate_1.8.0        xml2_1.3.3             splines_4.1.1         \n",
       " [22] cachem_1.0.6           GOSemSim_2.20.0        polyclip_1.10-0       \n",
       " [25] IRkernel_1.3           jsonlite_1.8.0         broom_0.8.0           \n",
       " [28] GO.db_3.14.0           dbplyr_2.2.0           png_0.1-7             \n",
       " [31] ggforce_0.3.3          compiler_4.1.1         httr_1.4.3            \n",
       " [34] backports_1.4.1        lazyeval_0.2.2         assertthat_0.2.1      \n",
       " [37] Matrix_1.4-1           fastmap_1.1.0          cli_3.3.0             \n",
       " [40] tweenr_1.0.2           htmltools_0.5.2        tools_4.1.1           \n",
       " [43] igraph_1.3.2           gtable_0.3.0           glue_1.6.2            \n",
       " [46] GenomeInfoDbData_1.2.6 reshape2_1.4.4         DO.db_2.9             \n",
       " [49] fastmatch_1.1-3        Rcpp_1.0.8.3           Biobase_2.52.0        \n",
       " [52] cellranger_1.1.0       vctrs_0.4.1            Biostrings_2.60.2     \n",
       " [55] nlme_3.1-158           ape_5.6-2              ggraph_2.0.5          \n",
       " [58] rvest_1.0.2            lifecycle_1.0.1        DOSE_3.20.1           \n",
       " [61] zlibbioc_1.38.0        MASS_7.3-57            scales_1.2.0          \n",
       " [64] tidygraph_1.2.1        vroom_1.5.7            hms_1.1.1             \n",
       " [67] parallel_4.1.1         RColorBrewer_1.1-3     memoise_2.0.1         \n",
       " [70] gridExtra_2.3          ggfun_0.0.6            downloader_0.4        \n",
       " [73] yulab.utils_0.0.4      stringi_1.7.6          RSQLite_2.2.14        \n",
       " [76] S4Vectors_0.30.2       tidytree_0.3.9         BiocGenerics_0.38.0   \n",
       " [79] zip_2.2.0              BiocParallel_1.26.2    repr_1.1.4            \n",
       " [82] GenomeInfoDb_1.28.4    rlang_1.0.2            pkgconfig_2.0.3       \n",
       " [85] bitops_1.0-7           evaluate_0.15          lattice_0.20-45       \n",
       " [88] treeio_1.18.1          patchwork_1.1.1        shadowtext_0.1.2      \n",
       " [91] bit_4.0.4              tidyselect_1.1.2       plyr_1.8.7            \n",
       " [94] magrittr_2.0.3         R6_2.5.1               IRanges_2.26.0        \n",
       " [97] generics_0.1.2         pbdZMQ_0.3-7           DBI_1.1.2             \n",
       "[100] pillar_1.7.0           haven_2.5.0            withr_2.5.0           \n",
       "[103] KEGGREST_1.32.0        RCurl_1.98-1.7         modelr_0.1.8          \n",
       "[106] crayon_1.5.1           uuid_1.1-0             utf8_1.2.2            \n",
       "[109] tzdb_0.3.0             viridis_0.6.2          grid_4.1.1            \n",
       "[112] data.table_1.14.2      blob_1.2.3             reprex_2.0.1          \n",
       "[115] digest_0.6.29          gridGraphics_0.5-1     stats4_4.1.1          \n",
       "[118] munsell_0.5.0          ggplotify_0.1.0        viridisLite_0.4.0     "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sessionInfo()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "arivale-r - R",
   "language": "R",
   "name": "conda-env-arivale-r-r"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.1.1"
  },
  "widgets": {
   "application/vnd.jupyter.widget-state+json": {
    "state": {},
    "version_major": 2,
    "version_minor": 0
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
