{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# GEM Reconstruction with LC M001-related Transcriptomics — Flux Comparison"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***by Kengo Watanabe***  \n",
    "\n",
    "Priyanka Baloni reconstructed mouse genome-scale metabolic models (GEMs; Khodaee, S. et al. Sci. Rep. 2020) with the preprocessed Longevity Consortium (LC) M001-related transcriptomics dataset (Tyshkovskiy, A. et al. Cell Metab. 2019; adjusted with sex and age), and calculated flux values using flux variability analysis (FVA).  \n",
    "–> This Jupyter Notebook (with Python 3 kernel) compared the calculated flux values between the control and intervention groups.  \n",
    "\n",
    "Input files:  \n",
    "- Flux data (maximum): merged_file_MaxFlux_reactions_Intervention-ori.csv  \n",
    "- Flux data (minimum): merged_file_MinFlux_reactions_Intervention-ori.csv  \n",
    "- Model–sample metadata: Metadata_LC.xlsx  \n",
    "- Sample–mouse metadata: 230215_LC-M001-related-TrOmics-DIRAC-ver3_Preprocessing_onWenc_sample-metadata.tsv  \n",
    "- Reaction metadata (iMM1865): iMM1685_data.xlsx  \n",
    "- Gene mapping metadata: iMM1685_data-Max.xlsx  \n",
    "- Reaction list for the diagram: Reactions-in-diagram.csv  \n",
    "\n",
    "Output figures and tables:  \n",
    "- Figure 6a, d  \n",
    "- Supplementary Data 8  \n",
    "\n",
    "Original notebook (memo for my future tracing):  \n",
    "- dalek:\\[JupyterLab HOME\\]/230315_LC-M001-related-TrOmics-GEM-ver3/230502_LC-M001-related-TrOmics-GEM-ver3-15_FluxAnalysis.ipynb  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# packages in environment at /opt/conda/envs/arivale-py3:\n",
      "#\n",
      "# Name                    Version                   Build  Channel\n",
      "_libgcc_mutex             0.1                 conda_forge    conda-forge\n",
      "_openmp_mutex             4.5                       1_gnu    conda-forge\n",
      "analytics                 0.1                      pypi_0    pypi\n",
      "argon2-cffi               21.1.0           py39h3811e60_0    conda-forge\n",
      "arivale-data-interface    0.1.0                    pypi_0    pypi\n",
      "async_generator           1.10                       py_0    conda-forge\n",
      "atk-1.0                   2.36.0               h3371d22_4    conda-forge\n",
      "attrs                     21.2.0             pyhd8ed1ab_0    conda-forge\n",
      "backcall                  0.2.0              pyh9f0ad1d_0    conda-forge\n",
      "backports                 1.0                        py_2    conda-forge\n",
      "backports.functools_lru_cache 1.6.4              pyhd8ed1ab_0    conda-forge\n",
      "biopython                 1.79             py39h3811e60_0    conda-forge\n",
      "bleach                    4.1.0              pyhd8ed1ab_0    conda-forge\n",
      "bokeh                     2.4.1            py39hf3d152e_1    conda-forge\n",
      "boost-cpp                 1.74.0               h312852a_4    conda-forge\n",
      "boto                      2.49.0                     py_0    conda-forge\n",
      "boto3                     1.19.2                   pypi_0    pypi\n",
      "botocore                  1.22.2                   pypi_0    pypi\n",
      "brotlipy                  0.7.0           py39h3811e60_1001    conda-forge\n",
      "bzip2                     1.0.8                h7f98852_4    conda-forge\n",
      "c-ares                    1.17.2               h7f98852_0    conda-forge\n",
      "ca-certificates           2022.12.7            ha878542_0    conda-forge\n",
      "cachecontrol              0.12.6                     py_0    conda-forge\n",
      "cairo                     1.16.0            h6cf1ce9_1008    conda-forge\n",
      "certifi                   2022.12.7          pyhd8ed1ab_0    conda-forge\n",
      "cffi                      1.14.6           py39h4bc2ebd_1    conda-forge\n",
      "chardet                   4.0.0            py39hf3d152e_1    conda-forge\n",
      "charset-normalizer        2.0.0              pyhd8ed1ab_0    conda-forge\n",
      "click                     8.0.3            py39hf3d152e_0    conda-forge\n",
      "colorama                  0.4.4              pyh9f0ad1d_0    conda-forge\n",
      "cryptography              35.0.0           py39h95dcef6_1    conda-forge\n",
      "cycler                    0.10.0                     py_2    conda-forge\n",
      "cython                    0.29.24          py39he80948d_0    conda-forge\n",
      "debugpy                   1.4.1            py39he80948d_0    conda-forge\n",
      "decorator                 4.4.2                      py_0    conda-forge\n",
      "defusedxml                0.7.1              pyhd8ed1ab_0    conda-forge\n",
      "english                   2020.7.0           pyhd8ed1ab_0    conda-forge\n",
      "entrypoints               0.3             py39hde42818_1002    conda-forge\n",
      "et_xmlfile                1.0.1                   py_1001    conda-forge\n",
      "expat                     2.4.1                h9c3ff4c_0    conda-forge\n",
      "font-ttf-dejavu-sans-mono 2.37                 hab24e00_0    conda-forge\n",
      "font-ttf-inconsolata      3.000                h77eed37_0    conda-forge\n",
      "font-ttf-source-code-pro  2.038                h77eed37_0    conda-forge\n",
      "font-ttf-ubuntu           0.83                 hab24e00_0    conda-forge\n",
      "fontconfig                2.13.1            hba837de_1005    conda-forge\n",
      "fonts-conda-ecosystem     1                             0    conda-forge\n",
      "fonts-conda-forge         1                             0    conda-forge\n",
      "freetype                  2.10.4               h0708190_1    conda-forge\n",
      "fribidi                   1.0.10               h36c2ea0_0    conda-forge\n",
      "future                    0.18.2           py39hf3d152e_3    conda-forge\n",
      "gdk-pixbuf                2.42.6               h04a7f16_0    conda-forge\n",
      "gettext                   0.19.8.1          h73d1719_1008    conda-forge\n",
      "giflib                    5.2.1                h36c2ea0_2    conda-forge\n",
      "goatools                  1.2.3              pyh7cba7a3_2    bioconda\n",
      "graphite2                 1.3.13            h58526e2_1001    conda-forge\n",
      "graphviz                  2.49.1               h85b4f2f_0    conda-forge\n",
      "gtk2                      2.24.33              h539f30e_1    conda-forge\n",
      "gts                       0.7.6                h64030ff_2    conda-forge\n",
      "harfbuzz                  3.0.0                h83ec7ef_1    conda-forge\n",
      "hdmedians                 0.14.2           py39hce5d2b2_0    conda-forge\n",
      "icu                       68.1                 h58526e2_0    conda-forge\n",
      "idna                      3.1                pyhd3deb0d_0    conda-forge\n",
      "importlib-metadata        4.8.1            py39hf3d152e_0    conda-forge\n",
      "iniconfig                 1.1.1              pyh9f0ad1d_0    conda-forge\n",
      "interchange               2021.0.4           pyhd8ed1ab_0    conda-forge\n",
      "ipykernel                 6.4.2            py39hef51801_0    conda-forge\n",
      "ipython                   7.28.0           py39hef51801_0    conda-forge\n",
      "ipython_genutils          0.2.0                      py_1    conda-forge\n",
      "ipywidgets                7.6.5              pyhd8ed1ab_0    conda-forge\n",
      "jbig                      2.1               h7f98852_2003    conda-forge\n",
      "jedi                      0.18.0           py39hf3d152e_2    conda-forge\n",
      "jinja2                    3.0.2              pyhd8ed1ab_0    conda-forge\n",
      "jmespath                  0.10.0                   pypi_0    pypi\n",
      "joblib                    1.1.0              pyhd8ed1ab_0    conda-forge\n",
      "jpeg                      9d                   h36c2ea0_0    conda-forge\n",
      "jsonschema                4.1.2              pyhd8ed1ab_0    conda-forge\n",
      "jupyter_client            7.0.6              pyhd8ed1ab_0    conda-forge\n",
      "jupyter_core              4.8.1            py39hf3d152e_0    conda-forge\n",
      "jupyterlab_pygments       0.1.2              pyh9f0ad1d_0    conda-forge\n",
      "jupyterlab_widgets        1.0.2              pyhd8ed1ab_0    conda-forge\n",
      "kiwisolver                1.3.2            py39h1a9c180_0    conda-forge\n",
      "krb5                      1.19.2               hcc1bbae_2    conda-forge\n",
      "lcms2                     2.12                 hddcbb42_0    conda-forge\n",
      "ld_impl_linux-64          2.36.1               hea4e1c9_2    conda-forge\n",
      "lerc                      2.2.1                h9c3ff4c_0    conda-forge\n",
      "libblas                   3.9.0           12_linux64_openblas    conda-forge\n",
      "libcblas                  3.9.0           12_linux64_openblas    conda-forge\n",
      "libcurl                   7.79.1               h2574ce0_1    conda-forge\n",
      "libdeflate                1.7                  h7f98852_5    conda-forge\n",
      "libedit                   3.1.20191231         he28a2e2_2    conda-forge\n",
      "libev                     4.33                 h516909a_1    conda-forge\n",
      "libffi                    3.4.2                h9c3ff4c_4    conda-forge\n",
      "libgcc-ng                 11.2.0              h1d223b6_11    conda-forge\n",
      "libgd                     2.3.3                h6ad9fb6_0    conda-forge\n",
      "libgfortran-ng            11.2.0              h69a702a_11    conda-forge\n",
      "libgfortran5              11.2.0              h5c6108e_11    conda-forge\n",
      "libglib                   2.70.0               h174f98d_1    conda-forge\n",
      "libgomp                   11.2.0              h1d223b6_11    conda-forge\n",
      "libiconv                  1.16                 h516909a_0    conda-forge\n",
      "liblapack                 3.9.0           12_linux64_openblas    conda-forge\n",
      "libnghttp2                1.43.0               h812cca2_1    conda-forge\n",
      "libopenblas               0.3.18          pthreads_h8fe5266_0    conda-forge\n",
      "libpng                    1.6.37               h21135ba_2    conda-forge\n",
      "libpq                     13.3                 hd57d9b9_1    conda-forge\n",
      "librsvg                   2.52.2               hc3c00ef_0    conda-forge\n",
      "libsodium                 1.0.18               h36c2ea0_1    conda-forge\n",
      "libssh2                   1.10.0               ha56f1ee_2    conda-forge\n",
      "libstdcxx-ng              11.2.0              he4da1e4_11    conda-forge\n",
      "libtiff                   4.3.0                hf544144_1    conda-forge\n",
      "libtool                   2.4.6             h9c3ff4c_1008    conda-forge\n",
      "libuuid                   2.32.1            h7f98852_1000    conda-forge\n",
      "libwebp                   1.2.1                h3452ae3_0    conda-forge\n",
      "libwebp-base              1.2.1                h7f98852_0    conda-forge\n",
      "libxcb                    1.13              h7f98852_1003    conda-forge\n",
      "libxml2                   2.9.12               h72842e0_0    conda-forge\n",
      "libxslt                   1.1.33               h15afd5d_2    conda-forge\n",
      "libzlib                   1.2.11            h36c2ea0_1013    conda-forge\n",
      "lockfile                  0.12.2                     py_1    conda-forge\n",
      "lxml                      4.6.3            py39h107f48f_0    conda-forge\n",
      "lz4-c                     1.9.3                h9c3ff4c_1    conda-forge\n",
      "markupsafe                2.0.1            py39h3811e60_0    conda-forge\n",
      "matplotlib-base           3.4.3            py39h2fa2bec_1    conda-forge\n",
      "matplotlib-inline         0.1.3              pyhd8ed1ab_0    conda-forge\n",
      "matplotlib-venn           0.11.7             pyhd8ed1ab_0    conda-forge\n",
      "missingpy                 0.2.0                    pypi_0    pypi\n",
      "mistune                   0.8.4           py39h3811e60_1004    conda-forge\n",
      "more-itertools            8.10.0             pyhd8ed1ab_0    conda-forge\n",
      "mscorefonts               0.0.1                         3    conda-forge\n",
      "msgpack-python            1.0.2            py39h1a9c180_1    conda-forge\n",
      "natsort                   7.1.1              pyhd8ed1ab_0    conda-forge\n",
      "nbclient                  0.5.4              pyhd8ed1ab_0    conda-forge\n",
      "nbconvert                 6.2.0            py39hf3d152e_0    conda-forge\n",
      "nbformat                  5.1.3              pyhd8ed1ab_0    conda-forge\n",
      "ncurses                   6.2                  h58526e2_4    conda-forge\n",
      "neo4j-python-driver       4.3.7              pyhd8ed1ab_0    conda-forge\n",
      "neobolt                   1.7.17           py39h3811e60_2    conda-forge\n",
      "neotime                   1.7.4                      py_0    conda-forge\n",
      "nest-asyncio              1.5.1              pyhd8ed1ab_0    conda-forge\n",
      "networkx                  2.6.3              pyhd8ed1ab_0    conda-forge\n",
      "notebook                  6.4.5              pyha770c72_0    conda-forge\n",
      "numpy                     1.21.3           py39hdbf815f_0    conda-forge\n",
      "olefile                   0.46               pyh9f0ad1d_1    conda-forge\n",
      "openjpeg                  2.4.0                hb52868f_1    conda-forge\n",
      "openpyxl                  3.0.9              pyhd8ed1ab_0    conda-forge\n",
      "openssl                   1.1.1o               h166bdaf_0    conda-forge\n",
      "packaging                 21.0               pyhd8ed1ab_0    conda-forge\n",
      "pandas                    1.3.4            py39hde0f152_0    conda-forge\n",
      "pandoc                    2.15                 h7f98852_0    conda-forge\n",
      "pandocfilters             1.5.0              pyhd8ed1ab_0    conda-forge\n",
      "pango                     1.48.10              h54213e6_2    conda-forge\n",
      "parso                     0.8.2              pyhd8ed1ab_0    conda-forge\n",
      "patsy                     0.5.2              pyhd8ed1ab_0    conda-forge\n",
      "pcre                      8.45                 h9c3ff4c_0    conda-forge\n",
      "pdf2image                 1.16.2                   pypi_0    pypi\n",
      "pexpect                   4.8.0              pyh9f0ad1d_2    conda-forge\n",
      "pickleshare               0.7.5           py39hde42818_1002    conda-forge\n",
      "pillow                    8.3.2            py39ha612740_0    conda-forge\n",
      "pip                       21.3.1             pyhd8ed1ab_0    conda-forge\n",
      "pixman                    0.40.0               h36c2ea0_0    conda-forge\n",
      "plotly                    5.3.1              pyhd8ed1ab_0    conda-forge\n",
      "pluggy                    1.0.0            py39hf3d152e_1    conda-forge\n",
      "poppler                   21.09.0              h0e1ea10_2    conda-forge\n",
      "poppler-data              0.4.12               hd8ed1ab_0    conda-forge\n",
      "prometheus_client         0.11.0             pyhd8ed1ab_0    conda-forge\n",
      "prompt-toolkit            3.0.21             pyha770c72_0    conda-forge\n",
      "prompt_toolkit            3.0.21               hd8ed1ab_0    conda-forge\n",
      "psycopg2                  2.9.1            py39h3811e60_0    conda-forge\n",
      "pthread-stubs             0.4               h36c2ea0_1001    conda-forge\n",
      "ptyprocess                0.7.0              pyhd3deb0d_0    conda-forge\n",
      "py                        1.10.0             pyhd3deb0d_0    conda-forge\n",
      "py2neo                    2021.2.3           pyhd8ed1ab_0    conda-forge\n",
      "pycparser                 2.20               pyh9f0ad1d_2    conda-forge\n",
      "pydot                     1.4.2            py39hf3d152e_3    conda-forge\n",
      "pygments                  2.10.0             pyhd8ed1ab_0    conda-forge\n",
      "pygraphviz                1.7              py39h78163bd_0    conda-forge\n",
      "pyopenssl                 21.0.0             pyhd8ed1ab_0    conda-forge\n",
      "pyparsing                 2.4.7              pyh9f0ad1d_0    conda-forge\n",
      "pyrsistent                0.17.3           py39h3811e60_2    conda-forge\n",
      "pysam                     0.17.0           py39h051187c_0    bioconda\n",
      "pysocks                   1.7.1            py39hf3d152e_3    conda-forge\n",
      "pytabix                   0.1              py39h98c8e45_1    bioconda\n",
      "pytest                    6.2.5            py39hf3d152e_0    conda-forge\n",
      "python                    3.9.7           hb7a2778_3_cpython    conda-forge\n",
      "python-dateutil           2.8.2              pyhd8ed1ab_0    conda-forge\n",
      "python-graphviz           0.17               pyhaef67bd_0    conda-forge\n",
      "python-wget               3.2                        py_0    conda-forge\n",
      "python_abi                3.9                      2_cp39    conda-forge\n",
      "pytz                      2021.3             pyhd8ed1ab_0    conda-forge\n",
      "pyvcf                     0.6.8           py39hde42818_1002    conda-forge\n",
      "pyyaml                    6.0              py39h3811e60_0    conda-forge\n",
      "pyzmq                     22.3.0           py39h37b5a0c_0    conda-forge\n",
      "readline                  8.1                  h46c0cb4_0    conda-forge\n",
      "requests                  2.26.0             pyhd8ed1ab_0    conda-forge\n",
      "s3transfer                0.5.0                    pypi_0    pypi\n",
      "scikit-bio                0.5.6            py39h16ac069_4    conda-forge\n",
      "scikit-learn              1.0.1            py39h7c5d8c9_0    conda-forge\n",
      "scipy                     1.7.1            py39hee8e79c_0    conda-forge\n",
      "seaborn                   0.11.2               hd8ed1ab_0    conda-forge\n",
      "seaborn-base              0.11.2             pyhd8ed1ab_0    conda-forge\n",
      "send2trash                1.8.0              pyhd8ed1ab_0    conda-forge\n",
      "setuptools                58.2.0           py39hf3d152e_0    conda-forge\n",
      "six                       1.16.0             pyh6c4a22f_0    conda-forge\n",
      "sqlite                    3.36.0               h9cd32fc_2    conda-forge\n",
      "statsmodels               0.13.0           py39hce5d2b2_0    conda-forge\n",
      "tenacity                  8.0.1              pyhd8ed1ab_0    conda-forge\n",
      "terminado                 0.12.1           py39hf3d152e_0    conda-forge\n",
      "testpath                  0.5.0              pyhd8ed1ab_0    conda-forge\n",
      "threadpoolctl             3.0.0              pyh8a188c0_0    conda-forge\n",
      "tk                        8.6.11               h27826a3_1    conda-forge\n",
      "toml                      0.10.2             pyhd8ed1ab_0    conda-forge\n",
      "tornado                   6.1              py39h3811e60_1    conda-forge\n",
      "traitlets                 5.1.0              pyhd8ed1ab_0    conda-forge\n",
      "typing_extensions         3.10.0.2           pyha770c72_0    conda-forge\n",
      "tzdata                    2021e                he74cb21_0    conda-forge\n",
      "urllib3                   1.26.7             pyhd8ed1ab_0    conda-forge\n",
      "venn                      0.1.3                    pypi_0    pypi\n",
      "wcwidth                   0.2.5              pyh9f0ad1d_2    conda-forge\n",
      "webencodings              0.5.1                      py_1    conda-forge\n",
      "wheel                     0.37.0             pyhd8ed1ab_1    conda-forge\n",
      "widgetsnbextension        3.5.1            py39hf3d152e_4    conda-forge\n",
      "wordcloud                 1.8.1                    pypi_0    pypi\n",
      "xlrd                      2.0.1              pyhd8ed1ab_3    conda-forge\n",
      "xlsxwriter                3.0.3              pyhd8ed1ab_0    conda-forge\n",
      "xorg-kbproto              1.0.7             h7f98852_1002    conda-forge\n",
      "xorg-libice               1.0.10               h7f98852_0    conda-forge\n",
      "xorg-libsm                1.2.3             hd9c2040_1000    conda-forge\n",
      "xorg-libx11               1.7.2                h7f98852_0    conda-forge\n",
      "xorg-libxau               1.0.9                h7f98852_0    conda-forge\n",
      "xorg-libxdmcp             1.1.3                h7f98852_0    conda-forge\n",
      "xorg-libxext              1.3.4                h7f98852_1    conda-forge\n",
      "xorg-libxrender           0.9.10            h7f98852_1003    conda-forge\n",
      "xorg-renderproto          0.11.1            h7f98852_1002    conda-forge\n",
      "xorg-xextproto            7.3.0             h7f98852_1002    conda-forge\n",
      "xorg-xproto               7.0.31            h7f98852_1007    conda-forge\n",
      "xz                        5.2.5                h516909a_1    conda-forge\n",
      "yaml                      0.2.5                h516909a_0    conda-forge\n",
      "zeromq                    4.3.4                h9c3ff4c_1    conda-forge\n",
      "zipp                      3.6.0              pyhd8ed1ab_0    conda-forge\n",
      "zlib                      1.2.11            h36c2ea0_1013    conda-forge\n",
      "zstd                      1.5.0                ha95c52a_0    conda-forge\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scipy.stats as stats\n",
    "import matplotlib.pyplot as plt\n",
    "%matplotlib inline\n",
    "import seaborn as sns\n",
    "#For Arial font\n",
    "#!conda install -c conda-forge -y mscorefonts\n",
    "##-> The below was also needed in matplotlib 3.4.2\n",
    "#import shutil\n",
    "#import matplotlib\n",
    "#shutil.rmtree(matplotlib.get_cachedir())\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')\n",
    "from IPython.display import display\n",
    "import time\n",
    "#For exporting .pdf file with editable text\n",
    "import matplotlib\n",
    "matplotlib.rcParams['pdf.fonttype']=42\n",
    "matplotlib.rcParams['ps.fonttype']=42\n",
    "\n",
    "from sklearn.preprocessing import RobustScaler\n",
    "from sklearn.decomposition import PCA\n",
    "from decimal import Decimal, ROUND_HALF_UP\n",
    "from statsmodels.stats import multitest as multi\n",
    "from sklearn.preprocessing import MaxAbsScaler\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "import re\n",
    "import matplotlib.patches as mpatches\n",
    "#!pip install venn\n",
    "from venn import venn\n",
    "\n",
    "!conda list"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. Clean the original tables"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1-1. Flux value data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tempD = {}\n",
    "for flux in ['MaxFlux', 'MinFlux']:\n",
    "    #Import raw flux data\n",
    "    fileDir = './ImportData/230426_FVA-ver3-2/'\n",
    "    fileName = 'merged_file_'+flux+'_reactions_Intervention-ori.csv'\n",
    "    tempDF = pd.read_csv(fileDir+fileName)\n",
    "    tempDF = tempDF.rename(columns={'Reaction':'ReactionID'})\n",
    "    tempDF = tempDF.set_index('ReactionID')\n",
    "    \n",
    "    #Clean the model label\n",
    "    tempDF1 = tempDF.columns.to_series().str.split(pat='_', expand=True)\n",
    "    tempDF1 = tempDF1.rename(columns={0:'Prefix', 1:'Model_i'})\n",
    "    tempDF.columns = 'Model_'+tempDF1['Model_i'].str.zfill(2)\n",
    "    tempDF.columns.name = None\n",
    "    \n",
    "    #Sort\n",
    "    tempL = ['Model_'+str(i+1).zfill(2) for i in range(len(tempDF.columns))]\n",
    "    tempDF = tempDF[tempL]\n",
    "    \n",
    "    print(flux)\n",
    "    display(tempDF)\n",
    "    display(tempDF.reset_index().describe(include='all'))\n",
    "    \n",
    "    tempD[flux] = tempDF\n",
    "\n",
    "#Calculate the average\n",
    "tempDF1 = tempD['MaxFlux']\n",
    "tempDF2 = tempD['MinFlux']\n",
    "##Check just in case\n",
    "print('nMatching rows:', (tempDF1.index==tempDF2.index).sum())\n",
    "print('nMatching columns:', (tempDF1.columns==tempDF2.columns).sum())\n",
    "tempDF = (tempDF1 + tempDF2) / 2\n",
    "\n",
    "print('Average')\n",
    "display(tempDF)\n",
    "display(tempDF.reset_index().describe(include='all'))\n",
    "\n",
    "#Save\n",
    "fileDir = './ExportData/'\n",
    "ipynbName = '230502_LC-M001-related-TrOmics-GEM-ver3-15_FluxAnalysis_'\n",
    "fileName = 'average-flux-data.tsv'\n",
    "tempDF.to_csv(fileDir+ipynbName+fileName, index=True, sep='\\t')\n",
    "\n",
    "fluxDF = tempDF"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1-2. Sample metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Import sample-model metadata\n",
    "fileDir = './ImportData/230419_FVA-ver2/'\n",
    "fileName = 'Metadata_LC.xlsx'\n",
    "sheetName = 'Sheet1'\n",
    "tempDF = pd.read_excel(fileDir+fileName, sheet_name=sheetName, engine='openpyxl')\n",
    "\n",
    "#Clean column names and model labels\n",
    "tempDF = tempDF.rename(columns={'Model':'ModelID', 'Data':'SampleID'})\n",
    "tempDF1 = tempDF['ModelID'].str.split(pat='_', expand=True)\n",
    "tempDF1 = tempDF1.rename(columns={0:'Prefix', 1:'Model_i'})\n",
    "tempDF['ModelID'] = 'Model_'+tempDF1['Model_i'].str.zfill(2)\n",
    "tempS = tempDF.set_index('ModelID')\n",
    "\n",
    "#Import the cleaned sample-model metadata\n",
    "fileDir = '../230224_LC-M001-PrOmics-vs-TrOmics-DIRAC-ver3/ImportData/'\n",
    "ipynbName = '230215_LC-M001-related-TrOmics-DIRAC-ver3_Preprocessing_onWenc_'\n",
    "fileName = 'sample-metadata.tsv'\n",
    "tempDF = pd.read_csv(fileDir+ipynbName+fileName, sep='\\t').set_index('SampleID')\n",
    "\n",
    "#Merge\n",
    "tempDF = pd.merge(tempS, tempDF, left_on='SampleID', right_index=True, how='inner')\n",
    "\n",
    "display(tempDF)\n",
    "display(tempDF.reset_index().describe(include='all'))\n",
    "display(tempDF['Group'].value_counts().sort_index(ascending=True))\n",
    "tempL = tempDF.sort_values(by='Category', ascending=True)['Category'].unique()\n",
    "print(' ->', len(tempL), 'categories\\n')\n",
    "for category in tempL:\n",
    "    tempDF1 = tempDF.loc[tempDF['Category']==category]\n",
    "    print(category+' category total:', len(tempDF1))\n",
    "    for condition in ['Intervention']:\n",
    "        display(tempDF1[condition].value_counts())\n",
    "    print('')\n",
    "\n",
    "#Save\n",
    "fileDir = './ExportData/'\n",
    "ipynbName = '230502_LC-M001-related-TrOmics-GEM-ver3-15_FluxAnalysis_'\n",
    "fileName = 'sample-metadata.tsv'\n",
    "tempDF.to_csv(fileDir+ipynbName+fileName, index=True, sep='\\t')\n",
    "\n",
    "sampleDF = tempDF\n",
    "categoryL = tempL"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1-3. Reaction metadata"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 1-3-1. iMM1865"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Import reaction metadata of the used model\n",
    "fileDir = '../220606_LC-M001-related-transcriptomics-GEM/ImportData/'\n",
    "fileName = 'iMM1685_data.xlsx'\n",
    "sheetName = 'Sheet1'\n",
    "tempDF = pd.read_excel(fileDir+fileName, sheet_name=sheetName, engine='openpyxl')\n",
    "tempDF = tempDF.rename(columns={'Reaction':'ReactionID', 'Unnamed: 3':'Reaction'})\n",
    "tempDF = tempDF.set_index('ReactionID')\n",
    "\n",
    "display(tempDF)\n",
    "print(' -> Unique rxn ID:', len(tempDF.index.unique()))\n",
    "\n",
    "rxnDF = tempDF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Check\n",
    "tempS1 = rxnDF['Subsystem'].value_counts()\n",
    "tempS1.name = 'Overall'\n",
    "tempS2 = rxnDF['Subsystem'].loc[fluxDF.index].value_counts()\n",
    "tempS2.name = 'FluxCalculated'\n",
    "tempDF = pd.merge(tempS1, tempS2, left_index=True, right_index=True, how='left')\n",
    "display(tempDF.describe())\n",
    "display(tempDF.loc[tempS1>100])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 1-3-2. Retrieve reaction name with BiGG Models API"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Download the latest file\n",
    "!wget -O ./ImportData/bigg_models_reactions.txt http://bigg.ucsd.edu/static/namespace/bigg_models_reactions.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Import reaction metadata\n",
    "fileDir = './ImportData/'\n",
    "fileName = 'bigg_models_reactions.txt'\n",
    "tempDF = pd.read_csv(fileDir+fileName, sep='\\t')\n",
    "tempDF = tempDF.rename(columns={'bigg_id':'ReactionID', 'name':'ReactionName'})\n",
    "tempDF = tempDF.set_index('ReactionID')\n",
    "print('Original nrow:', len(tempDF))\n",
    "print(' -> Unique rxn ID:', len(tempDF.index.unique()))\n",
    "\n",
    "#Add the reaction name\n",
    "tempDF = pd.merge(rxnDF, tempDF['ReactionName'], left_index=True, right_index=True, how='left')\n",
    "tempDF = tempDF[['Reaction', 'ReactionName', 'GPR', 'Subsystem']]\n",
    "\n",
    "display(tempDF)\n",
    "tempDF1 = tempDF.loc[tempDF['ReactionName'].isnull()]\n",
    "print(' -> No rxn name:', len(tempDF1))\n",
    "display(tempDF1)\n",
    "\n",
    "rxnDF = tempDF"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 1-3-3. Retrieve molecule name with BiGG Models API"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Take all molecules in the reactions\n",
    "tempS = rxnDF['Reaction'].str.replace('^\\'', '')\n",
    "tempS = tempS.str.replace('\\'$', '')\n",
    "tempDF = tempS.str.split(pat=' ', expand=True)\n",
    "print('Expanded DF:', tempDF.shape)\n",
    "t_start = time.time()\n",
    "tempS = set()\n",
    "for col_n in tempDF.columns.tolist():\n",
    "    tempS1 = tempDF[col_n].dropna()#None is considered an NA value in pandas\n",
    "    tempS1 = tempS1.loc[tempS1!='']#Delete empty\n",
    "    tempS1 = tempS1.loc[tempS1.str.contains('\\\\[.*\\\\]')]\n",
    "    #Character in square brackets indicates intracellular localization\n",
    "    tempL = tempS1.str.replace('\\\\[.*\\\\]', '').tolist()\n",
    "    tempS = tempS | set(tempL)\n",
    "t_elapsed = time.time() - t_start\n",
    "print(' -> Elapsed time:', round(t_elapsed//60), 'min', round(t_elapsed%60, 1), 'sec')\n",
    "print(' -> Unique molecule:', len(tempS))\n",
    "print(list(tempS)[:25])\n",
    "\n",
    "molS = tempS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Download the latest file\n",
    "!wget -O ./ImportData/bigg_models_metabolites.txt http://bigg.ucsd.edu/static/namespace/bigg_models_metabolites.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Import metabolite metadata\n",
    "fileDir = './ImportData/'\n",
    "fileName = 'bigg_models_metabolites.txt'\n",
    "tempDF = pd.read_csv(fileDir+fileName, sep='\\t')\n",
    "tempDF = tempDF.rename(columns={'universal_bigg_id':'MoleculeID', 'name':'MoleculeName'})\n",
    "tempDF = tempDF.set_index('MoleculeID')\n",
    "print('Original nrow:', len(tempDF))\n",
    "print(' -> Unique molecule ID:', len(tempDF.index.unique()))\n",
    "\n",
    "#Retrieve the molecules within the used model\n",
    "print('Molecules within the used model', len(molS))\n",
    "tempS = tempDF['MoleculeName']\n",
    "tempS = tempS.drop_duplicates()\n",
    "tempS = tempS.loc[tempS.index.isin(molS)]\n",
    "\n",
    "tempS = tempS.sort_index(ascending=True)\n",
    "display(tempS)\n",
    "\n",
    "molS = tempS"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 1-3-4. Gene mapping table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Import reaction metadata of the used model\n",
    "fileDir = '../220606_LC-M001-related-transcriptomics-GEM/ImportData/'\n",
    "fileName = 'iMM1685_data-Max.xlsx'\n",
    "sheetName = 'Sheet1'\n",
    "tempDF = pd.read_excel(fileDir+fileName, sheet_name=sheetName, engine='openpyxl', header=2)\n",
    "tempDF = tempDF.rename(columns={'Entrez ID':'EntrezID', 'Gene symbol':'GeneSymbol',\n",
    "                                'Ensembl gene ID':'EnsemblID'})\n",
    "tempDF = tempDF[['EntrezID', 'EnsemblID', 'GeneSymbol']]\n",
    "tempDF = tempDF.dropna()\n",
    "tempDF['EntrezID'] = tempDF['EntrezID'].astype('int64')\n",
    "tempDF = tempDF.set_index('EntrezID')\n",
    "tempDF = tempDF.sort_index(ascending=True)\n",
    "\n",
    "display(tempDF)\n",
    "print(' -> Unique Entrez ID:', len(tempDF.index.unique()))\n",
    "print(' -> Unique Ensembl ID:', len(tempDF['EnsemblID'].unique()))\n",
    "print(' -> Unique gene name:', len(tempDF['GeneSymbol'].unique()))\n",
    "\n",
    "geneDF = tempDF"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 1-3-5. Save as a single .xlsx file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Prepare a new .xlsx file\n",
    "fileDir = './ExportData/'\n",
    "ipynbName = '230502_LC-M001-related-TrOmics-GEM-ver3-15_FluxAnalysis_'\n",
    "fileName = 'reaction-metadata.xlsx'\n",
    "sheetName = 'Reaction'\n",
    "rxnDF.to_excel(fileDir+ipynbName+fileName, sheet_name=sheetName, header=True, index=True)\n",
    "\n",
    "#Appended the others to the above .xlsx file\n",
    "with pd.ExcelWriter(fileDir+ipynbName+fileName, mode='a', engine='openpyxl') as writer:\n",
    "    sheetName = 'Molecule'\n",
    "    molS.to_excel(writer, sheet_name=sheetName, header=True, index=True)\n",
    "    sheetName = 'Gene'\n",
    "    geneDF.to_excel(writer, sheet_name=sheetName, header=True, index=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1-4. Select the target samples"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> This study assesses only the Ctrl, Aca, Rapa, and CR groups, consistent with DIRAC analysis.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Select the target samples\n",
    "tempL = ['Ctrl1', 'Aca', 'Rapa', 'CRdiet']\n",
    "tempDF1 = sampleDF.loc[sampleDF['Intervention'].isin(tempL)]\n",
    "tempDF2 = fluxDF.loc[:, fluxDF.columns.isin(tempDF1.index.tolist())]\n",
    "\n",
    "#Re-define the control category\n",
    "tempL = tempDF1.sort_values(by='Category', ascending=True)['Category'].unique()\n",
    "\n",
    "display(tempDF1.describe(include='all'))\n",
    "display(tempDF1['Group'].value_counts().sort_index(ascending=True))\n",
    "print(' ->', len(tempL), 'categories\\n')\n",
    "for category in tempL:\n",
    "    tempDF = tempDF1.loc[tempDF1['Category']==category]\n",
    "    print(category+' category total:', len(tempDF))\n",
    "    for condition in ['Intervention']:\n",
    "        display(tempDF[condition].value_counts())\n",
    "    print('')\n",
    "\n",
    "display(tempDF2.describe(include='all'))\n",
    "\n",
    "#Update\n",
    "sampleDF = tempDF1\n",
    "categoryL = tempL\n",
    "fluxDF = tempDF2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1-5. Select biochemical reactions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> The following reacitons are removed in this analysis:  \n",
    "> - Reactions without subsytem annotation, because they are unfunctional (e.g., biomass reaction). Of note, Priyanka previously confirmed they were not errors.  \n",
    "> - Reactions in the \"Exchange/demand reaction\" subsystem, because they are pseudoreactions of GEM.  \n",
    "> - Reactions in the \"Miscellaneous\" subsystem, because they are difficult to be interpreted as a functional subsystem.  \n",
    "> - Reactions in the \"Transport, xxx\" subsystem, because their functional meaning is dependent on each transporting metabolite and covered by biochemical reactions of the other subsystems.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tempDF1 = rxnDF\n",
    "tempDF2 = fluxDF\n",
    "\n",
    "#Reactions without subsystem annotation\n",
    "tempDF = tempDF1.loc[tempDF1['Subsystem'].isnull()]\n",
    "print('Reactions without subsystem annotation')\n",
    "display(tempDF)\n",
    "tempL1 = tempDF.index.tolist()\n",
    "\n",
    "#Exchange/demand reactions\n",
    "tempDF = tempDF1.loc[tempDF1['Subsystem']=='Exchange/demand reaction']\n",
    "print('Excange/demand reactions')\n",
    "display(tempDF)\n",
    "tempL2 = tempDF.index.tolist()\n",
    "\n",
    "#Miscellaneous reactions\n",
    "tempDF = tempDF1.loc[tempDF1['Subsystem']=='Miscellaneous']\n",
    "print('Miscellaneous reactions')\n",
    "display(tempDF)\n",
    "tempL3 = tempDF.index.tolist()\n",
    "\n",
    "#Transport reactions\n",
    "tempDF = tempDF1.loc[~tempDF1['Subsystem'].isnull()]#To avoid NA error in the next line\n",
    "tempDF = tempDF.loc[tempDF['Subsystem'].str.contains('Transport, ')]\n",
    "print('Transport reactions')\n",
    "display(tempDF)\n",
    "tempL4 = tempDF.index.tolist()\n",
    "\n",
    "#Reactions to be removed\n",
    "tempL = [rxn for sublist in [tempL1, tempL2, tempL3, tempL4] for rxn in sublist]\n",
    "print('nReactions to be removed:', len(tempL))\n",
    "\n",
    "#Remove the reactions from flux data\n",
    "print('Flux DF before:', tempDF2.shape)\n",
    "tempDF = tempDF2.loc[~tempDF2.index.isin(tempL)]\n",
    "\n",
    "display(tempDF)\n",
    "display(tempDF.reset_index().describe(include='all'))\n",
    "\n",
    "#Save\n",
    "fileDir = './ExportData/'\n",
    "ipynbName = '230502_LC-M001-related-TrOmics-GEM-ver3-15_FluxAnalysis_'\n",
    "fileName = 'average-flux-data_selected.tsv'\n",
    "tempDF.to_csv(fileDir+ipynbName+fileName, index=True, sep='\\t')\n",
    "\n",
    "#Check\n",
    "tempS1 = tempDF1['Subsystem'].value_counts()\n",
    "tempS1.name = 'Overall'\n",
    "tempS2 = tempDF1['Subsystem'].loc[tempDF2.index].value_counts()\n",
    "tempS2.name = 'FluxCalculated'\n",
    "tempS3 = tempDF1['Subsystem'].loc[tempDF.index].value_counts()\n",
    "tempS3.name = 'Selected'\n",
    "tempDF3 = pd.concat([tempS1, tempS2, tempS3], axis=1)\n",
    "display(tempDF3.describe())\n",
    "display(tempDF3.loc[tempS1>30])\n",
    "display(tempDF3.loc[tempS1<=30])\n",
    "\n",
    "#Update\n",
    "fluxDF = tempDF"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. Check data structure of the flux values"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2-1. Distribution"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Skewness (and other summary metrics) per sample\n",
    "tempDF = fluxDF.describe()\n",
    "tempDF.loc['Skewness'] = fluxDF.agg(stats.skew, axis=0)\n",
    "display(tempDF.T.describe())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Distribution\n",
    "tempL = sampleDF.sort_values(by='Group', ascending=True)['Group'].unique()\n",
    "global_max = fluxDF.max(axis=1).max()\n",
    "global_min = fluxDF.min(axis=1).min()\n",
    "for group in tempL:\n",
    "    tempDF = sampleDF.loc[sampleDF['Group']==group]\n",
    "    tempDF = fluxDF.loc[:, tempDF.index.tolist()]\n",
    "    sns.set(style='ticks', font='Arial', context='notebook')\n",
    "    plt.figure(figsize=(4, 3))\n",
    "    for sample in tempDF.columns.tolist():\n",
    "        sns.distplot(tempDF[sample], label=sample)\n",
    "    sns.despine()\n",
    "    plt.xlim(global_min, global_max)\n",
    "    plt.ylabel('Density')\n",
    "    plt.xlabel('Maximum flux value')\n",
    "    plt.title('Group: '+group)\n",
    "    plt.legend(bbox_to_anchor=(1, 0.5), loc='center left', borderaxespad=1, ncol=1)\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> According to Priyanka, the 0 and ±1,000 values are truly meaningful because they were resolved at least. Also, the negative flux indicates that inverse direction is preferred at the steady state, which is possible only for the reaction that is defined as reversible one like A + B <=> C + D. In GEM, the “irreversible\" reaction is differently defined such as X + Y -> Z.  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2-2. PCA"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 2-2-1. Standardization"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> Clearly, flux value distribution across analytes is not normal distribution per model, but variables are set as analytes in this PCA. Because the value range is roughly consistent b/w samples, only the simple standardization is applied to analyte value before PCA.  \n",
    "> –> Instead of StandardScaler (z-score), RobustScaler (removed median and scaled by IQR) is used. (Of note, RobustScaler() returns just the centered (i.e., median-subtracted) values if IQR = 0.)  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Robust transformation\n",
    "scaler = RobustScaler(with_centering=True, with_scaling=True, quantile_range=(25, 75),\n",
    "                      copy=True, unit_variance=False)\n",
    "tempA = scaler.fit_transform(fluxDF.T)#axis: column\n",
    "fluxDF_scaled = pd.DataFrame(data=tempA, index=fluxDF.columns, columns=fluxDF.index)\n",
    "fluxDF_scaled.index.rename('SampleID', inplace=True)\n",
    "\n",
    "display(fluxDF_scaled)\n",
    "display(fluxDF_scaled.iloc[:, :15].describe())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 2-2-2. PCA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "nPCs = 10\n",
    "model = PCA(n_components=nPCs, svd_solver='randomized', iterated_power='auto', random_state=123)\n",
    "model.fit(fluxDF_scaled)\n",
    "\n",
    "#Explained variance\n",
    "tempS = pd.Series(data=model.explained_variance_ratio_*100,\n",
    "                  index=['PC'+str(i+1) for i in range(nPCs)], name='ExplainedVariance')\n",
    "print('Percentage of variance explained by each component:')\n",
    "display(tempS)\n",
    "\n",
    "#Scree plot\n",
    "tempDF = tempS.reset_index()\n",
    "tempDF['PC'] = [i+1 for i in range(nPCs)]\n",
    "sns.set(style='ticks', font='Arial', context='talk')\n",
    "plt.figure(figsize=(4, 3))\n",
    "p = sns.lineplot(data=tempDF, x='PC', y='ExplainedVariance', color='k')\n",
    "sns.despine()\n",
    "p.set(xlim=(0.5, 10.5), xticks=np.arange(1, 10.1, 1))\n",
    "plt.ylabel('Explained varaince [%]')\n",
    "plt.xlabel('Principal component number')\n",
    "plt.show()\n",
    "\n",
    "#Label\n",
    "tempL = []\n",
    "for i in range(nPCs):\n",
    "    round_value = Decimal(str(tempS['PC'+str(i+1)])).quantize(Decimal('0.1'), rounding=ROUND_HALF_UP)\n",
    "    tempL.append('PC'+str(i+1)+' ('+str(round_value)+'%)')\n",
    "\n",
    "#Projected space\n",
    "pca_space = pd.DataFrame(data=model.transform(fluxDF_scaled), index=fluxDF_scaled.index, columns=tempL)\n",
    "display(pca_space)\n",
    "\n",
    "#Principal axes in feature space\n",
    "pca_comp = pd.DataFrame(data=model.components_, index=tempL, columns=fluxDF_scaled.columns)\n",
    "display(pca_comp)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 2-2-3. Projected space"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "nPCs = 6\n",
    "\n",
    "#Prepare DF\n",
    "tempDF = sampleDF[['Intervention', 'Sex']]\n",
    "tempDF = pd.merge(tempDF, pca_space.iloc[:, :nPCs],\n",
    "                  left_index=True, right_index=True, how='left')\n",
    "tempDF = tempDF.reset_index()\n",
    "\n",
    "#Color\n",
    "tempD1 = {'Ctrl1':'tab:blue', 'Aca':'tab:red', '17aE2':'tab:green', 'Prot':'tab:brown',\n",
    "          'Rapa':'tab:purple', 'CRdiet':'tab:olive', 'Ctrl2':'tab:blue', 'MRdiet':'tab:orange',\n",
    "          'GHRWT':'tab:gray', 'GHRKO':'tab:pink', 'SnellWT':'black', 'SnellDW':'tab:cyan'}\n",
    "tempD2 = {'F':'o', 'M':'^'}\n",
    "\n",
    "#Visualization\n",
    "sns.set(style='ticks', font='Arial', context='talk')\n",
    "p = sns.PairGrid(data=tempDF, hue='Intervention', hue_order=tempD1.keys(), palette=tempD1)\n",
    "p.map_lower(sns.scatterplot, style=tempDF['Sex'], style_order=tempD2.keys(), markers=tempD2)\n",
    "for i, j in zip(*np.triu_indices_from(p.axes, 0)):\n",
    "    p.axes[i, j].set_visible(False)\n",
    "p.add_legend(bbox_to_anchor=(0.7, 0.7), loc='upper right', frameon=True, title='')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> –> Some points are missing???  \n",
    "> –> They were just completely overlapped.  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3. Compare flux values"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> 1. Compare flux values between control vs. each intervention using Mann–Whitney U-test.  \n",
    ">\n",
    "> To increase the statistical power, samples are grouped by intervention (i.e., strain, age, and sex are pooled). Note that the dataset was adjusted with them before GEM reconstruction. Although tricky, the P-value adjustment is NOT performed across reactions because this step is regarded as an intermediate process for the enrichment analysis to assess each intervention effect on GEM subsystems. As well, the P-value adjustment across interventions is just for reference purpose.  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3-1. Mann–Whitney U-test"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 3-1-1. Perform all statistical tests"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> Note that the scipy API (scipy.stats.mannwhitneyu) is used, because only the one-sided Mann–Whitney U-test seems implemented in the current statsmodels API (statsmodels.stats.nonparametric.rank_compare_2indep). Actually, the output objects of Mann–Whitney U-test are same b/w the two APIs, which is contrast to the case of t-test (degrees of freedom is not reported in the scipy API).  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tempL1 = ['Ctrl1', 'Aca', 'Rapa', 'CRdiet']#Target sample groups to be assessed\n",
    "tempL2 = categoryL#Control type for post-hoc comparisons\n",
    "tempDF1 = fluxDF\n",
    "tempDF2 = sampleDF.loc[sampleDF['Intervention'].isin(tempL1)]\n",
    "\n",
    "#Statistical tests per reaction\n",
    "t_start = time.time()\n",
    "tempL3 = []#For test summary\n",
    "for rxn in tempDF1.index:\n",
    "    #Select the target reaction\n",
    "    tempS = tempDF1.loc[rxn]\n",
    "    tempS.name = 'Flux'\n",
    "    #Add metadata while selecting the target samples\n",
    "    tempDF = pd.merge(tempS, tempDF2, left_index=True, right_index=True, how='inner')\n",
    "    \n",
    "    #Test per control vs. contrast\n",
    "    tempDF3 = pd.DataFrame(columns=['Ustat', 'Pval'])\n",
    "    for control in tempL2:\n",
    "        for contrast in tempL1:\n",
    "            tempA = tempDF2['Intervention'].loc[tempDF2['Category']==control].unique()\n",
    "            if (contrast in tempA) & (control!=contrast):\n",
    "                tempS1 = tempDF['Flux'].loc[tempDF['Intervention']==control]\n",
    "                tempS2 = tempDF['Flux'].loc[tempDF['Intervention']==contrast]\n",
    "                #Two-sided Mann–Whitney U-test\n",
    "                ustat, pval = stats.mannwhitneyu(tempS2, tempS1,#U-statistic corresponds to the contrast\n",
    "                                                 use_continuity=True, alternative='two-sided', method='auto')\n",
    "                tempDF3.loc[contrast+'-vs-'+control] = [ustat, pval]\n",
    "    ##P-value adjustment across all comparisons per reaction by using Benjamini–Hochberg method (just for reference)\n",
    "    tempDF3['AdjPval'] = multi.multipletests(tempDF3['Pval'], alpha=0.05, method='fdr_bh',\n",
    "                                             is_sorted=False, returnsorted=False)[1]\n",
    "    ##Convert to wide-format\n",
    "    tempL = []\n",
    "    for comparison in tempDF3.index:\n",
    "        tempS = tempDF3.loc[comparison]\n",
    "        tempS.index = comparison+'_'+tempS.index\n",
    "        tempS.name = rxn\n",
    "        tempL.append(tempS)\n",
    "    tempS = pd.concat(tempL, axis=0)\n",
    "    tempL3.append(tempS)\n",
    "t_elapsed = time.time() - t_start\n",
    "print('Elapsed time for', (len(tempL1)-len(tempL2))*len(tempDF1), 'tests (',\n",
    "      len(tempL1)-len(tempL2), 'comparisons x', len(tempDF1), 'reactions):',\n",
    "      round(t_elapsed//60), 'min', round(t_elapsed%60, 1), 'sec')\n",
    "\n",
    "#Generate test summary table\n",
    "tempDF3 = pd.concat(tempL3, axis=1).T\n",
    "tempDF3.index.name = tempDF1.index.name\n",
    "display(tempDF3)\n",
    "\n",
    "statDF = tempDF3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tempL1 = ['Ctrl1', 'Aca', 'Rapa', 'CRdiet']#Target sample groups to be summarized\n",
    "tempDF1 = fluxDF\n",
    "tempDF2 = sampleDF.loc[sampleDF['Intervention'].isin(tempL1)]\n",
    "\n",
    "#Calculate general statistics per intervention group\n",
    "tempL2 = []\n",
    "for intervention in tempL1:\n",
    "    #Select the target samples\n",
    "    tempL = tempDF2.loc[tempDF2['Intervention']==intervention].index.tolist()\n",
    "    tempDF = tempDF1[tempL]\n",
    "    #Calculate general statistics\n",
    "    tempS1 = len(tempL) - tempDF.isnull().sum(axis=1)\n",
    "    tempS1.name = intervention+'_N'\n",
    "    tempS2 = tempDF.mean(axis=1)\n",
    "    tempS2.name = intervention+'_FluxMean'\n",
    "    tempS3 = tempDF.median(axis=1)\n",
    "    tempS3.name = intervention+'_FluxMedian'\n",
    "    tempS4 = tempDF.apply(stats.median_absolute_deviation, axis=1)#Cf. pd.DataFrame.mad() is not median absolute deviation but mean absolute deviation\n",
    "    tempS4.name = intervention+'_FluxMAD'\n",
    "    #Merge\n",
    "    tempDF = pd.concat([tempS1, tempS2, tempS3, tempS4], axis=1)\n",
    "    tempL2.append(tempDF)\n",
    "tempDF = pd.concat(tempL2, axis=1)\n",
    "display(tempDF)\n",
    "\n",
    "#Merge all the tables\n",
    "print('General statistics table:', tempDF.shape)\n",
    "print('Test summary table:', statDF.shape)\n",
    "tempDF = pd.concat([tempDF, statDF], axis=1)\n",
    "tempS = rxnDF['ReactionName']\n",
    "tempDF = pd.merge(tempS, tempDF, left_index=True, right_index=True, how='right')\n",
    "\n",
    "display(tempDF)\n",
    "\n",
    "#Save\n",
    "fileDir = './ExportData/'\n",
    "ipynbName = '230502_LC-M001-related-TrOmics-GEM-ver3-15_FluxAnalysis_'\n",
    "fileName = 'flux-comparison_vs-each-control.tsv'\n",
    "tempDF.to_csv(fileDir+ipynbName+fileName, sep='\\t', index=True)\n",
    "\n",
    "#Update\n",
    "statDF = tempDF"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 3-1-2. Potentially changed reactions by each intervention (nominal P-value)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tempDF = statDF\n",
    "\n",
    "#Take nominal P-value\n",
    "tempDF1 = tempDF.loc[:, tempDF.columns.str.contains('-vs-.*_Pval$')]\n",
    "tempDF1.columns = tempDF1.columns.str.replace('_Pval$', '')\n",
    "tempDF1 = pd.merge(tempDF['ReactionName'], tempDF1,\n",
    "                   left_index=True, right_index=True, how='left')\n",
    "print('Nominal P-value:')\n",
    "display(tempDF1)\n",
    "display(tempDF1.describe())\n",
    "tempL = tempDF1.loc[:, tempDF1.columns.str.contains('-vs-')].columns.tolist()\n",
    "for comparison in tempL:\n",
    "    tempS = tempDF1[comparison]\n",
    "    tempS = tempS.loc[tempS<0.05]\n",
    "    print(' - '+comparison+':', len(tempS))\n",
    "\n",
    "#Take U-statistic difference from the null (for direction)\n",
    "tempDF2 = tempDF.loc[:, tempDF.columns.str.contains('-vs-.*_Ustat$')]\n",
    "tempDF2.columns = tempDF2.columns.str.replace('_Ustat$', '')\n",
    "for comparison in tempDF2.columns:\n",
    "    #Take total U of control and contrast\n",
    "    control = re.sub('^.*-vs-', '', comparison)\n",
    "    contrast = re.sub('-vs-.*', '', comparison)\n",
    "    tempS1 = tempDF[control+'_N']\n",
    "    tempS2 = tempDF[contrast+'_N']\n",
    "    tempS = tempS1 * tempS2\n",
    "    #Calculate the difference from the null\n",
    "    tempDF2[comparison] = tempDF2[comparison] - tempS/2\n",
    "tempDF2 = pd.merge(tempDF['ReactionName'], tempDF2,\n",
    "                   left_index=True, right_index=True, how='left')\n",
    "print('Changed direction (U-statistic difference from the null):')\n",
    "display(tempDF2)\n",
    "display(tempDF2.describe())\n",
    "\n",
    "pvalDF = tempDF1\n",
    "diffDF = tempDF2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> Check the changed reactions (based on the adjusted P-values across interventions) as reference.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tempDF = statDF\n",
    "\n",
    "#Take nominal P-value\n",
    "tempDF1 = tempDF.loc[:, tempDF.columns.str.contains('-vs-.*_AdjPval$')]\n",
    "tempDF1.columns = tempDF1.columns.str.replace('_AdjPval$', '')\n",
    "tempDF1 = pd.merge(tempDF['ReactionName'], tempDF1,\n",
    "                   left_index=True, right_index=True, how='left')\n",
    "print('Adjusted P-value:')\n",
    "display(tempDF1)\n",
    "display(tempDF1.describe())\n",
    "tempL = tempDF1.loc[:, tempDF1.columns.str.contains('-vs-')].columns.tolist()\n",
    "for comparison in tempL:\n",
    "    tempS = tempDF1[comparison]\n",
    "    tempS = tempS.loc[tempS<0.05]\n",
    "    print(' - '+comparison+':', len(tempS))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "### 3-2. Visualization: clustermap"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 3-2-1. Sample-wide"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> For visualization purpose, it would be better to use the centered values based on control groups, because around zero value simply indicates no change from control.  \n",
    "> –> At the same time, scaling is required for good clustering, but RobustScaler() returns just the centered (i.e., median-subtracted) values if IQR = 0. Hence, not IQR but the maximum absolete value is used for scaling. Of note, scaling eliminates the information about how largely an intervention group is different from the control group, but it rather reflects the rank-based statistical tests.  \n",
    "\n",
    "> ***Skip this!***  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 3-2-2. Group-wide 1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> Sample-based clustering would be difficult to be understood.  \n",
    "> –> Combine samples per intervention group. Of note, not median but mean is used as the summary statistic because the median loses many information in the case of small sample size.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Summarize samples per intervention group\n",
    "tempL = ['Ctrl1', 'Aca', 'Rapa', 'CRdiet']\n",
    "tempDF = pd.DataFrame(index=fluxDF.index)\n",
    "for intervention in tempL:\n",
    "    tempL = sampleDF.loc[sampleDF['Intervention']==intervention].index.tolist()\n",
    "    tempDF1 = fluxDF.loc[:, tempL]\n",
    "    #Calculate the mean\n",
    "    tempS = tempDF1.mean(axis=1)\n",
    "    tempS.name = intervention\n",
    "    #Merge\n",
    "    tempDF = pd.merge(tempDF, tempS, left_index=True, right_index=True, how='left')\n",
    "print('Summary per intervention group:')\n",
    "display(tempDF.describe())\n",
    "\n",
    "#Centering per each control category\n",
    "tempDF1 = pd.DataFrame(index=tempDF.index)\n",
    "for category in categoryL:\n",
    "    tempL = sampleDF['Intervention'].loc[sampleDF['Category']==category].unique().tolist()\n",
    "    tempDF2 = tempDF.loc[:, tempL]\n",
    "    #Centering\n",
    "    tempS = tempDF2[category]\n",
    "    tempDF2 = tempDF2.T - tempS\n",
    "    #Merge\n",
    "    tempDF1 = pd.merge(tempDF1, tempDF2.T, left_index=True, right_index=True, how='left')\n",
    "\n",
    "#Scaling\n",
    "scaler = MaxAbsScaler(copy=True)\n",
    "tempA = scaler.fit_transform(tempDF1.T)#axis: column\n",
    "tempDF2 = pd.DataFrame(data=tempA, index=tempDF1.columns, columns=tempDF1.index)\n",
    "tempDF2 = tempDF2.T\n",
    "tempL = ['Ctrl1', 'Aca', 'Rapa', 'CRdiet']\n",
    "tempDF2 = tempDF2[tempL]#Sort\n",
    "\n",
    "print('After centering and scaling:')\n",
    "display(tempDF2)\n",
    "display(tempDF2.describe())\n",
    "\n",
    "fluxDF_scaled = tempDF2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> –> Remove control groups because it is meaningless after centering.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Prepare DF to plot\n",
    "tempDF = fluxDF_scaled.copy()\n",
    "##Remove control groups\n",
    "tempDF = tempDF.drop(columns=categoryL)\n",
    "\n",
    "#Prepare color labels for the changed reactions\n",
    "regulation = 'Changed'\n",
    "tempD0 = {'Ctrl1':'Control', 'Aca':'Acarbose',\n",
    "          'Rapa':'Rapamycin', 'CRdiet':'CR diet'}\n",
    "tempDF.columns = tempDF.columns.map(tempD0)\n",
    "tempD1 = {'Control':'tab:blue', 'Acarbose':'tab:red',\n",
    "          'Rapamycin':'tab:purple', 'CR diet':'tab:olive'}\n",
    "tempDF1 = pvalDF.loc[tempDF.index.tolist(), pvalDF.columns.str.contains('-vs-')]\n",
    "for col_n in tempDF1.columns.tolist():\n",
    "    tempS1 = pvalDF[col_n]\n",
    "    tempS2 = diffDF[col_n]\n",
    "    if regulation=='Changed':\n",
    "        tempS2 = tempS2.loc[(tempS1<0.05)]\n",
    "    elif regulation=='Increased':\n",
    "        tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2>0)]\n",
    "    elif regulation=='Decreased':\n",
    "        tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2<0)]\n",
    "    #Replace dummy value (P-value) with color code\n",
    "    label = tempD0[re.sub('-vs-.*', '', col_n)]\n",
    "    tempL = []\n",
    "    count = 0#just for checking\n",
    "    for rxn in tempDF1.index.tolist():\n",
    "        if rxn in tempS2.index.tolist():\n",
    "            tempL.append(tempD1[label])\n",
    "            count += 1#Just for checking\n",
    "        else:\n",
    "            tempL.append('white')\n",
    "    tempDF1[label] = tempL\n",
    "    print(regulation+' reaction in '+col_n)\n",
    "    print(' -> nominal P < 0.05:', count)\n",
    "tempDF1 = tempDF1.loc[:, tempDF1.columns.isin(tempD1.keys())]\n",
    "##Remove the group showing no changed reactions\n",
    "#for col_n in tempDF1.columns.tolist():\n",
    "#    tempS = tempDF1[col_n]\n",
    "#    tempS = (tempS=='white')\n",
    "#    if tempS.sum()==len(tempS):\n",
    "#        tempDF1 = tempDF1.drop(columns=col_n)\n",
    "\n",
    "#Clustermap\n",
    "sns.set(style='ticks', font='Arial', context='talk')\n",
    "cm = sns.clustermap(tempDF.T, method='ward', metric='euclidean', cmap='RdBu_r',\n",
    "                    row_cluster=True, col_cluster=True, row_linkage=None, col_linkage=None,\n",
    "                    row_colors=None, col_colors=tempDF1, xticklabels=False, yticklabels=True,\n",
    "                    dendrogram_ratio=(0.025, 0.2), colors_ratio=(0.01, 0.1),\n",
    "                    cbar_pos=(1.02, 0.25, 0.3, 0.06), cbar_kws={'orientation': 'horizontal'},\n",
    "                    figsize=(15, 3), **{'vmin':-1, 'vmax':1})\n",
    "cm.cax.set_title('Group mean of flux values (vs. Control; scaled)', size='medium',\n",
    "                 verticalalignment='bottom', horizontalalignment='center')\n",
    "cm.cax.tick_params(labelsize='small')\n",
    "bottom, top = cm.ax_heatmap.get_ylim()\n",
    "#cm.ax_heatmap.set_ylim(bottom + 0.5, top - 0.5)##To avoid half cut of first and last rows\n",
    "hm = cm.ax_heatmap.get_position()\n",
    "rd = cm.ax_row_dendrogram.get_position()\n",
    "cd = cm.ax_col_dendrogram.get_position()\n",
    "cm.ax_heatmap.set_position([hm.x0, hm.y0, hm.width, hm.height])\n",
    "cm.ax_row_dendrogram.set_position([rd.x0, rd.y0, rd.width, rd.height])\n",
    "cm.ax_col_dendrogram.set_position([cd.x0, cd.y0, cd.width, cd.height])\n",
    "cm.ax_heatmap.set_xlabel('Reaction')\n",
    "cm.ax_heatmap.set_ylabel('')\n",
    "##row/column color bar legend (axis is same with cm.cax!)\n",
    "tempL = []\n",
    "for group in tempD1.keys():\n",
    "    if group in tempDF1.columns.tolist():\n",
    "        tempL.append(mpatches.Patch(color=tempD1[group], label='by '+group))\n",
    "legend = plt.legend(handles=tempL, fontsize='small', labelspacing=0.2, ncol=1,\n",
    "                    title='Changed reactions (vs. Control)', title_fontsize='medium',\n",
    "                    bbox_to_anchor=(0.5, 1.0), loc='lower center', borderaxespad=3, frameon=False)\n",
    "plt.gca().add_artist(legend)\n",
    "##Save\n",
    "fileDir = './ExportFigures/'\n",
    "ipynbName = '230502_LC-M001-related-TrOmics-GEM-ver3-15_FluxAnalysis_'\n",
    "fileName = 'flux-clustermap-across-groups.pdf'\n",
    "plt.gcf().savefig(fileDir+ipynbName+fileName, dpi=300, bbox_inches='tight', pad_inches=0.04, transparent=True)\n",
    "plt.show()\n",
    "\n",
    "#Save label order\n",
    "tempDF = pvalDF.loc[tempDF.index[cm.dendrogram_col.reordered_ind]]\n",
    "tempDF = tempDF.reset_index()\n",
    "tempDF.index.name = 'Xcoord'\n",
    "display(tempDF)\n",
    "fileDir = './ExportData/'\n",
    "ipynbName = '230502_LC-M001-related-TrOmics-GEM-ver3-15_FluxAnalysis_'\n",
    "fileName = 'flux-clustermap-across-groups_ticks-order.tsv'\n",
    "tempDF.to_csv(fileDir+ipynbName+fileName, index=True, sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 3-2-3. Group-wide 2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> Centering by each control may highlight the difference too strongly, because each control group is forced to be around zero.  \n",
    "> –> Check simple standardization (Z-score) across all groups.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Summarize samples per intervention group\n",
    "tempL = ['Ctrl1', 'Aca', 'Rapa', 'CRdiet']\n",
    "tempDF = pd.DataFrame(index=fluxDF.index)\n",
    "for intervention in tempL:\n",
    "    tempL = sampleDF.loc[sampleDF['Intervention']==intervention].index.tolist()\n",
    "    tempDF1 = fluxDF.loc[:, tempL]\n",
    "    #Calculate the mean\n",
    "    tempS = tempDF1.mean(axis=1)\n",
    "    tempS.name = intervention\n",
    "    #Merge\n",
    "    tempDF = pd.merge(tempDF, tempS, left_index=True, right_index=True, how='left')\n",
    "print('Summary per intervention group:')\n",
    "display(tempDF.describe())\n",
    "\n",
    "#Z-score transformation\n",
    "scaler = StandardScaler(copy=True, with_mean=True, with_std=True)\n",
    "tempA = scaler.fit_transform(tempDF.T)#axis: column\n",
    "tempDF1 = pd.DataFrame(data=tempA, index=tempDF.columns, columns=tempDF.index)\n",
    "tempDF1 = tempDF1.T\n",
    "tempL = ['Ctrl1', 'Aca', 'Rapa', 'CRdiet']\n",
    "tempDF1 = tempDF1[tempL]#Sort\n",
    "\n",
    "print('After centering and scaling:')\n",
    "display(tempDF1)\n",
    "display(tempDF1.describe())\n",
    "\n",
    "fluxDF_scaled = tempDF1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Prepare DF to plot\n",
    "tempDF = fluxDF_scaled.copy()\n",
    "##Remove control groups\n",
    "#tempDF = tempDF.drop(columns=categoryL)\n",
    "\n",
    "#Prepare color labels for the changed reactions\n",
    "regulation = 'Changed'\n",
    "tempD0 = {'Ctrl1':'Control', 'Aca':'Acarbose',\n",
    "          'Rapa':'Rapamycin', 'CRdiet':'CR diet'}\n",
    "tempDF.columns = tempDF.columns.map(tempD0)\n",
    "tempD1 = {'Control':'tab:blue', 'Acarbose':'tab:red',\n",
    "          'Rapamycin':'tab:purple', 'CR diet':'tab:olive'}\n",
    "tempDF1 = pvalDF.loc[tempDF.index.tolist(), pvalDF.columns.str.contains('-vs-')]\n",
    "for col_n in tempDF1.columns.tolist():\n",
    "    tempS1 = pvalDF[col_n]\n",
    "    tempS2 = diffDF[col_n]\n",
    "    if regulation=='Changed':\n",
    "        tempS2 = tempS2.loc[(tempS1<0.05)]\n",
    "    elif regulation=='Increased':\n",
    "        tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2>0)]\n",
    "    elif regulation=='Decreased':\n",
    "        tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2<0)]\n",
    "    #Replace dummy value (P-value) with color code\n",
    "    label = tempD0[re.sub('-vs-.*', '', col_n)]\n",
    "    tempL = []\n",
    "    count = 0#just for checking\n",
    "    for rxn in tempDF1.index.tolist():\n",
    "        if rxn in tempS2.index.tolist():\n",
    "            tempL.append(tempD1[label])\n",
    "            count += 1#Just for checking\n",
    "        else:\n",
    "            tempL.append('white')\n",
    "    tempDF1[label] = tempL\n",
    "    print(regulation+' reaction in '+col_n)\n",
    "    print(' -> nominal P < 0.05:', count)\n",
    "tempDF1 = tempDF1.loc[:, tempDF1.columns.isin(tempD1.keys())]\n",
    "##Remove the group showing no changed reactions\n",
    "#for col_n in tempDF1.columns.tolist():\n",
    "#    tempS = tempDF1[col_n]\n",
    "#    tempS = (tempS=='white')\n",
    "#    if tempS.sum()==len(tempS):\n",
    "#        tempDF1 = tempDF1.drop(columns=col_n)\n",
    "\n",
    "#Clustermap\n",
    "sns.set(style='ticks', font='Arial', context='talk')\n",
    "cm = sns.clustermap(tempDF.T, method='ward', metric='euclidean', cmap='RdBu_r',\n",
    "                    row_cluster=True, col_cluster=True, row_linkage=None, col_linkage=None,\n",
    "                    row_colors=None, col_colors=tempDF1, xticklabels=False, yticklabels=True,\n",
    "                    dendrogram_ratio=(0.025, 0.2), colors_ratio=(0.01, 0.1),\n",
    "                    cbar_pos=(1.02, 0.25, 0.3, 0.06), cbar_kws={'orientation': 'horizontal'},\n",
    "                    figsize=(15, 3), **{'vmin':-2, 'vmax':2})\n",
    "cm.cax.set_title('Group mean of flux values ('+r'$Z$'+'-score)', size='medium',\n",
    "                 verticalalignment='bottom', horizontalalignment='center')\n",
    "cm.cax.tick_params(labelsize='small')\n",
    "bottom, top = cm.ax_heatmap.get_ylim()\n",
    "#cm.ax_heatmap.set_ylim(bottom + 0.5, top - 0.5)##To avoid half cut of first and last rows\n",
    "hm = cm.ax_heatmap.get_position()\n",
    "rd = cm.ax_row_dendrogram.get_position()\n",
    "cd = cm.ax_col_dendrogram.get_position()\n",
    "cm.ax_heatmap.set_position([hm.x0, hm.y0, hm.width, hm.height])\n",
    "cm.ax_row_dendrogram.set_position([rd.x0, rd.y0, rd.width, rd.height])\n",
    "cm.ax_col_dendrogram.set_position([cd.x0, cd.y0, cd.width, cd.height])\n",
    "cm.ax_heatmap.set_xlabel('Reaction')\n",
    "cm.ax_heatmap.set_ylabel('')\n",
    "##row/column color bar legend (axis is same with cm.cax!)\n",
    "tempL = []\n",
    "for group in tempD1.keys():\n",
    "    if group in tempDF1.columns.tolist():\n",
    "        tempL.append(mpatches.Patch(color=tempD1[group], label='by '+group))\n",
    "legend = plt.legend(handles=tempL, fontsize='small', labelspacing=0.2, ncol=1,\n",
    "                    title='Changed reactions (vs. Control)', title_fontsize='medium',\n",
    "                    bbox_to_anchor=(0.5, 1.0), loc='lower center', borderaxespad=3, frameon=False)\n",
    "plt.gca().add_artist(legend)\n",
    "##Save\n",
    "fileDir = './ExportFigures/'\n",
    "ipynbName = '230502_LC-M001-related-TrOmics-GEM-ver3-15_FluxAnalysis_'\n",
    "fileName = 'flux-clustermap-across-groups.pdf'\n",
    "#plt.gcf().savefig(fileDir+ipynbName+fileName, dpi=300, bbox_inches='tight', pad_inches=0.04, transparent=True)\n",
    "plt.show()\n",
    "\n",
    "#Save label order\n",
    "tempDF = pvalDF.loc[tempDF.index[cm.dendrogram_col.reordered_ind]]\n",
    "tempDF = tempDF.reset_index()\n",
    "tempDF.index.name = 'Xcoord'\n",
    "display(tempDF)\n",
    "fileDir = './ExportData/'\n",
    "ipynbName = '230502_LC-M001-related-TrOmics-GEM-ver3-15_FluxAnalysis_'\n",
    "fileName = 'flux-clustermap-across-groups_ticks-order.tsv'\n",
    "#tempDF.to_csv(fileDir+ipynbName+fileName, index=True, sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> –> This representation is informative for us, but probably normal readers would feel difficulty to understand the figure message...  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3-3. Visualization: venn diagram"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Prepare label and color\n",
    "tempD0 = {'Ctrl1':'Control', 'Aca':'Acarbose',\n",
    "          'Rapa':'Rapamycin', 'CRdiet':'CR diet'}\n",
    "tempD1 = {'Acarbose':'tab:red', 'Rapamycin':'tab:purple', 'CR diet':'tab:olive'}\n",
    "\n",
    "#Visualization per direction\n",
    "for regulation in ['Changed', 'Increased', 'Decreased']:\n",
    "    #Prepare reaction sets\n",
    "    tempL = statDF.index.tolist()\n",
    "    tempDF1 = pvalDF.loc[tempL, pvalDF.columns.str.contains('-vs-')]\n",
    "    tempDF2 = diffDF.loc[tempL, diffDF.columns.str.contains('-vs-')]\n",
    "    tempD2 = {}\n",
    "    count = 0\n",
    "    for col_n in tempDF1.columns.tolist():\n",
    "        tempS1 = tempDF1[col_n]\n",
    "        tempS2 = tempDF2[col_n]\n",
    "        if regulation=='Changed':\n",
    "            tempS2 = tempS2.loc[(tempS1<0.05)]\n",
    "        elif regulation=='Increased':\n",
    "            tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2>0)]\n",
    "        elif regulation=='Decreased':\n",
    "            tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2<0)]\n",
    "        if len(tempS2)>0:\n",
    "            count += 1\n",
    "        label = tempD0[re.sub('-vs-.*', '', col_n)]\n",
    "        tempD2[label] = set(tempS2.index.tolist())\n",
    "    if count==0:\n",
    "        print(regulation+' reactions: no significance in any comparisons')\n",
    "        continue\n",
    "    ##Sort to make consistent order in manual legend generation\n",
    "    tempD = {}\n",
    "    count = 0\n",
    "    for label in tempD1.keys():\n",
    "        tempD[label] = tempD2[label]\n",
    "        if len(tempD2[label])>0:\n",
    "            count += 1\n",
    "    \n",
    "    #Skip the followings if no significant reaction\n",
    "    if count==0:\n",
    "        print(regulation+' reactions: no significance in any target comparisons')\n",
    "        continue\n",
    "    \n",
    "    #Venn diagram\n",
    "    sns.set(style='ticks', font='Arial', context='talk')\n",
    "    fig, ax = plt.subplots(figsize=(4, 4))\n",
    "    venn(tempD, fmt='{size:,}', cmap=list(tempD1.values()), legend_loc=None, ax=ax)\n",
    "    plt.setp(ax, ylim=(0.05, 0.975))#Otherwise, weird space...\n",
    "    ##Add legend annotation\n",
    "    x_coord = [0.1, 0.9, 0.8]\n",
    "    y_coord = [0.8, 0.8, 0.25]\n",
    "    h_align = ['right', 'left', 'left']\n",
    "    v_align = ['bottom', 'bottom', 'top']\n",
    "    for i in range(len(tempD1)):\n",
    "        key = list(tempD1.keys())[i]\n",
    "        total = f'{len(tempD[key]):,}'\n",
    "        ax.text(x_coord[i], y_coord[i], key+'\\n('+total+' modules)',\n",
    "                fontsize='small', multialignment='center',\n",
    "                horizontalalignment=h_align[i], verticalalignment=v_align[i],\n",
    "                bbox={'boxstyle':'round', 'facecolor':tempD1[key], 'pad':0.2, 'alpha':0.5})\n",
    "    title = regulation+' reactions (vs. Control)'\n",
    "    ax.set_title(title, fontsize='medium')\n",
    "    ##Save\n",
    "    if regulation!='Changed':\n",
    "        fileDir = './ExportFigures/'\n",
    "        ipynbName = '230502_LC-M001-related-TrOmics-GEM-ver3-15_FluxAnalysis_'\n",
    "        fileName = 'rxn-venn-'+regulation.lower()+'.pdf'\n",
    "        plt.gcf().savefig(fileDir+ipynbName+fileName, dpi=300, bbox_inches='tight', pad_inches=0.04, transparent=True)\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Export reaction list in each subset in the venn diagram\n",
    "for regulation in ['Increased', 'Decreased']:\n",
    "    #Prepare reaction sets\n",
    "    tempL = statDF.index.tolist()\n",
    "    tempDF1 = pvalDF.loc[tempL, pvalDF.columns.str.contains('-vs-')]\n",
    "    tempDF2 = diffDF.loc[tempL, diffDF.columns.str.contains('-vs-')]\n",
    "    tempD = {}\n",
    "    count = 0\n",
    "    for col_n in tempDF1.columns.tolist():\n",
    "        tempS1 = tempDF1[col_n]\n",
    "        tempS2 = tempDF2[col_n]\n",
    "        if regulation=='Changed':\n",
    "            tempS2 = tempS2.loc[(tempS1<0.05)]\n",
    "        elif regulation=='Increased':\n",
    "            tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2>0)]\n",
    "        elif regulation=='Decreased':\n",
    "            tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2<0)]\n",
    "        if len(tempS2)>0:\n",
    "            count += 1\n",
    "        tempD[col_n] = set(tempS2.index.tolist())\n",
    "    \n",
    "    #Skip the followings if no significant reaction\n",
    "    if count==0:\n",
    "        print(regulation+' reactions: no significance in any target comparisons')\n",
    "        continue\n",
    "    \n",
    "    #Prepare a new .xlsx file (dummy README)\n",
    "    tempL1 = [len(tempD[key]) for key in tempD.keys()]\n",
    "    tempDF = pd.DataFrame({'Group':tempD.keys(), 'nReactions':tempL1})\n",
    "    tempDF = tempDF.reset_index().rename(columns={'index':'VennOrder'})\n",
    "    tempDF['VennOrder'] = tempDF['VennOrder'] + 1\n",
    "    fileDir = './ExportData/'\n",
    "    ipynbName = '230502_LC-M001-related-TrOmics-GEM-ver3-15_FluxAnalysis_'\n",
    "    fileName = 'rxn-venn-'+regulation.lower()+'.xlsx'\n",
    "    tempDF.to_excel(fileDir+ipynbName+fileName, sheet_name='README', header=True, index=False)\n",
    "    display(tempDF)#Check\n",
    "    \n",
    "    t_start = time.time()\n",
    "    #Extract overall set\n",
    "    for key_i in range(len(tempD)):\n",
    "        key = list(tempD.keys())[key_i]\n",
    "        tempS = tempD[key]\n",
    "        tempDF = statDF.loc[statDF.index.isin(tempS)]\n",
    "        #Save summary table by appended to the above .xlsx file\n",
    "        ##Prepare sheet name\n",
    "        tempL1 = ['NA' for i in range(len(tempD))]\n",
    "        tempL1[key_i] = '1'\n",
    "        setName = '('+','.join(tempL1)+')'\n",
    "        with pd.ExcelWriter(fileDir+ipynbName+fileName, mode='a', engine='openpyxl') as writer:\n",
    "            tempDF.to_excel(writer, sheet_name=setName, header=True, index=True)\n",
    "        print(' - '+setName+':', len(tempDF))\n",
    "    \n",
    "    #Extract subset\n",
    "    tempL1 = ['1', '0']\n",
    "    tempL2 = [[k1, k2, k3] for k1 in tempL1 for k2 in tempL1 for k3 in tempL1]\n",
    "    #tempL2.remove(['0', '0', '0', '0', '0'])\n",
    "    for tempL1 in tempL2:\n",
    "        #Positive reaction set\n",
    "        tempL3 = [list(tempD.values())[key_i] for key_i, binary in enumerate(tempL1) if binary=='1']\n",
    "        tempS1 = set(statDF.index.tolist())#Initialize\n",
    "        for tempS in tempL3:\n",
    "            tempS1 = tempS1 & tempS\n",
    "        #Negative reaction set\n",
    "        tempL3 = [list(tempD.values())[key_i] for key_i, binary in enumerate(tempL1) if binary=='0']\n",
    "        tempS2 = set()#Initialize\n",
    "        for tempS in tempL3:\n",
    "            tempS2 = tempS2 | tempS\n",
    "        #Extract subset\n",
    "        tempS = tempS1 - tempS2\n",
    "        tempDF = statDF.loc[statDF.index.isin(tempS)]\n",
    "        #Save summary table by appended to the above .xlsx file\n",
    "        ##Prepare sheet name\n",
    "        setName = '('+','.join(tempL1)+')'\n",
    "        with pd.ExcelWriter(fileDir+ipynbName+fileName, mode='a', engine='openpyxl') as writer:\n",
    "            tempDF.to_excel(writer, sheet_name=setName, header=True, index=True)\n",
    "        print(' - '+setName+':', len(tempDF))\n",
    "    \n",
    "    t_elapsed = time.time() - t_start\n",
    "    print(' - Elapsed time:', round(t_elapsed//60), 'min', round(t_elapsed%60, 1), 'sec')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3-4. Visualization: boxplot"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 3-4-1. Reactions increased by Aca and CR"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Prepare the target reaction set\n",
    "posL = ['Aca-vs-Ctrl1', 'CRdiet-vs-Ctrl1']\n",
    "negL = ['']\n",
    "regulation = 'Increased'\n",
    "tempL = statDF.index.tolist()\n",
    "tempDF1 = pvalDF.loc[tempL, pvalDF.columns.str.contains('-vs-')]\n",
    "tempDF2 = diffDF.loc[tempL, diffDF.columns.str.contains('-vs-')]\n",
    "tempS = pd.Series(np.repeat(True, len(tempDF1)), index=tempDF1.index)#Initialize\n",
    "for col_n in tempDF1.columns.tolist():\n",
    "    tempS1 = tempDF1[col_n]\n",
    "    tempS2 = tempDF2[col_n]\n",
    "    if col_n in posL:\n",
    "        if regulation=='Increased':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2>0)\n",
    "        elif regulation=='Decreased':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2<0)\n",
    "    elif col_n in negL:\n",
    "        tempS3 = (tempS1>=0.05)\n",
    "        #Significance for inverse regulation\n",
    "        if regulation=='Increased':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2<0)\n",
    "        elif regulation=='Decreased':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2>0)\n",
    "        tempS1 = tempS3 | tempS1\n",
    "    else:\n",
    "        tempS1 = (tempS1>=0.0)\n",
    "    #Update True\n",
    "    tempS = tempS & tempS1\n",
    "tempL = tempS.loc[tempS.tolist()].index.tolist()\n",
    "print(len(tempL), regulation.lower()+' reactions with significance in', posL, 'but not in', negL)\n",
    "\n",
    "#Select representatives\n",
    "topX = np.min([30, len(tempL)])\n",
    "topX_plot = np.min([5, len(tempL)])\n",
    "tempDF = statDF.loc[:, statDF.columns.str.contains('Pval$')]\n",
    "tempDF = pd.merge(rxnDF[['ReactionName', 'Subsystem']], tempDF,\n",
    "                  left_index=True, right_index=True, how='left')\n",
    "tempL1 = [comparison+'_Pval' for comparison in posL]\n",
    "tempDF = tempDF.loc[tempL].sort_values(by=tempL1, ascending=True)\n",
    "print('Top', topX, 'reactions (sort by', posL, '):')\n",
    "display(tempDF.iloc[:topX])\n",
    "plotL = tempDF.index.tolist()[:topX_plot]\n",
    "\n",
    "#Prepare DF for plot\n",
    "tempDF = fluxDF.reset_index().melt(var_name='ModelID', value_name='Flux', id_vars='ReactionID')\n",
    "tempDF1 = sampleDF.reset_index()[['ModelID', 'Intervention']]\n",
    "tempDF = pd.merge(tempDF, tempDF1, on='ModelID', how='left')\n",
    "\n",
    "#Prepare label and color\n",
    "tempD0 = {'Ctrl1':'Control', 'Aca':'Acarbose',\n",
    "          'Rapa':'Rapamycin', 'CRdiet':'CR diet'}\n",
    "tempDF['Group'] = tempDF['Intervention'].map(tempD0)\n",
    "tempD1 = {'Control':'tab:blue', 'Acarbose':'tab:red',\n",
    "          'Rapamycin':'tab:purple', 'CR diet':'tab:olive'}\n",
    "\n",
    "#Prepare adjusted P-value across groups\n",
    "tempDF0 = statDF.loc[:, statDF.columns.str.contains('-vs-.*_AdjPval$')]\n",
    "tempDF0.columns = tempDF0.columns.str.replace('_AdjPval$', '')\n",
    "\n",
    "#Visualize each representative\n",
    "for rank_i in range(len(plotL)):\n",
    "    print(' - Rank '+str(rank_i+1)+' (sort by', posL, '):')\n",
    "    rxn = plotL[rank_i]\n",
    "    #Check reaction summary\n",
    "    tempDF1 = pd.DataFrame(rxnDF.loc[rxn]).T\n",
    "    display(tempDF1)\n",
    "    \n",
    "    #Select Flux\n",
    "    tempDF1 = tempDF.loc[tempDF['ReactionID']==rxn]\n",
    "    \n",
    "    #Check Flux summary\n",
    "    tempDF2 = tempDF1.groupby(['Group'])['Flux'].agg(['count', 'mean', 'std'])\n",
    "    tempL1 = []\n",
    "    tempL2 = []\n",
    "    for row_n in tempDF2.index.tolist():\n",
    "        count, mean, std = tempDF2.loc[row_n]\n",
    "        tempL1.append(mean - 1.96*std/np.sqrt(count))\n",
    "        tempL2.append(mean + 1.96*std/np.sqrt(count))\n",
    "    tempDF2['0.025'] = tempL1\n",
    "    tempDF2['0.975'] = tempL2\n",
    "    tempDF2 = tempDF2.loc[list(tempD1.keys())]#Sort\n",
    "    display(tempDF2)\n",
    "    \n",
    "    #Prepare significance labels\n",
    "    ##Retrieve statistical significance\n",
    "    tempS = tempDF0.loc[rxn, tempDF0.columns.str.contains('-vs-')]\n",
    "    tempS.name = 'AdjPval'\n",
    "    ##Clean\n",
    "    tempDF2 = tempS.index.to_series().str.split(pat='-vs-', expand=True)\n",
    "    tempDF2 = tempDF2.rename(columns={0:'Contrast', 1:'Baseline'})\n",
    "    tempDF2 = pd.merge(tempDF2, tempS, left_index=True, right_index=True, how='left')\n",
    "    tempDF2['Contrast'] = tempDF2['Contrast'].map(tempD0)\n",
    "    tempDF2['Baseline'] = tempDF2['Baseline'].map(tempD0)\n",
    "    ##Convert p-value to label\n",
    "    tempL = []\n",
    "    for row_i in range(len(tempDF2)):\n",
    "        pval = tempDF2['AdjPval'].iloc[row_i]\n",
    "        if pval<0.001:\n",
    "            tempL.append('***')\n",
    "        elif pval<0.01:\n",
    "            tempL.append('**')\n",
    "        elif pval<0.05:\n",
    "            tempL.append('*')\n",
    "        else:\n",
    "            pval_text = Decimal(str(pval)).quantize(Decimal('0.01'), rounding=ROUND_HALF_UP)\n",
    "            tempL.append(r'$P$ = '+str(pval_text))\n",
    "    tempDF2['SignifLabel'] = tempL\n",
    "    ##Add the y-position level in figure\n",
    "    tempDF2['YposLevel'] = [0, 1, 2]\n",
    "    display(tempDF2)\n",
    "    \n",
    "    #Visualization\n",
    "    sns.set(style='ticks', font='Arial', context='talk')\n",
    "    plt.figure(figsize=(2, 4))\n",
    "    sns.boxplot(data=tempDF1, x='Group', y='Flux', order=list(tempD1.keys()), palette=tempD1,\n",
    "                dodge=False, showfliers=False, showcaps=True, notch=False)\n",
    "    sns.pointplot(data=tempDF1, x='Group', y='Flux', order=list(tempD1.keys()),\n",
    "                  markers='x', color='0.25',#To match with the default sns.boxplot saturation=0.75\n",
    "                  dodge=False, join=False, estimator=np.mean, ci=None)#Add only the mean\n",
    "    p = sns.stripplot(data=tempDF1, x='Group', y='Flux',\n",
    "                      order=list(tempD1.keys()), palette=tempD1, dodge=False, jitter=0.3,\n",
    "                      size=5, edgecolor='black', linewidth=1, **{'marker':'o', 'alpha':0.5})\n",
    "    ##Set axis\n",
    "    sns.despine()\n",
    "    plt.setp(p.get_xticklabels(), rotation=70, horizontalalignment='right',\n",
    "             verticalalignment='center', rotation_mode='anchor')\n",
    "    ###Reset the axis range for the labels\n",
    "    ymin, ymax = p.get_ylim()\n",
    "    tempL = p.get_yticks().tolist()\n",
    "    yinter = int(tempL[1]) - int(tempL[0])\n",
    "    tempL = [int(y) for y in tempL if (y>ymin)&(y<ymax)]#seaborn seems to prepare wider range!?\n",
    "    p.set(ylim=(ymin, ymax+(yinter*1.5)), yticks=tempL)\n",
    "    p.set_yticklabels(['{:,}'.format(int(y)) for y in tempL])\n",
    "    ##Add significance labels\n",
    "    aline_ymin = tempDF1['Flux'].max()\n",
    "    aline_ymargin = yinter/2\n",
    "    for row_i in range(len(tempDF2)):\n",
    "        #Baseline\n",
    "        group_0 = tempDF2['Baseline'].iloc[row_i]\n",
    "        index_0 = list(tempD1.keys()).index(group_0)\n",
    "        xcoord_0 = index_0\n",
    "        #Contrast\n",
    "        group_1 = tempDF2['Contrast'].iloc[row_i]\n",
    "        index_1 = list(tempD1.keys()).index(group_1)\n",
    "        xcoord_1 = index_1\n",
    "        #Standard point of marker\n",
    "        xcoord = (xcoord_0+xcoord_1)/2\n",
    "        ycoord = aline_ymin + aline_ymargin*tempDF2['YposLevel'].iloc[row_i]\n",
    "        label = tempDF2['SignifLabel'].iloc[row_i]\n",
    "        #Add annotation lines\n",
    "        aline_offset = yinter/10\n",
    "        aline_length = yinter/10 + aline_offset\n",
    "        plt.plot([xcoord_0, xcoord_0, xcoord_1, xcoord_1],\n",
    "                 [ycoord+aline_offset, ycoord+aline_length, ycoord+aline_length, ycoord+aline_offset],\n",
    "                 lw=1.5, c='k')\n",
    "        #Add annotation text\n",
    "        if label in ['***', '**', '*']:\n",
    "            text_offset = yinter/4\n",
    "            p.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                       horizontalalignment='center', verticalalignment='center',\n",
    "                       fontsize='medium', color='k')\n",
    "        else:\n",
    "            text_offset = yinter/5\n",
    "            p.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                       horizontalalignment='center', verticalalignment='bottom',\n",
    "                       fontsize='x-small', color='k')\n",
    "    ##Set axis label and title\n",
    "    plt.setp(p, xlabel='', ylabel='Sample flux value [a.u.]')\n",
    "    p.set_title(rxn, {'fontsize':'medium'})\n",
    "    ##Save\n",
    "    fileDir = './ExportFigures/'\n",
    "    ipynbName = '230502_LC-M001-related-TrOmics-GEM-ver3-15_FluxAnalysis_'\n",
    "    fileName = 'flux-boxplot('+rxn+').pdf'\n",
    "    #plt.gcf().savefig(fileDir+ipynbName+fileName, dpi=300, bbox_inches='tight', pad_inches=0.04, transparent=True)\n",
    "    plt.show()\n",
    "    print('')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 3-4-2. Reactions decreased by Aca and CR"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Prepare the target reaction set\n",
    "posL = ['Aca-vs-Ctrl1', 'CRdiet-vs-Ctrl1']\n",
    "negL = ['']\n",
    "regulation = 'Decreased'\n",
    "tempL = statDF.index.tolist()\n",
    "tempDF1 = pvalDF.loc[tempL, pvalDF.columns.str.contains('-vs-')]\n",
    "tempDF2 = diffDF.loc[tempL, diffDF.columns.str.contains('-vs-')]\n",
    "tempS = pd.Series(np.repeat(True, len(tempDF1)), index=tempDF1.index)#Initialize\n",
    "for col_n in tempDF1.columns.tolist():\n",
    "    tempS1 = tempDF1[col_n]\n",
    "    tempS2 = tempDF2[col_n]\n",
    "    if col_n in posL:\n",
    "        if regulation=='Increased':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2>0)\n",
    "        elif regulation=='Decreased':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2<0)\n",
    "    elif col_n in negL:\n",
    "        tempS3 = (tempS1>=0.05)\n",
    "        #Significance for inverse regulation\n",
    "        if regulation=='Increased':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2<0)\n",
    "        elif regulation=='Decreased':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2>0)\n",
    "        tempS1 = tempS3 | tempS1\n",
    "    else:\n",
    "        tempS1 = (tempS1>=0.0)\n",
    "    #Update True\n",
    "    tempS = tempS & tempS1\n",
    "tempL = tempS.loc[tempS.tolist()].index.tolist()\n",
    "print(len(tempL), regulation.lower()+' reactions with significance in', posL, 'but not in', negL)\n",
    "\n",
    "#Select representatives\n",
    "topX = np.min([30, len(tempL)])\n",
    "topX_plot = np.min([5, len(tempL)])\n",
    "tempDF = statDF.loc[:, statDF.columns.str.contains('Pval$')]\n",
    "tempDF = pd.merge(rxnDF[['ReactionName', 'Subsystem']], tempDF,\n",
    "                  left_index=True, right_index=True, how='left')\n",
    "tempL1 = [comparison+'_Pval' for comparison in posL]\n",
    "tempDF = tempDF.loc[tempL].sort_values(by=tempL1, ascending=True)\n",
    "print('Top', topX, 'reactions (sort by', posL, '):')\n",
    "display(tempDF.iloc[:topX])\n",
    "plotL = tempDF.index.tolist()[:topX_plot]\n",
    "\n",
    "#Prepare DF for plot\n",
    "tempDF = fluxDF.reset_index().melt(var_name='ModelID', value_name='Flux', id_vars='ReactionID')\n",
    "tempDF1 = sampleDF.reset_index()[['ModelID', 'Intervention']]\n",
    "tempDF = pd.merge(tempDF, tempDF1, on='ModelID', how='left')\n",
    "\n",
    "#Prepare label and color\n",
    "tempD0 = {'Ctrl1':'Control', 'Aca':'Acarbose',\n",
    "          'Rapa':'Rapamycin', 'CRdiet':'CR diet'}\n",
    "tempDF['Group'] = tempDF['Intervention'].map(tempD0)\n",
    "tempD1 = {'Control':'tab:blue', 'Acarbose':'tab:red',\n",
    "          'Rapamycin':'tab:purple', 'CR diet':'tab:olive'}\n",
    "\n",
    "#Prepare adjusted P-value across groups\n",
    "tempDF0 = statDF.loc[:, statDF.columns.str.contains('-vs-.*_AdjPval$')]\n",
    "tempDF0.columns = tempDF0.columns.str.replace('_AdjPval$', '')\n",
    "\n",
    "#Visualize each representative\n",
    "for rank_i in range(len(plotL)):\n",
    "    print(' - Rank '+str(rank_i+1)+' (sort by', posL, '):')\n",
    "    rxn = plotL[rank_i]\n",
    "    #Check reaction summary\n",
    "    tempDF1 = pd.DataFrame(rxnDF.loc[rxn]).T\n",
    "    display(tempDF1)\n",
    "    \n",
    "    #Select Flux\n",
    "    tempDF1 = tempDF.loc[tempDF['ReactionID']==rxn]\n",
    "    \n",
    "    #Check Flux summary\n",
    "    tempDF2 = tempDF1.groupby(['Group'])['Flux'].agg(['count', 'mean', 'std'])\n",
    "    tempL1 = []\n",
    "    tempL2 = []\n",
    "    for row_n in tempDF2.index.tolist():\n",
    "        count, mean, std = tempDF2.loc[row_n]\n",
    "        tempL1.append(mean - 1.96*std/np.sqrt(count))\n",
    "        tempL2.append(mean + 1.96*std/np.sqrt(count))\n",
    "    tempDF2['0.025'] = tempL1\n",
    "    tempDF2['0.975'] = tempL2\n",
    "    tempDF2 = tempDF2.loc[list(tempD1.keys())]#Sort\n",
    "    display(tempDF2)\n",
    "    \n",
    "    #Prepare significance labels\n",
    "    ##Retrieve statistical significance\n",
    "    tempS = tempDF0.loc[rxn, tempDF0.columns.str.contains('-vs-')]\n",
    "    tempS.name = 'AdjPval'\n",
    "    ##Clean\n",
    "    tempDF2 = tempS.index.to_series().str.split(pat='-vs-', expand=True)\n",
    "    tempDF2 = tempDF2.rename(columns={0:'Contrast', 1:'Baseline'})\n",
    "    tempDF2 = pd.merge(tempDF2, tempS, left_index=True, right_index=True, how='left')\n",
    "    tempDF2['Contrast'] = tempDF2['Contrast'].map(tempD0)\n",
    "    tempDF2['Baseline'] = tempDF2['Baseline'].map(tempD0)\n",
    "    ##Convert p-value to label\n",
    "    tempL = []\n",
    "    for row_i in range(len(tempDF2)):\n",
    "        pval = tempDF2['AdjPval'].iloc[row_i]\n",
    "        if pval<0.001:\n",
    "            tempL.append('***')\n",
    "        elif pval<0.01:\n",
    "            tempL.append('**')\n",
    "        elif pval<0.05:\n",
    "            tempL.append('*')\n",
    "        else:\n",
    "            pval_text = Decimal(str(pval)).quantize(Decimal('0.01'), rounding=ROUND_HALF_UP)\n",
    "            tempL.append(r'$P$ = '+str(pval_text))\n",
    "    tempDF2['SignifLabel'] = tempL\n",
    "    ##Add the y-position level in figure\n",
    "    tempDF2['YposLevel'] = [0, 1, 2]\n",
    "    display(tempDF2)\n",
    "    \n",
    "    #Visualization\n",
    "    sns.set(style='ticks', font='Arial', context='talk')\n",
    "    plt.figure(figsize=(2, 4))\n",
    "    sns.boxplot(data=tempDF1, x='Group', y='Flux', order=list(tempD1.keys()), palette=tempD1,\n",
    "                dodge=False, showfliers=False, showcaps=True, notch=False)\n",
    "    sns.pointplot(data=tempDF1, x='Group', y='Flux', order=list(tempD1.keys()),\n",
    "                  markers='x', color='0.25',#To match with the default sns.boxplot saturation=0.75\n",
    "                  dodge=False, join=False, estimator=np.mean, ci=None)#Add only the mean\n",
    "    p = sns.stripplot(data=tempDF1, x='Group', y='Flux',\n",
    "                      order=list(tempD1.keys()), palette=tempD1, dodge=False, jitter=0.3,\n",
    "                      size=5, edgecolor='black', linewidth=1, **{'marker':'o', 'alpha':0.5})\n",
    "    ##Set axis\n",
    "    sns.despine()\n",
    "    plt.setp(p.get_xticklabels(), rotation=70, horizontalalignment='right',\n",
    "             verticalalignment='center', rotation_mode='anchor')\n",
    "    ###Reset the axis range for the labels\n",
    "    ymin, ymax = p.get_ylim()\n",
    "    tempL = p.get_yticks().tolist()\n",
    "    yinter = int(tempL[1]) - int(tempL[0])\n",
    "    tempL = [int(y) for y in tempL if (y>ymin)&(y<ymax)]#seaborn seems to prepare wider range!?\n",
    "    p.set(ylim=(ymin, ymax+(yinter*1.5)), yticks=tempL)\n",
    "    p.set_yticklabels(['{:,}'.format(int(y)) for y in tempL])\n",
    "    ##Add significance labels\n",
    "    aline_ymin = tempDF1['Flux'].max()\n",
    "    aline_ymargin = yinter/2\n",
    "    for row_i in range(len(tempDF2)):\n",
    "        #Baseline\n",
    "        group_0 = tempDF2['Baseline'].iloc[row_i]\n",
    "        index_0 = list(tempD1.keys()).index(group_0)\n",
    "        xcoord_0 = index_0\n",
    "        #Contrast\n",
    "        group_1 = tempDF2['Contrast'].iloc[row_i]\n",
    "        index_1 = list(tempD1.keys()).index(group_1)\n",
    "        xcoord_1 = index_1\n",
    "        #Standard point of marker\n",
    "        xcoord = (xcoord_0+xcoord_1)/2\n",
    "        ycoord = aline_ymin + aline_ymargin*tempDF2['YposLevel'].iloc[row_i]\n",
    "        label = tempDF2['SignifLabel'].iloc[row_i]\n",
    "        #Add annotation lines\n",
    "        aline_offset = yinter/10\n",
    "        aline_length = yinter/10 + aline_offset\n",
    "        plt.plot([xcoord_0, xcoord_0, xcoord_1, xcoord_1],\n",
    "                 [ycoord+aline_offset, ycoord+aline_length, ycoord+aline_length, ycoord+aline_offset],\n",
    "                 lw=1.5, c='k')\n",
    "        #Add annotation text\n",
    "        if label in ['***', '**', '*']:\n",
    "            text_offset = yinter/4\n",
    "            p.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                       horizontalalignment='center', verticalalignment='center',\n",
    "                       fontsize='medium', color='k')\n",
    "        else:\n",
    "            text_offset = yinter/5\n",
    "            p.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                       horizontalalignment='center', verticalalignment='bottom',\n",
    "                       fontsize='x-small', color='k')\n",
    "    ##Set axis label and title\n",
    "    plt.setp(p, xlabel='', ylabel='Sample flux value [a.u.]')\n",
    "    p.set_title(rxn, {'fontsize':'medium'})\n",
    "    ##Save\n",
    "    fileDir = './ExportFigures/'\n",
    "    ipynbName = '230502_LC-M001-related-TrOmics-GEM-ver3-15_FluxAnalysis_'\n",
    "    fileName = 'flux-boxplot('+rxn+').pdf'\n",
    "    #plt.gcf().savefig(fileDir+ipynbName+fileName, dpi=300, bbox_inches='tight', pad_inches=0.04, transparent=True)\n",
    "    plt.show()\n",
    "    print('')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 3-4-3. Reactions of special interest"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> Reactions included in the Priyanka's diagram, related to glycolysis, TCA cycle, fatty acid metabolism etc., are checked.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Import reactions in the diagram\n",
    "fileDir = './ImportData/'\n",
    "fileName = 'Reactions-in-diagram.csv'\n",
    "tempDF = pd.read_csv(fileDir+fileName)\n",
    "display(tempDF)\n",
    "print('-> Unique ID:', len(tempDF['ReactionID'].unique()))\n",
    "\n",
    "#Check\n",
    "tempDF1 = rxnDF\n",
    "tempDF2 = fluxDF\n",
    "tempS1 = tempDF1['Subsystem'].value_counts()\n",
    "tempS1.name = 'Overall'\n",
    "tempDF3 = tempDF1.loc[tempDF1.index.isin(tempDF['ReactionID'])]\n",
    "print('-> Included in the generic model:', len(tempDF3))\n",
    "display(tempDF.loc[~tempDF['ReactionID'].isin(tempDF1.index)])\n",
    "tempS2 = tempDF3['Subsystem'].value_counts()\n",
    "tempS2.name = 'InDiagram'\n",
    "tempDF3 = tempDF3.loc[tempDF3.index.isin(tempDF2.index)]\n",
    "print('-> Flux-calculated and selected:', len(tempDF3))\n",
    "tempS3 = tempDF3['Subsystem'].value_counts()\n",
    "tempS3.name = 'FluxCalculated-InDiagram'\n",
    "tempDF = pd.concat([tempS1, tempS2, tempS3], axis=1)\n",
    "tempDF = tempDF.dropna(subset=['InDiagram'])\n",
    "display(tempDF)\n",
    "\n",
    "plotL = tempDF3.index.tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Select the changed reactions\n",
    "tempDF = statDF.loc[:, statDF.columns.str.contains('Pval$')]\n",
    "tempDF = pd.merge(rxnDF[['ReactionName', 'Subsystem']], tempDF,\n",
    "                  left_index=True, right_index=True, how='left')\n",
    "tempDF = tempDF.loc[plotL]\n",
    "print(len(tempDF), 'target reactions:')\n",
    "tempL = statDF.loc[:, statDF.columns.str.contains('Pval$')].columns.tolist()\n",
    "tempS1 = set()\n",
    "for col_n in tempL:\n",
    "    tempS = tempDF[col_n]\n",
    "    tempS = tempS.loc[tempS<0.05]\n",
    "    print(' - '+col_n+':', len(tempS))\n",
    "    tempS1 = tempS1 | set(tempS.index.tolist())\n",
    "tempDF = tempDF.loc[tempS1]\n",
    "print(' -> Unique reactions:', len(tempDF))\n",
    "tempL = statDF.loc[:, statDF.columns.str.contains('_Pval$')].columns.tolist()\n",
    "tempDF = tempDF.sort_values(by=tempL, ascending=True)\n",
    "display(tempDF)\n",
    "plotL = tempDF.index.tolist()#Update by the sorted order\n",
    "\n",
    "#Prepare DF for plot\n",
    "tempDF = fluxDF.reset_index().melt(var_name='ModelID', value_name='Flux', id_vars='ReactionID')\n",
    "tempDF1 = sampleDF.reset_index()[['ModelID', 'Intervention']]\n",
    "tempDF = pd.merge(tempDF, tempDF1, on='ModelID', how='left')\n",
    "\n",
    "#Prepare label and color\n",
    "tempD0 = {'Ctrl1':'Control', 'Aca':'Acarbose',\n",
    "          'Rapa':'Rapamycin', 'CRdiet':'CR diet'}\n",
    "tempDF['Group'] = tempDF['Intervention'].map(tempD0)\n",
    "tempD1 = {'Control':'tab:blue', 'Acarbose':'tab:red',\n",
    "          'Rapamycin':'tab:purple', 'CR diet':'tab:olive'}\n",
    "\n",
    "#Prepare adjusted P-value across groups\n",
    "tempDF0 = statDF.loc[:, statDF.columns.str.contains('-vs-.*_AdjPval$')]\n",
    "tempDF0.columns = tempDF0.columns.str.replace('_AdjPval$', '')\n",
    "\n",
    "#Visualize each representative\n",
    "for rank_i in range(len(plotL)):\n",
    "    print(' - Rank '+str(rank_i+1)+' (sort by nominal P-value):')\n",
    "    rxn = plotL[rank_i]\n",
    "    #Check reaction summary\n",
    "    tempDF1 = pd.DataFrame(rxnDF.loc[rxn]).T\n",
    "    display(tempDF1)\n",
    "    \n",
    "    #Select Flux\n",
    "    tempDF1 = tempDF.loc[tempDF['ReactionID']==rxn]\n",
    "    \n",
    "    #Check Flux summary\n",
    "    tempDF2 = tempDF1.groupby(['Group'])['Flux'].agg(['count', 'mean', 'std'])\n",
    "    tempL1 = []\n",
    "    tempL2 = []\n",
    "    for row_n in tempDF2.index.tolist():\n",
    "        count, mean, std = tempDF2.loc[row_n]\n",
    "        tempL1.append(mean - 1.96*std/np.sqrt(count))\n",
    "        tempL2.append(mean + 1.96*std/np.sqrt(count))\n",
    "    tempDF2['0.025'] = tempL1\n",
    "    tempDF2['0.975'] = tempL2\n",
    "    tempDF2 = tempDF2.loc[list(tempD1.keys())]#Sort\n",
    "    display(tempDF2)\n",
    "    \n",
    "    #Prepare significance labels\n",
    "    ##Retrieve statistical significance\n",
    "    tempS = tempDF0.loc[rxn, tempDF0.columns.str.contains('-vs-')]\n",
    "    tempS.name = 'AdjPval'\n",
    "    ##Clean\n",
    "    tempDF2 = tempS.index.to_series().str.split(pat='-vs-', expand=True)\n",
    "    tempDF2 = tempDF2.rename(columns={0:'Contrast', 1:'Baseline'})\n",
    "    tempDF2 = pd.merge(tempDF2, tempS, left_index=True, right_index=True, how='left')\n",
    "    tempDF2['Contrast'] = tempDF2['Contrast'].map(tempD0)\n",
    "    tempDF2['Baseline'] = tempDF2['Baseline'].map(tempD0)\n",
    "    ##Convert p-value to label\n",
    "    tempL = []\n",
    "    for row_i in range(len(tempDF2)):\n",
    "        pval = tempDF2['AdjPval'].iloc[row_i]\n",
    "        if pval<0.001:\n",
    "            tempL.append('***')\n",
    "        elif pval<0.01:\n",
    "            tempL.append('**')\n",
    "        elif pval<0.05:\n",
    "            tempL.append('*')\n",
    "        else:\n",
    "            pval_text = Decimal(str(pval)).quantize(Decimal('0.01'), rounding=ROUND_HALF_UP)\n",
    "            tempL.append(r'$P$ = '+str(pval_text))\n",
    "    tempDF2['SignifLabel'] = tempL\n",
    "    ##Add the y-position level in figure\n",
    "    tempDF2['YposLevel'] = [0, 1, 2]\n",
    "    display(tempDF2)\n",
    "    \n",
    "    #Visualization\n",
    "    sns.set(style='ticks', font='Arial', context='talk')\n",
    "    plt.figure(figsize=(2, 4))\n",
    "    sns.boxplot(data=tempDF1, x='Group', y='Flux', order=list(tempD1.keys()), palette=tempD1,\n",
    "                dodge=False, showfliers=False, showcaps=True, notch=False)\n",
    "    sns.pointplot(data=tempDF1, x='Group', y='Flux', order=list(tempD1.keys()),\n",
    "                  markers='x', color='0.25',#To match with the default sns.boxplot saturation=0.75\n",
    "                  dodge=False, join=False, estimator=np.mean, ci=None)#Add only the mean\n",
    "    p = sns.stripplot(data=tempDF1, x='Group', y='Flux',\n",
    "                      order=list(tempD1.keys()), palette=tempD1, dodge=False, jitter=0.3,\n",
    "                      size=5, edgecolor='black', linewidth=1, **{'marker':'o', 'alpha':0.5})\n",
    "    ##Set axis\n",
    "    sns.despine()\n",
    "    plt.setp(p.get_xticklabels(), rotation=70, horizontalalignment='right',\n",
    "             verticalalignment='center', rotation_mode='anchor')\n",
    "    ###Reset the axis range for the labels\n",
    "    ymin, ymax = p.get_ylim()\n",
    "    tempL = p.get_yticks().tolist()\n",
    "    yinter = int(tempL[1]) - int(tempL[0])\n",
    "    tempL = [int(y) for y in tempL if (y>ymin)&(y<ymax)]#seaborn seems to prepare wider range!?\n",
    "    p.set(ylim=(ymin, ymax+(yinter*1.5)), yticks=tempL)\n",
    "    p.set_yticklabels(['{:,}'.format(int(y)) for y in tempL])\n",
    "    ##Add significance labels\n",
    "    aline_ymin = tempDF1['Flux'].max()\n",
    "    aline_ymargin = yinter/2\n",
    "    for row_i in range(len(tempDF2)):\n",
    "        #Baseline\n",
    "        group_0 = tempDF2['Baseline'].iloc[row_i]\n",
    "        index_0 = list(tempD1.keys()).index(group_0)\n",
    "        xcoord_0 = index_0\n",
    "        #Contrast\n",
    "        group_1 = tempDF2['Contrast'].iloc[row_i]\n",
    "        index_1 = list(tempD1.keys()).index(group_1)\n",
    "        xcoord_1 = index_1\n",
    "        #Standard point of marker\n",
    "        xcoord = (xcoord_0+xcoord_1)/2\n",
    "        ycoord = aline_ymin + aline_ymargin*tempDF2['YposLevel'].iloc[row_i]\n",
    "        label = tempDF2['SignifLabel'].iloc[row_i]\n",
    "        #Add annotation lines\n",
    "        aline_offset = yinter/10\n",
    "        aline_length = yinter/10 + aline_offset\n",
    "        plt.plot([xcoord_0, xcoord_0, xcoord_1, xcoord_1],\n",
    "                 [ycoord+aline_offset, ycoord+aline_length, ycoord+aline_length, ycoord+aline_offset],\n",
    "                 lw=1.5, c='k')\n",
    "        #Add annotation text\n",
    "        if label in ['***', '**', '*']:\n",
    "            text_offset = yinter/4\n",
    "            p.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                       horizontalalignment='center', verticalalignment='center',\n",
    "                       fontsize='medium', color='k')\n",
    "        else:\n",
    "            text_offset = yinter/5\n",
    "            p.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                       horizontalalignment='center', verticalalignment='bottom',\n",
    "                       fontsize='x-small', color='k')\n",
    "    ##Set axis label and title\n",
    "    plt.setp(p, xlabel='', ylabel='Sample flux value [a.u.]')\n",
    "    p.set_title(rxn, {'fontsize':'medium'})\n",
    "    ##Save\n",
    "    fileDir = './ExportFigures/'\n",
    "    ipynbName = '230502_LC-M001-related-TrOmics-GEM-ver3-15_FluxAnalysis_'\n",
    "    fileName = 'flux-boxplot('+rxn+').pdf'\n",
    "    plt.gcf().savefig(fileDir+ipynbName+fileName, dpi=300, bbox_inches='tight', pad_inches=0.04, transparent=True)\n",
    "    plt.show()\n",
    "    print('')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# — End of notebook —"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "arivale-py3 - Python",
   "language": "python",
   "name": "conda-env-arivale-py3-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  },
  "widgets": {
   "application/vnd.jupyter.widget-state+json": {
    "state": {},
    "version_major": 2,
    "version_minor": 0
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
