{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# GEM Reconstruction with LC M001-related Transcriptomics — Enrichment Analysis of Subsystems"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***by Kengo Watanabe***  \n",
    "\n",
    "The maximum flux values, which were calculated using mouse genome-scale metabolic models (GEMs; Khodaee, S. et al. Sci. Rep. 2020) with the Longevity Consortium (LC) M001-related transcriptomics dataset (Tyshkovskiy, A. et al. Cell Metab. 2019), were compared.  \n",
    "–> In this notebook, enrichment analysis is performed for the GEM subsystems in the significantly flux-changed reactions. This notebook is written by R.  \n",
    "\n",
    "Input:  \n",
    "* Cleaned reaction metadata: 220610_LC-M001-related-transcriptomics-GEM-ver2-4_FluxAnalysis_reaction-metadata.xlsx  \n",
    "* Cleaned flux data: 220610_LC-M001-related-transcriptomics-GEM-ver2-4_FluxAnalysis_max-flux-raw-data.tsv  \n",
    "* Flux comparison summary: 220610_LC-M001-related-transcriptomics-GEM-ver2-4_StatisticalTest_flux-comparison.xlsx  \n",
    "\n",
    "Output:  \n",
    "* Figure 4e  \n",
    "* Supplementary Figure 2a, b, g  \n",
    "* Supplementary Data 4  \n",
    "\n",
    "Original notebook (memo for my future tracing):  \n",
    "* dalek:[JupyterLab HOME]/220606_LC-M001-related-transcriptomics-GEM/220616_LC-M001-related-transcriptomics-GEM-ver2-4_Enrichment2.ipynb  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "library(\"tidyverse\")\n",
    "options(repr.plot.width=5, repr.plot.height=5)#Default=7x7\n",
    "\n",
    "#Bioconductor\n",
    "for (package in c(\"clusterProfiler\", \"enrichplot\")) {\n",
    "    #if (!requireNamespace(\"BiocManager\", quietly=TRUE))\n",
    "    #    install.packages(\"BiocManager\")\n",
    "    #BiocManager::install(package)\n",
    "    eval(bquote(library(.(package))))\n",
    "    print(str_c(package, \": \", as.character(packageVersion(package))))\n",
    "}\n",
    "#CRAN\n",
    "for (package in c(\"readxl\", \"openxlsx\")) {\n",
    "    #install.packages(package)\n",
    "    eval(bquote(library(.(package))))\n",
    "    print(str_c(package, \": \", as.character(packageVersion(package))))\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. All the successfully flux-calculated reactions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Import reaction metadata\n",
    "fileDir <- \"./ExportData/\"\n",
    "ipynbName <- \"220610_LC-M001-related-transcriptomics-GEM-ver2-4_FluxAnalysis_\"\n",
    "fileName <- \"reaction-metadata.xlsx\"\n",
    "sheetName <- \"Reaction\"\n",
    "temp <- read_excel(str_c(fileDir,ipynbName,fileName), sheet=sheetName)\n",
    "\n",
    "print(str_c(\"nrow: \",nrow(temp)))\n",
    "head(temp)\n",
    "print(str_c(\"Unique reaction: \", length(unique(temp$ReactionID))))\n",
    "print(str_c(\"Unique subsystem: \", length(unique(temp$Subsystem))))\n",
    "\n",
    "meta_tbl <- temp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Check\n",
    "meta_tbl %>% dplyr::filter(is.na(Subsystem))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> –> Reactions without subsytem annotation should be eliminated in this enrichment analysis. Of note, Priyanka confirmed they were not errors.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Eliminate reactions without subsystem\n",
    "temp <- meta_tbl %>%\n",
    "    tidyr::drop_na(Subsystem)\n",
    "\n",
    "print(str_c(\"nrow: \",nrow(temp)))\n",
    "print(str_c(\"Unique reaction: \", length(unique(temp$ReactionID))))\n",
    "print(str_c(\"Unique subsystem: \", length(unique(temp$Subsystem))))\n",
    "\n",
    "bgd_tbl <- temp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Import flux data\n",
    "fileDir <- \"./ExportData/\"\n",
    "ipynbName <- \"220610_LC-M001-related-transcriptomics-GEM-ver2-4_FluxAnalysis_\"\n",
    "fileName <- \"max-flux-raw-data.tsv\"\n",
    "temp <- read_delim(str_c(fileDir,ipynbName,fileName), delim=\"\\t\")\n",
    "print(str_c(\"nrow: \",nrow(temp)))\n",
    "\n",
    "#Take the successfully calculated reactions\n",
    "temp <- bgd_tbl %>%\n",
    "    dplyr::filter(ReactionID %in% temp$ReactionID)\n",
    "\n",
    "print(str_c(\"nrow: \",nrow(temp)))\n",
    "head(temp)\n",
    "print(str_c(\"Unique reaction: \", length(unique(temp$ReactionID))))\n",
    "print(str_c(\"Unique subsystem: \", length(unique(temp$Subsystem))))\n",
    "\n",
    "bgd_tbl <- temp"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. Significantly changed reactions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Import the summary tables\n",
    "fileDir <- \"./ExportData/\"\n",
    "ipynbName <- \"220610_LC-M001-related-transcriptomics-GEM-ver2-4_StatisticalTest_\"\n",
    "fileName <- \"flux-comparison.xlsx\"\n",
    "sheetName <- \"Main_vs-each-control\"\n",
    "temp <- read_excel(str_c(fileDir,ipynbName,fileName), sheet=sheetName)\n",
    "print(str_c(\"Original nrow: \",nrow(temp)))\n",
    "\n",
    "#Clean\n",
    "temp1 <- str_subset(names(temp), \"_AdjPval\")\n",
    "temp <- temp %>%\n",
    "    dplyr::select(ReactionID, all_of(temp1)) %>%\n",
    "    tidyr::drop_na(Intervention_AdjPval)#NA = reactions having invariable flux values\n",
    "names(temp) <- str_replace(names(temp), \"_AdjPval\", \"\")\n",
    "\n",
    "print(str_c(\"nrow: \",nrow(temp)))\n",
    "head(temp)\n",
    "print(str_c(\"Unique reaction: \", length(unique(temp$ReactionID))))\n",
    "\n",
    "pval_tbl1 <- temp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Import the summary tables\n",
    "fileDir <- \"./ExportData/\"\n",
    "ipynbName <- \"220610_LC-M001-related-transcriptomics-GEM-ver2-4_StatisticalTest_\"\n",
    "fileName <- \"flux-comparison.xlsx\"\n",
    "sheetName <- \"Posthoc2_MR-vs-CR\"\n",
    "temp <- read_excel(str_c(fileDir,ipynbName,fileName), sheet=sheetName)\n",
    "print(str_c(\"Original nrow: \",nrow(temp)))\n",
    "\n",
    "#Clean\n",
    "temp1 <- str_subset(names(temp), \"_AdjPval\")\n",
    "temp <- temp %>%\n",
    "    dplyr::select(ReactionID, all_of(temp1)) %>%\n",
    "    dplyr::filter(ReactionID %in% pval_tbl1$ReactionID)\n",
    "names(temp) <- str_replace(names(temp), \"_AdjPval\", \"\")\n",
    "\n",
    "print(str_c(\"nrow: \",nrow(temp)))\n",
    "head(temp)\n",
    "print(str_c(\"Unique reaction: \", length(unique(temp$ReactionID))))\n",
    "\n",
    "pval_tbl2 <- temp"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3. Enrichment Analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3-1. Aca, MR, GHRKO, SnellDW vs. each control"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> According to the number of significantly changed reactions, the four interventions (Aca, MR, GHRKO, SnellDW) are analyzed.  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 3-1-1. Save result objects"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "comparison_vec <- c(\"Acar-vs-Con1\", \"MetR-vs-Con2\", \"GHRk-vs-GHRw\", \"SneD-vs-SneW\")\n",
    "\n",
    "#Create a workbook object to save as one single .xlsx file\n",
    "workbook <- createWorkbook()\n",
    "\n",
    "#Summarize results per comparison\n",
    "nRxns_vec <- c()\n",
    "nMappedRxns_vec <- c()\n",
    "nRxnSystems_vec <- c()\n",
    "res_list <- list()\n",
    "for (i in 1:length(comparison_vec)) {\n",
    "    #Prepare input for enricher()\n",
    "    comparison <- comparison_vec[i]\n",
    "    rxns <- pval_tbl1 %>%\n",
    "        dplyr::filter(Intervention<0.05) %>%\n",
    "        dplyr::filter(!!as.name(comparison)<0.05) %>%\n",
    "        .$ReactionID\n",
    "    print(str_c(comparison,\": \",length(rxns),\" changed reactions\"))\n",
    "    bgds <- bgd_tbl %>%\n",
    "        dplyr::select(Subsystem, ReactionID)\n",
    "    labels <- bgd_tbl %>%\n",
    "        dplyr::mutate(Label=Subsystem) %>%#Dummy in this case\n",
    "        dplyr::select(Subsystem, Label) %>%\n",
    "        dplyr::distinct()\n",
    "    \n",
    "    #Save info\n",
    "    nRxns_vec <- c(nRxns_vec, length(rxns))\n",
    "    temp <- bgds %>%\n",
    "        dplyr::filter(ReactionID %in% rxns)\n",
    "    nMappedRxns_vec <- c(nMappedRxns_vec, length(unique(temp$ReactionID)))\n",
    "    nRxnSystems_vec <- c(nRxnSystems_vec, length(unique(temp$Subsystem)))\n",
    "    \n",
    "    #Enrichment analysis\n",
    "    temp <- enricher(gene=rxns,\n",
    "                     pvalueCutoff=1.0,#To export all\n",
    "                     pAdjustMethod=\"BH\",\n",
    "                     #universe=backgrounds,#Already managed\n",
    "                     minGSSize=4,\n",
    "                     maxGSSize=10000,\n",
    "                     qvalueCutoff=1.0,#To export all\n",
    "                     TERM2GENE=bgds,\n",
    "                     TERM2NAME=labels)\n",
    "    \n",
    "    #Add the summary table to the workbook object as an independent sheet\n",
    "    if (is.data.frame(temp[])) {\n",
    "        temp1 <- tibble(temp[]) %>%\n",
    "            dplyr::select(-Description) %>%\n",
    "            dplyr::rename(Subsystem=ID, Ratio2ChangedRxns=GeneRatio, Ratio2BGs=BgRatio,\n",
    "                          Pval=pvalue, AdjPval=p.adjust, Qval=qvalue, MappedChangedRxn=geneID, nMappedChangedRxns=Count)\n",
    "    } else {\n",
    "        temp1 <- tibble(`n/a`=NA)\n",
    "    }\n",
    "    addWorksheet(workbook, sheetName=comparison)\n",
    "    writeData(workbook, comparison, temp1)\n",
    "    \n",
    "    #Add result object to list\n",
    "    res_list <- c(res_list, list(temp))\n",
    "}\n",
    "\n",
    "#Save the workbook as one single .xlsx file\n",
    "fileDir <- \"./ExportData/\"\n",
    "ipynbName <- \"220616_LC-M001-related-transcriptomics-GEM-ver2-4_Enrichment2_\"\n",
    "fileName <- \"clusterProfiler-results.xlsx\"\n",
    "saveWorkbook(workbook, file=str_c(fileDir,ipynbName,fileName), overwrite=TRUE)\n",
    "\n",
    "print(str_c(\"nObjects: \", as.character(length(res_list))))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 3-1-2. Aca"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "obj_i <- 1\n",
    "figtitle <- \"Acarbose vs. Control 1\"\n",
    "\n",
    "#Retreive results\n",
    "comparison <- comparison_vec[obj_i]\n",
    "nRxns <- nRxns_vec[obj_i]\n",
    "nMappedRxns <- nMappedRxns_vec[obj_i]\n",
    "nRxnSystems <- nRxnSystems_vec[obj_i]\n",
    "res <- res_list[[obj_i]]\n",
    "\n",
    "#Check\n",
    "print(comparison)\n",
    "print(str_c(\" - # of the changed reactions: \",as.character(nRxns)))\n",
    "print(str_c(\" - # of the changed reactions that were mapped to any subsystem: \",as.character(nMappedRxns)))\n",
    "print(str_c(\" - # of subsystems having any changed reactions as a member: \",as.character(nRxnSystems)))\n",
    "res\n",
    "print(\" <- Note that the above 'X enriched terms found' is not correct. In this case, X indicates the number of all the tested terms.\")\n",
    "if (is.data.frame(res[])) {\n",
    "    tibble(res[]) %>%\n",
    "        dplyr::filter(pvalue<0.05) %>%#Display only nominal P-value < 0.05\n",
    "        dplyr::select(-Description) %>%\n",
    "        dplyr::rename(Subsystem=ID, Ratio2ChangedRxns=GeneRatio, Ratio2BGs=BgRatio,\n",
    "                      Pval=pvalue, AdjPval=p.adjust, Qval=qvalue, MappedChangedRxn=geneID, nMappedChangedRxns=Count)\n",
    "}\n",
    "\n",
    "#Visualization\n",
    "display <- tibble(res[]) %>%\n",
    "    dplyr::filter(pvalue<0.05) %>%\n",
    "    nrow()\n",
    "if (display>0) {\n",
    "    temp <- res %>%\n",
    "        dplyr::filter(pvalue<0.05) %>%#Display only nominal P-value < 0.05\n",
    "        dplyr::mutate(PvalLabel=str_c(\"AdjPval = \",scales::scientific(p.adjust, digits=2)),\n",
    "                      AdjSignif=ifelse(p.adjust<0.05, \"True\", \"False\")) %>%\n",
    "        barplot(., x=\"Count\", color=\"p.adjust\", showCategory=display) +\n",
    "        geom_text(aes(label=PvalLabel, color=AdjSignif), nudge_x=10, hjust=0) +\n",
    "        coord_cartesian(clip=\"off\") +\n",
    "        scale_x_continuous(limits=c(0, 275), breaks=seq(0, 250, by=50), expand=c(0, 0)) +\n",
    "        scale_y_discrete(labels=function(x) {str_wrap(x, width=50)}) +\n",
    "        scale_fill_viridis_c(begin=0, end=1, direction=1, option=\"plasma\",\n",
    "                             limits=c(0, 0.2), breaks=seq(0, 0.2, by=0.05), name=\"AdjPval\") +\n",
    "        scale_color_manual(values=c(\"True\"=\"#990000\", \"False\"=\"gray40\"), ) +\n",
    "        guides(fill=guide_colorbar(reverse=TRUE), color=\"none\") +\n",
    "        labs(x=\"Count of the changed reactions\",\n",
    "             y=\"\", title=str_c(\"Enriched subsystems: \",figtitle)) +\n",
    "        theme_classic(base_size=16, base_family=\"Helvetica\") +\n",
    "        theme(text=element_text(face=\"plain\", color=\"black\", family=\"Helvetica\")) +\n",
    "        theme(axis.text.x=element_text(face=\"plain\", color=\"black\", family=\"Helvetica\"),\n",
    "              axis.text.y=element_text(face=\"plain\", color=\"black\", family=\"Helvetica\"),\n",
    "              axis.title=element_text(face=\"plain\", color=\"black\", family=\"Helvetica\")) +\n",
    "        theme(plot.title=element_text(size=18, hjust=1.0)) +\n",
    "        theme(legend.direction=\"vertical\", legend.box=\"horizontal\",\n",
    "              legend.background=element_blank())\n",
    "    options(repr.plot.width=8.25, repr.plot.height=max(c(1+display*0.25, 2.5)))\n",
    "    plot(temp)\n",
    "    #Save\n",
    "    fileDir <- \"./ExportFigures/\"\n",
    "    ipynbName <- \"220616_LC-M001-related-transcriptomics-GEM-ver2-4_Enrichment2_\"\n",
    "    fileName <- str_c(comparison,\".pdf\")\n",
    "    ggsave(file=str_c(fileDir,ipynbName,fileName), plot=temp,\n",
    "           width=8.25, height=max(c(1+display*0.25, 2.5)), units=\"in\")\n",
    "    #(Font family is not reflected in JupyterLab output, but correctly done in .pdf file.)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 3-1-3. MR"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "obj_i <- 2\n",
    "figtitle <- \"MR diet vs. Control 2\"\n",
    "\n",
    "#Retreive results\n",
    "comparison <- comparison_vec[obj_i]\n",
    "nRxns <- nRxns_vec[obj_i]\n",
    "nMappedRxns <- nMappedRxns_vec[obj_i]\n",
    "nRxnSystems <- nRxnSystems_vec[obj_i]\n",
    "res <- res_list[[obj_i]]\n",
    "\n",
    "#Check\n",
    "print(comparison)\n",
    "print(str_c(\" - # of the changed reactions: \",as.character(nRxns)))\n",
    "print(str_c(\" - # of the changed reactions that were mapped to any subsystem: \",as.character(nMappedRxns)))\n",
    "print(str_c(\" - # of subsystems having any changed reactions as a member: \",as.character(nRxnSystems)))\n",
    "res\n",
    "print(\" <- Note that the above 'X enriched terms found' is not correct. In this case, X indicates the number of all the tested terms.\")\n",
    "if (is.data.frame(res[])) {\n",
    "    tibble(res[]) %>%\n",
    "        dplyr::filter(pvalue<0.05) %>%#Display only nominal P-value < 0.05\n",
    "        dplyr::select(-Description) %>%\n",
    "        dplyr::rename(Subsystem=ID, Ratio2ChangedRxns=GeneRatio, Ratio2BGs=BgRatio,\n",
    "                      Pval=pvalue, AdjPval=p.adjust, Qval=qvalue, MappedChangedRxn=geneID, nMappedChangedRxns=Count)\n",
    "}\n",
    "\n",
    "#Visualization\n",
    "display <- tibble(res[]) %>%\n",
    "    dplyr::filter(pvalue<0.05) %>%\n",
    "    nrow()\n",
    "if (display>0) {\n",
    "    temp <- res %>%\n",
    "        dplyr::filter(pvalue<0.05) %>%#Display only nominal P-value < 0.05\n",
    "        dplyr::mutate(PvalLabel=str_c(\"AdjPval = \",scales::scientific(p.adjust, digits=2)),\n",
    "                      AdjSignif=ifelse(p.adjust<0.05, \"True\", \"False\")) %>%\n",
    "        barplot(., x=\"Count\", color=\"p.adjust\", showCategory=display) +\n",
    "        geom_text(aes(label=PvalLabel, color=AdjSignif), nudge_x=10, hjust=0) +\n",
    "        coord_cartesian(clip=\"off\") +\n",
    "        scale_x_continuous(limits=c(0, 275), breaks=seq(0, 250, by=50), expand=c(0, 0)) +\n",
    "        scale_y_discrete(labels=function(x) {str_wrap(x, width=50)}) +\n",
    "        scale_fill_viridis_c(begin=0, end=1, direction=1, option=\"plasma\",\n",
    "                             limits=c(0, 0.2), breaks=seq(0, 0.2, by=0.05), name=\"AdjPval\") +\n",
    "        scale_color_manual(values=c(\"True\"=\"#990000\", \"False\"=\"gray40\"), ) +\n",
    "        guides(fill=guide_colorbar(reverse=TRUE), color=\"none\") +\n",
    "        labs(x=\"Count of the changed reactions\",\n",
    "             y=\"\", title=str_c(\"Enriched subsystems: \",figtitle)) +\n",
    "        theme_classic(base_size=16, base_family=\"Helvetica\") +\n",
    "        theme(text=element_text(face=\"plain\", color=\"black\", family=\"Helvetica\")) +\n",
    "        theme(axis.text.x=element_text(face=\"plain\", color=\"black\", family=\"Helvetica\"),\n",
    "              axis.text.y=element_text(face=\"plain\", color=\"black\", family=\"Helvetica\"),\n",
    "              axis.title=element_text(face=\"plain\", color=\"black\", family=\"Helvetica\")) +\n",
    "        theme(plot.title=element_text(size=18, hjust=1.0)) +\n",
    "        theme(legend.direction=\"vertical\", legend.box=\"horizontal\",\n",
    "              legend.background=element_blank())\n",
    "    options(repr.plot.width=8.25, repr.plot.height=max(c(1+display*0.25, 2.5)))\n",
    "    plot(temp)\n",
    "    #Save\n",
    "    fileDir <- \"./ExportFigures/\"\n",
    "    ipynbName <- \"220616_LC-M001-related-transcriptomics-GEM-ver2-4_Enrichment2_\"\n",
    "    fileName <- str_c(comparison,\".pdf\")\n",
    "    ggsave(file=str_c(fileDir,ipynbName,fileName), plot=temp,\n",
    "           width=8.25, height=max(c(1+display*0.25, 2.5)), units=\"in\")\n",
    "    #(Font family is not reflected in JupyterLab output, but correctly done in .pdf file.)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 3-1-3. GHRKO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "obj_i <- 3\n",
    "figtitle <- \"GHR knockout vs. wild-type\"\n",
    "\n",
    "#Retreive results\n",
    "comparison <- comparison_vec[obj_i]\n",
    "nRxns <- nRxns_vec[obj_i]\n",
    "nMappedRxns <- nMappedRxns_vec[obj_i]\n",
    "nRxnSystems <- nRxnSystems_vec[obj_i]\n",
    "res <- res_list[[obj_i]]\n",
    "\n",
    "#Check\n",
    "print(comparison)\n",
    "print(str_c(\" - # of the changed reactions: \",as.character(nRxns)))\n",
    "print(str_c(\" - # of the changed reactions that were mapped to any subsystem: \",as.character(nMappedRxns)))\n",
    "print(str_c(\" - # of subsystems having any changed reactions as a member: \",as.character(nRxnSystems)))\n",
    "res\n",
    "print(\" <- Note that the above 'X enriched terms found' is not correct. In this case, X indicates the number of all the tested terms.\")\n",
    "if (is.data.frame(res[])) {\n",
    "    tibble(res[]) %>%\n",
    "        dplyr::filter(pvalue<0.05) %>%#Display only nominal P-value < 0.05\n",
    "        dplyr::select(-Description) %>%\n",
    "        dplyr::rename(Subsystem=ID, Ratio2ChangedRxns=GeneRatio, Ratio2BGs=BgRatio,\n",
    "                      Pval=pvalue, AdjPval=p.adjust, Qval=qvalue, MappedChangedRxn=geneID, nMappedChangedRxns=Count)\n",
    "}\n",
    "\n",
    "#Visualization\n",
    "display <- tibble(res[]) %>%\n",
    "    dplyr::filter(pvalue<0.05) %>%\n",
    "    nrow()\n",
    "if (display>0) {\n",
    "    temp <- res %>%\n",
    "        dplyr::filter(pvalue<0.05) %>%#Display only nominal P-value < 0.05\n",
    "        dplyr::mutate(PvalLabel=str_c(\"AdjPval = \",scales::scientific(p.adjust, digits=2)),\n",
    "                      AdjSignif=ifelse(p.adjust<0.05, \"True\", \"False\")) %>%\n",
    "        barplot(., x=\"Count\", color=\"p.adjust\", showCategory=display) +\n",
    "        geom_text(aes(label=PvalLabel, color=AdjSignif), nudge_x=10, hjust=0) +\n",
    "        coord_cartesian(clip=\"off\") +\n",
    "        scale_x_continuous(limits=c(0, 275), breaks=seq(0, 250, by=50), expand=c(0, 0)) +\n",
    "        scale_y_discrete(labels=function(x) {str_wrap(x, width=50)}) +\n",
    "        scale_fill_viridis_c(begin=0, end=1, direction=1, option=\"plasma\",\n",
    "                             limits=c(0, 0.2), breaks=seq(0, 0.2, by=0.05), name=\"AdjPval\") +\n",
    "        scale_color_manual(values=c(\"True\"=\"#990000\", \"False\"=\"gray40\"), ) +\n",
    "        guides(fill=guide_colorbar(reverse=TRUE), color=\"none\") +\n",
    "        labs(x=\"Count of the changed reactions\",\n",
    "             y=\"\", title=str_c(\"Enriched subsystems: \",figtitle)) +\n",
    "        theme_classic(base_size=16, base_family=\"Helvetica\") +\n",
    "        theme(text=element_text(face=\"plain\", color=\"black\", family=\"Helvetica\")) +\n",
    "        theme(axis.text.x=element_text(face=\"plain\", color=\"black\", family=\"Helvetica\"),\n",
    "              axis.text.y=element_text(face=\"plain\", color=\"black\", family=\"Helvetica\"),\n",
    "              axis.title=element_text(face=\"plain\", color=\"black\", family=\"Helvetica\")) +\n",
    "        theme(plot.title=element_text(size=18, hjust=1.0)) +\n",
    "        theme(legend.direction=\"vertical\", legend.box=\"horizontal\",\n",
    "              legend.background=element_blank())\n",
    "    options(repr.plot.width=8.25, repr.plot.height=max(c(1+display*0.25, 2.5)))\n",
    "    plot(temp)\n",
    "    #Save\n",
    "    fileDir <- \"./ExportFigures/\"\n",
    "    ipynbName <- \"220616_LC-M001-related-transcriptomics-GEM-ver2-4_Enrichment2_\"\n",
    "    fileName <- str_c(comparison,\".pdf\")\n",
    "    ggsave(file=str_c(fileDir,ipynbName,fileName), plot=temp,\n",
    "           width=8.25, height=max(c(1+display*0.25, 2.5)), units=\"in\")\n",
    "    #(Font family is not reflected in JupyterLab output, but correctly done in .pdf file.)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 3-1-4. SnellDW"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "obj_i <- 4\n",
    "figtitle <- \"Snell dwarf vs. wild-type\"\n",
    "\n",
    "#Retreive results\n",
    "comparison <- comparison_vec[obj_i]\n",
    "nRxns <- nRxns_vec[obj_i]\n",
    "nMappedRxns <- nMappedRxns_vec[obj_i]\n",
    "nRxnSystems <- nRxnSystems_vec[obj_i]\n",
    "res <- res_list[[obj_i]]\n",
    "\n",
    "#Check\n",
    "print(comparison)\n",
    "print(str_c(\" - # of the changed reactions: \",as.character(nRxns)))\n",
    "print(str_c(\" - # of the changed reactions that were mapped to any subsystem: \",as.character(nMappedRxns)))\n",
    "print(str_c(\" - # of subsystems having any changed reactions as a member: \",as.character(nRxnSystems)))\n",
    "res\n",
    "print(\" <- Note that the above 'X enriched terms found' is not correct. In this case, X indicates the number of all the tested terms.\")\n",
    "if (is.data.frame(res[])) {\n",
    "    tibble(res[]) %>%\n",
    "        dplyr::filter(pvalue<0.05) %>%#Display only nominal P-value < 0.05\n",
    "        dplyr::select(-Description) %>%\n",
    "        dplyr::rename(Subsystem=ID, Ratio2ChangedRxns=GeneRatio, Ratio2BGs=BgRatio,\n",
    "                      Pval=pvalue, AdjPval=p.adjust, Qval=qvalue, MappedChangedRxn=geneID, nMappedChangedRxns=Count)\n",
    "}\n",
    "\n",
    "#Visualization\n",
    "display <- tibble(res[]) %>%\n",
    "    dplyr::filter(pvalue<0.05) %>%\n",
    "    nrow()\n",
    "if (display>0) {\n",
    "    temp <- res %>%\n",
    "        dplyr::filter(pvalue<0.05) %>%#Display only nominal P-value < 0.05\n",
    "        dplyr::mutate(PvalLabel=str_c(\"AdjPval = \",scales::scientific(p.adjust, digits=2)),\n",
    "                      AdjSignif=ifelse(p.adjust<0.05, \"True\", \"False\")) %>%\n",
    "        barplot(., x=\"Count\", color=\"p.adjust\", showCategory=display) +\n",
    "        geom_text(aes(label=PvalLabel, color=AdjSignif), nudge_x=10, hjust=0) +\n",
    "        coord_cartesian(clip=\"off\") +\n",
    "        scale_x_continuous(limits=c(0, 275), breaks=seq(0, 250, by=50), expand=c(0, 0)) +\n",
    "        scale_y_discrete(labels=function(x) {str_wrap(x, width=50)}) +\n",
    "        scale_fill_viridis_c(begin=0, end=1, direction=1, option=\"plasma\",\n",
    "                             limits=c(0, 0.2), breaks=seq(0, 0.2, by=0.05), name=\"AdjPval\") +\n",
    "        scale_color_manual(values=c(\"True\"=\"#990000\", \"False\"=\"gray40\"), ) +\n",
    "        guides(fill=guide_colorbar(reverse=TRUE), color=\"none\") +\n",
    "        labs(x=\"Count of the changed reactions\",\n",
    "             y=\"\", title=str_c(\"Enriched subsystems: \",figtitle)) +\n",
    "        theme_classic(base_size=16, base_family=\"Helvetica\") +\n",
    "        theme(text=element_text(face=\"plain\", color=\"black\", family=\"Helvetica\")) +\n",
    "        theme(axis.text.x=element_text(face=\"plain\", color=\"black\", family=\"Helvetica\"),\n",
    "              axis.text.y=element_text(face=\"plain\", color=\"black\", family=\"Helvetica\"),\n",
    "              axis.title=element_text(face=\"plain\", color=\"black\", family=\"Helvetica\")) +\n",
    "        theme(plot.title=element_text(size=18, hjust=1.0)) +\n",
    "        theme(legend.direction=\"vertical\", legend.box=\"horizontal\",\n",
    "              legend.background=element_blank())\n",
    "    options(repr.plot.width=8.25, repr.plot.height=max(c(1+display*0.25, 2.5)))\n",
    "    plot(temp)\n",
    "    #Save\n",
    "    fileDir <- \"./ExportFigures/\"\n",
    "    ipynbName <- \"220616_LC-M001-related-transcriptomics-GEM-ver2-4_Enrichment2_\"\n",
    "    fileName <- str_c(comparison,\".pdf\")\n",
    "    ggsave(file=str_c(fileDir,ipynbName,fileName), plot=temp,\n",
    "           width=8.25, height=max(c(1+display*0.25, 2.5)), units=\"in\")\n",
    "    #(Font family is not reflected in JupyterLab output, but correctly done in .pdf file.)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3-2. MR vs. CR"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 3-2-1. Save result objects"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> Add the result to the above file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "comparison_vec <- c(\"MetR-vs-CalR\")\n",
    "\n",
    "#Load the existing .xlsx file as a new workbook object\n",
    "fileDir <- \"./ExportData/\"\n",
    "ipynbName <- \"220616_LC-M001-related-transcriptomics-GEM-ver2-4_Enrichment2_\"\n",
    "fileName <- \"clusterProfiler-results.xlsx\"\n",
    "workbook <- loadWorkbook(str_c(fileDir,ipynbName,fileName))\n",
    "\n",
    "#Summarize results per comparison\n",
    "nRxns_vec <- c()\n",
    "nMappedRxns_vec <- c()\n",
    "nRxnSystems_vec <- c()\n",
    "res_list <- list()\n",
    "for (i in 1:length(comparison_vec)) {\n",
    "    #Prepare input for enricher()\n",
    "    comparison <- comparison_vec[i]\n",
    "    temp <- pval_tbl1 %>%\n",
    "        dplyr::filter(Intervention<0.05)\n",
    "    rxns <- pval_tbl2 %>%\n",
    "        dplyr::filter(ReactionID %in% temp$ReactionID) %>%\n",
    "        dplyr::filter(!!as.name(comparison)<0.05) %>%\n",
    "        .$ReactionID\n",
    "    print(str_c(comparison,\": \",length(rxns),\" changed reactions\"))\n",
    "    bgds <- bgd_tbl %>%\n",
    "        dplyr::select(Subsystem, ReactionID)\n",
    "    labels <- bgd_tbl %>%\n",
    "        dplyr::mutate(Label=Subsystem) %>%#Dummy in this case\n",
    "        dplyr::select(Subsystem, Label) %>%\n",
    "        dplyr::distinct()\n",
    "    \n",
    "    #Save info\n",
    "    nRxns_vec <- c(nRxns_vec, length(rxns))\n",
    "    temp <- bgds %>%\n",
    "        dplyr::filter(ReactionID %in% rxns)\n",
    "    nMappedRxns_vec <- c(nMappedRxns_vec, length(unique(temp$ReactionID)))\n",
    "    nRxnSystems_vec <- c(nRxnSystems_vec, length(unique(temp$Subsystem)))\n",
    "    \n",
    "    #Enrichment analysis\n",
    "    temp <- enricher(gene=rxns,\n",
    "                     pvalueCutoff=1.0,#To export all\n",
    "                     pAdjustMethod=\"BH\",\n",
    "                     #universe=backgrounds,#Already managed\n",
    "                     minGSSize=4,\n",
    "                     maxGSSize=10000,\n",
    "                     qvalueCutoff=1.0,#To export all\n",
    "                     TERM2GENE=bgds,\n",
    "                     TERM2NAME=labels)\n",
    "    \n",
    "    #Add the summary table to the workbook object as an independent sheet\n",
    "    if (is.data.frame(temp[])) {\n",
    "        temp1 <- tibble(temp[]) %>%\n",
    "            dplyr::select(-Description) %>%\n",
    "            dplyr::rename(Subsystem=ID, Ratio2ChangedRxns=GeneRatio, Ratio2BGs=BgRatio,\n",
    "                          Pval=pvalue, AdjPval=p.adjust, Qval=qvalue, MappedChangedRxn=geneID, nMappedChangedRxns=Count)\n",
    "    } else {\n",
    "        temp1 <- tibble(`n/a`=NA)\n",
    "    }\n",
    "    addWorksheet(workbook, sheetName=comparison)\n",
    "    writeData(workbook, comparison, temp1)\n",
    "    \n",
    "    #Add result object to list\n",
    "    res_list <- c(res_list, list(temp))\n",
    "}\n",
    "\n",
    "#Update the existing .xlsx file\n",
    "saveWorkbook(workbook, file=str_c(fileDir,ipynbName,fileName), overwrite=TRUE)\n",
    "\n",
    "print(str_c(\"nObjects: \", as.character(length(res_list))))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 3-2-2. MR vs. CR"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "obj_i <- 1\n",
    "figtitle <- \"MR diet vs. CR diet\"\n",
    "\n",
    "#Retreive results\n",
    "comparison <- comparison_vec[obj_i]\n",
    "nRxns <- nRxns_vec[obj_i]\n",
    "nMappedRxns <- nMappedRxns_vec[obj_i]\n",
    "nRxnSystems <- nRxnSystems_vec[obj_i]\n",
    "res <- res_list[[obj_i]]\n",
    "\n",
    "#Check\n",
    "print(comparison)\n",
    "print(str_c(\" - # of the changed reactions: \",as.character(nRxns)))\n",
    "print(str_c(\" - # of the changed reactions that were mapped to any subsystem: \",as.character(nMappedRxns)))\n",
    "print(str_c(\" - # of subsystems having any changed reactions as a member: \",as.character(nRxnSystems)))\n",
    "res\n",
    "print(\" <- Note that the above 'X enriched terms found' is not correct. In this case, X indicates the number of all the tested terms.\")\n",
    "if (is.data.frame(res[])) {\n",
    "    tibble(res[]) %>%\n",
    "        dplyr::filter(pvalue<0.05) %>%#Display only nominal P-value < 0.05\n",
    "        dplyr::select(-Description) %>%\n",
    "        dplyr::rename(Subsystem=ID, Ratio2ChangedRxns=GeneRatio, Ratio2BGs=BgRatio,\n",
    "                      Pval=pvalue, AdjPval=p.adjust, Qval=qvalue, MappedChangedRxn=geneID, nMappedChangedRxns=Count)\n",
    "}\n",
    "\n",
    "#Visualization\n",
    "display <- tibble(res[]) %>%\n",
    "    dplyr::filter(pvalue<0.05) %>%\n",
    "    nrow()\n",
    "if (display>0) {\n",
    "    temp <- res %>%\n",
    "        dplyr::filter(pvalue<0.05) %>%#Display only nominal P-value < 0.05\n",
    "        dplyr::mutate(PvalLabel=str_c(\"AdjPval = \",scales::scientific(p.adjust, digits=2)),\n",
    "                      AdjSignif=ifelse(p.adjust<0.05, \"True\", \"False\")) %>%\n",
    "        barplot(., x=\"Count\", color=\"p.adjust\", showCategory=display) +\n",
    "        geom_text(aes(label=PvalLabel, color=AdjSignif), nudge_x=10, hjust=0) +\n",
    "        coord_cartesian(clip=\"off\") +\n",
    "        scale_x_continuous(limits=c(0, 275), breaks=seq(0, 250, by=50), expand=c(0, 0)) +\n",
    "        scale_y_discrete(labels=function(x) {str_wrap(x, width=50)}) +\n",
    "        scale_fill_viridis_c(begin=0, end=1, direction=1, option=\"plasma\",\n",
    "                             limits=c(0, 0.2), breaks=seq(0, 0.2, by=0.05), name=\"AdjPval\") +\n",
    "        scale_color_manual(values=c(\"True\"=\"#990000\", \"False\"=\"gray40\"), ) +\n",
    "        guides(fill=guide_colorbar(reverse=TRUE), color=\"none\") +\n",
    "        labs(x=\"Count of the changed reactions\",\n",
    "             y=\"\", title=str_c(\"Enriched subsystems: \",figtitle)) +\n",
    "        theme_classic(base_size=16, base_family=\"Helvetica\") +\n",
    "        theme(text=element_text(face=\"plain\", color=\"black\", family=\"Helvetica\")) +\n",
    "        theme(axis.text.x=element_text(face=\"plain\", color=\"black\", family=\"Helvetica\"),\n",
    "              axis.text.y=element_text(face=\"plain\", color=\"black\", family=\"Helvetica\"),\n",
    "              axis.title=element_text(face=\"plain\", color=\"black\", family=\"Helvetica\")) +\n",
    "        theme(plot.title=element_text(size=18, hjust=1.0)) +\n",
    "        theme(legend.direction=\"vertical\", legend.box=\"horizontal\",\n",
    "              legend.background=element_blank())\n",
    "    options(repr.plot.width=8.25, repr.plot.height=max(c(1+display*0.25, 2.5)))\n",
    "    plot(temp)\n",
    "    #Save\n",
    "    fileDir <- \"./ExportFigures/\"\n",
    "    ipynbName <- \"220616_LC-M001-related-transcriptomics-GEM-ver2-4_Enrichment2_\"\n",
    "    fileName <- str_c(comparison,\".pdf\")\n",
    "    ggsave(file=str_c(fileDir,ipynbName,fileName), plot=temp,\n",
    "           width=8.25, height=max(c(1+display*0.25, 2.5)), units=\"in\")\n",
    "    #(Font family is not reflected in JupyterLab output, but correctly done in .pdf file.)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# — Session information —"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sessionInfo()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "arivale-r - R",
   "language": "R",
   "name": "conda-env-arivale-r-r"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.1.1"
  },
  "widgets": {
   "application/vnd.jupyter.widget-state+json": {
    "state": {},
    "version_major": 2,
    "version_minor": 0
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
