{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# GEM Reconstruction with LC M001-related Transcriptomics — Flux Comparison"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***by Kengo Watanabe***  \n",
    "\n",
    "Priyanka Baloni reconstructed mouse genome-scale metabolic models (GEMs; Khodaee, S. et al. Sci. Rep. 2020) with the Longevity Consortium (LC) M001-related transcriptomics dataset (Tyshkovskiy, A. et al. Cell Metab. 2019), and calculated flux values using flux variability analysis (FVA).  \n",
    "–> In this notebook, the calculated maximum flux values are compared between the context-specific GEMs (i.e., sample groups).  \n",
    "> To maintain the consistency with the DIRAC analyses, statistical tests are performed in a different notebook with R kernel.  \n",
    "\n",
    "Input:  \n",
    "* Flux data: merged_file_Maxflux_reactions_Intervention.csv  \n",
    "* Sample–model metadata: RNAseq_samples_annotation.csv  \n",
    "* Reaction metadata (iMM1865): iMM1685_data.xlsx  \n",
    "* Gene mapping metadata: iMM1685_data-Max.xlsx  \n",
    "* Statistical test summary: 220610_LC-M001-related-transcriptomics-GEM-ver2-4_StatisticalTest_flux-comparison.xlsx (Supplementary Data 3)  \n",
    "\n",
    "Output:  \n",
    "* Cleaned flux data, which is used in stasitical analysis (R sub-notebook)  \n",
    "* Cleaned sample–model metadata, which is used in R sub-notebook  \n",
    "* Cleaned reaction metadata, which is incorporated into Supplementary Data 3 in R sub-notebook  \n",
    "* Figure 4a–d  \n",
    "* Supplementary Figure 2c–f  \n",
    "\n",
    "Original notebook (memo for my future tracing):  \n",
    "* dalek:[JupyterLab HOME]/220606_LC-M001-related-transcriptomics-GEM/220610_LC-M001-related-transcriptomics-GEM-ver2-4_FluxAnalysis.ipynb  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scipy.stats as stats\n",
    "import matplotlib.pyplot as plt\n",
    "%matplotlib inline\n",
    "import seaborn as sns\n",
    "#For Arial font\n",
    "#!conda install -c conda-forge -y mscorefonts\n",
    "##-> The below was also needed in matplotlib 3.4.2\n",
    "#import shutil\n",
    "#import matplotlib\n",
    "#shutil.rmtree(matplotlib.get_cachedir())\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')\n",
    "from IPython.display import display\n",
    "import time\n",
    "\n",
    "from sklearn.preprocessing import RobustScaler\n",
    "from sklearn.decomposition import PCA\n",
    "from decimal import Decimal, ROUND_HALF_UP\n",
    "from sklearn.preprocessing import MaxAbsScaler\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "import re\n",
    "import matplotlib.patches as mpatches\n",
    "#!pip install venn\n",
    "from venn import venn\n",
    "\n",
    "!conda list"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. Clean the original tables"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1-1. Maximum flux value (raw) data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Import flux data\n",
    "fileDir = './ImportData/'\n",
    "fileName = 'merged_file_Maxflux_reactions_Intervention.csv'\n",
    "tempDF = pd.read_csv(fileDir+fileName)\n",
    "tempDF = tempDF.rename(columns={'V1':'ReactionID'})\n",
    "tempDF = tempDF.set_index('ReactionID')\n",
    "\n",
    "#Clean the model label\n",
    "tempDF1 = tempDF.columns.to_series().str.split(pat='_', expand=True)\n",
    "tempDF1 = tempDF1.rename(columns={0:'Prefix', 1:'Model_i'})\n",
    "tempDF.columns = 'Model_'+tempDF1['Model_i'].str.zfill(2)\n",
    "tempDF.columns.set_names('', inplace=True)\n",
    "\n",
    "#Sort\n",
    "tempL = ['Model_'+str(i+1).zfill(2) for i in range(len(tempDF.columns))]\n",
    "tempDF = tempDF[tempL]\n",
    "\n",
    "display(tempDF)\n",
    "\n",
    "fluxDF = tempDF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Save\n",
    "fileDir = './ExportData/'\n",
    "ipynbName = '220610_LC-M001-related-transcriptomics-GEM-ver2-4_FluxAnalysis_'\n",
    "fileName = 'max-flux-raw-data.tsv'\n",
    "fluxDF.to_csv(fileDir+ipynbName+fileName, index=True, sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1-2. Sample–model metadata"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> Prepare control-based category, which enables to easily make comparison labels.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Import sample-model metadata\n",
    "fileDir = './ImportData/'\n",
    "fileName = 'RNAseq_samples_annotation.csv'\n",
    "tempDF = pd.read_csv(fileDir+fileName)\n",
    "display(tempDF.describe(include='all'))\n",
    "\n",
    "#Clean the model label\n",
    "tempDF1 = tempDF['Model'].str.split(pat='_', expand=True)\n",
    "tempDF1 = tempDF1.rename(columns={0:'Prefix', 1:'Model_i'})\n",
    "tempDF['ModelID'] = 'Model_'+tempDF1['Model_i'].str.zfill(2)\n",
    "tempDF = tempDF.set_index('ModelID')\n",
    "\n",
    "#Clean the condition labels\n",
    "tempDF = tempDF.rename(columns={'ID':'SampleID'})\n",
    "tempDF = tempDF[['SampleID', 'Strain', 'Age', 'Sex', 'Intervention']]\n",
    "tempD = {'UMHet3':'U', '(C57BL6/JxBALB/cByJ)/F2':'C', '(PW/JxC3H/HeJ)/F2':'P'}\n",
    "tempDF['Strain'] = tempDF['Strain'].map(tempD)\n",
    "tempDF['Age'] = tempDF['Age'].astype(str).str.zfill(2)+'M'\n",
    "tempD = {'No intervention':'Con1', 'Acarbose':'Acar',\n",
    "         '17-alpha-estradiol':'17aE', 'Protandim':'Prot',\n",
    "         'Rapamycin':'Rapa', 'CR':'CalR',\n",
    "         'MR_control':'Con2', 'MR':'MetR',\n",
    "         'GHRKO_control':'GHRw', 'GHRKO':'GHRk',\n",
    "         'Snell_control':'SneW', 'Snell':'SneD'}\n",
    "tempDF['Intervention'] = tempDF['Intervention'].map(tempD)\n",
    "tempDF['Group'] = tempDF['Strain']+'-'+tempDF['Age']+'-'+tempDF['Sex']+'-'+tempDF['Intervention']\n",
    "\n",
    "#Corresponding control\n",
    "tempD = {'Con1':'Con1', 'Acar':'Con1', '17aE':'Con1', 'Prot':'Con1', 'Rapa':'Con1', 'CalR':'Con1',\n",
    "         'Con2':'Con2', 'MetR':'Con2', 'GHRw':'GHRw', 'GHRk':'GHRw',\n",
    "         'SneW':'SneW', 'SneD':'SneW'}\n",
    "tempDF['Control'] = tempDF['Strain']+'-'+tempDF['Age']+'-'+tempDF['Sex']+'-'+tempDF['Intervention'].map(tempD)\n",
    "\n",
    "#Categorize conditions\n",
    "tempDF['Category'] = tempDF['Control'].str.split(pat='-', expand=True).iloc[:, 3]\n",
    "\n",
    "display(tempDF)\n",
    "tempL = tempDF.sort_values(by='Category', ascending=True)['Category'].unique()\n",
    "print(' ->', len(tempL), 'categories\\n')\n",
    "for category in tempL:\n",
    "    tempDF1 = tempDF.loc[tempDF['Category']==category]\n",
    "    print(category+' category total:', len(tempDF1))\n",
    "    for condition in ['Intervention']:\n",
    "        display(tempDF1[condition].value_counts())\n",
    "    print('')\n",
    "\n",
    "sampleDF = tempDF\n",
    "categoryL = tempL"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Save\n",
    "fileDir = './ExportData/'\n",
    "ipynbName = '220610_LC-M001-related-transcriptomics-GEM-ver2-4_FluxAnalysis_'\n",
    "fileName = 'sample-metadata.tsv'\n",
    "sampleDF.to_csv(fileDir+ipynbName+fileName, index=True, sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1-3. Reaction metadata"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 1-3-1. iMM1865"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Import reaction metadata of the used model\n",
    "fileDir = './ImportData/'\n",
    "fileName = 'iMM1685_data.xlsx'\n",
    "sheetName = 'Sheet1'\n",
    "tempDF = pd.read_excel(fileDir+fileName, sheet_name=sheetName, engine='openpyxl')\n",
    "tempDF = tempDF.rename(columns={'Reaction':'ReactionID', 'Unnamed: 3':'Reaction'})\n",
    "tempDF = tempDF.set_index('ReactionID')\n",
    "\n",
    "display(tempDF)\n",
    "print(' -> Unique rxn ID:', len(tempDF.index.unique()))\n",
    "\n",
    "rxnDF = tempDF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Check\n",
    "tempS1 = rxnDF['Subsystem'].value_counts()\n",
    "tempS1.name = 'Overall'\n",
    "tempS2 = rxnDF['Subsystem'].loc[fluxDF.index].value_counts()\n",
    "tempS2.name = 'FluxCalculated'\n",
    "tempDF = pd.merge(tempS1, tempS2, left_index=True, right_index=True, how='left')\n",
    "display(tempDF.describe())\n",
    "display(tempDF.loc[tempS1>100])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 1-3-2. Retrieve reaction name with BiGG Models API"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Download the latest file\n",
    "!wget -O ./ImportData/bigg_models_reactions.txt http://bigg.ucsd.edu/static/namespace/bigg_models_reactions.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Import reaction metadata\n",
    "fileDir = './ImportData/'\n",
    "fileName = 'bigg_models_reactions.txt'\n",
    "tempDF = pd.read_csv(fileDir+fileName, sep='\\t')\n",
    "tempDF = tempDF.rename(columns={'bigg_id':'ReactionID', 'name':'ReactionName'})\n",
    "tempDF = tempDF.set_index('ReactionID')\n",
    "print('Original nrow:', len(tempDF))\n",
    "print(' -> Unique rxn ID:', len(tempDF.index.unique()))\n",
    "\n",
    "#Add the reaction name\n",
    "tempDF = pd.merge(rxnDF, tempDF['ReactionName'], left_index=True, right_index=True, how='left')\n",
    "tempDF = tempDF[['Reaction', 'ReactionName', 'GPR', 'Subsystem']]\n",
    "\n",
    "display(tempDF)\n",
    "tempDF1 = tempDF.loc[tempDF['ReactionName'].isnull()]\n",
    "print(' -> No rxn name:', len(tempDF1))\n",
    "display(tempDF1)\n",
    "\n",
    "rxnDF = tempDF"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 1-3-3. Retrieve molecule name with BiGG Models API"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Take all molecules in the reactions\n",
    "tempS = rxnDF['Reaction'].str.replace('^\\'', '')\n",
    "tempS = tempS.str.replace('\\'$', '')\n",
    "tempDF = tempS.str.split(pat=' ', expand=True)\n",
    "print('Expanded DF:', tempDF.shape)\n",
    "t_start = time.time()\n",
    "tempS = set()\n",
    "for col_n in tempDF.columns.tolist():\n",
    "    tempS1 = tempDF[col_n].dropna()#None is considered an NA value in pandas\n",
    "    tempS1 = tempS1.loc[tempS1!='']#Delete empty\n",
    "    tempS1 = tempS1.loc[tempS1.str.contains('\\\\[.*\\\\]')]\n",
    "    #Character in square brackets indicates intracellular localization\n",
    "    tempL = tempS1.str.replace('\\\\[.*\\\\]', '').tolist()\n",
    "    tempS = tempS | set(tempL)\n",
    "t_elapsed = time.time() - t_start\n",
    "print(' -> Elapsed time:', round(t_elapsed//60), 'min', round(t_elapsed%60, 1), 'sec')\n",
    "print(' -> Unique molecule:', len(tempS))\n",
    "print(list(tempS)[:25])\n",
    "\n",
    "molS = tempS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Download the latest file\n",
    "!wget -O ./ImportData/bigg_models_metabolites.txt http://bigg.ucsd.edu/static/namespace/bigg_models_metabolites.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Import metabolite metadata\n",
    "fileDir = './ImportData/'\n",
    "fileName = 'bigg_models_metabolites.txt'\n",
    "tempDF = pd.read_csv(fileDir+fileName, sep='\\t')\n",
    "tempDF = tempDF.rename(columns={'universal_bigg_id':'MoleculeID', 'name':'MoleculeName'})\n",
    "tempDF = tempDF.set_index('MoleculeID')\n",
    "print('Original nrow:', len(tempDF))\n",
    "print(' -> Unique molecule ID:', len(tempDF.index.unique()))\n",
    "\n",
    "#Retrieve the molecules within the used model\n",
    "print('Molecules within the used model', len(molS))\n",
    "tempS = tempDF['MoleculeName']\n",
    "tempS = tempS.drop_duplicates()\n",
    "tempS = tempS.loc[tempS.index.isin(molS)]\n",
    "\n",
    "tempS = tempS.sort_index(ascending=True)\n",
    "display(tempS)\n",
    "\n",
    "molS = tempS"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 1-3-4. Gene mapping table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Import reaction metadata of the used model\n",
    "fileDir = './ImportData/'\n",
    "fileName = 'iMM1685_data-Max.xlsx'\n",
    "sheetName = 'Sheet1'\n",
    "tempDF = pd.read_excel(fileDir+fileName, sheet_name=sheetName, engine='openpyxl', header=2)\n",
    "tempDF = tempDF.rename(columns={'Entrez ID':'EntrezID', 'Gene symbol':'GeneSymbol',\n",
    "                                'Ensembl gene ID':'EnsemblID'})\n",
    "tempDF = tempDF[['EntrezID', 'EnsemblID', 'GeneSymbol']]\n",
    "tempDF = tempDF.dropna()\n",
    "tempDF['EntrezID'] = tempDF['EntrezID'].astype('int64')\n",
    "tempDF = tempDF.set_index('EntrezID')\n",
    "tempDF = tempDF.sort_index(ascending=True)\n",
    "\n",
    "display(tempDF)\n",
    "print(' -> Unique Entrez ID:', len(tempDF.index.unique()))\n",
    "print(' -> Unique Ensembl ID:', len(tempDF['EnsemblID'].unique()))\n",
    "print(' -> Unique gene name:', len(tempDF['GeneSymbol'].unique()))\n",
    "\n",
    "geneDF = tempDF"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 1-3-5. Save as a single .xlsx file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Prepare a new .xlsx file\n",
    "fileDir = './ExportData/'\n",
    "ipynbName = '220610_LC-M001-related-transcriptomics-GEM-ver2-4_FluxAnalysis_'\n",
    "fileName = 'reaction-metadata.xlsx'\n",
    "sheetName = 'Reaction'\n",
    "rxnDF.to_excel(fileDir+ipynbName+fileName, sheet_name=sheetName, header=True, index=True)\n",
    "\n",
    "#Appended the others to the above .xlsx file\n",
    "with pd.ExcelWriter(fileDir+ipynbName+fileName, mode='a', engine='openpyxl') as writer:\n",
    "    sheetName = 'Molecule'\n",
    "    molS.to_excel(writer, sheet_name=sheetName, header=True, index=True)\n",
    "    sheetName = 'Gene'\n",
    "    geneDF.to_excel(writer, sheet_name=sheetName, header=True, index=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. Check data structure of the flux values"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> Note: this lengthy section was eliminated from the original notebook, because the outputs were not used in the manuscript.  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3. Compare flux values"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> 1. Test the rank of maximum flux values across interventions for each reaction using Kruskal–Wallis H-test.  \n",
    "> 2. Then, perform post-hoc comparisons between control vs. each intervention using Dunn's test.  \n",
    ">\n",
    "> To increase the statistical power, samples are grouped by intervention (i.e., strain, age, and sex are pooled) in the Kruskal–Wallis H-test. Note that Kruskal-Wallis H-test can shrink the variance utilizing all samples (per reaction), whose statistical power is better than the repeated Mann–Whitney U-tests (a.k.a., Wilcoxon rank-sum tests) in the case of small sample size. Although tricky, the P-value adjustment in (1) is performed across all reactions under the assumption that reactions are independent, which would be more conservative and less likely raise referees' eyebrows than using nominal P-value cutoff. Also, the reactions assigned with invariable flux values across all groups are eliminated from tests in advance, which can reduce the number of hypotheses. Because the post-hoc comparisons (2) are to address the effect of each intervention within a specific reaction, the P-values are adjusted across interventions only within the reaction (not across reactions). "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# — †1. Go to the top of the R sub-notebook —  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3-1. Kruskal–Wallis H-test (Flux ~ Intervention), followed by Dunn's test (Intervention)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 3-1-1/2/3/4. Import the summary tables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Import the summary tables\n",
    "fileDir = './ExportData/'\n",
    "ipynbName = '220610_LC-M001-related-transcriptomics-GEM-ver2-4_StatisticalTest_'\n",
    "fileName = 'flux-comparison.xlsx'\n",
    "sheetName = 'Main_vs-each-control'\n",
    "tempDF = pd.read_excel(fileDir+ipynbName+fileName, sheet_name=sheetName, engine='openpyxl')\n",
    "tempDF = tempDF.set_index('ReactionID')\n",
    "display(tempDF)\n",
    "\n",
    "statDF = tempDF"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 3-1-5. Changed reactions (Kruskal–Wallis H-test)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Prepare variables in the model\n",
    "tempS = statDF.loc[:, statDF.columns.str.contains('_Hstat')].columns.to_series()\n",
    "variableL = tempS.str.replace('_Hstat', '').tolist()\n",
    "\n",
    "#Changed reactions\n",
    "for variable in variableL:\n",
    "    tempDF = statDF.loc[statDF[variable+'_AdjPval']<0.05]\n",
    "    tempDF = tempDF.sort_values(by=variable+'_AdjPval', ascending=True)\n",
    "    tempL1 = tempDF.loc[:, tempDF.columns.str.contains('_FluxMedian')].columns.tolist()\n",
    "    tempL2 = tempDF.loc[:, tempDF.columns.str.contains('^'+variable+'_')].columns.tolist()\n",
    "    tempDF = tempDF[[col_n for subL in [['ReactionName'], tempL1, tempL2] for col_n in subL]]\n",
    "    print(variable+' (adjusted P < 0.05):', len(tempDF))\n",
    "    display(tempDF)\n",
    "    print('')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 3-1-6. Changed reactions by each intervention (Dunn's test)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Clean DF for adjusted P-value and z-statistic\n",
    "variable = 'Intervention'\n",
    "##Take adjusted P-value\n",
    "tempDF1 = statDF.loc[:, statDF.columns.str.contains('-vs-.*_AdjPval$')]\n",
    "tempDF1.columns = tempDF1.columns.str.replace('_AdjPval$', '')\n",
    "tempDF1 = pd.merge(statDF[['ReactionName', variable+'_AdjPval']], tempDF1,\n",
    "                   left_index=True, right_index=True, how='left')\n",
    "tempDF1 = tempDF1.sort_values(by=variable+'_AdjPval', ascending=True)\n",
    "print('Adjusted P-value:')\n",
    "display(tempDF1)\n",
    "##Take z-statistic\n",
    "tempDF2 = statDF.loc[:, statDF.columns.str.contains('-vs-.*_zStat$')]\n",
    "tempDF2.columns = tempDF2.columns.str.replace('_zStat$', '')\n",
    "tempDF2 = pd.merge(statDF[['ReactionName', variable+'_AdjPval']], tempDF2,\n",
    "                   left_index=True, right_index=True, how='left')\n",
    "tempDF2 = tempDF2.sort_values(by=variable+'_AdjPval', ascending=True)\n",
    "print('Changed direction (z-statistic):')\n",
    "display(tempDF2)\n",
    "\n",
    "pvalDF = tempDF1\n",
    "diffDF = tempDF2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> –> When z-statistic is positive, the flux median of contrast is larger than that of baseline.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Extract only the changed reactions\n",
    "variable = 'Intervention'\n",
    "tempDF = pvalDF.loc[pvalDF[variable+'_AdjPval']<0.05]\n",
    "print(variable+' (adjusted P < 0.05):', len(tempDF))\n",
    "\n",
    "tempL = tempDF.loc[:, tempDF.columns.str.contains('-vs-')].columns.tolist()\n",
    "for comparison in tempL:\n",
    "    tempS = tempDF[comparison]\n",
    "    tempS = tempS.loc[tempS<0.05]\n",
    "    print(' - '+comparison+':', len(tempS))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> Check the changed reactions (nominal P-value < 0.05) as reference.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Extract only the changed reactions\n",
    "variable = 'Intervention'\n",
    "tempL = statDF.loc[statDF[variable+'_Pval']<0.05].index.tolist()\n",
    "tempDF = pvalDF.loc[tempL]\n",
    "print(variable+' (nominal P < 0.05):', len(tempDF))\n",
    "\n",
    "tempL = tempDF.loc[:, tempDF.columns.str.contains('-vs-')].columns.tolist()\n",
    "for comparison in tempL:\n",
    "    tempS = tempDF[comparison]\n",
    "    tempS = tempS.loc[tempS<0.05]\n",
    "    print(' - '+comparison+':', len(tempS))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "### 3-2. Visualization: clustermap"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 3-2-1. Sample-wide"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> For visualization purpose, it would be better to use the centered values based on control groups, because around zero value simply indicates no change from control.  \n",
    "> –> At the same time, scaling is required for good clustering, but RobustScaler() returns just the centered (i.e., median-subtracted) values if IQR = 0. Hence, not IQR but the maximum absolete value is used for scaling. Of note, scaling eliminates the information about how largely an intervention group is different from the control group, but it rather reflects the rank-based statistical tests.  \n",
    "> –> In addition, only the changed reactions are used for good clustering.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Centering per each control category\n",
    "tempDF = pd.DataFrame(index=fluxDF.index)\n",
    "for category in categoryL:\n",
    "    tempL = sampleDF.loc[sampleDF['Category']==category].index.tolist()\n",
    "    tempDF1 = fluxDF.loc[:, tempL]\n",
    "    #Calculate the median of control group\n",
    "    tempL = sampleDF.loc[sampleDF['Group']==sampleDF['Control']].index.tolist()\n",
    "    tempDF2 = tempDF1.loc[:, tempDF1.columns.isin(tempL)]\n",
    "    tempS = tempDF2.median(axis=1)\n",
    "    #Centering\n",
    "    tempDF1 = tempDF1.T - tempS\n",
    "    #Merge\n",
    "    tempDF = pd.merge(tempDF, tempDF1.T, left_index=True, right_index=True, how='left')\n",
    "\n",
    "#Scaling\n",
    "scaler = MaxAbsScaler(copy=True)\n",
    "tempA = scaler.fit_transform(tempDF.T)#axis: column\n",
    "tempDF1 = pd.DataFrame(data=tempA, index=tempDF.columns, columns=tempDF.index)\n",
    "tempDF1 = tempDF1.T\n",
    "\n",
    "display(tempDF1)\n",
    "display(tempDF1.iloc[:, :15].describe())\n",
    "\n",
    "fluxDF_scaled = tempDF1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Prepare DF to plot only the changed reactions\n",
    "variable = 'Intervention'\n",
    "tempL = statDF.loc[statDF[variable+'_Pval']<0.05].index.tolist()\n",
    "tempDF = fluxDF_scaled.loc[tempL]\n",
    "print(variable+' (nominal P < 0.05):', len(tempDF))\n",
    "\n",
    "#Prepare color labels for the changed reactions\n",
    "regulation = 'Changed'\n",
    "tempD0 = {'Con1':'Control 1', 'Acar':'Acarbose',\n",
    "          '17aE':'17'+r'$\\alpha$'+'-Estradiol', 'Prot':'Protandim',\n",
    "          'Rapa':'Rapamycin', 'CalR':'CR diet',\n",
    "          'Con2':'Control 2', 'MetR':'MR diet',\n",
    "          'GHRw':'GHR WT', 'GHRk':'GHR KO',\n",
    "          'SneW':'Snell WT', 'SneD':'Snell dwarf'}\n",
    "tempD1 = {'Control 1':'tab:blue', 'Acarbose':'tab:red',\n",
    "          '17'+r'$\\alpha$'+'-Estradiol':'tab:green', 'Protandim':'tab:brown',\n",
    "          'Rapamycin':'tab:purple', 'CR diet':'tab:olive',\n",
    "          'Control 2':'tab:blue', 'MR diet':'tab:orange',\n",
    "          'GHR WT':'tab:gray', 'GHR KO':'tab:pink',\n",
    "          'Snell WT':'black', 'Snell dwarf':'tab:cyan'}\n",
    "tempD2 = {}\n",
    "for label in tempD1.keys():\n",
    "    if tempD1[label]=='tab:blue':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(1)\n",
    "    elif tempD1[label]=='tab:orange':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(3)\n",
    "    elif tempD1[label]=='tab:green':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(5)\n",
    "    elif tempD1[label]=='tab:red':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(7)\n",
    "    elif tempD1[label]=='tab:purple':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(9)\n",
    "    elif tempD1[label]=='tab:brown':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(11)\n",
    "    elif tempD1[label]=='tab:pink':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(13)\n",
    "    elif tempD1[label]=='tab:gray':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(15)\n",
    "    elif tempD1[label]=='tab:olive':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(17)\n",
    "    elif tempD1[label]=='tab:cyan':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(19)\n",
    "    else:\n",
    "        tempD2[label] = 'dimgray'\n",
    "tempL1 = statDF.loc[statDF[variable+'_AdjPval']<0.05].index.tolist()\n",
    "tempL2 = statDF.loc[statDF[variable+'_Pval']<0.05].index.tolist()\n",
    "tempDF1 = pvalDF.loc[tempDF.index.tolist(), pvalDF.columns.str.contains('-vs-')]\n",
    "for col_n in tempDF1.columns.tolist():\n",
    "    tempS1 = pvalDF[col_n]\n",
    "    tempS2 = diffDF[col_n]\n",
    "    if regulation=='Changed':\n",
    "        tempS2 = tempS2.loc[(tempS1<0.05)]\n",
    "    elif regulation=='Increased':\n",
    "        tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2>0)]\n",
    "    elif regulation=='Decreased':\n",
    "        tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2<0)]\n",
    "    #Replace dummy value (P-value) with color code\n",
    "    label = tempD0[re.sub('-vs-.*', '', col_n)]\n",
    "    tempL = []\n",
    "    count1 = 0#just for checking\n",
    "    count2 = 0#just for checking\n",
    "    for rxn in tempDF1.index.tolist():\n",
    "        if rxn in tempS2.index.tolist():\n",
    "            if rxn in tempL1:#Adjusted P-value\n",
    "                tempL.append(tempD1[label])\n",
    "                count1 += 1#Just for checking\n",
    "            else:#Nominal P-value\n",
    "                tempL.append(tempD2[label])\n",
    "            count2 += 1#Just for checking\n",
    "        else:\n",
    "            tempL.append('white')\n",
    "    tempDF1[label] = tempL\n",
    "    print(regulation+' reaction in '+col_n)\n",
    "    print(' -> adjusted P < 0.05:', count1)\n",
    "    print(' -> nominal P < 0.05:', count2)\n",
    "tempDF1 = tempDF1.loc[:, tempDF1.columns.isin(tempD1.keys())]\n",
    "##Remove the group showing no changed reactions\n",
    "for col_n in tempDF1.columns.tolist():\n",
    "    tempS = tempDF1[col_n]\n",
    "    tempS = (tempS=='white')\n",
    "    if tempS.sum()==len(tempS):\n",
    "        tempDF1 = tempDF1.drop(columns=col_n)\n",
    "\n",
    "#Prepare color labels for the models\n",
    "tempS = sampleDF['Intervention'].map(tempD0).map(tempD1)\n",
    "\n",
    "#Clustermap\n",
    "sns.set(style='ticks', font='Arial', context='talk')\n",
    "cm = sns.clustermap(tempDF, method='ward', metric='euclidean', cmap='RdBu_r',\n",
    "                    row_cluster=True, col_cluster=True, row_linkage=None, col_linkage=None,\n",
    "                    row_colors=tempDF1, col_colors=tempS, xticklabels=False, yticklabels=False,\n",
    "                    dendrogram_ratio=(0.1, 0.1), colors_ratio=(0.03, 0.03),\n",
    "                    cbar_pos=(0.05, -0.075, 0.3, 0.025), cbar_kws={'orientation': 'horizontal'},\n",
    "                    figsize=(12, 12), **{'vmin':-1, 'vmax':1})\n",
    "cm.cax.set_title('Scaled flux value', size='medium',\n",
    "                 verticalalignment='bottom', horizontalalignment='center')\n",
    "cm.cax.tick_params(labelsize='small')\n",
    "bottom, top = cm.ax_heatmap.get_ylim()\n",
    "#cm.ax_heatmap.set_ylim(bottom + 0.5, top - 0.5)##To avoid half cut of first and last rows\n",
    "hm = cm.ax_heatmap.get_position()\n",
    "rd = cm.ax_row_dendrogram.get_position()\n",
    "cd = cm.ax_col_dendrogram.get_position()\n",
    "cm.ax_heatmap.set_position([hm.x0, hm.y0, hm.width, hm.height])\n",
    "cm.ax_row_dendrogram.set_position([rd.x0, rd.y0, rd.width, rd.height])\n",
    "cm.ax_col_dendrogram.set_position([cd.x0, cd.y0, cd.width, cd.height])\n",
    "cm.ax_heatmap.set_xlabel('Sample')\n",
    "cm.ax_heatmap.set_ylabel('Reaction')\n",
    "##row/column color bar legend (axis is same with cm.cax!)\n",
    "tempL = []\n",
    "for group in tempD1.keys():\n",
    "    tempL.append(mpatches.Patch(color=tempD1[group], label=group))\n",
    "legend1 = plt.legend(handles=tempL, fontsize='small', labelspacing=0.2, ncol=2,\n",
    "                     title='Intervention', title_fontsize='medium',\n",
    "                     bbox_to_anchor=(1, -2), loc='lower left', borderaxespad=3.5, frameon=False)\n",
    "plt.gca().add_artist(legend1)\n",
    "tempL = []\n",
    "for group in tempD1.keys():\n",
    "    if group in tempDF1.columns.tolist():\n",
    "        tempL.append(mpatches.Patch(color=tempD1[group],\n",
    "                                    label='by '+group+' (adjusted '+r'$P$'+' < 0.05)'))\n",
    "for group in tempD2.keys():\n",
    "    if group in tempDF1.columns.tolist():\n",
    "        tempL.append(mpatches.Patch(color=tempD2[group], label='(nominal '+r'$P$'+' < 0.05)'))\n",
    "legend2 = plt.legend(handles=tempL, fontsize='small', labelspacing=0.2, ncol=2,\n",
    "                     title='Changed reactions (vs. each control)', title_fontsize='medium',\n",
    "                     bbox_to_anchor=(1, 2.5), loc='upper left', borderaxespad=3.5, frameon=False)\n",
    "plt.gca().add_artist(legend2)\n",
    "##Save\n",
    "fileDir = './ExportFigures/'\n",
    "ipynbName = '220610_LC-M001-related-transcriptomics-GEM-ver2-4_FluxAnalysis_'\n",
    "fileName = 'flux-clustermap-across-samples.tif'\n",
    "plt.gcf().savefig(fileDir+ipynbName+fileName, dpi=300, bbox_inches='tight', pad_inches=0.04,\n",
    "                  pil_kwargs={'compression':'tiff_lzw'})\n",
    "plt.show()\n",
    "\n",
    "#Save label order\n",
    "tempDF = pvalDF.loc[tempDF.index[cm.dendrogram_row.reordered_ind]]\n",
    "tempDF = tempDF.reset_index()\n",
    "tempDF.index.name = 'Ycoord'\n",
    "display(tempDF)\n",
    "fileDir = './ExportData/'\n",
    "ipynbName = '220610_LC-M001-related-transcriptomics-GEM-ver2-4_FluxAnalysis_'\n",
    "fileName = 'flux-clustermap-across-samples_ticks-order.tsv'\n",
    "tempDF.to_csv(fileDir+ipynbName+fileName, index=True, sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 3-2-2. Group-wide 1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> Sample-based clustering would be difficult to be understood.  \n",
    "> –> Combine samples per intervention group. Of note, not median but mean is used as the summary statistic because the median loses many information in the case of small sample size.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Summarize samples per intervention group\n",
    "tempL = ['Con1', 'Acar', '17aE', 'Prot', 'Rapa', 'CalR',\n",
    "         'Con2', 'MetR', 'GHRw', 'GHRk', 'SneW', 'SneD']\n",
    "tempDF = pd.DataFrame(index=fluxDF.index)\n",
    "for intervention in tempL:\n",
    "    tempL = sampleDF.loc[sampleDF['Intervention']==intervention].index.tolist()\n",
    "    tempDF1 = fluxDF.loc[:, tempL]\n",
    "    #Calculate the mean\n",
    "    tempS = tempDF1.mean(axis=1)\n",
    "    tempS.name = intervention\n",
    "    #Merge\n",
    "    tempDF = pd.merge(tempDF, tempS, left_index=True, right_index=True, how='left')\n",
    "print('Summary per intervention group:')\n",
    "display(tempDF.describe())\n",
    "\n",
    "#Centering per each control category\n",
    "tempDF1 = pd.DataFrame(index=tempDF.index)\n",
    "for category in categoryL:\n",
    "    tempL = sampleDF['Intervention'].loc[sampleDF['Category']==category].unique().tolist()\n",
    "    tempDF2 = tempDF.loc[:, tempL]\n",
    "    #Centering\n",
    "    tempS = tempDF2[category]\n",
    "    tempDF2 = tempDF2.T - tempS\n",
    "    #Merge\n",
    "    tempDF1 = pd.merge(tempDF1, tempDF2.T, left_index=True, right_index=True, how='left')\n",
    "\n",
    "#Scaling\n",
    "scaler = MaxAbsScaler(copy=True)\n",
    "tempA = scaler.fit_transform(tempDF1.T)#axis: column\n",
    "tempDF2 = pd.DataFrame(data=tempA, index=tempDF1.columns, columns=tempDF1.index)\n",
    "tempDF2 = tempDF2.T\n",
    "tempL = ['Con1', 'Acar', '17aE', 'Prot', 'Rapa', 'CalR',\n",
    "         'Con2', 'MetR', 'GHRw', 'GHRk', 'SneW', 'SneD']\n",
    "tempDF2 = tempDF2[tempL]#Sort\n",
    "\n",
    "print('After centering and scaling:')\n",
    "display(tempDF2)\n",
    "display(tempDF2.iloc[:, :15].describe())\n",
    "\n",
    "fluxDF_scaled = tempDF2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Prepare DF to plot only the changed reactions\n",
    "variable = 'Intervention'\n",
    "tempL = statDF.loc[statDF[variable+'_Pval']<0.05].index.tolist()\n",
    "tempDF = fluxDF_scaled.loc[tempL]\n",
    "print(variable+' (nominal P < 0.05):', len(tempDF))\n",
    "\n",
    "#Prepare color labels for the changed reactions\n",
    "regulation = 'Changed'\n",
    "tempD0 = {'Con1':'Control 1', 'Acar':'Acarbose',\n",
    "          '17aE':'17'+r'$\\alpha$'+'-Estradiol', 'Prot':'Protandim',\n",
    "          'Rapa':'Rapamycin', 'CalR':'CR diet',\n",
    "          'Con2':'Control 2', 'MetR':'MR diet',\n",
    "          'GHRw':'GHR WT', 'GHRk':'GHR KO',\n",
    "          'SneW':'Snell WT', 'SneD':'Snell dwarf'}\n",
    "tempDF.columns = tempDF.columns.map(tempD0)\n",
    "tempD1 = {'Control 1':'tab:blue', 'Acarbose':'tab:red',\n",
    "          '17'+r'$\\alpha$'+'-Estradiol':'tab:green', 'Protandim':'tab:brown',\n",
    "          'Rapamycin':'tab:purple', 'CR diet':'tab:olive',\n",
    "          'Control 2':'tab:blue', 'MR diet':'tab:orange',\n",
    "          'GHR WT':'tab:gray', 'GHR KO':'tab:pink',\n",
    "          'Snell WT':'black', 'Snell dwarf':'tab:cyan'}\n",
    "tempD2 = {}\n",
    "for label in tempD1.keys():\n",
    "    if tempD1[label]=='tab:blue':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(1)\n",
    "    elif tempD1[label]=='tab:orange':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(3)\n",
    "    elif tempD1[label]=='tab:green':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(5)\n",
    "    elif tempD1[label]=='tab:red':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(7)\n",
    "    elif tempD1[label]=='tab:purple':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(9)\n",
    "    elif tempD1[label]=='tab:brown':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(11)\n",
    "    elif tempD1[label]=='tab:pink':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(13)\n",
    "    elif tempD1[label]=='tab:gray':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(15)\n",
    "    elif tempD1[label]=='tab:olive':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(17)\n",
    "    elif tempD1[label]=='tab:cyan':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(19)\n",
    "    else:\n",
    "        tempD2[label] = 'dimgray'\n",
    "tempL1 = statDF.loc[statDF[variable+'_AdjPval']<0.05].index.tolist()\n",
    "tempL2 = statDF.loc[statDF[variable+'_Pval']<0.05].index.tolist()\n",
    "tempDF1 = pvalDF.loc[tempDF.index.tolist(), pvalDF.columns.str.contains('-vs-')]\n",
    "for col_n in tempDF1.columns.tolist():\n",
    "    tempS1 = pvalDF[col_n]\n",
    "    tempS2 = diffDF[col_n]\n",
    "    if regulation=='Changed':\n",
    "        tempS2 = tempS2.loc[(tempS1<0.05)]\n",
    "    elif regulation=='Increased':\n",
    "        tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2>0)]\n",
    "    elif regulation=='Decreased':\n",
    "        tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2<0)]\n",
    "    #Replace dummy value (P-value) with color code\n",
    "    label = tempD0[re.sub('-vs-.*', '', col_n)]\n",
    "    tempL = []\n",
    "    count1 = 0#just for checking\n",
    "    count2 = 0#just for checking\n",
    "    for rxn in tempDF1.index.tolist():\n",
    "        if rxn in tempS2.index.tolist():\n",
    "            if rxn in tempL1:#Adjusted P-value\n",
    "                tempL.append(tempD1[label])\n",
    "                count1 += 1#Just for checking\n",
    "            else:#Nominal P-value\n",
    "                tempL.append(tempD2[label])\n",
    "            count2 += 1#Just for checking\n",
    "        else:\n",
    "            tempL.append('white')\n",
    "    tempDF1[label] = tempL\n",
    "    print(regulation+' reaction in '+col_n)\n",
    "    print(' -> adjusted P < 0.05:', count1)\n",
    "    print(' -> nominal P < 0.05:', count2)\n",
    "tempDF1 = tempDF1.loc[:, tempDF1.columns.isin(tempD1.keys())]\n",
    "##Remove the group showing no changed reactions\n",
    "for col_n in tempDF1.columns.tolist():\n",
    "    tempS = tempDF1[col_n]\n",
    "    tempS = (tempS=='white')\n",
    "    if tempS.sum()==len(tempS):\n",
    "        tempDF1 = tempDF1.drop(columns=col_n)\n",
    "\n",
    "#Clustermap\n",
    "sns.set(style='ticks', font='Arial', context='talk')\n",
    "cm = sns.clustermap(tempDF.T, method='ward', metric='euclidean', cmap='RdBu_r',\n",
    "                    row_cluster=True, col_cluster=True, row_linkage=None, col_linkage=None,\n",
    "                    row_colors=None, col_colors=tempDF1, xticklabels=False, yticklabels=True,\n",
    "                    dendrogram_ratio=(0.02, 0.1), colors_ratio=(0.03, 0.05),\n",
    "                    cbar_pos=(0.08, -0.15, 0.3, 0.05), cbar_kws={'orientation': 'horizontal'},\n",
    "                    figsize=(15, 5), **{'vmin':-1, 'vmax':1})\n",
    "cm.cax.set_title('Group mean of flux values\\n(vs. each control; scaled)', size='medium',\n",
    "                 verticalalignment='bottom', horizontalalignment='center')\n",
    "cm.cax.tick_params(labelsize='small')\n",
    "bottom, top = cm.ax_heatmap.get_ylim()\n",
    "#cm.ax_heatmap.set_ylim(bottom + 0.5, top - 0.5)##To avoid half cut of first and last rows\n",
    "hm = cm.ax_heatmap.get_position()\n",
    "rd = cm.ax_row_dendrogram.get_position()\n",
    "cd = cm.ax_col_dendrogram.get_position()\n",
    "cm.ax_heatmap.set_position([hm.x0, hm.y0, hm.width, hm.height])\n",
    "cm.ax_row_dendrogram.set_position([rd.x0, rd.y0, rd.width, rd.height])\n",
    "cm.ax_col_dendrogram.set_position([cd.x0, cd.y0, cd.width, cd.height])\n",
    "cm.ax_heatmap.set_xlabel('Reaction')\n",
    "cm.ax_heatmap.set_ylabel('')\n",
    "##row/column color bar legend (axis is same with cm.cax!)\n",
    "tempL = []\n",
    "for group in tempD1.keys():\n",
    "    if group in tempDF1.columns.tolist():\n",
    "        tempL.append(mpatches.Patch(color=tempD1[group],\n",
    "                                    label='by '+group+' (adjusted '+r'$P$'+' < 0.05)'))\n",
    "for group in tempD2.keys():\n",
    "    if group in tempDF1.columns.tolist():\n",
    "        tempL.append(mpatches.Patch(color=tempD2[group], label='(nominal '+r'$P$'+' < 0.05)'))\n",
    "legend2 = plt.legend(handles=tempL, fontsize='small', labelspacing=0.2, ncol=2,\n",
    "                     title='Changed reactions (vs. each control)', title_fontsize='medium',\n",
    "                     bbox_to_anchor=(1, 0.5), loc='center left', borderaxespad=7, frameon=False)\n",
    "plt.gca().add_artist(legend2)\n",
    "##Save\n",
    "#fileDir = './ExportFigures/'\n",
    "#ipynbName = '220610_LC-M001-related-transcriptomics-GEM-ver2-4_FluxAnalysis_'\n",
    "#fileName = 'flux-clustermap-across-interventions(vs-each-control).tif'\n",
    "#plt.gcf().savefig(fileDir+ipynbName+fileName, dpi=300, bbox_inches='tight', pad_inches=0.04,\n",
    "#                  pil_kwargs={'compression':'tiff_lzw'})\n",
    "plt.show()\n",
    "\n",
    "#Save label order\n",
    "tempDF = pvalDF.loc[tempDF.index[cm.dendrogram_col.reordered_ind]]\n",
    "tempDF = tempDF.reset_index()\n",
    "tempDF.index.name = 'Xcoord'\n",
    "display(tempDF)\n",
    "#fileDir = './ExportData/'\n",
    "#ipynbName = '220610_LC-M001-related-transcriptomics-GEM-ver2-4_FluxAnalysis_'\n",
    "#fileName = 'flux-clustermap-across-interventions(vs-each-control)_ticks-order.tsv'\n",
    "#tempDF.to_csv(fileDir+ipynbName+fileName, index=True, sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> –> Remove control groups because it is meaningless after centering.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Prepare DF to plot only the changed reactions\n",
    "variable = 'Intervention'\n",
    "tempL = statDF.loc[statDF[variable+'_Pval']<0.05].index.tolist()\n",
    "tempDF = fluxDF_scaled.loc[tempL]\n",
    "print(variable+' (nominal P < 0.05):', len(tempDF))\n",
    "##Remove control groups\n",
    "tempDF = tempDF.drop(columns=categoryL)\n",
    "\n",
    "#Prepare color labels for the changed reactions\n",
    "regulation = 'Changed'\n",
    "tempD0 = {'Con1':'Control 1', 'Acar':'Acarbose',\n",
    "          '17aE':'17'+r'$\\alpha$'+'-Estradiol', 'Prot':'Protandim',\n",
    "          'Rapa':'Rapamycin', 'CalR':'CR diet',\n",
    "          'Con2':'Control 2', 'MetR':'MR diet',\n",
    "          'GHRw':'GHR WT', 'GHRk':'GHR KO',\n",
    "          'SneW':'Snell WT', 'SneD':'Snell dwarf'}\n",
    "tempDF.columns = tempDF.columns.map(tempD0)\n",
    "tempD1 = {'Control 1':'tab:blue', 'Acarbose':'tab:red',\n",
    "          '17'+r'$\\alpha$'+'-Estradiol':'tab:green', 'Protandim':'tab:brown',\n",
    "          'Rapamycin':'tab:purple', 'CR diet':'tab:olive',\n",
    "          'Control 2':'tab:blue', 'MR diet':'tab:orange',\n",
    "          'GHR WT':'tab:gray', 'GHR KO':'tab:pink',\n",
    "          'Snell WT':'black', 'Snell dwarf':'tab:cyan'}\n",
    "tempD2 = {}\n",
    "for label in tempD1.keys():\n",
    "    if tempD1[label]=='tab:blue':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(1)\n",
    "    elif tempD1[label]=='tab:orange':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(3)\n",
    "    elif tempD1[label]=='tab:green':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(5)\n",
    "    elif tempD1[label]=='tab:red':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(7)\n",
    "    elif tempD1[label]=='tab:purple':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(9)\n",
    "    elif tempD1[label]=='tab:brown':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(11)\n",
    "    elif tempD1[label]=='tab:pink':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(13)\n",
    "    elif tempD1[label]=='tab:gray':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(15)\n",
    "    elif tempD1[label]=='tab:olive':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(17)\n",
    "    elif tempD1[label]=='tab:cyan':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(19)\n",
    "    else:\n",
    "        tempD2[label] = 'dimgray'\n",
    "tempL1 = statDF.loc[statDF[variable+'_AdjPval']<0.05].index.tolist()\n",
    "tempL2 = statDF.loc[statDF[variable+'_Pval']<0.05].index.tolist()\n",
    "tempDF1 = pvalDF.loc[tempDF.index.tolist(), pvalDF.columns.str.contains('-vs-')]\n",
    "for col_n in tempDF1.columns.tolist():\n",
    "    tempS1 = pvalDF[col_n]\n",
    "    tempS2 = diffDF[col_n]\n",
    "    if regulation=='Changed':\n",
    "        tempS2 = tempS2.loc[(tempS1<0.05)]\n",
    "    elif regulation=='Increased':\n",
    "        tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2>0)]\n",
    "    elif regulation=='Decreased':\n",
    "        tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2<0)]\n",
    "    #Replace dummy value (P-value) with color code\n",
    "    label = tempD0[re.sub('-vs-.*', '', col_n)]\n",
    "    tempL = []\n",
    "    count1 = 0#just for checking\n",
    "    count2 = 0#just for checking\n",
    "    for rxn in tempDF1.index.tolist():\n",
    "        if rxn in tempS2.index.tolist():\n",
    "            if rxn in tempL1:#Adjusted P-value\n",
    "                tempL.append(tempD1[label])\n",
    "                count1 += 1#Just for checking\n",
    "            else:#Nominal P-value\n",
    "                tempL.append(tempD2[label])\n",
    "            count2 += 1#Just for checking\n",
    "        else:\n",
    "            tempL.append('white')\n",
    "    tempDF1[label] = tempL\n",
    "    print(regulation+' reaction in '+col_n)\n",
    "    print(' -> adjusted P < 0.05:', count1)\n",
    "    print(' -> nominal P < 0.05:', count2)\n",
    "tempDF1 = tempDF1.loc[:, tempDF1.columns.isin(tempD1.keys())]\n",
    "##Remove the group showing no changed reactions\n",
    "for col_n in tempDF1.columns.tolist():\n",
    "    tempS = tempDF1[col_n]\n",
    "    tempS = (tempS=='white')\n",
    "    if tempS.sum()==len(tempS):\n",
    "        tempDF1 = tempDF1.drop(columns=col_n)\n",
    "\n",
    "#Clustermap\n",
    "sns.set(style='ticks', font='Arial', context='talk')\n",
    "cm = sns.clustermap(tempDF.T, method='ward', metric='euclidean', cmap='RdBu_r',\n",
    "                    row_cluster=True, col_cluster=True, row_linkage=None, col_linkage=None,\n",
    "                    row_colors=None, col_colors=tempDF1, xticklabels=False, yticklabels=True,\n",
    "                    dendrogram_ratio=(0.02, 0.1), colors_ratio=(0.03, 0.06),\n",
    "                    cbar_pos=(0.08, -0.175, 0.3, 0.06), cbar_kws={'orientation': 'horizontal'},\n",
    "                    figsize=(15, 4), **{'vmin':-1, 'vmax':1})\n",
    "cm.cax.set_title('Group mean of flux values\\n(vs. each control; scaled)', size='medium',\n",
    "                 verticalalignment='bottom', horizontalalignment='center')\n",
    "cm.cax.tick_params(labelsize='small')\n",
    "bottom, top = cm.ax_heatmap.get_ylim()\n",
    "#cm.ax_heatmap.set_ylim(bottom + 0.5, top - 0.5)##To avoid half cut of first and last rows\n",
    "hm = cm.ax_heatmap.get_position()\n",
    "rd = cm.ax_row_dendrogram.get_position()\n",
    "cd = cm.ax_col_dendrogram.get_position()\n",
    "cm.ax_heatmap.set_position([hm.x0, hm.y0, hm.width, hm.height])\n",
    "cm.ax_row_dendrogram.set_position([rd.x0, rd.y0, rd.width, rd.height])\n",
    "cm.ax_col_dendrogram.set_position([cd.x0, cd.y0, cd.width, cd.height])\n",
    "cm.ax_heatmap.set_xlabel('Reaction')\n",
    "cm.ax_heatmap.set_ylabel('')\n",
    "##row/column color bar legend (axis is same with cm.cax!)\n",
    "tempL = []\n",
    "for group in tempD1.keys():\n",
    "    if group in tempDF1.columns.tolist():\n",
    "        tempL.append(mpatches.Patch(color=tempD1[group],\n",
    "                                    label='by '+group+' (adjusted '+r'$P$'+' < 0.05)'))\n",
    "for group in tempD2.keys():\n",
    "    if group in tempDF1.columns.tolist():\n",
    "        tempL.append(mpatches.Patch(color=tempD2[group], label='(nominal '+r'$P$'+' < 0.05)'))\n",
    "legend2 = plt.legend(handles=tempL, fontsize='small', labelspacing=0.2, ncol=2,\n",
    "                     title='Changed reactions (vs. each control)', title_fontsize='medium',\n",
    "                     bbox_to_anchor=(1, 0.5), loc='center left', borderaxespad=7, frameon=False)\n",
    "plt.gca().add_artist(legend2)\n",
    "##Save\n",
    "fileDir = './ExportFigures/'\n",
    "ipynbName = '220610_LC-M001-related-transcriptomics-GEM-ver2-4_FluxAnalysis_'\n",
    "fileName = 'flux-clustermap-across-interventions(vs-each-control).tif'\n",
    "plt.gcf().savefig(fileDir+ipynbName+fileName, dpi=300, bbox_inches='tight', pad_inches=0.04,\n",
    "                  pil_kwargs={'compression':'tiff_lzw'})\n",
    "plt.show()\n",
    "\n",
    "#Save label order\n",
    "tempDF = pvalDF.loc[tempDF.index[cm.dendrogram_col.reordered_ind]]\n",
    "tempDF = tempDF.reset_index()\n",
    "tempDF.index.name = 'Xcoord'\n",
    "display(tempDF)\n",
    "fileDir = './ExportData/'\n",
    "ipynbName = '220610_LC-M001-related-transcriptomics-GEM-ver2-4_FluxAnalysis_'\n",
    "fileName = 'flux-clustermap-across-interventions(vs-each-control)_ticks-order.tsv'\n",
    "tempDF.to_csv(fileDir+ipynbName+fileName, index=True, sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 3-2-3. Group-wide 2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> Centering by each control may highlight the difference too strongly, because each control group is forced to be around zero.  \n",
    "> –> Check simple standardization (Z-score) across all groups.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Summarize samples per intervention group\n",
    "tempL = ['Con1', 'Acar', '17aE', 'Prot', 'Rapa', 'CalR',\n",
    "         'Con2', 'MetR', 'GHRw', 'GHRk', 'SneW', 'SneD']\n",
    "tempDF = pd.DataFrame(index=fluxDF.index)\n",
    "for intervention in tempL:\n",
    "    tempL = sampleDF.loc[sampleDF['Intervention']==intervention].index.tolist()\n",
    "    tempDF1 = fluxDF.loc[:, tempL]\n",
    "    #Calculate the mean\n",
    "    tempS = tempDF1.mean(axis=1)\n",
    "    tempS.name = intervention\n",
    "    #Merge\n",
    "    tempDF = pd.merge(tempDF, tempS, left_index=True, right_index=True, how='left')\n",
    "print('Summary per intervention group:')\n",
    "display(tempDF.describe())\n",
    "\n",
    "#Z-score transformation\n",
    "scaler = StandardScaler(copy=True, with_mean=True, with_std=True)\n",
    "tempA = scaler.fit_transform(tempDF.T)#axis: column\n",
    "tempDF1 = pd.DataFrame(data=tempA, index=tempDF.columns, columns=tempDF.index)\n",
    "tempDF1 = tempDF1.T\n",
    "tempL = ['Con1', 'Acar', '17aE', 'Prot', 'Rapa', 'CalR',\n",
    "         'Con2', 'MetR', 'GHRw', 'GHRk', 'SneW', 'SneD']\n",
    "tempDF1 = tempDF1[tempL]#Sort\n",
    "\n",
    "print('After centering and scaling:')\n",
    "display(tempDF1)\n",
    "display(tempDF1.iloc[:, :15].describe())\n",
    "\n",
    "fluxDF_scaled = tempDF1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Prepare DF to plot only the changed reactions\n",
    "variable = 'Intervention'\n",
    "tempL = statDF.loc[statDF[variable+'_Pval']<0.05].index.tolist()\n",
    "tempDF = fluxDF_scaled.loc[tempL]\n",
    "print(variable+' (nominal P < 0.05):', len(tempDF))\n",
    "\n",
    "#Prepare color labels for the changed reactions\n",
    "regulation = 'Changed'\n",
    "tempD0 = {'Con1':'Control 1', 'Acar':'Acarbose',\n",
    "          '17aE':'17'+r'$\\alpha$'+'-Estradiol', 'Prot':'Protandim',\n",
    "          'Rapa':'Rapamycin', 'CalR':'CR diet',\n",
    "          'Con2':'Control 2', 'MetR':'MR diet',\n",
    "          'GHRw':'GHR WT', 'GHRk':'GHR KO',\n",
    "          'SneW':'Snell WT', 'SneD':'Snell dwarf'}\n",
    "tempDF.columns = tempDF.columns.map(tempD0)\n",
    "tempD1 = {'Control 1':'tab:blue', 'Acarbose':'tab:red',\n",
    "          '17'+r'$\\alpha$'+'-Estradiol':'tab:green', 'Protandim':'tab:brown',\n",
    "          'Rapamycin':'tab:purple', 'CR diet':'tab:olive',\n",
    "          'Control 2':'tab:blue', 'MR diet':'tab:orange',\n",
    "          'GHR WT':'tab:gray', 'GHR KO':'tab:pink',\n",
    "          'Snell WT':'black', 'Snell dwarf':'tab:cyan'}\n",
    "tempD2 = {}\n",
    "for label in tempD1.keys():\n",
    "    if tempD1[label]=='tab:blue':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(1)\n",
    "    elif tempD1[label]=='tab:orange':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(3)\n",
    "    elif tempD1[label]=='tab:green':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(5)\n",
    "    elif tempD1[label]=='tab:red':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(7)\n",
    "    elif tempD1[label]=='tab:purple':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(9)\n",
    "    elif tempD1[label]=='tab:brown':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(11)\n",
    "    elif tempD1[label]=='tab:pink':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(13)\n",
    "    elif tempD1[label]=='tab:gray':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(15)\n",
    "    elif tempD1[label]=='tab:olive':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(17)\n",
    "    elif tempD1[label]=='tab:cyan':\n",
    "        tempD2[label] = plt.get_cmap('tab20')(19)\n",
    "    else:\n",
    "        tempD2[label] = 'dimgray'\n",
    "tempL1 = statDF.loc[statDF[variable+'_AdjPval']<0.05].index.tolist()\n",
    "tempL2 = statDF.loc[statDF[variable+'_Pval']<0.05].index.tolist()\n",
    "tempDF1 = pvalDF.loc[tempDF.index.tolist(), pvalDF.columns.str.contains('-vs-')]\n",
    "for col_n in tempDF1.columns.tolist():\n",
    "    tempS1 = pvalDF[col_n]\n",
    "    tempS2 = diffDF[col_n]\n",
    "    if regulation=='Changed':\n",
    "        tempS2 = tempS2.loc[(tempS1<0.05)]\n",
    "    elif regulation=='Increased':\n",
    "        tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2>0)]\n",
    "    elif regulation=='Decreased':\n",
    "        tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2<0)]\n",
    "    #Replace dummy value (P-value) with color code\n",
    "    label = tempD0[re.sub('-vs-.*', '', col_n)]\n",
    "    tempL = []\n",
    "    count1 = 0#just for checking\n",
    "    count2 = 0#just for checking\n",
    "    for rxn in tempDF1.index.tolist():\n",
    "        if rxn in tempS2.index.tolist():\n",
    "            if rxn in tempL1:#Adjusted P-value\n",
    "                tempL.append(tempD1[label])\n",
    "                count1 += 1#Just for checking\n",
    "            else:#Nominal P-value\n",
    "                tempL.append(tempD2[label])\n",
    "            count2 += 1#Just for checking\n",
    "        else:\n",
    "            tempL.append('white')\n",
    "    tempDF1[label] = tempL\n",
    "    print(regulation+' reaction in '+col_n)\n",
    "    print(' -> adjusted P < 0.05:', count1)\n",
    "    print(' -> nominal P < 0.05:', count2)\n",
    "tempDF1 = tempDF1.loc[:, tempDF1.columns.isin(tempD1.keys())]\n",
    "##Remove the group showing no changed reactions\n",
    "for col_n in tempDF1.columns.tolist():\n",
    "    tempS = tempDF1[col_n]\n",
    "    tempS = (tempS=='white')\n",
    "    if tempS.sum()==len(tempS):\n",
    "        tempDF1 = tempDF1.drop(columns=col_n)\n",
    "\n",
    "#Clustermap\n",
    "sns.set(style='ticks', font='Arial', context='talk')\n",
    "cm = sns.clustermap(tempDF.T, method='ward', metric='euclidean', cmap='RdBu_r',\n",
    "                    row_cluster=True, col_cluster=True, row_linkage=None, col_linkage=None,\n",
    "                    row_colors=None, col_colors=tempDF1, xticklabels=False, yticklabels=True,\n",
    "                    dendrogram_ratio=(0.02, 0.1), colors_ratio=(0.03, 0.05),\n",
    "                    cbar_pos=(0.08, -0.15, 0.3, 0.05), cbar_kws={'orientation': 'horizontal'},\n",
    "                    figsize=(15, 5), **{'vmin':-3, 'vmax':3})\n",
    "cm.cax.set_title('Group mean of flux values\\n('+r'$Z$'+'-score)', size='medium',\n",
    "                 verticalalignment='bottom', horizontalalignment='center')\n",
    "cm.cax.tick_params(labelsize='small')\n",
    "bottom, top = cm.ax_heatmap.get_ylim()\n",
    "#cm.ax_heatmap.set_ylim(bottom + 0.5, top - 0.5)##To avoid half cut of first and last rows\n",
    "hm = cm.ax_heatmap.get_position()\n",
    "rd = cm.ax_row_dendrogram.get_position()\n",
    "cd = cm.ax_col_dendrogram.get_position()\n",
    "cm.ax_heatmap.set_position([hm.x0, hm.y0, hm.width, hm.height])\n",
    "cm.ax_row_dendrogram.set_position([rd.x0, rd.y0, rd.width, rd.height])\n",
    "cm.ax_col_dendrogram.set_position([cd.x0, cd.y0, cd.width, cd.height])\n",
    "cm.ax_heatmap.set_xlabel('Reaction')\n",
    "cm.ax_heatmap.set_ylabel('')\n",
    "##row/column color bar legend (axis is same with cm.cax!)\n",
    "tempL = []\n",
    "for group in tempD1.keys():\n",
    "    if group in tempDF1.columns.tolist():\n",
    "        tempL.append(mpatches.Patch(color=tempD1[group],\n",
    "                                    label='by '+group+' (adjusted '+r'$P$'+' < 0.05)'))\n",
    "for group in tempD2.keys():\n",
    "    if group in tempDF1.columns.tolist():\n",
    "        tempL.append(mpatches.Patch(color=tempD2[group], label='(nominal '+r'$P$'+' < 0.05)'))\n",
    "legend2 = plt.legend(handles=tempL, fontsize='small', labelspacing=0.2, ncol=2,\n",
    "                     title='Changed reactions (vs. each control)', title_fontsize='medium',\n",
    "                     bbox_to_anchor=(1, 0.5), loc='center left', borderaxespad=7, frameon=False)\n",
    "plt.gca().add_artist(legend2)\n",
    "##Save\n",
    "fileDir = './ExportFigures/'\n",
    "ipynbName = '220610_LC-M001-related-transcriptomics-GEM-ver2-4_FluxAnalysis_'\n",
    "fileName = 'flux-clustermap-across-interventions.tif'\n",
    "plt.gcf().savefig(fileDir+ipynbName+fileName, dpi=300, bbox_inches='tight', pad_inches=0.04,\n",
    "                  pil_kwargs={'compression':'tiff_lzw'})\n",
    "plt.show()\n",
    "\n",
    "#Save label order\n",
    "tempDF = pvalDF.loc[tempDF.index[cm.dendrogram_col.reordered_ind]]\n",
    "tempDF = tempDF.reset_index()\n",
    "tempDF.index.name = 'Xcoord'\n",
    "display(tempDF)\n",
    "fileDir = './ExportData/'\n",
    "ipynbName = '220610_LC-M001-related-transcriptomics-GEM-ver2-4_FluxAnalysis_'\n",
    "fileName = 'flux-clustermap-across-interventions_ticks-order.tsv'\n",
    "tempDF.to_csv(fileDir+ipynbName+fileName, index=True, sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> –> This representation is informative for us, but probably normal readers would feel difficulty to understand the figure message...  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3-3. Visualization: venn diagram"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Prepare label and color\n",
    "tempD0 = {'Con1':'Control 1', 'Acar':'Acarbose',\n",
    "          '17aE':'17'+r'$\\alpha$'+'-Estradiol', 'Prot':'Protandim',\n",
    "          'Rapa':'Rapamycin', 'CalR':'CR diet',\n",
    "          'Con2':'Control 2', 'MetR':'MR diet',\n",
    "          'GHRw':'GHR WT', 'GHRk':'GHR KO',\n",
    "          'SneW':'Snell WT', 'SneD':'Snell dwarf'}\n",
    "tempD1 = {'Acarbose':'tab:red', 'Rapamycin':'tab:purple', 'MR diet':'tab:orange',\n",
    "          'GHR KO':'tab:pink', 'Snell dwarf':'tab:cyan'}\n",
    "\n",
    "#Visualization per direction\n",
    "for regulation in ['Changed', 'Increased', 'Decreased']:\n",
    "    #Prepare reaction sets\n",
    "    variable = 'Intervention'\n",
    "    tempL = statDF.loc[statDF[variable+'_AdjPval']<0.05].index.tolist()\n",
    "    tempDF1 = pvalDF.loc[tempL, pvalDF.columns.str.contains('-vs-')]\n",
    "    tempDF2 = diffDF.loc[tempL, diffDF.columns.str.contains('-vs-')]\n",
    "    tempD2 = {}\n",
    "    count = 0\n",
    "    for col_n in tempDF1.columns.tolist():\n",
    "        tempS1 = tempDF1[col_n]\n",
    "        tempS2 = tempDF2[col_n]\n",
    "        if regulation=='Changed':\n",
    "            tempS2 = tempS2.loc[(tempS1<0.05)]\n",
    "        elif regulation=='Increased':\n",
    "            tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2>0)]\n",
    "        elif regulation=='Decreased':\n",
    "            tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2<0)]\n",
    "        if len(tempS2)>0:\n",
    "            count += 1\n",
    "        label = tempD0[re.sub('-vs-.*', '', col_n)]\n",
    "        tempD2[label] = set(tempS2.index.tolist())\n",
    "    if count==0:\n",
    "        print(regulation+' reactions: no significance in any comparisons')\n",
    "        continue\n",
    "    ##Sort to make consistent order in manual legend generation\n",
    "    tempD = {}\n",
    "    count = 0\n",
    "    for label in tempD1.keys():\n",
    "        tempD[label] = tempD2[label]\n",
    "        if len(tempD2[label])>0:\n",
    "            count += 1\n",
    "    \n",
    "    #Skip the followings if no significant reaction\n",
    "    if count==0:\n",
    "        print(regulation+' reactions: no significance in any target comparisons')\n",
    "        continue\n",
    "    \n",
    "    #Venn diagram\n",
    "    sns.set(style='ticks', font='Arial', context='talk')\n",
    "    fig, ax = plt.subplots(figsize=(4, 4))\n",
    "    venn(tempD, fmt='{size:,}', cmap=list(tempD1.values()), legend_loc=None, ax=ax)\n",
    "    plt.setp(ax, ylim=(0.0, 1.0))#Otherwise, weird space...\n",
    "    ##Add legend annotation\n",
    "    x_coord = [0.25, 0.75, 0.9, 0.5, 0.1]\n",
    "    y_coord = [0.8, 0.8, 0.35, 0.0, 0.35]\n",
    "    h_align = ['right', 'left', 'left', 'center', 'right']\n",
    "    v_align = ['bottom', 'bottom', 'top', 'top', 'top']\n",
    "    for i in range(len(tempD1)):\n",
    "        key = list(tempD1.keys())[i]\n",
    "        total = f'{len(tempD[key]):,}'\n",
    "        ax.text(x_coord[i], y_coord[i], key+'\\n('+total+' reactions)',\n",
    "                fontsize='small', multialignment='center',\n",
    "                horizontalalignment=h_align[i], verticalalignment=v_align[i],\n",
    "                bbox={'boxstyle':'round', 'facecolor':tempD1[key], 'pad':0.2, 'alpha':0.5})\n",
    "    title = regulation+' reactions (vs. each control)'\n",
    "    ax.set_title(title, fontsize='medium')\n",
    "    ##Save\n",
    "    if regulation!='Changed':\n",
    "        fileDir = './ExportFigures/'\n",
    "        ipynbName = '220610_LC-M001-related-transcriptomics-GEM-ver2-4_FluxAnalysis_'\n",
    "        fileName = 'rxn-venn-'+regulation.lower()+'.tif'\n",
    "        plt.gcf().savefig(fileDir+ipynbName+fileName, dpi=300, bbox_inches='tight', pad_inches=0.04,\n",
    "                          pil_kwargs={'compression':'tiff_lzw'})\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Export reaction list in each subset in the venn diagram\n",
    "for regulation in ['Increased', 'Decreased']:\n",
    "    #Prepare reaction sets\n",
    "    variable = 'Intervention'\n",
    "    tempL1 = statDF.loc[statDF[variable+'_AdjPval']<0.05].index.tolist()\n",
    "    tempL2 = ['Acar-vs-Con1', 'Rapa-vs-Con1', 'MetR-vs-Con2', 'GHRk-vs-GHRw', 'SneD-vs-SneW']\n",
    "    tempDF1 = pvalDF.loc[tempL1, tempL2]\n",
    "    tempDF2 = diffDF.loc[tempL1, tempL2]\n",
    "    tempD = {}\n",
    "    count = 0\n",
    "    for col_n in tempDF1.columns.tolist():\n",
    "        tempS1 = tempDF1[col_n]\n",
    "        tempS2 = tempDF2[col_n]\n",
    "        if regulation=='Changed':\n",
    "            tempS2 = tempS2.loc[(tempS1<0.05)]\n",
    "        elif regulation=='Increased':\n",
    "            tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2>0)]\n",
    "        elif regulation=='Decreased':\n",
    "            tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2<0)]\n",
    "        if len(tempS2)>0:\n",
    "            count += 1\n",
    "        tempD[col_n] = set(tempS2.index.tolist())\n",
    "    \n",
    "    #Skip the followings if no significant reaction\n",
    "    if count==0:\n",
    "        print(regulation+' reactions: no significance in any target comparisons')\n",
    "        continue\n",
    "    \n",
    "    #Prepare a new .xlsx file (dummy README)\n",
    "    tempL1 = [len(tempD[key]) for key in tempD.keys()]\n",
    "    tempDF = pd.DataFrame({'Group':tempD.keys(), 'nReactions':tempL1})\n",
    "    tempDF = tempDF.reset_index().rename(columns={'index':'VennOrder'})\n",
    "    tempDF['VennOrder'] = tempDF['VennOrder'] + 1\n",
    "    fileDir = './ExportData/'\n",
    "    ipynbName = '220610_LC-M001-related-transcriptomics-GEM-ver2-4_FluxAnalysis_'\n",
    "    fileName = 'rxn-venn-'+regulation.lower()+'.xlsx'\n",
    "    tempDF.to_excel(fileDir+ipynbName+fileName, sheet_name='README', header=True, index=False)\n",
    "    display(tempDF)#Check\n",
    "    \n",
    "    t_start = time.time()\n",
    "    #Extract overall set\n",
    "    for key_i in range(len(tempD)):\n",
    "        key = list(tempD.keys())[key_i]\n",
    "        tempS = tempD[key]\n",
    "        tempDF = statDF.loc[statDF.index.isin(tempS)]\n",
    "        #Save summary table by appended to the above .xlsx file\n",
    "        ##Prepare sheet name\n",
    "        tempL1 = ['NA' for i in range(len(tempD))]\n",
    "        tempL1[key_i] = '1'\n",
    "        setName = '('+','.join(tempL1)+')'\n",
    "        with pd.ExcelWriter(fileDir+ipynbName+fileName, mode='a', engine='openpyxl') as writer:\n",
    "            tempDF.to_excel(writer, sheet_name=setName, header=True, index=True)\n",
    "        print(' - '+setName+':', len(tempDF))\n",
    "    \n",
    "    #Extract subset\n",
    "    tempL1 = ['1', '0']\n",
    "    tempL2 = [[k1, k2, k3, k4, k5] for k1 in tempL1 for k2 in tempL1 for k3 in tempL1 for k4 in tempL1 for k5 in tempL1]\n",
    "    #tempL2.remove(['0', '0', '0', '0', '0'])\n",
    "    for tempL1 in tempL2:\n",
    "        #Positive reaction set\n",
    "        tempL3 = [list(tempD.values())[key_i] for key_i, binary in enumerate(tempL1) if binary=='1']\n",
    "        tempS1 = set(statDF.loc[statDF[variable+'_AdjPval']<0.05].index.tolist())#Initialize\n",
    "        for tempS in tempL3:\n",
    "            tempS1 = tempS1 & tempS\n",
    "        #Negative reaction set\n",
    "        tempL3 = [list(tempD.values())[key_i] for key_i, binary in enumerate(tempL1) if binary=='0']\n",
    "        tempS2 = set()#Initialize\n",
    "        for tempS in tempL3:\n",
    "            tempS2 = tempS2 | tempS\n",
    "        #Extract subset\n",
    "        tempS = tempS1 - tempS2\n",
    "        tempDF = statDF.loc[statDF.index.isin(tempS)]\n",
    "        #Save summary table by appended to the above .xlsx file\n",
    "        ##Prepare sheet name\n",
    "        setName = '('+','.join(tempL1)+')'\n",
    "        with pd.ExcelWriter(fileDir+ipynbName+fileName, mode='a', engine='openpyxl') as writer:\n",
    "            tempDF.to_excel(writer, sheet_name=setName, header=True, index=True)\n",
    "        print(' - '+setName+':', len(tempDF))\n",
    "    \n",
    "    t_elapsed = time.time() - t_start\n",
    "    print(' - Elapsed time:', round(t_elapsed//60), 'min', round(t_elapsed%60, 1), 'sec')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3-4. Visualization: boxplot"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 3-4-1. Reactions increased by MR, GHRKO, SnellDW"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Prepare the target reaction set\n",
    "posL = ['MetR-vs-Con2', 'GHRk-vs-GHRw', 'SneD-vs-SneW']\n",
    "negL = ['']\n",
    "regulation = 'Increased'\n",
    "tempL = statDF.loc[statDF[variable+'_AdjPval']<0.05].index.tolist()\n",
    "tempDF1 = pvalDF.loc[tempL, pvalDF.columns.str.contains('-vs-')]\n",
    "tempDF2 = diffDF.loc[tempL, diffDF.columns.str.contains('-vs-')]\n",
    "tempS = pd.Series(np.repeat(True, len(tempDF1)), index=tempDF1.index)#Initialize\n",
    "for col_n in tempDF1.columns.tolist():\n",
    "    tempS1 = tempDF1[col_n]\n",
    "    tempS2 = tempDF2[col_n]\n",
    "    if col_n in posL:\n",
    "        if regulation=='Increased':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2>0)\n",
    "        elif regulation=='Decreased':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2<0)\n",
    "    elif col_n in negL:\n",
    "        tempS3 = (tempS1>=0.05)\n",
    "        #Significance for inverse regulation\n",
    "        if regulation=='Increased':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2<0)\n",
    "        elif regulation=='Decreased':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2>0)\n",
    "        tempS1 = tempS3 | tempS1\n",
    "    else:\n",
    "        tempS1 = (tempS1>=0.0)\n",
    "    #Update True\n",
    "    tempS = tempS & tempS1\n",
    "tempL = tempS.loc[tempS.tolist()].index.tolist()\n",
    "print(len(tempL), regulation.lower()+' reactions with significance in', posL, 'but not in', negL)\n",
    "\n",
    "#Select representatives\n",
    "topX = np.min([30, len(tempL)])\n",
    "topX_plot = np.min([5, len(tempL)])\n",
    "tempDF = statDF.loc[:, statDF.columns.str.contains('_AdjPval$')]\n",
    "tempDF = pd.merge(statDF[['ReactionName', variable+'_Pval']], tempDF,\n",
    "                  left_index=True, right_index=True, how='left')\n",
    "tempDF = tempDF.loc[tempL].sort_values(by=variable+'_AdjPval', ascending=True)\n",
    "print('Top', topX, 'reactions (sort by the main effect of '+variable+'):')\n",
    "display(tempDF.iloc[:topX])\n",
    "plotL = tempDF.index.tolist()[:topX_plot]\n",
    "\n",
    "#Prepare DF for plot\n",
    "tempDF = fluxDF.reset_index().melt(var_name='ModelID', value_name='Flux', id_vars='ReactionID')\n",
    "tempDF1 = sampleDF.reset_index()[['ModelID', 'Intervention']]\n",
    "tempDF = pd.merge(tempDF, tempDF1, on='ModelID', how='left')\n",
    "\n",
    "#Prepare label and color\n",
    "tempD0 = {'Con1':'Control 1', 'Acar':'Acarbose',\n",
    "          '17aE':'17'+r'$\\alpha$'+'-Estradiol', 'Prot':'Protandim',\n",
    "          'Rapa':'Rapamycin', 'CalR':'CR diet',\n",
    "          'Con2':'Control 2', 'MetR':'MR diet',\n",
    "          'GHRw':'GHR WT', 'GHRk':'GHR KO',\n",
    "          'SneW':'Snell WT', 'SneD':'Snell dwarf'}\n",
    "tempDF['Group'] = tempDF['Intervention'].map(tempD0)\n",
    "tempD1 = {'Control 1':'tab:blue', 'Acarbose':'tab:red',\n",
    "          '17'+r'$\\alpha$'+'-Estradiol':'tab:green', 'Protandim':'tab:brown',\n",
    "          'Rapamycin':'tab:purple', 'CR diet':'tab:olive',\n",
    "          'Control 2':'tab:blue', 'MR diet':'tab:orange',\n",
    "          'GHR WT':'tab:gray', 'GHR KO':'tab:pink',\n",
    "          'Snell WT':'black', 'Snell dwarf':'tab:cyan'}\n",
    "\n",
    "#Visualize each representative\n",
    "for rank_i in range(len(plotL)):\n",
    "    print(' - Rank '+str(rank_i+1)+' (sort by the main effect of '+variable+'):')\n",
    "    rxn = plotL[rank_i]\n",
    "    #Check reaction summary\n",
    "    tempDF1 = pd.DataFrame(rxnDF.loc[rxn]).T\n",
    "    display(tempDF1)\n",
    "    \n",
    "    #Select Flux\n",
    "    tempDF1 = tempDF.loc[tempDF['ReactionID']==rxn]\n",
    "    \n",
    "    #Check Flux summary\n",
    "    tempDF2 = tempDF1.groupby(['Group'])['Flux'].agg(['count', 'mean', 'std'])\n",
    "    tempL1 = []\n",
    "    tempL2 = []\n",
    "    for row_n in tempDF2.index.tolist():\n",
    "        count, mean, std = tempDF2.loc[row_n]\n",
    "        tempL1.append(mean - 1.96*std/np.sqrt(count))\n",
    "        tempL2.append(mean + 1.96*std/np.sqrt(count))\n",
    "    tempDF2['0.025'] = tempL1\n",
    "    tempDF2['0.975'] = tempL2\n",
    "    tempDF2 = tempDF2.loc[list(tempD1.keys())]#Sort\n",
    "    display(tempDF2)\n",
    "    \n",
    "    #Prepare significance labels\n",
    "    ##Retrieve statistical significance\n",
    "    tempS = pvalDF.loc[rxn, pvalDF.columns.str.contains('-vs-')]\n",
    "    tempS.name = 'AdjPval'\n",
    "    ##Clean\n",
    "    tempDF2 = tempS.index.to_series().str.split(pat='-vs-', expand=True)\n",
    "    tempDF2 = tempDF2.rename(columns={0:'Contrast', 1:'Baseline'})\n",
    "    tempDF2 = pd.merge(tempDF2, tempS, left_index=True, right_index=True, how='left')\n",
    "    tempDF2['Contrast'] = tempDF2['Contrast'].map(tempD0)\n",
    "    tempDF2['Baseline'] = tempDF2['Baseline'].map(tempD0)\n",
    "    ##Convert p-value to label\n",
    "    tempL = []\n",
    "    for row_i in range(len(tempDF2)):\n",
    "        pval = tempDF2['AdjPval'].iloc[row_i]\n",
    "        if pval<0.001:\n",
    "            tempL.append('***')\n",
    "        elif pval<0.01:\n",
    "            tempL.append('**')\n",
    "        elif pval<0.05:\n",
    "            tempL.append('*')\n",
    "        else:\n",
    "            pval_text = Decimal(str(pval)).quantize(Decimal('0.01'), rounding=ROUND_HALF_UP)\n",
    "            tempL.append(r'$P$ = '+str(pval_text))\n",
    "    tempDF2['SignifLabel'] = tempL\n",
    "    ##Add the y-position level in figure\n",
    "    tempDF2['YposLevel'] = [0, 1, 2, 3, 4, 2, 3, 4]\n",
    "    display(tempDF2)\n",
    "    \n",
    "    #Visualization\n",
    "    sns.set(style='ticks', font='Arial', context='talk')\n",
    "    plt.figure(figsize=(4, 4))\n",
    "    sns.boxplot(data=tempDF1, x='Group', y='Flux', order=list(tempD1.keys()), palette=tempD1,\n",
    "                dodge=False, showfliers=False, showcaps=True, notch=False)\n",
    "    p = sns.stripplot(data=tempDF1, x='Group', y='Flux',\n",
    "                      order=list(tempD1.keys()), palette=tempD1, dodge=False, jitter=0.3,\n",
    "                      size=5, edgecolor='black', linewidth=1, **{'marker':'o', 'alpha':0.5})\n",
    "    ##Add border line\n",
    "    for xcoord in [5.5, 7.5, 9.5]:\n",
    "        p.axvline(x=xcoord, **{'linestyle':'dotted', 'color':'black', 'zorder':0})\n",
    "    ##Set axis\n",
    "    sns.despine()\n",
    "    plt.setp(p.get_xticklabels(), rotation=70, horizontalalignment='right',\n",
    "             verticalalignment='center', rotation_mode='anchor')\n",
    "    ###Reset the axis range for the labels\n",
    "    ymin, ymax = p.get_ylim()\n",
    "    tempL = p.get_yticks().tolist()\n",
    "    yinter = int(tempL[1]) - int(tempL[0])\n",
    "    tempL = [int(y) for y in tempL if (y>ymin)&(y<ymax)]#seaborn seems to prepare wider range!?\n",
    "    p.set(ylim=(ymin, ymax+(yinter*2.5)), yticks=tempL)\n",
    "    p.set_yticklabels(['{:,}'.format(int(y)) for y in tempL])\n",
    "    ##Add significance labels\n",
    "    aline_ymin = tempDF1['Flux'].max()\n",
    "    aline_ymargin = yinter/2\n",
    "    for row_i in range(len(tempDF2)):\n",
    "        #Baseline\n",
    "        group_0 = tempDF2['Baseline'].iloc[row_i]\n",
    "        index_0 = list(tempD1.keys()).index(group_0)\n",
    "        xcoord_0 = index_0\n",
    "        #Contrast\n",
    "        group_1 = tempDF2['Contrast'].iloc[row_i]\n",
    "        index_1 = list(tempD1.keys()).index(group_1)\n",
    "        xcoord_1 = index_1\n",
    "        #Standard point of marker\n",
    "        xcoord = (xcoord_0+xcoord_1)/2\n",
    "        ycoord = aline_ymin + aline_ymargin*tempDF2['YposLevel'].iloc[row_i]\n",
    "        label = tempDF2['SignifLabel'].iloc[row_i]\n",
    "        #Add annotation lines\n",
    "        aline_offset = yinter/10\n",
    "        aline_length = yinter/10 + aline_offset\n",
    "        plt.plot([xcoord_0, xcoord_0, xcoord_1, xcoord_1],\n",
    "                 [ycoord+aline_offset, ycoord+aline_length, ycoord+aline_length, ycoord+aline_offset],\n",
    "                 lw=1.5, c='k')\n",
    "        #Add annotation text\n",
    "        if label in ['***', '**', '*']:\n",
    "            text_offset = yinter/4\n",
    "            p.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                       horizontalalignment='center', verticalalignment='center',\n",
    "                       fontsize='medium', color='k')\n",
    "        else:\n",
    "            text_offset = yinter/5\n",
    "            p.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                       horizontalalignment='center', verticalalignment='bottom',\n",
    "                       fontsize='x-small', color='k')\n",
    "    ##Set axis label and title\n",
    "    plt.setp(p, xlabel='', ylabel='Sample flux value [a.u.]')\n",
    "    p.set_title(rxn, {'fontsize':'medium'})\n",
    "    ##Save\n",
    "    fileDir = './ExportFigures/'\n",
    "    ipynbName = '220610_LC-M001-related-transcriptomics-GEM-ver2-4_FluxAnalysis_'\n",
    "    fileName = 'flux-boxplot('+rxn+').tif'\n",
    "    plt.gcf().savefig(fileDir+ipynbName+fileName, dpi=300, bbox_inches='tight', pad_inches=0.04,\n",
    "                      pil_kwargs={'compression':'tiff_lzw'})\n",
    "    plt.show()\n",
    "    print('')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 3-4-2. Reactions decreased by MR, GHRKO, SnellDW"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Prepare the target reaction set\n",
    "posL = ['MetR-vs-Con2', 'GHRk-vs-GHRw', 'SneD-vs-SneW']\n",
    "negL = ['']\n",
    "regulation = 'Decreased'\n",
    "tempL = statDF.loc[statDF[variable+'_AdjPval']<0.05].index.tolist()\n",
    "tempDF1 = pvalDF.loc[tempL, pvalDF.columns.str.contains('-vs-')]\n",
    "tempDF2 = diffDF.loc[tempL, diffDF.columns.str.contains('-vs-')]\n",
    "tempS = pd.Series(np.repeat(True, len(tempDF1)), index=tempDF1.index)#Initialize\n",
    "for col_n in tempDF1.columns.tolist():\n",
    "    tempS1 = tempDF1[col_n]\n",
    "    tempS2 = tempDF2[col_n]\n",
    "    if col_n in posL:\n",
    "        if regulation=='Increased':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2>0)\n",
    "        elif regulation=='Decreased':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2<0)\n",
    "    elif col_n in negL:\n",
    "        tempS3 = (tempS1>=0.05)\n",
    "        #Significance for inverse regulation\n",
    "        if regulation=='Increased':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2<0)\n",
    "        elif regulation=='Decreased':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2>0)\n",
    "        tempS1 = tempS3 | tempS1\n",
    "    else:\n",
    "        tempS1 = (tempS1>=0.0)\n",
    "    #Update True\n",
    "    tempS = tempS & tempS1\n",
    "tempL = tempS.loc[tempS.tolist()].index.tolist()\n",
    "print(len(tempL), regulation.lower()+' reactions with significance in', posL, 'but not in', negL)\n",
    "\n",
    "#Select representatives\n",
    "topX = np.min([30, len(tempL)])\n",
    "topX_plot = np.min([5, len(tempL)])\n",
    "tempDF = statDF.loc[:, statDF.columns.str.contains('_AdjPval$')]\n",
    "tempDF = pd.merge(statDF[['ReactionName', variable+'_Pval']], tempDF,\n",
    "                  left_index=True, right_index=True, how='left')\n",
    "tempDF = tempDF.loc[tempL].sort_values(by=variable+'_AdjPval', ascending=True)\n",
    "print('Top', topX, 'reactions (sort by the main effect of '+variable+'):')\n",
    "display(tempDF.iloc[:topX])\n",
    "plotL = tempDF.index.tolist()[:topX_plot]\n",
    "\n",
    "#Prepare DF for plot\n",
    "tempDF = fluxDF.reset_index().melt(var_name='ModelID', value_name='Flux', id_vars='ReactionID')\n",
    "tempDF1 = sampleDF.reset_index()[['ModelID', 'Intervention']]\n",
    "tempDF = pd.merge(tempDF, tempDF1, on='ModelID', how='left')\n",
    "\n",
    "#Prepare label and color\n",
    "tempD0 = {'Con1':'Control 1', 'Acar':'Acarbose',\n",
    "          '17aE':'17'+r'$\\alpha$'+'-Estradiol', 'Prot':'Protandim',\n",
    "          'Rapa':'Rapamycin', 'CalR':'CR diet',\n",
    "          'Con2':'Control 2', 'MetR':'MR diet',\n",
    "          'GHRw':'GHR WT', 'GHRk':'GHR KO',\n",
    "          'SneW':'Snell WT', 'SneD':'Snell dwarf'}\n",
    "tempDF['Group'] = tempDF['Intervention'].map(tempD0)\n",
    "tempD1 = {'Control 1':'tab:blue', 'Acarbose':'tab:red',\n",
    "          '17'+r'$\\alpha$'+'-Estradiol':'tab:green', 'Protandim':'tab:brown',\n",
    "          'Rapamycin':'tab:purple', 'CR diet':'tab:olive',\n",
    "          'Control 2':'tab:blue', 'MR diet':'tab:orange',\n",
    "          'GHR WT':'tab:gray', 'GHR KO':'tab:pink',\n",
    "          'Snell WT':'black', 'Snell dwarf':'tab:cyan'}\n",
    "\n",
    "#Visualize each representative\n",
    "for rank_i in range(len(plotL)):\n",
    "    print(' - Rank '+str(rank_i+1)+' (sort by the main effect of '+variable+'):')\n",
    "    rxn = plotL[rank_i]\n",
    "    #Check reaction summary\n",
    "    tempDF1 = pd.DataFrame(rxnDF.loc[rxn]).T\n",
    "    display(tempDF1)\n",
    "    \n",
    "    #Select Flux\n",
    "    tempDF1 = tempDF.loc[tempDF['ReactionID']==rxn]\n",
    "    \n",
    "    #Check Flux summary\n",
    "    tempDF2 = tempDF1.groupby(['Group'])['Flux'].agg(['count', 'mean', 'std'])\n",
    "    tempL1 = []\n",
    "    tempL2 = []\n",
    "    for row_n in tempDF2.index.tolist():\n",
    "        count, mean, std = tempDF2.loc[row_n]\n",
    "        tempL1.append(mean - 1.96*std/np.sqrt(count))\n",
    "        tempL2.append(mean + 1.96*std/np.sqrt(count))\n",
    "    tempDF2['0.025'] = tempL1\n",
    "    tempDF2['0.975'] = tempL2\n",
    "    tempDF2 = tempDF2.loc[list(tempD1.keys())]#Sort\n",
    "    display(tempDF2)\n",
    "    \n",
    "    #Prepare significance labels\n",
    "    ##Retrieve statistical significance\n",
    "    tempS = pvalDF.loc[rxn, pvalDF.columns.str.contains('-vs-')]\n",
    "    tempS.name = 'AdjPval'\n",
    "    ##Clean\n",
    "    tempDF2 = tempS.index.to_series().str.split(pat='-vs-', expand=True)\n",
    "    tempDF2 = tempDF2.rename(columns={0:'Contrast', 1:'Baseline'})\n",
    "    tempDF2 = pd.merge(tempDF2, tempS, left_index=True, right_index=True, how='left')\n",
    "    tempDF2['Contrast'] = tempDF2['Contrast'].map(tempD0)\n",
    "    tempDF2['Baseline'] = tempDF2['Baseline'].map(tempD0)\n",
    "    ##Convert p-value to label\n",
    "    tempL = []\n",
    "    for row_i in range(len(tempDF2)):\n",
    "        pval = tempDF2['AdjPval'].iloc[row_i]\n",
    "        if pval<0.001:\n",
    "            tempL.append('***')\n",
    "        elif pval<0.01:\n",
    "            tempL.append('**')\n",
    "        elif pval<0.05:\n",
    "            tempL.append('*')\n",
    "        else:\n",
    "            pval_text = Decimal(str(pval)).quantize(Decimal('0.01'), rounding=ROUND_HALF_UP)\n",
    "            tempL.append(r'$P$ = '+str(pval_text))\n",
    "    tempDF2['SignifLabel'] = tempL\n",
    "    ##Add the y-position level in figure\n",
    "    tempDF2['YposLevel'] = [0, 1, 2, 3, 4, 2, 3, 4]\n",
    "    display(tempDF2)\n",
    "    \n",
    "    #Visualization\n",
    "    sns.set(style='ticks', font='Arial', context='talk')\n",
    "    plt.figure(figsize=(4, 4))\n",
    "    sns.boxplot(data=tempDF1, x='Group', y='Flux', order=list(tempD1.keys()), palette=tempD1,\n",
    "                dodge=False, showfliers=False, showcaps=True, notch=False)\n",
    "    p = sns.stripplot(data=tempDF1, x='Group', y='Flux',\n",
    "                      order=list(tempD1.keys()), palette=tempD1, dodge=False, jitter=0.3,\n",
    "                      size=5, edgecolor='black', linewidth=1, **{'marker':'o', 'alpha':0.5})\n",
    "    ##Add border line\n",
    "    for xcoord in [5.5, 7.5, 9.5]:\n",
    "        p.axvline(x=xcoord, **{'linestyle':'dotted', 'color':'black', 'zorder':0})\n",
    "    ##Set axis\n",
    "    sns.despine()\n",
    "    plt.setp(p.get_xticklabels(), rotation=70, horizontalalignment='right',\n",
    "             verticalalignment='center', rotation_mode='anchor')\n",
    "    ###Reset the axis range for the labels\n",
    "    ymin, ymax = p.get_ylim()\n",
    "    tempL = p.get_yticks().tolist()\n",
    "    yinter = int(tempL[1]) - int(tempL[0])\n",
    "    tempL = [int(y) for y in tempL if (y>ymin)&(y<ymax)]#seaborn seems to prepare wider range!?\n",
    "    p.set(ylim=(ymin, ymax+(yinter*2.5)), yticks=tempL)\n",
    "    p.set_yticklabels(['{:,}'.format(int(y)) for y in tempL])\n",
    "    ##Add significance labels\n",
    "    aline_ymin = tempDF1['Flux'].max()\n",
    "    aline_ymargin = yinter/2\n",
    "    for row_i in range(len(tempDF2)):\n",
    "        #Baseline\n",
    "        group_0 = tempDF2['Baseline'].iloc[row_i]\n",
    "        index_0 = list(tempD1.keys()).index(group_0)\n",
    "        xcoord_0 = index_0\n",
    "        #Contrast\n",
    "        group_1 = tempDF2['Contrast'].iloc[row_i]\n",
    "        index_1 = list(tempD1.keys()).index(group_1)\n",
    "        xcoord_1 = index_1\n",
    "        #Standard point of marker\n",
    "        xcoord = (xcoord_0+xcoord_1)/2\n",
    "        ycoord = aline_ymin + aline_ymargin*tempDF2['YposLevel'].iloc[row_i]\n",
    "        label = tempDF2['SignifLabel'].iloc[row_i]\n",
    "        #Add annotation lines\n",
    "        aline_offset = yinter/10\n",
    "        aline_length = yinter/10 + aline_offset\n",
    "        plt.plot([xcoord_0, xcoord_0, xcoord_1, xcoord_1],\n",
    "                 [ycoord+aline_offset, ycoord+aline_length, ycoord+aline_length, ycoord+aline_offset],\n",
    "                 lw=1.5, c='k')\n",
    "        #Add annotation text\n",
    "        if label in ['***', '**', '*']:\n",
    "            text_offset = yinter/4\n",
    "            p.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                       horizontalalignment='center', verticalalignment='center',\n",
    "                       fontsize='medium', color='k')\n",
    "        else:\n",
    "            text_offset = yinter/5\n",
    "            p.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                       horizontalalignment='center', verticalalignment='bottom',\n",
    "                       fontsize='x-small', color='k')\n",
    "    ##Set axis label and title\n",
    "    plt.setp(p, xlabel='', ylabel='Sample flux value [a.u.]')\n",
    "    p.set_title(rxn, {'fontsize':'medium'})\n",
    "    ##Save\n",
    "    fileDir = './ExportFigures/'\n",
    "    ipynbName = '220610_LC-M001-related-transcriptomics-GEM-ver2-4_FluxAnalysis_'\n",
    "    fileName = 'flux-boxplot('+rxn+').tif'\n",
    "    plt.gcf().savefig(fileDir+ipynbName+fileName, dpi=300, bbox_inches='tight', pad_inches=0.04,\n",
    "                      pil_kwargs={'compression':'tiff_lzw'})\n",
    "    plt.show()\n",
    "    print('')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 3-4-3. Reactions increased by Aca, GHRKO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Prepare the target reaction set\n",
    "posL = ['Acar-vs-Con1', 'GHRk-vs-GHRw']\n",
    "negL = ['']\n",
    "regulation = 'Increased'\n",
    "tempL = statDF.loc[statDF[variable+'_AdjPval']<0.05].index.tolist()\n",
    "tempDF1 = pvalDF.loc[tempL, pvalDF.columns.str.contains('-vs-')]\n",
    "tempDF2 = diffDF.loc[tempL, diffDF.columns.str.contains('-vs-')]\n",
    "tempS = pd.Series(np.repeat(True, len(tempDF1)), index=tempDF1.index)#Initialize\n",
    "for col_n in tempDF1.columns.tolist():\n",
    "    tempS1 = tempDF1[col_n]\n",
    "    tempS2 = tempDF2[col_n]\n",
    "    if col_n in posL:\n",
    "        if regulation=='Increased':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2>0)\n",
    "        elif regulation=='Decreased':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2<0)\n",
    "    elif col_n in negL:\n",
    "        tempS3 = (tempS1>=0.05)\n",
    "        #Significance for inverse regulation\n",
    "        if regulation=='Increased':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2<0)\n",
    "        elif regulation=='Decreased':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2>0)\n",
    "        tempS1 = tempS3 | tempS1\n",
    "    else:\n",
    "        tempS1 = (tempS1>=0.0)\n",
    "    #Update True\n",
    "    tempS = tempS & tempS1\n",
    "tempL = tempS.loc[tempS.tolist()].index.tolist()\n",
    "print(len(tempL), regulation.lower()+' reactions with significance in', posL, 'but not in', negL)\n",
    "\n",
    "#Select representatives\n",
    "topX = np.min([30, len(tempL)])\n",
    "topX_plot = np.min([5, len(tempL)])\n",
    "tempDF = statDF.loc[:, statDF.columns.str.contains('_AdjPval$')]\n",
    "tempDF = pd.merge(statDF[['ReactionName', variable+'_Pval']], tempDF,\n",
    "                  left_index=True, right_index=True, how='left')\n",
    "tempDF = tempDF.loc[tempL].sort_values(by=variable+'_AdjPval', ascending=True)\n",
    "print('Top', topX, 'reactions (sort by the main effect of '+variable+'):')\n",
    "display(tempDF.iloc[:topX])\n",
    "plotL = tempDF.index.tolist()[:topX_plot]\n",
    "\n",
    "#Prepare DF for plot\n",
    "tempDF = fluxDF.reset_index().melt(var_name='ModelID', value_name='Flux', id_vars='ReactionID')\n",
    "tempDF1 = sampleDF.reset_index()[['ModelID', 'Intervention']]\n",
    "tempDF = pd.merge(tempDF, tempDF1, on='ModelID', how='left')\n",
    "\n",
    "#Prepare label and color\n",
    "tempD0 = {'Con1':'Control 1', 'Acar':'Acarbose',\n",
    "          '17aE':'17'+r'$\\alpha$'+'-Estradiol', 'Prot':'Protandim',\n",
    "          'Rapa':'Rapamycin', 'CalR':'CR diet',\n",
    "          'Con2':'Control 2', 'MetR':'MR diet',\n",
    "          'GHRw':'GHR WT', 'GHRk':'GHR KO',\n",
    "          'SneW':'Snell WT', 'SneD':'Snell dwarf'}\n",
    "tempDF['Group'] = tempDF['Intervention'].map(tempD0)\n",
    "tempD1 = {'Control 1':'tab:blue', 'Acarbose':'tab:red',\n",
    "          '17'+r'$\\alpha$'+'-Estradiol':'tab:green', 'Protandim':'tab:brown',\n",
    "          'Rapamycin':'tab:purple', 'CR diet':'tab:olive',\n",
    "          'Control 2':'tab:blue', 'MR diet':'tab:orange',\n",
    "          'GHR WT':'tab:gray', 'GHR KO':'tab:pink',\n",
    "          'Snell WT':'black', 'Snell dwarf':'tab:cyan'}\n",
    "\n",
    "#Visualize each representative\n",
    "for rank_i in range(len(plotL)):\n",
    "    print(' - Rank '+str(rank_i+1)+' (sort by the main effect of '+variable+'):')\n",
    "    rxn = plotL[rank_i]\n",
    "    #Check reaction summary\n",
    "    tempDF1 = pd.DataFrame(rxnDF.loc[rxn]).T\n",
    "    display(tempDF1)\n",
    "    \n",
    "    #Select Flux\n",
    "    tempDF1 = tempDF.loc[tempDF['ReactionID']==rxn]\n",
    "    \n",
    "    #Check Flux summary\n",
    "    tempDF2 = tempDF1.groupby(['Group'])['Flux'].agg(['count', 'mean', 'std'])\n",
    "    tempL1 = []\n",
    "    tempL2 = []\n",
    "    for row_n in tempDF2.index.tolist():\n",
    "        count, mean, std = tempDF2.loc[row_n]\n",
    "        tempL1.append(mean - 1.96*std/np.sqrt(count))\n",
    "        tempL2.append(mean + 1.96*std/np.sqrt(count))\n",
    "    tempDF2['0.025'] = tempL1\n",
    "    tempDF2['0.975'] = tempL2\n",
    "    tempDF2 = tempDF2.loc[list(tempD1.keys())]#Sort\n",
    "    display(tempDF2)\n",
    "    \n",
    "    #Prepare significance labels\n",
    "    ##Retrieve statistical significance\n",
    "    tempS = pvalDF.loc[rxn, pvalDF.columns.str.contains('-vs-')]\n",
    "    tempS.name = 'AdjPval'\n",
    "    ##Clean\n",
    "    tempDF2 = tempS.index.to_series().str.split(pat='-vs-', expand=True)\n",
    "    tempDF2 = tempDF2.rename(columns={0:'Contrast', 1:'Baseline'})\n",
    "    tempDF2 = pd.merge(tempDF2, tempS, left_index=True, right_index=True, how='left')\n",
    "    tempDF2['Contrast'] = tempDF2['Contrast'].map(tempD0)\n",
    "    tempDF2['Baseline'] = tempDF2['Baseline'].map(tempD0)\n",
    "    ##Convert p-value to label\n",
    "    tempL = []\n",
    "    for row_i in range(len(tempDF2)):\n",
    "        pval = tempDF2['AdjPval'].iloc[row_i]\n",
    "        if pval<0.001:\n",
    "            tempL.append('***')\n",
    "        elif pval<0.01:\n",
    "            tempL.append('**')\n",
    "        elif pval<0.05:\n",
    "            tempL.append('*')\n",
    "        else:\n",
    "            pval_text = Decimal(str(pval)).quantize(Decimal('0.01'), rounding=ROUND_HALF_UP)\n",
    "            tempL.append(r'$P$ = '+str(pval_text))\n",
    "    tempDF2['SignifLabel'] = tempL\n",
    "    ##Add the y-position level in figure\n",
    "    tempDF2['YposLevel'] = [0, 1, 2, 3, 4, 2, 3, 4]\n",
    "    display(tempDF2)\n",
    "    \n",
    "    #Visualization\n",
    "    sns.set(style='ticks', font='Arial', context='talk')\n",
    "    plt.figure(figsize=(4, 4))\n",
    "    sns.boxplot(data=tempDF1, x='Group', y='Flux', order=list(tempD1.keys()), palette=tempD1,\n",
    "                dodge=False, showfliers=False, showcaps=True, notch=False)\n",
    "    p = sns.stripplot(data=tempDF1, x='Group', y='Flux',\n",
    "                      order=list(tempD1.keys()), palette=tempD1, dodge=False, jitter=0.3,\n",
    "                      size=5, edgecolor='black', linewidth=1, **{'marker':'o', 'alpha':0.5})\n",
    "    ##Add border line\n",
    "    for xcoord in [5.5, 7.5, 9.5]:\n",
    "        p.axvline(x=xcoord, **{'linestyle':'dotted', 'color':'black', 'zorder':0})\n",
    "    ##Set axis\n",
    "    sns.despine()\n",
    "    plt.setp(p.get_xticklabels(), rotation=70, horizontalalignment='right',\n",
    "             verticalalignment='center', rotation_mode='anchor')\n",
    "    ###Reset the axis range for the labels\n",
    "    ymin, ymax = p.get_ylim()\n",
    "    tempL = p.get_yticks().tolist()\n",
    "    yinter = int(tempL[1]) - int(tempL[0])\n",
    "    tempL = [int(y) for y in tempL if (y>ymin)&(y<ymax)]#seaborn seems to prepare wider range!?\n",
    "    p.set(ylim=(ymin, ymax+(yinter*2.5)), yticks=tempL)\n",
    "    p.set_yticklabels(['{:,}'.format(int(y)) for y in tempL])\n",
    "    ##Add significance labels\n",
    "    aline_ymin = tempDF1['Flux'].max()\n",
    "    aline_ymargin = yinter/2\n",
    "    for row_i in range(len(tempDF2)):\n",
    "        #Baseline\n",
    "        group_0 = tempDF2['Baseline'].iloc[row_i]\n",
    "        index_0 = list(tempD1.keys()).index(group_0)\n",
    "        xcoord_0 = index_0\n",
    "        #Contrast\n",
    "        group_1 = tempDF2['Contrast'].iloc[row_i]\n",
    "        index_1 = list(tempD1.keys()).index(group_1)\n",
    "        xcoord_1 = index_1\n",
    "        #Standard point of marker\n",
    "        xcoord = (xcoord_0+xcoord_1)/2\n",
    "        ycoord = aline_ymin + aline_ymargin*tempDF2['YposLevel'].iloc[row_i]\n",
    "        label = tempDF2['SignifLabel'].iloc[row_i]\n",
    "        #Add annotation lines\n",
    "        aline_offset = yinter/10\n",
    "        aline_length = yinter/10 + aline_offset\n",
    "        plt.plot([xcoord_0, xcoord_0, xcoord_1, xcoord_1],\n",
    "                 [ycoord+aline_offset, ycoord+aline_length, ycoord+aline_length, ycoord+aline_offset],\n",
    "                 lw=1.5, c='k')\n",
    "        #Add annotation text\n",
    "        if label in ['***', '**', '*']:\n",
    "            text_offset = yinter/4\n",
    "            p.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                       horizontalalignment='center', verticalalignment='center',\n",
    "                       fontsize='medium', color='k')\n",
    "        else:\n",
    "            text_offset = yinter/5\n",
    "            p.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                       horizontalalignment='center', verticalalignment='bottom',\n",
    "                       fontsize='x-small', color='k')\n",
    "    ##Set axis label and title\n",
    "    plt.setp(p, xlabel='', ylabel='Sample flux value [a.u.]')\n",
    "    p.set_title(rxn, {'fontsize':'medium'})\n",
    "    ##Save\n",
    "    fileDir = './ExportFigures/'\n",
    "    ipynbName = '220610_LC-M001-related-transcriptomics-GEM-ver2-4_FluxAnalysis_'\n",
    "    fileName = 'flux-boxplot('+rxn+').tif'\n",
    "    plt.gcf().savefig(fileDir+ipynbName+fileName, dpi=300, bbox_inches='tight', pad_inches=0.04,\n",
    "                      pil_kwargs={'compression':'tiff_lzw'})\n",
    "    plt.show()\n",
    "    print('')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 3-4-4. Reactions decreased by Aca, GHRKO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Prepare the target reaction set\n",
    "posL = ['Acar-vs-Con1', 'GHRk-vs-GHRw']\n",
    "negL = ['']\n",
    "regulation = 'Decreased'\n",
    "tempL = statDF.loc[statDF[variable+'_AdjPval']<0.05].index.tolist()\n",
    "tempDF1 = pvalDF.loc[tempL, pvalDF.columns.str.contains('-vs-')]\n",
    "tempDF2 = diffDF.loc[tempL, diffDF.columns.str.contains('-vs-')]\n",
    "tempS = pd.Series(np.repeat(True, len(tempDF1)), index=tempDF1.index)#Initialize\n",
    "for col_n in tempDF1.columns.tolist():\n",
    "    tempS1 = tempDF1[col_n]\n",
    "    tempS2 = tempDF2[col_n]\n",
    "    if col_n in posL:\n",
    "        if regulation=='Increased':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2>0)\n",
    "        elif regulation=='Decreased':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2<0)\n",
    "    elif col_n in negL:\n",
    "        tempS3 = (tempS1>=0.05)\n",
    "        #Significance for inverse regulation\n",
    "        if regulation=='Increased':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2<0)\n",
    "        elif regulation=='Decreased':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2>0)\n",
    "        tempS1 = tempS3 | tempS1\n",
    "    else:\n",
    "        tempS1 = (tempS1>=0.0)\n",
    "    #Update True\n",
    "    tempS = tempS & tempS1\n",
    "tempL = tempS.loc[tempS.tolist()].index.tolist()\n",
    "print(len(tempL), regulation.lower()+' reactions with significance in', posL, 'but not in', negL)\n",
    "\n",
    "#Select representatives\n",
    "topX = np.min([30, len(tempL)])\n",
    "topX_plot = np.min([5, len(tempL)])\n",
    "tempDF = statDF.loc[:, statDF.columns.str.contains('_AdjPval$')]\n",
    "tempDF = pd.merge(statDF[['ReactionName', variable+'_Pval']], tempDF,\n",
    "                  left_index=True, right_index=True, how='left')\n",
    "tempDF = tempDF.loc[tempL].sort_values(by=variable+'_AdjPval', ascending=True)\n",
    "print('Top', topX, 'reactions (sort by the main effect of '+variable+'):')\n",
    "display(tempDF.iloc[:topX])\n",
    "plotL = tempDF.index.tolist()[:topX_plot]\n",
    "\n",
    "#Prepare DF for plot\n",
    "tempDF = fluxDF.reset_index().melt(var_name='ModelID', value_name='Flux', id_vars='ReactionID')\n",
    "tempDF1 = sampleDF.reset_index()[['ModelID', 'Intervention']]\n",
    "tempDF = pd.merge(tempDF, tempDF1, on='ModelID', how='left')\n",
    "\n",
    "#Prepare label and color\n",
    "tempD0 = {'Con1':'Control 1', 'Acar':'Acarbose',\n",
    "          '17aE':'17'+r'$\\alpha$'+'-Estradiol', 'Prot':'Protandim',\n",
    "          'Rapa':'Rapamycin', 'CalR':'CR diet',\n",
    "          'Con2':'Control 2', 'MetR':'MR diet',\n",
    "          'GHRw':'GHR WT', 'GHRk':'GHR KO',\n",
    "          'SneW':'Snell WT', 'SneD':'Snell dwarf'}\n",
    "tempDF['Group'] = tempDF['Intervention'].map(tempD0)\n",
    "tempD1 = {'Control 1':'tab:blue', 'Acarbose':'tab:red',\n",
    "          '17'+r'$\\alpha$'+'-Estradiol':'tab:green', 'Protandim':'tab:brown',\n",
    "          'Rapamycin':'tab:purple', 'CR diet':'tab:olive',\n",
    "          'Control 2':'tab:blue', 'MR diet':'tab:orange',\n",
    "          'GHR WT':'tab:gray', 'GHR KO':'tab:pink',\n",
    "          'Snell WT':'black', 'Snell dwarf':'tab:cyan'}\n",
    "\n",
    "#Visualize each representative\n",
    "for rank_i in range(len(plotL)):\n",
    "    print(' - Rank '+str(rank_i+1)+' (sort by the main effect of '+variable+'):')\n",
    "    rxn = plotL[rank_i]\n",
    "    #Check reaction summary\n",
    "    tempDF1 = pd.DataFrame(rxnDF.loc[rxn]).T\n",
    "    display(tempDF1)\n",
    "    \n",
    "    #Select Flux\n",
    "    tempDF1 = tempDF.loc[tempDF['ReactionID']==rxn]\n",
    "    \n",
    "    #Check Flux summary\n",
    "    tempDF2 = tempDF1.groupby(['Group'])['Flux'].agg(['count', 'mean', 'std'])\n",
    "    tempL1 = []\n",
    "    tempL2 = []\n",
    "    for row_n in tempDF2.index.tolist():\n",
    "        count, mean, std = tempDF2.loc[row_n]\n",
    "        tempL1.append(mean - 1.96*std/np.sqrt(count))\n",
    "        tempL2.append(mean + 1.96*std/np.sqrt(count))\n",
    "    tempDF2['0.025'] = tempL1\n",
    "    tempDF2['0.975'] = tempL2\n",
    "    tempDF2 = tempDF2.loc[list(tempD1.keys())]#Sort\n",
    "    display(tempDF2)\n",
    "    \n",
    "    #Prepare significance labels\n",
    "    ##Retrieve statistical significance\n",
    "    tempS = pvalDF.loc[rxn, pvalDF.columns.str.contains('-vs-')]\n",
    "    tempS.name = 'AdjPval'\n",
    "    ##Clean\n",
    "    tempDF2 = tempS.index.to_series().str.split(pat='-vs-', expand=True)\n",
    "    tempDF2 = tempDF2.rename(columns={0:'Contrast', 1:'Baseline'})\n",
    "    tempDF2 = pd.merge(tempDF2, tempS, left_index=True, right_index=True, how='left')\n",
    "    tempDF2['Contrast'] = tempDF2['Contrast'].map(tempD0)\n",
    "    tempDF2['Baseline'] = tempDF2['Baseline'].map(tempD0)\n",
    "    ##Convert p-value to label\n",
    "    tempL = []\n",
    "    for row_i in range(len(tempDF2)):\n",
    "        pval = tempDF2['AdjPval'].iloc[row_i]\n",
    "        if pval<0.001:\n",
    "            tempL.append('***')\n",
    "        elif pval<0.01:\n",
    "            tempL.append('**')\n",
    "        elif pval<0.05:\n",
    "            tempL.append('*')\n",
    "        else:\n",
    "            pval_text = Decimal(str(pval)).quantize(Decimal('0.01'), rounding=ROUND_HALF_UP)\n",
    "            tempL.append(r'$P$ = '+str(pval_text))\n",
    "    tempDF2['SignifLabel'] = tempL\n",
    "    ##Add the y-position level in figure\n",
    "    tempDF2['YposLevel'] = [0, 1, 2, 3, 4, 2, 3, 4]\n",
    "    display(tempDF2)\n",
    "    \n",
    "    #Visualization\n",
    "    sns.set(style='ticks', font='Arial', context='talk')\n",
    "    plt.figure(figsize=(4, 4))\n",
    "    sns.boxplot(data=tempDF1, x='Group', y='Flux', order=list(tempD1.keys()), palette=tempD1,\n",
    "                dodge=False, showfliers=False, showcaps=True, notch=False)\n",
    "    p = sns.stripplot(data=tempDF1, x='Group', y='Flux',\n",
    "                      order=list(tempD1.keys()), palette=tempD1, dodge=False, jitter=0.3,\n",
    "                      size=5, edgecolor='black', linewidth=1, **{'marker':'o', 'alpha':0.5})\n",
    "    ##Add border line\n",
    "    for xcoord in [5.5, 7.5, 9.5]:\n",
    "        p.axvline(x=xcoord, **{'linestyle':'dotted', 'color':'black', 'zorder':0})\n",
    "    ##Set axis\n",
    "    sns.despine()\n",
    "    plt.setp(p.get_xticklabels(), rotation=70, horizontalalignment='right',\n",
    "             verticalalignment='center', rotation_mode='anchor')\n",
    "    ###Reset the axis range for the labels\n",
    "    ymin, ymax = p.get_ylim()\n",
    "    tempL = p.get_yticks().tolist()\n",
    "    yinter = int(tempL[1]) - int(tempL[0])\n",
    "    tempL = [int(y) for y in tempL if (y>ymin)&(y<ymax)]#seaborn seems to prepare wider range!?\n",
    "    p.set(ylim=(ymin, ymax+(yinter*2.5)), yticks=tempL)\n",
    "    p.set_yticklabels(['{:,}'.format(int(y)) for y in tempL])\n",
    "    ##Add significance labels\n",
    "    aline_ymin = tempDF1['Flux'].max()\n",
    "    aline_ymargin = yinter/2\n",
    "    for row_i in range(len(tempDF2)):\n",
    "        #Baseline\n",
    "        group_0 = tempDF2['Baseline'].iloc[row_i]\n",
    "        index_0 = list(tempD1.keys()).index(group_0)\n",
    "        xcoord_0 = index_0\n",
    "        #Contrast\n",
    "        group_1 = tempDF2['Contrast'].iloc[row_i]\n",
    "        index_1 = list(tempD1.keys()).index(group_1)\n",
    "        xcoord_1 = index_1\n",
    "        #Standard point of marker\n",
    "        xcoord = (xcoord_0+xcoord_1)/2\n",
    "        ycoord = aline_ymin + aline_ymargin*tempDF2['YposLevel'].iloc[row_i]\n",
    "        label = tempDF2['SignifLabel'].iloc[row_i]\n",
    "        #Add annotation lines\n",
    "        aline_offset = yinter/10\n",
    "        aline_length = yinter/10 + aline_offset\n",
    "        plt.plot([xcoord_0, xcoord_0, xcoord_1, xcoord_1],\n",
    "                 [ycoord+aline_offset, ycoord+aline_length, ycoord+aline_length, ycoord+aline_offset],\n",
    "                 lw=1.5, c='k')\n",
    "        #Add annotation text\n",
    "        if label in ['***', '**', '*']:\n",
    "            text_offset = yinter/4\n",
    "            p.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                       horizontalalignment='center', verticalalignment='center',\n",
    "                       fontsize='medium', color='k')\n",
    "        else:\n",
    "            text_offset = yinter/5\n",
    "            p.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                       horizontalalignment='center', verticalalignment='bottom',\n",
    "                       fontsize='x-small', color='k')\n",
    "    ##Set axis label and title\n",
    "    plt.setp(p, xlabel='', ylabel='Sample flux value [a.u.]')\n",
    "    p.set_title(rxn, {'fontsize':'medium'})\n",
    "    ##Save\n",
    "    fileDir = './ExportFigures/'\n",
    "    ipynbName = '220610_LC-M001-related-transcriptomics-GEM-ver2-4_FluxAnalysis_'\n",
    "    fileName = 'flux-boxplot('+rxn+').tif'\n",
    "    plt.gcf().savefig(fileDir+ipynbName+fileName, dpi=300, bbox_inches='tight', pad_inches=0.04,\n",
    "                      pil_kwargs={'compression':'tiff_lzw'})\n",
    "    plt.show()\n",
    "    print('')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 3-4-5. Reactions of special interest"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> 25 reactions included in the Priyanka's diagram, related to fatty acid metabolism etc. Of note, ACS would not be used for any represetnation because she could not map ACS to the diagram.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plotL = ['ACITL', 'ACS', 'ACYP', 'ALDD21', 'CBPPer',\n",
    "         'CRNtx', 'DPGase', 'DPGM', 'FACOAL140', 'FACOAL180',\n",
    "         'FACOAL184', 'FACOAL191', 'FAOXC80', 'FAOXC140', 'FAOXC160',\n",
    "         'FBP', 'HEX1', 'MGSA', 'MGSA2', 'PGK',\n",
    "         'PGM', 'PRPPS', 'STRDNCCPT2', 'SUCOAS1m', 'TKT1',\n",
    "         'TKT2']\n",
    "plotL = list(set(plotL))\n",
    "plotL.sort()\n",
    "for rxn in plotL:\n",
    "    if rxn not in fluxDF.index.tolist():\n",
    "        print(rxn+' was NOT included?')\n",
    "        plotL.remove(rxn)\n",
    "tempDF = statDF.loc[:, statDF.columns.str.contains('_AdjPval$')]\n",
    "tempDF = pd.merge(statDF[['ReactionName', variable+'_Pval']], tempDF,\n",
    "                  left_index=True, right_index=True, how='left')\n",
    "tempDF = tempDF.loc[plotL].sort_values(by=variable+'_AdjPval', ascending=True)\n",
    "print(len(plotL), 'target reactions:')\n",
    "print(' -> '+variable+' (adjusted P < 0.05):', len(tempDF.loc[tempDF[variable+'_AdjPval']<0.05]))\n",
    "print(' -> '+variable+' (nominal P < 0.05):', len(tempDF.loc[tempDF[variable+'_Pval']<0.05]))\n",
    "display(tempDF)\n",
    "plotL = tempDF.index.tolist()#Update by the sorted order\n",
    "\n",
    "#Prepare DF for plot\n",
    "tempDF = fluxDF.reset_index().melt(var_name='ModelID', value_name='Flux', id_vars='ReactionID')\n",
    "tempDF1 = sampleDF.reset_index()[['ModelID', 'Intervention']]\n",
    "tempDF = pd.merge(tempDF, tempDF1, on='ModelID', how='left')\n",
    "\n",
    "#Prepare label and color\n",
    "tempD0 = {'Con1':'Control 1', 'Acar':'Acarbose',\n",
    "          '17aE':'17'+r'$\\alpha$'+'-Estradiol', 'Prot':'Protandim',\n",
    "          'Rapa':'Rapamycin', 'CalR':'CR diet',\n",
    "          'Con2':'Control 2', 'MetR':'MR diet',\n",
    "          'GHRw':'GHR WT', 'GHRk':'GHR KO',\n",
    "          'SneW':'Snell WT', 'SneD':'Snell dwarf'}\n",
    "tempDF['Group'] = tempDF['Intervention'].map(tempD0)\n",
    "tempD1 = {'Control 1':'tab:blue', 'Acarbose':'tab:red',\n",
    "          '17'+r'$\\alpha$'+'-Estradiol':'tab:green', 'Protandim':'tab:brown',\n",
    "          'Rapamycin':'tab:purple', 'CR diet':'tab:olive',\n",
    "          'Control 2':'tab:blue', 'MR diet':'tab:orange',\n",
    "          'GHR WT':'tab:gray', 'GHR KO':'tab:pink',\n",
    "          'Snell WT':'black', 'Snell dwarf':'tab:cyan'}\n",
    "\n",
    "#Visualize each representative\n",
    "for rank_i in range(len(plotL)):\n",
    "    print(' - Rank '+str(rank_i+1)+' (sort by the main effect of '+variable+'):')\n",
    "    rxn = plotL[rank_i]\n",
    "    #Check reaction summary\n",
    "    tempDF1 = pd.DataFrame(rxnDF.loc[rxn]).T\n",
    "    display(tempDF1)\n",
    "    \n",
    "    #Select Flux\n",
    "    tempDF1 = tempDF.loc[tempDF['ReactionID']==rxn]\n",
    "    \n",
    "    #Check Flux summary\n",
    "    tempDF2 = tempDF1.groupby(['Group'])['Flux'].agg(['count', 'mean', 'std'])\n",
    "    tempL1 = []\n",
    "    tempL2 = []\n",
    "    for row_n in tempDF2.index.tolist():\n",
    "        count, mean, std = tempDF2.loc[row_n]\n",
    "        tempL1.append(mean - 1.96*std/np.sqrt(count))\n",
    "        tempL2.append(mean + 1.96*std/np.sqrt(count))\n",
    "    tempDF2['0.025'] = tempL1\n",
    "    tempDF2['0.975'] = tempL2\n",
    "    tempDF2 = tempDF2.loc[list(tempD1.keys())]#Sort\n",
    "    display(tempDF2)\n",
    "    \n",
    "    #Prepare significance labels\n",
    "    ##Retrieve statistical significance\n",
    "    tempS = pvalDF.loc[rxn, pvalDF.columns.str.contains('-vs-')]\n",
    "    tempS.name = 'AdjPval'\n",
    "    ##Clean\n",
    "    tempDF2 = tempS.index.to_series().str.split(pat='-vs-', expand=True)\n",
    "    tempDF2 = tempDF2.rename(columns={0:'Contrast', 1:'Baseline'})\n",
    "    tempDF2 = pd.merge(tempDF2, tempS, left_index=True, right_index=True, how='left')\n",
    "    tempDF2['Contrast'] = tempDF2['Contrast'].map(tempD0)\n",
    "    tempDF2['Baseline'] = tempDF2['Baseline'].map(tempD0)\n",
    "    ##Convert p-value to label\n",
    "    tempL = []\n",
    "    for row_i in range(len(tempDF2)):\n",
    "        pval = tempDF2['AdjPval'].iloc[row_i]\n",
    "        if pval<0.001:\n",
    "            tempL.append('***')\n",
    "        elif pval<0.01:\n",
    "            tempL.append('**')\n",
    "        elif pval<0.05:\n",
    "            tempL.append('*')\n",
    "        else:\n",
    "            pval_text = Decimal(str(pval)).quantize(Decimal('0.01'), rounding=ROUND_HALF_UP)\n",
    "            tempL.append(r'$P$ = '+str(pval_text))\n",
    "    tempDF2['SignifLabel'] = tempL\n",
    "    ##Add the y-position level in figure\n",
    "    tempDF2['YposLevel'] = [0, 1, 2, 3, 4, 2, 3, 4]\n",
    "    display(tempDF2)\n",
    "    \n",
    "    #Visualization\n",
    "    sns.set(style='ticks', font='Arial', context='talk')\n",
    "    plt.figure(figsize=(4, 4))\n",
    "    sns.boxplot(data=tempDF1, x='Group', y='Flux', order=list(tempD1.keys()), palette=tempD1,\n",
    "                dodge=False, showfliers=False, showcaps=True, notch=False)\n",
    "    p = sns.stripplot(data=tempDF1, x='Group', y='Flux',\n",
    "                      order=list(tempD1.keys()), palette=tempD1, dodge=False, jitter=0.3,\n",
    "                      size=5, edgecolor='black', linewidth=1, **{'marker':'o', 'alpha':0.5})\n",
    "    ##Add border line\n",
    "    for xcoord in [5.5, 7.5, 9.5]:\n",
    "        p.axvline(x=xcoord, **{'linestyle':'dotted', 'color':'black', 'zorder':0})\n",
    "    ##Set axis\n",
    "    sns.despine()\n",
    "    plt.setp(p.get_xticklabels(), rotation=70, horizontalalignment='right',\n",
    "             verticalalignment='center', rotation_mode='anchor')\n",
    "    ###Reset the axis range for the labels\n",
    "    ymin, ymax = p.get_ylim()\n",
    "    tempL = p.get_yticks().tolist()\n",
    "    yinter = int(tempL[1]) - int(tempL[0])\n",
    "    tempL = [int(y) for y in tempL if (y>ymin)&(y<ymax)]#seaborn seems to prepare wider range!?\n",
    "    p.set(ylim=(ymin, ymax+(yinter*2.5)), yticks=tempL)\n",
    "    p.set_yticklabels(['{:,}'.format(int(y)) for y in tempL])\n",
    "    ##Add significance labels\n",
    "    aline_ymin = tempDF1['Flux'].max()\n",
    "    aline_ymargin = yinter/2\n",
    "    for row_i in range(len(tempDF2)):\n",
    "        #Baseline\n",
    "        group_0 = tempDF2['Baseline'].iloc[row_i]\n",
    "        index_0 = list(tempD1.keys()).index(group_0)\n",
    "        xcoord_0 = index_0\n",
    "        #Contrast\n",
    "        group_1 = tempDF2['Contrast'].iloc[row_i]\n",
    "        index_1 = list(tempD1.keys()).index(group_1)\n",
    "        xcoord_1 = index_1\n",
    "        #Standard point of marker\n",
    "        xcoord = (xcoord_0+xcoord_1)/2\n",
    "        ycoord = aline_ymin + aline_ymargin*tempDF2['YposLevel'].iloc[row_i]\n",
    "        label = tempDF2['SignifLabel'].iloc[row_i]\n",
    "        #Add annotation lines\n",
    "        aline_offset = yinter/10\n",
    "        aline_length = yinter/10 + aline_offset\n",
    "        plt.plot([xcoord_0, xcoord_0, xcoord_1, xcoord_1],\n",
    "                 [ycoord+aline_offset, ycoord+aline_length, ycoord+aline_length, ycoord+aline_offset],\n",
    "                 lw=1.5, c='k')\n",
    "        #Add annotation text\n",
    "        if label in ['***', '**', '*']:\n",
    "            text_offset = yinter/4\n",
    "            p.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                       horizontalalignment='center', verticalalignment='center',\n",
    "                       fontsize='medium', color='k')\n",
    "        else:\n",
    "            text_offset = yinter/5\n",
    "            p.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                       horizontalalignment='center', verticalalignment='bottom',\n",
    "                       fontsize='x-small', color='k')\n",
    "    ##Set axis label and title\n",
    "    plt.setp(p, xlabel='', ylabel='Sample flux value [a.u.]')\n",
    "    p.set_title(rxn, {'fontsize':'medium'})\n",
    "    ##Save\n",
    "    fileDir = './ExportFigures/'\n",
    "    ipynbName = '220610_LC-M001-related-transcriptomics-GEM-ver2-4_FluxAnalysis_'\n",
    "    fileName = 'flux-boxplot('+rxn+').tif'\n",
    "    plt.gcf().savefig(fileDir+ipynbName+fileName, dpi=300, bbox_inches='tight', pad_inches=0.04,\n",
    "                      pil_kwargs={'compression':'tiff_lzw'})\n",
    "    plt.show()\n",
    "    print('')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4. Compare flux values between MR and CR"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> Model is same as the above Dunn's test, but post-hoc comparisons are differently retrieved to compare MR vs. CR.  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# — †2. Go to †1 of the R sub-notebook —  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 4-1. Another post-hoc Dunn's test"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 4-1-1/2. Import the summary tables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Import the summary tables\n",
    "fileDir = './ExportData/'\n",
    "ipynbName = '220610_LC-M001-related-transcriptomics-GEM-ver2-4_StatisticalTest_'\n",
    "fileName = 'flux-comparison.xlsx'\n",
    "sheetName = 'Posthoc2_MR-vs-CR'\n",
    "tempDF = pd.read_excel(fileDir+ipynbName+fileName, sheet_name=sheetName, engine='openpyxl')\n",
    "tempDF = tempDF.set_index('ReactionID')\n",
    "display(tempDF)\n",
    "\n",
    "statDF2 = tempDF"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 4-1-3. Changed reactions by each intervention (Dunn's test)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Clean DF for adjusted P-value and z-statistic\n",
    "variable = 'Intervention'\n",
    "##Take adjusted P-value\n",
    "tempDF1 = statDF2.loc[:, statDF2.columns.str.contains('-vs-.*_AdjPval$')]\n",
    "tempDF1.columns = tempDF1.columns.str.replace('_AdjPval$', '')\n",
    "tempDF1 = pd.merge(statDF[['ReactionName', variable+'_AdjPval']], tempDF1,\n",
    "                   left_index=True, right_index=True, how='left')\n",
    "tempDF1 = tempDF1.sort_values(by=variable+'_AdjPval', ascending=True)\n",
    "print('Adjusted P-value:')\n",
    "display(tempDF1)\n",
    "##Take z-statistic\n",
    "tempDF2 = statDF2.loc[:, statDF2.columns.str.contains('-vs-.*_zStat$')]\n",
    "tempDF2.columns = tempDF2.columns.str.replace('_zStat$', '')\n",
    "tempDF2 = pd.merge(statDF[['ReactionName', variable+'_AdjPval']], tempDF2,\n",
    "                   left_index=True, right_index=True, how='left')\n",
    "tempDF2 = tempDF2.sort_values(by=variable+'_AdjPval', ascending=True)\n",
    "print('Changed direction (z-statistic):')\n",
    "display(tempDF2)\n",
    "\n",
    "pvalDF2 = tempDF1\n",
    "diffDF2 = tempDF2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> –> When z-statistic is positive, the flux median of contrast is larger than that of baseline.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Extract only the changed reactions\n",
    "variable = 'Intervention'\n",
    "tempDF = pvalDF2.loc[pvalDF2[variable+'_AdjPval']<0.05]\n",
    "print(variable+' (adjusted P < 0.05):', len(tempDF))\n",
    "\n",
    "tempL = tempDF.loc[:, tempDF.columns.str.contains('-vs-')].columns.tolist()\n",
    "for comparison in tempL:\n",
    "    tempS = tempDF[comparison]\n",
    "    tempS = tempS.loc[tempS<0.05]\n",
    "    print(' - '+comparison+':', len(tempS))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> Check the changed reactions (nominal P-value < 0.05) as reference.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Extract only the changed reactions\n",
    "variable = 'Intervention'\n",
    "tempL = statDF.loc[statDF[variable+'_Pval']<0.05].index.tolist()\n",
    "tempDF = pvalDF2.loc[tempL]\n",
    "print(variable+' (nominal P < 0.05):', len(tempDF))\n",
    "\n",
    "tempL = tempDF.loc[:, tempDF.columns.str.contains('-vs-')].columns.tolist()\n",
    "for comparison in tempL:\n",
    "    tempS = tempDF[comparison]\n",
    "    tempS = tempS.loc[tempS<0.05]\n",
    "    print(' - '+comparison+':', len(tempS))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 4-2. Visualization: venn diagram"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Prepare label and color\n",
    "tempD0 = {'Con1':'Control 1', 'Acar':'Acarbose',\n",
    "          '17aE':'17'+r'$\\alpha$'+'-Estradiol', 'Prot':'Protandim',\n",
    "          'Rapa':'Rapamycin', 'CalR':'CR diet',\n",
    "          'Con2':'Control 2', 'MetR':'MR diet',\n",
    "          'GHRw':'GHR WT', 'GHRk':'GHR KO',\n",
    "          'SneW':'Snell WT', 'SneD':'Snell dwarf'}\n",
    "tempD1 = {'CR diet\\nvs.\\nControl 1':'tab:olive',\n",
    "          'MR diet\\nvs.\\nControl 2':'tab:orange',\n",
    "          'MR diet\\nvs.\\nCR diet':'turquoise'}\n",
    "\n",
    "#Visualization per direction\n",
    "for regulation in ['Changed', 'Increased', 'Decreased']:\n",
    "    #Prepare reaction sets\n",
    "    variable = 'Intervention'\n",
    "    tempL = statDF.loc[statDF[variable+'_AdjPval']<0.05].index.tolist()\n",
    "    tempDF1 = pvalDF2.loc[tempL, pvalDF2.columns.str.contains('-vs-')]\n",
    "    tempDF2 = diffDF2.loc[tempL, diffDF2.columns.str.contains('-vs-')]\n",
    "    tempD2 = {}\n",
    "    count = 0\n",
    "    for col_n in tempDF1.columns.tolist():\n",
    "        tempS1 = tempDF1[col_n]\n",
    "        tempS2 = tempDF2[col_n]\n",
    "        if regulation=='Changed':\n",
    "            tempS2 = tempS2.loc[(tempS1<0.05)]\n",
    "        elif regulation=='Increased':\n",
    "            tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2>0)]\n",
    "        elif regulation=='Decreased':\n",
    "            tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2<0)]\n",
    "        if len(tempS2)>0:\n",
    "            count += 1\n",
    "        contrast, baseline = col_n.split(sep='-vs-')\n",
    "        label = tempD0[contrast]+'\\nvs.\\n'+tempD0[baseline]\n",
    "        tempD2[label] = set(tempS2.index.tolist())\n",
    "    if count==0:\n",
    "        print(regulation+' reactions: no significance in any comparisons')\n",
    "        continue\n",
    "    ##Sort to make consistent order in manual legend generation\n",
    "    tempD = {}\n",
    "    count = 0\n",
    "    for label in tempD1.keys():\n",
    "        tempD[label] = tempD2[label]\n",
    "        if len(tempD2[label])>0:\n",
    "            count += 1\n",
    "    \n",
    "    #Skip the followings if no significant reaction\n",
    "    if count==0:\n",
    "        print(regulation+' reactions: no significance in any target comparisons')\n",
    "        continue\n",
    "    \n",
    "    #Venn diagram\n",
    "    sns.set(style='ticks', font='Arial', context='talk')\n",
    "    fig, ax = plt.subplots(figsize=(4, 4))\n",
    "    venn(tempD, fmt='{size:,}', cmap=list(tempD1.values()), legend_loc=None, ax=ax)\n",
    "    plt.setp(ax, ylim=(0.05, 0.975))#Otherwise, weird space...\n",
    "    ##Add legend annotation\n",
    "    x_coord = [0.05, 0.95, 0.85]\n",
    "    y_coord = [0.65, 0.65, 0.4]\n",
    "    h_align = ['right', 'left', 'left']\n",
    "    v_align = ['bottom', 'bottom', 'top']\n",
    "    for i in range(len(tempD1)):\n",
    "        key = list(tempD1.keys())[i]\n",
    "        total = f'{len(tempD[key]):,}'\n",
    "        ax.text(x_coord[i], y_coord[i], key+'\\n('+total+' reactions)',\n",
    "                fontsize='small', multialignment='center',\n",
    "                horizontalalignment=h_align[i], verticalalignment=v_align[i],\n",
    "                bbox={'boxstyle':'round', 'facecolor':tempD1[key], 'pad':0.2, 'alpha':0.5})\n",
    "    title = regulation+' reactions'\n",
    "    ax.set_title(title, fontsize='medium')\n",
    "    ##Save\n",
    "    if regulation!='Changed':\n",
    "        fileDir = './ExportFigures/'\n",
    "        ipynbName = '220610_LC-M001-related-transcriptomics-GEM-ver2-4_FluxAnalysis_'\n",
    "        fileName = 'MR-vs-CR-venn-'+regulation.lower()+'.tif'\n",
    "        plt.gcf().savefig(fileDir+ipynbName+fileName, dpi=300, bbox_inches='tight', pad_inches=0.04,\n",
    "                          pil_kwargs={'compression':'tiff_lzw'})\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Export reaction list in each subset in the venn diagram\n",
    "for regulation in ['Increased', 'Decreased']:\n",
    "    #Prepare reaction sets\n",
    "    variable = 'Intervention'\n",
    "    tempL1 = statDF.loc[statDF[variable+'_AdjPval']<0.05].index.tolist()\n",
    "    tempL2 = pvalDF2.loc[:, pvalDF2.columns.str.contains('-vs-')].columns.tolist()\n",
    "    tempDF1 = pvalDF2.loc[tempL1, tempL2]\n",
    "    tempDF2 = diffDF2.loc[tempL1, tempL2]\n",
    "    tempD = {}\n",
    "    count = 0\n",
    "    for col_n in tempDF1.columns.tolist():\n",
    "        tempS1 = tempDF1[col_n]\n",
    "        tempS2 = tempDF2[col_n]\n",
    "        if regulation=='Changed':\n",
    "            tempS2 = tempS2.loc[(tempS1<0.05)]\n",
    "        elif regulation=='Increased':\n",
    "            tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2>0)]\n",
    "        elif regulation=='Decreased':\n",
    "            tempS2 = tempS2.loc[(tempS1<0.05)&(tempS2<0)]\n",
    "        if len(tempS2)>0:\n",
    "            count += 1\n",
    "        tempD[col_n] = set(tempS2.index.tolist())\n",
    "    \n",
    "    #Skip the followings if no significant reaction\n",
    "    if count==0:\n",
    "        print(regulation+' reactions: no significance in any target comparisons')\n",
    "        continue\n",
    "    \n",
    "    #Prepare a new .xlsx file (dummy README)\n",
    "    tempL1 = [len(tempD[key]) for key in tempD.keys()]\n",
    "    tempDF = pd.DataFrame({'Group':tempD.keys(), 'nReactions':tempL1})\n",
    "    tempDF = tempDF.reset_index().rename(columns={'index':'VennOrder'})\n",
    "    tempDF['VennOrder'] = tempDF['VennOrder'] + 1\n",
    "    fileDir = './ExportData/'\n",
    "    ipynbName = '220610_LC-M001-related-transcriptomics-GEM-ver2-4_FluxAnalysis_'\n",
    "    fileName = 'MR-vs-CR-venn-'+regulation.lower()+'.xlsx'\n",
    "    tempDF.to_excel(fileDir+ipynbName+fileName, sheet_name='README', header=True, index=False)\n",
    "    display(tempDF)#Check\n",
    "    \n",
    "    t_start = time.time()\n",
    "    #Extract overall set\n",
    "    for key_i in range(len(tempD)):\n",
    "        key = list(tempD.keys())[key_i]\n",
    "        tempS = tempD[key]\n",
    "        tempDF = statDF2.loc[statDF2.index.isin(tempS)]\n",
    "        #Save summary table by appended to the above .xlsx file\n",
    "        ##Prepare sheet name\n",
    "        tempL1 = ['NA' for i in range(len(tempD))]\n",
    "        tempL1[key_i] = '1'\n",
    "        setName = '('+','.join(tempL1)+')'\n",
    "        with pd.ExcelWriter(fileDir+ipynbName+fileName, mode='a', engine='openpyxl') as writer:\n",
    "            tempDF.to_excel(writer, sheet_name=setName, header=True, index=True)\n",
    "        print(' - '+setName+':', len(tempDF))\n",
    "    \n",
    "    #Extract subset\n",
    "    tempL1 = ['1', '0']\n",
    "    tempL2 = [[k1, k2, k3] for k1 in tempL1 for k2 in tempL1 for k3 in tempL1]\n",
    "    #tempL2.remove(['0', '0', '0'])\n",
    "    for tempL1 in tempL2:\n",
    "        #Positive reaction set\n",
    "        tempL3 = [list(tempD.values())[key_i] for key_i, binary in enumerate(tempL1) if binary=='1']\n",
    "        tempS1 = set(statDF.loc[statDF[variable+'_AdjPval']<0.05].index.tolist())#Initialize\n",
    "        for tempS in tempL3:\n",
    "            tempS1 = tempS1 & tempS\n",
    "        #Negative reaction set\n",
    "        tempL3 = [list(tempD.values())[key_i] for key_i, binary in enumerate(tempL1) if binary=='0']\n",
    "        tempS2 = set()#Initialize\n",
    "        for tempS in tempL3:\n",
    "            tempS2 = tempS2 | tempS\n",
    "        #Extract subset\n",
    "        tempS = tempS1 - tempS2\n",
    "        tempDF = statDF2.loc[statDF2.index.isin(tempS)]\n",
    "        #Save summary table by appended to the above .xlsx file\n",
    "        ##Prepare sheet name\n",
    "        setName = '('+','.join(tempL1)+')'\n",
    "        with pd.ExcelWriter(fileDir+ipynbName+fileName, mode='a', engine='openpyxl') as writer:\n",
    "            tempDF.to_excel(writer, sheet_name=setName, header=True, index=True)\n",
    "        print(' - '+setName+':', len(tempDF))\n",
    "    \n",
    "    t_elapsed = time.time() - t_start\n",
    "    print(' - Elapsed time:', round(t_elapsed//60), 'min', round(t_elapsed%60, 1), 'sec')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 4-3. Visualization: boxplot"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 4-3-1. Reactions increased by MR but not CR"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Prepare the target reaction set\n",
    "posL = ['MetR-vs-Con2', 'MetR-vs-CalR']\n",
    "negL = ['CalR-vs-Con1']\n",
    "regulation = 'Increased'\n",
    "tempL = statDF.loc[statDF[variable+'_AdjPval']<0.05].index.tolist()\n",
    "tempDF1 = pvalDF2.loc[tempL, pvalDF2.columns.str.contains('-vs-')]\n",
    "tempDF2 = diffDF2.loc[tempL, diffDF2.columns.str.contains('-vs-')]\n",
    "tempS = pd.Series(np.repeat(True, len(tempDF1)), index=tempDF1.index)#Initialize\n",
    "for col_n in tempDF1.columns.tolist():\n",
    "    tempS1 = tempDF1[col_n]\n",
    "    tempS2 = tempDF2[col_n]\n",
    "    if col_n in posL:\n",
    "        if regulation=='Increased':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2>0)\n",
    "        elif regulation=='Decreased':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2<0)\n",
    "    elif col_n in negL:\n",
    "        tempS3 = (tempS1>=0.05)\n",
    "        #Significance for inverse regulation\n",
    "        if regulation=='Increased':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2<0)\n",
    "        elif regulation=='Decreased':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2>0)\n",
    "        tempS1 = tempS3 | tempS1\n",
    "    else:\n",
    "        tempS1 = (tempS1>=0.0)\n",
    "    #Update True\n",
    "    tempS = tempS & tempS1\n",
    "tempL = tempS.loc[tempS.tolist()].index.tolist()\n",
    "print(len(tempL), regulation.lower()+' reactions with significance in', posL, 'but not in', negL)\n",
    "\n",
    "#Select representatives\n",
    "topX = np.min([30, len(tempL)])\n",
    "topX_plot = np.min([5, len(tempL)])\n",
    "tempDF = statDF2.loc[:, statDF2.columns.str.contains('_AdjPval$')]\n",
    "tempDF = pd.merge(statDF[['ReactionName', variable+'_Pval', variable+'_AdjPval']], tempDF,\n",
    "                  left_index=True, right_index=True, how='left')\n",
    "tempDF = tempDF.loc[tempL].sort_values(by=variable+'_AdjPval', ascending=True)\n",
    "print('Top', topX, 'reactions (sort by the main effect of '+variable+'):')\n",
    "display(tempDF.iloc[:topX])\n",
    "plotL = tempDF.index.tolist()[:topX_plot]\n",
    "\n",
    "#Prepare DF for plot\n",
    "tempDF = fluxDF.reset_index().melt(var_name='ModelID', value_name='Flux', id_vars='ReactionID')\n",
    "tempDF1 = sampleDF.reset_index()[['ModelID', 'Intervention']]\n",
    "tempDF = pd.merge(tempDF, tempDF1, on='ModelID', how='left')\n",
    "\n",
    "#Prepare label and color\n",
    "tempD0 = {'Con1':'Control 1', 'Acar':'Acarbose',\n",
    "          '17aE':'17'+r'$\\alpha$'+'-Estradiol', 'Prot':'Protandim',\n",
    "          'Rapa':'Rapamycin', 'CalR':'CR diet',\n",
    "          'Con2':'Control 2', 'MetR':'MR diet',\n",
    "          'GHRw':'GHR WT', 'GHRk':'GHR KO',\n",
    "          'SneW':'Snell WT', 'SneD':'Snell dwarf'}\n",
    "tempDF['Group'] = tempDF['Intervention'].map(tempD0)\n",
    "tempD1 = {'Control 1':'tab:blue', 'CR diet':'tab:olive',\n",
    "          'Control 2':'tab:blue', 'MR diet':'tab:orange'}\n",
    "\n",
    "#Select only the target groups\n",
    "tempDF = tempDF.loc[tempDF['Group'].isin(tempD1.keys())]\n",
    "\n",
    "#Visualize each representative\n",
    "for rank_i in range(len(plotL)):\n",
    "    print(' - Rank '+str(rank_i+1)+' (sort by the main effect of '+variable+'):')\n",
    "    rxn = plotL[rank_i]\n",
    "    #Check reaction summary\n",
    "    tempDF1 = pd.DataFrame(rxnDF.loc[rxn]).T\n",
    "    display(tempDF1)\n",
    "    \n",
    "    #Select Flux\n",
    "    tempDF1 = tempDF.loc[tempDF['ReactionID']==rxn]\n",
    "    \n",
    "    #Check Flux summary\n",
    "    tempDF2 = tempDF1.groupby(['Group'])['Flux'].agg(['count', 'mean', 'std'])\n",
    "    tempL1 = []\n",
    "    tempL2 = []\n",
    "    for row_n in tempDF2.index.tolist():\n",
    "        count, mean, std = tempDF2.loc[row_n]\n",
    "        tempL1.append(mean - 1.96*std/np.sqrt(count))\n",
    "        tempL2.append(mean + 1.96*std/np.sqrt(count))\n",
    "    tempDF2['0.025'] = tempL1\n",
    "    tempDF2['0.975'] = tempL2\n",
    "    tempDF2 = tempDF2.loc[list(tempD1.keys())]#Sort\n",
    "    display(tempDF2)\n",
    "    \n",
    "    #Prepare significance labels\n",
    "    ##Retrieve statistical significance\n",
    "    tempS = pvalDF2.loc[rxn, pvalDF2.columns.str.contains('-vs-')]\n",
    "    tempS.name = 'AdjPval'\n",
    "    ##Clean\n",
    "    tempDF2 = tempS.index.to_series().str.split(pat='-vs-', expand=True)\n",
    "    tempDF2 = tempDF2.rename(columns={0:'Contrast', 1:'Baseline'})\n",
    "    tempDF2 = pd.merge(tempDF2, tempS, left_index=True, right_index=True, how='left')\n",
    "    tempDF2['Contrast'] = tempDF2['Contrast'].map(tempD0)\n",
    "    tempDF2['Baseline'] = tempDF2['Baseline'].map(tempD0)\n",
    "    ##Convert p-value to label\n",
    "    tempL = []\n",
    "    for row_i in range(len(tempDF2)):\n",
    "        pval = tempDF2['AdjPval'].iloc[row_i]\n",
    "        if pval<0.001:\n",
    "            tempL.append('***')\n",
    "        elif pval<0.01:\n",
    "            tempL.append('**')\n",
    "        elif pval<0.05:\n",
    "            tempL.append('*')\n",
    "        else:\n",
    "            pval_text = Decimal(str(pval)).quantize(Decimal('0.01'), rounding=ROUND_HALF_UP)\n",
    "            tempL.append(r'$P$ = '+str(pval_text))\n",
    "    tempDF2['SignifLabel'] = tempL\n",
    "    ##Add the y-position level in figure\n",
    "    tempDF2['YposLevel'] = [0, 0, 1]\n",
    "    display(tempDF2)\n",
    "    \n",
    "    #Visualization\n",
    "    sns.set(style='ticks', font='Arial', context='talk')\n",
    "    plt.figure(figsize=(2, 4))\n",
    "    sns.boxplot(data=tempDF1, x='Group', y='Flux', order=list(tempD1.keys()), palette=tempD1,\n",
    "                dodge=False, showfliers=False, showcaps=True, notch=False)\n",
    "    p = sns.stripplot(data=tempDF1, x='Group', y='Flux',\n",
    "                      order=list(tempD1.keys()), palette=tempD1, dodge=False, jitter=0.3,\n",
    "                      size=5, edgecolor='black', linewidth=1, **{'marker':'o', 'alpha':0.5})\n",
    "    ##Add border line\n",
    "    for xcoord in [1.5]:\n",
    "        p.axvline(x=xcoord, **{'linestyle':'dotted', 'color':'black', 'zorder':0})\n",
    "    ##Set axis\n",
    "    sns.despine()\n",
    "    plt.setp(p.get_xticklabels(), rotation=70, horizontalalignment='right',\n",
    "             verticalalignment='center', rotation_mode='anchor')\n",
    "    ###Reset the axis range for the labels\n",
    "    ymin, ymax = p.get_ylim()\n",
    "    tempL = p.get_yticks().tolist()\n",
    "    yinter = int(tempL[1]) - int(tempL[0])\n",
    "    tempL = [int(y) for y in tempL if (y>ymin)&(y<ymax)]#seaborn seems to prepare wider range!?\n",
    "    p.set(ylim=(ymin, ymax+(yinter*0.8)), yticks=tempL)\n",
    "    p.set_yticklabels(['{:,}'.format(int(y)) for y in tempL])\n",
    "    ##Add significance labels\n",
    "    aline_ymin = tempDF1['Flux'].max()\n",
    "    aline_ymargin = yinter/2\n",
    "    for row_i in range(len(tempDF2)):\n",
    "        #Baseline\n",
    "        group_0 = tempDF2['Baseline'].iloc[row_i]\n",
    "        index_0 = list(tempD1.keys()).index(group_0)\n",
    "        xcoord_0 = index_0\n",
    "        #Contrast\n",
    "        group_1 = tempDF2['Contrast'].iloc[row_i]\n",
    "        index_1 = list(tempD1.keys()).index(group_1)\n",
    "        xcoord_1 = index_1\n",
    "        #Standard point of marker\n",
    "        xcoord = (xcoord_0+xcoord_1)/2\n",
    "        ycoord = aline_ymin + aline_ymargin*tempDF2['YposLevel'].iloc[row_i]\n",
    "        label = tempDF2['SignifLabel'].iloc[row_i]\n",
    "        #Add annotation lines\n",
    "        aline_offset = yinter/10\n",
    "        aline_length = yinter/10 + aline_offset\n",
    "        plt.plot([xcoord_0, xcoord_0, xcoord_1, xcoord_1],\n",
    "                 [ycoord+aline_offset, ycoord+aline_length, ycoord+aline_length, ycoord+aline_offset],\n",
    "                 lw=1.5, c='k')\n",
    "        #Add annotation text\n",
    "        if label in ['***', '**', '*']:\n",
    "            text_offset = yinter/4\n",
    "            p.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                       horizontalalignment='center', verticalalignment='center',\n",
    "                       fontsize='medium', color='k')\n",
    "        else:\n",
    "            text_offset = yinter/5\n",
    "            p.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                       horizontalalignment='center', verticalalignment='bottom',\n",
    "                       fontsize='x-small', color='k')\n",
    "    ##Set axis label and title\n",
    "    plt.setp(p, xlabel='', ylabel='Sample flux value [a.u.]')\n",
    "    p.set_title(rxn, {'fontsize':'medium'})\n",
    "    ##Save\n",
    "    fileDir = './ExportFigures/'\n",
    "    ipynbName = '220610_LC-M001-related-transcriptomics-GEM-ver2-4_FluxAnalysis_'\n",
    "    fileName = 'MR-vs-CR-boxplot('+rxn+').tif'\n",
    "    plt.gcf().savefig(fileDir+ipynbName+fileName, dpi=300, bbox_inches='tight', pad_inches=0.04,\n",
    "                      pil_kwargs={'compression':'tiff_lzw'})\n",
    "    plt.show()\n",
    "    print('')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 4-3-2. Reactions decreased by MR but not CR"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Prepare the target reaction set\n",
    "posL = ['MetR-vs-Con2', 'MetR-vs-CalR']\n",
    "negL = ['CalR-vs-Con1']\n",
    "regulation = 'Decreased'\n",
    "tempL = statDF.loc[statDF[variable+'_AdjPval']<0.05].index.tolist()\n",
    "tempDF1 = pvalDF2.loc[tempL, pvalDF2.columns.str.contains('-vs-')]\n",
    "tempDF2 = diffDF2.loc[tempL, diffDF2.columns.str.contains('-vs-')]\n",
    "tempS = pd.Series(np.repeat(True, len(tempDF1)), index=tempDF1.index)#Initialize\n",
    "for col_n in tempDF1.columns.tolist():\n",
    "    tempS1 = tempDF1[col_n]\n",
    "    tempS2 = tempDF2[col_n]\n",
    "    if col_n in posL:\n",
    "        if regulation=='Increased':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2>0)\n",
    "        elif regulation=='Decreased':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2<0)\n",
    "    elif col_n in negL:\n",
    "        tempS3 = (tempS1>=0.05)\n",
    "        #Significance for inverse regulation\n",
    "        if regulation=='Increased':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2<0)\n",
    "        elif regulation=='Decreased':\n",
    "            tempS1 = (tempS1<0.05) & (tempS2>0)\n",
    "        tempS1 = tempS3 | tempS1\n",
    "    else:\n",
    "        tempS1 = (tempS1>=0.0)\n",
    "    #Update True\n",
    "    tempS = tempS & tempS1\n",
    "tempL = tempS.loc[tempS.tolist()].index.tolist()\n",
    "print(len(tempL), regulation.lower()+' reactions with significance in', posL, 'but not in', negL)\n",
    "\n",
    "#Select representatives\n",
    "topX = np.min([30, len(tempL)])\n",
    "topX_plot = np.min([5, len(tempL)])\n",
    "tempDF = statDF2.loc[:, statDF2.columns.str.contains('_AdjPval$')]\n",
    "tempDF = pd.merge(statDF[['ReactionName', variable+'_Pval', variable+'_AdjPval']], tempDF,\n",
    "                  left_index=True, right_index=True, how='left')\n",
    "tempDF = tempDF.loc[tempL].sort_values(by=variable+'_AdjPval', ascending=True)\n",
    "print('Top', topX, 'reactions (sort by the main effect of '+variable+'):')\n",
    "display(tempDF.iloc[:topX])\n",
    "plotL = tempDF.index.tolist()[:topX_plot]\n",
    "\n",
    "#Prepare DF for plot\n",
    "tempDF = fluxDF.reset_index().melt(var_name='ModelID', value_name='Flux', id_vars='ReactionID')\n",
    "tempDF1 = sampleDF.reset_index()[['ModelID', 'Intervention']]\n",
    "tempDF = pd.merge(tempDF, tempDF1, on='ModelID', how='left')\n",
    "\n",
    "#Prepare label and color\n",
    "tempD0 = {'Con1':'Control 1', 'Acar':'Acarbose',\n",
    "          '17aE':'17'+r'$\\alpha$'+'-Estradiol', 'Prot':'Protandim',\n",
    "          'Rapa':'Rapamycin', 'CalR':'CR diet',\n",
    "          'Con2':'Control 2', 'MetR':'MR diet',\n",
    "          'GHRw':'GHR WT', 'GHRk':'GHR KO',\n",
    "          'SneW':'Snell WT', 'SneD':'Snell dwarf'}\n",
    "tempDF['Group'] = tempDF['Intervention'].map(tempD0)\n",
    "tempD1 = {'Control 1':'tab:blue', 'CR diet':'tab:olive',\n",
    "          'Control 2':'tab:blue', 'MR diet':'tab:orange'}\n",
    "\n",
    "#Select only the target groups\n",
    "tempDF = tempDF.loc[tempDF['Group'].isin(tempD1.keys())]\n",
    "\n",
    "#Visualize each representative\n",
    "for rank_i in range(len(plotL)):\n",
    "    print(' - Rank '+str(rank_i+1)+' (sort by the main effect of '+variable+'):')\n",
    "    rxn = plotL[rank_i]\n",
    "    #Check reaction summary\n",
    "    tempDF1 = pd.DataFrame(rxnDF.loc[rxn]).T\n",
    "    display(tempDF1)\n",
    "    \n",
    "    #Select Flux\n",
    "    tempDF1 = tempDF.loc[tempDF['ReactionID']==rxn]\n",
    "    \n",
    "    #Check Flux summary\n",
    "    tempDF2 = tempDF1.groupby(['Group'])['Flux'].agg(['count', 'mean', 'std'])\n",
    "    tempL1 = []\n",
    "    tempL2 = []\n",
    "    for row_n in tempDF2.index.tolist():\n",
    "        count, mean, std = tempDF2.loc[row_n]\n",
    "        tempL1.append(mean - 1.96*std/np.sqrt(count))\n",
    "        tempL2.append(mean + 1.96*std/np.sqrt(count))\n",
    "    tempDF2['0.025'] = tempL1\n",
    "    tempDF2['0.975'] = tempL2\n",
    "    tempDF2 = tempDF2.loc[list(tempD1.keys())]#Sort\n",
    "    display(tempDF2)\n",
    "    \n",
    "    #Prepare significance labels\n",
    "    ##Retrieve statistical significance\n",
    "    tempS = pvalDF2.loc[rxn, pvalDF2.columns.str.contains('-vs-')]\n",
    "    tempS.name = 'AdjPval'\n",
    "    ##Clean\n",
    "    tempDF2 = tempS.index.to_series().str.split(pat='-vs-', expand=True)\n",
    "    tempDF2 = tempDF2.rename(columns={0:'Contrast', 1:'Baseline'})\n",
    "    tempDF2 = pd.merge(tempDF2, tempS, left_index=True, right_index=True, how='left')\n",
    "    tempDF2['Contrast'] = tempDF2['Contrast'].map(tempD0)\n",
    "    tempDF2['Baseline'] = tempDF2['Baseline'].map(tempD0)\n",
    "    ##Convert p-value to label\n",
    "    tempL = []\n",
    "    for row_i in range(len(tempDF2)):\n",
    "        pval = tempDF2['AdjPval'].iloc[row_i]\n",
    "        if pval<0.001:\n",
    "            tempL.append('***')\n",
    "        elif pval<0.01:\n",
    "            tempL.append('**')\n",
    "        elif pval<0.05:\n",
    "            tempL.append('*')\n",
    "        else:\n",
    "            pval_text = Decimal(str(pval)).quantize(Decimal('0.01'), rounding=ROUND_HALF_UP)\n",
    "            tempL.append(r'$P$ = '+str(pval_text))\n",
    "    tempDF2['SignifLabel'] = tempL\n",
    "    ##Add the y-position level in figure\n",
    "    tempDF2['YposLevel'] = [0, 0, 1]\n",
    "    display(tempDF2)\n",
    "    \n",
    "    #Visualization\n",
    "    sns.set(style='ticks', font='Arial', context='talk')\n",
    "    plt.figure(figsize=(2, 4))\n",
    "    sns.boxplot(data=tempDF1, x='Group', y='Flux', order=list(tempD1.keys()), palette=tempD1,\n",
    "                dodge=False, showfliers=False, showcaps=True, notch=False)\n",
    "    p = sns.stripplot(data=tempDF1, x='Group', y='Flux',\n",
    "                      order=list(tempD1.keys()), palette=tempD1, dodge=False, jitter=0.3,\n",
    "                      size=5, edgecolor='black', linewidth=1, **{'marker':'o', 'alpha':0.5})\n",
    "    ##Add border line\n",
    "    for xcoord in [1.5]:\n",
    "        p.axvline(x=xcoord, **{'linestyle':'dotted', 'color':'black', 'zorder':0})\n",
    "    ##Set axis\n",
    "    sns.despine()\n",
    "    plt.setp(p.get_xticklabels(), rotation=70, horizontalalignment='right',\n",
    "             verticalalignment='center', rotation_mode='anchor')\n",
    "    ###Reset the axis range for the labels\n",
    "    ymin, ymax = p.get_ylim()\n",
    "    tempL = p.get_yticks().tolist()\n",
    "    yinter = int(tempL[1]) - int(tempL[0])\n",
    "    tempL = [int(y) for y in tempL if (y>ymin)&(y<ymax)]#seaborn seems to prepare wider range!?\n",
    "    p.set(ylim=(ymin, ymax+(yinter*0.8)), yticks=tempL)\n",
    "    p.set_yticklabels(['{:,}'.format(int(y)) for y in tempL])\n",
    "    ##Add significance labels\n",
    "    aline_ymin = tempDF1['Flux'].max()\n",
    "    aline_ymargin = yinter/2\n",
    "    for row_i in range(len(tempDF2)):\n",
    "        #Baseline\n",
    "        group_0 = tempDF2['Baseline'].iloc[row_i]\n",
    "        index_0 = list(tempD1.keys()).index(group_0)\n",
    "        xcoord_0 = index_0\n",
    "        #Contrast\n",
    "        group_1 = tempDF2['Contrast'].iloc[row_i]\n",
    "        index_1 = list(tempD1.keys()).index(group_1)\n",
    "        xcoord_1 = index_1\n",
    "        #Standard point of marker\n",
    "        xcoord = (xcoord_0+xcoord_1)/2\n",
    "        ycoord = aline_ymin + aline_ymargin*tempDF2['YposLevel'].iloc[row_i]\n",
    "        label = tempDF2['SignifLabel'].iloc[row_i]\n",
    "        #Add annotation lines\n",
    "        aline_offset = yinter/10\n",
    "        aline_length = yinter/10 + aline_offset\n",
    "        plt.plot([xcoord_0, xcoord_0, xcoord_1, xcoord_1],\n",
    "                 [ycoord+aline_offset, ycoord+aline_length, ycoord+aline_length, ycoord+aline_offset],\n",
    "                 lw=1.5, c='k')\n",
    "        #Add annotation text\n",
    "        if label in ['***', '**', '*']:\n",
    "            text_offset = yinter/4\n",
    "            p.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                       horizontalalignment='center', verticalalignment='center',\n",
    "                       fontsize='medium', color='k')\n",
    "        else:\n",
    "            text_offset = yinter/5\n",
    "            p.annotate(label, xy=(xcoord, ycoord+text_offset),\n",
    "                       horizontalalignment='center', verticalalignment='bottom',\n",
    "                       fontsize='x-small', color='k')\n",
    "    ##Set axis label and title\n",
    "    plt.setp(p, xlabel='', ylabel='Sample flux value [a.u.]')\n",
    "    p.set_title(rxn, {'fontsize':'medium'})\n",
    "    ##Save\n",
    "    fileDir = './ExportFigures/'\n",
    "    ipynbName = '220610_LC-M001-related-transcriptomics-GEM-ver2-4_FluxAnalysis_'\n",
    "    fileName = 'MR-vs-CR-boxplot('+rxn+').tif'\n",
    "    plt.gcf().savefig(fileDir+ipynbName+fileName, dpi=300, bbox_inches='tight', pad_inches=0.04,\n",
    "                      pil_kwargs={'compression':'tiff_lzw'})\n",
    "    plt.show()\n",
    "    print('')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# — End of notebook —"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "arivale-py3 - Python",
   "language": "python",
   "name": "conda-env-arivale-py3-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  },
  "widgets": {
   "application/vnd.jupyter.widget-state+json": {
    "state": {},
    "version_major": 2,
    "version_minor": 0
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
